{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# allele_specific_analyses(10)\n",
    "\n",
    "1/27/2021\n",
    "\n",
    "run gatk (see one note for details)\n",
    "\n",
    "now see if overlap with mpra results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import os,sys, glob\n",
    "import pybedtools\n",
    "import scipy.stats as stats\n",
    "import re\n",
    "import numpy as np\n",
    "from statsmodels.stats.multitest import fdrcorrection\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_vcf(path):\n",
    "    with open(path, 'r') as f:\n",
    "        lines = f.readlines()\n",
    "        lines = [l.strip().split('\\t') for l in lines if not l.startswith('##')]\n",
    "    df=  pd.DataFrame(lines[1: ],columns=lines[0]).rename(columns={'#CHROM': 'CHROM'})\n",
    "    if not df['CHROM'][0].startswith('chr'):\n",
    "        df['CHROM'] = 'chr' + df.CHROM.map(str)\n",
    "\n",
    "    df['rowname'] =  df.CHROM + '_' + df.POS.map(str)\n",
    "    print('read',path)\n",
    "    print('shape',df.shape)\n",
    "    return df\n",
    "\n",
    "def preprocess_vcf(row,tissue):\n",
    "    # get info quality metrics\n",
    "    info_str = row['INFO']\n",
    "    info_str_arr = re.split('[=;]',info_str)\n",
    "    qual =  pd.Series(dict(zip(info_str_arr[::2],info_str_arr[1::2] )))\n",
    "    qual = qual[['DP','SOR','FS']]\n",
    "    # get \n",
    "    tissue_cols = [x for x in row.index if re.search('B[1-3]$',x) is not None]    #x.startswith(tissue[:min(len(tissue),4)])]\n",
    "    if len(tissue_cols)==0:\n",
    "        tissue_cols=['sample','sample']\n",
    "#     print('tissue_cols',tissue_cols)\n",
    "    try:\n",
    "        sample_1_info = row[tissue_cols[0]].split(':')\n",
    "        sample_2_info = row[tissue_cols[1]].split(':')\n",
    "\n",
    "        qual['GT_1'] = sample_1_info[0]\n",
    "        qual['GT_2'] = sample_2_info[0]\n",
    "        qual['AD_1_ref'] = int(sample_1_info[1].split(',')[0])\n",
    "        qual['AD_1_alt'] = int(sample_1_info[1].split(',')[1])\n",
    "        qual['AD_2_ref'] = int(sample_2_info[1].split(',')[0])\n",
    "        qual['AD_2_alt'] = int(sample_2_info[1].split(',')[1])\n",
    "    except:\n",
    "        print(row)\n",
    "        print(tissue_cols)\n",
    "        raise\n",
    "    return qual\n",
    "#     format_arr = row['FORMAT'].split(':')\n",
    "#     \n",
    "#     format_series = pd.Series()\n",
    "#     for idx,sample in enumerate(tissue_cols):\n",
    "#         format_arr_sample = [x+\"_\"+str(idx)for x in format_arr]\n",
    "#         format_series=pd.concat([format_series,pd.Series(dict(zip(format_arr_sample,row[sample].split(':'))))])\n",
    "#     return pd.concat([qual,format_series])\n",
    "\n",
    "def filter_vcf(vcf_df):\n",
    "    # filter depth count (DP)>=10\n",
    "    vcf_df = vcf_df[vcf_df.DP.map(int)>=10]\n",
    "    # the allele  only biallelic SNP sites (GT: 0/1) w\n",
    "    vcf_df = vcf_df[(vcf_df.GT_1 =='0/1')& (vcf_df.GT_2 =='0/1')]\n",
    "    \n",
    "    # minimum reference or alternative allele count >=2 (AD)\n",
    "    vcf_df['AD_1_min'] = vcf_df[['AD_1_ref','AD_1_alt']].min(axis=1)\n",
    "    vcf_df = vcf_df[vcf_df.AD_1_min>=2]\n",
    "    vcf_df['AD_2_min'] = vcf_df[['AD_2_ref','AD_2_alt']].min(axis=1)\n",
    "    vcf_df = vcf_df[vcf_df.AD_2_min>=2]\n",
    "    return vcf_df\n",
    "\n",
    "def preprocess_and_filter(row,tissue):\n",
    "    # for hichip, do preproces and filter flag at same time\n",
    "    binom_p = 1 # pass file is the probability of passing the binomial test\n",
    "    # if they don't pass criterion give. a-1 for ease of filtering out \n",
    "\n",
    "    # get info quality metrics\n",
    "    info_str = row['INFO']\n",
    "    info_str_arr = re.split('[=;]',info_str)\n",
    "    qual_dict =  dict(zip(info_str_arr[::2],info_str_arr[1::2] ))\n",
    "#     qual = qual[['DP','SOR','FS']]\n",
    "    # get tissue cols\n",
    "    tissue_cols = [x for x in row.index if re.search('B[1-3]$',x) is not None]    #x.startswith(tissue[:min(len(tissue),4)])]\n",
    "    if len(tissue_cols)==0:\n",
    "        tissue_cols=['sample','sample']\n",
    "#     print('tissue_cols',tissue_cols)\n",
    "\n",
    "    try:\n",
    "        # filter depth count (DP)>=10   \n",
    "        if( int(qual_dict['DP'])<10):\n",
    "            return -1\n",
    "#         print('qual_dict dp')\n",
    "        # the allele  only biallelic SNP sites (GT: 0/1) w        \n",
    "        sample_1_info = row[tissue_cols[0]].split(':')\n",
    "        sample_2_info = row[tissue_cols[1]].split(':')\n",
    "        if ((sample_1_info[0]!='0/1')| (sample_2_info[0]!='0/1')):\n",
    "            return -1\n",
    "#         print('GT')\n",
    "        \n",
    "        # minimum reference or alternative allele count >=2 (AD)\n",
    "        AD_arr = [  int(sample_1_info[1].split(',')[0]), #AD_1_ref\n",
    "                    int(sample_1_info[1].split(',')[1]), #AD_1_alt\n",
    "                    int(sample_2_info[1].split(',')[0]), #AD_2_ref\n",
    "                    int(sample_2_info[1].split(',')[1])] #AD_2_alt\n",
    "        binom_p = stats.binom_test(x=AD_arr[1] , n=sum(AD_arr)/2, p=0.5, alternative='two-sided')\n",
    "        if min(AD_arr)<2:\n",
    "            return -1\n",
    "    except:\n",
    "        binom_p = -1\n",
    "        print('error, warning****, skipped..')\n",
    "        print(row)\n",
    "        print(tissue_cols)\n",
    "    return binom_p    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_dir = '../data/processed/vcf_files/analyses'\n",
    "if not os.path.exists(save_dir):\n",
    "    os.makedirs(save_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "lib_df_bed_df (35203, 4) lib_df_bed_df1 (5016, 4)\n",
      "lib_df_bed_df shape post merge (38031, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>rowname</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>44171211</td>\n",
       "      <td>44171212</td>\n",
       "      <td>chr1_44171211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>8086527</td>\n",
       "      <td>8086528</td>\n",
       "      <td>chr1_8086527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>108330356</td>\n",
       "      <td>108330357</td>\n",
       "      <td>chr1_108330356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>chr5</td>\n",
       "      <td>44876507</td>\n",
       "      <td>44876508</td>\n",
       "      <td>chr5_44876507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>chr5</td>\n",
       "      <td>50735307</td>\n",
       "      <td>50735308</td>\n",
       "      <td>chr5_50735307</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr      start       stop         rowname\n",
       "0  chr1   44171211   44171212   chr1_44171211\n",
       "1  chr1    8086527    8086528    chr1_8086527\n",
       "2  chr1  108330356  108330357  chr1_108330356\n",
       "3  chr5   44876507   44876508   chr5_44876507\n",
       "4  chr5   50735307   50735308   chr5_50735307"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# option1\n",
    "lib_df1  = pd.read_csv('../data/external/GWAS/cancer_mpra_lib_snps_bedformat_121520.bed',sep='\\t',header=None)\n",
    "lib_df1.columns = ['chr','start','stop','rsid']\n",
    "lib_df1['rowname'] = lib_df1.chr + '_' + lib_df1['start'].map(str)\n",
    "lib_df_bed_df1 = lib_df1[['chr','start','stop','rowname']]\n",
    "# option2\n",
    "lib_df  = pd.read_csv('../data/external/GWAS/snp_list_combined.csv',index_col=0)\n",
    "lib_df.columns = ['rsid','chr','start','disease']\n",
    "lib_df['stop'] = lib_df['start'] + 1 \n",
    "lib_df['rowname'] = lib_df.chr + '_' + lib_df['start'].map(str)\n",
    "lib_df_bed_df = lib_df[['chr','start','stop','rowname']]\n",
    "print('lib_df_bed_df',lib_df_bed_df.shape,'lib_df_bed_df1', lib_df_bed_df1.shape)\n",
    "lib_df_bed_df = pd.concat([lib_df_bed_df,lib_df_bed_df1]).drop_duplicates()\n",
    "print('lib_df_bed_df shape post merge',lib_df_bed_df.shape)\n",
    "display(lib_df_bed_df[:5])\n",
    "#common processing\n",
    "lib_df_bed = pybedtools.BedTool.from_dataframe(lib_df_bed_df)\n",
    "# print(lib_df.disease.unique())\n",
    "# lib_df[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Causal_SNP</th>\n",
       "      <th>rowname</th>\n",
       "      <th>disease</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>locus</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>locus_1000</td>\n",
       "      <td>rs4866783</td>\n",
       "      <td>chr5_44876507</td>\n",
       "      <td>hmec</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>locus_1001</td>\n",
       "      <td>rs75282042</td>\n",
       "      <td>chr5_50735307</td>\n",
       "      <td>hmec</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>locus_1002</td>\n",
       "      <td>rs115010969</td>\n",
       "      <td>chr5_52535211</td>\n",
       "      <td>hmec</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>locus_1003</td>\n",
       "      <td>rs17343328</td>\n",
       "      <td>chr5_52615305</td>\n",
       "      <td>hmec</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>locus_1004</td>\n",
       "      <td>rs10940312</td>\n",
       "      <td>chr5_52667982</td>\n",
       "      <td>hmec</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>locus_996</td>\n",
       "      <td>rs16901937</td>\n",
       "      <td>chr5_44709141</td>\n",
       "      <td>hmec</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>locus_997</td>\n",
       "      <td>rs1866406</td>\n",
       "      <td>chr5_44809945</td>\n",
       "      <td>hmec</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>locus_999</td>\n",
       "      <td>rs10070037</td>\n",
       "      <td>chr5_44870237</td>\n",
       "      <td>hmec</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>locus_99</td>\n",
       "      <td>rs4915073</td>\n",
       "      <td>chr1_108325879</td>\n",
       "      <td>thy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>locus_9</td>\n",
       "      <td>rs408320</td>\n",
       "      <td>chr1_8085328</td>\n",
       "      <td>hmec</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4853 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Causal_SNP         rowname disease\n",
       "locus                                          \n",
       "locus_1000    rs4866783   chr5_44876507    hmec\n",
       "locus_1001   rs75282042   chr5_50735307    hmec\n",
       "locus_1002  rs115010969   chr5_52535211    hmec\n",
       "locus_1003   rs17343328   chr5_52615305    hmec\n",
       "locus_1004   rs10940312   chr5_52667982    hmec\n",
       "...                 ...             ...     ...\n",
       "locus_996    rs16901937   chr5_44709141    hmec\n",
       "locus_997     rs1866406   chr5_44809945    hmec\n",
       "locus_999    rs10070037   chr5_44870237    hmec\n",
       "locus_99      rs4915073  chr1_108325879     thy\n",
       "locus_9        rs408320    chr1_8085328    hmec\n",
       "\n",
       "[4853 rows x 3 columns]"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mpra_info_df = pd.read_table('../data/external/GWAS/cancer_mpra_snps_results_120320.tsv',index_col=0)\n",
    "mpra_info_df['rowname'] = 'chr'+mpra_info_df.Chr_37.map(str) + '_' +mpra_info_df.Start_37.map(str)\n",
    "mpra_info_df[['Causal_SNP', 'rowname','disease']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr_37</th>\n",
       "      <th>Start_37</th>\n",
       "      <th>disease</th>\n",
       "      <th>hit_tissue</th>\n",
       "      <th>rowname</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Causal_SNP</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>rs111640187</td>\n",
       "      <td>5</td>\n",
       "      <td>82428995</td>\n",
       "      <td>airway</td>\n",
       "      <td>airway</td>\n",
       "      <td>chr5_82428995</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>rs34064842</td>\n",
       "      <td>6</td>\n",
       "      <td>27688625</td>\n",
       "      <td>airway,hmec</td>\n",
       "      <td>airway</td>\n",
       "      <td>chr6_27688625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>rs71546575</td>\n",
       "      <td>6</td>\n",
       "      <td>27805022</td>\n",
       "      <td>airway,hmec</td>\n",
       "      <td>airway</td>\n",
       "      <td>chr6_27805022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>rs35353359</td>\n",
       "      <td>6</td>\n",
       "      <td>28324378</td>\n",
       "      <td>airway,hmec</td>\n",
       "      <td>airway</td>\n",
       "      <td>chr6_28324378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>rs3095311</td>\n",
       "      <td>6</td>\n",
       "      <td>31051675</td>\n",
       "      <td>airway</td>\n",
       "      <td>airway</td>\n",
       "      <td>chr6_31051675</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             Chr_37  Start_37      disease hit_tissue        rowname\n",
       "Causal_SNP                                                          \n",
       "rs111640187       5  82428995       airway     airway  chr5_82428995\n",
       "rs34064842        6  27688625  airway,hmec     airway  chr6_27688625\n",
       "rs71546575        6  27805022  airway,hmec     airway  chr6_27805022\n",
       "rs35353359        6  28324378  airway,hmec     airway  chr6_28324378\n",
       "rs3095311         6  31051675       airway     airway  chr6_31051675"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mpra_res_df = pd.read_table('../data/external/GWAS/cancer_mpra_significant_hits_coords_1.27.21.tsv',index_col=0)\n",
    "mpra_res_df['rowname'] = 'chr'+mpra_res_df.Chr_37.map(str) + '_' +mpra_res_df.Start_37.map(str)\n",
    "# mpra_res_df = mpra_res_df.merge(lib_df[['Linked_SNP','rowname']],how='left',on='rowname' )\n",
    "mpra_res_df[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(451, 5)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "HMEC           147\n",
       "Airway          68\n",
       "Prostate        51\n",
       "Colon           31\n",
       "Ovarian         24\n",
       "Melanocytes     23\n",
       "Uterine         22\n",
       "GDSD6           19\n",
       "Pancreas        17\n",
       "GM12878         16\n",
       "Thyroid         16\n",
       "Astrocytes      10\n",
       "Esophageal       4\n",
       "Renal            3\n",
       "Name: mapped_tissue, dtype: int64"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(mpra_res_df.shape)\n",
    "mpra_res_df.hit_tissue.unique()\n",
    "# # from asATAC to tissue\n",
    "mpra_tissue_mapper = {\n",
    "    'Airway':'airway',\n",
    "    'Astrocytes':'ast',\n",
    "    'Colon':'colon',\n",
    "    'Esophageal':'eso',\n",
    "    'GM12878':'gm',\n",
    "    'HMEC':'hmec',\n",
    "    'GDSD6':'kc',\n",
    "    'Melanocytes':'mc',\n",
    "    'Ovarian':'ov',\n",
    "    'Pancreas':'panc',\n",
    "    'Prostate':'pros',\n",
    "    'Renal':'renal',\n",
    "    'Thyroid':'thy',\n",
    "    'Uterine':'endo'\n",
    "}\n",
    "inv_mpra_tissue_mapper = {v: k for k, v in mpra_tissue_mapper.items()}\n",
    "mpra_res_df['mapped_tissue'] = mpra_res_df.hit_tissue.map(inv_mpra_tissue_mapper )\n",
    "mpra_res_df.mapped_tissue.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "# mpra_rownames_sig = list(mpra_res_df.rowname.unique())\n",
    "# len(mpra_rownames_sig)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "# num_mpra_sig_hits = mpra_res_df.rowname.unique().shape[0]\n",
    "num_lib_tested = lib_df.rowname.unique().shape[0]\n",
    "# print(num_mpra_sig_hits, num_lib_tested)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. atac"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['../data/interim/merged/atac/Airway_merged.bed',\n",
       " '../data/interim/merged/atac/Astrocytes_merged.bed',\n",
       " '../data/interim/merged/atac/Bladder_merged.bed',\n",
       " '../data/interim/merged/atac/Colon_merged.bed',\n",
       " '../data/interim/merged/atac/Esophageal_merged.bed',\n",
       " '../data/interim/merged/atac/GDSD0_merged.bed',\n",
       " '../data/interim/merged/atac/GDSD3_merged.bed',\n",
       " '../data/interim/merged/atac/GDSD6_merged.bed',\n",
       " '../data/interim/merged/atac/GM12878_merged.bed',\n",
       " '../data/interim/merged/atac/HMEC_merged.bed',\n",
       " '../data/interim/merged/atac/Melanocytes_merged.bed',\n",
       " '../data/interim/merged/atac/Ovarian_merged.bed',\n",
       " '../data/interim/merged/atac/Pancreas_merged.bed',\n",
       " '../data/interim/merged/atac/Prostate_merged.bed',\n",
       " '../data/interim/merged/atac/Renal_merged.bed',\n",
       " '../data/interim/merged/atac/Thyroid_merged.bed',\n",
       " '../data/interim/merged/atac/Uterine_merged.bed']"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "atac_dir = '../data/interim/merged/atac/'\n",
    "atac_bed_files = glob.glob(os.path.join(atac_dir, '*bed'))\n",
    "atac_bed_files_sel =sorted( [ '../data/interim/merged/atac/Renal_merged.bed',\n",
    " '../data/interim/merged/atac/Prostate_merged.bed',\n",
    " '../data/interim/merged/atac/Uterine_merged.bed',\n",
    " '../data/interim/merged/atac/GDSD6_merged.bed',\n",
    " '../data/interim/merged/atac/Pancreas_merged.bed',\n",
    " '../data/interim/merged/atac/Melanocytes_merged.bed',\n",
    " '../data/interim/merged/atac/Ovarian_merged.bed',\n",
    " '../data/interim/merged/atac/Bladder_merged.bed',\n",
    " '../data/interim/merged/atac/GDSD3_merged.bed',\n",
    " '../data/interim/merged/atac/Esophageal_merged.bed',\n",
    " '../data/interim/merged/atac/GM12878_merged.bed',\n",
    " '../data/interim/merged/atac/Astrocytes_merged.bed',\n",
    " '../data/interim/merged/atac/Colon_merged.bed',\n",
    " '../data/interim/merged/atac/GDSD0_merged.bed',\n",
    " '../data/interim/merged/atac/Thyroid_merged.bed',\n",
    " '../data/interim/merged/atac/Airway_merged.bed',\n",
    " '../data/interim/merged/atac/HMEC_merged.bed'])\n",
    "atac_bed_files_sel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #testing\n",
    "# atac_file = '/Users/mguo123/Documents/pan_omics_psych/data/interim/merged/atac/SL_D0_merged.bed'\n",
    "# tissue_mapper = {\n",
    "#     'H9_D2':'H9D2',\n",
    "#     'SLC_D0':'SLC',\n",
    "#     'SL_D0':'SL',\n",
    "#     'Astrocytes':'AST1',\n",
    "#     'H9_D10':'H9D10',\n",
    "#     'H9_D0':'H9D0'\n",
    "# }\n",
    "# tissue = os.path.basename(atac_file).split('_merged')[0]\n",
    "# atac_bed = pybedtools.BedTool(atac_file)\n",
    "# lib_df_bed_atac = lib_df_bed.intersect(atac_bed).to_dataframe()\n",
    "# lib_df_bed_atac  =lib_df_bed_atac[['name']].drop_duplicates().reset_index(drop=True)\n",
    "# lib_df_bed_atac['tissue'] = tissue_mapper[tissue]\n",
    "# lib_df_bed_atac.name[lib_df_bed_atac.name.str.startswith('chr3_503')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "# lib_atac_df.rowname[lib_atac_df.rowname.str.startswith('chr3_503')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Airway\n",
      "lib_atac_df (3067, 2)\n",
      "Astrocytes\n",
      "lib_atac_df (4414, 2)\n",
      "Bladder\n",
      "lib_atac_df (7692, 2)\n",
      "Colon\n",
      "lib_atac_df (11160, 2)\n",
      "Esophageal\n",
      "lib_atac_df (14751, 2)\n",
      "GDSD0\n",
      "lib_atac_df (17860, 2)\n",
      "GDSD3\n",
      "lib_atac_df (21137, 2)\n",
      "GDSD6\n",
      "lib_atac_df (24326, 2)\n",
      "GM12878\n",
      "lib_atac_df (27188, 2)\n",
      "HMEC\n",
      "lib_atac_df (30094, 2)\n",
      "Melanocytes\n",
      "lib_atac_df (31847, 2)\n",
      "Ovarian\n",
      "lib_atac_df (33669, 2)\n",
      "Pancreas\n",
      "lib_atac_df (37299, 2)\n",
      "Prostate\n",
      "lib_atac_df (39143, 2)\n",
      "Renal\n",
      "lib_atac_df (42699, 2)\n",
      "Thyroid\n",
      "lib_atac_df (46083, 2)\n",
      "Uterine\n",
      "lib_atac_df (50089, 2)\n"
     ]
    }
   ],
   "source": [
    "# aside get a table which lists for all the snps whether it's in an atac peak and which tissues\n",
    "\n",
    "lib_atac_df = pd.DataFrame()\n",
    "for atac_file in sorted(atac_bed_files_sel):\n",
    "    tissue = os.path.basename(atac_file).split('_merged')[0]\n",
    "    print(tissue)\n",
    "#     if tissue not in tissue_mapper.keys():\n",
    "#         print(tissue, 'not considered')\n",
    "#         continue\n",
    "    atac_bed = pybedtools.BedTool(atac_file)\n",
    "    lib_df_bed_atac = lib_df_bed.intersect(atac_bed).to_dataframe()\n",
    "    lib_df_bed_atac  =lib_df_bed_atac[['name']].drop_duplicates().reset_index(drop=True)\n",
    "    lib_df_bed_atac['tissue'] = tissue#tissue_mapper[tissue]\n",
    "    lib_atac_df = pd.concat([lib_atac_df, lib_df_bed_atac])\n",
    "    print('lib_atac_df', lib_atac_df.shape)\n",
    "    \n",
    "lib_atac_df = lib_atac_df.groupby('name').agg({'tissue':'|'.join}).reset_index()\n",
    "lib_atac_df.columns = ['rowname','atac_tissues']\n",
    "lib_atac_df['bool_in_atac_pk'] = True\n",
    "lib_atac_df.to_csv(os.path.join(save_dir,'lib_atac_annotation.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rowname</th>\n",
       "      <th>atac_tissues</th>\n",
       "      <th>bool_in_atac_pk</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>chr10_101860630</td>\n",
       "      <td>Uterine</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>chr10_101860631</td>\n",
       "      <td>Uterine</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>chr10_101946033</td>\n",
       "      <td>Airway|Astrocytes|Bladder|Colon|Esophageal|GDS...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>chr10_101989509</td>\n",
       "      <td>Airway|Astrocytes|Bladder|Colon|Esophageal|GDS...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>chr10_101989510</td>\n",
       "      <td>Airway|Astrocytes|Bladder|Colon|Esophageal|GDS...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           rowname                                       atac_tissues  \\\n",
       "0  chr10_101860630                                            Uterine   \n",
       "1  chr10_101860631                                            Uterine   \n",
       "2  chr10_101946033  Airway|Astrocytes|Bladder|Colon|Esophageal|GDS...   \n",
       "3  chr10_101989509  Airway|Astrocytes|Bladder|Colon|Esophageal|GDS...   \n",
       "4  chr10_101989510  Airway|Astrocytes|Bladder|Colon|Esophageal|GDS...   \n",
       "\n",
       "   bool_in_atac_pk  \n",
       "0             True  \n",
       "1             True  \n",
       "2             True  \n",
       "3             True  \n",
       "4             True  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "((7125, 3), (38031, 4))"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "display(lib_atac_df[:5])\n",
    "lib_atac_df.shape, lib_df_bed_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['../data/processed/vcf_files/atac/Airway_postvqsr.vcf',\n",
       " '../data/processed/vcf_files/atac/Astrocytes_postvqsr.vcf',\n",
       " '../data/processed/vcf_files/atac/Bladder_postvqsr.vcf',\n",
       " '../data/processed/vcf_files/atac/Colon_postvqsr.vcf',\n",
       " '../data/processed/vcf_files/atac/Esophageal_postvqsr.vcf',\n",
       " '../data/processed/vcf_files/atac/GDSD0_postvqsr.vcf',\n",
       " '../data/processed/vcf_files/atac/GDSD3_postvqsr.vcf',\n",
       " '../data/processed/vcf_files/atac/GDSD6_postvqsr.vcf',\n",
       " '../data/processed/vcf_files/atac/GM12878_postvqsr.vcf',\n",
       " '../data/processed/vcf_files/atac/HMEC_postvqsr.vcf',\n",
       " '../data/processed/vcf_files/atac/Melanocytes_postvqsr.vcf',\n",
       " '../data/processed/vcf_files/atac/Ovarian_postvqsr.vcf',\n",
       " '../data/processed/vcf_files/atac/Pancreas_postvqsr.vcf',\n",
       " '../data/processed/vcf_files/atac/Prostate_postvqsr.vcf',\n",
       " '../data/processed/vcf_files/atac/Renal_postvqsr.vcf',\n",
       " '../data/processed/vcf_files/atac/Thyroid_postvqsr.vcf',\n",
       " '../data/processed/vcf_files/atac/Uterine_postvqsr.vcf']"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcf_files = sorted(glob.glob('../data/processed/vcf_files/atac/*vcf'))\n",
    "vcf_files "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Airway\n",
      "Astrocytes\n",
      "Bladder\n",
      "Colon\n",
      "Esophageal\n",
      "GDSD0\n",
      "GDSD3\n",
      "GDSD6\n",
      "GM12878\n",
      "HMEC\n",
      "Melanocytes\n",
      "Ovarian\n",
      "Pancreas\n",
      "Prostate\n",
      "Renal\n",
      "Thyroid\n",
      "Uterine\n"
     ]
    }
   ],
   "source": [
    "tissues = []\n",
    "for vcf_file in vcf_files:\n",
    "    tissue = os.path.basename(vcf_file).split('_post')[0]\n",
    "    print(tissue)\n",
    "    tissues.append(tissue)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Airway': 'Airway',\n",
       " 'Astrocytes': 'Astrocytes',\n",
       " 'Bladder': 'Bladder',\n",
       " 'Colon': 'Colon',\n",
       " 'Esophageal': 'Esophageal',\n",
       " 'GDSD0': 'GDSD0',\n",
       " 'GDSD3': 'GDSD3',\n",
       " 'GDSD6': 'GDSD6',\n",
       " 'GM12878': 'GM12878',\n",
       " 'HMEC': 'HMEC',\n",
       " 'Melanocytes': 'Melanocytes',\n",
       " 'Ovarian': 'Ovarian',\n",
       " 'Pancreas': 'Pancreas',\n",
       " 'Prostate': 'Prostate',\n",
       " 'Renal': 'Renal',\n",
       " 'Thyroid': 'Thyroid',\n",
       " 'Uterine': 'Uterine'}"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "atac_tissue_mapper = dict(zip(tissues,tissues))\n",
    "atac_tissue_mapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [],
   "source": [
    "### FROM TESTING\n",
    "# vcf_df[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "hypergeometric test between two sets\n",
    "- phyper=(overlap-1,list1,PopSize-list1,list2,lower.tail = FALSE, log.p = FALSE)\n",
    "- scipy.stats.hypergeom.cdf(overlap, pop, list1, list2)\n",
    "\n",
    "- \n",
    "phyper=(88,598,23000-598,5500,lower.tail = FALSE, log.p = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [],
   "source": [
    "def hypergeometric_test(x, M, n, N):\n",
    "    \"\"\"\n",
    "    The hypergeometric distribution models drawing objects from a bin.\n",
    "    - M is total number of objects\n",
    "    - n is total number of Type I objects. \n",
    "    - x (random variate) represents the number of Type I objects in N drawn without replacement from the total population\n",
    "\n",
    "    - http://en.wikipedia.org/wiki/Hypergeometric_distribution\n",
    "    - https://www.biostars.org/p/66729/\n",
    "    - http://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.stats.hypergeom.html\n",
    "    - http://docs.scipy.org/doc/numpy/reference/generated/numpy.random.hypergeometric.html\n",
    "    - http://stackoverflow.com/questions/6594840/what-are-equivalents-to-rs-phyper-function-in-python\n",
    "    \"\"\"\n",
    "\n",
    "    assert n <= M\n",
    "    assert x <= n\n",
    "    assert N <= M\n",
    "#     pv_le = stats.hypergeom.cdf(x+1, M, n, N)\n",
    "    pv_gt = stats.hypergeom.sf(x-1, M, n, N)# 1-cdf sometimes more accurate\n",
    "    return pv_gt#pv_gt pv_gt \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "hyper geometric test 1\n",
    "- check overlap between geneset of all asATAC snp locations with location with all of library snps over background of total # of snps? \n",
    "- check overlap between geneset of asATAC snps (tissue specific) in the library with the mpra hits in the library (tissue specific) \n",
    "- check overlap between geneset of asATAC snps (overall tissues) in the library with the mpra hits in the library all tissues)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "************\n",
      "Airway\n",
      "read ../data/processed/vcf_files/atac/Airway_postvqsr.vcf\n",
      "shape (75646, 11)\n",
      "num_vcf_hits PRE FILTER 75646\n",
      "num_vcf_hits POST FILTER 31211\n",
      "num_vcf_hits POST FDR 1772\n",
      "background library hits that are asATAC 33066 34 0.001028246537228573\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:30: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "atac filter 1605 out of 1772 allele-specific atac 0.9057562076749436\n",
      "atac filter 3067 out of 33066 mpra alleles tested 0.09275388616705982\n",
      "background library hits-atac filt 3067 32 0.010433648516465601\n",
      "************\n",
      "Astrocytes\n",
      "read ../data/processed/vcf_files/atac/Astrocytes_postvqsr.vcf\n",
      "shape (59122, 11)\n",
      "num_vcf_hits PRE FILTER 59122\n",
      "num_vcf_hits POST FILTER 21337\n",
      "num_vcf_hits POST FDR 2017\n",
      "background library hits that are asATAC 33066 51 0.0015423698058428597\n",
      "atac filter 587 out of 2017 allele-specific atac 0.29102627664848785\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:30: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "atac filter 1347 out of 33066 mpra alleles tested 0.04073670840137906\n",
      "background library hits-atac filt 1347 14 0.010393466963622866\n",
      "************\n",
      "Bladder\n",
      "read ../data/processed/vcf_files/atac/Bladder_postvqsr.vcf\n",
      "shape (76857, 11)\n",
      "num_vcf_hits PRE FILTER 76857\n",
      "num_vcf_hits POST FILTER 26112\n",
      "num_vcf_hits POST FDR 1214\n",
      "background library hits that are asATAC 33066 24 0.0007258210851025222\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:30: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "atac filter 1118 out of 1214 allele-specific atac 0.9209225700164745\n",
      "atac filter 3278 out of 33066 mpra alleles tested 0.09913506320691949\n",
      "background library hits-atac filt 3278 22 0.006711409395973154\n",
      "************\n",
      "Colon\n",
      "read ../data/processed/vcf_files/atac/Colon_postvqsr.vcf\n",
      "shape (81035, 11)\n",
      "num_vcf_hits PRE FILTER 81035\n",
      "num_vcf_hits POST FILTER 26331\n",
      "num_vcf_hits POST FDR 7390\n",
      "background library hits that are asATAC 33066 192 0.005806568680820178\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:30: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "atac filter 6632 out of 7390 allele-specific atac 0.8974289580514209\n",
      "atac filter 3468 out of 33066 mpra alleles tested 0.10488114679731446\n",
      "background library hits-atac filt 3468 177 0.05103806228373702\n",
      "************\n",
      "Esophageal\n",
      "read ../data/processed/vcf_files/atac/Esophageal_postvqsr.vcf\n",
      "shape (80599, 11)\n",
      "num_vcf_hits PRE FILTER 80599\n",
      "num_vcf_hits POST FILTER 26104\n",
      "num_vcf_hits POST FDR 8674\n",
      "background library hits that are asATAC 33066 216 0.0065323897659227\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:30: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "atac filter 7884 out of 8674 allele-specific atac 0.9089232188148489\n",
      "atac filter 3591 out of 33066 mpra alleles tested 0.10860097985846488\n",
      "background library hits-atac filt 3591 199 0.05541631857421331\n",
      "************\n",
      "GDSD0\n",
      "read ../data/processed/vcf_files/atac/GDSD0_postvqsr.vcf\n",
      "shape (66677, 11)\n",
      "num_vcf_hits PRE FILTER 66677\n",
      "num_vcf_hits POST FILTER 24220\n",
      "num_vcf_hits POST FDR 2215\n",
      "background library hits that are asATAC 33066 41 0.0012399443537168089\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:30: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "atac filter 1578 out of 2215 allele-specific atac 0.7124153498871332\n",
      "atac filter 3109 out of 33066 mpra alleles tested 0.09402407306598923\n",
      "background library hits-atac filt 3109 30 0.00964940495336121\n",
      "************\n",
      "GDSD3\n",
      "read ../data/processed/vcf_files/atac/GDSD3_postvqsr.vcf\n",
      "shape (72690, 11)\n",
      "num_vcf_hits PRE FILTER 72690\n",
      "num_vcf_hits POST FILTER 26235\n",
      "num_vcf_hits POST FDR 2749\n",
      "background library hits that are asATAC 33066 63 0.001905280348394121\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:30: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "atac filter 2062 out of 2749 allele-specific atac 0.7500909421607858\n",
      "atac filter 3277 out of 33066 mpra alleles tested 0.0991048206617069\n",
      "background library hits-atac filt 3277 38 0.011595971925541654\n",
      "************\n",
      "GDSD6\n",
      "read ../data/processed/vcf_files/atac/GDSD6_postvqsr.vcf\n",
      "shape (71290, 11)\n",
      "num_vcf_hits PRE FILTER 71290\n",
      "num_vcf_hits POST FILTER 25634\n",
      "num_vcf_hits POST FDR 2497\n",
      "background library hits that are asATAC 33066 50 0.0015121272606302546\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:30: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "atac filter 1895 out of 2497 allele-specific atac 0.7589106928313977\n",
      "atac filter 3189 out of 33066 mpra alleles tested 0.09644347668299764\n",
      "background library hits-atac filt 3189 36 0.011288805268109126\n",
      "************\n",
      "GM12878\n",
      "read ../data/processed/vcf_files/atac/GM12878_postvqsr.vcf\n",
      "shape (63362, 11)\n",
      "num_vcf_hits PRE FILTER 63362\n",
      "num_vcf_hits POST FILTER 23755\n",
      "num_vcf_hits POST FDR 1407\n",
      "background library hits that are asATAC 33066 36 0.0010887316276537834\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:30: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "atac filter 1310 out of 1407 allele-specific atac 0.9310589907604833\n",
      "atac filter 2862 out of 33066 mpra alleles tested 0.08655416439847577\n",
      "background library hits-atac filt 2862 34 0.011879804332634521\n",
      "************\n",
      "HMEC\n",
      "read ../data/processed/vcf_files/atac/HMEC_postvqsr.vcf\n",
      "shape (59800, 11)\n",
      "num_vcf_hits PRE FILTER 59800\n",
      "num_vcf_hits POST FILTER 22109\n",
      "num_vcf_hits POST FDR 2272\n",
      "background library hits that are asATAC 33066 52 0.0015726123510554648\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:30: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "atac filter 1779 out of 2272 allele-specific atac 0.7830105633802817\n",
      "atac filter 2906 out of 33066 mpra alleles tested 0.0878848363878304\n",
      "background library hits-atac filt 2906 43 0.014796971782518927\n",
      "************\n",
      "Melanocytes\n",
      "read ../data/processed/vcf_files/atac/Melanocytes_postvqsr.vcf\n",
      "shape (66427, 11)\n",
      "num_vcf_hits PRE FILTER 66427\n",
      "num_vcf_hits POST FILTER 36822\n",
      "num_vcf_hits POST FDR 16189\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:30: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "background library hits that are asATAC 33066 563 0.017026552954696666\n",
      "atac filter 9662 out of 16189 allele-specific atac 0.5968250046327753\n",
      "atac filter 1753 out of 33066 mpra alleles tested 0.053015181757696726\n",
      "background library hits-atac filt 1753 354 0.20193953223046207\n",
      "************\n",
      "Ovarian\n",
      "read ../data/processed/vcf_files/atac/Ovarian_postvqsr.vcf\n",
      "shape (55824, 11)\n",
      "num_vcf_hits PRE FILTER 55824\n",
      "num_vcf_hits POST FILTER 17853\n",
      "num_vcf_hits POST FDR 3233\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:30: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "background library hits that are asATAC 33066 99 0.0029940119760479044\n",
      "atac filter 1580 out of 3233 allele-specific atac 0.4887101763068358\n",
      "atac filter 1822 out of 33066 mpra alleles tested 0.05510191737736648\n",
      "background library hits-atac filt 1822 53 0.029088913282107574\n",
      "************\n",
      "Pancreas\n",
      "read ../data/processed/vcf_files/atac/Pancreas_postvqsr.vcf\n",
      "shape (82065, 11)\n",
      "num_vcf_hits PRE FILTER 82065\n",
      "num_vcf_hits POST FILTER 27878\n",
      "num_vcf_hits POST FDR 7837\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:30: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "background library hits that are asATAC 33066 211 0.006381177039859675\n",
      "atac filter 7120 out of 7837 allele-specific atac 0.9085109097869083\n",
      "atac filter 3630 out of 33066 mpra alleles tested 0.10978043912175649\n",
      "background library hits-atac filt 3630 195 0.05371900826446281\n",
      "************\n",
      "Prostate\n",
      "read ../data/processed/vcf_files/atac/Prostate_postvqsr.vcf\n",
      "shape (41596, 11)\n",
      "num_vcf_hits PRE FILTER 41596\n",
      "num_vcf_hits POST FILTER 11277\n",
      "num_vcf_hits POST FDR 2092\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:30: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "background library hits that are asATAC 33066 59 0.0017843101675437004\n",
      "atac filter 1344 out of 2092 allele-specific atac 0.6424474187380497\n",
      "atac filter 1844 out of 33066 mpra alleles tested 0.055767253372043794\n",
      "background library hits-atac filt 1844 52 0.028199566160520606\n",
      "************\n",
      "Renal\n",
      "read ../data/processed/vcf_files/atac/Renal_postvqsr.vcf\n",
      "shape (79237, 11)\n",
      "num_vcf_hits PRE FILTER 79237\n",
      "num_vcf_hits POST FILTER 26721\n",
      "num_vcf_hits POST FDR 7856\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:30: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "background library hits that are asATAC 33066 208 0.006290449404221859\n",
      "atac filter 7170 out of 7856 allele-specific atac 0.9126782077393075\n",
      "atac filter 3556 out of 33066 mpra alleles tested 0.10754249077602371\n",
      "background library hits-atac filt 3556 193 0.05427446569178852\n",
      "************\n",
      "Thyroid\n",
      "read ../data/processed/vcf_files/atac/Thyroid_postvqsr.vcf\n",
      "shape (79958, 11)\n",
      "num_vcf_hits PRE FILTER 79958\n",
      "num_vcf_hits POST FILTER 26163\n",
      "num_vcf_hits POST FDR 7366\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:30: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "background library hits that are asATAC 33066 197 0.005957781406883203\n",
      "atac filter 6550 out of 7366 allele-specific atac 0.8892207439587293\n",
      "atac filter 3384 out of 33066 mpra alleles tested 0.10234077299945564\n",
      "background library hits-atac filt 3384 176 0.05200945626477541\n",
      "************\n",
      "Uterine\n",
      "read ../data/processed/vcf_files/atac/Uterine_postvqsr.vcf\n",
      "shape (96367, 11)\n",
      "num_vcf_hits PRE FILTER 96367\n",
      "num_vcf_hits POST FILTER 44282\n",
      "num_vcf_hits POST FDR 4929\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:30: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "background library hits that are asATAC 33066 110 0.00332667997338656\n",
      "atac filter 4309 out of 4929 allele-specific atac 0.8742138364779874\n",
      "atac filter 4006 out of 33066 mpra alleles tested 0.12115163612169601\n",
      "background library hits-atac filt 4006 101 0.025212181727408887\n",
      "CPU times: user 5min 32s, sys: 3.48 s, total: 5min 35s\n",
      "Wall time: 5min 40s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "asatac_lib_df = pd.DataFrame()\n",
    "vcf_df_all = pd.DataFrame()\n",
    "vcf_bed_df_atac_all = pd.DataFrame()\n",
    "vcf_df_lib_all = pd.DataFrame()\n",
    "vcf_df_lib_atac_all = pd.DataFrame()\n",
    "for vcf_file in sorted(vcf_files):\n",
    "    tissue = os.path.basename(vcf_file).split('_post')[0]\n",
    "    print('************')\n",
    "    print(tissue)\n",
    "   # read vcf\n",
    "    vcf_df = read_vcf(vcf_file)\n",
    "    num_vcf_hits = vcf_df.shape[0]\n",
    "    print('num_vcf_hits PRE FILTER', num_vcf_hits)  \n",
    "    \n",
    "    # filter vcf for those that meet the heterogenous requiremnts\n",
    "    vcf_df['binom_p'] = vcf_df.apply(lambda row: preprocess_and_filter(row,tissue), axis=1)\n",
    "    vcf_df = vcf_df[vcf_df.binom_p>=0]\n",
    "    vcf_df['tissue'] = tissue\n",
    "    num_vcf_hits_postfilt = vcf_df.shape[0]\n",
    "    print('num_vcf_hits POST FILTER', num_vcf_hits_postfilt)      \n",
    "    \n",
    "    \n",
    "    # binomial test\n",
    "    vcf_df['pass_filt'], vcf_df['p_adjust'] = fdrcorrection(vcf_df.binom_p, alpha=0.05)\n",
    "    vcf_df = vcf_df[vcf_df.pass_filt]\n",
    "    num_vcf_hits_passfdr = vcf_df.shape[0]\n",
    "    print('num_vcf_hits POST FDR', num_vcf_hits_passfdr)      \n",
    "\n",
    "    vcf_bed_df = vcf_df[['CHROM','POS','rowname']]\n",
    "    vcf_bed_df['stop'] = vcf_df.POS.map(int) + 1\n",
    "    vcf_bed_df = vcf_bed_df[['CHROM','POS','stop','rowname']]\n",
    "    vcf_bed_df.columns = ['chr','start','stop','name']\n",
    "    vcf_df_all = pd.concat([vcf_df_all, vcf_df])\n",
    "\n",
    "    vcf_df_lib = vcf_df.merge(lib_df_bed_df, how='inner',on='rowname')\n",
    "    vcf_df_lib_all = pd.concat([vcf_df_lib_all, vcf_df_lib])\n",
    "    lib_hits = vcf_df_lib.rowname.unique()\n",
    "    num_lib_hits = lib_hits.shape[0]\n",
    "    print('background library hits that are asATAC', num_lib_tested,num_lib_hits, num_lib_hits/ num_lib_tested)\n",
    "    \n",
    "#     if tissue in mpra_tissue_mapper:\n",
    "#         mpra_res_df_tissue = mpra_res_df[mpra_res_df.tissue==mpra_tissue_mapper[tissue]]\n",
    "#         mpra_hits_tissue = mpra_res_df_tissue.rowname.unique()\n",
    "#         num_mpra_hits_tissue= mpra_hits_tissue.shape[0]\n",
    "# #         mpra_sig_df_lib = vcf_df.merge(mpra_res_df_tissue, how='inner',on='rowname')\n",
    "\n",
    "#         num_mpra_asatac_hits = len(set(lib_hits).intersection(set(mpra_hits_tissue)))\n",
    "#         asatac_lib_df_tissue=pd.DataFrame()\n",
    "#         asatac_lib_df_tissue['rowname'] = sorted(set(lib_hits).intersection(set(mpra_hits_tissue)))\n",
    "#         asatac_lib_df_tissue['tissue'] =  mpra_tissue_mapper[tissue]\n",
    "#         asatac_lib_df = pd.concat([asatac_lib_df, asatac_lib_df_tissue])\n",
    "#         # do a hypergeometric test between being asATAC and being an mpra hit over a background of being in the mpra dataset\n",
    "    \n",
    "#         print('mpra sig library hits in tissue ', num_mpra_hits_tissue)\n",
    "#         print('overlap:', num_mpra_asatac_hits)\n",
    "# #         oddsratio, pvalue = stats.fisher_exact([[num_lib_hits, num_lib_tested- num_lib_hits],\n",
    "# #                                                 [num_mpra_hits, num_mpra_sig_hits - num_mpra_hits]])\n",
    "# #         print('fisher for mpra and as annotation association', pvalue, oddsratio)\n",
    "#         phyper =  hypergeometric_test(num_mpra_asatac_hits, num_lib_tested, num_mpra_hits_tissue, num_lib_hits)\n",
    "#         #stats.hypergeom.cdf(num_mpra_asatac_hits, num_lib_tested, num_lib_hits, num_mpra_hits)\n",
    "#         print('hypergeometric test between being asATAC and MPRA hit over backgroun of being in mpra_dataset', phyper )\n",
    "#         phyper =  hypergeometric_test(num_mpra_asatac_hits, num_vcf_hits_postfilt, num_mpra_hits_tissue, num_lib_hits)\n",
    "#         print('hypergeometric test between being asATAC and MPRA hit over backgroun of being in an asatac', phyper )\n",
    "    \n",
    "    ## filter through atac\n",
    "    atac_tissue = atac_tissue_mapper[tissue]\n",
    "    atac_bed_file = os.path.join(atac_dir, atac_tissue+'_merged.bed')\n",
    "    if not os.path.exists(atac_bed_file):\n",
    "        print('atac file',atac_bed_file, 'does not exist')\n",
    "        continue\n",
    "    atac_bed = pybedtools.BedTool(atac_bed_file)\n",
    "\n",
    "    vcf_bed_df_atac = pybedtools.BedTool.from_dataframe(vcf_bed_df).intersect(atac_bed).to_dataframe()\n",
    "    num_atac_vcf = vcf_bed_df_atac.name.unique().shape[0]\n",
    "    vcf_bed_df_atac['tissue'] = tissue\n",
    "    vcf_bed_df_atac_all = pd.concat([vcf_bed_df_atac_all, vcf_bed_df_atac])\n",
    "    print('atac filter', num_atac_vcf, 'out of', vcf_bed_df.shape[0], 'allele-specific atac', num_atac_vcf/vcf_bed_df.shape[0])\n",
    "    lib_df_bed_atac = lib_df_bed.intersect(atac_bed).to_dataframe()\n",
    "    lib_atac  = lib_df_bed_atac.name.unique()\n",
    "    num_lib_atac = lib_atac.shape[0]\n",
    "    print('atac filter', num_lib_atac, 'out of', num_lib_tested, 'mpra alleles tested', num_lib_atac/num_lib_tested)\n",
    "#     mpra_atac = lib_df_bed_atac[lib_df_bed_atac.name.isin(mpra_rownames_sig)].name.unique()\n",
    "#     num_mpra_atac = mpra_atac.shape[0]\n",
    "#     print('atac filter', num_mpra_atac, 'out of', num_mpra_sig_hits, 'mpra alleles sig',num_mpra_atac/num_mpra_sig_hits)\n",
    "    \n",
    "    vcf_df_lib_atac = vcf_bed_df_atac.merge(lib_df_bed_df, how='inner',left_on='name',right_on='rowname')\n",
    "    num_lib_hits_atac = vcf_df_lib_atac.rowname.unique().shape[0]\n",
    "    vcf_df_lib_atac_all = pd.concat([vcf_df_lib_atac_all, vcf_df_lib_atac])\n",
    "#     mpra_sig_df_lib_atac = vcf_bed_df_atac.merge(mpra_res_df, how='inner',left_on='name',right_on='rowname')\n",
    "#     num_mpra_hits_atac = mpra_sig_df_lib_atac.rowname.unique().shape[0]\n",
    "    \n",
    "    print('background library hits-atac filt', num_lib_atac,num_lib_hits_atac, num_lib_hits_atac/ num_lib_atac)\n",
    "#     print('mpra sig library hits-atac filt', num_mpra_atac,num_mpra_hits_atac, num_mpra_hits_atac/ num_mpra_atac)\n",
    "#     if tissue in mpra_tissue_mapper:\n",
    "#         mpra_hits_tissue_atac = set(mpra_hits_tissue).intersection(set(lib_atac))\n",
    "#         num_mpra_hits_tissue_atac= mpra_hits_tissue.shape[0]\n",
    "# #         mpra_sig_df_lib = vcf_df.merge(mpra_res_df_tissue, how='inner',on='rowname')\n",
    "\n",
    "#         num_mpra_asatac_filtatac_hits = len(set(lib_hits).intersection(set(mpra_hits_tissue_atac)))\n",
    "#         # do a hypergeometric test between being asATAC and being an mpra hit over a background of being in the mpra dataset\n",
    "\n",
    "#         print('mpra sig library hits in tissue atac filt ', num_mpra_hits_tissue_atac)\n",
    "#         print('overlap atac:', num_mpra_asatac_filtatac_hits)\n",
    "# #         oddsratio, pvalue = stats.fisher_exact([[num_lib_hits, num_lib_tested- num_lib_hits],\n",
    "# #                                                 [num_mpra_hits, num_mpra_sig_hits - num_mpra_hits]])\n",
    "# #         print('fisher for mpra and as annotation association', pvalue, oddsratio)\n",
    "#         phyper =  hypergeometric_test(num_mpra_asatac_filtatac_hits, num_atac_vcf, num_mpra_hits_tissue_atac, num_lib_atac)\n",
    "#         #stats.hypergeom.cdf(num_mpra_asatac_hits, num_lib_tested, num_lib_hits, num_mpra_hits)\n",
    "#         print('hypergeometric test between being asATAC and MPRA hit over backgroun of being asatac', phyper )\n",
    "\n",
    "\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "we can see by the % that the mpra significant hits are overenriched compared to library background\n",
    "- choice of library background can be changed to all gwas non cancer hits??? (will do next \n",
    "\n",
    "\n",
    "\n",
    "that means that MPRA signal is semi effected by "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.040283070223189985"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "1332/33066"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO', 'FORMAT',\n",
       "       'sample', 'rowname', 'binom_p', 'tissue', 'pass_filt', 'p_adjust'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcf_df_all.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(81709, 15)\n",
      "(248925, 5)\n"
     ]
    }
   ],
   "source": [
    "print(vcf_df_all.shape)\n",
    "vcf_df_all.to_csv(os.path.join(save_dir, 'vcf_df_all_atac.csv'))\n",
    "print(vcf_bed_df_atac_all.shape)\n",
    "vcf_bed_df_atac_all.to_csv(os.path.join(save_dir, 'vcf_bed_df_atac_all.csv'))\n",
    "vcf_df_lib_all.to_csv(os.path.join(save_dir, 'vcf_df_lib_all_atac.csv'))\n",
    "vcf_df_lib_atac_all.to_csv(os.path.join(save_dir, 'vcf_df_lib_atac_all.csv'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start_x</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>tissue</th>\n",
       "      <th>chr</th>\n",
       "      <th>start_y</th>\n",
       "      <th>stop</th>\n",
       "      <th>rowname</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>38461821</td>\n",
       "      <td>38461822</td>\n",
       "      <td>chr1_38461821</td>\n",
       "      <td>Airway</td>\n",
       "      <td>chr1</td>\n",
       "      <td>38461821</td>\n",
       "      <td>38461822</td>\n",
       "      <td>chr1_38461821</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>154839799</td>\n",
       "      <td>154839800</td>\n",
       "      <td>chr1_154839799</td>\n",
       "      <td>Airway</td>\n",
       "      <td>chr1</td>\n",
       "      <td>154839799</td>\n",
       "      <td>154839800</td>\n",
       "      <td>chr1_154839799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>209979635</td>\n",
       "      <td>209979636</td>\n",
       "      <td>chr1_209979635</td>\n",
       "      <td>Airway</td>\n",
       "      <td>chr1</td>\n",
       "      <td>209979635</td>\n",
       "      <td>209979636</td>\n",
       "      <td>chr1_209979635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>chr2</td>\n",
       "      <td>29119930</td>\n",
       "      <td>29119931</td>\n",
       "      <td>chr2_29119930</td>\n",
       "      <td>Airway</td>\n",
       "      <td>chr2</td>\n",
       "      <td>29119930</td>\n",
       "      <td>29119931</td>\n",
       "      <td>chr2_29119930</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>chr2</td>\n",
       "      <td>110316285</td>\n",
       "      <td>110316286</td>\n",
       "      <td>chr2_110316285</td>\n",
       "      <td>Airway</td>\n",
       "      <td>chr2</td>\n",
       "      <td>110316285</td>\n",
       "      <td>110316286</td>\n",
       "      <td>chr2_110316285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>346</td>\n",
       "      <td>chr19</td>\n",
       "      <td>49217305</td>\n",
       "      <td>49217306</td>\n",
       "      <td>chr19_49217305</td>\n",
       "      <td>Uterine</td>\n",
       "      <td>chr19</td>\n",
       "      <td>49217305</td>\n",
       "      <td>49217306</td>\n",
       "      <td>chr19_49217305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>347</td>\n",
       "      <td>chr19</td>\n",
       "      <td>49217305</td>\n",
       "      <td>49217306</td>\n",
       "      <td>chr19_49217305</td>\n",
       "      <td>Uterine</td>\n",
       "      <td>chr19</td>\n",
       "      <td>49217305</td>\n",
       "      <td>49217306</td>\n",
       "      <td>chr19_49217305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>348</td>\n",
       "      <td>chr19</td>\n",
       "      <td>49217305</td>\n",
       "      <td>49217306</td>\n",
       "      <td>chr19_49217305</td>\n",
       "      <td>Uterine</td>\n",
       "      <td>chr19</td>\n",
       "      <td>49217305</td>\n",
       "      <td>49217306</td>\n",
       "      <td>chr19_49217305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>349</td>\n",
       "      <td>chr19</td>\n",
       "      <td>49217305</td>\n",
       "      <td>49217306</td>\n",
       "      <td>chr19_49217305</td>\n",
       "      <td>Uterine</td>\n",
       "      <td>chr19</td>\n",
       "      <td>49217305</td>\n",
       "      <td>49217306</td>\n",
       "      <td>chr19_49217305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>350</td>\n",
       "      <td>chr19</td>\n",
       "      <td>49217305</td>\n",
       "      <td>49217306</td>\n",
       "      <td>chr19_49217305</td>\n",
       "      <td>Uterine</td>\n",
       "      <td>chr19</td>\n",
       "      <td>49217305</td>\n",
       "      <td>49217306</td>\n",
       "      <td>chr19_49217305</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7184 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom    start_x        end            name   tissue    chr    start_y  \\\n",
       "0     chr1   38461821   38461822   chr1_38461821   Airway   chr1   38461821   \n",
       "1     chr1  154839799  154839800  chr1_154839799   Airway   chr1  154839799   \n",
       "2     chr1  209979635  209979636  chr1_209979635   Airway   chr1  209979635   \n",
       "3     chr2   29119930   29119931   chr2_29119930   Airway   chr2   29119930   \n",
       "4     chr2  110316285  110316286  chr2_110316285   Airway   chr2  110316285   \n",
       "..     ...        ...        ...             ...      ...    ...        ...   \n",
       "346  chr19   49217305   49217306  chr19_49217305  Uterine  chr19   49217305   \n",
       "347  chr19   49217305   49217306  chr19_49217305  Uterine  chr19   49217305   \n",
       "348  chr19   49217305   49217306  chr19_49217305  Uterine  chr19   49217305   \n",
       "349  chr19   49217305   49217306  chr19_49217305  Uterine  chr19   49217305   \n",
       "350  chr19   49217305   49217306  chr19_49217305  Uterine  chr19   49217305   \n",
       "\n",
       "          stop         rowname  \n",
       "0     38461822   chr1_38461821  \n",
       "1    154839800  chr1_154839799  \n",
       "2    209979636  chr1_209979635  \n",
       "3     29119931   chr2_29119930  \n",
       "4    110316286  chr2_110316285  \n",
       "..         ...             ...  \n",
       "346   49217306  chr19_49217305  \n",
       "347   49217306  chr19_49217305  \n",
       "348   49217306  chr19_49217305  \n",
       "349   49217306  chr19_49217305  \n",
       "350   49217306  chr19_49217305  \n",
       "\n",
       "[7184 rows x 9 columns]"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcf_df_lib_atac_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(898, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rowname</th>\n",
       "      <th>tissue_asatac</th>\n",
       "      <th>bool_is_asatac</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>chr10_104239100</td>\n",
       "      <td>Renal</td>\n",
       "      <td>True</td>\n",
       "      <td>chr10</td>\n",
       "      <td>104239100</td>\n",
       "      <td>104239101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>chr10_104262628</td>\n",
       "      <td>Colon|Esophageal|Melanocytes|Pancreas|Prostate...</td>\n",
       "      <td>True</td>\n",
       "      <td>chr10</td>\n",
       "      <td>104262628</td>\n",
       "      <td>104262629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>chr10_104263675</td>\n",
       "      <td>Colon|Esophageal|Melanocytes|Pancreas|Renal|Th...</td>\n",
       "      <td>True</td>\n",
       "      <td>chr10</td>\n",
       "      <td>104263675</td>\n",
       "      <td>104263676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>chr10_104269217</td>\n",
       "      <td>Renal</td>\n",
       "      <td>True</td>\n",
       "      <td>chr10</td>\n",
       "      <td>104269217</td>\n",
       "      <td>104269218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>chr10_104269301</td>\n",
       "      <td>Colon|Renal</td>\n",
       "      <td>True</td>\n",
       "      <td>chr10</td>\n",
       "      <td>104269301</td>\n",
       "      <td>104269302</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           rowname                                      tissue_asatac  \\\n",
       "0  chr10_104239100                                              Renal   \n",
       "1  chr10_104262628  Colon|Esophageal|Melanocytes|Pancreas|Prostate...   \n",
       "2  chr10_104263675  Colon|Esophageal|Melanocytes|Pancreas|Renal|Th...   \n",
       "3  chr10_104269217                                              Renal   \n",
       "4  chr10_104269301                                        Colon|Renal   \n",
       "\n",
       "   bool_is_asatac    chr      start       stop  \n",
       "0            True  chr10  104239100  104239101  \n",
       "1            True  chr10  104262628  104262629  \n",
       "2            True  chr10  104263675  104263676  \n",
       "3            True  chr10  104269217  104269218  \n",
       "4            True  chr10  104269301  104269302  "
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "asatac_lib_df = vcf_df_lib_atac_all.merge(vcf_df_all[['rowname','tissue']],how='left',on=['rowname','tissue']).drop_duplicates().sort_values('tissue').groupby('rowname').agg({'tissue':'|'.join}).reset_index()\n",
    "asatac_lib_df.columns = ['rowname','tissue_asatac']\n",
    "asatac_lib_df['bool_is_asatac'] = True\n",
    "asatac_lib_df= asatac_lib_df.merge(lib_df_bed_df, how='left',on='rowname')\n",
    "asatac_lib_df.to_csv(os.path.join(save_dir,'lib_asatac_annotation_postfilter.csv'))\n",
    "print(asatac_lib_df.shape)\n",
    "asatac_lib_df[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(29,)"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcf_bed_df_atac_all.name[vcf_bed_df_atac_all.name.isin(mpra_res_df.rowname.unique())].unique().shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1B. ATAC hypergeometric test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_df_all = pd.read_csv(os.path.join(save_dir, 'vcf_df_all_atac.csv'),index_col=0)\n",
    "vcf_df_lib_all = pd.read_csv(os.path.join(save_dir, 'vcf_df_lib_all_atac.csv'),index_col=0)\n",
    "vcf_bed_df_atac_all = pd.read_csv(os.path.join(save_dir, 'vcf_bed_df_atac_all.csv'),index_col=0)\n",
    "asatac_lib_df = pd.read_csv(os.path.join(save_dir,'lib_asatac_annotation_postfilter.csv'),index_col=0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>disease</th>\n",
       "      <th>hit</th>\n",
       "      <th>snp_type</th>\n",
       "      <th>rsid</th>\n",
       "      <th>seqnames</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>width</th>\n",
       "      <th>strand</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>chr</th>\n",
       "      <th>seq_id</th>\n",
       "      <th>rowname</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>hmec</td>\n",
       "      <td>NaN</td>\n",
       "      <td>linked_snp</td>\n",
       "      <td>rs4866783</td>\n",
       "      <td>chr5</td>\n",
       "      <td>44876507</td>\n",
       "      <td>44876507</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>5</td>\n",
       "      <td>5:44876507</td>\n",
       "      <td>chr5_44876507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>hmec</td>\n",
       "      <td>NaN</td>\n",
       "      <td>linked_snp</td>\n",
       "      <td>rs75282042</td>\n",
       "      <td>chr5</td>\n",
       "      <td>50735307</td>\n",
       "      <td>50735307</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>5</td>\n",
       "      <td>5:50735307</td>\n",
       "      <td>chr5_50735307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>hmec</td>\n",
       "      <td>NaN</td>\n",
       "      <td>linked_snp</td>\n",
       "      <td>rs115010969</td>\n",
       "      <td>chr5</td>\n",
       "      <td>52535211</td>\n",
       "      <td>52535211</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>5</td>\n",
       "      <td>5:52535211</td>\n",
       "      <td>chr5_52535211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>10</td>\n",
       "      <td>hmec</td>\n",
       "      <td>NaN</td>\n",
       "      <td>linked_snp</td>\n",
       "      <td>rs17343328</td>\n",
       "      <td>chr5</td>\n",
       "      <td>52615305</td>\n",
       "      <td>52615305</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>5</td>\n",
       "      <td>5:52615305</td>\n",
       "      <td>chr5_52615305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10</td>\n",
       "      <td>11</td>\n",
       "      <td>hmec</td>\n",
       "      <td>NaN</td>\n",
       "      <td>linked_snp</td>\n",
       "      <td>rs10940312</td>\n",
       "      <td>chr5</td>\n",
       "      <td>52667982</td>\n",
       "      <td>52667982</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>5</td>\n",
       "      <td>5:52667982</td>\n",
       "      <td>chr5_52667982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7663</td>\n",
       "      <td>7664</td>\n",
       "      <td>hmec</td>\n",
       "      <td>NaN</td>\n",
       "      <td>linked_snp</td>\n",
       "      <td>rs16901937</td>\n",
       "      <td>chr5</td>\n",
       "      <td>44709141</td>\n",
       "      <td>44709141</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>5</td>\n",
       "      <td>5:44709141</td>\n",
       "      <td>chr5_44709141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7664</td>\n",
       "      <td>7665</td>\n",
       "      <td>hmec</td>\n",
       "      <td>NaN</td>\n",
       "      <td>linked_snp</td>\n",
       "      <td>rs1866406</td>\n",
       "      <td>chr5</td>\n",
       "      <td>44809945</td>\n",
       "      <td>44809945</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>5</td>\n",
       "      <td>5:44809945</td>\n",
       "      <td>chr5_44809945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7667</td>\n",
       "      <td>7668</td>\n",
       "      <td>hmec</td>\n",
       "      <td>NaN</td>\n",
       "      <td>linked_snp</td>\n",
       "      <td>rs10070037</td>\n",
       "      <td>chr5</td>\n",
       "      <td>44870237</td>\n",
       "      <td>44870237</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>5</td>\n",
       "      <td>5:44870237</td>\n",
       "      <td>chr5_44870237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7668</td>\n",
       "      <td>7669</td>\n",
       "      <td>thy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>linked_snp</td>\n",
       "      <td>rs4915073</td>\n",
       "      <td>chr1</td>\n",
       "      <td>108325879</td>\n",
       "      <td>108325879</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>1:108325879</td>\n",
       "      <td>chr1_108325879</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7669</td>\n",
       "      <td>7670</td>\n",
       "      <td>hmec</td>\n",
       "      <td>NaN</td>\n",
       "      <td>linked_snp</td>\n",
       "      <td>rs408320</td>\n",
       "      <td>chr1</td>\n",
       "      <td>8085328</td>\n",
       "      <td>8085328</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>1:8085328</td>\n",
       "      <td>chr1_8085328</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4654 rows Ã— 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Unnamed: 0 disease  hit    snp_type         rsid seqnames      start  \\\n",
       "0              1    hmec  NaN  linked_snp    rs4866783     chr5   44876507   \n",
       "3              4    hmec  NaN  linked_snp   rs75282042     chr5   50735307   \n",
       "6              7    hmec  NaN  linked_snp  rs115010969     chr5   52535211   \n",
       "9             10    hmec  NaN  linked_snp   rs17343328     chr5   52615305   \n",
       "10            11    hmec  NaN  linked_snp   rs10940312     chr5   52667982   \n",
       "...          ...     ...  ...         ...          ...      ...        ...   \n",
       "7663        7664    hmec  NaN  linked_snp   rs16901937     chr5   44709141   \n",
       "7664        7665    hmec  NaN  linked_snp    rs1866406     chr5   44809945   \n",
       "7667        7668    hmec  NaN  linked_snp   rs10070037     chr5   44870237   \n",
       "7668        7669     thy  NaN  linked_snp    rs4915073     chr1  108325879   \n",
       "7669        7670    hmec  NaN  linked_snp     rs408320     chr1    8085328   \n",
       "\n",
       "            end  width strand REF ALT  chr       seq_id         rowname  \n",
       "0      44876507      1      *   C   T    5   5:44876507   chr5_44876507  \n",
       "3      50735307      1      *   T   C    5   5:50735307   chr5_50735307  \n",
       "6      52535211      1      *   T   A    5   5:52535211   chr5_52535211  \n",
       "9      52615305      1      *   G   A    5   5:52615305   chr5_52615305  \n",
       "10     52667982      1      *   A   T    5   5:52667982   chr5_52667982  \n",
       "...         ...    ...    ...  ..  ..  ...          ...             ...  \n",
       "7663   44709141      1      *   A   G    5   5:44709141   chr5_44709141  \n",
       "7664   44809945      1      *   G   C    5   5:44809945   chr5_44809945  \n",
       "7667   44870237      1      *   A   T    5   5:44870237   chr5_44870237  \n",
       "7668  108325879      1      *   A   T    1  1:108325879  chr1_108325879  \n",
       "7669    8085328      1      *   T   C    1    1:8085328    chr1_8085328  \n",
       "\n",
       "[4654 rows x 15 columns]"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lib_df = pd.read_csv('/Users/mguo123/Google Drive/1_khavari/omics_project-LD/pan_omics/data/external/GWAS/cancer_mpra_snps_withlead_annon.csv')\n",
    "lib_df = lib_df[lib_df.snp_type=='linked_snp']\n",
    "lib_df['rowname'] = lib_df.seqnames + '_'+lib_df.start.map(str)\n",
    "lib_df#.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr_37</th>\n",
       "      <th>Start_37</th>\n",
       "      <th>disease</th>\n",
       "      <th>hit_tissue</th>\n",
       "      <th>rowname</th>\n",
       "      <th>mapped_tissue</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Causal_SNP</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>rs111640187</td>\n",
       "      <td>5</td>\n",
       "      <td>82428995</td>\n",
       "      <td>airway</td>\n",
       "      <td>airway</td>\n",
       "      <td>chr5_82428995</td>\n",
       "      <td>Airway</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>rs34064842</td>\n",
       "      <td>6</td>\n",
       "      <td>27688625</td>\n",
       "      <td>airway,hmec</td>\n",
       "      <td>airway</td>\n",
       "      <td>chr6_27688625</td>\n",
       "      <td>Airway</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>rs71546575</td>\n",
       "      <td>6</td>\n",
       "      <td>27805022</td>\n",
       "      <td>airway,hmec</td>\n",
       "      <td>airway</td>\n",
       "      <td>chr6_27805022</td>\n",
       "      <td>Airway</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>rs35353359</td>\n",
       "      <td>6</td>\n",
       "      <td>28324378</td>\n",
       "      <td>airway,hmec</td>\n",
       "      <td>airway</td>\n",
       "      <td>chr6_28324378</td>\n",
       "      <td>Airway</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>rs3095311</td>\n",
       "      <td>6</td>\n",
       "      <td>31051675</td>\n",
       "      <td>airway</td>\n",
       "      <td>airway</td>\n",
       "      <td>chr6_31051675</td>\n",
       "      <td>Airway</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>rs34992253</td>\n",
       "      <td>14</td>\n",
       "      <td>36647163</td>\n",
       "      <td>thy</td>\n",
       "      <td>thy</td>\n",
       "      <td>chr14_36647163</td>\n",
       "      <td>Thyroid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>rs17293632</td>\n",
       "      <td>15</td>\n",
       "      <td>67442596</td>\n",
       "      <td>thy</td>\n",
       "      <td>thy</td>\n",
       "      <td>chr15_67442596</td>\n",
       "      <td>Thyroid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>rs13406698</td>\n",
       "      <td>2</td>\n",
       "      <td>218283085</td>\n",
       "      <td>hmec,thy</td>\n",
       "      <td>thy</td>\n",
       "      <td>chr2_218283085</td>\n",
       "      <td>Thyroid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>rs16857609</td>\n",
       "      <td>2</td>\n",
       "      <td>218296508</td>\n",
       "      <td>hmec,thy</td>\n",
       "      <td>thy</td>\n",
       "      <td>chr2_218296508</td>\n",
       "      <td>Thyroid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>rs12617808</td>\n",
       "      <td>2</td>\n",
       "      <td>218302217</td>\n",
       "      <td>thy</td>\n",
       "      <td>thy</td>\n",
       "      <td>chr2_218302217</td>\n",
       "      <td>Thyroid</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>451 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Chr_37   Start_37      disease hit_tissue         rowname  \\\n",
       "Causal_SNP                                                               \n",
       "rs111640187       5   82428995       airway     airway   chr5_82428995   \n",
       "rs34064842        6   27688625  airway,hmec     airway   chr6_27688625   \n",
       "rs71546575        6   27805022  airway,hmec     airway   chr6_27805022   \n",
       "rs35353359        6   28324378  airway,hmec     airway   chr6_28324378   \n",
       "rs3095311         6   31051675       airway     airway   chr6_31051675   \n",
       "...             ...        ...          ...        ...             ...   \n",
       "rs34992253       14   36647163          thy        thy  chr14_36647163   \n",
       "rs17293632       15   67442596          thy        thy  chr15_67442596   \n",
       "rs13406698        2  218283085     hmec,thy        thy  chr2_218283085   \n",
       "rs16857609        2  218296508     hmec,thy        thy  chr2_218296508   \n",
       "rs12617808        2  218302217          thy        thy  chr2_218302217   \n",
       "\n",
       "            mapped_tissue  \n",
       "Causal_SNP                 \n",
       "rs111640187        Airway  \n",
       "rs34064842         Airway  \n",
       "rs71546575         Airway  \n",
       "rs35353359         Airway  \n",
       "rs3095311          Airway  \n",
       "...                   ...  \n",
       "rs34992253        Thyroid  \n",
       "rs17293632        Thyroid  \n",
       "rs13406698        Thyroid  \n",
       "rs16857609        Thyroid  \n",
       "rs12617808        Thyroid  \n",
       "\n",
       "[451 rows x 6 columns]"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mpra_res_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [],
   "source": [
    "#         phyper =  hypergeometric_test(num_mpra_asatac_hits, num_lib_tested, num_mpra_hits_tissue, num_lib_hits)\n",
    "#         #stats.hypergeom.cdf(num_mpra_asatac_hits, num_lib_tested, num_lib_hits, num_mpra_hits)\n",
    "#         print('hypergeometric test between being asATAC and MPRA hit over backgroun of being in mpra_dataset', phyper )\n",
    "#         phyper =  hypergeometric_test(num_mpra_asatac_hits, num_vcf_hits_postfilt, num_mpra_hits_tissue, num_lib_hits)\n",
    "#         print('hypergeometric test between being asATAC and MPRA hit over backgroun of being in an asatac', phyper )\n",
    " #         # do a hypergeometric test between being asATAC and being an mpra hit over a background of being in the mpra dataset\n",
    "#     - M is total number of objects\n",
    "#     - n is total number of Type I objects. \n",
    "#     - x (random variate) represents the number of Type I objects in N drawn without replacement from the total population\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True     403\n",
       "False     48\n",
       "Name: rowname, dtype: int64"
      ]
     },
     "execution_count": 163,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mpra_res_df.rowname.isin(lib_df.rowname).value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr3_194872824\n",
      "chr6_167370211\n",
      "chr12_48395043\n",
      "chr2_202223566\n",
      "chr13_31314455\n",
      "chr17_76165104\n",
      "chr2_70141519\n",
      "chr7_97914305\n",
      "chr10_38645117\n",
      "chr15_67442596\n",
      "chr4_164415734\n",
      "chr7_97924226\n",
      "chr11_18343822\n",
      "chr19_13950294\n",
      "chr1_155197268\n",
      "chr5_67615731\n",
      "chr6_4403131\n",
      "chr19_41857404\n",
      "chr2_67624302\n",
      "chr10_38383147\n",
      "chr10_38383793\n",
      "chr10_96122826\n",
      "chr11_69354931\n",
      "chr15_40337462\n",
      "chr16_50100253\n",
      "chr2_67878108\n",
      "chr19_13955161\n",
      "chr1_10556447\n",
      "chr6_117803538\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "29"
      ]
     },
     "execution_count": 164,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "asatac_snps =[] \n",
    "for rowname in vcf_df_lib_atac_all.rowname.unique() :\n",
    "    if rowname in mpra_res_df.rowname.unique():\n",
    "        asatac_snps.append(rowname)\n",
    "        print(rowname)\n",
    "len(asatac_snps)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "168"
      ]
     },
     "execution_count": 165,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "asatac_snps_lib =[] \n",
    "for rowname in vcf_df_lib_atac_all.rowname.unique() :\n",
    "    if rowname in lib_df.rowname.unique():\n",
    "        asatac_snps_lib.append(rowname)\n",
    "#         print(rowname)\n",
    "len(asatac_snps_lib)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_df_all['rowname'] = vcf_df_all.CHROM + '_' + vcf_df_all.POS.map(str)\n",
    "vcf_df_all[vcf_df_all.rowname.isin(asatac_snps)].to_csv(os.path.join(save_dir, 'asatac-mprahits-wpval.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-67-c3fb6ce7aa7a>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0masatac_lib_snps\u001b[0m \u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mrowname\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mvcf_df_lib_atac_all\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrowname\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m     \u001b[0;32mif\u001b[0m \u001b[0mrowname\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mlib_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrowname\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m         \u001b[0masatac_lib_snps\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrowname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;31m#         print(rowname)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/pandas/core/series.py\u001b[0m in \u001b[0;36munique\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1986\u001b[0m         \u001b[0mCategories\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;36m3\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0ma\u001b[0m \u001b[0;34m<\u001b[0m \u001b[0mb\u001b[0m \u001b[0;34m<\u001b[0m \u001b[0mc\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1987\u001b[0m         \"\"\"\n\u001b[0;32m-> 1988\u001b[0;31m         \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msuper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1989\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1990\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/pandas/core/base.py\u001b[0m in \u001b[0;36munique\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1403\u001b[0m             \u001b[0;32mfrom\u001b[0m \u001b[0mpandas\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcore\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0malgorithms\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0munique1d\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1404\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1405\u001b[0;31m             \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0munique1d\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1406\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1407\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/pandas/core/algorithms.py\u001b[0m in \u001b[0;36munique\u001b[0;34m(values)\u001b[0m\n\u001b[1;32m    403\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    404\u001b[0m     \u001b[0mtable\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mhtable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 405\u001b[0;31m     \u001b[0muniques\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtable\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    406\u001b[0m     \u001b[0muniques\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_reconstruct_data\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0muniques\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moriginal\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    407\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0muniques\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "asatac_lib_snps =[] \n",
    "for rowname in vcf_df_lib_atac_all.rowname.unique() :\n",
    "    if rowname in lib_df.rowname.unique():\n",
    "        asatac_lib_snps.append(rowname)\n",
    "#         print(rowname)\n",
    "len(asatac_lib_snps)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rowname</th>\n",
       "      <th>tissue_asatac</th>\n",
       "      <th>bool_is_asatac</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>chr10_104239100</td>\n",
       "      <td>Renal|Thyroid</td>\n",
       "      <td>True</td>\n",
       "      <td>chr10</td>\n",
       "      <td>104239100</td>\n",
       "      <td>104239101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>chr10_104262628</td>\n",
       "      <td>Bladder|Colon|Esophageal|Melanocytes|Pancreas|...</td>\n",
       "      <td>True</td>\n",
       "      <td>chr10</td>\n",
       "      <td>104262628</td>\n",
       "      <td>104262629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>chr10_104263675</td>\n",
       "      <td>Colon|Esophageal|Melanocytes|Ovarian|Pancreas|...</td>\n",
       "      <td>True</td>\n",
       "      <td>chr10</td>\n",
       "      <td>104263675</td>\n",
       "      <td>104263676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>chr10_104269217</td>\n",
       "      <td>Renal</td>\n",
       "      <td>True</td>\n",
       "      <td>chr10</td>\n",
       "      <td>104269217</td>\n",
       "      <td>104269218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>chr10_104269301</td>\n",
       "      <td>Airway|Colon|Renal</td>\n",
       "      <td>True</td>\n",
       "      <td>chr10</td>\n",
       "      <td>104269301</td>\n",
       "      <td>104269302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1342</td>\n",
       "      <td>chr9_33817617</td>\n",
       "      <td>Melanocytes</td>\n",
       "      <td>True</td>\n",
       "      <td>chr9</td>\n",
       "      <td>33817617</td>\n",
       "      <td>33817618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1343</td>\n",
       "      <td>chr9_34074476</td>\n",
       "      <td>Prostate|Uterine</td>\n",
       "      <td>True</td>\n",
       "      <td>chr9</td>\n",
       "      <td>34074476</td>\n",
       "      <td>34074477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1344</td>\n",
       "      <td>chr9_35538626</td>\n",
       "      <td>GDSD0|GDSD3|GDSD6</td>\n",
       "      <td>True</td>\n",
       "      <td>chr9</td>\n",
       "      <td>35538626</td>\n",
       "      <td>35538627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1345</td>\n",
       "      <td>chr9_6008571</td>\n",
       "      <td>Melanocytes</td>\n",
       "      <td>True</td>\n",
       "      <td>chr9</td>\n",
       "      <td>6008571</td>\n",
       "      <td>6008572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1346</td>\n",
       "      <td>chr9_93940390</td>\n",
       "      <td>Bladder</td>\n",
       "      <td>True</td>\n",
       "      <td>chr9</td>\n",
       "      <td>93940390</td>\n",
       "      <td>93940391</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1347 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              rowname                                      tissue_asatac  \\\n",
       "0     chr10_104239100                                      Renal|Thyroid   \n",
       "1     chr10_104262628  Bladder|Colon|Esophageal|Melanocytes|Pancreas|...   \n",
       "2     chr10_104263675  Colon|Esophageal|Melanocytes|Ovarian|Pancreas|...   \n",
       "3     chr10_104269217                                              Renal   \n",
       "4     chr10_104269301                                 Airway|Colon|Renal   \n",
       "...               ...                                                ...   \n",
       "1342    chr9_33817617                                        Melanocytes   \n",
       "1343    chr9_34074476                                   Prostate|Uterine   \n",
       "1344    chr9_35538626                                  GDSD0|GDSD3|GDSD6   \n",
       "1345     chr9_6008571                                        Melanocytes   \n",
       "1346    chr9_93940390                                            Bladder   \n",
       "\n",
       "      bool_is_asatac    chr      start       stop  \n",
       "0               True  chr10  104239100  104239101  \n",
       "1               True  chr10  104262628  104262629  \n",
       "2               True  chr10  104263675  104263676  \n",
       "3               True  chr10  104269217  104269218  \n",
       "4               True  chr10  104269301  104269302  \n",
       "...              ...    ...        ...        ...  \n",
       "1342            True   chr9   33817617   33817618  \n",
       "1343            True   chr9   34074476   34074477  \n",
       "1344            True   chr9   35538626   35538627  \n",
       "1345            True   chr9    6008571    6008572  \n",
       "1346            True   chr9   93940390   93940391  \n",
       "\n",
       "[1347 rows x 6 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "mpra_info_df[['Causal_SNP', 'rowname','disease']]\n",
    "asatac_lib_df#.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Airway\n",
      "2 5016 68 133\n",
      "hypergeometric test between being asATAC and MPRA hit in a tissue over background of being in mpra_dataset 0.5428690370111467\n",
      "Astrocytes\n",
      "0 5016 10 45\n",
      "hypergeometric test between being asATAC and MPRA hit in a tissue over background of being in mpra_dataset 1.0\n",
      "Colon\n",
      "1 5016 31 252\n",
      "hypergeometric test between being asATAC and MPRA hit in a tissue over background of being in mpra_dataset 0.7986728154382556\n",
      "Esophageal\n",
      "0 5016 4 267\n",
      "hypergeometric test between being asATAC and MPRA hit in a tissue over background of being in mpra_dataset 1.0\n",
      "GDSD6\n",
      "0 5016 19 93\n",
      "hypergeometric test between being asATAC and MPRA hit in a tissue over background of being in mpra_dataset 1.0\n",
      "GM12878\n",
      "0 5016 16 113\n",
      "hypergeometric test between being asATAC and MPRA hit in a tissue over background of being in mpra_dataset 1.0\n",
      "HMEC\n",
      "2 5016 138 96\n",
      "hypergeometric test between being asATAC and MPRA hit in a tissue over background of being in mpra_dataset 0.7477135595069265\n",
      "Melanocytes\n",
      "1 5016 23 412\n",
      "hypergeometric test between being asATAC and MPRA hit in a tissue over background of being in mpra_dataset 0.8613490918705736\n",
      "Ovarian\n",
      "0 5016 23 79\n",
      "hypergeometric test between being asATAC and MPRA hit in a tissue over background of being in mpra_dataset 1.0\n",
      "Pancreas\n",
      "1 5016 17 270\n",
      "hypergeometric test between being asATAC and MPRA hit in a tissue over background of being in mpra_dataset 0.6102177115690189\n",
      "Prostate\n",
      "0 5016 51 90\n",
      "hypergeometric test between being asATAC and MPRA hit in a tissue over background of being in mpra_dataset 1.0\n",
      "Renal\n",
      "0 5016 3 276\n",
      "hypergeometric test between being asATAC and MPRA hit in a tissue over background of being in mpra_dataset 1.0\n",
      "Thyroid\n",
      "1 5016 16 253\n",
      "hypergeometric test between being asATAC and MPRA hit in a tissue over background of being in mpra_dataset 0.5636692657237252\n",
      "Uterine\n",
      "1 5016 22 227\n",
      "hypergeometric test between being asATAC and MPRA hit in a tissue over background of being in mpra_dataset 0.6397781971291123\n"
     ]
    }
   ],
   "source": [
    "# testing for given tissue\n",
    "tissue_num_hits_dict = {}\n",
    "\n",
    "for tissue in sorted(mpra_tissue_mapper.keys()):\n",
    "    print(tissue)\n",
    "    mpra_tissue = mpra_tissue_mapper[tissue]\n",
    "    lib_tested_rownames = set(mpra_info_df[mpra_info_df.disease.str.contains(mpra_tissue)].rowname.values)\n",
    "    mpra_hit_rownames = set(mpra_res_df[mpra_res_df.mapped_tissue==tissue].rowname.values)\n",
    "    asatac_hit_rowanmes = set(asatac_lib_df[asatac_lib_df.tissue_asatac.str.contains(tissue)].rowname.values)\n",
    "\n",
    "\n",
    "    num_mpra_asatac_hits = len(mpra_hit_rownames.intersection(asatac_hit_rowanmes))\n",
    "    num_lib_tested =lib_df1.shape[0] #len(lib_tested_rownames)\n",
    "    num_mpra_hits = len(mpra_hit_rownames)\n",
    "    num_asatac_hits = len(asatac_hit_rowanmes)\n",
    "    print(num_mpra_asatac_hits,  num_lib_tested, num_mpra_hits, num_asatac_hits)\n",
    "    phyper =  hypergeometric_test(num_mpra_asatac_hits, num_lib_tested, num_mpra_hits, num_asatac_hits)\n",
    "    print('hypergeometric test between being asATAC and MPRA hit in a tissue over background of being in mpra_dataset', phyper )\n",
    "\n",
    "    tissue_num_hits_dict[tissue]= num_mpra_asatac_hits\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Airway': 2,\n",
       " 'Astrocytes': 0,\n",
       " 'Colon': 1,\n",
       " 'Esophageal': 0,\n",
       " 'GDSD6': 0,\n",
       " 'GM12878': 0,\n",
       " 'HMEC': 2,\n",
       " 'Melanocytes': 1,\n",
       " 'Ovarian': 0,\n",
       " 'Pancreas': 1,\n",
       " 'Prostate': 0,\n",
       " 'Renal': 0,\n",
       " 'Thyroid': 1,\n",
       " 'Uterine': 1}"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tissue_num_hits_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>rsid</th>\n",
       "      <th>rowname</th>\n",
       "      <th>Causal_SNP</th>\n",
       "      <th>disease</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>46505785</td>\n",
       "      <td>46505785</td>\n",
       "      <td>rs71062735</td>\n",
       "      <td>chr1_46505785</td>\n",
       "      <td>rs71062735</td>\n",
       "      <td>hmec</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>46358009</td>\n",
       "      <td>46358010</td>\n",
       "      <td>rs4539075</td>\n",
       "      <td>chr1_46358009</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>46361176</td>\n",
       "      <td>46361177</td>\n",
       "      <td>rs7512395</td>\n",
       "      <td>chr1_46361176</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>88423171</td>\n",
       "      <td>88423172</td>\n",
       "      <td>rs7514001</td>\n",
       "      <td>chr1_88423171</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>108368897</td>\n",
       "      <td>108368902</td>\n",
       "      <td>rs11278684</td>\n",
       "      <td>chr1_108368897</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5023</td>\n",
       "      <td>chr22</td>\n",
       "      <td>40436972</td>\n",
       "      <td>40436973</td>\n",
       "      <td>rs11704416</td>\n",
       "      <td>chr22_40436972</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5024</td>\n",
       "      <td>chr22</td>\n",
       "      <td>38568832</td>\n",
       "      <td>38568833</td>\n",
       "      <td>rs738321</td>\n",
       "      <td>chr22_38568832</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5025</td>\n",
       "      <td>chr22</td>\n",
       "      <td>30236242</td>\n",
       "      <td>30236243</td>\n",
       "      <td>rs2105870</td>\n",
       "      <td>chr22_30236242</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5026</td>\n",
       "      <td>chr22</td>\n",
       "      <td>40948874</td>\n",
       "      <td>40948875</td>\n",
       "      <td>rs55662398</td>\n",
       "      <td>chr22_40948874</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5027</td>\n",
       "      <td>chr22</td>\n",
       "      <td>40889907</td>\n",
       "      <td>40889908</td>\n",
       "      <td>rs77426923</td>\n",
       "      <td>chr22_40889907</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5028 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr      start       stop        rsid         rowname  Causal_SNP  \\\n",
       "0      chr1   46505785   46505785  rs71062735   chr1_46505785  rs71062735   \n",
       "1      chr1   46358009   46358010   rs4539075   chr1_46358009         NaN   \n",
       "2      chr1   46361176   46361177   rs7512395   chr1_46361176         NaN   \n",
       "3      chr1   88423171   88423172   rs7514001   chr1_88423171         NaN   \n",
       "4      chr1  108368897  108368902  rs11278684  chr1_108368897         NaN   \n",
       "...     ...        ...        ...         ...             ...         ...   \n",
       "5023  chr22   40436972   40436973  rs11704416  chr22_40436972         NaN   \n",
       "5024  chr22   38568832   38568833    rs738321  chr22_38568832         NaN   \n",
       "5025  chr22   30236242   30236243   rs2105870  chr22_30236242         NaN   \n",
       "5026  chr22   40948874   40948875  rs55662398  chr22_40948874         NaN   \n",
       "5027  chr22   40889907   40889908  rs77426923  chr22_40889907         NaN   \n",
       "\n",
       "     disease  \n",
       "0       hmec  \n",
       "1        NaN  \n",
       "2        NaN  \n",
       "3        NaN  \n",
       "4        NaN  \n",
       "...      ...  \n",
       "5023     NaN  \n",
       "5024     NaN  \n",
       "5025     NaN  \n",
       "5026     NaN  \n",
       "5027     NaN  \n",
       "\n",
       "[5028 rows x 7 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lib_df1.merge(mpra_info_df[['Causal_SNP', 'rowname','disease']],how='left',on='rowname')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Hichip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['../data/interim/merged/anchors_bed_sort/COLO_SCR_DMSO_sort.bed',\n",
       " '../data/interim/merged/anchors_bed_sort/Ovarian_sort.bed',\n",
       " '../data/interim/merged/anchors_bed_sort/GDSD3_sort.bed',\n",
       " '../data/interim/merged/anchors_bed_sort/COLO_shMITF_DMSO_sort.bed',\n",
       " '../data/interim/merged/anchors_bed_sort/Melanocytes_sort.bed',\n",
       " '../data/interim/merged/anchors_bed_sort/Pancreas_sort.bed',\n",
       " '../data/interim/merged/anchors_bed_sort/Colon_sort.bed',\n",
       " '../data/interim/merged/anchors_bed_sort/SCC13-CTRLi_sort.bed',\n",
       " '../data/interim/merged/anchors_bed_sort/CAL27-CTRLi_sort.bed',\n",
       " '../data/interim/merged/anchors_bed_sort/Thyroid_sort.bed',\n",
       " '../data/interim/merged/anchors_bed_sort/Uterine_sort.bed',\n",
       " '../data/interim/merged/anchors_bed_sort/D6-p63i_sort.bed',\n",
       " '../data/interim/merged/anchors_bed_sort/GM12878_sort.bed',\n",
       " '../data/interim/merged/anchors_bed_sort/HMEC_sort.bed',\n",
       " '../data/interim/merged/anchors_bed_sort/A431-p63i_sort.bed',\n",
       " '../data/interim/merged/anchors_bed_sort/Astrocytes_sort.bed',\n",
       " '../data/interim/merged/anchors_bed_sort/D3-p63i_sort.bed',\n",
       " '../data/interim/merged/anchors_bed_sort/D0-CTRLi_sort.bed',\n",
       " '../data/interim/merged/anchors_bed_sort/Airway_sort.bed',\n",
       " '../data/interim/merged/anchors_bed_sort/GDSD0_sort.bed',\n",
       " '../data/interim/merged/anchors_bed_sort/SCC13-p63i_sort.bed',\n",
       " '../data/interim/merged/anchors_bed_sort/D0-p63i_sort.bed',\n",
       " '../data/interim/merged/anchors_bed_sort/WM_SCR_DMSO_sort.bed',\n",
       " '../data/interim/merged/anchors_bed_sort/D6-CTRLi_sort.bed',\n",
       " '../data/interim/merged/anchors_bed_sort/WM_shMITF_PLX_sort.bed',\n",
       " '../data/interim/merged/anchors_bed_sort/GDSD6_sort.bed',\n",
       " '../data/interim/merged/anchors_bed_sort/A431-CTRLi_sort.bed',\n",
       " '../data/interim/merged/anchors_bed_sort/Renal_sort.bed',\n",
       " '../data/interim/merged/anchors_bed_sort/D3-CTRLi_sort.bed',\n",
       " '../data/interim/merged/anchors_bed_sort/WM_SCR_PLX_sort.bed',\n",
       " '../data/interim/merged/anchors_bed_sort/CAL27-p63i_sort.bed',\n",
       " '../data/interim/merged/anchors_bed_sort/Bladder_sort.bed',\n",
       " '../data/interim/merged/anchors_bed_sort/WM_shMITF_DMSO_sort.bed',\n",
       " '../data/interim/merged/anchors_bed_sort/Prostate_sort.bed',\n",
       " '../data/interim/merged/anchors_bed_sort/COLO_shMITF_PLX_sort.bed',\n",
       " '../data/interim/merged/anchors_bed_sort/COLO_SCR_PLX_sort.bed',\n",
       " '../data/interim/merged/anchors_bed_sort/Esophageal_sort.bed']"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hichip_peaks_dir = '../data/interim/merged/anchors_bed_sort'\n",
    "hichip_bed_files = glob.glob(os.path.join(hichip_peaks_dir, '*_sort.bed'))\n",
    "hichip_bed_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for hichip_file in sorted(hichip_bed_files):\n",
    "#     tissue = os.path.basename(hichip_file).split('_diffloop')[0]\n",
    "#     print(tissue)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A431-CTRLi\n",
      "A431-CTRLi not considered\n",
      "A431-p63i\n",
      "A431-p63i not considered\n",
      "Airway\n",
      "lib_hichip_df (7094, 2)\n",
      "Astrocytes\n",
      "lib_hichip_df (19811, 2)\n",
      "Bladder\n",
      "lib_hichip_df (32719, 2)\n",
      "CAL27-CTRLi\n",
      "CAL27-CTRLi not considered\n",
      "CAL27-p63i\n",
      "CAL27-p63i not considered\n",
      "COLO_SCR_DMSO\n",
      "COLO_SCR_DMSO not considered\n",
      "COLO_SCR_PLX\n",
      "COLO_SCR_PLX not considered\n",
      "COLO_shMITF_DMSO\n",
      "COLO_shMITF_DMSO not considered\n",
      "COLO_shMITF_PLX\n",
      "COLO_shMITF_PLX not considered\n",
      "Colon\n",
      "lib_hichip_df (48275, 2)\n",
      "D0-CTRLi\n",
      "D0-CTRLi not considered\n",
      "D0-p63i\n",
      "D0-p63i not considered\n",
      "D3-CTRLi\n",
      "D3-CTRLi not considered\n",
      "D3-p63i\n",
      "D3-p63i not considered\n",
      "D6-CTRLi\n",
      "D6-CTRLi not considered\n",
      "D6-p63i\n",
      "D6-p63i not considered\n",
      "Esophageal\n",
      "lib_hichip_df (62259, 2)\n",
      "GDSD0\n",
      "lib_hichip_df (75640, 2)\n",
      "GDSD3\n",
      "lib_hichip_df (89399, 2)\n",
      "GDSD6\n",
      "lib_hichip_df (102456, 2)\n",
      "GM12878\n",
      "lib_hichip_df (115908, 2)\n",
      "HMEC\n",
      "lib_hichip_df (126392, 2)\n",
      "Melanocytes\n",
      "lib_hichip_df (135862, 2)\n",
      "Ovarian\n",
      "lib_hichip_df (150199, 2)\n",
      "Pancreas\n",
      "lib_hichip_df (162783, 2)\n",
      "Prostate\n",
      "lib_hichip_df (179240, 2)\n",
      "Renal\n",
      "lib_hichip_df (195056, 2)\n",
      "SCC13-CTRLi\n",
      "SCC13-CTRLi not considered\n",
      "SCC13-p63i\n",
      "SCC13-p63i not considered\n",
      "Thyroid\n",
      "lib_hichip_df (209232, 2)\n",
      "Uterine\n",
      "lib_hichip_df (228791, 2)\n",
      "WM_SCR_DMSO\n",
      "WM_SCR_DMSO not considered\n",
      "WM_SCR_PLX\n",
      "WM_SCR_PLX not considered\n",
      "WM_shMITF_DMSO\n",
      "WM_shMITF_DMSO not considered\n",
      "WM_shMITF_PLX\n",
      "WM_shMITF_PLX not considered\n"
     ]
    }
   ],
   "source": [
    "# aside get a table which lists for all the snps whether it's in a hichip anchor peak and which tissues\n",
    "tissue_mapper = atac_tissue_mapper\n",
    "\n",
    "lib_hichip_df = pd.DataFrame()\n",
    "for hichip_file in sorted(hichip_bed_files):# sorted(hichip_bed_files):\n",
    "    tissue = os.path.basename(hichip_file).split('_sort')[0]\n",
    "    print(tissue)\n",
    "    if tissue not in tissue_mapper.keys():\n",
    "        print(tissue, 'not considered')\n",
    "        continue\n",
    "    hichip_bed = pybedtools.BedTool(hichip_file).to_dataframe()\n",
    "    if type(hichip_bed['chrom'][0])==int:\n",
    "        hichip_bed['chrom'] = 'chr' + hichip_bed.chrom.map(str)\n",
    "        hichip_bed = pybedtools.BedTool.from_dataframe(hichip_bed)\n",
    "    else:\n",
    "        hichip_bed = pybedtools.BedTool(hichip_file)\n",
    "    lib_df_bed_hichip = lib_df_bed.intersect(hichip_bed).to_dataframe()\n",
    "    lib_df_bed_hichip  =lib_df_bed_hichip[['name']].drop_duplicates().reset_index(drop=True)\n",
    "    lib_df_bed_hichip['tissue'] = tissue_mapper[tissue]\n",
    "    lib_hichip_df = pd.concat([lib_hichip_df, lib_df_bed_hichip])\n",
    "    print('lib_hichip_df', lib_hichip_df.shape)\n",
    "    \n",
    "lib_hichip_df = lib_hichip_df.groupby('name').agg({'tissue':'|'.join}).reset_index()\n",
    "lib_hichip_df.columns = ['rowname','hichip_tissues']\n",
    "lib_hichip_df['bool_in_hichip_pk'] = True\n",
    "lib_hichip_df.to_csv(os.path.join(save_dir, 'lib_hichip_annotation.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['../data/processed/vcf_files/hichip/GDSD3_postvqsr.vcf',\n",
       " '../data/processed/vcf_files/hichip/GM12878_postvqsr.vcf',\n",
       " '../data/processed/vcf_files/hichip/Airway_postvqsr.vcf',\n",
       " '../data/processed/vcf_files/hichip/Bladder_postvqsr.vcf',\n",
       " '../data/processed/vcf_files/hichip/Astrocytes_postvqsr.vcf',\n",
       " '../data/processed/vcf_files/hichip/Esophageal_postvqsr.vcf',\n",
       " '../data/processed/vcf_files/hichip/Ovarian_postvqsr.vcf',\n",
       " '../data/processed/vcf_files/hichip/Melanocytes_postvqsr.vcf',\n",
       " '../data/processed/vcf_files/hichip/Colon_postvqsr.vcf',\n",
       " '../data/processed/vcf_files/hichip/Thyroid_postvqsr.vcf',\n",
       " '../data/processed/vcf_files/hichip/HMEC_postvqsr.vcf',\n",
       " '../data/processed/vcf_files/hichip/Uterine_postvqsr.vcf',\n",
       " '../data/processed/vcf_files/hichip/GDSD0_postvqsr.vcf',\n",
       " '../data/processed/vcf_files/hichip/GDSD6_postvqsr.vcf',\n",
       " '../data/processed/vcf_files/hichip/Renal_postvqsr.vcf',\n",
       " '../data/processed/vcf_files/hichip/Pancreas_postvqsr.vcf',\n",
       " '../data/processed/vcf_files/hichip/Prostate_postvqsr.vcf']"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcf_files_hichip = glob.glob('../data/processed/vcf_files/hichip/*vcf')\n",
    "vcf_files_hichip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GDSD3\n",
      "GM12878\n",
      "Airway\n",
      "Bladder\n",
      "Astrocytes\n",
      "Esophageal\n",
      "Ovarian\n",
      "Melanocytes\n",
      "Colon\n",
      "Thyroid\n",
      "HMEC\n",
      "Uterine\n",
      "GDSD0\n",
      "GDSD6\n",
      "Renal\n",
      "Pancreas\n",
      "Prostate\n"
     ]
    }
   ],
   "source": [
    "for vcf_file in vcf_files_hichip:\n",
    "    tissue = os.path.basename(vcf_file).split('_post')[0]\n",
    "    print(tissue)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "hichip_tissue_mapper =atac_tissue_mapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "************\n",
      "tissue Airway\n",
      "read ../data/processed/vcf_files/hichip/Airway_postvqsr.vcf\n",
      "shape (1173693, 12)\n",
      "num_vcf_hits PRE FILTER 1173693\n",
      "num_vcf_hits POST FILTER 110841\n",
      "num_vcf_hits POST FDR 1431\n",
      "background library hits that are asHiChIP 5016 12 0.0023923444976076554\n",
      "hichip filter 702 out of 1431 allele-specific hichip 0.49056603773584906\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:31: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hichip filter 7094 out of 5016 mpra alleles tested 1.414274322169059\n",
      "background library hits-hichip filt 7094 4 0.000563856780377784\n",
      "************\n",
      "tissue Astrocytes\n",
      "read ../data/processed/vcf_files/hichip/Astrocytes_postvqsr.vcf\n",
      "shape (1348725, 12)\n",
      "num_vcf_hits PRE FILTER 1348725\n",
      "num_vcf_hits POST FILTER 135366\n",
      "num_vcf_hits POST FDR 3697\n",
      "background library hits that are asHiChIP 5016 19 0.003787878787878788\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:31: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:34: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:37: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hichip filter 2270 out of 3697 allele-specific hichip 0.6140113605626183\n",
      "hichip filter 12717 out of 5016 mpra alleles tested 2.535287081339713\n",
      "background library hits-hichip filt 12717 13 0.0010222536761814894\n",
      "************\n",
      "tissue Bladder\n",
      "read ../data/processed/vcf_files/hichip/Bladder_postvqsr.vcf\n",
      "shape (1253005, 12)\n",
      "num_vcf_hits PRE FILTER 1253005\n",
      "num_vcf_hits POST FILTER 518064\n",
      "num_vcf_hits POST FDR 11810\n",
      "background library hits that are asHiChIP 5016 50 0.009968102073365232\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:31: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:34: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:37: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hichip filter 7576 out of 11810 allele-specific hichip 0.6414902624894158\n",
      "hichip filter 12908 out of 5016 mpra alleles tested 2.573365231259968\n",
      "background library hits-hichip filt 12908 45 0.003486210102262163\n",
      "************\n",
      "tissue Colon\n",
      "read ../data/processed/vcf_files/hichip/Colon_postvqsr.vcf\n",
      "shape (953402, 12)\n",
      "num_vcf_hits PRE FILTER 953402\n",
      "num_vcf_hits POST FILTER 310610\n",
      "num_vcf_hits POST FDR 81574\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:31: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:34: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:37: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "background library hits that are asHiChIP 5016 667 0.1329744816586922\n",
      "hichip filter 42577 out of 81574 allele-specific hichip 0.5219432662367911\n",
      "hichip filter 15556 out of 5016 mpra alleles tested 3.101275917065391\n",
      "background library hits-hichip filt 15556 470 0.03021342247364361\n",
      "************\n",
      "tissue Esophageal\n",
      "read ../data/processed/vcf_files/hichip/Esophageal_postvqsr.vcf\n",
      "shape (866052, 12)\n",
      "num_vcf_hits PRE FILTER 866052\n",
      "num_vcf_hits POST FILTER 211493\n",
      "num_vcf_hits POST FDR 58226\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:31: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:34: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:37: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "background library hits that are asHiChIP 5016 447 0.08911483253588516\n",
      "hichip filter 29118 out of 58226 allele-specific hichip 0.5000858722907292\n",
      "hichip filter 13984 out of 5016 mpra alleles tested 2.787878787878788\n",
      "background library hits-hichip filt 13984 280 0.02002288329519451\n",
      "************\n",
      "tissue GDSD0\n",
      "read ../data/processed/vcf_files/hichip/GDSD0_postvqsr.vcf\n",
      "shape (1308563, 12)\n",
      "num_vcf_hits PRE FILTER 1308563\n",
      "num_vcf_hits POST FILTER 148582\n",
      "num_vcf_hits POST FDR 8920\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:31: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:34: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:37: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "background library hits that are asHiChIP 5016 38 0.007575757575757576\n",
      "hichip filter 4821 out of 8920 allele-specific hichip 0.5404708520179372\n",
      "hichip filter 13381 out of 5016 mpra alleles tested 2.667663476874003\n",
      "background library hits-hichip filt 13381 19 0.0014199237725132652\n",
      "************\n",
      "tissue GDSD3\n",
      "read ../data/processed/vcf_files/hichip/GDSD3_postvqsr.vcf\n",
      "shape (1276709, 12)\n",
      "num_vcf_hits PRE FILTER 1276709\n",
      "num_vcf_hits POST FILTER 140246\n",
      "num_vcf_hits POST FDR 8306\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:31: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:34: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:37: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "background library hits that are asHiChIP 5016 28 0.005582137161084529\n",
      "hichip filter 4638 out of 8306 allele-specific hichip 0.558391524199374\n",
      "hichip filter 13759 out of 5016 mpra alleles tested 2.743022328548644\n",
      "background library hits-hichip filt 13759 17 0.0012355549095137727\n",
      "************\n",
      "tissue GDSD6\n",
      "read ../data/processed/vcf_files/hichip/GDSD6_postvqsr.vcf\n",
      "shape (1247380, 12)\n",
      "num_vcf_hits PRE FILTER 1247380\n",
      "num_vcf_hits POST FILTER 133980\n",
      "num_vcf_hits POST FDR 6080\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:31: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:34: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:37: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "background library hits that are asHiChIP 5016 26 0.00518341307814992\n",
      "hichip filter 3527 out of 6080 allele-specific hichip 0.5800986842105263\n",
      "hichip filter 13057 out of 5016 mpra alleles tested 2.6030701754385963\n",
      "background library hits-hichip filt 13057 17 0.0013019836103239642\n",
      "************\n",
      "tissue GM12878\n",
      "read ../data/processed/vcf_files/hichip/GM12878_postvqsr.vcf\n",
      "shape (1015175, 12)\n",
      "num_vcf_hits PRE FILTER 1015175\n",
      "num_vcf_hits POST FILTER 394674\n",
      "num_vcf_hits POST FDR 14080\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:31: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:34: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:37: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "background library hits that are asHiChIP 5016 102 0.02033492822966507\n",
      "hichip filter 7859 out of 14080 allele-specific hichip 0.5581676136363637\n",
      "hichip filter 13452 out of 5016 mpra alleles tested 2.6818181818181817\n",
      "background library hits-hichip filt 13452 62 0.004608980077311924\n",
      "************\n",
      "tissue HMEC\n",
      "read ../data/processed/vcf_files/hichip/HMEC_postvqsr.vcf\n",
      "shape (1300369, 12)\n",
      "num_vcf_hits PRE FILTER 1300369\n",
      "num_vcf_hits POST FILTER 145379\n",
      "num_vcf_hits POST FDR 3849\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:31: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:34: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:37: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "background library hits that are asHiChIP 5016 16 0.003189792663476874\n",
      "hichip filter 1906 out of 3849 allele-specific hichip 0.4951935567679917\n",
      "hichip filter 10484 out of 5016 mpra alleles tested 2.0901116427432216\n",
      "background library hits-hichip filt 10484 12 0.0011446012972148034\n",
      "************\n",
      "tissue Melanocytes\n",
      "read ../data/processed/vcf_files/hichip/Melanocytes_postvqsr.vcf\n",
      "shape (1314344, 12)\n",
      "num_vcf_hits PRE FILTER 1314344\n",
      "num_vcf_hits POST FILTER 264814\n",
      "num_vcf_hits POST FDR 85511\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:31: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:34: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:37: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "background library hits that are asHiChIP 5016 789 0.15729665071770335\n",
      "hichip filter 27448 out of 85511 allele-specific hichip 0.3209879430716516\n",
      "hichip filter 9470 out of 5016 mpra alleles tested 1.8879585326953747\n",
      "background library hits-hichip filt 9470 320 0.0337909186906019\n",
      "************\n",
      "tissue Ovarian\n",
      "read ../data/processed/vcf_files/hichip/Ovarian_postvqsr.vcf\n",
      "shape (885465, 12)\n",
      "num_vcf_hits PRE FILTER 885465\n",
      "num_vcf_hits POST FILTER 255422\n",
      "num_vcf_hits POST FDR 55842\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:31: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:34: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:37: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "background library hits that are asHiChIP 5016 560 0.11164274322169059\n",
      "hichip filter 28052 out of 55842 allele-specific hichip 0.5023459045163139\n",
      "hichip filter 14337 out of 5016 mpra alleles tested 2.8582535885167464\n",
      "background library hits-hichip filt 14337 367 0.025598102810908838\n",
      "************\n",
      "tissue Pancreas\n",
      "read ../data/processed/vcf_files/hichip/Pancreas_postvqsr.vcf\n",
      "shape (999275, 12)\n",
      "num_vcf_hits PRE FILTER 999275\n",
      "num_vcf_hits POST FILTER 208362\n",
      "num_vcf_hits POST FDR 54619\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:31: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:34: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:37: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "background library hits that are asHiChIP 5016 457 0.09110845295055821\n",
      "hichip filter 25437 out of 54619 allele-specific hichip 0.46571705816657205\n",
      "hichip filter 12584 out of 5016 mpra alleles tested 2.508771929824561\n",
      "background library hits-hichip filt 12584 283 0.022488874761602034\n",
      "************\n",
      "tissue Prostate\n",
      "read ../data/processed/vcf_files/hichip/Prostate_postvqsr.vcf\n",
      "shape (1060235, 12)\n",
      "num_vcf_hits PRE FILTER 1060235\n",
      "num_vcf_hits POST FILTER 441257\n",
      "num_vcf_hits POST FDR 15885\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:31: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:34: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:37: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "background library hits that are asHiChIP 5016 69 0.01375598086124402\n",
      "hichip filter 10906 out of 15885 allele-specific hichip 0.686559647466163\n",
      "hichip filter 16457 out of 5016 mpra alleles tested 3.280901116427432\n",
      "background library hits-hichip filt 16457 57 0.0034635717323935103\n",
      "************\n",
      "tissue Renal\n",
      "read ../data/processed/vcf_files/hichip/Renal_postvqsr.vcf\n",
      "shape (1188479, 12)\n",
      "num_vcf_hits PRE FILTER 1188479\n",
      "num_vcf_hits POST FILTER 93193\n",
      "num_vcf_hits POST FDR 6535\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:31: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:34: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:37: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "background library hits that are asHiChIP 5016 48 0.009569377990430622\n",
      "hichip filter 4173 out of 6535 allele-specific hichip 0.6385615914307574\n",
      "hichip filter 15816 out of 5016 mpra alleles tested 3.15311004784689\n",
      "background library hits-hichip filt 15816 35 0.002212948912493677\n",
      "************\n",
      "tissue Thyroid\n",
      "read ../data/processed/vcf_files/hichip/Thyroid_postvqsr.vcf\n",
      "shape (983295, 12)\n",
      "num_vcf_hits PRE FILTER 983295\n",
      "num_vcf_hits POST FILTER 313153\n",
      "num_vcf_hits POST FDR 74753\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:31: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:34: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:37: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "background library hits that are asHiChIP 5016 645 0.12858851674641147\n",
      "hichip filter 34351 out of 74753 allele-specific hichip 0.45952670795820905\n",
      "hichip filter 14176 out of 5016 mpra alleles tested 2.8261562998405103\n",
      "background library hits-hichip filt 14176 400 0.028216704288939052\n",
      "************\n",
      "tissue Uterine\n",
      "read ../data/processed/vcf_files/hichip/Uterine_postvqsr.vcf\n",
      "shape (1248162, 12)\n",
      "num_vcf_hits PRE FILTER 1248162\n",
      "num_vcf_hits POST FILTER 110783\n",
      "num_vcf_hits POST FDR 8598\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:31: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:34: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:37: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "background library hits that are asHiChIP 5016 44 0.008771929824561403\n",
      "hichip filter 6865 out of 8598 allele-specific hichip 0.798441498022796\n",
      "hichip filter 19559 out of 5016 mpra alleles tested 3.899322169059011\n",
      "background library hits-hichip filt 19559 37 0.0018917122552277723\n",
      "CPU times: user 58min 22s, sys: 23.4 s, total: 58min 45s\n",
      "Wall time: 58min\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "ashichip_lib_df = pd.DataFrame()\n",
    "vcf_df_all_hichip = pd.DataFrame()\n",
    "vcf_bed_df_all_hichip = pd.DataFrame()\n",
    "vcf_df_lib_all_hichip = pd.DataFrame()\n",
    "vcf_df_lib_hichip_all_hichip = pd.DataFrame()\n",
    "for vcf_file in sorted(vcf_files_hichip):\n",
    "    tissue = os.path.basename(vcf_file).split('_post')[0]\n",
    "    print('************')\n",
    "    print('tissue', tissue)\n",
    "    \n",
    "   # read vcf\n",
    "    vcf_df = read_vcf(vcf_file)\n",
    "    num_vcf_hits = vcf_df.shape[0]\n",
    "    print('num_vcf_hits PRE FILTER', num_vcf_hits)  \n",
    "    \n",
    "    # filter vcf for those that meet the heterogenous requiremnts\n",
    "    vcf_df['binom_p'] = vcf_df.apply(lambda row: preprocess_and_filter(row,tissue), axis=1)\n",
    "    vcf_df = vcf_df[vcf_df.binom_p>=0]\n",
    "    vcf_df['tissue'] = tissue\n",
    "    num_vcf_hits_postfilt = vcf_df.shape[0]\n",
    "    print('num_vcf_hits POST FILTER', num_vcf_hits_postfilt)      \n",
    "    \n",
    "    \n",
    "    # binomial test\n",
    "    vcf_df['pass_filt'], vcf_df['p_adjust'] = fdrcorrection(vcf_df.binom_p, alpha=0.05)\n",
    "    vcf_df = vcf_df[vcf_df.pass_filt]\n",
    "    num_vcf_hits_passfdr = vcf_df.shape[0]\n",
    "    print('num_vcf_hits POST FDR', num_vcf_hits_passfdr)      \n",
    "\n",
    "    vcf_bed_df = vcf_df[['CHROM','POS','rowname']]\n",
    "    vcf_bed_df['stop'] = vcf_df.POS.map(int) + 1\n",
    "    vcf_bed_df = vcf_bed_df[['CHROM','POS','stop','rowname']]\n",
    "    vcf_bed_df.columns = ['chr','start','stop','name']\n",
    "    vcf_df_all_hichip = pd.concat([vcf_df_all_hichip, vcf_df])\n",
    "\n",
    "    vcf_df_lib = vcf_df.merge(lib_df_bed_df, how='inner',on='rowname')\n",
    "    vcf_df_lib_all_hichip = pd.concat([vcf_df_lib_all_hichip, vcf_df_lib])\n",
    "    lib_hits = vcf_df_lib.rowname.unique()\n",
    "    num_lib_hits = lib_hits.shape[0]\n",
    "    print('background library hits that are asHiChIP', num_lib_tested,num_lib_hits, num_lib_hits/ num_lib_tested)\n",
    "    \n",
    "#     if tissue in mpra_tissue_mapper:\n",
    "#         mpra_res_df_tissue = mpra_res_df[mpra_res_df.tissue==mpra_tissue_mapper[tissue]]\n",
    "#         mpra_hits_tissue = mpra_res_df_tissue.rowname.unique()\n",
    "#         num_mpra_hits_tissue= mpra_hits_tissue.shape[0]\n",
    "# #         mpra_sig_df_lib = vcf_df.merge(mpra_res_df_tissue, how='inner',on='rowname')\n",
    "\n",
    "#         num_mpra_ashichip_hits = len(set(lib_hits).intersection(set(mpra_hits_tissue)))\n",
    "#         ashichip_lib_df_tissue=pd.DataFrame()\n",
    "#         ashichip_lib_df_tissue['rowname'] = sorted(set(lib_hits).intersection(set(mpra_hits_tissue)))\n",
    "#         ashichip_lib_df_tissue['tissue'] =  mpra_tissue_mapper[tissue]\n",
    "#         ashichip_lib_df = pd.concat([ashichip_lib_df, ashichip_lib_df_tissue])\n",
    "#         # do a hypergeometric test between being asHiChIP and being an mpra hit over a background of being in the mpra dataset\n",
    "    \n",
    "#         print('mpra sig library hits in tissue ', num_mpra_hits_tissue)\n",
    "#         print('overlap:', num_mpra_ashichip_hits)\n",
    "# #         oddsratio, pvalue = stats.fisher_exact([[num_lib_hits, num_lib_tested- num_lib_hits],\n",
    "# #                                                 [num_mpra_hits, num_mpra_sig_hits - num_mpra_hits]])\n",
    "# #         print('fisher for mpra and as annotation association', pvalue, oddsratio)\n",
    "#         phyper =  hypergeometric_test(num_mpra_ashichip_hits, num_lib_tested, num_mpra_hits_tissue, num_lib_hits)\n",
    "#         #stats.hypergeom.cdf(num_mpra_ashichip_hits, num_lib_tested, num_lib_hits, num_mpra_hits)\n",
    "#         print('hypergeometric test between being asHiChIP and MPRA hit over backgroun of being in mpra_dataset', phyper )\n",
    "#         phyper =  hypergeometric_test(num_mpra_ashichip_hits, num_vcf_hits, num_mpra_hits_tissue, num_lib_hits)\n",
    "#         print('hypergeometric test between being asHiChIP and MPRA hit over backgroun of being in an ashichip', phyper )\n",
    "    \n",
    "    ## filter through hichip\n",
    "    if  tissue not in hichip_tissue_mapper:\n",
    "        print(tissue,'not in hichip tissue mapper')\n",
    "        continue\n",
    "    hichip_tissue = hichip_tissue_mapper[tissue]\n",
    "    hichip_bed_file = os.path.join(hichip_peaks_dir, hichip_tissue+'_sort.bed')\n",
    "    if not os.path.exists(hichip_bed_file):\n",
    "        print('hichip file',hichip_bed_file, 'does not exist')\n",
    "        continue\n",
    "    hichip_bed = pybedtools.BedTool(hichip_bed_file).to_dataframe()\n",
    "    if type(hichip_bed['chrom'][0])==int:\n",
    "        hichip_bed['chrom'] = 'chr' + hichip_bed.chrom.map(str)\n",
    "        hichip_bed = pybedtools.BedTool.from_dataframe(hichip_bed)\n",
    "    else:\n",
    "        hichip_bed = pybedtools.BedTool(hichip_bed_file)\n",
    "        \n",
    "    vcf_bed_df_hichip = pybedtools.BedTool.from_dataframe(vcf_bed_df).intersect(hichip_bed).to_dataframe()\n",
    "    num_hichip_vcf = vcf_bed_df_hichip.name.unique().shape[0]\n",
    "    vcf_bed_df_hichip['tissue'] = tissue\n",
    "    vcf_bed_df_all_hichip = pd.concat([vcf_bed_df_all_hichip,vcf_bed_df_hichip])\n",
    "    print('hichip filter', num_hichip_vcf, 'out of', vcf_bed_df.shape[0], 'allele-specific hichip', num_hichip_vcf/vcf_bed_df.shape[0])\n",
    "    lib_df_bed_hichip = lib_df_bed.intersect(hichip_bed).to_dataframe()\n",
    "    lib_hichip  = lib_df_bed_hichip.name.unique()\n",
    "    num_lib_hichip = lib_hichip.shape[0]\n",
    "    print('hichip filter', num_lib_hichip, 'out of', num_lib_tested, 'mpra alleles tested', num_lib_hichip/num_lib_tested)\n",
    "#     mpra_hichip = lib_df_bed_hichip[lib_df_bed_hichip.name.isin(mpra_rownames_sig)].name.unique()\n",
    "#     num_mpra_hichip = mpra_hichip.shape[0]\n",
    "#     print('hichip filter', num_mpra_hichip, 'out of', num_mpra_sig_hits, 'mpra alleles sig',num_mpra_hichip/num_mpra_sig_hits)\n",
    "    \n",
    "    vcf_df_lib_hichip = vcf_bed_df_hichip.merge(lib_df_bed_df, how='inner',left_on='name',right_on='rowname')\n",
    "    num_lib_hits_hichip = vcf_df_lib_hichip.rowname.unique().shape[0]\n",
    "    vcf_df_lib_hichip_all_hichip = pd.concat([vcf_df_lib_hichip_all_hichip, vcf_df_lib_hichip])\n",
    "#     mpra_sig_df_lib_hichip = vcf_bed_df_hichip.merge(mpra_res_df, how='inner',left_on='name',right_on='rowname')\n",
    "#     num_mpra_hits_hichip = mpra_sig_df_lib_hichip.rowname.unique().shape[0]\n",
    "    \n",
    "    print('background library hits-hichip filt', num_lib_hichip,num_lib_hits_hichip, num_lib_hits_hichip/ num_lib_hichip)\n",
    "#     print('mpra sig library hits-hichip filt', num_mpra_hichip,num_mpra_hits_hichip, num_mpra_hits_hichip/ num_mpra_hichip)\n",
    "#     if tissue in mpra_tissue_mapper:\n",
    "#         mpra_hits_tissue_hichip = set(mpra_hits_tissue).intersection(set(lib_hichip))\n",
    "#         num_mpra_hits_tissue_hichip= mpra_hits_tissue.shape[0]\n",
    "# #         mpra_sig_df_lib = vcf_df.merge(mpra_res_df_tissue, how='inner',on='rowname')\n",
    "\n",
    "#         num_mpra_ashichip_filthichip_hits = len(set(lib_hits).intersection(set(mpra_hits_tissue_hichip)))\n",
    "#         # do a hypergeometric test between being asHiChIP and being an mpra hit over a background of being in the mpra dataset\n",
    "\n",
    "#         print('mpra sig library hits in tissue hichip filt ', num_mpra_hits_tissue_hichip)\n",
    "#         print('overlap hichip:', num_mpra_ashichip_filthichip_hits)\n",
    "# #         oddsratio, pvalue = stats.fisher_exact([[num_lib_hits, num_lib_tested- num_lib_hits],\n",
    "# #                                                 [num_mpra_hits, num_mpra_sig_hits - num_mpra_hits]])\n",
    "# #         print('fisher for mpra and as annotation association', pvalue, oddsratio)\n",
    "#         phyper =  hypergeometric_test(num_mpra_ashichip_filthichip_hits, num_hichip_vcf, num_mpra_hits_tissue_hichip, num_lib_hichip)\n",
    "#         #stats.hypergeom.cdf(num_mpra_ashichip_hits, num_lib_tested, num_lib_hits, num_mpra_hits)\n",
    "#         print('hypergeometric test between being asHiChIP and MPRA hit over backgroun of being ashichip', phyper )\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(499716, 48)\n",
      "(242226, 5)\n"
     ]
    }
   ],
   "source": [
    "print(vcf_df_all_hichip.shape)\n",
    "print(vcf_bed_df_all_hichip.shape)\n",
    "vcf_df_all_hichip.to_csv(os.path.join(save_dir, 'vcf_df_all_hichip.csv'))\n",
    "vcf_bed_df_all_hichip.to_csv(os.path.join(save_dir, 'vcf_bed_df_all_hichip.csv'))\n",
    "vcf_df_lib_all_hichip.to_csv(os.path.join(save_dir, 'vcf_df_lib_all_hichip.csv'))\n",
    "vcf_df_lib_hichip_all_hichip.to_csv(os.path.join(save_dir, 'vcf_df_lib_hichip_all_hichip.csv'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "ashichip_lib_df = vcf_df_lib_hichip_all_hichip.merge(vcf_df_all[['rowname','tissue']],how='left',on=['rowname','tissue']).drop_duplicates().sort_values('tissue').groupby('rowname').agg({'tissue':'|'.join}).reset_index()\n",
    "ashichip_lib_df.columns = ['rowname','tissue_ashichip']\n",
    "ashichip_lib_df['bool_is_ashichip'] = True\n",
    "ashichip_lib_df= ashichip_lib_df.merge(lib_df_bed_df, how='left',on='rowname')\n",
    "ashichip_lib_df.to_csv(os.path.join(save_dir,'lib_ashichip_annotation_post_filt.csv'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rowname</th>\n",
       "      <th>tissue_ashichip</th>\n",
       "      <th>bool_is_ashichip</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>chr1_101328083</td>\n",
       "      <td>Colon|Esophageal|Ovarian</td>\n",
       "      <td>True</td>\n",
       "      <td>chr1</td>\n",
       "      <td>101328083</td>\n",
       "      <td>101328084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>chr1_101328230</td>\n",
       "      <td>Thyroid</td>\n",
       "      <td>True</td>\n",
       "      <td>chr1</td>\n",
       "      <td>101328230</td>\n",
       "      <td>101328231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>chr1_101390090</td>\n",
       "      <td>Melanocytes</td>\n",
       "      <td>True</td>\n",
       "      <td>chr1</td>\n",
       "      <td>101390090</td>\n",
       "      <td>101390091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>chr1_101393409</td>\n",
       "      <td>Bladder|Colon|Esophageal|Ovarian|Thyroid</td>\n",
       "      <td>True</td>\n",
       "      <td>chr1</td>\n",
       "      <td>101393409</td>\n",
       "      <td>101393410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>chr1_101466054</td>\n",
       "      <td>Esophageal|Thyroid</td>\n",
       "      <td>True</td>\n",
       "      <td>chr1</td>\n",
       "      <td>101466054</td>\n",
       "      <td>101466055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1243</td>\n",
       "      <td>chr9_22112241</td>\n",
       "      <td>Melanocytes</td>\n",
       "      <td>True</td>\n",
       "      <td>chr9</td>\n",
       "      <td>22112241</td>\n",
       "      <td>22112242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1244</td>\n",
       "      <td>chr9_35519359</td>\n",
       "      <td>Melanocytes</td>\n",
       "      <td>True</td>\n",
       "      <td>chr9</td>\n",
       "      <td>35519359</td>\n",
       "      <td>35519360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1245</td>\n",
       "      <td>chr9_73863654</td>\n",
       "      <td>Melanocytes</td>\n",
       "      <td>True</td>\n",
       "      <td>chr9</td>\n",
       "      <td>73863654</td>\n",
       "      <td>73863655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1246</td>\n",
       "      <td>chr9_93958422</td>\n",
       "      <td>Melanocytes</td>\n",
       "      <td>True</td>\n",
       "      <td>chr9</td>\n",
       "      <td>93958422</td>\n",
       "      <td>93958423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1247</td>\n",
       "      <td>chr9_96327691</td>\n",
       "      <td>Melanocytes</td>\n",
       "      <td>True</td>\n",
       "      <td>chr9</td>\n",
       "      <td>96327691</td>\n",
       "      <td>96327692</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1248 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             rowname                           tissue_ashichip  \\\n",
       "0     chr1_101328083                  Colon|Esophageal|Ovarian   \n",
       "1     chr1_101328230                                   Thyroid   \n",
       "2     chr1_101390090                               Melanocytes   \n",
       "3     chr1_101393409  Bladder|Colon|Esophageal|Ovarian|Thyroid   \n",
       "4     chr1_101466054                        Esophageal|Thyroid   \n",
       "...              ...                                       ...   \n",
       "1243   chr9_22112241                               Melanocytes   \n",
       "1244   chr9_35519359                               Melanocytes   \n",
       "1245   chr9_73863654                               Melanocytes   \n",
       "1246   chr9_93958422                               Melanocytes   \n",
       "1247   chr9_96327691                               Melanocytes   \n",
       "\n",
       "      bool_is_ashichip   chr      start       stop  \n",
       "0                 True  chr1  101328083  101328084  \n",
       "1                 True  chr1  101328230  101328231  \n",
       "2                 True  chr1  101390090  101390091  \n",
       "3                 True  chr1  101393409  101393410  \n",
       "4                 True  chr1  101466054  101466055  \n",
       "...                ...   ...        ...        ...  \n",
       "1243              True  chr9   22112241   22112242  \n",
       "1244              True  chr9   35519359   35519360  \n",
       "1245              True  chr9   73863654   73863655  \n",
       "1246              True  chr9   93958422   93958423  \n",
       "1247              True  chr9   96327691   96327692  \n",
       "\n",
       "[1248 rows x 6 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ashichip_lib_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
