{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# preprocess_annon_file\n",
    "\n",
    "for pan_omics project\n",
    "\n",
    "04/09/2020"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import subprocess\n",
    "import seaborn as sns\n",
    "import pybedtools\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For human, mouse, zebrafish, rat and pig, Ensembl not only shows transcripts that are annotated automatically using the Ensembl genebuild pipeline, but also transcripts that are manually annotated by the HAVANA team. If the Ensembl and Havana annotation agree with each other the transcripts are combined into an Ensembl/Havana merged transcript. When a transcript is only annotated by Ensembl or Havana it is named an Ensembl or Havana transcript, respectively. Transcripts that do match a species-specific entry in the UniProtKB/Swiss-Prot or RefSeq databases are categorised as known, those that do not as categorised as novel. For more detailed information, please have a look at our genebuild documentation.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "tss_file = '../data/external/TSS_annon_ALL.csv'\n",
    "# gencode_file = '../data/external/gene_annon_ALL.csv'\n",
    "gene_annon_file = '../data/external/gencode.v19.annotation.gtf'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 11.4 s, sys: 725 ms, total: 12.1 s\n",
      "Wall time: 10.7 s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(57820, 14)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "gene_annon_df = pd.read_csv(gene_annon_file, header=None, skiprows=5,sep='\\t')\n",
    "gene_annon_df.columns = ['chr','annotation_source', 'type', 'start','end','qual','strand', 'misc', 'gene_info']\n",
    "gene_annon_df=gene_annon_df[gene_annon_df.type=='gene']\n",
    "#this line takes awhile\n",
    "gene_annon_df[['ensg_id','enst_id','gene_type','status','gene_sym']] = gene_annon_df.gene_info.str.split('[ \";]',expand=True).loc[:,[2,7,12,17,22]]\n",
    "gene_annon_df.shape#(57820, 14)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(55765, 6)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_annon_df_file6.drop_duplicates('gene_sym').shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# bed6 format\n",
    "gene_annon_df_file6 = gene_annon_df[['chr','start','end','gene_sym', 'qual','strand']]\n",
    "gene_annon_df_file6.drop_duplicates('gene_sym').to_csv('../data/external/gencode.v19_bed6.gene.bed',sep='\\t',header=None,index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "# bed4 format\n",
    "gene_annon_df_file = gene_annon_df[['chr','start','end','gene_sym']]\n",
    "gene_annon_df_file.to_csv('../data/external/gencode.v19.gene.bed',sep='\\t',header=None,index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "exon_annon_df = pd.read_csv(gene_annon_file, header=None, skiprows=5,sep='\\t')\n",
    "exon_annon_df.columns = ['chr','annotation_source', 'type', 'start','end','qual','strand', 'misc', 'gene_info']\n",
    "exon_annon_df=exon_annon_df[exon_annon_df.type=='exon']\n",
    "#this line takes awhile\n",
    "exon_annon_df[['ensg_id','enst_id','gene_type','status','gene_sym']] = exon_annon_df.gene_info.str.split('[ \";]',expand=True).loc[:,[2,7,12,17,22]]#(1196293, 14)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "exon_annon_df_filt = exon_annon_df[['chr','start','end','gene_sym']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "exon_annon_df_filt.to_csv('../data/external/gencode.v19.exon.bed',sep='\\t',header=None,index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1196293, 14)"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_annon_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>annotation_source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>qual</th>\n",
       "      <th>strand</th>\n",
       "      <th>misc</th>\n",
       "      <th>gene_info</th>\n",
       "      <th>ensg_id</th>\n",
       "      <th>enst_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>status</th>\n",
       "      <th>gene_sym</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.4\"; transcript_id \"EN...</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.4\"; transcript_id \"EN...</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.4\"; transcript_id \"EN...</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>11872</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.4\"; transcript_id \"EN...</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>ENST00000515242.2</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.4\"; transcript_id \"EN...</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>ENST00000515242.2</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr annotation_source  type  start    end qual strand misc  \\\n",
       "2  chr1            HAVANA  exon  11869  12227    .      +    .   \n",
       "3  chr1            HAVANA  exon  12613  12721    .      +    .   \n",
       "4  chr1            HAVANA  exon  13221  14409    .      +    .   \n",
       "6  chr1           ENSEMBL  exon  11872  12227    .      +    .   \n",
       "7  chr1           ENSEMBL  exon  12613  12721    .      +    .   \n",
       "\n",
       "                                           gene_info            ensg_id  \\\n",
       "2  gene_id \"ENSG00000223972.4\"; transcript_id \"EN...  ENSG00000223972.4   \n",
       "3  gene_id \"ENSG00000223972.4\"; transcript_id \"EN...  ENSG00000223972.4   \n",
       "4  gene_id \"ENSG00000223972.4\"; transcript_id \"EN...  ENSG00000223972.4   \n",
       "6  gene_id \"ENSG00000223972.4\"; transcript_id \"EN...  ENSG00000223972.4   \n",
       "7  gene_id \"ENSG00000223972.4\"; transcript_id \"EN...  ENSG00000223972.4   \n",
       "\n",
       "             enst_id   gene_type status gene_sym  \n",
       "2  ENST00000456328.2  pseudogene  KNOWN  DDX11L1  \n",
       "3  ENST00000456328.2  pseudogene  KNOWN  DDX11L1  \n",
       "4  ENST00000456328.2  pseudogene  KNOWN  DDX11L1  \n",
       "6  ENST00000515242.2  pseudogene  KNOWN  DDX11L1  \n",
       "7  ENST00000515242.2  pseudogene  KNOWN  DDX11L1  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_annon_df[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "protein_coding              20345\n",
       "pseudogene                  13931\n",
       "lincRNA                      7114\n",
       "antisense                    5276\n",
       "miRNA                        3055\n",
       "misc_RNA                     2034\n",
       "snRNA                        1916\n",
       "snoRNA                       1457\n",
       "sense_intronic                742\n",
       "rRNA                          527\n",
       "processed_transcript          515\n",
       "sense_overlapping             202\n",
       "IG_V_pseudogene               187\n",
       "IG_V_gene                     138\n",
       "TR_V_gene                      97\n",
       "TR_J_gene                      74\n",
       "polymorphic_pseudogene         45\n",
       "IG_D_gene                      37\n",
       "TR_V_pseudogene                27\n",
       "Mt_tRNA                        22\n",
       "3prime_overlapping_ncrna       21\n",
       "IG_J_gene                      18\n",
       "IG_C_gene                      14\n",
       "IG_C_pseudogene                 9\n",
       "TR_C_gene                       5\n",
       "TR_J_pseudogene                 4\n",
       "IG_J_pseudogene                 3\n",
       "TR_D_gene                       3\n",
       "Mt_rRNA                         2\n",
       "Name: gene_type, dtype: int64"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_annon_df.gene_type.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "count    5.782000e+04\n",
      "mean     2.990738e+04\n",
      "std      8.709646e+04\n",
      "min      7.000000e+00\n",
      "25%      4.710000e+02\n",
      "50%      2.782500e+03\n",
      "75%      2.262200e+04\n",
      "max      2.304637e+06\n",
      "dtype: float64\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7f910a509050>"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZkAAAD4CAYAAAA+epuFAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAf+UlEQVR4nO3df5BV5Z3n8fenb3cDKj+bxiWAgrGdpEkmxLRiNqNboxvFVCa4tWTSakUzay2brO7OrrtTws5mJuOYqmGndplNjRrNaKJuCDDuWPZmJiHJajKzqQRtlKDgIA0qthBpBPmh0HC7v/vHfdrcXO8vmj7dCJ9X1a177vc8z/d5zuF2f/v8uBdFBGZmZlloGOsJmJnZ6ctFxszMMuMiY2ZmmXGRMTOzzLjImJlZZhrHegKnmunTp8fcuXPHehpmZu8pGzZs2BsRraVxF5kSc+fOpbu7e6ynYWb2niLplXJxny4zM7PMuMiYmVlmXGTMzCwzLjJmZpYZFxkzM8uMi4yZmWXGRcbMzDLjImNmZplxkTEzs8z4E/8jbNX6nWXjNyw8b5RnYmY29nwkY2ZmmXGRMTOzzLjImJlZZlxkzMwsMy4yZmaWmbqKjKRFkrZK6pG0rMz6cZLWpPXrJc0tWrc8xbdKuqZWTknzUo5tKWdzil8h6RlJeUlLitr/tqSNRY+jkq5L674l6aWidQuGs5PMzGx4ahYZSTngbuBaoB24XlJ7SbNbgP0RcSGwEliR+rYDncB8YBFwj6RcjZwrgJUR0QbsT7kBdgJfAFYVDxwRT0bEgohYAFwJvA38oKjJHwytj4iNtbbXzMxGTj1HMpcCPRGxIyKOAauBxSVtFgMPpeVHgaskKcVXR0R/RLwE9KR8ZXOmPlemHKSc1wFExMsRsQkYrDLXJcD3IuLtOrbLzMwyVk+RmQW8WvS6N8XKtomIPHAAaKnSt1K8BXgz5ag0VjWdwHdKYl+VtEnSSknjynWStFRSt6Tuvr6+ExjOzMyqqafIqEws6mwzUvGaJM0EPgysKwovBz4AXAJMA+4o1zci7o+IjojoaG1trWc4MzOrQz1FpheYU/R6NrCrUhtJjcBkYF+VvpXie4EpKUelsSr5XeCxiDg+FIiI3VHQD3yTwmk6MzMbJfUUmaeBtnTXVzOFU1JdJW26gJvT8hLgiYiIFO9Md5/NA9qApyrlTH2eTDlIOR+vc1uup+RUWTq6IV3ruQ54vs5cZmY2Amp+QWZE5CXdRuE0VA54MCI2S7oT6I6ILuAB4BFJPRSOYDpT382S1gJbgDxwa0QMAJTLmYa8A1gt6S7g2ZQbSZcAjwFTgd+R9CcRMT+tm0vhyOgnJdP/tqRWCqfhNgJfPMH9Y2ZmJ0GFgwcb0tHREd3d3cPu729hNrMzkaQNEdFRGvcn/s3MLDMuMmZmlhkXGTMzy4yLjJmZZcZFxszMMuMiY2ZmmXGRMTOzzLjImJlZZlxkzMwsMy4yZmaWGRcZMzPLjIuMmZllxkXGzMwy4yJjZmaZcZExM7PMuMiYmVlmXGTMzCwzLjJmZpYZFxkzM8tMXUVG0iJJWyX1SFpWZv04SWvS+vWS5hatW57iWyVdUyunpHkpx7aUsznFr5D0jKS8pCUl4w9I2pgeXbVymZnZ6KhZZCTlgLuBa4F24HpJ7SXNbgH2R8SFwEpgRerbDnQC84FFwD2ScjVyrgBWRkQbsD/lBtgJfAFYVWaaRyJiQXp8piheKZeZmY2Ceo5kLgV6ImJHRBwDVgOLS9osBh5Ky48CV0lSiq+OiP6IeAnoSfnK5kx9rkw5SDmvA4iIlyNiEzBYz4ZVy2VmZqOjniIzC3i16HVvipVtExF54ADQUqVvpXgL8GbKUWmscsZL6pb0c0lDhaTuXJKWpv7dfX19dQxnZmb1aKyjjcrEos42leLlilu19rWcFxG7JF0APCHpOeBgvbki4n7gfoCOjo56xjMzszrUcyTTC8wpej0b2FWpjaRGYDKwr0rfSvG9wJSUo9JY7xIRu9LzDuDHwEeHm8vMzEZOPUXmaaAt3anVTOFCfldJmy7g5rS8BHgiIiLFO9PdZ/OANuCpSjlTnydTDlLOx6tNTtJUSePS8nTgE8CW4eQyM7ORVbPIpGsatwHrgBeAtRGxWdKdkobu5HoAaJHUA9wOLEt9NwNrgS3A94FbI2KgUs6U6w7g9pSrJeVG0iWSeoHPAvdJGmr/QaBb0i8oFJU/i4gt1XKZmdnoUOEPfhvS0dER3d3dw+6/av3OsvEbFp437JxmZqc6SRsioqM07k/8m5lZZlxkzMwsMy4yZmaWGRcZMzPLjIuMmZllxkXGzMwy4yJjZmaZcZExM7PMuMiYmVlmXGTMzCwzLjJmZpYZFxkzM8uMi4yZmWXGRcbMzDLjImNmZplxkTEzs8y4yJiZWWZcZMzMLDMuMmZmlpm6ioykRZK2SuqRtKzM+nGS1qT16yXNLVq3PMW3SrqmVk5J81KObSlnc4pfIekZSXlJS4raL5D0M0mbJW2S9Lmidd+S9JKkjemx4ER3kJmZDV/NIiMpB9wNXAu0A9dLai9pdguwPyIuBFYCK1LfdqATmA8sAu6RlKuRcwWwMiLagP0pN8BO4AvAqpKx3wZuioihMf5C0pSi9X8QEQvSY2Ot7TUzs5FTz5HMpUBPROyIiGPAamBxSZvFwENp+VHgKklK8dUR0R8RLwE9KV/ZnKnPlSkHKed1ABHxckRsAgaLB46IFyNiW1reBewBWuveA2Zmlpl6isws4NWi170pVrZNROSBA0BLlb6V4i3AmylHpbEqknQp0AxsLwp/NZ1GWylpXIV+SyV1S+ru6+urdzgzM6uhniKjMrGos81IxWuSNBN4BPi9iBg62lkOfAC4BJgG3FGub0TcHxEdEdHR2uqDIDOzkVJPkekF5hS9ng3sqtRGUiMwGdhXpW+l+F5gSspRaax3kTQJ+Fvgv0bEz4fiEbE7CvqBb1I4TWdmZqOkniLzNNCW7vpqpnAhv6ukTRdwc1peAjwREZHinenus3lAG/BUpZypz5MpBynn49Uml/o/BjwcEX9dsm5mehaFazvP17G9ZmY2QmoWmXR95DZgHfACsDYiNku6U9JnUrMHgBZJPcDtwLLUdzOwFtgCfB+4NSIGKuVMue4Abk+5WlJuJF0iqRf4LHCfpKH2vwtcAXyhzK3K35b0HPAcMB24axj7yMzMhkmFgwcb0tHREd3d3cPuv2r9zrLxGxaeN+ycZmanOkkbIqKjNO5P/JuZWWZcZMzMLDMuMmZmlhkXGTMzy4yLjJmZZcZFxszMMuMiY2ZmmXGRMTOzzLjImJlZZlxkzMwsMy4yZmaWGRcZMzPLjIuMmZllxkXGzMwy4yJjZmaZcZExM7PMuMiYmVlmXGTMzCwzLjJmZpYZFxkzM8tMXUVG0iJJWyX1SFpWZv04SWvS+vWS5hatW57iWyVdUyunpHkpx7aUsznFr5D0jKS8pCUl49+c2m+TdHNR/GOSnktjfE2STmTnmJnZyalZZCTlgLuBa4F24HpJ7SXNbgH2R8SFwEpgRerbDnQC84FFwD2ScjVyrgBWRkQbsD/lBtgJfAFYVTK/acAfAwuBS4E/ljQ1rb4XWAq0pceiWttrZmYjp54jmUuBnojYERHHgNXA4pI2i4GH0vKjwFXpqGExsDoi+iPiJaAn5SubM/W5MuUg5bwOICJejohNwGDJ2NcAP4yIfRGxH/ghsEjSTGBSRPwsIgJ4eCiXmZmNjnqKzCzg1aLXvSlWtk1E5IEDQEuVvpXiLcCbKUelseqd36y0XG3eAEhaKqlbUndfX1+N4czMrF71FJly1zGizjYjFa/mpHNFxP0R0RERHa2trTWGMzOzetVTZHqBOUWvZwO7KrWR1AhMBvZV6VspvheYknJUGqve+fWm5WrzNjOzDNVTZJ4G2tJdX80ULuR3lbTpAobu6loCPJGug3QBnenus3kULr4/VSln6vNkykHK+XiN+a0DrpY0NV3wvxpYFxG7gUOSLkvXem6qI5eZmY2gmkUmXR+5jcIv8xeAtRGxWdKdkj6Tmj0AtEjqAW4HlqW+m4G1wBbg+8CtETFQKWfKdQdwe8rVknIj6RJJvcBngfskbU5j7AP+lELhehq4M8UAvgT8FYUbDrYD3xvGPjIzs2FS4eDBhnR0dER3d/ew+69av7Ns/IaF5w07p5nZqU7ShojoKI37E/9mZpYZFxkzM8uMi4yZmWXGRcbMzDLjImNmZplxkTEzs8y4yJiZWWZcZMzMLDMuMmZmlhkXGTMzy4yLjJmZZcZFxszMMuMiY2ZmmXGRMTOzzLjImJlZZlxkzMwsMy4yZmaWGRcZMzPLjIuMmZllpq4iI2mRpK2SeiQtK7N+nKQ1af16SXOL1i1P8a2SrqmVU9K8lGNbytlcbQxJN0raWPQYlLQgrftxGmNo3Yzh7SYzMxuOmkVGUg64G7gWaAeul9Re0uwWYH9EXAisBFakvu1AJzAfWATcIylXI+cKYGVEtAH7U+6KY0TEtyNiQUQsAD4PvBwRG4vmduPQ+ojYU/eeMTOzk1bPkcylQE9E7IiIY8BqYHFJm8XAQ2n5UeAqSUrx1RHRHxEvAT0pX9mcqc+VKQcp53U1xih2PfCdOrbJzMxGQT1FZhbwatHr3hQr2yYi8sABoKVK30rxFuDNlKN0rEpjFPsc7y4y30ynyr5cpigBIGmppG5J3X19feWamJnZMNRTZMr9Yo4624xUvOY8JC0E3o6I54vW3xgRHwYuT4/Pl8lBRNwfER0R0dHa2lquiZmZDUM9RaYXmFP0ejawq1IbSY3AZGBflb6V4nuBKSlH6ViVxhjSSclRTES8lp4PAasonKYzM7NRUk+ReRpoS3d9NVP4Zd5V0qYLuDktLwGeiIhI8c50Z9g8oA14qlLO1OfJlIOU8/EaYyCpAfgshWs7pFijpOlpuQn4NFB8lGNmZhlrrNUgIvKSbgPWATngwYjYLOlOoDsiuoAHgEck9VA4uuhMfTdLWgtsAfLArRExAFAuZxryDmC1pLuAZ1NuKo2RXAH0RsSOotg4YF0qMDngR8A3TmDfmJnZSVI6GLCko6Mjuru7h91/1fqdZeM3LDxv2DnNzE51kjZEREdp3J/4NzOzzLjImJlZZlxkzMwsMy4yZmaWGRcZMzPLjIuMmZllxkXGzMwy4yJjZmaZcZExM7PMuMiYmVlmXGTMzCwzLjJmZpYZFxkzM8uMi4yZmWXGRcbMzDLjImNmZplxkTEzs8y4yJiZWWZcZMzMLDN1FRlJiyRtldQjaVmZ9eMkrUnr10uaW7RueYpvlXRNrZyS5qUc21LO5mpjSJor6Yikjenx9aJcH5P0XOrzNUk68V1kZmbDVbPISMoBdwPXAu3A9ZLaS5rdAuyPiAuBlcCK1Lcd6ATmA4uAeyTlauRcAayMiDZgf8pdcYxke0QsSI8vFsXvBZYCbemxqNb2mpnZyKnnSOZSoCcidkTEMWA1sLikzWLgobT8KHBVOmpYDKyOiP6IeAnoSfnK5kx9rkw5SDmvqzFGWZJmApMi4mcREcDDRbnMzGwU1FNkZgGvFr3uTbGybSIiDxwAWqr0rRRvAd5MOUrHqjQGwDxJz0r6iaTLi9r31pg3AJKWSuqW1N3X11euiZmZDUM9Rabc0ULU2Wak4tXG2A2cFxEfBW4HVkmaVOe8C8GI+yOiIyI6WltbyzUxM7NhqKfI9AJzil7PBnZVaiOpEZgM7KvSt1J8LzAl5Sgdq+wY6VTcGwARsQHYDlyU2s+uMW8zM8tQPUXmaaAt3fXVTOFCfldJmy7g5rS8BHgiXQfpAjrTnWHzKFx8f6pSztTnyZSDlPPxamNIak03EiDpgjTGjojYDRySdFm6dnNTUS4zMxsFjbUaRERe0m3AOiAHPBgRmyXdCXRHRBfwAPCIpB4KRzCdqe9mSWuBLUAeuDUiBgDK5UxD3gGslnQX8GzKTaUxgCuAOyXlgQHgixGxL637EvAtYALwvfQwM7NRosLBgw3p6OiI7u7uYfdftX5n2fgNC88bdk4zs1OdpA0R0VEa9yf+zcwsMy4yZmaWGRcZMzPLjIuMmZllxkXGzMwy4yJjZmaZcZExM7PMuMiYmVlmXGTMzCwzLjJmZpYZFxkzM8uMi4yZmWXGRcbMzDLjImNmZplxkTEzs8y4yJiZWWZcZMzMLDMuMmZmlhkXmVEQEQwM+r+5NrMzT11FRtIiSVsl9UhaVmb9OElr0vr1kuYWrVue4lslXVMrp6R5Kce2lLO52hiSPilpg6Tn0vOVRbl+nMbYmB4zTnwXnbyf9uzlt1Y8waALjZmdYWoWGUk54G7gWqAduF5Se0mzW4D9EXEhsBJYkfq2A53AfGARcI+kXI2cK4CVEdEG7E+5K44B7AV+JyI+DNwMPFIytxsjYkF67Km5RzKw6bUD7D5wlN79R8ZieDOzMVPPkcylQE9E7IiIY8BqYHFJm8XAQ2n5UeAqSUrx1RHRHxEvAT0pX9mcqc+VKQcp53XVxoiIZyNiV4pvBsZLGlfvDsja4f48r6Xism3PoTGejZnZ6KqnyMwCXi163ZtiZdtERB44ALRU6Vsp3gK8mXKUjlVpjGL/Eng2IvqLYt9Mp8q+nIrYqNr2+iGGTpJt23N4tIc3MxtT9RSZcr+YSy8uVGozUvGa85A0n8IptH9TtP7GdBrt8vT4fJkcSFoqqVtSd19fX7kmw7Ztz2HObs7ROnEc2153kTGzM0s9RaYXmFP0ejawq1IbSY3AZGBflb6V4nuBKSlH6ViVxkDSbOAx4KaI2D6UNCJeS8+HgFUUTtO9S0TcHxEdEdHR2tpaZVecmMEIXnz9EG3nTuQ3zp3o02Vmdsapp8g8DbSlu76aKVzI7ypp00XhojvAEuCJiIgU70x3hs0D2oCnKuVMfZ5MOUg5H682hqQpwN8CyyPip0MTktQoaXpabgI+DTxfx/aOmNf2H+HtYwNcdO5ELpxxDj17DvsOMzM7ozTWahAReUm3AeuAHPBgRGyWdCfQHRFdwAPAI5J6KBxddKa+myWtBbYAeeDWiBgAKJczDXkHsFrSXcCzKTeVxgBuAy4Evizpyyl2NfAWsC4VmBzwI+AbJ7yHTsKLew4hoG3GOZw9rpG3jw2w68ARZk89azSnYWY2ZmoWGYCI+Dvg70pif1S0fBT4bIW+XwW+Wk/OFN9BmdNalcaIiLuAuypM/WMV4qPixV8eYtbUCZw9rpG2c88BCtdoXGTM7EzhT/xnJCL45cGjnD+tUFDaZqQi87qvy5jZmcNFJiNHjg9wfCCYfFYzAFPOavYdZmZ2xnGRycjBI4WP+kwa/6szkm0zzvFnZczsjOIik5GDR48DMHlC0zuxtnSHWeEmOjOz05+LTEYOHikUmUnji4rMuRM53J9n94GjYzUtM7NRVdfdZXbiDqQjmYkTCrt41fqdvPLG2wDc//c7uOjciQDcsPC8sZmgmdko8JFMRg4eyXN2c47Ghl/t4pmTx9OUE8/s3D+GMzMzGz0uMhk5eOQ4k4quxwCMb8rxifdPZ1PvAXa96a/9N7PTn4tMRg4ePf5r12OGXN7WyoSmHD/Y8ssxmJWZ2ehykclIuSMZgAnNOf7ZRa28+PphXtr71hjMzMxs9LjIZCA/MMhbxwaYNKH8fRUff38Lk8Y38t1Nuzh6fGCUZ2dmNnpcZDJw6Gjhg5iTy5wuA2jKNbB4wSx2HzjKf3nsOX9uxsxOWy4yGRj6IGa502VDPjhzEld9YAZ/88xrPPyzV0ZramZmo8qfk8nAgTIfxCzntz8wAwn+9LtbaDmnmU//5vtGY3pmZqPGRzIZOJhOl1W6JjOkQWLl5xbw0fOm8O++8ywP/+zl7CdnZjaKfCSTgYNHjtPYICY05Wq2/T+/2M2nf/N9HDqa548e38y653/JNfP/CTf907nZT9TMLGM+ksnAwaOF25cl1dW+KdfAjQvP57ILpvHT7W9w70+207PH/++Mmb33uchk4OCR8h/ErCbXID7zkVl8/rLzOXDkOIv+4h/4j2s28sLugxnN0swsez5dloGDR/PMnjphWH0/OHMSv39VG3sO9fOdp3by2LOv8aFZk7j2QzP55x88l7YZ59DQUN8RkpnZWHORGSH5gUFyDSIiOHjkOJNnThp2ronjm5g4von/9MnfoPuVfTz/2gH+fN1W/nzdVqae1cQlc6ex8IIWFs6bxpypZ3HWuBxNOR+Umtmpp64iI2kR8D+BHPBXEfFnJevHAQ8DHwPeAD4XES+ndcuBW4AB4N9HxLpqOSXNA1YD04BngM9HxLGRHCMLd353C3sP9zN/5mTyg1H1MzL1mtCc4/K2Vi5va+XAkeNs31P4KpruV/bzgy2v/1rbyROa+MicKXx0zhQuOnci86afzcTxjRw6mmdgMJg7/SwmnuApPDOzk1WzyEjKAXcDnwR6gacldUXElqJmtwD7I+JCSZ3ACuBzktqBTmA+8D7gR5IuSn0q5VwBrIyI1ZK+nnLfO8JjjKiIYPbUCfyvn7/C37+4F6j+QczhmDyhiYvPn8rF508FCp/FefmNtzh8NM+xgUHefPsYL/7yEP/wYh+Vvj/gfZPH09TYwFv9eSQxa8oEZk2dwNnNhSOh/EBw8Ohx+vODTDu7mdaJ4wr/XUGugcYG0ZRroDFXeG7KicaGoufGBpoaVGibE00NDTQ1isYGMRiQHwgGBoOBCAYGBwuvI2jONTC+KceE5hwTmnKMb8pRejZQvCvw6y9VcRUNErkGIaVlFZYlMTAYvH0sz/GBINdQmOvQ3BvSUWkEBLzzrQyRxijkLLTpzw+SH4x39lHOpzNtGIbeY8U3DA0MBhFBYzpTMTgYHM0PkGsQzSl29Pggh/qPM6EpxznjGhkMeOOtfg4dzTP1rGamTGji4NHjbO97i0NHjzNv+tmcO2k8z7yynyf+cQ8BXPmBGVw6b1omZ0TqOZK5FOiJiB0AklYDi4HiX9aLga+k5UeBv1RhTy0GVkdEP/CSpJ6Uj3I5Jb0AXAnckNo8lPLeO1JjlMx7REhi6RXv50OzJvOvH+oGYPL4bM9ETp7QxEdmT3lX/Fh+kL2H+9l7uJ/jA8G4xgYk6DvUz55D/QxGMGPieCKC/W8f4+fb3yY/GO+c7hvflKOxQRzuz3O4P8/gafqNNxKczLf5SNDU0MCxgcGy6xobREOddxdmZSyHf9cfBqM59hhudwQMRhT+0CtaHiogDUV/5CjNtUEiPxgcHxgkgnf+2MkPBPn0AzgUO3r8V++3XEPhj6bi9+DQ3zfFP7cNouLPcXNjAwIe+H8vMXF8I2uWfpz29w3/VH859fwmnAW8WvS6F1hYqU1E5CUdAFpS/OclfWel5XI5W4A3IyJfpv1IjfEukpYCS9PLw5K2lmtXp+nA3uUrTiLD6WE6sHesJzHGvA8KvB8KTvn9MP9PTqr7+eWC9RSZcn8XlNbFSm0qxcsdk1VrP5JjvDsYcT9wf7l1J0pSd0R0jESu9zLvB++DId4PBWfqfqjnBFwvMKfo9WxgV6U2khqBycC+Kn0rxfcCU1KO0rFGagwzMxsl9RSZp4E2SfMkNVO4yN5V0qYLuDktLwGeiMJJyC6gU9K4dNdYG/BUpZypz5MpBynn4yM5Rn27xczMRkLN02Xp+sdtwDoKtwI/GBGbJd0JdEdEF/AA8Ei66L6Pwi90Uru1FC6254FbI2IAoFzONOQdwGpJdwHPptyM8BhZGpHTbqcB7wfvgyHeDwVn5H6Q/8MsMzPLij8mbmZmmXGRMTOzzLjIjBBJiyRtldQjadlYz+dkSHpZ0nOSNkrqTrFpkn4oaVt6nprikvS1tN2bJF1clOfm1H6bpJuL4h9L+XtSX1UbY5S2+UFJeyQ9XxQbs22uNsYY7IevSHotvR82SvpU0brlaY5bJV1TFC/785BuxFmftndNuimHdOPOmtR+vaS5tcbIcB/MkfSkpBckbZb0+yl+xr0fRkThqzP8OJkHhRsLtgMXAM3AL4D2sZ7XSWzPy8D0kth/A5al5WXAirT8KeB7FD6vdBmwPsWnATvS89S0PDWtewr4eOrzPeDaamOM0jZfAVwMPH8qbHOlMcZoP3wF+M9l2ran9/o4YF76GchV+3kA1gKdafnrwJfS8r8Fvp6WO4E11cbIeB/MBC5OyxOBF9M8zrj3w4jsz7GewOnwSG+WdUWvlwPLx3peJ7E9L/PuIrMVmJmWZwJb0/J9wPWl7YDrgfuK4vel2EzgH4vi77SrNMYobvfckl+uY7bNlcYYo/3wFcoXmV97n1O4k/PjlX4e0i/IvUBjir/TbqhvWm5M7VRpjFF+XzxO4TsQz8j3w8k+fLpsZJT76p1ZFdq+FwTwA0kbVPjKHYBzI2I3QHqekeKVtr1avLdMvNoYY2Ust/lUe0/dlk7TPFh0GvNE90PdXxsFFH9t1Jjth3Ta7qPAevx+GBYXmZFRz1fvvJd8IiIuBq4FbpV0RZW2J/p1P6fDvhqNbT6V9tO9wPuBBcBu4L+n+Ejuh1Pu/SLpHOB/A/8hIqr9F7Vn2vvhhLjIjIzT6itsImJXet4DPEbhW61flzQTID3vSc1P9Gt9etNyaZwqY4yVsdzmU+Y9FRGvR8RARAwC3+BX33I+ll8blSlJTRQKzLcj4m9S2O+HYXCRGRmnzVfYSDpb0sShZeBq4Hl+/Wt9Sr/u56Z098tlwIF0mL8OuFrS1HR65WoK5993A4ckXZbuqLmJ8l8dVDzGWBnLba40xqgb+qWX/AsK7wcY26+Nykz6N3oAeCEi/kfRKr8fhmOsLwqdLg8Kd3+8SOHulz8c6/mcxHZcQOFunl8Am4e2hcL58f8LbEvP01JcFP5zuO3Ac0BHUa5/BfSkx+8VxTso/KLaDvwlv/rmibJjjNJ2f4fCqaDjFP5qvGUst7naGGOwHx5Jc9hE4ZfdzKL2f5jmuJV0h1S1n4f0/noq7Z+/Bsal+Pj0uietv6DWGBnug9+icCpqE7AxPT51Jr4fRuLhr5UxM7PM+HSZmZllxkXGzMwy4yJjZmaZcZExM7PMuMiYmVlmXGTMzCwzLjJmZpaZ/w++b/AX0nJvSQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# gene lengths\n",
    "print((gene_annon_df.end-gene_annon_df.start).describe())\n",
    "sns.distplot((gene_annon_df.end-gene_annon_df.start))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "HAVANA     47970\n",
       "ENSEMBL     9850\n",
       "Name: annotation_source, dtype: int64"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_annon_df.annotation_source.value_counts()#.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] File b'../data/external/gene_annon_ALL.csv' does not exist: b'../data/external/gene_annon_ALL.csv'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-5-95b064fd103c>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# make gencode list\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mgencode_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgencode_file\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mindex_col\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mgencode_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m'chr'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'start'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'end'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'width'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'strand'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'gene'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mgencode_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'chr'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'chr'\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mgencode_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'chr'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mgencode_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;36m5\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36mparser_f\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, dialect, error_bad_lines, warn_bad_lines, delim_whitespace, low_memory, memory_map, float_precision)\u001b[0m\n\u001b[1;32m    683\u001b[0m         )\n\u001b[1;32m    684\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 685\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0m_read\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    686\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    687\u001b[0m     \u001b[0mparser_f\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    455\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    456\u001b[0m     \u001b[0;31m# Create the parser.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 457\u001b[0;31m     \u001b[0mparser\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mTextFileReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfp_or_buf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    458\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    459\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mchunksize\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0miterator\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m    893\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"has_index_names\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"has_index_names\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    894\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 895\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    896\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    897\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m_make_engine\u001b[0;34m(self, engine)\u001b[0m\n\u001b[1;32m   1133\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mengine\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"c\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1134\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mengine\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"c\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1135\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mCParserWrapper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1136\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1137\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mengine\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"python\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, src, **kwds)\u001b[0m\n\u001b[1;32m   1915\u001b[0m         \u001b[0mkwds\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"usecols\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0musecols\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1916\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1917\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_reader\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparsers\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mTextReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msrc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1918\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munnamed_cols\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_reader\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munnamed_cols\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1919\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/parsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader.__cinit__\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/parsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader._setup_parser_source\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] File b'../data/external/gene_annon_ALL.csv' does not exist: b'../data/external/gene_annon_ALL.csv'"
     ]
    }
   ],
   "source": [
    "# make gencode list\n",
    "# gencode_df = pd.read_csv(gencode_file,index_col=0)\n",
    "# gencode_df.columns = ['chr', 'start', 'end', 'width', 'strand', 'gene']\n",
    "# gencode_df['chr'] = 'chr' + gencode_df['chr'].astype(str)\n",
    "# gencode_df[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>width</th>\n",
       "      <th>strand</th>\n",
       "      <th>TSS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1846266</td>\n",
       "      <td>1846266</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>CALML6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>231376933</td>\n",
       "      <td>231376933</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>C1orf131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>241683061</td>\n",
       "      <td>241683061</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>FH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>120202421</td>\n",
       "      <td>120202421</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>PHGDH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>204135465</td>\n",
       "      <td>204135465</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>REN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr      start        end  width strand       TSS\n",
       "0  chr1    1846266    1846266      1      *    CALML6\n",
       "1  chr1  231376933  231376933      1      *  C1orf131\n",
       "2  chr1  241683061  241683061      1      *        FH\n",
       "3  chr1  120202421  120202421      1      *     PHGDH\n",
       "4  chr1  204135465  204135465      1      *       REN"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# make tss list\n",
    "tss_df = pd.read_csv(tss_file,index_col=0)\n",
    "tss_df.columns = ['chr', 'start', 'end', 'width', 'strand', 'TSS']\n",
    "tss_df['chr'] = 'chr' + tss_df['chr'].astype(str)\n",
    "tss_df.sort_values('chr',inplace=True)\n",
    "tss_df.reset_index(drop=True,inplace=True)\n",
    "tss_df[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(20178, 1)\n",
      "(69,)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Index(['ACE', 'ALG9', 'ANKRD18A', 'CCDC177', 'CDRT1', 'CELF6', 'CHTF8',\n",
       "       'CKS1B', 'CLN3', 'CRIP1', 'CRYBG3', 'DEFB130', 'FAM47E-STBD1', 'GALT',\n",
       "       'GOLGA7B', 'HOXC4', 'IER3IP1', 'ITFG3', 'ITGB3', 'KBTBD4', 'KIAA0391',\n",
       "       'KLK9', 'KRBOX1', 'KRT222', 'LCN6', 'LIMS3', 'LSP1', 'LTB4R2', 'MDGA2',\n",
       "       'MFRP', 'MRPL30', 'MRPS17', 'MUC3A', 'NAA60', 'NDUFA7', 'NDUFB8',\n",
       "       'NPIPA7', 'OC90', 'PAGR1', 'PIK3R2', 'POLR2J2', 'PSMA1', 'PSMA2',\n",
       "       'RBL1', 'RNASE11', 'RPP14', 'SDHD', 'SEPT1', 'SHPK', 'SIGLEC5',\n",
       "       'SLC25A10', 'SNURF', 'SOGA3', 'TAP2', 'TIMM10B', 'TLR9', 'TM4SF19',\n",
       "       'TMBIM4', 'TMEM236', 'TUBB3', 'UGT2A1', 'UQCR11', 'ZFP41', 'ZNF223',\n",
       "       'ZNF286A', 'ZNF668', 'ZNF709', 'ZNF747', 'ZNF8'],\n",
       "      dtype='object', name='TSS')"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "check_tss_unique = pd.DataFrame(tss_df.groupby(\"TSS\")['end'].nunique())\n",
    "print(check_tss_unique.shape)\n",
    "print(check_tss_unique[check_tss_unique.end>1].index.shape)\n",
    "check_tss_unique[check_tss_unique.end>1].index.sort_values()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " note that there are 69 of 20178 unique TSS symbols that have two TSS locations, there is some discrepancy as to where those TSS's actually are located so we'll just use both and not this for future reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [],
   "source": [
    "PROMOTER_LEN_UPSTREAM=2000\n",
    "PROMOTER_LEN_DOWNSTREAM=500"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {},
   "outputs": [],
   "source": [
    "promoter_df = tss_df\n",
    "promoter_df['start_n']= promoter_df['start'].map(int) - PROMOTER_LEN_UPSTREAM\n",
    "promoter_df['end_n']= promoter_df.end + PROMOTER_LEN_DOWNSTREAM\n",
    "promoter_df = promoter_df[['chr','start_n','end_n','TSS']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start_n</th>\n",
       "      <th>end_n</th>\n",
       "      <th>TSS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1834266</td>\n",
       "      <td>1836766</td>\n",
       "      <td>CALML6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>231364933</td>\n",
       "      <td>231367433</td>\n",
       "      <td>C1orf131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>241671061</td>\n",
       "      <td>241673561</td>\n",
       "      <td>FH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>120190421</td>\n",
       "      <td>120192921</td>\n",
       "      <td>PHGDH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>204123465</td>\n",
       "      <td>204125965</td>\n",
       "      <td>REN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr    start_n      end_n       TSS\n",
       "0  chr1    1834266    1836766    CALML6\n",
       "1  chr1  231364933  231367433  C1orf131\n",
       "2  chr1  241671061  241673561        FH\n",
       "3  chr1  120190421  120192921     PHGDH\n",
       "4  chr1  204123465  204125965       REN"
      ]
     },
     "execution_count": 211,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "promoter_df[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [],
   "source": [
    "promoter_df.to_csv('promoter_hg19_{}_{}.bed'.format(PROMOTER_LEN_UPSTREAM,PROMOTER_LEN_DOWNSTREAM),\n",
    "                   header=None,index=None,sep='\\t')\n",
    "                   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {},
   "outputs": [],
   "source": [
    "# cmd = \"bedtools sort -i promoter_hg19_5000_5000.bed\"\n",
    "# cmd_arr = cmd.split(' ')\n",
    "# print(cmd_arr)\n",
    "# with open(\"promoter_hg19_2500_500_sort.bed\", \"w\") as file:\n",
    "#     subprocess.run(cmd_arr, shell=True, check=True, stdout=file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BedTool(promoter_hg19_2000_500_sort.bed)>"
      ]
     },
     "execution_count": 214,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "promoter = pybedtools.BedTool('promoter_hg19_{}_{}.bed'.format(PROMOTER_LEN_UPSTREAM,PROMOTER_LEN_DOWNSTREAM))\n",
    "\n",
    "promoter.sort().saveas('promoter_hg19_{}_{}_sort.bed'.format(PROMOTER_LEN_UPSTREAM,PROMOTER_LEN_DOWNSTREAM))\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# fix tf annotation file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_tpm_file = '../data/interim/rna/tissue_tpm_sym.csv'\n",
    "rna_df = pd.read_csv(rna_tpm_file, index_col=0,header=0)\n",
    "rna_df.index = [x.upper() for x in rna_df.index.values]\n",
    "def get_exp(rna_df, sym, tissue,verbose=True):\n",
    "    try:\n",
    "        return rna_df.loc[sym,tissue]\n",
    "    except KeyError:\n",
    "        if verbose:\n",
    "            print(sym, 'not found')\n",
    "        return 0\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "hoco_annon_file = '../data/external/HOCOMOCOv11_full_annotation_HUMAN_mono.tsv'\n",
    "hoco_annon = pd.read_table(hoco_annon_file,header=0)\n",
    "# hoco_annon[['motif_abbr','motif_info']]=hoco_annon['Model'].str.split(\"_\",expand=True)\n",
    "hoco_id_to_tf = pd.Series(hoco_annon['Transcription factor'].values, index=hoco_annon.Model.values).to_dict()\n",
    "# hoco_df = pd.read_table(hoco_annon_file, index_col=None, \n",
    "#                             names=['anchor_chr', 'anchor_start', 'anchor_end', \n",
    "#                                     'foot_chr', 'foot_start', 'foot_end', 'motif', 'score', 'strand'])\n",
    "# hoco_df['transcription_factor'] = hoco_df['motif'].map(hoco_motif_to_tf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "T not found\n",
      "BRAC_HUMAN.H11MO.0.A T\n",
      "T not found\n",
      "BRAC_HUMAN.H11MO.1.B T\n",
      "HNF1B not found\n",
      "HNF1B_HUMAN.H11MO.0.A HNF1B\n",
      "HNF1B not found\n",
      "HNF1B_HUMAN.H11MO.1.A HNF1B\n",
      "HSFY1; HSFY2 not found\n",
      "HSFY1_HUMAN.H11MO.0.D HSFY1; HSFY2\n",
      "MEF2B not found\n",
      "MEF2B_HUMAN.H11MO.0.A MEF2B\n",
      "NR2E3 not found\n",
      "NR2E3_HUMAN.H11MO.0.C NR2E3\n",
      "ZNF8 not found\n",
      "ZNF8_HUMAN.H11MO.0.C ZNF8\n"
     ]
    }
   ],
   "source": [
    "# check to see if the rna expression for these tfs is in the table\n",
    "for motif, tf in hoco_id_to_tf.items():\n",
    "    exp = get_exp(rna_df, tf, 'GDSD0')\n",
    "    if type(exp)==int:\n",
    "        print(motif, tf)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "769"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(hoco_id_to_tf) # 769\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So of the 769 human Transcription factors in HOCOMOCO,5 are not found in RNA_df by direct lookup, By manual search we find the following mappings\n",
    "\n",
    "-  BRAC_HUMAN.H11MO.0.A, T --> TBXT\n",
    "- BRAC_HUMAN.H11MO.1.B T --> TBXT\n",
    "- HNF1B_HUMAN.H11MO.0.A --> n/a\n",
    "- HNF1B_HUMAN.H11MO.1.A --> n/a\n",
    "- HSFY1_HUMAN.H11MO.0.D --> n/a\n",
    "- MEF2B_HUMAN.H11MO.0.A --> n/a\n",
    "- NR2E3_HUMAN.H11MO.0.C --> n/a\n",
    "- ZNF8_HUMAN.H11MO.0.C --> ZNF8-ERVK3-1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# how to check for manually \n",
    "'IRF6' in rna_df.index\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HNF1B not found\n",
      "HNF1B_HUMAN.H11MO.0.A HNF1B\n",
      "HNF1B not found\n",
      "HNF1B_HUMAN.H11MO.1.A HNF1B\n",
      "HSFY1; HSFY2 not found\n",
      "HSFY1_HUMAN.H11MO.0.D HSFY1; HSFY2\n",
      "MEF2B not found\n",
      "MEF2B_HUMAN.H11MO.0.A MEF2B\n",
      "NR2E3 not found\n",
      "NR2E3_HUMAN.H11MO.0.C NR2E3\n"
     ]
    }
   ],
   "source": [
    "# manual correction of mapping dict\n",
    "# Add in IRF6\n",
    "hoco_id_to_tf['IRF6_HUMAN.MANUAL.0.A'] = 'IRF6'\n",
    "hoco_id_to_tf['BRAC_HUMAN.H11MO.0.A'] = 'TBXT'\n",
    "hoco_id_to_tf['BRAC_HUMAN.H11MO.1.B'] = 'TBXT'\n",
    "hoco_id_to_tf['ZNF8_HUMAN.H11MO.0.C'] = 'ZNF8-ERVK3-1'\n",
    "for motif, tf in hoco_id_to_tf.items():\n",
    "    exp = get_exp(rna_df, tf, 'GDSD0')\n",
    "    if type(exp)==int:\n",
    "        print(motif, tf)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "hoco_id_to_tf = pd.DataFrame(pd.Series(hoco_id_to_tf).reset_index())\n",
    "hoco_id_to_tf.columns=['id','tf']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>tf</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>AHR_HUMAN.H11MO.0.B</td>\n",
       "      <td>AHR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>AIRE_HUMAN.H11MO.0.C</td>\n",
       "      <td>AIRE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>ALX1_HUMAN.H11MO.0.B</td>\n",
       "      <td>ALX1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>ALX3_HUMAN.H11MO.0.D</td>\n",
       "      <td>ALX3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>ALX4_HUMAN.H11MO.0.D</td>\n",
       "      <td>ALX4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     id    tf\n",
       "0   AHR_HUMAN.H11MO.0.B   AHR\n",
       "1  AIRE_HUMAN.H11MO.0.C  AIRE\n",
       "2  ALX1_HUMAN.H11MO.0.B  ALX1\n",
       "3  ALX3_HUMAN.H11MO.0.D  ALX3\n",
       "4  ALX4_HUMAN.H11MO.0.D  ALX4"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hoco_id_to_tf[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "hoco_id_to_tf.to_csv('../data/external/HOCOMOCOv11_annotation.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# cluster the hocomoco stuff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "130"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clusters_df = pd.read_csv('/Users/mguo123/Documents/rsat/hocomoco_human_cluster/clusters_tf.csv')\n",
    "clusters_df.columns = ['id','cluster']\n",
    "# clusters_df[['motif_abbr','motif_info']]=clusters_df['id'].str.split(\"_\",expand=True)\n",
    "clusters_df['tf'] = clusters_df['id'].map(hoco_id_to_tf)\n",
    "clusters_df.sort_values('cluster',inplace=True)\n",
    "clusters_df_tf = clusters_df.groupby('tf').first() \n",
    "clusters_df_tf.sort_values('cluster',inplace=True)\n",
    "len(set(clusters_df_tf['cluster'])) # number of clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "clusters_df_tf = clusters_df_tf.merge(hoco_annon, how='left',left_on = 'id', right_on='Model')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "clusters_df_tf.to_csv('../data/external/hocomoco_tf_clusters_prenaming.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "manually make cluster names\n",
    "\n",
    "save as `'../data/external/hocomoco_tf_clusters_postnaming.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>cluster</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tf</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>AHR</td>\n",
       "      <td>AHR_HUMAN.H11MO.0.B</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>EPAS1</td>\n",
       "      <td>EPAS1_HUMAN.H11MO.0.B</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ARNT2</td>\n",
       "      <td>ARNT2_HUMAN.H11MO.0.D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>HIF1A</td>\n",
       "      <td>HIF1A_HUMAN.H11MO.0.C</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ARNT</td>\n",
       "      <td>ARNT_HUMAN.H11MO.0.B</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ZIC3</td>\n",
       "      <td>ZIC3_HUMAN.H11MO.0.B</td>\n",
       "      <td>129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ZNF257</td>\n",
       "      <td>ZN257_HUMAN.H11MO.0.C</td>\n",
       "      <td>130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ZNF410</td>\n",
       "      <td>ZN410_HUMAN.H11MO.0.D</td>\n",
       "      <td>131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ZNF282</td>\n",
       "      <td>ZN282_HUMAN.H11MO.0.D</td>\n",
       "      <td>131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ZNF652</td>\n",
       "      <td>ZN652_HUMAN.H11MO.0.D</td>\n",
       "      <td>132</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>677 rows  2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           id  cluster\n",
       "tf                                    \n",
       "AHR       AHR_HUMAN.H11MO.0.B        1\n",
       "EPAS1   EPAS1_HUMAN.H11MO.0.B        1\n",
       "ARNT2   ARNT2_HUMAN.H11MO.0.D        1\n",
       "HIF1A   HIF1A_HUMAN.H11MO.0.C        1\n",
       "ARNT     ARNT_HUMAN.H11MO.0.B        1\n",
       "...                       ...      ...\n",
       "ZIC3     ZIC3_HUMAN.H11MO.0.B      129\n",
       "ZNF257  ZN257_HUMAN.H11MO.0.C      130\n",
       "ZNF410  ZN410_HUMAN.H11MO.0.D      131\n",
       "ZNF282  ZN282_HUMAN.H11MO.0.D      131\n",
       "ZNF652  ZN652_HUMAN.H11MO.0.D      132\n",
       "\n",
       "[677 rows x 2 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clusters_df_tf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# hocomoco human trim motifs --> jaspar.txt file\n",
    "\n",
    "like '../data/external/HOCOMOCOv11_core_HUMAN_mono_jaspar_format.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob, os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_jaspar_file = '../data/external/hocomoco_human_trim_jaspar_format.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AHR_HUMAN.H11MO.0.B\n",
      "AIRE_HUMAN.H11MO.0.C\n",
      "ALX1_HUMAN.H11MO.0.B\n",
      "ALX3_HUMAN.H11MO.0.D\n",
      "ALX4_HUMAN.H11MO.0.D\n",
      "ANDR_HUMAN.H11MO.0.A\n",
      "ANDR_HUMAN.H11MO.1.A\n",
      "ANDR_HUMAN.H11MO.2.A\n",
      "AP2A_HUMAN.H11MO.0.A\n",
      "AP2B_HUMAN.H11MO.0.B\n",
      "AP2C_HUMAN.H11MO.0.A\n",
      "AP2D_HUMAN.H11MO.0.D\n",
      "ARI3A_HUMAN.H11MO.0.D\n",
      "ARI5B_HUMAN.H11MO.0.C\n",
      "ARNT2_HUMAN.H11MO.0.D\n",
      "ARNT_HUMAN.H11MO.0.B\n",
      "ARX_HUMAN.H11MO.0.D\n",
      "ASCL1_HUMAN.H11MO.0.A\n",
      "ASCL2_HUMAN.H11MO.0.D\n",
      "ATF1_HUMAN.H11MO.0.B\n",
      "ATF2_HUMAN.H11MO.0.B\n",
      "ATF2_HUMAN.H11MO.1.B\n",
      "ATF2_HUMAN.H11MO.2.C\n",
      "ATF3_HUMAN.H11MO.0.A\n",
      "ATF4_HUMAN.H11MO.0.A\n",
      "ATF6A_HUMAN.H11MO.0.B\n",
      "ATF7_HUMAN.H11MO.0.D\n",
      "ATOH1_HUMAN.H11MO.0.B\n",
      "BACH1_HUMAN.H11MO.0.A\n",
      "BACH2_HUMAN.H11MO.0.A\n",
      "BARH1_HUMAN.H11MO.0.D\n",
      "BARH2_HUMAN.H11MO.0.D\n",
      "BARX1_HUMAN.H11MO.0.D\n",
      "BARX2_HUMAN.H11MO.0.D\n",
      "BATF3_HUMAN.H11MO.0.B\n",
      "BATF_HUMAN.H11MO.0.A\n",
      "BATF_HUMAN.H11MO.1.A\n",
      "BC11A_HUMAN.H11MO.0.A\n",
      "BCL6B_HUMAN.H11MO.0.D\n",
      "BCL6_HUMAN.H11MO.0.A\n",
      "BHA15_HUMAN.H11MO.0.B\n",
      "BHE22_HUMAN.H11MO.0.D\n",
      "BHE23_HUMAN.H11MO.0.D\n",
      "BHE40_HUMAN.H11MO.0.A\n",
      "BHE41_HUMAN.H11MO.0.D\n",
      "BMAL1_HUMAN.H11MO.0.A\n",
      "BPTF_HUMAN.H11MO.0.D\n",
      "BRAC_HUMAN.H11MO.0.A\n",
      "BRAC_HUMAN.H11MO.1.B\n",
      "BRCA1_HUMAN.H11MO.0.D\n",
      "BSH_HUMAN.H11MO.0.D\n",
      "CDC5L_HUMAN.H11MO.0.D\n",
      "CDX1_HUMAN.H11MO.0.C\n",
      "CDX2_HUMAN.H11MO.0.A\n",
      "CEBPA_HUMAN.H11MO.0.A\n",
      "CEBPB_HUMAN.H11MO.0.A\n",
      "CEBPD_HUMAN.H11MO.0.C\n",
      "CEBPE_HUMAN.H11MO.0.A\n",
      "CEBPG_HUMAN.H11MO.0.B\n",
      "CEBPZ_HUMAN.H11MO.0.D\n",
      "CENPB_HUMAN.H11MO.0.D\n",
      "CLOCK_HUMAN.H11MO.0.C\n",
      "COE1_HUMAN.H11MO.0.A\n",
      "COT1_HUMAN.H11MO.0.C\n",
      "COT1_HUMAN.H11MO.1.C\n",
      "COT2_HUMAN.H11MO.0.A\n",
      "COT2_HUMAN.H11MO.1.A\n",
      "CPEB1_HUMAN.H11MO.0.D\n",
      "CR3L1_HUMAN.H11MO.0.D\n",
      "CR3L2_HUMAN.H11MO.0.D\n",
      "CREB1_HUMAN.H11MO.0.A\n",
      "CREB3_HUMAN.H11MO.0.D\n",
      "CREB5_HUMAN.H11MO.0.D\n",
      "CREM_HUMAN.H11MO.0.C\n",
      "CRX_HUMAN.H11MO.0.B\n",
      "CTCFL_HUMAN.H11MO.0.A\n",
      "CTCF_HUMAN.H11MO.0.A\n",
      "CUX1_HUMAN.H11MO.0.C\n",
      "CUX2_HUMAN.H11MO.0.D\n",
      "CXXC1_HUMAN.H11MO.0.D\n",
      "DBP_HUMAN.H11MO.0.B\n",
      "DDIT3_HUMAN.H11MO.0.D\n",
      "DLX1_HUMAN.H11MO.0.D\n",
      "DLX2_HUMAN.H11MO.0.D\n",
      "DLX3_HUMAN.H11MO.0.C\n",
      "DLX4_HUMAN.H11MO.0.D\n",
      "DLX5_HUMAN.H11MO.0.D\n",
      "DLX6_HUMAN.H11MO.0.D\n",
      "DMBX1_HUMAN.H11MO.0.D\n",
      "DMRT1_HUMAN.H11MO.0.D\n",
      "DPRX_HUMAN.H11MO.0.D\n",
      "DRGX_HUMAN.H11MO.0.D\n",
      "DUX4_HUMAN.H11MO.0.A\n",
      "DUXA_HUMAN.H11MO.0.D\n",
      "E2F1_HUMAN.H11MO.0.A\n",
      "E2F2_HUMAN.H11MO.0.B\n",
      "E2F3_HUMAN.H11MO.0.A\n",
      "E2F4_HUMAN.H11MO.0.A\n",
      "E2F4_HUMAN.H11MO.1.A\n",
      "E2F5_HUMAN.H11MO.0.B\n",
      "E2F6_HUMAN.H11MO.0.A\n",
      "E2F7_HUMAN.H11MO.0.B\n",
      "E2F8_HUMAN.H11MO.0.D\n",
      "E4F1_HUMAN.H11MO.0.D\n",
      "EGR1_HUMAN.H11MO.0.A\n",
      "EGR2_HUMAN.H11MO.0.A\n",
      "EGR2_HUMAN.H11MO.1.A\n",
      "EGR3_HUMAN.H11MO.0.D\n",
      "EGR4_HUMAN.H11MO.0.D\n",
      "EHF_HUMAN.H11MO.0.B\n",
      "ELF1_HUMAN.H11MO.0.A\n",
      "ELF2_HUMAN.H11MO.0.C\n",
      "ELF3_HUMAN.H11MO.0.A\n",
      "ELF5_HUMAN.H11MO.0.A\n",
      "ELK1_HUMAN.H11MO.0.B\n",
      "ELK3_HUMAN.H11MO.0.D\n",
      "ELK4_HUMAN.H11MO.0.A\n",
      "EMX1_HUMAN.H11MO.0.D\n",
      "EMX2_HUMAN.H11MO.0.D\n",
      "EOMES_HUMAN.H11MO.0.D\n",
      "EPAS1_HUMAN.H11MO.0.B\n",
      "ERG_HUMAN.H11MO.0.A\n",
      "ERR1_HUMAN.H11MO.0.A\n",
      "ERR2_HUMAN.H11MO.0.A\n",
      "ERR3_HUMAN.H11MO.0.B\n",
      "ESR1_HUMAN.H11MO.0.A\n",
      "ESR1_HUMAN.H11MO.1.A\n",
      "ESR2_HUMAN.H11MO.0.A\n",
      "ESR2_HUMAN.H11MO.1.A\n",
      "ESX1_HUMAN.H11MO.0.D\n",
      "ETS1_HUMAN.H11MO.0.A\n",
      "ETS2_HUMAN.H11MO.0.B\n",
      "ETV1_HUMAN.H11MO.0.A\n",
      "ETV2_HUMAN.H11MO.0.B\n",
      "ETV3_HUMAN.H11MO.0.D\n",
      "ETV4_HUMAN.H11MO.0.B\n",
      "ETV5_HUMAN.H11MO.0.C\n",
      "ETV6_HUMAN.H11MO.0.D\n",
      "ETV7_HUMAN.H11MO.0.D\n",
      "EVI1_HUMAN.H11MO.0.B\n",
      "FEV_HUMAN.H11MO.0.B\n",
      "FEZF1_HUMAN.H11MO.0.C\n",
      "FIGLA_HUMAN.H11MO.0.D\n",
      "FLI1_HUMAN.H11MO.0.A\n",
      "FLI1_HUMAN.H11MO.1.A\n",
      "FOSB_HUMAN.H11MO.0.A\n",
      "FOSL1_HUMAN.H11MO.0.A\n",
      "FOSL2_HUMAN.H11MO.0.A\n",
      "FOS_HUMAN.H11MO.0.A\n",
      "FOXA1_HUMAN.H11MO.0.A\n",
      "FOXA2_HUMAN.H11MO.0.A\n",
      "FOXA3_HUMAN.H11MO.0.B\n",
      "FOXB1_HUMAN.H11MO.0.D\n",
      "FOXC1_HUMAN.H11MO.0.C\n",
      "FOXC2_HUMAN.H11MO.0.D\n",
      "FOXD1_HUMAN.H11MO.0.D\n",
      "FOXD2_HUMAN.H11MO.0.D\n",
      "FOXD3_HUMAN.H11MO.0.D\n",
      "FOXF1_HUMAN.H11MO.0.D\n",
      "FOXF2_HUMAN.H11MO.0.D\n",
      "FOXG1_HUMAN.H11MO.0.D\n",
      "FOXH1_HUMAN.H11MO.0.A\n",
      "FOXI1_HUMAN.H11MO.0.B\n",
      "FOXJ2_HUMAN.H11MO.0.C\n",
      "FOXJ3_HUMAN.H11MO.0.A\n",
      "FOXJ3_HUMAN.H11MO.1.B\n",
      "FOXK1_HUMAN.H11MO.0.A\n",
      "FOXL1_HUMAN.H11MO.0.D\n",
      "FOXM1_HUMAN.H11MO.0.A\n",
      "FOXO1_HUMAN.H11MO.0.A\n",
      "FOXO3_HUMAN.H11MO.0.B\n",
      "FOXO4_HUMAN.H11MO.0.C\n",
      "FOXO6_HUMAN.H11MO.0.D\n",
      "FOXP1_HUMAN.H11MO.0.A\n",
      "FOXP2_HUMAN.H11MO.0.C\n",
      "FOXP3_HUMAN.H11MO.0.D\n",
      "FOXQ1_HUMAN.H11MO.0.C\n",
      "FUBP1_HUMAN.H11MO.0.D\n",
      "GABPA_HUMAN.H11MO.0.A\n",
      "GATA1_HUMAN.H11MO.0.A\n",
      "GATA1_HUMAN.H11MO.1.A\n",
      "GATA2_HUMAN.H11MO.0.A\n",
      "GATA2_HUMAN.H11MO.1.A\n",
      "GATA3_HUMAN.H11MO.0.A\n",
      "GATA4_HUMAN.H11MO.0.A\n",
      "GATA5_HUMAN.H11MO.0.D\n",
      "GATA6_HUMAN.H11MO.0.A\n",
      "GBX1_HUMAN.H11MO.0.D\n",
      "GBX2_HUMAN.H11MO.0.D\n",
      "GCM1_HUMAN.H11MO.0.D\n",
      "GCM2_HUMAN.H11MO.0.D\n",
      "GCR_HUMAN.H11MO.0.A\n",
      "GCR_HUMAN.H11MO.1.A\n",
      "GFI1B_HUMAN.H11MO.0.A\n",
      "GFI1_HUMAN.H11MO.0.C\n",
      "GLI1_HUMAN.H11MO.0.D\n",
      "GLI2_HUMAN.H11MO.0.D\n",
      "GLI3_HUMAN.H11MO.0.B\n",
      "GLIS1_HUMAN.H11MO.0.D\n",
      "GLIS2_HUMAN.H11MO.0.D\n",
      "GLIS3_HUMAN.H11MO.0.D\n",
      "GMEB2_HUMAN.H11MO.0.D\n",
      "GRHL1_HUMAN.H11MO.0.D\n",
      "GRHL2_HUMAN.H11MO.0.A\n",
      "GSC2_HUMAN.H11MO.0.D\n",
      "GSC_HUMAN.H11MO.0.D\n",
      "GSX1_HUMAN.H11MO.0.D\n",
      "GSX2_HUMAN.H11MO.0.D\n",
      "HAND1_HUMAN.H11MO.0.D\n",
      "HAND1_HUMAN.H11MO.1.D\n",
      "HBP1_HUMAN.H11MO.0.D\n",
      "HEN1_HUMAN.H11MO.0.C\n",
      "HES1_HUMAN.H11MO.0.D\n",
      "HES5_HUMAN.H11MO.0.D\n",
      "HES7_HUMAN.H11MO.0.D\n",
      "HESX1_HUMAN.H11MO.0.D\n",
      "HEY1_HUMAN.H11MO.0.D\n",
      "HEY2_HUMAN.H11MO.0.D\n",
      "HIC1_HUMAN.H11MO.0.C\n",
      "HIC2_HUMAN.H11MO.0.D\n",
      "HIF1A_HUMAN.H11MO.0.C\n",
      "HINFP_HUMAN.H11MO.0.C\n",
      "HLF_HUMAN.H11MO.0.C\n",
      "HLTF_HUMAN.H11MO.0.D\n",
      "HMBX1_HUMAN.H11MO.0.D\n",
      "HME1_HUMAN.H11MO.0.D\n",
      "HME2_HUMAN.H11MO.0.D\n",
      "HMGA1_HUMAN.H11MO.0.D\n",
      "HMGA2_HUMAN.H11MO.0.D\n",
      "HMX1_HUMAN.H11MO.0.D\n",
      "HMX2_HUMAN.H11MO.0.D\n",
      "HMX3_HUMAN.H11MO.0.D\n",
      "HNF1A_HUMAN.H11MO.0.C\n",
      "HNF1B_HUMAN.H11MO.0.A\n",
      "HNF1B_HUMAN.H11MO.1.A\n",
      "HNF4A_HUMAN.H11MO.0.A\n",
      "HNF4G_HUMAN.H11MO.0.B\n",
      "HNF6_HUMAN.H11MO.0.B\n",
      "HOMEZ_HUMAN.H11MO.0.D\n",
      "HSF1_HUMAN.H11MO.0.A\n",
      "HSF1_HUMAN.H11MO.1.A\n",
      "HSF2_HUMAN.H11MO.0.A\n",
      "HSF4_HUMAN.H11MO.0.D\n",
      "HSFY1_HUMAN.H11MO.0.D\n",
      "HTF4_HUMAN.H11MO.0.A\n",
      "HXA10_HUMAN.H11MO.0.C\n",
      "HXA11_HUMAN.H11MO.0.D\n",
      "HXA13_HUMAN.H11MO.0.C\n",
      "HXA1_HUMAN.H11MO.0.C\n",
      "HXA2_HUMAN.H11MO.0.D\n",
      "HXA5_HUMAN.H11MO.0.D\n",
      "HXA7_HUMAN.H11MO.0.D\n",
      "HXA9_HUMAN.H11MO.0.B\n",
      "HXB13_HUMAN.H11MO.0.A\n",
      "HXB1_HUMAN.H11MO.0.D\n",
      "HXB2_HUMAN.H11MO.0.D\n",
      "HXB3_HUMAN.H11MO.0.D\n",
      "HXB4_HUMAN.H11MO.0.B\n",
      "HXB6_HUMAN.H11MO.0.D\n",
      "HXB7_HUMAN.H11MO.0.C\n",
      "HXB8_HUMAN.H11MO.0.C\n",
      "HXC10_HUMAN.H11MO.0.D\n",
      "HXC11_HUMAN.H11MO.0.D\n",
      "HXC12_HUMAN.H11MO.0.D\n",
      "HXC13_HUMAN.H11MO.0.D\n",
      "HXC6_HUMAN.H11MO.0.D\n",
      "HXC8_HUMAN.H11MO.0.D\n",
      "HXC9_HUMAN.H11MO.0.C\n",
      "HXD10_HUMAN.H11MO.0.D\n",
      "HXD11_HUMAN.H11MO.0.D\n",
      "HXD12_HUMAN.H11MO.0.D\n",
      "HXD13_HUMAN.H11MO.0.D\n",
      "HXD3_HUMAN.H11MO.0.D\n",
      "HXD4_HUMAN.H11MO.0.D\n",
      "HXD8_HUMAN.H11MO.0.D\n",
      "HXD9_HUMAN.H11MO.0.D\n",
      "ID4_HUMAN.H11MO.0.D\n",
      "IKZF1_HUMAN.H11MO.0.C\n",
      "INSM1_HUMAN.H11MO.0.C\n",
      "IRF1_HUMAN.H11MO.0.A\n",
      "IRF2_HUMAN.H11MO.0.A\n",
      "IRF3_HUMAN.H11MO.0.B\n",
      "IRF4_HUMAN.H11MO.0.A\n",
      "IRF5_HUMAN.H11MO.0.D\n",
      "IRF6_HUMAN.MANUAL.0.A\n",
      "IRF7_HUMAN.H11MO.0.C\n",
      "IRF8_HUMAN.H11MO.0.B\n",
      "IRF9_HUMAN.H11MO.0.C\n",
      "IRX2_HUMAN.H11MO.0.D\n",
      "IRX3_HUMAN.H11MO.0.D\n",
      "ISL1_HUMAN.H11MO.0.A\n",
      "ISL2_HUMAN.H11MO.0.D\n",
      "ISX_HUMAN.H11MO.0.D\n",
      "ITF2_HUMAN.H11MO.0.C\n",
      "JDP2_HUMAN.H11MO.0.D\n",
      "JUNB_HUMAN.H11MO.0.A\n",
      "JUND_HUMAN.H11MO.0.A\n",
      "JUN_HUMAN.H11MO.0.A\n",
      "KAISO_HUMAN.H11MO.0.A\n",
      "KAISO_HUMAN.H11MO.1.A\n",
      "KAISO_HUMAN.H11MO.2.A\n",
      "KLF12_HUMAN.H11MO.0.C\n",
      "KLF13_HUMAN.H11MO.0.D\n",
      "KLF14_HUMAN.H11MO.0.D\n",
      "KLF15_HUMAN.H11MO.0.A\n",
      "KLF16_HUMAN.H11MO.0.D\n",
      "KLF1_HUMAN.H11MO.0.A\n",
      "KLF3_HUMAN.H11MO.0.B\n",
      "KLF4_HUMAN.H11MO.0.A\n",
      "KLF5_HUMAN.H11MO.0.A\n",
      "KLF6_HUMAN.H11MO.0.A\n",
      "KLF8_HUMAN.H11MO.0.C\n",
      "KLF9_HUMAN.H11MO.0.C\n",
      "LBX2_HUMAN.H11MO.0.D\n",
      "LEF1_HUMAN.H11MO.0.A\n",
      "LHX2_HUMAN.H11MO.0.A\n",
      "LHX3_HUMAN.H11MO.0.C\n",
      "LHX4_HUMAN.H11MO.0.D\n",
      "LHX6_HUMAN.H11MO.0.D\n",
      "LHX8_HUMAN.H11MO.0.D\n",
      "LHX9_HUMAN.H11MO.0.D\n",
      "LMX1A_HUMAN.H11MO.0.D\n",
      "LMX1B_HUMAN.H11MO.0.D\n",
      "LYL1_HUMAN.H11MO.0.A\n",
      "MAFA_HUMAN.H11MO.0.D\n",
      "MAFB_HUMAN.H11MO.0.B\n",
      "MAFF_HUMAN.H11MO.0.B\n",
      "MAFF_HUMAN.H11MO.1.B\n",
      "MAFG_HUMAN.H11MO.0.A\n",
      "MAFG_HUMAN.H11MO.1.A\n",
      "MAFK_HUMAN.H11MO.0.A\n",
      "MAFK_HUMAN.H11MO.1.A\n",
      "MAF_HUMAN.H11MO.0.A\n",
      "MAF_HUMAN.H11MO.1.B\n",
      "MAX_HUMAN.H11MO.0.A\n",
      "MAZ_HUMAN.H11MO.0.A\n",
      "MAZ_HUMAN.H11MO.1.A\n",
      "MBD2_HUMAN.H11MO.0.B\n",
      "MCR_HUMAN.H11MO.0.D\n",
      "MECP2_HUMAN.H11MO.0.C\n",
      "MEF2A_HUMAN.H11MO.0.A\n",
      "MEF2B_HUMAN.H11MO.0.A\n",
      "MEF2C_HUMAN.H11MO.0.A\n",
      "MEF2D_HUMAN.H11MO.0.A\n",
      "MEIS1_HUMAN.H11MO.0.A\n",
      "MEIS1_HUMAN.H11MO.1.B\n",
      "MEIS2_HUMAN.H11MO.0.B\n",
      "MEIS3_HUMAN.H11MO.0.D\n",
      "MEOX1_HUMAN.H11MO.0.D\n",
      "MEOX2_HUMAN.H11MO.0.D\n",
      "MESP1_HUMAN.H11MO.0.D\n",
      "MGAP_HUMAN.H11MO.0.D\n",
      "MITF_HUMAN.H11MO.0.A\n",
      "MIXL1_HUMAN.H11MO.0.D\n",
      "MLXPL_HUMAN.H11MO.0.D\n",
      "MLX_HUMAN.H11MO.0.D\n",
      "MNX1_HUMAN.H11MO.0.D\n",
      "MSX1_HUMAN.H11MO.0.D\n",
      "MSX2_HUMAN.H11MO.0.D\n",
      "MTF1_HUMAN.H11MO.0.C\n",
      "MXI1_HUMAN.H11MO.0.A\n",
      "MXI1_HUMAN.H11MO.1.A\n",
      "MYBA_HUMAN.H11MO.0.D\n",
      "MYBB_HUMAN.H11MO.0.D\n",
      "MYB_HUMAN.H11MO.0.A\n",
      "MYCN_HUMAN.H11MO.0.A\n",
      "MYC_HUMAN.H11MO.0.A\n",
      "MYF6_HUMAN.H11MO.0.C\n",
      "MYNN_HUMAN.H11MO.0.D\n",
      "MYOD1_HUMAN.H11MO.0.A\n",
      "MYOD1_HUMAN.H11MO.1.A\n",
      "MYOG_HUMAN.H11MO.0.B\n",
      "MZF1_HUMAN.H11MO.0.B\n",
      "NANOG_HUMAN.H11MO.0.A\n",
      "NANOG_HUMAN.H11MO.1.B\n",
      "NDF1_HUMAN.H11MO.0.A\n",
      "NDF2_HUMAN.H11MO.0.B\n",
      "NF2L1_HUMAN.H11MO.0.C\n",
      "NF2L2_HUMAN.H11MO.0.A\n",
      "NFAC1_HUMAN.H11MO.0.B\n",
      "NFAC1_HUMAN.H11MO.1.B\n",
      "NFAC2_HUMAN.H11MO.0.B\n",
      "NFAC3_HUMAN.H11MO.0.B\n",
      "NFAC4_HUMAN.H11MO.0.C\n",
      "NFAT5_HUMAN.H11MO.0.D\n",
      "NFE2_HUMAN.H11MO.0.A\n",
      "NFIA_HUMAN.H11MO.0.C\n",
      "NFIA_HUMAN.H11MO.1.D\n",
      "NFIB_HUMAN.H11MO.0.D\n",
      "NFIC_HUMAN.H11MO.0.A\n",
      "NFIC_HUMAN.H11MO.1.A\n",
      "NFIL3_HUMAN.H11MO.0.D\n",
      "NFKB1_HUMAN.H11MO.1.B\n",
      "NFKB2_HUMAN.H11MO.0.B\n",
      "NFYA_HUMAN.H11MO.0.A\n",
      "NFYB_HUMAN.H11MO.0.A\n",
      "NFYC_HUMAN.H11MO.0.A\n",
      "NGN2_HUMAN.H11MO.0.D\n",
      "NKX21_HUMAN.H11MO.0.A\n",
      "NKX22_HUMAN.H11MO.0.D\n",
      "NKX23_HUMAN.H11MO.0.D\n",
      "NKX25_HUMAN.H11MO.0.B\n",
      "NKX28_HUMAN.H11MO.0.C\n",
      "NKX31_HUMAN.H11MO.0.C\n",
      "NKX32_HUMAN.H11MO.0.C\n",
      "NKX61_HUMAN.H11MO.0.B\n",
      "NKX61_HUMAN.H11MO.1.B\n",
      "NKX62_HUMAN.H11MO.0.D\n",
      "NOBOX_HUMAN.H11MO.0.C\n",
      "NOTO_HUMAN.H11MO.0.D\n",
      "NR0B1_HUMAN.H11MO.0.D\n",
      "NR1D1_HUMAN.H11MO.0.B\n",
      "NR1D1_HUMAN.H11MO.1.D\n",
      "NR1H2_HUMAN.H11MO.0.D\n",
      "NR1H3_HUMAN.H11MO.0.B\n",
      "NR1H3_HUMAN.H11MO.1.B\n",
      "NR1H4_HUMAN.H11MO.0.B\n",
      "NR1H4_HUMAN.H11MO.1.B\n",
      "NR1I2_HUMAN.H11MO.0.C\n",
      "NR1I2_HUMAN.H11MO.1.D\n",
      "NR1I3_HUMAN.H11MO.0.C\n",
      "NR1I3_HUMAN.H11MO.1.D\n",
      "NR2C1_HUMAN.H11MO.0.C\n",
      "NR2C2_HUMAN.H11MO.0.B\n",
      "NR2E1_HUMAN.H11MO.0.D\n",
      "NR2E3_HUMAN.H11MO.0.C\n",
      "NR2F6_HUMAN.H11MO.0.D\n",
      "NR4A1_HUMAN.H11MO.0.A\n",
      "NR4A2_HUMAN.H11MO.0.C\n",
      "NR4A3_HUMAN.H11MO.0.D\n",
      "NR5A2_HUMAN.H11MO.0.B\n",
      "NR6A1_HUMAN.H11MO.0.B\n",
      "NRF1_HUMAN.H11MO.0.A\n",
      "NRL_HUMAN.H11MO.0.D\n",
      "OLIG1_HUMAN.H11MO.0.D\n",
      "OLIG2_HUMAN.H11MO.0.B\n",
      "OLIG2_HUMAN.H11MO.1.B\n",
      "OLIG3_HUMAN.H11MO.0.D\n",
      "ONEC2_HUMAN.H11MO.0.D\n",
      "ONEC3_HUMAN.H11MO.0.D\n",
      "OSR2_HUMAN.H11MO.0.C\n",
      "OTX1_HUMAN.H11MO.0.D\n",
      "OTX2_HUMAN.H11MO.0.A\n",
      "OVOL1_HUMAN.H11MO.0.C\n",
      "OVOL2_HUMAN.H11MO.0.D\n",
      "OZF_HUMAN.H11MO.0.C\n",
      "P53_HUMAN.H11MO.0.A\n",
      "P53_HUMAN.H11MO.1.A\n",
      "P5F1B_HUMAN.H11MO.0.D\n",
      "P63_HUMAN.H11MO.0.A\n",
      "P63_HUMAN.H11MO.1.A\n",
      "P73_HUMAN.H11MO.0.A\n",
      "P73_HUMAN.H11MO.1.A\n",
      "PATZ1_HUMAN.H11MO.0.C\n",
      "PATZ1_HUMAN.H11MO.1.C\n",
      "PAX1_HUMAN.H11MO.0.D\n",
      "PAX2_HUMAN.H11MO.0.D\n",
      "PAX3_HUMAN.H11MO.0.D\n",
      "PAX4_HUMAN.H11MO.0.D\n",
      "PAX5_HUMAN.H11MO.0.A\n",
      "PAX6_HUMAN.H11MO.0.C\n",
      "PAX7_HUMAN.H11MO.0.D\n",
      "PAX8_HUMAN.H11MO.0.D\n",
      "PBX1_HUMAN.H11MO.0.A\n",
      "PBX1_HUMAN.H11MO.1.C\n",
      "PBX2_HUMAN.H11MO.0.C\n",
      "PBX3_HUMAN.H11MO.0.A\n",
      "PBX3_HUMAN.H11MO.1.A\n",
      "PDX1_HUMAN.H11MO.0.A\n",
      "PDX1_HUMAN.H11MO.1.A\n",
      "PEBB_HUMAN.H11MO.0.C\n",
      "PHX2A_HUMAN.H11MO.0.D\n",
      "PHX2B_HUMAN.H11MO.0.D\n",
      "PIT1_HUMAN.H11MO.0.C\n",
      "PITX1_HUMAN.H11MO.0.D\n",
      "PITX2_HUMAN.H11MO.0.D\n",
      "PITX3_HUMAN.H11MO.0.D\n",
      "PKNX1_HUMAN.H11MO.0.B\n",
      "PLAG1_HUMAN.H11MO.0.D\n",
      "PLAL1_HUMAN.H11MO.0.D\n",
      "PO2F1_HUMAN.H11MO.0.C\n",
      "PO2F2_HUMAN.H11MO.0.A\n",
      "PO2F3_HUMAN.H11MO.0.D\n",
      "PO3F1_HUMAN.H11MO.0.C\n",
      "PO3F2_HUMAN.H11MO.0.A\n",
      "PO3F3_HUMAN.H11MO.0.D\n",
      "PO3F4_HUMAN.H11MO.0.D\n",
      "PO4F1_HUMAN.H11MO.0.D\n",
      "PO4F2_HUMAN.H11MO.0.D\n",
      "PO4F3_HUMAN.H11MO.0.D\n",
      "PO5F1_HUMAN.H11MO.0.A\n",
      "PO5F1_HUMAN.H11MO.1.A\n",
      "PO6F1_HUMAN.H11MO.0.D\n",
      "PO6F2_HUMAN.H11MO.0.D\n",
      "PPARA_HUMAN.H11MO.0.B\n",
      "PPARA_HUMAN.H11MO.1.B\n",
      "PPARD_HUMAN.H11MO.0.D\n",
      "PPARG_HUMAN.H11MO.0.A\n",
      "PPARG_HUMAN.H11MO.1.A\n",
      "PRD14_HUMAN.H11MO.0.A\n",
      "PRDM1_HUMAN.H11MO.0.A\n",
      "PRDM4_HUMAN.H11MO.0.D\n",
      "PRDM6_HUMAN.H11MO.0.C\n",
      "PRGR_HUMAN.H11MO.0.A\n",
      "PRGR_HUMAN.H11MO.1.A\n",
      "PROP1_HUMAN.H11MO.0.D\n",
      "PROX1_HUMAN.H11MO.0.D\n",
      "PRRX1_HUMAN.H11MO.0.D\n",
      "PRRX2_HUMAN.H11MO.0.C\n",
      "PTF1A_HUMAN.H11MO.0.B\n",
      "PTF1A_HUMAN.H11MO.1.B\n",
      "PURA_HUMAN.H11MO.0.D\n",
      "RARA_HUMAN.H11MO.0.A\n",
      "RARA_HUMAN.H11MO.1.A\n",
      "RARA_HUMAN.H11MO.2.A\n",
      "RARB_HUMAN.H11MO.0.D\n",
      "RARG_HUMAN.H11MO.0.B\n",
      "RARG_HUMAN.H11MO.1.B\n",
      "RARG_HUMAN.H11MO.2.D\n",
      "RAX2_HUMAN.H11MO.0.D\n",
      "RELB_HUMAN.H11MO.0.C\n",
      "REL_HUMAN.H11MO.0.B\n",
      "REST_HUMAN.H11MO.0.A\n",
      "RFX1_HUMAN.H11MO.0.B\n",
      "RFX1_HUMAN.H11MO.1.B\n",
      "RFX2_HUMAN.H11MO.0.A\n",
      "RFX2_HUMAN.H11MO.1.A\n",
      "RFX3_HUMAN.H11MO.0.B\n",
      "RFX4_HUMAN.H11MO.0.D\n",
      "RFX5_HUMAN.H11MO.0.A\n",
      "RFX5_HUMAN.H11MO.1.A\n",
      "RHXF1_HUMAN.H11MO.0.D\n",
      "RORA_HUMAN.H11MO.0.C\n",
      "RORG_HUMAN.H11MO.0.C\n",
      "RREB1_HUMAN.H11MO.0.D\n",
      "RUNX1_HUMAN.H11MO.0.A\n",
      "RUNX2_HUMAN.H11MO.0.A\n",
      "RUNX3_HUMAN.H11MO.0.A\n",
      "RXRA_HUMAN.H11MO.0.A\n",
      "RXRA_HUMAN.H11MO.1.A\n",
      "RXRB_HUMAN.H11MO.0.C\n",
      "RXRG_HUMAN.H11MO.0.B\n",
      "RX_HUMAN.H11MO.0.D\n",
      "SALL4_HUMAN.H11MO.0.B\n",
      "SCRT1_HUMAN.H11MO.0.D\n",
      "SCRT2_HUMAN.H11MO.0.D\n",
      "SHOX2_HUMAN.H11MO.0.D\n",
      "SHOX_HUMAN.H11MO.0.D\n",
      "SIX1_HUMAN.H11MO.0.A\n",
      "SIX2_HUMAN.H11MO.0.A\n",
      "SMAD1_HUMAN.H11MO.0.D\n",
      "SMAD2_HUMAN.H11MO.0.A\n",
      "SMAD3_HUMAN.H11MO.0.B\n",
      "SMAD4_HUMAN.H11MO.0.B\n",
      "SMCA1_HUMAN.H11MO.0.C\n",
      "SMCA5_HUMAN.H11MO.0.C\n",
      "SNAI1_HUMAN.H11MO.0.C\n",
      "SNAI2_HUMAN.H11MO.0.A\n",
      "SOX10_HUMAN.H11MO.0.B\n",
      "SOX10_HUMAN.H11MO.1.A\n",
      "SOX11_HUMAN.H11MO.0.D\n",
      "SOX13_HUMAN.H11MO.0.D\n",
      "SOX15_HUMAN.H11MO.0.D\n",
      "SOX17_HUMAN.H11MO.0.C\n",
      "SOX18_HUMAN.H11MO.0.D\n",
      "SOX1_HUMAN.H11MO.0.D\n",
      "SOX21_HUMAN.H11MO.0.D\n",
      "SOX2_HUMAN.H11MO.0.A\n",
      "SOX2_HUMAN.H11MO.1.A\n",
      "SOX3_HUMAN.H11MO.0.B\n",
      "SOX4_HUMAN.H11MO.0.B\n",
      "SOX5_HUMAN.H11MO.0.C\n",
      "SOX7_HUMAN.H11MO.0.D\n",
      "SOX8_HUMAN.H11MO.0.D\n",
      "SOX9_HUMAN.H11MO.0.B\n",
      "SOX9_HUMAN.H11MO.1.B\n",
      "SP1_HUMAN.H11MO.0.A\n",
      "SP1_HUMAN.H11MO.1.A\n",
      "SP2_HUMAN.H11MO.0.A\n",
      "SP2_HUMAN.H11MO.1.B\n",
      "SP3_HUMAN.H11MO.0.B\n",
      "SP4_HUMAN.H11MO.0.A\n",
      "SP4_HUMAN.H11MO.1.A\n",
      "SPDEF_HUMAN.H11MO.0.D\n",
      "SPI1_HUMAN.H11MO.0.A\n",
      "SPIB_HUMAN.H11MO.0.A\n",
      "SPIC_HUMAN.H11MO.0.D\n",
      "SPZ1_HUMAN.H11MO.0.D\n",
      "SRBP1_HUMAN.H11MO.0.A\n",
      "SRBP2_HUMAN.H11MO.0.B\n",
      "SRF_HUMAN.H11MO.0.A\n",
      "SRY_HUMAN.H11MO.0.B\n",
      "STA5A_HUMAN.H11MO.0.A\n",
      "STA5B_HUMAN.H11MO.0.A\n",
      "STAT1_HUMAN.H11MO.0.A\n",
      "STAT1_HUMAN.H11MO.1.A\n",
      "STAT2_HUMAN.H11MO.0.A\n",
      "STAT3_HUMAN.H11MO.0.A\n",
      "STAT4_HUMAN.H11MO.0.A\n",
      "STAT6_HUMAN.H11MO.0.B\n",
      "STF1_HUMAN.H11MO.0.B\n",
      "SUH_HUMAN.H11MO.0.A\n",
      "TAF1_HUMAN.H11MO.0.A\n",
      "TAL1_HUMAN.H11MO.0.A\n",
      "TAL1_HUMAN.H11MO.1.A\n",
      "TBP_HUMAN.H11MO.0.A\n",
      "TBR1_HUMAN.H11MO.0.D\n",
      "TBX15_HUMAN.H11MO.0.D\n",
      "TBX19_HUMAN.H11MO.0.D\n",
      "TBX1_HUMAN.H11MO.0.D\n",
      "TBX20_HUMAN.H11MO.0.D\n",
      "TBX21_HUMAN.H11MO.0.A\n",
      "TBX2_HUMAN.H11MO.0.D\n",
      "TBX3_HUMAN.H11MO.0.C\n",
      "TBX4_HUMAN.H11MO.0.D\n",
      "TBX5_HUMAN.H11MO.0.D\n",
      "TCF7_HUMAN.H11MO.0.A\n",
      "TEAD1_HUMAN.H11MO.0.A\n",
      "TEAD2_HUMAN.H11MO.0.D\n",
      "TEAD3_HUMAN.H11MO.0.D\n",
      "TEAD4_HUMAN.H11MO.0.A\n",
      "TEF_HUMAN.H11MO.0.D\n",
      "TF2LX_HUMAN.H11MO.0.D\n",
      "TF65_HUMAN.H11MO.0.A\n",
      "TF7L1_HUMAN.H11MO.0.B\n",
      "TF7L2_HUMAN.H11MO.0.A\n",
      "TFAP4_HUMAN.H11MO.0.A\n",
      "TFCP2_HUMAN.H11MO.0.D\n",
      "TFDP1_HUMAN.H11MO.0.C\n",
      "TFE2_HUMAN.H11MO.0.A\n",
      "TFE3_HUMAN.H11MO.0.B\n",
      "TFEB_HUMAN.H11MO.0.C\n",
      "TGIF1_HUMAN.H11MO.0.A\n",
      "TGIF2_HUMAN.H11MO.0.D\n",
      "THA11_HUMAN.H11MO.0.B\n",
      "THAP1_HUMAN.H11MO.0.C\n",
      "THA_HUMAN.H11MO.0.C\n",
      "THA_HUMAN.H11MO.1.D\n",
      "THB_HUMAN.H11MO.0.C\n",
      "THB_HUMAN.H11MO.1.D\n",
      "TLX1_HUMAN.H11MO.0.D\n",
      "TWST1_HUMAN.H11MO.0.A\n",
      "TWST1_HUMAN.H11MO.1.A\n",
      "TYY1_HUMAN.H11MO.0.A\n",
      "TYY2_HUMAN.H11MO.0.D\n",
      "UBIP1_HUMAN.H11MO.0.D\n",
      "UNC4_HUMAN.H11MO.0.D\n",
      "USF1_HUMAN.H11MO.0.A\n",
      "USF2_HUMAN.H11MO.0.A\n",
      "VAX1_HUMAN.H11MO.0.D\n",
      "VAX2_HUMAN.H11MO.0.D\n",
      "VDR_HUMAN.H11MO.0.A\n",
      "VDR_HUMAN.H11MO.1.A\n",
      "VENTX_HUMAN.H11MO.0.D\n",
      "VEZF1_HUMAN.H11MO.0.C\n",
      "VEZF1_HUMAN.H11MO.1.C\n",
      "VSX1_HUMAN.H11MO.0.D\n",
      "VSX2_HUMAN.H11MO.0.D\n",
      "WT1_HUMAN.H11MO.0.C\n",
      "WT1_HUMAN.H11MO.1.B\n",
      "XBP1_HUMAN.H11MO.0.D\n",
      "Z324A_HUMAN.H11MO.0.C\n",
      "Z354A_HUMAN.H11MO.0.C\n",
      "ZBED1_HUMAN.H11MO.0.D\n",
      "ZBT14_HUMAN.H11MO.0.C\n",
      "ZBT17_HUMAN.H11MO.0.A\n",
      "ZBT18_HUMAN.H11MO.0.C\n",
      "ZBT48_HUMAN.H11MO.0.C\n",
      "ZBT49_HUMAN.H11MO.0.D\n",
      "ZBT7A_HUMAN.H11MO.0.A\n",
      "ZBT7B_HUMAN.H11MO.0.D\n",
      "ZBTB4_HUMAN.H11MO.0.D\n",
      "ZBTB4_HUMAN.H11MO.1.D\n",
      "ZBTB6_HUMAN.H11MO.0.C\n",
      "ZEB1_HUMAN.H11MO.0.A\n",
      "ZEP1_HUMAN.H11MO.0.D\n",
      "ZEP2_HUMAN.H11MO.0.D\n",
      "ZF64A_HUMAN.H11MO.0.D\n",
      "ZFHX3_HUMAN.H11MO.0.D\n",
      "ZFP28_HUMAN.H11MO.0.C\n",
      "ZFP42_HUMAN.H11MO.0.A\n",
      "ZFP82_HUMAN.H11MO.0.C\n",
      "ZFX_HUMAN.H11MO.0.A\n",
      "ZFX_HUMAN.H11MO.1.A\n",
      "ZIC1_HUMAN.H11MO.0.B\n",
      "ZIC2_HUMAN.H11MO.0.D\n",
      "ZIC3_HUMAN.H11MO.0.B\n",
      "ZIC4_HUMAN.H11MO.0.D\n",
      "ZIM3_HUMAN.H11MO.0.C\n",
      "ZKSC1_HUMAN.H11MO.0.B\n",
      "ZKSC3_HUMAN.H11MO.0.D\n",
      "ZN121_HUMAN.H11MO.0.C\n",
      "ZN134_HUMAN.H11MO.0.C\n",
      "ZN134_HUMAN.H11MO.1.C\n",
      "ZN136_HUMAN.H11MO.0.C\n",
      "ZN140_HUMAN.H11MO.0.C\n",
      "ZN143_HUMAN.H11MO.0.A\n",
      "ZN148_HUMAN.H11MO.0.D\n",
      "ZN214_HUMAN.H11MO.0.C\n",
      "ZN219_HUMAN.H11MO.0.D\n",
      "ZN232_HUMAN.H11MO.0.D\n",
      "ZN250_HUMAN.H11MO.0.C\n",
      "ZN257_HUMAN.H11MO.0.C\n",
      "ZN260_HUMAN.H11MO.0.C\n",
      "ZN263_HUMAN.H11MO.0.A\n",
      "ZN263_HUMAN.H11MO.1.A\n",
      "ZN264_HUMAN.H11MO.0.C\n",
      "ZN274_HUMAN.H11MO.0.A\n",
      "ZN281_HUMAN.H11MO.0.A\n",
      "ZN282_HUMAN.H11MO.0.D\n",
      "ZN317_HUMAN.H11MO.0.C\n",
      "ZN320_HUMAN.H11MO.0.C\n",
      "ZN322_HUMAN.H11MO.0.B\n",
      "ZN329_HUMAN.H11MO.0.C\n",
      "ZN331_HUMAN.H11MO.0.C\n",
      "ZN333_HUMAN.H11MO.0.D\n",
      "ZN335_HUMAN.H11MO.0.A\n",
      "ZN335_HUMAN.H11MO.1.A\n",
      "ZN341_HUMAN.H11MO.0.C\n",
      "ZN341_HUMAN.H11MO.1.C\n",
      "ZN350_HUMAN.H11MO.0.C\n",
      "ZN350_HUMAN.H11MO.1.D\n",
      "ZN382_HUMAN.H11MO.0.C\n",
      "ZN384_HUMAN.H11MO.0.C\n",
      "ZN394_HUMAN.H11MO.0.C\n",
      "ZN394_HUMAN.H11MO.1.D\n",
      "ZN410_HUMAN.H11MO.0.D\n",
      "ZN418_HUMAN.H11MO.0.C\n",
      "ZN418_HUMAN.H11MO.1.D\n",
      "ZN423_HUMAN.H11MO.0.D\n",
      "ZN436_HUMAN.H11MO.0.C\n",
      "ZN449_HUMAN.H11MO.0.C\n",
      "ZN467_HUMAN.H11MO.0.C\n",
      "ZN490_HUMAN.H11MO.0.C\n",
      "ZN502_HUMAN.H11MO.0.C\n",
      "ZN524_HUMAN.H11MO.0.D\n",
      "ZN528_HUMAN.H11MO.0.C\n",
      "ZN547_HUMAN.H11MO.0.C\n",
      "ZN549_HUMAN.H11MO.0.C\n",
      "ZN554_HUMAN.H11MO.0.C\n",
      "ZN554_HUMAN.H11MO.1.D\n",
      "ZN563_HUMAN.H11MO.0.C\n",
      "ZN563_HUMAN.H11MO.1.C\n",
      "ZN582_HUMAN.H11MO.0.C\n",
      "ZN586_HUMAN.H11MO.0.C\n",
      "ZN589_HUMAN.H11MO.0.D\n",
      "ZN652_HUMAN.H11MO.0.D\n",
      "ZN667_HUMAN.H11MO.0.C\n",
      "ZN680_HUMAN.H11MO.0.C\n",
      "ZN708_HUMAN.H11MO.0.C\n",
      "ZN708_HUMAN.H11MO.1.D\n",
      "ZN713_HUMAN.H11MO.0.D\n",
      "ZN740_HUMAN.H11MO.0.D\n",
      "ZN768_HUMAN.H11MO.0.C\n",
      "ZN770_HUMAN.H11MO.0.C\n",
      "ZN770_HUMAN.H11MO.1.C\n",
      "ZN784_HUMAN.H11MO.0.D\n",
      "ZN816_HUMAN.H11MO.0.C\n",
      "ZN816_HUMAN.H11MO.1.C\n",
      "ZNF18_HUMAN.H11MO.0.C\n",
      "ZNF41_HUMAN.H11MO.0.C\n",
      "ZNF41_HUMAN.H11MO.1.C\n",
      "ZNF76_HUMAN.H11MO.0.C\n",
      "ZNF85_HUMAN.H11MO.0.C\n",
      "ZNF85_HUMAN.H11MO.1.C\n",
      "ZNF8_HUMAN.H11MO.0.C\n",
      "ZSC16_HUMAN.H11MO.0.D\n",
      "ZSC22_HUMAN.H11MO.0.C\n",
      "ZSC31_HUMAN.H11MO.0.C\n"
     ]
    }
   ],
   "source": [
    "with open(output_jaspar_file, 'w') as f:\n",
    "    for file in sorted(glob.glob('../data/external/hocomoco_human_trim/*')):\n",
    "        motif_id = os.path.basename(file).split('.pwm')[0]\n",
    "        f.write('>' + motif_id + '\\n')\n",
    "        with open(file, 'r') as g:\n",
    "            print(motif_id)\n",
    "            for line in g:\n",
    "                f.write('\\t'.join(line.strip().split())+'\\n')\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
