{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 8B_footprint_target_genes\n",
    "\n",
    "7/07/2021\n",
    "\n",
    "Can I get the gene lists for TF footprints of ATF4 and CEBPG in KCD6 and Prostate pro-loops? \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pybedtools\n",
    "import glob, os\n",
    "import pandas as pd\n",
    "\n",
    "save_dir = '../data/processed/fig7_vocab_go/'\n",
    "vocab_geneset_dir = '../data/processed/fig7_vocab_go/vocab_geneset_all_tissue'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# tfs \n",
    "tf_annon_df = pd.read_csv('../data/external/HOCOMOCOv11_annotation.csv',index_col=0)\n",
    "tf_annon_df['id_trim'] = tf_annon_df['id'] + '.pwm.trim'\n",
    "tf_name_to_id_dict = pd.Series(tf_annon_df.id_trim.values, index=tf_annon_df.tf.values).to_dict()\n",
    "tf_id_to_name_dict = pd.Series(tf_annon_df.tf.values, index=tf_annon_df.id_trim.values).to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_instances(vocab, tissue ,anc_prom, loop_df, anc_foot, pro_anc_foot_df, gene=None,verbose=False):\n",
    "    if verbose:\n",
    "        print('*** getting instances for gene: ', gene)\n",
    "        print('anc_prom', anc_prom.shape)\n",
    "    \n",
    "    # get anchors that loop to promoter anchors\n",
    "    if gene is not None:\n",
    "        anc_prom_filt = anc_prom[anc_prom.name==gene]\n",
    "    else:\n",
    "        anc_prom_filt = anc_prom\n",
    "    anc_prom_arr = anc_prom_filt.thickEnd.unique()\n",
    "    if verbose:\n",
    "        print('num promoter anchors', len(anc_prom_arr))\n",
    "    \n",
    "    # get loops\n",
    "    loop_df_filt = loop_df[loop_df.source.isin(anc_prom_arr)|loop_df.target.isin(anc_prom_arr)]\n",
    "    all_anc = sorted(set(anc_prom_filt.thickEnd.to_list() +loop_df_filt.source.to_list() + loop_df_filt.target.to_list() ))\n",
    "    if verbose:\n",
    "        print('num all anchors', len(all_anc))#all_anc)\n",
    "        display(loop_df)\n",
    "    \n",
    "    # get footprinting anchor data\n",
    "    anc_foot_filt = anc_foot[anc_foot.name.isin(all_anc)]\n",
    "    if verbose:\n",
    "        print('number of footprinted anchors with vocab',anc_foot.shape)\n",
    "    \n",
    "#     # get promoter footprint data\n",
    "#     pro_anc_foot = pd.DataFrame()\n",
    "#     if gene is not None:\n",
    "#         pro_anc_foot1 = pro_anc_foot_df[(pro_anc_foot_df.name==gene)]# not perfect will catch later???\n",
    "#         pro_anc_foot = pd.concat([pro_anc_foot, pro_anc_foot1])\n",
    "\n",
    "    genomic_instances = anc_foot_filt#pd.concat([anc_foot_filt,pro_anc_foot])\n",
    "    genomic_instances.columns = ['chr', 'start', 'stop', 'TSS', 'chr_m', 'start_m', 'stop_m', 'id_trim',\n",
    "       'score', 'strand', 'qual','tf']\n",
    "    genomic_instances.drop('qual',axis=1,inplace=True)\n",
    "    genomic_instances['vocab'] = vocab\n",
    "    genomic_instances['tissue'] = tissue\n",
    "    genomic_instances['config'] = 'pro_loop'\n",
    "    genomic_instances['gene'] = gene\n",
    "    return genomic_instances\n",
    "    \n",
    "\n",
    "def get_genomic_instances_pro_loop(vocab, tissue, list_genes,verbose=False, \n",
    "                          tf_id_to_name_dict=tf_id_to_name_dict,\n",
    "                          save_dir=None):\n",
    "    # get genes with this vocab (any configuraton)\n",
    "    if verbose:\n",
    "        print('=======================')\n",
    "        print( vocab, tissue)\n",
    "        print(list_genes)\n",
    "        print(len(list_genes))\n",
    "        print('loading files...')\n",
    "    # get files\n",
    "    anc_prom = pybedtools.BedTool('../data/interim/annon/promoter_anchors/promoter_'+tissue+'_annon.bed').to_dataframe()\n",
    "    loop_df = pd.read_csv('../data/interim/merged/loops/'+tissue+'.loops.csv',index_col=0)\n",
    "    anc_foot = pybedtools.BedTool('../data/interim/annon/anchor_footprinting/'+tissue+'_annon.bed').to_dataframe()\n",
    "    anc_foot['tf'] = anc_foot.thickEnd.map(tf_id_to_name_dict)\n",
    "    anc_foot = anc_foot[anc_foot.tf.isin(vocab.split('::'))]\n",
    "    pro_anc_foot_df = pybedtools.BedTool('../data/interim/annon/promoter_footprinting/promoter_'+tissue+'_annon.bed').to_dataframe()\n",
    "    pro_anc_foot_df['tf'] = pro_anc_foot_df.thickEnd.map(tf_id_to_name_dict)\n",
    "\n",
    "    genomic_instances_all = pd.DataFrame()\n",
    "    \n",
    "    if list_genes is not None:\n",
    "        for gene in list_genes:\n",
    " \n",
    "            genomic_instances = get_instances(vocab, tissue, anc_prom, loop_df, anc_foot, pro_anc_foot_df, gene,verbose=verbose)\n",
    "            genomic_instances_all = pd.concat([genomic_instances_all, genomic_instances])\n",
    "    else:\n",
    "        genomic_instances_all = get_instances(vocab, tissue, anc_prom, loop_df, anc_foot, pro_anc_foot_df, gene,verbose=verbose)\n",
    "\n",
    "\n",
    "    \n",
    "    if verbose:\n",
    "        print('*** genomic instances',genomic_instances_all.shape)\n",
    "        display(genomic_instances_all)\n",
    "    if save_dir is not None:\n",
    "        genomic_instances_all.to_csv(os.path.join(save_dir, '_'.join(vocab.split('::'))+'_genomic_instances.csv'))\n",
    "    return genomic_instances_all\n",
    "\n",
    "\n",
    "\n",
    "def get_target_genes(tf,tissue, not_tissues=None, verbose=False):\n",
    "    vocab_df = pd.read_csv('../data/processed/tissue_crms/pro_loop_tissue/'+tissue+'_crm.csv',index_col=0)\n",
    "    target_genes =  vocab_df.index[vocab_df[tf+'_loop']>0].values\n",
    "    print(len(target_genes), 'target genes found for', tf, 'in tissue: ', tissue)\n",
    "    if not_tissues is not None:\n",
    "        not_target_genes = []\n",
    "        for t in not_tissues:\n",
    "            not_vocab_df = pd.read_csv('../data/processed/tissue_crms/pro_loop_tissue/'+t+'_crm.csv',index_col=0)\n",
    "            not_target_genes += not_vocab_df.index[not_vocab_df[tf+'_loop']>0].to_list()\n",
    "        not_target_genes = set(not_target_genes)\n",
    "        print(len(not_target_genes), 'off-target genes found for', tf, 'in tissues: ', not_tissues)\n",
    "    target_genes =  sorted(set(target_genes) - not_target_genes)\n",
    "    print(len(target_genes), 'filtered target genes found for', tf, 'in tissue: ', tissue)\n",
    "    \n",
    "    # get genomic instances\n",
    "    df_all_instances = get_genomic_instances_pro_loop(tf, tissue, list_genes=target_genes,verbose=verbose)\n",
    "    print(len(df_all_instances.gene.unique()), 'final target genes with instances found for', tf, 'in tissue: ', tissue)\n",
    "\n",
    "#     df = df_all_instances[~df_all_instances.TSS.str.startswith('chr')]\n",
    "#     if len(df)>0:\n",
    "#         print(len(df.TSS.unique()), 'final target genes with instances found for', tf, 'in tissue: ', tissue)\n",
    "#     else:\n",
    "#         print('no target genes found for ', tf, 'in tissue: ',tissue)\n",
    "#     print(sorted(df.TSS.unique()))\n",
    "    print('done...')\n",
    "    return df_all_instances"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<!-- Can you get me the footprint DNA sequence and genomic location of the footrpint CEBPG and ATF4 for STK11? It potentially can work...\n",
    "\n",
    " -->\n",
    " \n",
    " basically what we want is we want target genes with loops in one tissue (KCD6) and not in other tissue (prostate) and vice versa where the differential loop in both cases uses the same motif"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get_genomic_instances_pro_loop('CEBPG', 'Prostate', list_genes=['STK11'])\n",
    "# get_genomic_instances_pro_loop('ATF4', 'Prostate', list_genes=['STK11'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'atf4_kcd6' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-5-b443ceebe897>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0matf4_kcd6\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgene\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'atf4_kcd6' is not defined"
     ]
    }
   ],
   "source": [
    "atf4_kcd6.gene.unique().shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "atf4_kcd6 = get_target_genes('ATF4', 'GDSD6', not_tissues=['Prostate'])\n",
    "atf4_pros = get_target_genes('ATF4', 'Prostate', not_tissues=['GDSD6'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1065 target genes found for CEBPG in tissue:  GDSD6\n",
      "376 off-target genes found for CEBPG in tissues:  ['Prostate']\n",
      "887 filtered target genes found for CEBPG in tissue:  GDSD6\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/pandas/core/frame.py:4102: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  errors=errors,\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:37: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:38: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:39: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:40: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "887 final target genes with instances found for CEBPG in tissue:  GDSD6\n",
      "done...\n",
      "376 target genes found for CEBPG in tissue:  Prostate\n",
      "1065 off-target genes found for CEBPG in tissues:  ['GDSD6']\n",
      "198 filtered target genes found for CEBPG in tissue:  Prostate\n",
      "198 final target genes with instances found for CEBPG in tissue:  Prostate\n",
      "done...\n"
     ]
    }
   ],
   "source": [
    "cebpg_kcd6 = get_target_genes('CEBPG', 'GDSD6', not_tissues=['Prostate'])\n",
    "cebpg_pros = get_target_genes('CEBPG', 'Prostate', not_tissues=['GDSD6'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "764 target genes found for SMAD4 in tissue:  GDSD6\n",
      "456 off-target genes found for SMAD4 in tissues:  ['Prostate']\n",
      "566 filtered target genes found for SMAD4 in tissue:  GDSD6\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:37: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:38: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:39: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:40: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "566 final target genes with instances found for SMAD4 in tissue:  GDSD6\n",
      "done...\n",
      "456 target genes found for SMAD4 in tissue:  Prostate\n",
      "764 off-target genes found for SMAD4 in tissues:  ['GDSD6']\n",
      "258 filtered target genes found for SMAD4 in tissue:  Prostate\n",
      "258 final target genes with instances found for SMAD4 in tissue:  Prostate\n",
      "done...\n"
     ]
    }
   ],
   "source": [
    "smad4_kcd6 = get_target_genes('SMAD4', 'GDSD6', not_tissues=['Prostate'])\n",
    "smad4_pros = get_target_genes('SMAD4', 'Prostate', not_tissues=['GDSD6'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "atf4_kcd6.to_csv('../data/processed/fig7_vocab_go/genomic_instance_overlap/ATF4_kcd6_manual.csv')\n",
    "atf4_pros.to_csv('../data/processed/fig7_vocab_go/genomic_instance_overlap/ATF4_prostate_manual.csv')\n",
    "cebpg_kcd6.to_csv('../data/processed/fig7_vocab_go/genomic_instance_overlap/CEBPG_kcd6_manual.csv')\n",
    "cebpg_pros.to_csv('../data/processed/fig7_vocab_go/genomic_instance_overlap/CEBPG_prostate_manual.csv')\n",
    "smad4_kcd6.to_csv('../data/processed/fig7_vocab_go/genomic_instance_overlap/SMAD4_kcd6_manual.csv')\n",
    "smad4_pros.to_csv('../data/processed/fig7_vocab_go/genomic_instance_overlap/SMAD4_prostate_manual.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:37: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:38: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:39: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:40: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>TSS</th>\n",
       "      <th>chr_m</th>\n",
       "      <th>start_m</th>\n",
       "      <th>stop_m</th>\n",
       "      <th>id_trim</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>tf</th>\n",
       "      <th>vocab</th>\n",
       "      <th>tissue</th>\n",
       "      <th>config</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>1222741</td>\n",
       "      <td>chr2</td>\n",
       "      <td>37550000</td>\n",
       "      <td>37555000</td>\n",
       "      <td>chr2_37550000_37555000</td>\n",
       "      <td>chr2</td>\n",
       "      <td>37551938</td>\n",
       "      <td>37551959</td>\n",
       "      <td>SP1_HUMAN.H11MO.0.A.pwm.trim</td>\n",
       "      <td>12.651939</td>\n",
       "      <td>-</td>\n",
       "      <td>SP1</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>SCC13-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "      <td>CYP1B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1223261</td>\n",
       "      <td>chr2</td>\n",
       "      <td>37895000</td>\n",
       "      <td>37900000</td>\n",
       "      <td>chr2_37895000_37900000</td>\n",
       "      <td>chr2</td>\n",
       "      <td>37898508</td>\n",
       "      <td>37898518</td>\n",
       "      <td>SP1_HUMAN.H11MO.1.A.pwm.trim</td>\n",
       "      <td>10.291843</td>\n",
       "      <td>+</td>\n",
       "      <td>SP1</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>SCC13-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "      <td>CYP1B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1223298</td>\n",
       "      <td>chr2</td>\n",
       "      <td>37895000</td>\n",
       "      <td>37900000</td>\n",
       "      <td>chr2_37895000_37900000</td>\n",
       "      <td>chr2</td>\n",
       "      <td>37898822</td>\n",
       "      <td>37898843</td>\n",
       "      <td>SP1_HUMAN.H11MO.0.A.pwm.trim</td>\n",
       "      <td>22.014516</td>\n",
       "      <td>+</td>\n",
       "      <td>SP1</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>SCC13-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "      <td>CYP1B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1223307</td>\n",
       "      <td>chr2</td>\n",
       "      <td>37895000</td>\n",
       "      <td>37900000</td>\n",
       "      <td>chr2_37895000_37900000</td>\n",
       "      <td>chr2</td>\n",
       "      <td>37898823</td>\n",
       "      <td>37898833</td>\n",
       "      <td>SP1_HUMAN.H11MO.1.A.pwm.trim</td>\n",
       "      <td>15.157784</td>\n",
       "      <td>+</td>\n",
       "      <td>SP1</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>SCC13-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "      <td>CYP1B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1223321</td>\n",
       "      <td>chr2</td>\n",
       "      <td>37895000</td>\n",
       "      <td>37900000</td>\n",
       "      <td>chr2_37895000_37900000</td>\n",
       "      <td>chr2</td>\n",
       "      <td>37898825</td>\n",
       "      <td>37898846</td>\n",
       "      <td>SP1_HUMAN.H11MO.0.A.pwm.trim</td>\n",
       "      <td>7.886173</td>\n",
       "      <td>+</td>\n",
       "      <td>SP1</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>SCC13-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "      <td>CYP1B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1223327</td>\n",
       "      <td>chr2</td>\n",
       "      <td>37895000</td>\n",
       "      <td>37900000</td>\n",
       "      <td>chr2_37895000_37900000</td>\n",
       "      <td>chr2</td>\n",
       "      <td>37898827</td>\n",
       "      <td>37898848</td>\n",
       "      <td>SP1_HUMAN.H11MO.0.A.pwm.trim</td>\n",
       "      <td>17.743111</td>\n",
       "      <td>+</td>\n",
       "      <td>SP1</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>SCC13-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "      <td>CYP1B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1223332</td>\n",
       "      <td>chr2</td>\n",
       "      <td>37895000</td>\n",
       "      <td>37900000</td>\n",
       "      <td>chr2_37895000_37900000</td>\n",
       "      <td>chr2</td>\n",
       "      <td>37898828</td>\n",
       "      <td>37898838</td>\n",
       "      <td>SP1_HUMAN.H11MO.1.A.pwm.trim</td>\n",
       "      <td>12.310072</td>\n",
       "      <td>+</td>\n",
       "      <td>SP1</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>SCC13-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "      <td>CYP1B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1223340</td>\n",
       "      <td>chr2</td>\n",
       "      <td>37895000</td>\n",
       "      <td>37900000</td>\n",
       "      <td>chr2_37895000_37900000</td>\n",
       "      <td>chr2</td>\n",
       "      <td>37898830</td>\n",
       "      <td>37898851</td>\n",
       "      <td>SP1_HUMAN.H11MO.0.A.pwm.trim</td>\n",
       "      <td>7.260751</td>\n",
       "      <td>+</td>\n",
       "      <td>SP1</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>SCC13-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "      <td>CYP1B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1223344</td>\n",
       "      <td>chr2</td>\n",
       "      <td>37895000</td>\n",
       "      <td>37900000</td>\n",
       "      <td>chr2_37895000_37900000</td>\n",
       "      <td>chr2</td>\n",
       "      <td>37898832</td>\n",
       "      <td>37898853</td>\n",
       "      <td>SP1_HUMAN.H11MO.0.A.pwm.trim</td>\n",
       "      <td>16.885970</td>\n",
       "      <td>+</td>\n",
       "      <td>SP1</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>SCC13-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "      <td>CYP1B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1223347</td>\n",
       "      <td>chr2</td>\n",
       "      <td>37895000</td>\n",
       "      <td>37900000</td>\n",
       "      <td>chr2_37895000_37900000</td>\n",
       "      <td>chr2</td>\n",
       "      <td>37898833</td>\n",
       "      <td>37898843</td>\n",
       "      <td>SP1_HUMAN.H11MO.1.A.pwm.trim</td>\n",
       "      <td>12.682930</td>\n",
       "      <td>+</td>\n",
       "      <td>SP1</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>SCC13-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "      <td>CYP1B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1223359</td>\n",
       "      <td>chr2</td>\n",
       "      <td>37895000</td>\n",
       "      <td>37900000</td>\n",
       "      <td>chr2_37895000_37900000</td>\n",
       "      <td>chr2</td>\n",
       "      <td>37898837</td>\n",
       "      <td>37898858</td>\n",
       "      <td>SP1_HUMAN.H11MO.0.A.pwm.trim</td>\n",
       "      <td>24.419791</td>\n",
       "      <td>+</td>\n",
       "      <td>SP1</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>SCC13-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "      <td>CYP1B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1223369</td>\n",
       "      <td>chr2</td>\n",
       "      <td>37895000</td>\n",
       "      <td>37900000</td>\n",
       "      <td>chr2_37895000_37900000</td>\n",
       "      <td>chr2</td>\n",
       "      <td>37898838</td>\n",
       "      <td>37898848</td>\n",
       "      <td>SP1_HUMAN.H11MO.1.A.pwm.trim</td>\n",
       "      <td>18.026685</td>\n",
       "      <td>+</td>\n",
       "      <td>SP1</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>SCC13-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "      <td>CYP1B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1223401</td>\n",
       "      <td>chr2</td>\n",
       "      <td>37895000</td>\n",
       "      <td>37900000</td>\n",
       "      <td>chr2_37895000_37900000</td>\n",
       "      <td>chr2</td>\n",
       "      <td>37898873</td>\n",
       "      <td>37898883</td>\n",
       "      <td>SP1_HUMAN.H11MO.1.A.pwm.trim</td>\n",
       "      <td>11.706440</td>\n",
       "      <td>+</td>\n",
       "      <td>SP1</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>SCC13-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "      <td>CYP1B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1223939</td>\n",
       "      <td>chr2</td>\n",
       "      <td>38300000</td>\n",
       "      <td>38305000</td>\n",
       "      <td>chr2_38300000_38305000</td>\n",
       "      <td>chr2</td>\n",
       "      <td>38304040</td>\n",
       "      <td>38304061</td>\n",
       "      <td>SP1_HUMAN.H11MO.0.A.pwm.trim</td>\n",
       "      <td>13.623659</td>\n",
       "      <td>+</td>\n",
       "      <td>SP1</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>SCC13-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "      <td>CYP1B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1223943</td>\n",
       "      <td>chr2</td>\n",
       "      <td>38300000</td>\n",
       "      <td>38305000</td>\n",
       "      <td>chr2_38300000_38305000</td>\n",
       "      <td>chr2</td>\n",
       "      <td>38304041</td>\n",
       "      <td>38304051</td>\n",
       "      <td>SP1_HUMAN.H11MO.1.A.pwm.trim</td>\n",
       "      <td>10.828229</td>\n",
       "      <td>+</td>\n",
       "      <td>SP1</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>SCC13-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "      <td>CYP1B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1224510</td>\n",
       "      <td>chr2</td>\n",
       "      <td>38600000</td>\n",
       "      <td>38605000</td>\n",
       "      <td>chr2_38600000_38605000</td>\n",
       "      <td>chr2</td>\n",
       "      <td>38604474</td>\n",
       "      <td>38604495</td>\n",
       "      <td>SP1_HUMAN.H11MO.0.A.pwm.trim</td>\n",
       "      <td>6.759938</td>\n",
       "      <td>+</td>\n",
       "      <td>SP1</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>SCC13-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "      <td>CYP1B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1224536</td>\n",
       "      <td>chr2</td>\n",
       "      <td>38600000</td>\n",
       "      <td>38605000</td>\n",
       "      <td>chr2_38600000_38605000</td>\n",
       "      <td>chr2</td>\n",
       "      <td>38604665</td>\n",
       "      <td>38604675</td>\n",
       "      <td>SP1_HUMAN.H11MO.1.A.pwm.trim</td>\n",
       "      <td>15.225751</td>\n",
       "      <td>+</td>\n",
       "      <td>SP1</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>SCC13-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "      <td>CYP1B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1224994</td>\n",
       "      <td>chr2</td>\n",
       "      <td>38830000</td>\n",
       "      <td>38835000</td>\n",
       "      <td>chr2_38830000_38835000</td>\n",
       "      <td>chr2</td>\n",
       "      <td>38830304</td>\n",
       "      <td>38830325</td>\n",
       "      <td>SP1_HUMAN.H11MO.0.A.pwm.trim</td>\n",
       "      <td>7.553580</td>\n",
       "      <td>-</td>\n",
       "      <td>SP1</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>SCC13-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "      <td>CYP1B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1225014</td>\n",
       "      <td>chr2</td>\n",
       "      <td>38830000</td>\n",
       "      <td>38835000</td>\n",
       "      <td>chr2_38830000_38835000</td>\n",
       "      <td>chr2</td>\n",
       "      <td>38830675</td>\n",
       "      <td>38830696</td>\n",
       "      <td>SP1_HUMAN.H11MO.0.A.pwm.trim</td>\n",
       "      <td>8.006493</td>\n",
       "      <td>+</td>\n",
       "      <td>SP1</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>SCC13-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "      <td>CYP1B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1225025</td>\n",
       "      <td>chr2</td>\n",
       "      <td>38830000</td>\n",
       "      <td>38835000</td>\n",
       "      <td>chr2_38830000_38835000</td>\n",
       "      <td>chr2</td>\n",
       "      <td>38830686</td>\n",
       "      <td>38830696</td>\n",
       "      <td>SP1_HUMAN.H11MO.1.A.pwm.trim</td>\n",
       "      <td>11.896678</td>\n",
       "      <td>-</td>\n",
       "      <td>SP1</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>SCC13-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "      <td>CYP1B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1225211</td>\n",
       "      <td>chr2</td>\n",
       "      <td>38975000</td>\n",
       "      <td>38980000</td>\n",
       "      <td>chr2_38975000_38980000</td>\n",
       "      <td>chr2</td>\n",
       "      <td>38978059</td>\n",
       "      <td>38978080</td>\n",
       "      <td>SP1_HUMAN.H11MO.0.A.pwm.trim</td>\n",
       "      <td>18.964402</td>\n",
       "      <td>+</td>\n",
       "      <td>SP1</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>SCC13-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "      <td>CYP1B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1225227</td>\n",
       "      <td>chr2</td>\n",
       "      <td>38975000</td>\n",
       "      <td>38980000</td>\n",
       "      <td>chr2_38975000_38980000</td>\n",
       "      <td>chr2</td>\n",
       "      <td>38978060</td>\n",
       "      <td>38978070</td>\n",
       "      <td>SP1_HUMAN.H11MO.1.A.pwm.trim</td>\n",
       "      <td>12.356850</td>\n",
       "      <td>+</td>\n",
       "      <td>SP1</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>SCC13-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "      <td>CYP1B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1225272</td>\n",
       "      <td>chr2</td>\n",
       "      <td>38975000</td>\n",
       "      <td>38980000</td>\n",
       "      <td>chr2_38975000_38980000</td>\n",
       "      <td>chr2</td>\n",
       "      <td>38978063</td>\n",
       "      <td>38978084</td>\n",
       "      <td>SP1_HUMAN.H11MO.0.A.pwm.trim</td>\n",
       "      <td>10.673023</td>\n",
       "      <td>+</td>\n",
       "      <td>SP1</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>SCC13-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "      <td>CYP1B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1225291</td>\n",
       "      <td>chr2</td>\n",
       "      <td>38975000</td>\n",
       "      <td>38980000</td>\n",
       "      <td>chr2_38975000_38980000</td>\n",
       "      <td>chr2</td>\n",
       "      <td>38978065</td>\n",
       "      <td>38978086</td>\n",
       "      <td>SP1_HUMAN.H11MO.0.A.pwm.trim</td>\n",
       "      <td>9.500081</td>\n",
       "      <td>+</td>\n",
       "      <td>SP1</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>SCC13-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "      <td>CYP1B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1225309</td>\n",
       "      <td>chr2</td>\n",
       "      <td>38975000</td>\n",
       "      <td>38980000</td>\n",
       "      <td>chr2_38975000_38980000</td>\n",
       "      <td>chr2</td>\n",
       "      <td>38978067</td>\n",
       "      <td>38978088</td>\n",
       "      <td>SP1_HUMAN.H11MO.0.A.pwm.trim</td>\n",
       "      <td>6.819492</td>\n",
       "      <td>+</td>\n",
       "      <td>SP1</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>SCC13-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "      <td>CYP1B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1225317</td>\n",
       "      <td>chr2</td>\n",
       "      <td>38975000</td>\n",
       "      <td>38980000</td>\n",
       "      <td>chr2_38975000_38980000</td>\n",
       "      <td>chr2</td>\n",
       "      <td>38978068</td>\n",
       "      <td>38978089</td>\n",
       "      <td>SP1_HUMAN.H11MO.0.A.pwm.trim</td>\n",
       "      <td>14.015691</td>\n",
       "      <td>+</td>\n",
       "      <td>SP1</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>SCC13-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "      <td>CYP1B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1225325</td>\n",
       "      <td>chr2</td>\n",
       "      <td>38975000</td>\n",
       "      <td>38980000</td>\n",
       "      <td>chr2_38975000_38980000</td>\n",
       "      <td>chr2</td>\n",
       "      <td>38978069</td>\n",
       "      <td>38978079</td>\n",
       "      <td>SP1_HUMAN.H11MO.1.A.pwm.trim</td>\n",
       "      <td>10.581895</td>\n",
       "      <td>+</td>\n",
       "      <td>SP1</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>SCC13-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "      <td>CYP1B1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr     start      stop                     TSS chr_m   start_m  \\\n",
       "1222741  chr2  37550000  37555000  chr2_37550000_37555000  chr2  37551938   \n",
       "1223261  chr2  37895000  37900000  chr2_37895000_37900000  chr2  37898508   \n",
       "1223298  chr2  37895000  37900000  chr2_37895000_37900000  chr2  37898822   \n",
       "1223307  chr2  37895000  37900000  chr2_37895000_37900000  chr2  37898823   \n",
       "1223321  chr2  37895000  37900000  chr2_37895000_37900000  chr2  37898825   \n",
       "1223327  chr2  37895000  37900000  chr2_37895000_37900000  chr2  37898827   \n",
       "1223332  chr2  37895000  37900000  chr2_37895000_37900000  chr2  37898828   \n",
       "1223340  chr2  37895000  37900000  chr2_37895000_37900000  chr2  37898830   \n",
       "1223344  chr2  37895000  37900000  chr2_37895000_37900000  chr2  37898832   \n",
       "1223347  chr2  37895000  37900000  chr2_37895000_37900000  chr2  37898833   \n",
       "1223359  chr2  37895000  37900000  chr2_37895000_37900000  chr2  37898837   \n",
       "1223369  chr2  37895000  37900000  chr2_37895000_37900000  chr2  37898838   \n",
       "1223401  chr2  37895000  37900000  chr2_37895000_37900000  chr2  37898873   \n",
       "1223939  chr2  38300000  38305000  chr2_38300000_38305000  chr2  38304040   \n",
       "1223943  chr2  38300000  38305000  chr2_38300000_38305000  chr2  38304041   \n",
       "1224510  chr2  38600000  38605000  chr2_38600000_38605000  chr2  38604474   \n",
       "1224536  chr2  38600000  38605000  chr2_38600000_38605000  chr2  38604665   \n",
       "1224994  chr2  38830000  38835000  chr2_38830000_38835000  chr2  38830304   \n",
       "1225014  chr2  38830000  38835000  chr2_38830000_38835000  chr2  38830675   \n",
       "1225025  chr2  38830000  38835000  chr2_38830000_38835000  chr2  38830686   \n",
       "1225211  chr2  38975000  38980000  chr2_38975000_38980000  chr2  38978059   \n",
       "1225227  chr2  38975000  38980000  chr2_38975000_38980000  chr2  38978060   \n",
       "1225272  chr2  38975000  38980000  chr2_38975000_38980000  chr2  38978063   \n",
       "1225291  chr2  38975000  38980000  chr2_38975000_38980000  chr2  38978065   \n",
       "1225309  chr2  38975000  38980000  chr2_38975000_38980000  chr2  38978067   \n",
       "1225317  chr2  38975000  38980000  chr2_38975000_38980000  chr2  38978068   \n",
       "1225325  chr2  38975000  38980000  chr2_38975000_38980000  chr2  38978069   \n",
       "\n",
       "           stop_m                       id_trim      score strand   tf  \\\n",
       "1222741  37551959  SP1_HUMAN.H11MO.0.A.pwm.trim  12.651939      -  SP1   \n",
       "1223261  37898518  SP1_HUMAN.H11MO.1.A.pwm.trim  10.291843      +  SP1   \n",
       "1223298  37898843  SP1_HUMAN.H11MO.0.A.pwm.trim  22.014516      +  SP1   \n",
       "1223307  37898833  SP1_HUMAN.H11MO.1.A.pwm.trim  15.157784      +  SP1   \n",
       "1223321  37898846  SP1_HUMAN.H11MO.0.A.pwm.trim   7.886173      +  SP1   \n",
       "1223327  37898848  SP1_HUMAN.H11MO.0.A.pwm.trim  17.743111      +  SP1   \n",
       "1223332  37898838  SP1_HUMAN.H11MO.1.A.pwm.trim  12.310072      +  SP1   \n",
       "1223340  37898851  SP1_HUMAN.H11MO.0.A.pwm.trim   7.260751      +  SP1   \n",
       "1223344  37898853  SP1_HUMAN.H11MO.0.A.pwm.trim  16.885970      +  SP1   \n",
       "1223347  37898843  SP1_HUMAN.H11MO.1.A.pwm.trim  12.682930      +  SP1   \n",
       "1223359  37898858  SP1_HUMAN.H11MO.0.A.pwm.trim  24.419791      +  SP1   \n",
       "1223369  37898848  SP1_HUMAN.H11MO.1.A.pwm.trim  18.026685      +  SP1   \n",
       "1223401  37898883  SP1_HUMAN.H11MO.1.A.pwm.trim  11.706440      +  SP1   \n",
       "1223939  38304061  SP1_HUMAN.H11MO.0.A.pwm.trim  13.623659      +  SP1   \n",
       "1223943  38304051  SP1_HUMAN.H11MO.1.A.pwm.trim  10.828229      +  SP1   \n",
       "1224510  38604495  SP1_HUMAN.H11MO.0.A.pwm.trim   6.759938      +  SP1   \n",
       "1224536  38604675  SP1_HUMAN.H11MO.1.A.pwm.trim  15.225751      +  SP1   \n",
       "1224994  38830325  SP1_HUMAN.H11MO.0.A.pwm.trim   7.553580      -  SP1   \n",
       "1225014  38830696  SP1_HUMAN.H11MO.0.A.pwm.trim   8.006493      +  SP1   \n",
       "1225025  38830696  SP1_HUMAN.H11MO.1.A.pwm.trim  11.896678      -  SP1   \n",
       "1225211  38978080  SP1_HUMAN.H11MO.0.A.pwm.trim  18.964402      +  SP1   \n",
       "1225227  38978070  SP1_HUMAN.H11MO.1.A.pwm.trim  12.356850      +  SP1   \n",
       "1225272  38978084  SP1_HUMAN.H11MO.0.A.pwm.trim  10.673023      +  SP1   \n",
       "1225291  38978086  SP1_HUMAN.H11MO.0.A.pwm.trim   9.500081      +  SP1   \n",
       "1225309  38978088  SP1_HUMAN.H11MO.0.A.pwm.trim   6.819492      +  SP1   \n",
       "1225317  38978089  SP1_HUMAN.H11MO.0.A.pwm.trim  14.015691      +  SP1   \n",
       "1225325  38978079  SP1_HUMAN.H11MO.1.A.pwm.trim  10.581895      +  SP1   \n",
       "\n",
       "             vocab       tissue    config    gene  \n",
       "1222741  ARNT::SP1  SCC13-CTRLi  pro_loop  CYP1B1  \n",
       "1223261  ARNT::SP1  SCC13-CTRLi  pro_loop  CYP1B1  \n",
       "1223298  ARNT::SP1  SCC13-CTRLi  pro_loop  CYP1B1  \n",
       "1223307  ARNT::SP1  SCC13-CTRLi  pro_loop  CYP1B1  \n",
       "1223321  ARNT::SP1  SCC13-CTRLi  pro_loop  CYP1B1  \n",
       "1223327  ARNT::SP1  SCC13-CTRLi  pro_loop  CYP1B1  \n",
       "1223332  ARNT::SP1  SCC13-CTRLi  pro_loop  CYP1B1  \n",
       "1223340  ARNT::SP1  SCC13-CTRLi  pro_loop  CYP1B1  \n",
       "1223344  ARNT::SP1  SCC13-CTRLi  pro_loop  CYP1B1  \n",
       "1223347  ARNT::SP1  SCC13-CTRLi  pro_loop  CYP1B1  \n",
       "1223359  ARNT::SP1  SCC13-CTRLi  pro_loop  CYP1B1  \n",
       "1223369  ARNT::SP1  SCC13-CTRLi  pro_loop  CYP1B1  \n",
       "1223401  ARNT::SP1  SCC13-CTRLi  pro_loop  CYP1B1  \n",
       "1223939  ARNT::SP1  SCC13-CTRLi  pro_loop  CYP1B1  \n",
       "1223943  ARNT::SP1  SCC13-CTRLi  pro_loop  CYP1B1  \n",
       "1224510  ARNT::SP1  SCC13-CTRLi  pro_loop  CYP1B1  \n",
       "1224536  ARNT::SP1  SCC13-CTRLi  pro_loop  CYP1B1  \n",
       "1224994  ARNT::SP1  SCC13-CTRLi  pro_loop  CYP1B1  \n",
       "1225014  ARNT::SP1  SCC13-CTRLi  pro_loop  CYP1B1  \n",
       "1225025  ARNT::SP1  SCC13-CTRLi  pro_loop  CYP1B1  \n",
       "1225211  ARNT::SP1  SCC13-CTRLi  pro_loop  CYP1B1  \n",
       "1225227  ARNT::SP1  SCC13-CTRLi  pro_loop  CYP1B1  \n",
       "1225272  ARNT::SP1  SCC13-CTRLi  pro_loop  CYP1B1  \n",
       "1225291  ARNT::SP1  SCC13-CTRLi  pro_loop  CYP1B1  \n",
       "1225309  ARNT::SP1  SCC13-CTRLi  pro_loop  CYP1B1  \n",
       "1225317  ARNT::SP1  SCC13-CTRLi  pro_loop  CYP1B1  \n",
       "1225325  ARNT::SP1  SCC13-CTRLi  pro_loop  CYP1B1  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab = 'ARNT::SP1'\n",
    "df =get_genomic_instances_pro_loop(vocab, 'SCC13-CTRLi', ['CYP1B1'])\n",
    "df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>TSS</th>\n",
       "      <th>chr_m</th>\n",
       "      <th>start_m</th>\n",
       "      <th>stop_m</th>\n",
       "      <th>id_trim</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>tf</th>\n",
       "      <th>vocab</th>\n",
       "      <th>tissue</th>\n",
       "      <th>config</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [chr, start, stop, TSS, chr_m, start_m, stop_m, id_trim, score, strand, tf, vocab, tissue, config, gene]\n",
       "Index: []"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prom_anc_arr = ['chr2_38320000_38325000', 'chr2_38325000_38330000',\n",
    "       'chr2_38330000_38335000']\n",
    "\n",
    "df[df.TSS.isin(prom_anc_arr)]\n",
    "# df[df.TSS.isin(['chr2_38320000_38325000'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>546518</td>\n",
       "      <td>chr2_38320000_38325000</td>\n",
       "      <td>chr2_38465000_38470000</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        source                  target  count\n",
       "546518  chr2_38320000_38325000  chr2_38465000_38470000     22"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loop_df[loop_df.source.isin(['chr2_38465000_38470000'])|loop_df.target.isin(['chr2_38465000_38470000'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:23: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    }
   ],
   "source": [
    "tissue='CAL27-CTRLi'\n",
    "list_genes = ['CUL9']\n",
    "anc_prom = pybedtools.BedTool('../data/interim/annon/promoter_anchors/promoter_'+tissue+'_annon.bed').to_dataframe()\n",
    "anc_prom = anc_prom[anc_prom.name.isin(list_genes)]\n",
    "anc_prom_arr = anc_prom.thickEnd.unique()\n",
    "    \n",
    "loop_df = pd.read_csv('../data/interim/merged/loops/'+tissue+'.loops.csv',index_col=0)\n",
    "loop_df = loop_df[loop_df.source.isin(anc_prom_arr)|loop_df.target.isin(anc_prom_arr)]\n",
    "all_anc = sorted(set(anc_prom.thickEnd.to_list() +loop_df.source.to_list() + loop_df.target.to_list() ))\n",
    "\n",
    "anc_foot = pybedtools.BedTool('../data/interim/annon/anchor_footprinting/'+tissue+'_annon.bed').to_dataframe()\n",
    "anc_foot['tf'] = anc_foot.thickEnd.map(tf_id_to_name_dict)\n",
    "anc_foot = anc_foot[anc_foot.tf.isin(vocab.split('::'))]\n",
    "anc_foot = anc_foot[anc_foot.name.isin(all_anc)]\n",
    "\n",
    "\n",
    "pro_anc_foot_df = pybedtools.BedTool('../data/interim/annon/promoter_footprinting/promoter_'+tissue+'_annon.bed').to_dataframe()\n",
    "pro_anc_foot_df['tf'] = pro_anc_foot_df.thickEnd.map(tf_id_to_name_dict)\n",
    "pro_anc_foot = pd.DataFrame()\n",
    "\n",
    "for gene in list_genes:\n",
    "    pro_anc_foot1 = pro_anc_foot_df[(pro_anc_foot_df.name==gene)|(pro_anc_foot_df.tf.isin(vocab.split('::')))]# not perfect will catch later???\n",
    "    pro_anc_foot1['anchor'] = pro_anc_foot_df.score + \"_\" + pro_anc_foot_df.strand.map(str)+\"_\"+ pro_anc_foot_df.thickStart.map(str)\n",
    "    pro_anc_foot1 = pro_anc_foot1[pro_anc_foot1.anchor.isin(all_anc)]\n",
    "    pro_anc_foot1.drop(columns=['anchor'],inplace=True)\n",
    "    pro_anc_foot = pd.concat([pro_anc_foot, pro_anc_foot1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>thickStart</th>\n",
       "      <th>thickEnd</th>\n",
       "      <th>itemRgb</th>\n",
       "      <th>blockCount</th>\n",
       "      <th>blockSizes</th>\n",
       "      <th>tf</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>16</td>\n",
       "      <td>chr1</td>\n",
       "      <td>740000</td>\n",
       "      <td>745000</td>\n",
       "      <td>chr1_740000_745000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>740299</td>\n",
       "      <td>740305</td>\n",
       "      <td>ARNT_HUMAN.H11MO.0.B.pwm.trim</td>\n",
       "      <td>10.397177</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARNT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>56</td>\n",
       "      <td>chr1</td>\n",
       "      <td>760000</td>\n",
       "      <td>765000</td>\n",
       "      <td>chr1_760000_765000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>762629</td>\n",
       "      <td>762650</td>\n",
       "      <td>SP1_HUMAN.H11MO.0.A.pwm.trim</td>\n",
       "      <td>6.556889</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>110</td>\n",
       "      <td>chr1</td>\n",
       "      <td>760000</td>\n",
       "      <td>765000</td>\n",
       "      <td>chr1_760000_765000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>762921</td>\n",
       "      <td>762931</td>\n",
       "      <td>SP1_HUMAN.H11MO.1.A.pwm.trim</td>\n",
       "      <td>18.026685</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>123</td>\n",
       "      <td>chr1</td>\n",
       "      <td>760000</td>\n",
       "      <td>765000</td>\n",
       "      <td>chr1_760000_765000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>762960</td>\n",
       "      <td>762981</td>\n",
       "      <td>SP1_HUMAN.H11MO.0.A.pwm.trim</td>\n",
       "      <td>18.764170</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>127</td>\n",
       "      <td>chr1</td>\n",
       "      <td>760000</td>\n",
       "      <td>765000</td>\n",
       "      <td>chr1_760000_765000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>762961</td>\n",
       "      <td>762971</td>\n",
       "      <td>SP1_HUMAN.H11MO.1.A.pwm.trim</td>\n",
       "      <td>11.564458</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2553907</td>\n",
       "      <td>chrX</td>\n",
       "      <td>154255000</td>\n",
       "      <td>154260000</td>\n",
       "      <td>chrX_154255000_154260000</td>\n",
       "      <td>chrX</td>\n",
       "      <td>154255348</td>\n",
       "      <td>154255369</td>\n",
       "      <td>SP1_HUMAN.H11MO.0.A.pwm.trim</td>\n",
       "      <td>10.018741</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2554020</td>\n",
       "      <td>chrX</td>\n",
       "      <td>154295000</td>\n",
       "      <td>154300000</td>\n",
       "      <td>chrX_154295000_154300000</td>\n",
       "      <td>chrX</td>\n",
       "      <td>154299437</td>\n",
       "      <td>154299458</td>\n",
       "      <td>SP1_HUMAN.H11MO.0.A.pwm.trim</td>\n",
       "      <td>14.421472</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2554061</td>\n",
       "      <td>chrX</td>\n",
       "      <td>154295000</td>\n",
       "      <td>154300000</td>\n",
       "      <td>chrX_154295000_154300000</td>\n",
       "      <td>chrX</td>\n",
       "      <td>154299481</td>\n",
       "      <td>154299502</td>\n",
       "      <td>SP1_HUMAN.H11MO.0.A.pwm.trim</td>\n",
       "      <td>9.501379</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2554076</td>\n",
       "      <td>chrX</td>\n",
       "      <td>154295000</td>\n",
       "      <td>154300000</td>\n",
       "      <td>chrX_154295000_154300000</td>\n",
       "      <td>chrX</td>\n",
       "      <td>154299491</td>\n",
       "      <td>154299501</td>\n",
       "      <td>SP1_HUMAN.H11MO.1.A.pwm.trim</td>\n",
       "      <td>11.706440</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2554081</td>\n",
       "      <td>chrX</td>\n",
       "      <td>154295000</td>\n",
       "      <td>154300000</td>\n",
       "      <td>chrX_154295000_154300000</td>\n",
       "      <td>chrX</td>\n",
       "      <td>154299495</td>\n",
       "      <td>154299516</td>\n",
       "      <td>SP1_HUMAN.H11MO.0.A.pwm.trim</td>\n",
       "      <td>10.224412</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SP1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>24553 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom      start        end                      name score  \\\n",
       "16       chr1     740000     745000        chr1_740000_745000  chr1   \n",
       "56       chr1     760000     765000        chr1_760000_765000  chr1   \n",
       "110      chr1     760000     765000        chr1_760000_765000  chr1   \n",
       "123      chr1     760000     765000        chr1_760000_765000  chr1   \n",
       "127      chr1     760000     765000        chr1_760000_765000  chr1   \n",
       "...       ...        ...        ...                       ...   ...   \n",
       "2553907  chrX  154255000  154260000  chrX_154255000_154260000  chrX   \n",
       "2554020  chrX  154295000  154300000  chrX_154295000_154300000  chrX   \n",
       "2554061  chrX  154295000  154300000  chrX_154295000_154300000  chrX   \n",
       "2554076  chrX  154295000  154300000  chrX_154295000_154300000  chrX   \n",
       "2554081  chrX  154295000  154300000  chrX_154295000_154300000  chrX   \n",
       "\n",
       "            strand  thickStart                       thickEnd    itemRgb  \\\n",
       "16          740299      740305  ARNT_HUMAN.H11MO.0.B.pwm.trim  10.397177   \n",
       "56          762629      762650   SP1_HUMAN.H11MO.0.A.pwm.trim   6.556889   \n",
       "110         762921      762931   SP1_HUMAN.H11MO.1.A.pwm.trim  18.026685   \n",
       "123         762960      762981   SP1_HUMAN.H11MO.0.A.pwm.trim  18.764170   \n",
       "127         762961      762971   SP1_HUMAN.H11MO.1.A.pwm.trim  11.564458   \n",
       "...            ...         ...                            ...        ...   \n",
       "2553907  154255348   154255369   SP1_HUMAN.H11MO.0.A.pwm.trim  10.018741   \n",
       "2554020  154299437   154299458   SP1_HUMAN.H11MO.0.A.pwm.trim  14.421472   \n",
       "2554061  154299481   154299502   SP1_HUMAN.H11MO.0.A.pwm.trim   9.501379   \n",
       "2554076  154299491   154299501   SP1_HUMAN.H11MO.1.A.pwm.trim  11.706440   \n",
       "2554081  154299495   154299516   SP1_HUMAN.H11MO.0.A.pwm.trim  10.224412   \n",
       "\n",
       "        blockCount blockSizes    tf  \n",
       "16               -          .  ARNT  \n",
       "56               +          .   SP1  \n",
       "110              +          .   SP1  \n",
       "123              +          .   SP1  \n",
       "127              +          .   SP1  \n",
       "...            ...        ...   ...  \n",
       "2553907          +          .   SP1  \n",
       "2554020          -          .   SP1  \n",
       "2554061          -          .   SP1  \n",
       "2554076          -          .   SP1  \n",
       "2554081          -          .   SP1  \n",
       "\n",
       "[24553 rows x 12 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "anc_foot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
