{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "08/25/2020\n",
    "\n",
    "check vocab words\n",
    "1. make sure they are in the rna expression\n",
    "\n",
    "\n",
    "08/30 finish gathering genomic instances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [],
   "source": [
    "# basic packages\n",
    "import re\n",
    "import os, glob\n",
    "import pandas as pd\n",
    "import numpy as np; np.random.seed(0)\n",
    "import itertools\n",
    "from collections import Counter, defaultdict\n",
    "import time\n",
    "# Import tools needed for visualization\n",
    "import seaborn as sns; sns.set()\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import pybedtools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_dir = '../data/processed/fig4_modelling/vocab_sum_final/'\n",
    "if not os.path.exists(save_dir):\n",
    "    os.makedirs(save_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1 Get the vocab words"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 904,
   "metadata": {},
   "outputs": [],
   "source": [
    "vocab_dict = {'Colon':pd.read_excel('../data/processed/fig4_modelling/vocab_sum/Colon_filtered_vocab_list.xlsx',header=None).loc[:,0].values,\n",
    "'Melanocytes':pd.read_excel('../data/processed/fig4_modelling/vocab_sum/Melanocytes_filtered_vocab_list.xlsx',header=None).loc[:,0].values,\n",
    "'GDSD6':pd.read_excel('../data/processed/fig4_modelling/vocab_sum/GDSD6_filtered_vocab_list.xlsx',header=None).loc[:,0].values,\n",
    " 'GM12878':pd.read_excel('../data/processed/fig4_modelling/vocab_sum/GM12878_filtered_vocab_list.xlsx',header=None).loc[:,0].values,\n",
    "'SCC':pd.read_excel('../data/processed/fig4_modelling/vocab_sum/SCC_filtered_vocab.xlsx',header=None).loc[:,0].values,\n",
    " 'MM':pd.read_excel('../data/processed/fig4_modelling/vocab_sum/MM_filtered_vocab_list.xlsx',header=None).loc[:,0].values\n",
    "\n",
    "}\n",
    "\n",
    "vocab_df = pd.DataFrame(columns=['tissue','vocab'])\n",
    "for tissue, vocab_arr in vocab_dict.items():\n",
    "    for vocab in vocab_arr:\n",
    "        vocab_df = vocab_df.append({'tissue':tissue, 'vocab':vocab},ignore_index=True)\n",
    "vocab_df['source'] = 'enrich'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 905,
   "metadata": {},
   "outputs": [],
   "source": [
    "irf_df = pd.read_excel('../data/processed/fig4_modelling/irf_manual_lookup.xlsx')\n",
    "irf_df = irf_df[irf_df['Y/n?']=='Y']\n",
    "irf_df = irf_df[['Vocab','Tissue']]\n",
    "irf_df.columns = ['vocab', 'tissue']\n",
    "irf_df['vocab'] = irf_df.vocab.str.split('[:_]',expand=True).iloc[:,[0,3]].agg('::'.join, axis=1)\n",
    "irf_df['source'] = 'irf'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 906,
   "metadata": {},
   "outputs": [],
   "source": [
    "overall_df = pd.concat([vocab_df, irf_df],axis=0,sort=False)\n",
    "overall_df[['vocab1','vocab2']] = overall_df.vocab.str.split('::',expand=True)\n",
    "overall_df.drop_duplicates(inplace=True)\n",
    "overall_df.reset_index(inplace=True, drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 907,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tissue</th>\n",
       "      <th>vocab</th>\n",
       "      <th>source</th>\n",
       "      <th>vocab1</th>\n",
       "      <th>vocab2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>Colon</td>\n",
       "      <td>BARX2::DLX4</td>\n",
       "      <td>enrich</td>\n",
       "      <td>BARX2</td>\n",
       "      <td>DLX4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>Colon</td>\n",
       "      <td>BARX2::FOXF2</td>\n",
       "      <td>enrich</td>\n",
       "      <td>BARX2</td>\n",
       "      <td>FOXF2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>Colon</td>\n",
       "      <td>BARX2::FOXM1</td>\n",
       "      <td>enrich</td>\n",
       "      <td>BARX2</td>\n",
       "      <td>FOXM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>Colon</td>\n",
       "      <td>BARX2::THAP1</td>\n",
       "      <td>enrich</td>\n",
       "      <td>BARX2</td>\n",
       "      <td>THAP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>Colon</td>\n",
       "      <td>BATF::KLF15</td>\n",
       "      <td>enrich</td>\n",
       "      <td>BATF</td>\n",
       "      <td>KLF15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  tissue         vocab  source vocab1 vocab2\n",
       "0  Colon   BARX2::DLX4  enrich  BARX2   DLX4\n",
       "1  Colon  BARX2::FOXF2  enrich  BARX2  FOXF2\n",
       "2  Colon  BARX2::FOXM1  enrich  BARX2  FOXM1\n",
       "3  Colon  BARX2::THAP1  enrich  BARX2  THAP1\n",
       "4  Colon   BATF::KLF15  enrich   BATF  KLF15"
      ]
     },
     "execution_count": 907,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overall_df[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 908,
   "metadata": {},
   "outputs": [],
   "source": [
    "overall_df.to_csv(os.path.join(save_dir, 'overall_df.csv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##  1B. some basic stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 909,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Melanocytes    132\n",
       "GDSD6           94\n",
       "Colon           85\n",
       "GM12878         64\n",
       "MM              53\n",
       "SCC             45\n",
       "Name: tissue, dtype: int64"
      ]
     },
     "execution_count": 909,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overall_df.tissue.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 910,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(473, 5)"
      ]
     },
     "execution_count": 910,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overall_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 911,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "472"
      ]
     },
     "execution_count": 911,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(overall_df.vocab.unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "currently 472 unique vocab pairs for each tissue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 913,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "248"
      ]
     },
     "execution_count": 913,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "motifs =set( list(overall_df.vocab1)+ list(overall_df.vocab2))\n",
    "len(motifs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "using 248 vocabulary words"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 914,
   "metadata": {},
   "outputs": [],
   "source": [
    "motif_to_tissue = {}\n",
    "for motif in motifs:\n",
    "    motif_to_tissue[motif] = {'Colon':0,'GDSD6':0,'GM12878':0, 'Melanocytes':0,'SCC':0,'MM':0}\n",
    "for idx, row in overall_df.iterrows():\n",
    "    motif_to_tissue[row['vocab1']][row['tissue']]+=1\n",
    "    motif_to_tissue[row['vocab2']][row['tissue']]+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 915,
   "metadata": {},
   "outputs": [],
   "source": [
    "motif_to_tissue_df = pd.DataFrame.from_dict(motif_to_tissue,orient='index')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 916,
   "metadata": {},
   "outputs": [],
   "source": [
    "motif_to_tissue_df.to_csv(os.path.join(save_dir, 'motif_to_tissue.csv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1C. check to see if the vocab words are in the irf model somewhere"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 917,
   "metadata": {},
   "outputs": [],
   "source": [
    "tissue_to_irf_file  = {\n",
    "    'Colon': '../data/processed/fig4_modelling/tf_tf_pairs_stability_sherlock/expr_Colon_Colon_boosted_stability_score.csv',\n",
    "    'GDSD6': '../data/processed/fig4_modelling/tf_tf_pairs_stability_sherlock/expr_GDSD6_GDSD6_boosted_stability_score.csv',\n",
    "    'GM12878': '../data/processed/fig4_modelling/tf_tf_pairs_stability_sherlock/expr_GM12878_GM12878_boosted_stability_score.csv',\n",
    "    'Melanocytes': '../data/processed/fig4_modelling/tf_tf_pairs_stability_sherlock/expr_Melanocytes_Melanocytes_boosted_stability_score.csv',\n",
    "    'MM': '../data/processed/fig4_modelling/irf_manual_mm/expr_cancer_boosted_stability_score.csv',#'../data/processed/fig4_modelling/irf_manual_mm/diff_cancer_boosted_stability_score.csv'\n",
    "    'SCC': '../data/processed/fig4_modelling/irf_manual_scc/expr_cancer_boosted_stability_score.csv',#'../data/processed/fig4_modelling/irf_manual_scc/diff_cancer_boosted_stability_score.csv'\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 918,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "1\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "10\n",
      "11\n",
      "12\n",
      "13\n",
      "14\n",
      "15\n",
      "16\n",
      "17\n",
      "18\n",
      "19\n",
      "20\n",
      "21\n",
      "22\n",
      "23\n",
      "24\n",
      "25\n",
      "26\n",
      "27\n",
      "28\n",
      "29\n",
      "30\n",
      "31\n",
      "32\n",
      "33\n",
      "34\n",
      "35\n",
      "36\n",
      "37\n",
      "38\n",
      "39\n",
      "40\n",
      "41\n",
      "42\n",
      "43\n",
      "44\n",
      "45\n",
      "46\n",
      "47\n",
      "48\n",
      "49\n",
      "50\n",
      "51\n",
      "52\n",
      "53\n",
      "54\n",
      "55\n",
      "56\n",
      "57\n",
      "58\n",
      "59\n",
      "60\n",
      "61\n",
      "62\n",
      "63\n",
      "64\n",
      "65\n",
      "66\n",
      "67\n",
      "68\n",
      "69\n",
      "70\n",
      "71\n",
      "72\n",
      "73\n",
      "74\n",
      "75\n",
      "76\n",
      "77\n",
      "78\n",
      "79\n",
      "80\n",
      "81\n",
      "82\n",
      "83\n",
      "84\n",
      "85\n",
      "86\n",
      "87\n",
      "88\n",
      "89\n",
      "90\n",
      "91\n",
      "92\n",
      "93\n",
      "94\n",
      "95\n",
      "96\n",
      "97\n",
      "98\n",
      "99\n",
      "100\n",
      "101\n",
      "102\n",
      "103\n",
      "104\n",
      "105\n",
      "106\n",
      "107\n",
      "108\n",
      "109\n",
      "110\n",
      "111\n",
      "112\n",
      "113\n",
      "114\n",
      "115\n",
      "116\n",
      "117\n",
      "118\n",
      "119\n",
      "120\n",
      "121\n",
      "122\n",
      "123\n",
      "124\n",
      "125\n",
      "126\n",
      "127\n",
      "128\n",
      "129\n",
      "130\n",
      "131\n",
      "132\n",
      "133\n",
      "134\n",
      "135\n",
      "136\n",
      "137\n",
      "138\n",
      "139\n",
      "140\n",
      "141\n",
      "142\n",
      "143\n",
      "144\n",
      "145\n",
      "146\n",
      "147\n",
      "148\n",
      "149\n",
      "150\n",
      "151\n",
      "152\n",
      "153\n",
      "154\n",
      "155\n",
      "156\n",
      "157\n",
      "158\n",
      "159\n",
      "160\n",
      "161\n",
      "162\n",
      "163\n",
      "164\n",
      "165\n",
      "166\n",
      "167\n",
      "168\n",
      "169\n",
      "170\n",
      "171\n",
      "172\n",
      "173\n",
      "174\n",
      "175\n",
      "176\n",
      "177\n",
      "178\n",
      "179\n",
      "180\n",
      "181\n",
      "182\n",
      "183\n",
      "184\n",
      "185\n",
      "186\n",
      "187\n",
      "188\n",
      "189\n",
      "190\n",
      "191\n",
      "192\n",
      "193\n",
      "194\n",
      "195\n",
      "196\n",
      "197\n",
      "198\n",
      "199\n",
      "200\n",
      "201\n",
      "202\n",
      "203\n",
      "204\n",
      "205\n",
      "206\n",
      "207\n",
      "208\n",
      "209\n",
      "210\n",
      "211\n",
      "212\n",
      "213\n",
      "214\n",
      "215\n",
      "216\n",
      "217\n",
      "218\n",
      "219\n",
      "220\n",
      "221\n",
      "222\n",
      "223\n",
      "224\n",
      "225\n",
      "226\n",
      "227\n",
      "228\n",
      "229\n",
      "230\n",
      "231\n",
      "232\n",
      "233\n",
      "234\n",
      "235\n",
      "236\n",
      "237\n",
      "238\n",
      "239\n",
      "240\n",
      "241\n",
      "242\n",
      "243\n",
      "244\n",
      "245\n",
      "246\n",
      "247\n",
      "248\n",
      "249\n",
      "250\n",
      "251\n",
      "252\n",
      "253\n",
      "254\n",
      "255\n",
      "256\n",
      "257\n",
      "258\n",
      "259\n",
      "260\n",
      "261\n",
      "262\n",
      "263\n",
      "264\n",
      "265\n",
      "266\n",
      "267\n",
      "268\n",
      "269\n",
      "270\n",
      "271\n",
      "272\n",
      "273\n",
      "274\n",
      "275\n",
      "276\n",
      "277\n",
      "278\n",
      "279\n",
      "280\n",
      "281\n",
      "282\n",
      "283\n",
      "284\n",
      "285\n",
      "286\n",
      "287\n",
      "288\n",
      "289\n",
      "290\n",
      "291\n",
      "292\n",
      "293\n",
      "294\n",
      "295\n",
      "296\n",
      "297\n",
      "298\n",
      "299\n",
      "300\n",
      "301\n",
      "302\n",
      "303\n",
      "304\n",
      "305\n",
      "306\n",
      "307\n",
      "308\n",
      "309\n",
      "310\n",
      "311\n",
      "312\n",
      "313\n",
      "314\n",
      "315\n",
      "316\n",
      "317\n",
      "318\n",
      "319\n",
      "320\n",
      "321\n",
      "322\n",
      "323\n",
      "324\n",
      "325\n",
      "326\n",
      "327\n",
      "328\n",
      "329\n",
      "330\n",
      "331\n",
      "332\n",
      "333\n",
      "334\n",
      "335\n",
      "336\n",
      "337\n",
      "338\n",
      "339\n",
      "340\n",
      "341\n",
      "342\n",
      "343\n",
      "344\n",
      "345\n",
      "346\n",
      "347\n",
      "348\n",
      "349\n",
      "350\n",
      "351\n",
      "352\n",
      "353\n",
      "354\n",
      "355\n",
      "356\n",
      "357\n",
      "358\n",
      "359\n",
      "360\n",
      "361\n",
      "362\n",
      "363\n",
      "364\n",
      "365\n",
      "366\n",
      "367\n",
      "368\n",
      "369\n",
      "370\n",
      "371\n",
      "372\n",
      "373\n",
      "374\n",
      "375\n",
      "376\n",
      "377\n",
      "378\n",
      "379\n",
      "380\n",
      "381\n",
      "382\n",
      "383\n",
      "384\n",
      "385\n",
      "386\n",
      "387\n",
      "388\n",
      "389\n",
      "390\n",
      "391\n",
      "392\n",
      "393\n",
      "394\n",
      "395\n",
      "396\n",
      "397\n",
      "398\n",
      "399\n",
      "400\n",
      "401\n",
      "402\n",
      "403\n",
      "404\n",
      "405\n",
      "406\n",
      "407\n",
      "408\n",
      "409\n",
      "410\n",
      "411\n",
      "412\n",
      "413\n",
      "414\n",
      "415\n",
      "416\n",
      "417\n",
      "418\n",
      "419\n",
      "420\n",
      "421\n",
      "422\n",
      "423\n",
      "424\n",
      "425\n",
      "426\n",
      "427\n",
      "428\n",
      "429\n",
      "430\n",
      "431\n",
      "432\n",
      "433\n",
      "434\n",
      "435\n",
      "436\n",
      "437\n",
      "438\n",
      "439\n",
      "440\n",
      "441\n",
      "442\n",
      "443\n",
      "444\n",
      "445\n",
      "446\n",
      "447\n",
      "448\n",
      "449\n",
      "450\n",
      "451\n",
      "452\n",
      "453\n",
      "454\n",
      "455\n",
      "456\n",
      "457\n",
      "458\n",
      "459\n",
      "460\n",
      "461\n",
      "462\n",
      "463\n",
      "464\n",
      "465\n",
      "466\n",
      "467\n",
      "468\n",
      "469\n",
      "470\n",
      "471\n",
      "472\n",
      "CPU times: user 2min 28s, sys: 1.07 s, total: 2min 29s\n",
      "Wall time: 2min 29s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# takes about 5 min, slow \n",
    "overall_df['irf_found'] = ''\n",
    "overall_df['irf_score_found'] = 0.0\n",
    "for idx, row in overall_df.iterrows():\n",
    "    print(idx)\n",
    "    vocab1, vocab2 = row['vocab1'], row['vocab2']\n",
    "    tissue = row['tissue']\n",
    "    irf_file = tissue_to_irf_file[tissue]\n",
    "    \n",
    "    irf_df = pd.read_csv(irf_file).sort_values('score',ascending=False)\n",
    "    found_vocab_pair = False\n",
    "    jdx = 0\n",
    "    while (not found_vocab_pair )and (jdx < irf_df.shape[0]) :\n",
    "        row_irf = irf_df.loc[irf_df.index[jdx],:]\n",
    "        irf_vocab = row_irf['index']\n",
    "        score = row_irf['score']\n",
    "        if (vocab1 in irf_vocab) & (vocab2 in irf_vocab):\n",
    "            found_vocab_pair = True\n",
    "            overall_df.at[idx, 'irf_found'] = irf_vocab\n",
    "            overall_df.at[idx,'irf_score_found'] = score\n",
    "        jdx+=1\n",
    "    \n",
    "\n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 919,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(63, 7)"
      ]
     },
     "execution_count": 919,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overall_df[overall_df.irf_score_found>0].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 920,
   "metadata": {},
   "outputs": [],
   "source": [
    "irf_vocab = overall_df[overall_df.irf_score_found>0].vocab.values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "63 of the 455 are from irf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 929,
   "metadata": {},
   "outputs": [],
   "source": [
    "overall_df.to_csv(os.path.join(save_dir, 'overall_df.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. check RNA expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 922,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['A431-CTRLi', 'A431-p63i', 'Airway', 'Astrocytes', 'Bladder',\n",
      "       'CAL27-CTRLi', 'CAL27-p63i', 'COLO_SCR_DMSO', 'COLO_SCR_PLX',\n",
      "       'COLO_shMITF_DMSO', 'COLO_shMITF_PLX', 'Colon', 'D0-CTRLi', 'D0-p63i',\n",
      "       'D3-CTRLi', 'D3-p63i', 'Esophageal', 'GDSD0', 'GDSD3', 'GDSD6',\n",
      "       'GM12878', 'HMEC', 'Melanocytes', 'Ovarian', 'Pancreas', 'Prostate',\n",
      "       'Renal', 'SCC13-CTRLi', 'SCC13-p63i', 'SKMEL5_SCR_DMSO',\n",
      "       'SKMEL5_SCR_PLX', 'SKMEL5_shMITF-DMSO', 'SKMEL5_shMITF_PLX', 'Thyroid',\n",
      "       'Uterine', 'WM_SCR_DMSO', 'WM_SCR_PLX', 'WM_shMITF_DMSO',\n",
      "       'WM_shMITF_PLX'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "rna_df = pd.read_csv('../data/interim/rna/tissue_tpm_sym.csv',index_col=0)\n",
    "print(rna_df.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 923,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "248"
      ]
     },
     "execution_count": 923,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "motifs_with_rna_sym = motif_to_tissue_df.index[[True if x in rna_df.index.values else False for x in motif_to_tissue_df.index]]\n",
    "len(motifs_with_rna_sym)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 924,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index([], dtype='object')"
      ]
     },
     "execution_count": 924,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "motif_to_tissue_df.index[~np.array([True if x in rna_df.index.values else False for x in motif_to_tissue_df.index])]\n",
    "# all motifs are rna symbols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 925,
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_df_tf = rna_df.loc[sorted(motif_to_tissue_df.index.values),['Colon','GDSD6','GM12878','Melanocytes','A431-CTRLi','CAL27-CTRLi','SCC13-CTRLi','COLO_SCR_DMSO','WM_SCR_DMSO']]\n",
    "rna_df_tf.to_csv(os.path.join(save_dir, 'rna_df_tf.csv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2B. manual check for RNA TF to have TPM >1\n",
    "\n",
    "files to compare: `'../data/processed/fig4_modelling/vocab_sum_final/motif_to_tissue_log.pdf'`\n",
    "'../data/processed/fig4_modelling/vocab_sum_final/rna_df_tf_bin.pdf'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 926,
   "metadata": {},
   "outputs": [],
   "source": [
    "THRES=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 927,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tissue</th>\n",
       "      <th>motif</th>\n",
       "      <th>pass_exp_filt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>133</td>\n",
       "      <td>Colon</td>\n",
       "      <td>NR1D1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>191</td>\n",
       "      <td>Colon</td>\n",
       "      <td>STAT5A</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>439</td>\n",
       "      <td>GDSD6</td>\n",
       "      <td>STAT5A</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>554</td>\n",
       "      <td>GM12878</td>\n",
       "      <td>GATA2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>613</td>\n",
       "      <td>GM12878</td>\n",
       "      <td>MITF</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>834</td>\n",
       "      <td>Melanocytes</td>\n",
       "      <td>JDP2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          tissue   motif  pass_exp_filt\n",
       "133        Colon   NR1D1          False\n",
       "191        Colon  STAT5A          False\n",
       "439        GDSD6  STAT5A          False\n",
       "554      GM12878   GATA2          False\n",
       "613      GM12878    MITF          False\n",
       "834  Melanocytes    JDP2          False"
      ]
     },
     "execution_count": 927,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "check_rna = pd.DataFrame(rna_df_tf[motif_to_tissue_df==1].fillna(2)>THRES).unstack().reset_index()\n",
    "check_rna.columns = ['tissue','motif', 'pass_exp_filt']\n",
    "check_rna[check_rna.pass_exp_filt==False]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "manually pick motifs to remove"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 930,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tissue</th>\n",
       "      <th>vocab</th>\n",
       "      <th>source</th>\n",
       "      <th>vocab1</th>\n",
       "      <th>vocab2</th>\n",
       "      <th>irf_found</th>\n",
       "      <th>irf_score_found</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>423</td>\n",
       "      <td>Colon</td>\n",
       "      <td>NFYC::NR1D1</td>\n",
       "      <td>irf</td>\n",
       "      <td>NFYC</td>\n",
       "      <td>NR1D1</td>\n",
       "      <td>NFYC_loop::NR1D1_loop</td>\n",
       "      <td>0.892857</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    tissue        vocab source vocab1 vocab2              irf_found  \\\n",
       "423  Colon  NFYC::NR1D1    irf   NFYC  NR1D1  NFYC_loop::NR1D1_loop   \n",
       "\n",
       "     irf_score_found  \n",
       "423         0.892857  "
      ]
     },
     "execution_count": 930,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overall_df[(overall_df.vocab1=='NR1D1') |(overall_df.vocab2=='NR1D1') ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 931,
   "metadata": {},
   "outputs": [],
   "source": [
    "bad_tfs = ['NR1D1','STAT5A','GATA2', 'JDP2'] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 932,
   "metadata": {},
   "outputs": [],
   "source": [
    "# testing\n",
    "# rna_df.loc['STAT5A',:] # not expressed in KRT (0.045)\n",
    "# rna_df.loc['GATA2',:] # not expressed in GM12878 (0.35)\n",
    "# rna_df.loc['FOSL1',:] # not expressed in GM12878 (0.635)\n",
    "# rna_df.loc['JDP2',:] # not expressed in Melanocytes (0.434)\n",
    "# rna_df.loc['NR1D1',:] # not expressed in Colon (0.160)\n",
    "# rna_df.loc['MITF',:] # not expressed in GM but going to keep b/c needed in KRT\n",
    "# rna_df.loc['SMAD4',:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 933,
   "metadata": {},
   "outputs": [],
   "source": [
    "overall_df_filt = overall_df[(~overall_df.vocab1.isin(bad_tfs)) & (~overall_df.vocab2.isin(bad_tfs))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 934,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(473, 7)\n",
      "(467, 7)\n"
     ]
    }
   ],
   "source": [
    "print(overall_df.shape)\n",
    "print(overall_df_filt.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 935,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Melanocytes    131\n",
       "GDSD6           93\n",
       "Colon           82\n",
       "GM12878         63\n",
       "MM              53\n",
       "SCC             45\n",
       "Name: tissue, dtype: int64"
      ]
     },
     "execution_count": 935,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overall_df_filt.tissue.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 936,
   "metadata": {},
   "outputs": [],
   "source": [
    "overall_df_filt.to_csv(os.path.join(save_dir, 'overall_df_filt.csv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2C plots in 8E1R "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. pull the genomic instances and pick 10 each"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3A get genes\n",
    "(melanoma is differential genes\n",
    "scc is higher threshold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 937,
   "metadata": {},
   "outputs": [],
   "source": [
    "# melanoma genes\n",
    "cancer_tissues = ['COLO_SCR_DMSO', 'WM_SCR_DMSO']  # SKMEL5_SCR_DMSO doesn't have HiChIP data\n",
    "normal_tissues = ['Melanocytes']#'D0-CTRLi', 'D3-CTRLi']\n",
    "sel_tissues = cancer_tissues+normal_tissues\n",
    "tissues_dict = dict(zip(sel_tissues,range(len(sel_tissues))))\n",
    "rna_df_sel = rna_df[sel_tissues]\n",
    "\n",
    "\n",
    "rna_df_sel_high = rna_df_sel[rna_df_sel.max(axis=1)>1]\n",
    "rna_df_sel_high_pseudo = rna_df_sel_high+1\n",
    "\n",
    "rna_df_sel_high_pseudo['avg_ratio'] = (rna_df_sel_high_pseudo.COLO_SCR_DMSO+rna_df_sel_high_pseudo.WM_SCR_DMSO)/(2*rna_df_sel_high_pseudo.Melanocytes)\n",
    "rna_df_sel_high_pseudo['min_ratio'] = rna_df_sel_high_pseudo[['COLO_SCR_DMSO','WM_SCR_DMSO']].min(axis=1)/(rna_df_sel_high_pseudo.Melanocytes)\n",
    "\n",
    "rna_df_sel_high_pseudo.sort_values('min_ratio',ascending=False, inplace=True)\n",
    "MM_genes = sorted(rna_df_sel_high_pseudo[rna_df_sel_high_pseudo.min_ratio>2].index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 938,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "colon_geneset = list(rna_df['Colon'][rna_df['Colon']>5].index)\n",
    "GDSD6_geneset = list(rna_df['GDSD6'][rna_df['GDSD6']>5].index)\n",
    "GM12878_geneset = list(rna_df['GM12878'][rna_df['GM12878']>5].index)\n",
    "MC_geneset = list(rna_df['Melanocytes'][rna_df['Melanocytes']>3].index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 939,
   "metadata": {},
   "outputs": [],
   "source": [
    "## SCC genes\n",
    "THRES=10\n",
    "SCC_genes = rna_df.index[rna_df[['A431-CTRLi','CAL27-CTRLi','SCC13-CTRLi']].min(axis=1)>THRES].values\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 940,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Colon 6826\n",
      "GDSD6 6442\n",
      "GM12878 6166\n",
      "Melanocytes 8688\n",
      "SCC 3916\n",
      "MM 1143\n"
     ]
    }
   ],
   "source": [
    "geneset_dict  = {\n",
    "    'Colon':colon_geneset,\n",
    "    'GDSD6':GDSD6_geneset,\n",
    "    'GM12878':GM12878_geneset,\n",
    "    'Melanocytes':MC_geneset,\n",
    "    'SCC':SCC_genes,\n",
    "    'MM':MM_genes\n",
    "}\n",
    "for tissue,geneset in geneset_dict.items():\n",
    "    print(tissue, len(geneset))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3B set up global variables and dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 941,
   "metadata": {},
   "outputs": [],
   "source": [
    "tissue_to_celltype_mapper = {\n",
    "    'Colon':['Colon'],\n",
    "    'GDSD6':['GDSD6'],\n",
    "    'GM12878':['GM12878'],\n",
    "    'Melanocytes':['Melanocytes'],\n",
    "    'SCC':['A431-CTRLi','CAL27-CTRLi','SCC13-CTRLi'],\n",
    "    'MM':['COLO-SCR-DMSO','WM-SCR-DMSO']\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 942,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1min 27s, sys: 24.2 s, total: 1min 52s\n",
      "Wall time: 1min 35s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# import\n",
    "data_all = pd.read_csv('/Users/mguo123/Google Drive/1_khavari/omics_project-LD/pan_omics/data/processed/tissue_crms/all_count_comb_overall.csv',index_col=0,header=0)\n",
    "data_all = data_all[data_all.tissue.isin(['Colon','GDSD6','GM12878','Melanocytes','A431-CTRLi','CAL27-CTRLi','SCC13-CTRLi','COLO_SCR_DMSO','WM_SCR_DMSO'])]\n",
    "data_all = data_all[data_all.iloc[:,2:].sum(axis=1)>1e-1]\n",
    "\n",
    "# expression labels\n",
    "exp_label = list(np.log10(data_all.exp.values+1e-2))\n",
    "labels_all  = np.array(np.array(exp_label)>THRES)\n",
    "\n",
    "\n",
    "tissues_label  = data_all.tissue.values#np.array((data_all.exp>THRES).values)\n",
    "tissue_num_labels =  data_all.tissue.map(tissues_dict).values\n",
    "is_cancer_label = np.array(['cancer' if x in cancer_tissues  else 'normal' for x in tissues_label ])\n",
    "tissues_label[tissues_label=='COLO_SCR_DMSO'] = 'COLO-SCR-DMSO'\n",
    "tissues_label[tissues_label=='WM_SCR_DMSO'] = 'WM-SCR-DMSO'\n",
    "\n",
    "genes_all = data_all.index.values\n",
    "gene_to_num_dict = dict(zip(np.unique(genes_all),range(len(np.unique(genes_all)))))\n",
    "genes_num_all = np.vectorize(gene_to_num_dict.get)(genes_all)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 943,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 35.5 s, sys: 2.58 s, total: 38.1 s\n",
      "Wall time: 25.3 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "tissue_to_foot_df = {}\n",
    "foot_tissues = ['Colon','GDSD6','GM12878','Melanocytes','A431-CTRLi','CAL27-CTRLi','SCC13-CTRLi','COLO-SCR-DMSO','WM-SCR-DMSO']\n",
    "for tissue in foot_tissues:\n",
    "    foot_df = pd.read_csv(os.path.join('../data/interim/annon/promoter_motif_scan/','promoter_'+tissue+'_annon.bed'),sep='\\t',header=None)\n",
    "    foot_df.columns = ['chr','start','stop','TSS','chr_m','start_m','stop_m','id_trim','score','strand']\n",
    "    tissue_to_foot_df[tissue] = foot_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "tf_annon_df = pd.read_csv('../data/external/HOCOMOCOv11_annotation.csv',index_col=0)\n",
    "tf_annon_df['id_trim'] = tf_annon_df['id'] + '.pwm.trim'\n",
    "tf_name_to_id_dict = pd.Series(tf_annon_df.id_trim.values, index=tf_annon_df.tf.values).to_dict()\n",
    "tf_id_to_name_dict = pd.Series(tf_annon_df.tf.values, index=tf_annon_df.id_trim.values).to_dict()\n",
    "tf_name_to_id_abbr_dict = pd.Series(tf_annon_df.id.values, index=tf_annon_df.tf.values).to_dict()\n",
    "tf_id_abbr_to_name_dict = pd.Series(tf_annon_df.tf.values, index=tf_annon_df.id.values).to_dict()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 945,
   "metadata": {},
   "outputs": [],
   "source": [
    "MAX_DIST=135\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3C helper functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 946,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_tf_tf_loop_occur(tf_tf_pair, tissue, geneset, data_all=data_all,tissues_label=tissues_label):  #suffixes='loop', 'pro_loop'\n",
    "    tf_1, tf_2 = tf_tf_pair.split('::')\n",
    "    \n",
    "    # get features to look at together, list of lists\n",
    "    feat_list_type = {\n",
    "        'pro_pro':[tf_1 + '_pro', tf_2+'_pro'],\n",
    "        'loop_loop':[tf_1 + '_loop', tf_2+'_loop'],\n",
    "        'pro_loop':[tf_1 + '_pro', tf_2+'_loop'],\n",
    "        'loop_pro':[tf_1 + '_loop', tf_2+'_pro']}\n",
    "  \n",
    "    data_all_sel = data_all[data_all.index.isin(geneset) & (tissues_label==tissue)]\n",
    "    feat_type_dict = {}\n",
    "    for type_feat, feat_list in feat_list_type.items():\n",
    "        try:\n",
    "            check_series = pd.DataFrame(data_all_sel[feat_list]>0).all(axis=1)\n",
    "            feat_type_dict[type_feat+\"_count\"] = check_series.value_counts()[True]  \n",
    "            feat_type_dict[type_feat+\"_genes\"] = '|'.join(check_series.index[check_series])\n",
    "            \n",
    "        except:\n",
    "            feat_type_dict[type_feat+\"_count\"] = 0\n",
    "            feat_type_dict[type_feat+\"_genes\"] = ''\n",
    "            \n",
    "    return feat_type_dict\n",
    "\n",
    "def get_other_vocab_word(row, next_row):\n",
    "    if row['vocab']!=next_row['vocab']:\n",
    "        return False\n",
    "#     else row['vocab']==next_row['vocab']\n",
    "#     vocab_word = set(row['tf'])\n",
    "#     vocab_set = set(row['vocab'].split('::'))\n",
    "#     other_vocab = list(vocab_set -  vocab_word)[0]\n",
    "    return row['tf'] != next_row['tf']\n",
    "\n",
    "def check_distance(row,next_row,max_dist=MAX_DIST):\n",
    "    if row['chr_m']==next_row['chr_m']:\n",
    "        if row['stop_m']<next_row['start_m']:\n",
    "            tot_dist = next_row['stop_m'] - row['start_m']\n",
    "            btn_dist = next_row['start_m'] - row['stop_m']\n",
    "            return (tot_dist < max_dist), tot_dist, btn_dist\n",
    "    return False,-1,-1\n",
    "\n",
    "def check_tissue(row,next_row,tfs=tf_name_to_id_dict.keys()):\n",
    "    if (row['tf'] in tfs) & (next_row['tf'] in tfs):\n",
    "        return 'both'\n",
    "    elif (row['tf'] in tfs) | (next_row['tf'] in tfs):\n",
    "        return 'one'    \n",
    "    else:\n",
    "        return 'none'\n",
    "    \n",
    "def get_hits(vocab_df,tfs=tf_name_to_id_dict.keys()):\n",
    "#     print('**** reading', vocab_file)\n",
    "#     vocab_df = pd.read_csv(vocab_file)\n",
    "    print(vocab_df.shape)\n",
    "    vocab_df.sort_values(['chr','start','stop'],inplace=True)\n",
    "    idx = 0\n",
    "    idx_hits = 0\n",
    "    results_dict = {}\n",
    "    while idx < (vocab_df.shape[0]-1):\n",
    "        # look at next \n",
    "        row = vocab_df.iloc[idx,:]\n",
    "        next_row = vocab_df.iloc[idx+1,:]\n",
    "\n",
    "        check_vocab_pair = get_other_vocab_word(row,next_row)\n",
    "        check_dist,tot_dist, btn_dist = check_distance(row, next_row)\n",
    "        check_tissue_tf = check_tissue(row,next_row,tfs)\n",
    "        if (check_dist and check_vocab_pair):\n",
    "    #         print('hi',idx)\n",
    "    #         print(row)\n",
    "    #         print(next_row)\n",
    "\n",
    "            results_dict[idx_hits] = {'vocab_pair':row['vocab'],'tot_dist':tot_dist,'btn_dist':btn_dist,\n",
    "                                      'chr':row['chr'],'start':row['start_m'],'stop':next_row['stop_m'], \n",
    "                                      'vocab1':row['tf'],'vocab1_start':row['start_m'], 'vocab1_stop': row['stop_m'],\n",
    "                                      'vocab2':next_row['tf'],'vocab2_start':next_row['start_m'], 'vocab2_stop': next_row['stop_m'],\n",
    "                                      'genes':row['genes'],'num_genes':len(row['genes'].split('|')), 'tissue':row['tissue'],#'foot_tissue':row['foot_tissue'],\n",
    "                                     'check_tissuetf':check_tissue_tf, 'loop_type':row['type_loop']}\n",
    "            idx_hits+=1\n",
    "\n",
    "        idx+=1\n",
    "    print('num_hits',idx_hits)\n",
    "    results_df = pd.DataFrame.from_dict(results_dict, orient='index').drop_duplicates()\n",
    "    return results_df\n",
    "    \n",
    "def filter_results(results_df,min_hits_per_vocab=10):\n",
    "    print('shape', results_df.shape)\n",
    "    vocab_counts = results_df[['vocab_pair','chr','start','stop']].drop_duplicates().vocab_pair.value_counts()\n",
    "    print('original num vocab',vocab_counts.shape[0])\n",
    "    vocab_to_include = vocab_counts[vocab_counts>min_hits_per_vocab].index.values\n",
    "    print('filt num vocab',vocab_to_include.shape[0])\n",
    "    \n",
    "    results_df_filt = results_df[results_df.vocab_pair.isin(vocab_to_include)]\n",
    "    return results_df_filt\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 947,
   "metadata": {},
   "outputs": [],
   "source": [
    "# debugging get_hits()\n",
    "# vocab_df = vocab_tissue_dict['MM']\n",
    "# vocab_tissue_dict['MM'][vocab_tissue_dict['MM'].vocab =='IRF3::CEBPZ']\n",
    "# vocab_instances_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3D get genomic instances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 948,
   "metadata": {},
   "outputs": [],
   "source": [
    "vocab_tissue_dict = {}\n",
    "for tissue in geneset_dict.keys():\n",
    "    vocab_tissue_dict[tissue] = pd.DataFrame()\n",
    "\n",
    "#     print(tissue, len(geneset))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 949,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "***** 0 BARX2::DLX4 Colon\n",
      "Colon pro_pro (12, 16)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:19: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Colon loop_loop (108, 16)\n",
      "***** 1 BARX2::FOXF2 Colon\n",
      "Colon pro_pro (80, 16)\n",
      "Colon loop_loop (408, 16)\n",
      "***** 2 BARX2::FOXM1 Colon\n",
      "Colon pro_pro (54, 16)\n",
      "Colon loop_loop (360, 16)\n",
      "***** 3 BARX2::THAP1 Colon\n",
      "Colon pro_pro (108, 16)\n",
      "Colon loop_loop (1223, 16)\n",
      "***** 4 BATF::KLF15 Colon\n",
      "Colon pro_pro (33, 16)\n",
      "Colon loop_loop (1130, 16)\n",
      "***** 5 DLX4::SOX13 Colon\n",
      "Colon loop_loop (134, 16)\n",
      "***** 6 E2F7::THAP1 Colon\n",
      "Colon pro_pro (106, 16)\n",
      "Colon loop_loop (715, 16)\n",
      "***** 7 E2F7::FOXM1 Colon\n",
      "Colon loop_loop (63, 16)\n",
      "***** 8 FLI1::FOXM1 Colon\n",
      "Colon pro_pro (14, 16)\n",
      "Colon loop_loop (269, 16)\n",
      "***** 9 FOXA1::THAP1 Colon\n",
      "Colon loop_loop (193, 16)\n",
      "***** 10 FOXA2::THAP1 Colon\n",
      "Colon loop_loop (196, 16)\n",
      "***** 11 FOXF2::KLF5 Colon\n",
      "Colon pro_pro (15, 16)\n",
      "Colon loop_loop (263, 16)\n",
      "***** 12 FOXF2::SP4 Colon\n",
      "Colon pro_pro (33, 16)\n",
      "Colon loop_loop (657, 16)\n",
      "***** 13 FOXF2::THAP1 Colon\n",
      "Colon loop_loop (202, 16)\n",
      "***** 14 FOXM1::KLF15 Colon\n",
      "Colon pro_pro (36, 16)\n",
      "Colon loop_loop (686, 16)\n",
      "***** 15 FOXM1::KLF5 Colon\n",
      "Colon pro_pro (16, 16)\n",
      "Colon loop_loop (280, 16)\n",
      "***** 16 FOXM1::KLF9 Colon\n",
      "Colon pro_pro (12, 16)\n",
      "Colon loop_loop (253, 16)\n",
      "***** 17 FOXM1::MESP1 Colon\n",
      "Colon pro_pro (14, 16)\n",
      "Colon loop_loop (136, 16)\n",
      "***** 18 FOXM1::NR2F1 Colon\n",
      "Colon pro_pro (6, 16)\n",
      "Colon loop_loop (140, 16)\n",
      "***** 19 FOXM1::PATZ1 Colon\n",
      "Colon pro_pro (51, 16)\n",
      "Colon loop_loop (950, 16)\n",
      "***** 20 FOXM1::PLAG1 Colon\n",
      "Colon pro_pro (7, 16)\n",
      "Colon loop_loop (168, 16)\n",
      "***** 21 FOXM1::PPARG Colon\n",
      "Colon pro_pro (4, 16)\n",
      "Colon loop_loop (116, 16)\n",
      "***** 22 FOXM1::RARG Colon\n",
      "Colon pro_pro (5, 16)\n",
      "Colon loop_loop (138, 16)\n",
      "***** 23 FOXM1::RORA Colon\n",
      "Colon pro_pro (6, 16)\n",
      "Colon loop_loop (98, 16)\n",
      "***** 24 FOXM1::SNAI1 Colon\n",
      "Colon pro_pro (14, 16)\n",
      "Colon loop_loop (136, 16)\n",
      "***** 25 FOXM1::SOX13 Colon\n",
      "Colon pro_pro (83, 16)\n",
      "Colon loop_loop (625, 16)\n",
      "***** 26 FOXM1::SP2 Colon\n",
      "Colon pro_pro (32, 16)\n",
      "Colon loop_loop (1192, 16)\n",
      "***** 27 FOXM1::SP4 Colon\n",
      "Colon pro_pro (30, 16)\n",
      "Colon loop_loop (768, 16)\n",
      "***** 28 FOXM1::THAP1 Colon\n",
      "Colon loop_loop (215, 16)\n",
      "***** 29 FOXM1::WT1 Colon\n",
      "Colon pro_pro (53, 16)\n",
      "Colon loop_loop (1069, 16)\n",
      "***** 31 GATA4::THAP1 Colon\n",
      "Colon loop_loop (58, 16)\n",
      "***** 32 GRHL2::THAP1 Colon\n",
      "Colon loop_loop (127, 16)\n",
      "***** 33 HOXA11::KLF15 Colon\n",
      "Colon pro_pro (12, 16)\n",
      "Colon loop_loop (203, 16)\n",
      "***** 34 HOXA11::KLF5 Colon\n",
      "Colon pro_pro (3, 16)\n",
      "Colon loop_loop (80, 16)\n",
      "***** 35 HOXB3::KLF5 Colon\n",
      "Colon pro_pro (11, 16)\n",
      "Colon loop_loop (126, 16)\n",
      "***** 36 HOXC8::KLF5 Colon\n",
      "Colon loop_loop (147, 16)\n",
      "***** 37 HOXC8::THAP1 Colon\n",
      "Colon loop_loop (129, 16)\n",
      "***** 38 KLF15::KLF5 Colon\n",
      "Colon pro_pro (1050, 16)\n",
      "Colon loop_loop (3718, 16)\n",
      "***** 39 KLF15::MSX2 Colon\n",
      "Colon pro_pro (159, 16)\n",
      "Colon loop_loop (1522, 16)\n",
      "***** 40 KLF15::OSR2 Colon\n",
      "Colon pro_pro (98, 16)\n",
      "Colon loop_loop (989, 16)\n",
      "***** 41 KLF15::OTX1 Colon\n",
      "Colon pro_pro (98, 16)\n",
      "Colon loop_loop (1602, 16)\n",
      "***** 42 KLF15::OVOL2 Colon\n",
      "Colon pro_pro (22, 16)\n",
      "Colon loop_loop (391, 16)\n",
      "***** 43 KLF15::RFX2 Colon\n",
      "Colon pro_pro (48, 16)\n",
      "Colon loop_loop (1039, 16)\n",
      "***** 44 KLF15::SNAI2 Colon\n",
      "Colon pro_pro (108, 16)\n",
      "Colon loop_loop (1203, 16)\n",
      "***** 45 KLF15::THAP1 Colon\n",
      "Colon pro_pro (568, 16)\n",
      "Colon loop_loop (2620, 16)\n",
      "***** 46 KLF15::ZFP28 Colon\n",
      "Colon pro_pro (83, 16)\n",
      "Colon loop_loop (822, 16)\n",
      "***** 47 KLF5::NR2F1 Colon\n",
      "Colon pro_pro (92, 16)\n",
      "Colon loop_loop (1022, 16)\n",
      "***** 48 KLF5::NR4A2 Colon\n",
      "Colon pro_pro (16, 16)\n",
      "Colon loop_loop (214, 16)\n",
      "***** 49 KLF5::RUNX3 Colon\n",
      "Colon pro_pro (39, 16)\n",
      "Colon loop_loop (237, 16)\n",
      "***** 50 KLF5::SOX13 Colon\n",
      "Colon pro_pro (374, 16)\n",
      "Colon loop_loop (2941, 16)\n",
      "***** 51 KLF5::ZFP28 Colon\n",
      "Colon pro_pro (25, 16)\n",
      "Colon loop_loop (369, 16)\n",
      "***** 52 KLF5::ZNF257 Colon\n",
      "Colon pro_pro (24, 16)\n",
      "Colon loop_loop (456, 16)\n",
      "***** 53 MAFB::THAP1 Colon\n",
      "Colon pro_pro (20, 16)\n",
      "Colon loop_loop (376, 16)\n",
      "***** 54 MESP1::THAP1 Colon\n",
      "Colon pro_pro (61, 16)\n",
      "Colon loop_loop (803, 16)\n",
      "***** 55 MSX2::THAP1 Colon\n",
      "Colon pro_pro (33, 16)\n",
      "Colon loop_loop (499, 16)\n",
      "***** 56 NR2F1::THAP1 Colon\n",
      "Colon pro_pro (72, 16)\n",
      "Colon loop_loop (761, 16)\n",
      "***** 57 NR4A2::THAP1 Colon\n",
      "Colon pro_pro (13, 16)\n",
      "Colon loop_loop (152, 16)\n",
      "***** 58 OSR2::THAP1 Colon\n",
      "Colon pro_pro (24, 16)\n",
      "Colon loop_loop (316, 16)\n",
      "***** 59 OTX1::THAP1 Colon\n",
      "Colon pro_pro (36, 16)\n",
      "Colon loop_loop (492, 16)\n",
      "***** 60 OVOL2::THAP1 Colon\n",
      "Colon loop_loop (83, 16)\n",
      "***** 61 RARA::THAP1 Colon\n",
      "Colon pro_pro (60, 16)\n",
      "Colon loop_loop (720, 16)\n",
      "***** 62 RARG::THAP1 Colon\n",
      "Colon pro_pro (64, 16)\n",
      "Colon loop_loop (686, 16)\n",
      "***** 63 RFX2::THAP1 Colon\n",
      "Colon pro_pro (16, 16)\n",
      "Colon loop_loop (346, 16)\n",
      "***** 64 RORA::THAP1 Colon\n",
      "Colon pro_pro (71, 16)\n",
      "Colon loop_loop (750, 16)\n",
      "***** 65 RUNX2::THAP1 Colon\n",
      "Colon pro_pro (3, 16)\n",
      "Colon loop_loop (239, 16)\n",
      "***** 66 SNAI2::THAP1 Colon\n",
      "Colon pro_pro (14, 16)\n",
      "Colon loop_loop (419, 16)\n",
      "***** 67 SNAI2::WT1 Colon\n",
      "Colon pro_pro (251, 16)\n",
      "Colon loop_loop (1754, 16)\n",
      "***** 68 SOX13::THAP1 Colon\n",
      "Colon pro_pro (142, 16)\n",
      "Colon loop_loop (1928, 16)\n",
      "***** 69 SP2::THAP1 Colon\n",
      "Colon pro_pro (1424, 16)\n",
      "Colon loop_loop (4577, 16)\n",
      "***** 70 SP4::THAP1 Colon\n",
      "Colon pro_pro (749, 16)\n",
      "Colon loop_loop (3014, 16)\n",
      "***** 71 TCF7::THAP1 Colon\n",
      "Colon pro_pro (2, 16)\n",
      "Colon loop_loop (160, 16)\n",
      "***** 72 THAP1::THRB Colon\n",
      "Colon pro_pro (15, 16)\n",
      "Colon loop_loop (355, 16)\n",
      "***** 73 THAP1::ZFP28 Colon\n",
      "Colon pro_pro (31, 16)\n",
      "Colon loop_loop (246, 16)\n",
      "***** 74 THAP1::ZNF333 Colon\n",
      "Colon loop_loop (88, 16)\n",
      "***** 75 ALX1::PRRX1 Melanocytes\n",
      "Melanocytes pro_pro (14, 16)\n",
      "Melanocytes loop_loop (111, 16)\n",
      "***** 76 E2F4::KLF15 Melanocytes\n",
      "Melanocytes pro_pro (419, 16)\n",
      "Melanocytes loop_loop (462, 16)\n",
      "***** 77 EN2::PRRX1 Melanocytes\n",
      "Melanocytes pro_pro (539, 16)\n",
      "Melanocytes loop_loop (546, 16)\n",
      "***** 78 ETV2::PRRX1 Melanocytes\n",
      "Melanocytes pro_pro (282, 16)\n",
      "Melanocytes loop_loop (503, 16)\n",
      "***** 79 ETV2::SP2 Melanocytes\n",
      "Melanocytes pro_pro (417, 16)\n",
      "Melanocytes loop_loop (612, 16)\n",
      "***** 80 FOXD3::PRRX1 Melanocytes\n",
      "Melanocytes pro_pro (248, 16)\n",
      "Melanocytes loop_loop (347, 16)\n",
      "***** 81 FOXD3::ZNF589 Melanocytes\n",
      "Melanocytes pro_pro (60, 16)\n",
      "Melanocytes loop_loop (100, 16)\n",
      "***** 82 FOXL1::PRRX1 Melanocytes\n",
      "Melanocytes pro_pro (284, 16)\n",
      "Melanocytes loop_loop (312, 16)\n",
      "***** 83 HES7::PRRX1 Melanocytes\n",
      "Melanocytes pro_pro (11, 16)\n",
      "Melanocytes loop_loop (58, 16)\n",
      "***** 84 IRF4::KLF12 Melanocytes\n",
      "Melanocytes pro_pro (71, 16)\n",
      "Melanocytes loop_loop (135, 16)\n",
      "***** 85 IRF4::PRRX1 Melanocytes\n",
      "Melanocytes pro_pro (287, 16)\n",
      "Melanocytes loop_loop (428, 16)\n",
      "***** 86 KLF15::PRRX1 Melanocytes\n",
      "Melanocytes pro_pro (1164, 16)\n",
      "Melanocytes loop_loop (1935, 16)\n",
      "***** 87 KLF15::SALL4 Melanocytes\n",
      "Melanocytes pro_pro (335, 16)\n",
      "Melanocytes loop_loop (419, 16)\n",
      "***** 88 KLF15::TFAP2B Melanocytes\n",
      "Melanocytes pro_pro (366, 16)\n",
      "Melanocytes loop_loop (458, 16)\n",
      "***** 89 KLF15::ZNF219 Melanocytes\n",
      "Melanocytes pro_pro (817, 16)\n",
      "Melanocytes loop_loop (852, 16)\n",
      "***** 90 KLF15::ZNF589 Melanocytes\n",
      "Melanocytes pro_pro (697, 16)\n",
      "Melanocytes loop_loop (1252, 16)\n",
      "***** 91 MITF::PRRX1 Melanocytes\n",
      "Melanocytes pro_pro (101, 16)\n",
      "Melanocytes loop_loop (170, 16)\n",
      "***** 92 MITF::SP2 Melanocytes\n",
      "Melanocytes pro_pro (78, 16)\n",
      "Melanocytes loop_loop (151, 16)\n",
      "***** 93 NFATC1::PRRX1 Melanocytes\n",
      "Melanocytes pro_pro (211, 16)\n",
      "Melanocytes loop_loop (372, 16)\n",
      "***** 94 PAX3::PRRX1 Melanocytes\n",
      "Melanocytes pro_pro (131, 16)\n",
      "Melanocytes loop_loop (91, 16)\n",
      "***** 95 PBX1::PRRX1 Melanocytes\n",
      "Melanocytes pro_pro (242, 16)\n",
      "Melanocytes loop_loop (337, 16)\n",
      "***** 96 PRRX1::RARA Melanocytes\n",
      "Melanocytes pro_pro (159, 16)\n",
      "Melanocytes loop_loop (604, 16)\n",
      "***** 97 PRRX1::SALL4 Melanocytes\n",
      "Melanocytes pro_pro (326, 16)\n",
      "Melanocytes loop_loop (540, 16)\n",
      "***** 98 PRRX1::SOX10 Melanocytes\n",
      "Melanocytes pro_pro (82, 16)\n",
      "Melanocytes loop_loop (232, 16)\n",
      "***** 99 PRRX1::TBX2 Melanocytes\n",
      "Melanocytes pro_pro (46, 16)\n",
      "Melanocytes loop_loop (60, 16)\n",
      "***** 100 PRRX1::TFAP2B Melanocytes\n",
      "Melanocytes pro_pro (185, 16)\n",
      "Melanocytes loop_loop (687, 16)\n",
      "***** 101 PRRX1::TFEB Melanocytes\n",
      "Melanocytes pro_pro (29, 16)\n",
      "Melanocytes loop_loop (53, 16)\n",
      "***** 102 SP2::TFAP2B Melanocytes\n",
      "Melanocytes pro_pro (982, 16)\n",
      "Melanocytes loop_loop (991, 16)\n",
      "***** 103 SP2::ZNF324 Melanocytes\n",
      "Melanocytes pro_pro (375, 16)\n",
      "Melanocytes loop_loop (622, 16)\n",
      "***** 104 TFAP2B::ZNF589 Melanocytes\n",
      "Melanocytes pro_pro (72, 16)\n",
      "Melanocytes loop_loop (275, 16)\n",
      "***** 105 NFATC4::SOX10 Melanocytes\n",
      "Melanocytes pro_pro (23, 16)\n",
      "Melanocytes loop_loop (33, 16)\n",
      "***** 106 PAX3::ZNF589 Melanocytes\n",
      "Melanocytes pro_pro (31, 16)\n",
      "Melanocytes loop_loop (24, 16)\n",
      "***** 107 SOX10::ZNF589 Melanocytes\n",
      "Melanocytes pro_pro (19, 16)\n",
      "Melanocytes loop_loop (62, 16)\n",
      "***** 108 ETV2::ZNF589 Melanocytes\n",
      "Melanocytes pro_pro (86, 16)\n",
      "Melanocytes loop_loop (201, 16)\n",
      "***** 109 LEF1::MITF Melanocytes\n",
      "Melanocytes pro_pro (2, 16)\n",
      "Melanocytes loop_loop (7, 16)\n",
      "***** 110 IRF4::MITF Melanocytes\n",
      "Melanocytes pro_pro (2, 16)\n",
      "Melanocytes loop_loop (13, 16)\n",
      "***** 111 MITF::PAX3 Melanocytes\n",
      "Melanocytes loop_loop (0, 16)\n",
      "***** 112 MITF::TFAP2B Melanocytes\n",
      "Melanocytes pro_pro (10, 16)\n",
      "Melanocytes loop_loop (12, 16)\n",
      "***** 113 MITF::SOX10 Melanocytes\n",
      "Melanocytes loop_loop (1, 16)\n",
      "***** 114 MITF::SOX11 Melanocytes\n",
      "Melanocytes pro_pro (2, 16)\n",
      "Melanocytes loop_loop (7, 16)\n",
      "***** 115 MITF::SOX5 Melanocytes\n",
      "Melanocytes loop_loop (0, 16)\n",
      "***** 116 ALX1::MITF Melanocytes\n",
      "Melanocytes loop_loop (0, 16)\n",
      "***** 117 GLI3::MITF Melanocytes\n",
      "Melanocytes pro_pro (2, 16)\n",
      "Melanocytes loop_loop (2, 16)\n",
      "***** 118 CPEB1::MITF Melanocytes\n",
      "Melanocytes loop_loop (0, 16)\n",
      "***** 119 CREB3L1::MITF Melanocytes\n",
      "Melanocytes pro_pro (8, 16)\n",
      "Melanocytes loop_loop (8, 16)\n",
      "***** 120 MITF::TCF4 Melanocytes\n",
      "Melanocytes loop_loop (0, 16)\n",
      "***** 121 LEF1::PAX3 Melanocytes\n",
      "Melanocytes loop_loop (0, 16)\n",
      "***** 122 PAX3::SOX10 Melanocytes\n",
      "Melanocytes loop_loop (0, 16)\n",
      "***** 123 PAX3::TCF4 Melanocytes\n",
      "Melanocytes pro_pro (2, 16)\n",
      "Melanocytes loop_loop (0, 16)\n",
      "***** 124 PAX3::TFAP2B Melanocytes\n",
      "Melanocytes loop_loop (2, 16)\n",
      "***** 125 LEF1::SOX10 Melanocytes\n",
      "Melanocytes loop_loop (6, 16)\n",
      "***** 126 LEF1::TCF4 Melanocytes\n",
      "Melanocytes loop_loop (0, 16)\n",
      "***** 127 LEF1::TFAP2B Melanocytes\n",
      "Melanocytes loop_loop (15, 16)\n",
      "***** 128 GLI3::LEF1 Melanocytes\n",
      "Melanocytes loop_loop (1, 16)\n",
      "***** 129 ARNT2::MITF Melanocytes\n",
      "Melanocytes pro_pro (19, 16)\n",
      "Melanocytes loop_loop (26, 16)\n",
      "***** 130 BATF3::MITF Melanocytes\n",
      "Melanocytes pro_pro (2, 16)\n",
      "Melanocytes loop_loop (2, 16)\n",
      "***** 131 E2F1::MITF Melanocytes\n",
      "Melanocytes pro_pro (4, 16)\n",
      "Melanocytes loop_loop (4, 16)\n",
      "***** 132 E2F2::MITF Melanocytes\n",
      "Melanocytes pro_pro (2, 16)\n",
      "Melanocytes loop_loop (0, 16)\n",
      "***** 133 E2F4::MITF Melanocytes\n",
      "Melanocytes pro_pro (7, 16)\n",
      "Melanocytes loop_loop (7, 16)\n",
      "***** 134 E2F7::MITF Melanocytes\n",
      "Melanocytes pro_pro (4, 16)\n",
      "Melanocytes loop_loop (6, 16)\n",
      "***** 135 EBF1::MITF Melanocytes\n",
      "Melanocytes loop_loop (12, 16)\n",
      "***** 136 EN2::MITF Melanocytes\n",
      "Melanocytes loop_loop (0, 16)\n",
      "***** 137 ETV1::MITF Melanocytes\n",
      "Melanocytes loop_loop (0, 16)\n",
      "***** 138 ETV2::MITF Melanocytes\n",
      "Melanocytes pro_pro (2, 16)\n",
      "Melanocytes loop_loop (6, 16)\n",
      "***** 139 FOXD3::MITF Melanocytes\n",
      "Melanocytes loop_loop (3, 16)\n",
      "***** 140 FOXF2::MITF Melanocytes\n",
      "Melanocytes pro_pro (2, 16)\n",
      "Melanocytes loop_loop (1, 16)\n",
      "***** 141 FOXL1::MITF Melanocytes\n",
      "Melanocytes loop_loop (0, 16)\n",
      "***** 142 GLIS3::MITF Melanocytes\n",
      "Melanocytes loop_loop (0, 16)\n",
      "***** 143 HES7::MITF Melanocytes\n",
      "Melanocytes pro_pro (3, 16)\n",
      "Melanocytes loop_loop (3, 16)\n",
      "***** 144 HEY1::MITF Melanocytes\n",
      "Melanocytes pro_pro (16, 16)\n",
      "Melanocytes loop_loop (16, 16)\n",
      "***** 145 HOXB3::MITF Melanocytes\n",
      "Melanocytes loop_loop (0, 16)\n",
      "***** 146 HOXB8::MITF Melanocytes\n",
      "Melanocytes loop_loop (0, 16)\n",
      "***** 147 HOXC10::MITF Melanocytes\n",
      "Melanocytes loop_loop (0, 16)\n",
      "***** 148 HOXC11::MITF Melanocytes\n",
      "Melanocytes pro_pro (2, 16)\n",
      "Melanocytes loop_loop (0, 16)\n",
      "***** 149 HOXC6::MITF Melanocytes\n",
      "Melanocytes loop_loop (0, 16)\n",
      "***** 150 HOXC8::MITF Melanocytes\n",
      "Melanocytes loop_loop (2, 16)\n",
      "***** 151 HOXD3::MITF Melanocytes\n",
      "Melanocytes loop_loop (0, 16)\n",
      "***** 152 HOXD8::MITF Melanocytes\n",
      "Melanocytes pro_pro (3, 16)\n",
      "Melanocytes loop_loop (3, 16)\n",
      "***** 153 HSF4::MITF Melanocytes\n",
      "Melanocytes loop_loop (0, 16)\n",
      "***** 154 ID4::MITF Melanocytes\n",
      "Melanocytes pro_pro (3, 16)\n",
      "Melanocytes loop_loop (3, 16)\n",
      "***** 155 KLF12::MITF Melanocytes\n",
      "Melanocytes pro_pro (23, 16)\n",
      "Melanocytes loop_loop (36, 16)\n",
      "***** 156 KLF15::MITF Melanocytes\n",
      "Melanocytes pro_pro (33, 16)\n",
      "Melanocytes loop_loop (94, 16)\n",
      "***** 157 KLF9::MITF Melanocytes\n",
      "Melanocytes pro_pro (12, 16)\n",
      "Melanocytes loop_loop (27, 16)\n",
      "***** 158 LYL1::MITF Melanocytes\n",
      "Melanocytes loop_loop (1, 16)\n",
      "***** 159 MAF::MITF Melanocytes\n",
      "Melanocytes loop_loop (0, 16)\n",
      "***** 160 MEF2C::MITF Melanocytes\n",
      "Melanocytes loop_loop (1, 16)\n",
      "***** 161 MEIS1::MITF Melanocytes\n",
      "Melanocytes pro_pro (7, 16)\n",
      "Melanocytes loop_loop (0, 16)\n",
      "***** 162 MEIS3::MITF Melanocytes\n",
      "Melanocytes loop_loop (0, 16)\n",
      "***** 163 MITF::MYOD1 Melanocytes\n",
      "Melanocytes loop_loop (2, 16)\n",
      "***** 164 MITF::NFATC1 Melanocytes\n",
      "Melanocytes pro_pro (3, 16)\n",
      "Melanocytes loop_loop (8, 16)\n",
      "***** 165 MITF::NFATC2 Melanocytes\n",
      "Melanocytes pro_pro (4, 16)\n",
      "Melanocytes loop_loop (8, 16)\n",
      "***** 166 MITF::NFATC4 Melanocytes\n",
      "Melanocytes pro_pro (24, 16)\n",
      "Melanocytes loop_loop (56, 16)\n",
      "***** 167 MITF::NR4A3 Melanocytes\n",
      "Melanocytes loop_loop (0, 16)\n",
      "***** 168 MITF::PLAG1 Melanocytes\n",
      "Melanocytes pro_pro (5, 16)\n",
      "Melanocytes loop_loop (6, 16)\n",
      "***** 169 MITF::POU3F1 Melanocytes\n",
      "Melanocytes loop_loop (0, 16)\n",
      "***** 170 MITF::RARA Melanocytes\n",
      "Melanocytes pro_pro (2, 16)\n",
      "Melanocytes loop_loop (12, 16)\n",
      "***** 171 MITF::RARB Melanocytes\n",
      "Melanocytes loop_loop (2, 16)\n",
      "***** 172 MITF::RFX3 Melanocytes\n",
      "Melanocytes loop_loop (8, 16)\n",
      "***** 173 MITF::SALL4 Melanocytes\n",
      "Melanocytes pro_pro (8, 16)\n",
      "Melanocytes loop_loop (12, 16)\n",
      "***** 174 MITF::TBX2 Melanocytes\n",
      "Melanocytes loop_loop (10, 16)\n",
      "***** 175 MITF::TFEB Melanocytes\n",
      "Melanocytes pro_pro (4, 16)\n",
      "Melanocytes loop_loop (4, 16)\n",
      "***** 176 MITF::TWIST1 Melanocytes\n",
      "Melanocytes loop_loop (4, 16)\n",
      "***** 177 MITF::ZBTB33 Melanocytes\n",
      "Melanocytes loop_loop (0, 16)\n",
      "***** 178 MITF::ZBTB6 Melanocytes\n",
      "Melanocytes pro_pro (3, 16)\n",
      "Melanocytes loop_loop (3, 16)\n",
      "***** 179 MITF::ZFP82 Melanocytes\n",
      "Melanocytes loop_loop (1, 16)\n",
      "***** 180 MITF::ZNF136 Melanocytes\n",
      "Melanocytes loop_loop (0, 16)\n",
      "***** 181 MITF::ZNF18 Melanocytes\n",
      "Melanocytes pro_pro (5, 16)\n",
      "Melanocytes loop_loop (13, 16)\n",
      "***** 182 MITF::ZNF219 Melanocytes\n",
      "Melanocytes pro_pro (14, 16)\n",
      "Melanocytes loop_loop (25, 16)\n",
      "***** 183 MITF::ZNF324 Melanocytes\n",
      "Melanocytes pro_pro (4, 16)\n",
      "Melanocytes loop_loop (5, 16)\n",
      "***** 184 MITF::ZNF333 Melanocytes\n",
      "Melanocytes loop_loop (0, 16)\n",
      "***** 185 MITF::ZNF382 Melanocytes\n",
      "Melanocytes loop_loop (0, 16)\n",
      "***** 186 MITF::ZNF41 Melanocytes\n",
      "Melanocytes loop_loop (10, 16)\n",
      "***** 187 MITF::ZNF418 Melanocytes\n",
      "Melanocytes pro_pro (4, 16)\n",
      "Melanocytes loop_loop (1, 16)\n",
      "***** 188 MITF::ZNF423 Melanocytes\n",
      "Melanocytes pro_pro (2, 16)\n",
      "Melanocytes loop_loop (0, 16)\n",
      "***** 189 MITF::ZNF436 Melanocytes\n",
      "Melanocytes loop_loop (2, 16)\n",
      "***** 190 MITF::ZNF449 Melanocytes\n",
      "Melanocytes pro_pro (3, 16)\n",
      "Melanocytes loop_loop (6, 16)\n",
      "***** 191 MITF::ZNF549 Melanocytes\n",
      "Melanocytes loop_loop (2, 16)\n",
      "***** 192 MITF::ZNF554 Melanocytes\n",
      "Melanocytes pro_pro (7, 16)\n",
      "Melanocytes loop_loop (8, 16)\n",
      "***** 193 MITF::ZNF589 Melanocytes\n",
      "Melanocytes pro_pro (32, 16)\n",
      "Melanocytes loop_loop (78, 16)\n",
      "***** 194 MITF::ZNF708 Melanocytes\n",
      "Melanocytes loop_loop (2, 16)\n",
      "***** 195 MITF::ZNF76 Melanocytes\n",
      "Melanocytes loop_loop (7, 16)\n",
      "***** 196 ARID5B::KLF5 GDSD6\n",
      "GDSD6 pro_pro (111, 16)\n",
      "GDSD6 loop_loop (1274, 16)\n",
      "***** 197 ASCL2::KLF4 GDSD6\n",
      "GDSD6 pro_pro (40, 16)\n",
      "GDSD6 loop_loop (222, 16)\n",
      "***** 198 ASCL2::KLF5 GDSD6\n",
      "GDSD6 pro_pro (51, 16)\n",
      "GDSD6 loop_loop (272, 16)\n",
      "***** 199 ATF2::KLF5 GDSD6\n",
      "GDSD6 pro_pro (30, 16)\n",
      "GDSD6 loop_loop (295, 16)\n",
      "***** 200 ATF2::RORA GDSD6\n",
      "GDSD6 pro_pro (24, 16)\n",
      "GDSD6 loop_loop (117, 16)\n",
      "***** 201 ATF4::EGR2 GDSD6\n",
      "GDSD6 pro_pro (3, 16)\n",
      "GDSD6 loop_loop (343, 16)\n",
      "***** 202 ATF4::KLF5 GDSD6\n",
      "GDSD6 loop_loop (167, 16)\n",
      "***** 203 BACH1::KLF5 GDSD6\n",
      "GDSD6 pro_pro (13, 16)\n",
      "GDSD6 loop_loop (180, 16)\n",
      "***** 204 BACH1::RORA GDSD6\n",
      "GDSD6 pro_pro (6, 16)\n",
      "GDSD6 loop_loop (70, 16)\n",
      "***** 205 CEBPB::KLF5 GDSD6\n",
      "GDSD6 loop_loop (90, 16)\n",
      "***** 206 CEBPG::KLF5 GDSD6\n",
      "GDSD6 pro_pro (2, 16)\n",
      "GDSD6 loop_loop (194, 16)\n",
      "***** 207 CEBPG::RORA GDSD6\n",
      "GDSD6 pro_pro (3, 16)\n",
      "GDSD6 loop_loop (56, 16)\n",
      "***** 208 CREM::KLF5 GDSD6\n",
      "GDSD6 pro_pro (19, 16)\n",
      "GDSD6 loop_loop (136, 16)\n",
      "***** 209 DBP::KLF5 GDSD6\n",
      "GDSD6 pro_pro (6, 16)\n",
      "GDSD6 loop_loop (105, 16)\n",
      "***** 210 DDIT3::KLF5 GDSD6\n",
      "GDSD6 loop_loop (200, 16)\n",
      "***** 211 DDIT3::RORA GDSD6\n",
      "GDSD6 loop_loop (55, 16)\n",
      "***** 212 DLX5::KLF5 GDSD6\n",
      "GDSD6 loop_loop (50, 16)\n",
      "***** 213 EGR2::FOSL2 GDSD6\n",
      "GDSD6 pro_pro (35, 16)\n",
      "GDSD6 loop_loop (816, 16)\n",
      "***** 214 EGR2::JUN GDSD6\n",
      "GDSD6 pro_pro (35, 16)\n",
      "GDSD6 loop_loop (750, 16)\n",
      "***** 215 EGR2::JUNB GDSD6\n",
      "GDSD6 pro_pro (35, 16)\n",
      "GDSD6 loop_loop (662, 16)\n",
      "***** 216 EGR2::KLF4 GDSD6\n",
      "GDSD6 pro_pro (766, 16)\n",
      "GDSD6 loop_loop (1831, 16)\n",
      "***** 217 EGR2::KLF5 GDSD6\n",
      "GDSD6 pro_pro (790, 16)\n",
      "GDSD6 loop_loop (1968, 16)\n",
      "***** 218 EGR2::RARG GDSD6\n",
      "GDSD6 pro_pro (33, 16)\n",
      "GDSD6 loop_loop (714, 16)\n",
      "***** 219 EGR2::RORA GDSD6\n",
      "GDSD6 pro_pro (175, 16)\n",
      "GDSD6 loop_loop (893, 16)\n",
      "***** 220 EGR2::SNAI2 GDSD6\n",
      "GDSD6 pro_pro (75, 16)\n",
      "GDSD6 loop_loop (427, 16)\n",
      "***** 221 EGR2::TP63 GDSD6\n",
      "GDSD6 pro_pro (90, 16)\n",
      "GDSD6 loop_loop (839, 16)\n",
      "***** 222 EGR3::KLF4 GDSD6\n",
      "GDSD6 pro_pro (109, 16)\n",
      "GDSD6 loop_loop (465, 16)\n",
      "***** 223 EGR3::TP63 GDSD6\n",
      "GDSD6 pro_pro (10, 16)\n",
      "GDSD6 loop_loop (169, 16)\n",
      "***** 224 FOS::RORA GDSD6\n",
      "GDSD6 pro_pro (13, 16)\n",
      "GDSD6 loop_loop (147, 16)\n",
      "***** 225 FOSL2::SREBF2 GDSD6\n",
      "GDSD6 pro_pro (15, 16)\n",
      "GDSD6 loop_loop (140, 16)\n",
      "***** 226 GATA3::KLF5 GDSD6\n",
      "GDSD6 pro_pro (3, 16)\n",
      "GDSD6 loop_loop (167, 16)\n",
      "***** 227 HBP1::KLF5 GDSD6\n",
      "GDSD6 pro_pro (32, 16)\n",
      "GDSD6 loop_loop (400, 16)\n",
      "***** 228 HBP1::RORA GDSD6\n",
      "GDSD6 loop_loop (129, 16)\n",
      "***** 229 HMGA1::KLF5 GDSD6\n",
      "GDSD6 pro_pro (42, 16)\n",
      "GDSD6 loop_loop (600, 16)\n",
      "***** 230 HMGA1::RORA GDSD6\n",
      "GDSD6 pro_pro (18, 16)\n",
      "GDSD6 loop_loop (196, 16)\n",
      "***** 231 HOXA1::KLF5 GDSD6\n",
      "GDSD6 loop_loop (69, 16)\n",
      "***** 232 ID4::TP63 GDSD6\n",
      "GDSD6 pro_pro (22, 16)\n",
      "GDSD6 loop_loop (238, 16)\n",
      "***** 233 IRF6::SP2 GDSD6\n",
      "GDSD6 pro_pro (35, 16)\n",
      "GDSD6 loop_loop (143, 16)\n",
      "***** 234 IRF7::KLF5 GDSD6\n",
      "GDSD6 pro_pro (21, 16)\n",
      "GDSD6 loop_loop (133, 16)\n",
      "***** 235 IRF7::RORA GDSD6\n",
      "GDSD6 loop_loop (54, 16)\n",
      "***** 236 IRX3::KLF5 GDSD6\n",
      "GDSD6 pro_pro (9, 16)\n",
      "GDSD6 loop_loop (136, 16)\n",
      "***** 237 IRX3::RORA GDSD6\n",
      "GDSD6 pro_pro (4, 16)\n",
      "GDSD6 loop_loop (51, 16)\n",
      "***** 238 JUN::RORA GDSD6\n",
      "GDSD6 pro_pro (13, 16)\n",
      "GDSD6 loop_loop (138, 16)\n",
      "***** 239 JUN::SREBF2 GDSD6\n",
      "GDSD6 pro_pro (15, 16)\n",
      "GDSD6 loop_loop (130, 16)\n",
      "***** 240 JUNB::RORA GDSD6\n",
      "GDSD6 pro_pro (12, 16)\n",
      "GDSD6 loop_loop (118, 16)\n",
      "***** 241 JUND::RORA GDSD6\n",
      "GDSD6 pro_pro (12, 16)\n",
      "GDSD6 loop_loop (118, 16)\n",
      "***** 242 KLF4::POU2F3 GDSD6\n",
      "GDSD6 loop_loop (71, 16)\n",
      "***** 243 KLF4::SALL4 GDSD6\n",
      "GDSD6 pro_pro (218, 16)\n",
      "GDSD6 loop_loop (584, 16)\n",
      "***** 244 KLF4::SREBF2 GDSD6\n",
      "GDSD6 pro_pro (43, 16)\n",
      "GDSD6 loop_loop (314, 16)\n",
      "***** 245 KLF4::STAT3 GDSD6\n",
      "GDSD6 pro_pro (24, 16)\n",
      "GDSD6 loop_loop (124, 16)\n",
      "***** 246 KLF4::TFE3 GDSD6\n",
      "GDSD6 pro_pro (22, 16)\n",
      "GDSD6 loop_loop (125, 16)\n",
      "***** 247 KLF4::ZNF143 GDSD6\n",
      "GDSD6 pro_pro (2, 16)\n",
      "GDSD6 loop_loop (80, 16)\n",
      "***** 248 KLF5::MAFG GDSD6\n",
      "GDSD6 pro_pro (56, 16)\n",
      "GDSD6 loop_loop (290, 16)\n",
      "***** 249 KLF5::MAFK GDSD6\n",
      "GDSD6 pro_pro (14, 16)\n",
      "GDSD6 loop_loop (260, 16)\n",
      "***** 250 KLF5::MEIS2 GDSD6\n",
      "GDSD6 loop_loop (95, 16)\n",
      "***** 251 KLF5::NFIC GDSD6\n",
      "GDSD6 pro_pro (69, 16)\n",
      "GDSD6 loop_loop (524, 16)\n",
      "***** 252 KLF5::NFIL3 GDSD6\n",
      "GDSD6 pro_pro (4, 16)\n",
      "GDSD6 loop_loop (130, 16)\n",
      "***** 253 KLF5::NR3C1 GDSD6\n",
      "GDSD6 pro_pro (29, 16)\n",
      "GDSD6 loop_loop (284, 16)\n",
      "***** 254 KLF5::PITX1 GDSD6\n",
      "GDSD6 loop_loop (88, 16)\n",
      "***** 255 KLF5::POU2F1 GDSD6\n",
      "GDSD6 pro_pro (2, 16)\n",
      "GDSD6 loop_loop (89, 16)\n",
      "***** 256 KLF5::POU2F3 GDSD6\n",
      "GDSD6 loop_loop (89, 16)\n",
      "***** 257 KLF5::SOX7 GDSD6\n",
      "GDSD6 pro_pro (5, 16)\n",
      "GDSD6 loop_loop (61, 16)\n",
      "***** 258 KLF5::SREBF2 GDSD6\n",
      "GDSD6 pro_pro (54, 16)\n",
      "GDSD6 loop_loop (380, 16)\n",
      "***** 259 KLF5::STAT3 GDSD6\n",
      "GDSD6 pro_pro (26, 16)\n",
      "GDSD6 loop_loop (156, 16)\n",
      "***** 260 KLF5::TCF7L2 GDSD6\n",
      "GDSD6 pro_pro (7, 16)\n",
      "GDSD6 loop_loop (85, 16)\n",
      "***** 261 KLF5::TGIF1 GDSD6\n",
      "GDSD6 pro_pro (4, 16)\n",
      "GDSD6 loop_loop (220, 16)\n",
      "***** 262 KLF5::UBP1 GDSD6\n",
      "GDSD6 pro_pro (372, 16)\n",
      "GDSD6 loop_loop (1972, 16)\n",
      "***** 263 KLF5::ZNF143 GDSD6\n",
      "GDSD6 pro_pro (3, 16)\n",
      "GDSD6 loop_loop (98, 16)\n",
      "***** 264 MAF::RORA GDSD6\n",
      "GDSD6 pro_pro (4, 16)\n",
      "GDSD6 loop_loop (126, 16)\n",
      "***** 265 MAFF::RORA GDSD6\n",
      "GDSD6 pro_pro (17, 16)\n",
      "GDSD6 loop_loop (127, 16)\n",
      "***** 266 MAFG::RORA GDSD6\n",
      "GDSD6 pro_pro (15, 16)\n",
      "GDSD6 loop_loop (111, 16)\n",
      "***** 267 NFATC1::RORA GDSD6\n",
      "GDSD6 loop_loop (94, 16)\n",
      "***** 268 NFE2L1::RORA GDSD6\n",
      "GDSD6 pro_pro (9, 16)\n",
      "GDSD6 loop_loop (257, 16)\n",
      "***** 269 NFIC::RARG GDSD6\n",
      "GDSD6 pro_pro (8, 16)\n",
      "GDSD6 loop_loop (202, 16)\n",
      "***** 270 NFIC::RORA GDSD6\n",
      "GDSD6 pro_pro (11, 16)\n",
      "GDSD6 loop_loop (174, 16)\n",
      "***** 271 NR3C1::RORA GDSD6\n",
      "GDSD6 pro_pro (4, 16)\n",
      "GDSD6 loop_loop (93, 16)\n",
      "***** 272 PRDM1::RORA GDSD6\n",
      "GDSD6 pro_pro (5, 16)\n",
      "GDSD6 loop_loop (87, 16)\n",
      "***** 273 RARG::SREBF2 GDSD6\n",
      "GDSD6 pro_pro (5, 16)\n",
      "GDSD6 loop_loop (142, 16)\n",
      "***** 274 RORA::SMAD3 GDSD6\n",
      "GDSD6 pro_pro (17, 16)\n",
      "GDSD6 loop_loop (140, 16)\n",
      "***** 275 RORA::SREBF2 GDSD6\n",
      "GDSD6 pro_pro (17, 16)\n",
      "GDSD6 loop_loop (108, 16)\n",
      "***** 276 RORA::TP53 GDSD6\n",
      "GDSD6 pro_pro (38, 16)\n",
      "GDSD6 loop_loop (133, 16)\n",
      "***** 277 RORA::TP63 GDSD6\n",
      "GDSD6 pro_pro (11, 16)\n",
      "GDSD6 loop_loop (140, 16)\n",
      "***** 278 RORA::UBP1 GDSD6\n",
      "GDSD6 pro_pro (86, 16)\n",
      "GDSD6 loop_loop (751, 16)\n",
      "***** 279 SALL4::TP63 GDSD6\n",
      "GDSD6 pro_pro (19, 16)\n",
      "GDSD6 loop_loop (215, 16)\n",
      "***** 280 STAT3::UBP1 GDSD6\n",
      "GDSD6 pro_pro (41, 16)\n",
      "GDSD6 loop_loop (311, 16)\n",
      "***** 281 TFE3::UBP1 GDSD6\n",
      "GDSD6 pro_pro (20, 16)\n",
      "GDSD6 loop_loop (353, 16)\n",
      "***** 282 BATF::IKZF1 GM12878\n",
      "GM12878 pro_pro (8, 16)\n",
      "GM12878 loop_loop (105, 16)\n",
      "***** 283 BATF::ZNF589 GM12878\n",
      "GM12878 pro_pro (138, 16)\n",
      "GM12878 loop_loop (586, 16)\n",
      "***** 284 CBFB::IRF8 GM12878\n",
      "GM12878 pro_pro (8, 16)\n",
      "GM12878 loop_loop (142, 16)\n",
      "***** 285 EBF1::SP4 GM12878\n",
      "GM12878 pro_pro (147, 16)\n",
      "GM12878 loop_loop (758, 16)\n",
      "***** 286 EBF1::ZNF816 GM12878\n",
      "GM12878 pro_pro (7, 16)\n",
      "GM12878 loop_loop (155, 16)\n",
      "***** 287 EGR1::IKZF1 GM12878\n",
      "GM12878 pro_pro (47, 16)\n",
      "GM12878 loop_loop (352, 16)\n",
      "***** 288 EGR1::IRF2 GM12878\n",
      "GM12878 loop_loop (155, 16)\n",
      "***** 289 EGR1::IRF4 GM12878\n",
      "GM12878 pro_pro (34, 16)\n",
      "GM12878 loop_loop (242, 16)\n",
      "***** 290 EGR3::IRF8 GM12878\n",
      "GM12878 loop_loop (139, 16)\n",
      "***** 291 ETS1::IKZF1 GM12878\n",
      "GM12878 pro_pro (13, 16)\n",
      "GM12878 loop_loop (244, 16)\n",
      "***** 292 ETS1::IRF4 GM12878\n",
      "GM12878 pro_pro (32, 16)\n",
      "GM12878 loop_loop (198, 16)\n",
      "***** 293 ETV5::IRF8 GM12878\n",
      "GM12878 pro_pro (57, 16)\n",
      "GM12878 loop_loop (313, 16)\n",
      "***** 294 FOXP1::IRF8 GM12878\n",
      "GM12878 pro_pro (13, 16)\n",
      "GM12878 loop_loop (80, 16)\n",
      "***** 295 HBP1::IRF8 GM12878\n",
      "GM12878 pro_pro (25, 16)\n",
      "GM12878 loop_loop (171, 16)\n",
      "***** 296 IKZF1::IRF1 GM12878\n",
      "GM12878 pro_pro (11, 16)\n",
      "GM12878 loop_loop (258, 16)\n",
      "***** 297 IKZF1::IRF2 GM12878\n",
      "GM12878 pro_pro (10, 16)\n",
      "GM12878 loop_loop (197, 16)\n",
      "***** 298 IKZF1::IRF3 GM12878\n",
      "GM12878 pro_pro (19, 16)\n",
      "GM12878 loop_loop (370, 16)\n",
      "***** 299 IKZF1::IRF4 GM12878\n",
      "GM12878 pro_pro (23, 16)\n",
      "GM12878 loop_loop (210, 16)\n",
      "***** 300 IKZF1::IRF7 GM12878\n",
      "GM12878 pro_pro (12, 16)\n",
      "GM12878 loop_loop (139, 16)\n",
      "***** 301 IKZF1::IRF8 GM12878\n",
      "GM12878 pro_pro (24, 16)\n",
      "GM12878 loop_loop (208, 16)\n",
      "***** 302 IKZF1::MEF2A GM12878\n",
      "GM12878 pro_pro (5, 16)\n",
      "GM12878 loop_loop (87, 16)\n",
      "***** 303 IKZF1::MYBL2 GM12878\n",
      "GM12878 pro_pro (33, 16)\n",
      "GM12878 loop_loop (378, 16)\n",
      "***** 304 IKZF1::NFKB1 GM12878\n",
      "GM12878 pro_pro (6, 16)\n",
      "GM12878 loop_loop (113, 16)\n",
      "***** 305 IKZF1::PAX5 GM12878\n",
      "GM12878 pro_pro (13, 16)\n",
      "GM12878 loop_loop (103, 16)\n",
      "***** 306 IKZF1::RUNX3 GM12878\n",
      "GM12878 pro_pro (9, 16)\n",
      "GM12878 loop_loop (93, 16)\n",
      "***** 307 IKZF1::SP4 GM12878\n",
      "GM12878 pro_pro (224, 16)\n",
      "GM12878 loop_loop (1428, 16)\n",
      "***** 308 IKZF1::TP73 GM12878\n",
      "GM12878 pro_pro (9, 16)\n",
      "GM12878 loop_loop (110, 16)\n",
      "***** 309 IKZF1::ZBTB33 GM12878\n",
      "GM12878 pro_pro (4, 16)\n",
      "GM12878 loop_loop (80, 16)\n",
      "***** 310 IKZF1::ZNF563 GM12878\n",
      "GM12878 pro_pro (38, 16)\n",
      "GM12878 loop_loop (221, 16)\n",
      "***** 311 IKZF1::ZNF586 GM12878\n",
      "GM12878 pro_pro (5, 16)\n",
      "GM12878 loop_loop (83, 16)\n",
      "***** 312 IKZF1::ZNF589 GM12878\n",
      "GM12878 pro_pro (120, 16)\n",
      "GM12878 loop_loop (1247, 16)\n",
      "***** 313 IRF2::IRF4 GM12878\n",
      "GM12878 pro_pro (38, 16)\n",
      "GM12878 loop_loop (210, 16)\n",
      "***** 314 IRF2::TP73 GM12878\n",
      "GM12878 pro_pro (5, 16)\n",
      "GM12878 loop_loop (73, 16)\n",
      "***** 315 IRF4::IRF7 GM12878\n",
      "GM12878 pro_pro (31, 16)\n",
      "GM12878 loop_loop (122, 16)\n",
      "***** 316 IRF4::IRF8 GM12878\n",
      "GM12878 pro_pro (83, 16)\n",
      "GM12878 loop_loop (236, 16)\n",
      "***** 317 IRF4::NR3C1 GM12878\n",
      "GM12878 pro_pro (20, 16)\n",
      "GM12878 loop_loop (122, 16)\n",
      "***** 318 IRF4::OSR2 GM12878\n",
      "GM12878 pro_pro (4, 16)\n",
      "GM12878 loop_loop (104, 16)\n",
      "***** 319 IRF4::PLAG1 GM12878\n",
      "GM12878 pro_pro (2, 16)\n",
      "GM12878 loop_loop (225, 16)\n",
      "***** 320 IRF4::SP4 GM12878\n",
      "GM12878 pro_pro (127, 16)\n",
      "GM12878 loop_loop (1070, 16)\n",
      "***** 321 IRF4::ZBTB49 GM12878\n",
      "GM12878 pro_pro (12, 16)\n",
      "GM12878 loop_loop (171, 16)\n",
      "***** 322 IRF4::ZNF816 GM12878\n",
      "GM12878 pro_pro (17, 16)\n",
      "GM12878 loop_loop (239, 16)\n",
      "***** 323 IRF8::LEF1 GM12878\n",
      "GM12878 pro_pro (29, 16)\n",
      "GM12878 loop_loop (77, 16)\n",
      "***** 324 IRF8::MEF2A GM12878\n",
      "GM12878 pro_pro (11, 16)\n",
      "GM12878 loop_loop (70, 16)\n",
      "***** 325 IRF8::MYBL2 GM12878\n",
      "GM12878 pro_pro (45, 16)\n",
      "GM12878 loop_loop (268, 16)\n",
      "***** 326 IRF8::PRDM1 GM12878\n",
      "GM12878 pro_pro (90, 16)\n",
      "GM12878 loop_loop (213, 16)\n",
      "***** 327 IRF8::SPIB GM12878\n",
      "GM12878 pro_pro (29, 16)\n",
      "GM12878 loop_loop (180, 16)\n",
      "***** 328 IRF8::STAT1 GM12878\n",
      "GM12878 pro_pro (120, 16)\n",
      "GM12878 loop_loop (384, 16)\n",
      "***** 329 IRF8::STAT2 GM12878\n",
      "GM12878 pro_pro (93, 16)\n",
      "GM12878 loop_loop (330, 16)\n",
      "***** 330 IRF8::UBP1 GM12878\n",
      "GM12878 pro_pro (276, 16)\n",
      "GM12878 loop_loop (1085, 16)\n",
      "***** 331 IRF8::ZBTB49 GM12878\n",
      "GM12878 pro_pro (16, 16)\n",
      "GM12878 loop_loop (157, 16)\n",
      "***** 332 IRF8::ZNF394 GM12878\n",
      "GM12878 pro_pro (29, 16)\n",
      "GM12878 loop_loop (173, 16)\n",
      "***** 333 IRF8::ZNF589 GM12878\n",
      "GM12878 pro_pro (137, 16)\n",
      "GM12878 loop_loop (985, 16)\n",
      "***** 334 IRF8::ZNF816 GM12878\n",
      "GM12878 pro_pro (12, 16)\n",
      "GM12878 loop_loop (222, 16)\n",
      "***** 335 PAX5::SP4 GM12878\n",
      "GM12878 pro_pro (184, 16)\n",
      "GM12878 loop_loop (783, 16)\n",
      "***** 336 PAX5::ZNF589 GM12878\n",
      "GM12878 pro_pro (20, 16)\n",
      "GM12878 loop_loop (432, 16)\n",
      "***** 337 SP4::TCF3 GM12878\n",
      "GM12878 pro_pro (73, 16)\n",
      "GM12878 loop_loop (678, 16)\n",
      "***** 338 TCF3::ZNF589 GM12878\n",
      "GM12878 pro_pro (35, 16)\n",
      "GM12878 loop_loop (553, 16)\n",
      "***** 339 ARNT::SP1 SCC\n",
      "A431-CTRLi pro_pro (59, 16)\n",
      "A431-CTRLi loop_loop (1745, 16)\n",
      "CAL27-CTRLi pro_pro (62, 16)\n",
      "CAL27-CTRLi loop_loop (2011, 16)\n",
      "SCC13-CTRLi pro_pro (76, 16)\n",
      "SCC13-CTRLi loop_loop (1478, 16)\n",
      "***** 340 E2F1::SP1 SCC\n",
      "A431-CTRLi pro_pro (415, 16)\n",
      "A431-CTRLi loop_loop (2784, 16)\n",
      "CAL27-CTRLi pro_pro (343, 16)\n",
      "CAL27-CTRLi loop_loop (2600, 16)\n",
      "SCC13-CTRLi pro_pro (316, 16)\n",
      "SCC13-CTRLi loop_loop (2268, 16)\n",
      "***** 341 E2F6::SP1 SCC\n",
      "A431-CTRLi pro_pro (407, 16)\n",
      "A431-CTRLi loop_loop (3293, 16)\n",
      "CAL27-CTRLi pro_pro (315, 16)\n",
      "CAL27-CTRLi loop_loop (3156, 16)\n",
      "SCC13-CTRLi pro_pro (336, 16)\n",
      "SCC13-CTRLi loop_loop (2766, 16)\n",
      "***** 342 EGR1::SP1 SCC\n",
      "A431-CTRLi pro_pro (955, 16)\n",
      "A431-CTRLi loop_loop (3524, 16)\n",
      "CAL27-CTRLi pro_pro (658, 16)\n",
      "CAL27-CTRLi loop_loop (3412, 16)\n",
      "SCC13-CTRLi pro_pro (634, 16)\n",
      "SCC13-CTRLi loop_loop (2981, 16)\n",
      "***** 343 EGR1::SP3 SCC\n",
      "A431-CTRLi pro_pro (354, 16)\n",
      "A431-CTRLi loop_loop (1337, 16)\n",
      "CAL27-CTRLi pro_pro (236, 16)\n",
      "CAL27-CTRLi loop_loop (1331, 16)\n",
      "SCC13-CTRLi pro_pro (270, 16)\n",
      "SCC13-CTRLi loop_loop (1182, 16)\n",
      "***** 344 ETS1::SP1 SCC\n",
      "A431-CTRLi pro_pro (357, 16)\n",
      "A431-CTRLi loop_loop (3248, 16)\n",
      "CAL27-CTRLi pro_pro (396, 16)\n",
      "CAL27-CTRLi loop_loop (2900, 16)\n",
      "SCC13-CTRLi pro_pro (256, 16)\n",
      "SCC13-CTRLi loop_loop (2606, 16)\n",
      "***** 345 KLF16::SP1 SCC\n",
      "A431-CTRLi pro_pro (1778, 16)\n",
      "A431-CTRLi loop_loop (4760, 16)\n",
      "CAL27-CTRLi pro_pro (1346, 16)\n",
      "CAL27-CTRLi loop_loop (4676, 16)\n",
      "SCC13-CTRLi pro_pro (1418, 16)\n",
      "SCC13-CTRLi loop_loop (4146, 16)\n",
      "***** 346 KLF16::SP3 SCC\n",
      "A431-CTRLi pro_pro (884, 16)\n",
      "A431-CTRLi loop_loop (2493, 16)\n",
      "CAL27-CTRLi pro_pro (642, 16)\n",
      "CAL27-CTRLi loop_loop (2467, 16)\n",
      "SCC13-CTRLi pro_pro (700, 16)\n",
      "SCC13-CTRLi loop_loop (2243, 16)\n",
      "***** 347 KLF16::TFAP2C SCC\n",
      "A431-CTRLi pro_pro (144, 16)\n",
      "A431-CTRLi loop_loop (1590, 16)\n",
      "CAL27-CTRLi pro_pro (108, 16)\n",
      "CAL27-CTRLi loop_loop (1518, 16)\n",
      "SCC13-CTRLi pro_pro (105, 16)\n",
      "SCC13-CTRLi loop_loop (1285, 16)\n",
      "***** 348 KLF4::KLF5 SCC\n",
      "A431-CTRLi pro_pro (412, 16)\n",
      "A431-CTRLi loop_loop (1375, 16)\n",
      "CAL27-CTRLi pro_pro (398, 16)\n",
      "CAL27-CTRLi loop_loop (1313, 16)\n",
      "SCC13-CTRLi pro_pro (393, 16)\n",
      "SCC13-CTRLi loop_loop (1241, 16)\n",
      "***** 349 KLF4::SP1 SCC\n",
      "A431-CTRLi pro_pro (1372, 16)\n",
      "A431-CTRLi loop_loop (3893, 16)\n",
      "CAL27-CTRLi pro_pro (1073, 16)\n",
      "CAL27-CTRLi loop_loop (3716, 16)\n",
      "SCC13-CTRLi pro_pro (1234, 16)\n",
      "SCC13-CTRLi loop_loop (3294, 16)\n",
      "***** 350 KLF4::SP3 SCC\n",
      "A431-CTRLi pro_pro (568, 16)\n",
      "A431-CTRLi loop_loop (1624, 16)\n",
      "CAL27-CTRLi pro_pro (466, 16)\n",
      "CAL27-CTRLi loop_loop (1554, 16)\n",
      "SCC13-CTRLi pro_pro (561, 16)\n",
      "SCC13-CTRLi loop_loop (1430, 16)\n",
      "***** 351 KLF4::TFAP2C SCC\n",
      "A431-CTRLi pro_pro (80, 16)\n",
      "A431-CTRLi loop_loop (835, 16)\n",
      "CAL27-CTRLi pro_pro (46, 16)\n",
      "CAL27-CTRLi loop_loop (786, 16)\n",
      "SCC13-CTRLi pro_pro (99, 16)\n",
      "SCC13-CTRLi loop_loop (693, 16)\n",
      "***** 352 KLF5::SP3 SCC\n",
      "A431-CTRLi pro_pro (542, 16)\n",
      "A431-CTRLi loop_loop (1753, 16)\n",
      "CAL27-CTRLi pro_pro (459, 16)\n",
      "CAL27-CTRLi loop_loop (1703, 16)\n",
      "SCC13-CTRLi pro_pro (454, 16)\n",
      "SCC13-CTRLi loop_loop (1614, 16)\n",
      "***** 353 KLF5::TFAP2C SCC\n",
      "A431-CTRLi pro_pro (88, 16)\n",
      "A431-CTRLi loop_loop (949, 16)\n",
      "CAL27-CTRLi pro_pro (64, 16)\n",
      "CAL27-CTRLi loop_loop (915, 16)\n",
      "SCC13-CTRLi pro_pro (116, 16)\n",
      "SCC13-CTRLi loop_loop (830, 16)\n",
      "***** 354 KLF5::ZNF219 SCC\n",
      "A431-CTRLi pro_pro (247, 16)\n",
      "A431-CTRLi loop_loop (1255, 16)\n",
      "CAL27-CTRLi pro_pro (235, 16)\n",
      "CAL27-CTRLi loop_loop (1169, 16)\n",
      "SCC13-CTRLi pro_pro (231, 16)\n",
      "SCC13-CTRLi loop_loop (1083, 16)\n",
      "***** 355 KLF6::SP1 SCC\n",
      "A431-CTRLi pro_pro (1589, 16)\n",
      "A431-CTRLi loop_loop (4444, 16)\n",
      "CAL27-CTRLi pro_pro (1250, 16)\n",
      "CAL27-CTRLi loop_loop (4271, 16)\n",
      "SCC13-CTRLi pro_pro (1484, 16)\n",
      "SCC13-CTRLi loop_loop (3884, 16)\n",
      "***** 356 MAZ::SP1 SCC\n",
      "A431-CTRLi pro_pro (2876, 16)\n",
      "A431-CTRLi loop_loop (6969, 16)\n",
      "CAL27-CTRLi pro_pro (2448, 16)\n",
      "CAL27-CTRLi loop_loop (6942, 16)\n",
      "SCC13-CTRLi pro_pro (2695, 16)\n",
      "SCC13-CTRLi loop_loop (6292, 16)\n",
      "***** 357 NFAT5::SP1 SCC\n",
      "A431-CTRLi pro_pro (286, 16)\n",
      "A431-CTRLi loop_loop (4109, 16)\n",
      "CAL27-CTRLi pro_pro (450, 16)\n",
      "CAL27-CTRLi loop_loop (4061, 16)\n",
      "SCC13-CTRLi pro_pro (474, 16)\n",
      "SCC13-CTRLi loop_loop (3746, 16)\n",
      "***** 358 NR2F1::SP1 SCC\n",
      "A431-CTRLi pro_pro (301, 16)\n",
      "A431-CTRLi loop_loop (3295, 16)\n",
      "CAL27-CTRLi pro_pro (358, 16)\n",
      "CAL27-CTRLi loop_loop (3187, 16)\n",
      "SCC13-CTRLi pro_pro (478, 16)\n",
      "SCC13-CTRLi loop_loop (2678, 16)\n",
      "***** 359 NRF1::SP1 SCC\n",
      "A431-CTRLi pro_pro (478, 16)\n",
      "A431-CTRLi loop_loop (2541, 16)\n",
      "CAL27-CTRLi pro_pro (130, 16)\n",
      "CAL27-CTRLi loop_loop (2428, 16)\n",
      "SCC13-CTRLi pro_pro (213, 16)\n",
      "SCC13-CTRLi loop_loop (1897, 16)\n",
      "***** 360 PURA::SP1 SCC\n",
      "A431-CTRLi pro_pro (783, 16)\n",
      "A431-CTRLi loop_loop (3841, 16)\n",
      "CAL27-CTRLi pro_pro (781, 16)\n",
      "CAL27-CTRLi loop_loop (3810, 16)\n",
      "SCC13-CTRLi pro_pro (847, 16)\n",
      "SCC13-CTRLi loop_loop (3219, 16)\n",
      "***** 361 SNAI2::SP1 SCC\n",
      "A431-CTRLi pro_pro (168, 16)\n",
      "A431-CTRLi loop_loop (2346, 16)\n",
      "CAL27-CTRLi pro_pro (124, 16)\n",
      "CAL27-CTRLi loop_loop (2325, 16)\n",
      "SCC13-CTRLi pro_pro (180, 16)\n",
      "SCC13-CTRLi loop_loop (1808, 16)\n",
      "***** 362 SP1::SP2 SCC\n",
      "A431-CTRLi pro_pro (3124, 16)\n",
      "A431-CTRLi loop_loop (7076, 16)\n",
      "CAL27-CTRLi pro_pro (2797, 16)\n",
      "CAL27-CTRLi loop_loop (6792, 16)\n",
      "SCC13-CTRLi pro_pro (2980, 16)\n",
      "SCC13-CTRLi loop_loop (6198, 16)\n",
      "***** 363 SP1::SP3 SCC\n",
      "A431-CTRLi pro_pro (1868, 16)\n",
      "A431-CTRLi loop_loop (4297, 16)\n",
      "CAL27-CTRLi pro_pro (1527, 16)\n",
      "CAL27-CTRLi loop_loop (4129, 16)\n",
      "SCC13-CTRLi pro_pro (1813, 16)\n",
      "SCC13-CTRLi loop_loop (3834, 16)\n",
      "***** 364 SP1::SP4 SCC\n",
      "A431-CTRLi pro_pro (2181, 16)\n",
      "A431-CTRLi loop_loop (5598, 16)\n",
      "CAL27-CTRLi pro_pro (1886, 16)\n",
      "CAL27-CTRLi loop_loop (5470, 16)\n",
      "SCC13-CTRLi pro_pro (2217, 16)\n",
      "SCC13-CTRLi loop_loop (4937, 16)\n",
      "***** 365 SP1::TBX1 SCC\n",
      "A431-CTRLi pro_pro (1536, 16)\n",
      "A431-CTRLi loop_loop (4397, 16)\n",
      "CAL27-CTRLi pro_pro (1139, 16)\n",
      "CAL27-CTRLi loop_loop (4290, 16)\n",
      "SCC13-CTRLi pro_pro (981, 16)\n",
      "SCC13-CTRLi loop_loop (3773, 16)\n",
      "***** 366 SP1::TCF3 SCC\n",
      "A431-CTRLi pro_pro (57, 16)\n",
      "A431-CTRLi loop_loop (2139, 16)\n",
      "CAL27-CTRLi pro_pro (74, 16)\n",
      "CAL27-CTRLi loop_loop (2124, 16)\n",
      "SCC13-CTRLi pro_pro (127, 16)\n",
      "SCC13-CTRLi loop_loop (1654, 16)\n",
      "***** 367 SP1::TFAP2C SCC\n",
      "A431-CTRLi pro_pro (304, 16)\n",
      "A431-CTRLi loop_loop (3139, 16)\n",
      "CAL27-CTRLi pro_pro (340, 16)\n",
      "CAL27-CTRLi loop_loop (2919, 16)\n",
      "SCC13-CTRLi pro_pro (365, 16)\n",
      "SCC13-CTRLi loop_loop (2551, 16)\n",
      "***** 368 SP1::TFDP1 SCC\n",
      "A431-CTRLi pro_pro (436, 16)\n",
      "A431-CTRLi loop_loop (2887, 16)\n",
      "CAL27-CTRLi pro_pro (333, 16)\n",
      "CAL27-CTRLi loop_loop (2754, 16)\n",
      "SCC13-CTRLi pro_pro (281, 16)\n",
      "SCC13-CTRLi loop_loop (2224, 16)\n",
      "***** 369 SP1::VEZF1 SCC\n",
      "A431-CTRLi pro_pro (2000, 16)\n",
      "A431-CTRLi loop_loop (6132, 16)\n",
      "CAL27-CTRLi pro_pro (1810, 16)\n",
      "CAL27-CTRLi loop_loop (6062, 16)\n",
      "SCC13-CTRLi pro_pro (2031, 16)\n",
      "SCC13-CTRLi loop_loop (5417, 16)\n",
      "***** 370 SP1::ZBTB14 SCC\n",
      "A431-CTRLi pro_pro (376, 16)\n",
      "A431-CTRLi loop_loop (2609, 16)\n",
      "CAL27-CTRLi pro_pro (416, 16)\n",
      "CAL27-CTRLi loop_loop (2464, 16)\n",
      "SCC13-CTRLi pro_pro (420, 16)\n",
      "SCC13-CTRLi loop_loop (1979, 16)\n",
      "***** 371 SP1::ZBTB17 SCC\n",
      "A431-CTRLi pro_pro (1417, 16)\n",
      "A431-CTRLi loop_loop (4330, 16)\n",
      "CAL27-CTRLi pro_pro (1152, 16)\n",
      "CAL27-CTRLi loop_loop (4216, 16)\n",
      "SCC13-CTRLi pro_pro (1154, 16)\n",
      "SCC13-CTRLi loop_loop (3850, 16)\n",
      "***** 372 SP1::ZNF219 SCC\n",
      "A431-CTRLi pro_pro (1150, 16)\n",
      "A431-CTRLi loop_loop (3755, 16)\n",
      "CAL27-CTRLi pro_pro (1093, 16)\n",
      "CAL27-CTRLi loop_loop (3524, 16)\n",
      "SCC13-CTRLi pro_pro (1024, 16)\n",
      "SCC13-CTRLi loop_loop (3198, 16)\n",
      "***** 373 SP1::ZNF770 SCC\n",
      "A431-CTRLi pro_pro (1215, 16)\n",
      "A431-CTRLi loop_loop (4590, 16)\n",
      "CAL27-CTRLi pro_pro (947, 16)\n",
      "CAL27-CTRLi loop_loop (4674, 16)\n",
      "SCC13-CTRLi pro_pro (816, 16)\n",
      "SCC13-CTRLi loop_loop (4209, 16)\n",
      "***** 374 SP3::TFAP2C SCC\n",
      "A431-CTRLi pro_pro (117, 16)\n",
      "A431-CTRLi loop_loop (1156, 16)\n",
      "CAL27-CTRLi pro_pro (87, 16)\n",
      "CAL27-CTRLi loop_loop (1106, 16)\n",
      "SCC13-CTRLi pro_pro (153, 16)\n",
      "SCC13-CTRLi loop_loop (1012, 16)\n",
      "***** 375 SP3::ZBTB17 SCC\n",
      "A431-CTRLi pro_pro (633, 16)\n",
      "A431-CTRLi loop_loop (2059, 16)\n",
      "CAL27-CTRLi pro_pro (532, 16)\n",
      "CAL27-CTRLi loop_loop (2058, 16)\n",
      "SCC13-CTRLi pro_pro (596, 16)\n",
      "SCC13-CTRLi loop_loop (1908, 16)\n",
      "***** 376 SP3::ZNF219 SCC\n",
      "A431-CTRLi pro_pro (479, 16)\n",
      "A431-CTRLi loop_loop (1501, 16)\n",
      "CAL27-CTRLi pro_pro (401, 16)\n",
      "CAL27-CTRLi loop_loop (1427, 16)\n",
      "SCC13-CTRLi pro_pro (416, 16)\n",
      "SCC13-CTRLi loop_loop (1304, 16)\n",
      "***** 377 TFAP2C::ZBTB14 SCC\n",
      "A431-CTRLi pro_pro (50, 16)\n",
      "A431-CTRLi loop_loop (338, 16)\n",
      "CAL27-CTRLi pro_pro (37, 16)\n",
      "CAL27-CTRLi loop_loop (338, 16)\n",
      "SCC13-CTRLi pro_pro (44, 16)\n",
      "SCC13-CTRLi loop_loop (274, 16)\n",
      "***** 378 TFAP2C::ZNF219 SCC\n",
      "A431-CTRLi pro_pro (85, 16)\n",
      "A431-CTRLi loop_loop (742, 16)\n",
      "CAL27-CTRLi pro_pro (65, 16)\n",
      "CAL27-CTRLi loop_loop (688, 16)\n",
      "SCC13-CTRLi pro_pro (69, 16)\n",
      "SCC13-CTRLi loop_loop (609, 16)\n",
      "***** 379 NFE2L1::PRRX1 MM\n",
      "WM-SCR-DMSO pro_pro (292, 16)\n",
      "WM-SCR-DMSO loop_loop (1584, 16)\n",
      "***** 380 NFE2L1::UBP1 MM\n",
      "COLO-SCR-DMSO pro_pro (136, 16)\n",
      "COLO-SCR-DMSO loop_loop (512, 16)\n",
      "WM-SCR-DMSO pro_pro (105, 16)\n",
      "WM-SCR-DMSO loop_loop (663, 16)\n",
      "***** 381 NFE2L1::SOX13 MM\n",
      "COLO-SCR-DMSO pro_pro (68, 16)\n",
      "COLO-SCR-DMSO loop_loop (455, 16)\n",
      "WM-SCR-DMSO pro_pro (59, 16)\n",
      "WM-SCR-DMSO loop_loop (567, 16)\n",
      "***** 382 STAT1::UBP1 MM\n",
      "COLO-SCR-DMSO pro_pro (120, 16)\n",
      "COLO-SCR-DMSO loop_loop (396, 16)\n",
      "WM-SCR-DMSO pro_pro (94, 16)\n",
      "WM-SCR-DMSO loop_loop (538, 16)\n",
      "***** 383 ARID5B::NFE2L1 MM\n",
      "COLO-SCR-DMSO pro_pro (84, 16)\n",
      "COLO-SCR-DMSO loop_loop (366, 16)\n",
      "WM-SCR-DMSO pro_pro (64, 16)\n",
      "WM-SCR-DMSO loop_loop (487, 16)\n",
      "***** 384 PRRX1::STAT1 MM\n",
      "WM-SCR-DMSO pro_pro (144, 16)\n",
      "WM-SCR-DMSO loop_loop (1170, 16)\n",
      "***** 385 SOX13::STAT1 MM\n",
      "COLO-SCR-DMSO pro_pro (74, 16)\n",
      "COLO-SCR-DMSO loop_loop (355, 16)\n",
      "WM-SCR-DMSO pro_pro (73, 16)\n",
      "WM-SCR-DMSO loop_loop (491, 16)\n",
      "***** 386 NFAT5::NFE2L1 MM\n",
      "COLO-SCR-DMSO pro_pro (23, 16)\n",
      "COLO-SCR-DMSO loop_loop (256, 16)\n",
      "WM-SCR-DMSO pro_pro (44, 16)\n",
      "WM-SCR-DMSO loop_loop (383, 16)\n",
      "***** 387 NFATC4::NFE2L1 MM\n",
      "COLO-SCR-DMSO pro_pro (23, 16)\n",
      "COLO-SCR-DMSO loop_loop (256, 16)\n",
      "WM-SCR-DMSO pro_pro (44, 16)\n",
      "WM-SCR-DMSO loop_loop (383, 16)\n",
      "***** 388 ARID5B::STAT1 MM\n",
      "COLO-SCR-DMSO pro_pro (88, 16)\n",
      "COLO-SCR-DMSO loop_loop (305, 16)\n",
      "WM-SCR-DMSO pro_pro (12, 16)\n",
      "WM-SCR-DMSO loop_loop (398, 16)\n",
      "***** 389 ARID5B::SOX10 MM\n",
      "COLO-SCR-DMSO pro_pro (34, 16)\n",
      "COLO-SCR-DMSO loop_loop (240, 16)\n",
      "WM-SCR-DMSO pro_pro (47, 16)\n",
      "WM-SCR-DMSO loop_loop (381, 16)\n",
      "***** 390 NFAT5::STAT1 MM\n",
      "COLO-SCR-DMSO pro_pro (54, 16)\n",
      "COLO-SCR-DMSO loop_loop (244, 16)\n",
      "WM-SCR-DMSO pro_pro (21, 16)\n",
      "WM-SCR-DMSO loop_loop (340, 16)\n",
      "***** 391 NFATC4::STAT1 MM\n",
      "COLO-SCR-DMSO pro_pro (54, 16)\n",
      "COLO-SCR-DMSO loop_loop (244, 16)\n",
      "WM-SCR-DMSO pro_pro (21, 16)\n",
      "WM-SCR-DMSO loop_loop (340, 16)\n",
      "***** 392 SOX10::UBP1 MM\n",
      "COLO-SCR-DMSO pro_pro (58, 16)\n",
      "COLO-SCR-DMSO loop_loop (347, 16)\n",
      "WM-SCR-DMSO pro_pro (74, 16)\n",
      "WM-SCR-DMSO loop_loop (524, 16)\n",
      "***** 393 HMGA1::NFE2L1 MM\n",
      "COLO-SCR-DMSO pro_pro (39, 16)\n",
      "COLO-SCR-DMSO loop_loop (172, 16)\n",
      "WM-SCR-DMSO pro_pro (19, 16)\n",
      "WM-SCR-DMSO loop_loop (228, 16)\n",
      "***** 394 SOX10::SOX13 MM\n",
      "COLO-SCR-DMSO pro_pro (38, 16)\n",
      "COLO-SCR-DMSO loop_loop (272, 16)\n",
      "WM-SCR-DMSO pro_pro (57, 16)\n",
      "WM-SCR-DMSO loop_loop (430, 16)\n",
      "***** 395 EGR2::NFE2L1 MM\n",
      "WM-SCR-DMSO pro_pro (23, 16)\n",
      "WM-SCR-DMSO loop_loop (447, 16)\n",
      "***** 396 LEF1::PRRX1 MM\n",
      "WM-SCR-DMSO pro_pro (67, 16)\n",
      "WM-SCR-DMSO loop_loop (922, 16)\n",
      "***** 397 LEF1::UBP1 MM\n",
      "COLO-SCR-DMSO pro_pro (47, 16)\n",
      "COLO-SCR-DMSO loop_loop (256, 16)\n",
      "WM-SCR-DMSO pro_pro (28, 16)\n",
      "WM-SCR-DMSO loop_loop (363, 16)\n",
      "***** 398 NFE2L1::TBX15 MM\n",
      "WM-SCR-DMSO pro_pro (40, 16)\n",
      "WM-SCR-DMSO loop_loop (418, 16)\n",
      "***** 399 HMGA1::STAT1 MM\n",
      "COLO-SCR-DMSO pro_pro (58, 16)\n",
      "COLO-SCR-DMSO loop_loop (170, 16)\n",
      "WM-SCR-DMSO pro_pro (8, 16)\n",
      "WM-SCR-DMSO loop_loop (206, 16)\n",
      "***** 400 NFE2L1::SP4 MM\n",
      "COLO-SCR-DMSO pro_pro (5, 16)\n",
      "COLO-SCR-DMSO loop_loop (403, 16)\n",
      "WM-SCR-DMSO pro_pro (18, 16)\n",
      "WM-SCR-DMSO loop_loop (564, 16)\n",
      "***** 401 SP4::STAT1 MM\n",
      "COLO-SCR-DMSO pro_pro (32, 16)\n",
      "COLO-SCR-DMSO loop_loop (295, 16)\n",
      "WM-SCR-DMSO pro_pro (72, 16)\n",
      "WM-SCR-DMSO loop_loop (599, 16)\n",
      "***** 402 HMGA1::SOX10 MM\n",
      "COLO-SCR-DMSO pro_pro (11, 16)\n",
      "COLO-SCR-DMSO loop_loop (106, 16)\n",
      "WM-SCR-DMSO pro_pro (20, 16)\n",
      "WM-SCR-DMSO loop_loop (178, 16)\n",
      "***** 403 NFAT5::SOX10 MM\n",
      "COLO-SCR-DMSO pro_pro (13, 16)\n",
      "COLO-SCR-DMSO loop_loop (151, 16)\n",
      "WM-SCR-DMSO pro_pro (39, 16)\n",
      "WM-SCR-DMSO loop_loop (280, 16)\n",
      "***** 404 NFATC4::SOX10 MM\n",
      "COLO-SCR-DMSO pro_pro (13, 16)\n",
      "COLO-SCR-DMSO loop_loop (151, 16)\n",
      "WM-SCR-DMSO pro_pro (39, 16)\n",
      "WM-SCR-DMSO loop_loop (280, 16)\n",
      "***** 405 ARID5B::LEF1 MM\n",
      "COLO-SCR-DMSO pro_pro (38, 16)\n",
      "COLO-SCR-DMSO loop_loop (205, 16)\n",
      "WM-SCR-DMSO pro_pro (29, 16)\n",
      "WM-SCR-DMSO loop_loop (282, 16)\n",
      "***** 406 LEF1::SOX13 MM\n",
      "COLO-SCR-DMSO pro_pro (37, 16)\n",
      "COLO-SCR-DMSO loop_loop (219, 16)\n",
      "WM-SCR-DMSO pro_pro (26, 16)\n",
      "WM-SCR-DMSO loop_loop (312, 16)\n",
      "***** 407 EGR4::NFE2L1 MM\n",
      "WM-SCR-DMSO pro_pro (6, 16)\n",
      "WM-SCR-DMSO loop_loop (200, 16)\n",
      "***** 408 EN2::STAT1 MM\n",
      "COLO-SCR-DMSO pro_pro (16, 16)\n",
      "COLO-SCR-DMSO loop_loop (113, 16)\n",
      "WM-SCR-DMSO pro_pro (4, 16)\n",
      "WM-SCR-DMSO loop_loop (148, 16)\n",
      "***** 409 STAT1::TBX15 MM\n",
      "WM-SCR-DMSO pro_pro (50, 16)\n",
      "WM-SCR-DMSO loop_loop (444, 16)\n",
      "***** 410 HBP1::STAT1 MM\n",
      "COLO-SCR-DMSO pro_pro (9, 16)\n",
      "COLO-SCR-DMSO loop_loop (79, 16)\n",
      "WM-SCR-DMSO pro_pro (9, 16)\n",
      "WM-SCR-DMSO loop_loop (119, 16)\n",
      "***** 411 EGR2::STAT1 MM\n",
      "WM-SCR-DMSO pro_pro (34, 16)\n",
      "WM-SCR-DMSO loop_loop (438, 16)\n",
      "***** 412 EGR2::LEF1 MM\n",
      "WM-SCR-DMSO loop_loop (241, 16)\n",
      "***** 413 EGR2::SOX10 MM\n",
      "WM-SCR-DMSO pro_pro (35, 16)\n",
      "WM-SCR-DMSO loop_loop (338, 16)\n",
      "***** 414 NFE2L1::NR3C1 MM\n",
      "COLO-SCR-DMSO pro_pro (10, 16)\n",
      "COLO-SCR-DMSO loop_loop (67, 16)\n",
      "WM-SCR-DMSO loop_loop (95, 16)\n",
      "***** 415 EN2::NFE2L1 MM\n",
      "COLO-SCR-DMSO pro_pro (8, 16)\n",
      "COLO-SCR-DMSO loop_loop (99, 16)\n",
      "WM-SCR-DMSO pro_pro (10, 16)\n",
      "WM-SCR-DMSO loop_loop (115, 16)\n",
      "***** 416 LEF1::NFAT5 MM\n",
      "COLO-SCR-DMSO pro_pro (15, 16)\n",
      "COLO-SCR-DMSO loop_loop (122, 16)\n",
      "WM-SCR-DMSO pro_pro (13, 16)\n",
      "WM-SCR-DMSO loop_loop (206, 16)\n",
      "***** 417 NFE2L1::ZBTB49 MM\n",
      "COLO-SCR-DMSO pro_pro (8, 16)\n",
      "COLO-SCR-DMSO loop_loop (99, 16)\n",
      "WM-SCR-DMSO pro_pro (10, 16)\n",
      "WM-SCR-DMSO loop_loop (115, 16)\n",
      "***** 418 SOX10::SP4 MM\n",
      "COLO-SCR-DMSO loop_loop (196, 16)\n",
      "WM-SCR-DMSO pro_pro (47, 16)\n",
      "WM-SCR-DMSO loop_loop (474, 16)\n",
      "***** 419 HMGA1::LEF1 MM\n",
      "COLO-SCR-DMSO pro_pro (19, 16)\n",
      "COLO-SCR-DMSO loop_loop (95, 16)\n",
      "WM-SCR-DMSO pro_pro (16, 16)\n",
      "WM-SCR-DMSO loop_loop (127, 16)\n",
      "***** 420 CEBPG::NR3C1 Colon\n",
      "Colon pro_pro (9, 16)\n",
      "Colon loop_loop (55, 16)\n",
      "***** 421 MEIS2::MGA Colon\n",
      "Colon pro_pro (3, 16)\n",
      "Colon loop_loop (7, 16)\n",
      "***** 422 NFE2L1::NFIA Colon\n",
      "Colon pro_pro (143, 16)\n",
      "Colon loop_loop (1146, 16)\n",
      "***** 424 E2F6::PPARD Colon\n",
      "Colon pro_pro (2, 16)\n",
      "Colon loop_loop (118, 16)\n",
      "***** 425 ATF1::NFE2L1 Colon\n",
      "Colon pro_pro (6, 16)\n",
      "Colon loop_loop (64, 16)\n",
      "***** 426 POU2F1::PPARD Colon\n",
      "Colon loop_loop (10, 16)\n",
      "***** 427 BCL6::NFYC Colon\n",
      "Colon loop_loop (13, 16)\n",
      "***** 429 ZNF354A::MEIS2 Colon\n",
      "Colon pro_pro (2, 16)\n",
      "Colon loop_loop (15, 16)\n",
      "***** 430 REL::SOX4 GDSD6\n",
      "GDSD6 loop_loop (10, 16)\n",
      "***** 431 OSR2::RELB GDSD6\n",
      "GDSD6 loop_loop (8, 16)\n",
      "***** 432 OVOL2::RUNX1 GDSD6\n",
      "GDSD6 loop_loop (1, 16)\n",
      "***** 433 REL::SP1 GDSD6\n",
      "GDSD6 pro_pro (20, 16)\n",
      "GDSD6 loop_loop (303, 16)\n",
      "***** 435 RUNX1::VEZF1 GDSD6\n",
      "GDSD6 pro_pro (60, 16)\n",
      "GDSD6 loop_loop (238, 16)\n",
      "***** 436 SOX4::ZBTB7A GDSD6\n",
      "GDSD6 loop_loop (1, 16)\n",
      "***** 437 SP1::XBP1 GDSD6\n",
      "GDSD6 pro_pro (169, 16)\n",
      "GDSD6 loop_loop (726, 16)\n",
      "***** 438 NFKB1::NR1I3 Melanocytes\n",
      "Melanocytes loop_loop (2, 16)\n",
      "***** 439 NFKB1::RFX2 Melanocytes\n",
      "Melanocytes pro_pro (6, 16)\n",
      "Melanocytes loop_loop (11, 16)\n",
      "***** 441 IRF2::KLF5 Melanocytes\n",
      "Melanocytes pro_pro (13, 16)\n",
      "Melanocytes loop_loop (31, 16)\n",
      "***** 442 NFE2L2::NFKB1 Melanocytes\n",
      "Melanocytes loop_loop (0, 16)\n",
      "***** 443 IRF6::NFYB Melanocytes\n",
      "***** 444 NFE2L2::RFX2 Melanocytes\n",
      "Melanocytes loop_loop (0, 16)\n",
      "***** 445 IRF6::KLF5 Melanocytes\n",
      "***** 446 IRF2::IRF6 Melanocytes\n",
      "***** 447 IRF2::NFYB Melanocytes\n",
      "Melanocytes pro_pro (5, 16)\n",
      "Melanocytes loop_loop (5, 16)\n",
      "***** 448 KLF5::NFYB Melanocytes\n",
      "Melanocytes pro_pro (58, 16)\n",
      "Melanocytes loop_loop (104, 16)\n",
      "***** 449 IRF7::E2F4 GM12878\n",
      "GM12878 loop_loop (90, 16)\n",
      "***** 450 MAZ::FOSL1 GM12878\n",
      "***** 452 KLF6::FOSL1 GM12878\n",
      "***** 453 MEF2D::FOXO3 GM12878\n",
      "GM12878 loop_loop (10, 16)\n",
      "***** 454 IRF7::SMARCA5 GM12878\n",
      "GM12878 pro_pro (4, 16)\n",
      "GM12878 loop_loop (23, 16)\n",
      "***** 455 SMARCA5::E2F4 GM12878\n",
      "GM12878 loop_loop (59, 16)\n",
      "***** 456 E2F4::FOXO1 MM\n",
      "COLO-SCR-DMSO loop_loop (4, 16)\n",
      "WM-SCR-DMSO loop_loop (23, 16)\n",
      "***** 457 E2F8::SMAD4 MM\n",
      "COLO-SCR-DMSO pro_pro (2, 16)\n",
      "COLO-SCR-DMSO loop_loop (5, 16)\n",
      "WM-SCR-DMSO pro_pro (2, 16)\n",
      "WM-SCR-DMSO loop_loop (9, 16)\n",
      "***** 458 MAZ::E2F8 MM\n",
      "COLO-SCR-DMSO pro_pro (16, 16)\n",
      "COLO-SCR-DMSO loop_loop (159, 16)\n",
      "WM-SCR-DMSO pro_pro (16, 16)\n",
      "WM-SCR-DMSO loop_loop (527, 16)\n",
      "***** 459 CUX1::FOXO1 MM\n",
      "COLO-SCR-DMSO loop_loop (1, 16)\n",
      "WM-SCR-DMSO loop_loop (1, 16)\n",
      "***** 460 NFIC::ETV3 MM\n",
      "COLO-SCR-DMSO loop_loop (8, 16)\n",
      "WM-SCR-DMSO pro_pro (3, 16)\n",
      "WM-SCR-DMSO loop_loop (30, 16)\n",
      "***** 461 BPTF::ETS2 MM\n",
      "COLO-SCR-DMSO loop_loop (21, 16)\n",
      "WM-SCR-DMSO loop_loop (22, 16)\n",
      "***** 462 CEBPZ::ETS2 MM\n",
      "COLO-SCR-DMSO pro_pro (3, 16)\n",
      "COLO-SCR-DMSO loop_loop (17, 16)\n",
      "WM-SCR-DMSO loop_loop (28, 16)\n",
      "***** 463 IRF3::CEBPZ MM\n",
      "COLO-SCR-DMSO pro_pro (7, 16)\n",
      "COLO-SCR-DMSO loop_loop (71, 16)\n",
      "WM-SCR-DMSO pro_pro (7, 16)\n",
      "WM-SCR-DMSO loop_loop (90, 16)\n",
      "***** 464 CUX1::RFX2 MM\n",
      "COLO-SCR-DMSO loop_loop (1, 16)\n",
      "WM-SCR-DMSO loop_loop (0, 16)\n",
      "***** 465 ETV3::HOXA9 MM\n",
      "***** 466 E2F4::GATA3 MM\n",
      "COLO-SCR-DMSO loop_loop (7, 16)\n",
      "***** 467 MEF2D::GATA3 MM\n",
      "COLO-SCR-DMSO pro_pro (4, 16)\n",
      "COLO-SCR-DMSO loop_loop (11, 16)\n",
      "***** 468 ZFX::E2F6 SCC\n",
      "A431-CTRLi pro_pro (61, 16)\n",
      "A431-CTRLi loop_loop (917, 16)\n",
      "CAL27-CTRLi pro_pro (78, 16)\n",
      "CAL27-CTRLi loop_loop (957, 16)\n",
      "SCC13-CTRLi pro_pro (60, 16)\n",
      "SCC13-CTRLi loop_loop (802, 16)\n",
      "***** 469 E2F3::TP63 SCC\n",
      "A431-CTRLi loop_loop (297, 16)\n",
      "CAL27-CTRLi pro_pro (17, 16)\n",
      "CAL27-CTRLi loop_loop (341, 16)\n",
      "SCC13-CTRLi pro_pro (10, 16)\n",
      "SCC13-CTRLi loop_loop (266, 16)\n",
      "***** 470 IRX3::TP63 SCC\n",
      "A431-CTRLi pro_pro (6, 16)\n",
      "A431-CTRLi loop_loop (128, 16)\n",
      "CAL27-CTRLi pro_pro (7, 16)\n",
      "CAL27-CTRLi loop_loop (158, 16)\n",
      "SCC13-CTRLi pro_pro (8, 16)\n",
      "SCC13-CTRLi loop_loop (125, 16)\n",
      "***** 471 NR4A1::TP63 SCC\n",
      "A431-CTRLi pro_pro (3, 16)\n",
      "A431-CTRLi loop_loop (116, 16)\n",
      "CAL27-CTRLi pro_pro (3, 16)\n",
      "CAL27-CTRLi loop_loop (119, 16)\n",
      "SCC13-CTRLi pro_pro (6, 16)\n",
      "SCC13-CTRLi loop_loop (105, 16)\n",
      "***** 472 TP63::ZNF354A SCC\n",
      "A431-CTRLi pro_pro (8, 16)\n",
      "A431-CTRLi loop_loop (185, 16)\n",
      "CAL27-CTRLi pro_pro (7, 16)\n",
      "CAL27-CTRLi loop_loop (140, 16)\n",
      "SCC13-CTRLi loop_loop (131, 16)\n",
      "CPU times: user 10min 1s, sys: 19.9 s, total: 10min 20s\n",
      "Wall time: 3min 46s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# vocab_tissue_df = pd.DataFrame()\n",
    "# loop through vocab words\n",
    "for idx, row in overall_df_filt.iterrows():\n",
    "    vocab = row['vocab']\n",
    "    tissue = row['tissue']\n",
    "    foot_tissues = tissue_to_celltype_mapper[tissue]\n",
    "    print('*****', idx, vocab, tissue)\n",
    "\n",
    "    # loop through the possible tissues that can use that vocab word\n",
    "    for foot_tissue in foot_tissues:\n",
    "        foot_df = tissue_to_foot_df[foot_tissue]\n",
    "        tf_tf_loop_orient_dict =  check_tf_tf_loop_occur(vocab, foot_tissue ,geneset_dict[tissue])\n",
    "        \n",
    "        # loop through two types of configurations (priority, pro-pro then loop-loop)\n",
    "        for type_loop in ['pro_pro','loop_loop']:\n",
    "            if tf_tf_loop_orient_dict[type_loop+'_count']>0:\n",
    "                genes_loop_type = tf_tf_loop_orient_dict[type_loop+'_genes'].split('|')\n",
    "                foot_df_sel = foot_df[foot_df.TSS.isin(genes_loop_type)]\n",
    "                foot_df_sel['tf'] = foot_df_sel.id_trim.map(tf_id_to_name_dict)\n",
    "                foot_df_sel = foot_df_sel[foot_df_sel.tf.isin(vocab.split('::'))].drop_duplicates()\n",
    "                foot_df_sel['vocab'] = vocab\n",
    "                foot_df_sel['genes'] = tf_tf_loop_orient_dict[type_loop+'_genes']\n",
    "                foot_df_sel['tissue'] = tissue\n",
    "                foot_df_sel['foot_tissue'] = foot_tissue\n",
    "                foot_df_sel['type_loop'] = type_loop\n",
    "                print(foot_tissue, type_loop, foot_df_sel.shape)\n",
    "#                 vocab_tissue_df = pd.concat([vocab_tissue_df, foot_df_sel],axis=0,sort=False)\n",
    "                vocab_tissue_dict[tissue] = pd.concat([ vocab_tissue_dict[tissue], foot_df_sel],axis=0,sort=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 950,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df = get_hits(pd.read_csv('../data/processed/fig4_modelling/tf_tf_pairs/expr_Colon_pro_pro_vocab_info.csv'))\n",
    "# df.vocab_pair.value_counts()['HOXA11::KLF5']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 951,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df2 = get_hits(pd.read_csv('../data/processed/fig4_modelling/tf_tf_pairs/expr_Colon_loop_loop_vocab_info.csv'))\n",
    "# df2.vocab_pair.value_counts()['HOXA11::KLF5']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 952,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df = pd.DataFrame()\n",
    "# # loop through vocab words\n",
    "# # for idx, row in overall_df_filt.iterrows():\n",
    "# vocab = 'NFATC4::STAT1'#row['vocab']\n",
    "# tissue = 'MM'#row['tissue']\n",
    "# foot_tissues = tissue_to_celltype_mapper[tissue]\n",
    "# print('*****', idx, vocab, tissue)\n",
    "\n",
    "# # loop through the possible tissues that can use that vocab word\n",
    "# for foot_tissue in foot_tissues:\n",
    "#     foot_df = tissue_to_foot_df[foot_tissue]\n",
    "#     tf_tf_loop_orient_dict =  check_tf_tf_loop_occur(vocab, foot_tissue ,geneset_dict[tissue])\n",
    "#     print(tf_tf_loop_orient_dict)\n",
    "#     # loop through two types of configurations (priority, pro-pro then loop-loop)\n",
    "#     for type_loop in ['pro_pro','loop_loop']:\n",
    "#         if tf_tf_loop_orient_dict[type_loop+'_count']>0:\n",
    "#             genes_loop_type = tf_tf_loop_orient_dict[type_loop+'_genes'].split('|')\n",
    "#             foot_df_sel = foot_df[foot_df.TSS.isin(genes_loop_type)]\n",
    "#             foot_df_sel['tf'] = foot_df_sel.id_trim.map(tf_id_to_name_dict)\n",
    "#             foot_df_sel = foot_df_sel[foot_df_sel.tf.isin(vocab.split('::'))].drop_duplicates()\n",
    "#             foot_df_sel['vocab'] = vocab\n",
    "#             foot_df_sel['genes'] = tf_tf_loop_orient_dict[type_loop+'_genes']\n",
    "#             foot_df_sel['tissue'] = tissue\n",
    "#             foot_df_sel['foot_tissue'] = foot_tissue\n",
    "#             foot_df_sel['type_loop'] = type_loop\n",
    "#             print(foot_tissue, type_loop, foot_df_sel.shape)\n",
    "#             df = pd.concat([df, foot_df_sel],axis=0,sort=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 953,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 954,
   "metadata": {},
   "outputs": [],
   "source": [
    "# vocab_instances_df = get_hits(vocab_tissue_df[vocab_tissue_df.tissue=='Melanocytes'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 955,
   "metadata": {},
   "outputs": [],
   "source": [
    "# vocab_instances_df.vocab_pair.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 956,
   "metadata": {},
   "outputs": [],
   "source": [
    "# vocab_instances_df[vocab_instances_df.vocab_pair=='HOXA11::KLF5']#.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 957,
   "metadata": {},
   "outputs": [],
   "source": [
    "# vocab_instances_df.vocab_pair.value_counts()['FOXM1::SP4']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 958,
   "metadata": {},
   "outputs": [],
   "source": [
    "# vocab_instances_df_filt = filter_results(vocab_instances_df,min_hits_per_vocab=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 959,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Colon (60018, 16)\n",
      "GDSD6 (31694, 16)\n",
      "GM12878 (20945, 16)\n",
      "Melanocytes (24377, 16)\n",
      "SCC (434942, 16)\n",
      "MM (28204, 16)\n",
      "CPU times: user 1min 39s, sys: 3.38 s, total: 1min 43s\n",
      "Wall time: 1min 43s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "for tissue, df in vocab_tissue_dict.items():\n",
    "    print(tissue,  df.shape)\n",
    "    df.to_csv(os.path.join(save_dir, 'vocab_tissue_df_'+tissue+'.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 960,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%time\n",
    "# print(vocab_tissue_df.shape)\n",
    "# vocab_tissue_df.to_csv(os.path.join(save_dir, 'vocab_tissue_df.csv'))\n",
    "# vocab_tissue_df[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3F filter vocab for 10 genomic instances < 135 bp away ( stringent filter < 50 bp away)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1014,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "********** Colon\n",
      "(60018, 16)\n",
      "num_hits 4515\n",
      "num motifs in tissue vocab 52\n",
      "pre genomic instance filter\n",
      "num vocab in expression enrichment:  76\n",
      "shape (4344, 14)\n",
      "original num vocab 76\n",
      "filt num vocab 42\n",
      "post genomic instance filter\n",
      "num vocab in expression enrichment:  42\n",
      "********** GDSD6\n",
      "(31694, 16)\n",
      "num_hits 2918\n",
      "num motifs in tissue vocab 60\n",
      "pre genomic instance filter\n",
      "num vocab in expression enrichment:  86\n",
      "shape (2787, 14)\n",
      "original num vocab 86\n",
      "filt num vocab 49\n",
      "post genomic instance filter\n",
      "num vocab in expression enrichment:  49\n",
      "********** GM12878\n",
      "(20945, 16)\n",
      "num_hits 1920\n",
      "num motifs in tissue vocab 42\n",
      "pre genomic instance filter\n",
      "num vocab in expression enrichment:  60\n",
      "shape (1789, 14)\n",
      "original num vocab 60\n",
      "filt num vocab 39\n",
      "post genomic instance filter\n",
      "num vocab in expression enrichment:  39\n",
      "********** Melanocytes\n",
      "(24377, 16)\n",
      "num_hits 2524\n",
      "num motifs in tissue vocab 47\n",
      "pre genomic instance filter\n",
      "num vocab in expression enrichment:  65\n",
      "shape (2427, 14)\n",
      "original num vocab 65\n",
      "filt num vocab 26\n",
      "post genomic instance filter\n",
      "num vocab in expression enrichment:  26\n",
      "********** SCC\n",
      "(434942, 16)\n",
      "num_hits 31445\n",
      "num motifs in tissue vocab 34\n",
      "pre genomic instance filter\n",
      "num vocab in expression enrichment:  45\n",
      "shape (14378, 14)\n",
      "original num vocab 45\n",
      "filt num vocab 43\n",
      "post genomic instance filter\n",
      "num vocab in expression enrichment:  43\n",
      "********** MM\n",
      "(28204, 16)\n",
      "num_hits 2748\n",
      "num motifs in tissue vocab 28\n",
      "pre genomic instance filter\n",
      "num vocab in expression enrichment:  46\n",
      "shape (2082, 14)\n",
      "original num vocab 46\n",
      "filt num vocab 40\n",
      "post genomic instance filter\n",
      "num vocab in expression enrichment:  40\n",
      "CPU times: user 3min 39s, sys: 236 ms, total: 3min 39s\n",
      "Wall time: 3min 39s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "vocab_instances_tissue_dict= {}\n",
    "for tissue, df in vocab_tissue_dict.items():\n",
    "    print('**********', tissue)\n",
    "    vocab_instances_df = get_hits(df)\n",
    "        \n",
    "    motifs_in_tissue = sorted(set(list(vocab_instances_df.vocab1.unique())+list(vocab_instances_df.vocab2.unique())))\n",
    "    print('num motifs in tissue vocab', len(motifs_in_tissue))\n",
    "    \n",
    "    vocab_instances_df = vocab_instances_df[['vocab_pair','tot_dist','btn_dist','chr','start','stop',\n",
    "     'vocab1','vocab1_start','vocab1_stop', 'vocab2','vocab2_start','vocab2_stop',\n",
    "     'tissue','loop_type' ]].drop_duplicates()\n",
    "\n",
    "    # print(motifs_pro_pro)\n",
    "    #filter expression enriched vocab words if then have at least 10 genomic instances then get stats\n",
    "    print('pre genomic instance filter')\n",
    "    print('num vocab in expression enrichment: ', vocab_instances_df.vocab_pair.unique().shape[0])\n",
    "    vocab_instances_df_filt = filter_results(vocab_instances_df,min_hits_per_vocab=10)\n",
    "    print('post genomic instance filter')\n",
    "    print('num vocab in expression enrichment: ', vocab_instances_df_filt.vocab_pair.unique().shape[0])\n",
    "\n",
    "    vocab_instances_tissue_dict[tissue] = vocab_instances_df_filt\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1015,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "***** Colon ../data/processed/fig4_modelling/vocab_sum_final/vocab_instances_df_filt_Colon.csv\n",
      "num vocab in expression enrichment:  42\n",
      "['E2F7::THAP1' 'KLF15::THAP1' 'SP2::THAP1' 'KLF15::OSR2' 'MESP1::THAP1'\n",
      " 'RARG::THAP1' 'SOX13::THAP1' 'THAP1::THRB' 'NR2F1::THAP1' 'BARX2::THAP1'\n",
      " 'KLF15::KLF5' 'KLF15::OTX1' 'KLF5::RUNX3' 'OTX1::THAP1' 'SP4::THAP1'\n",
      " 'SNAI2::THAP1' 'FLI1::FOXM1' 'FOXM1::PATZ1' 'FOXM1::SP2' 'KLF15::SNAI2'\n",
      " 'KLF5::SOX13' 'MAFB::THAP1' 'OSR2::THAP1' 'RARA::THAP1' 'RORA::THAP1'\n",
      " 'SNAI2::WT1' 'NFE2L1::NFIA' 'KLF15::MSX2' 'KLF5::ZNF257' 'KLF5::NR2F1'\n",
      " 'BARX2::FOXF2' 'BARX2::FOXM1' 'FOXM1::KLF15' 'FOXM1::SOX13' 'FOXM1::SP4'\n",
      " 'FOXM1::WT1' 'KLF15::RFX2' 'RFX2::THAP1' 'KLF15::ZFP28' 'KLF5::ZFP28'\n",
      " 'MSX2::THAP1' 'BATF::KLF15']\n",
      "***** GDSD6 ../data/processed/fig4_modelling/vocab_sum_final/vocab_instances_df_filt_GDSD6.csv\n",
      "num vocab in expression enrichment:  49\n",
      "['KLF5::NFIC' 'KLF5::UBP1' 'NFIC::RORA' 'RORA::UBP1' 'ARID5B::KLF5'\n",
      " 'EGR2::KLF4' 'EGR2::KLF5' 'HBP1::KLF5' 'SP1::XBP1' 'KLF4::SREBF2'\n",
      " 'KLF5::SREBF2' 'EGR2::RORA' 'FOS::RORA' 'JUN::RORA' 'JUNB::RORA'\n",
      " 'JUND::RORA' 'KLF5::MAFG' 'KLF5::MAFK' 'MAFF::RORA' 'MAFG::RORA'\n",
      " 'RORA::TP63' 'EGR2::RARG' 'BACH1::KLF5' 'HMGA1::KLF5' 'EGR2::TP63'\n",
      " 'STAT3::UBP1' 'KLF5::NR3C1' 'HBP1::RORA' 'HMGA1::RORA' 'PRDM1::RORA'\n",
      " 'KLF4::SALL4' 'EGR2::FOSL2' 'EGR2::JUN' 'EGR2::JUNB' 'KLF5::TGIF1'\n",
      " 'NFE2L1::RORA' 'ASCL2::KLF5' 'KLF4::TFE3' 'ASCL2::KLF4' 'EGR2::SNAI2'\n",
      " 'FOSL2::SREBF2' 'JUN::SREBF2' 'NFIC::RARG' 'SALL4::TP63' 'MAF::RORA'\n",
      " 'EGR3::KLF4' 'ID4::TP63' 'EGR3::TP63' 'RORA::SMAD3']\n",
      "***** GM12878 ../data/processed/fig4_modelling/vocab_sum_final/vocab_instances_df_filt_GM12878.csv\n",
      "num vocab in expression enrichment:  39\n",
      "['TCF3::ZNF589' 'EGR1::IKZF1' 'IKZF1::SP4' 'IKZF1::ZNF589' 'SP4::TCF3'\n",
      " 'PAX5::ZNF589' 'BATF::ZNF589' 'IRF2::IRF4' 'IRF4::ZBTB49' 'ETV5::IRF8'\n",
      " 'IKZF1::IRF3' 'IRF8::UBP1' 'IRF8::ZNF589' 'IKZF1::IRF8' 'IKZF1::ZNF563'\n",
      " 'IRF4::IRF8' 'IRF8::MYBL2' 'ETS1::IRF4' 'IRF8::SPIB' 'IRF8::ZNF394'\n",
      " 'IKZF1::MYBL2' 'IRF4::SP4' 'ETS1::IKZF1' 'IRF4::PLAG1' 'IRF4::ZNF816'\n",
      " 'IKZF1::IRF4' 'IKZF1::PAX5' 'PAX5::SP4' 'EBF1::SP4' 'EBF1::ZNF816'\n",
      " 'EGR1::IRF4' 'HBP1::IRF8' 'IRF8::STAT1' 'IRF8::STAT2' 'IRF8::ZNF816'\n",
      " 'CBFB::IRF8' 'IKZF1::IRF1' 'IKZF1::IRF2' 'IRF8::ZBTB49']\n",
      "***** Melanocytes ../data/processed/fig4_modelling/vocab_sum_final/vocab_instances_df_filt_Melanocytes.csv\n",
      "num vocab in expression enrichment:  26\n",
      "['E2F4::KLF15' 'KLF15::PRRX1' 'KLF15::ZNF219' 'KLF15::ZNF589'\n",
      " 'SP2::ZNF324' 'FOXL1::PRRX1' 'ETV2::SP2' 'ETV2::ZNF589' 'ETV2::PRRX1'\n",
      " 'NFATC1::PRRX1' 'IRF4::PRRX1' 'KLF15::SALL4' 'PRRX1::RARA' 'PRRX1::SALL4'\n",
      " 'EN2::PRRX1' 'PRRX1::TFAP2B' 'PBX1::PRRX1' 'MITF::ZNF589' 'KLF15::TFAP2B'\n",
      " 'SP2::TFAP2B' 'TFAP2B::ZNF589' 'MITF::PRRX1' 'KLF5::NFYB' 'FOXD3::PRRX1'\n",
      " 'FOXD3::ZNF589' 'PRRX1::SOX10']\n",
      "***** SCC ../data/processed/fig4_modelling/vocab_sum_final/vocab_instances_df_filt_SCC.csv\n",
      "num vocab in expression enrichment:  43\n",
      "['EGR1::SP1' 'KLF16::SP1' 'KLF16::SP3' 'KLF5::SP3' 'KLF6::SP1' 'MAZ::SP1'\n",
      " 'PURA::SP1' 'SP1::SP2' 'SP1::SP3' 'SP1::SP4' 'SP1::TBX1' 'SP1::TFAP2C'\n",
      " 'SP1::VEZF1' 'SP1::ZBTB17' 'SP1::ZNF770' 'SP3::TFAP2C' 'SP3::ZBTB17'\n",
      " 'ZFX::E2F6' 'EGR1::SP3' 'KLF16::TFAP2C' 'KLF4::TFAP2C' 'KLF5::TFAP2C'\n",
      " 'TFAP2C::ZNF219' 'NFAT5::SP1' 'SNAI2::SP1' 'E2F3::TP63' 'IRX3::TP63'\n",
      " 'ARNT::SP1' 'SP1::ZBTB14' 'TFAP2C::ZBTB14' 'E2F1::SP1' 'KLF4::KLF5'\n",
      " 'KLF4::SP1' 'KLF4::SP3' 'KLF5::ZNF219' 'NR2F1::SP1' 'SP1::TFDP1'\n",
      " 'SP1::ZNF219' 'SP3::ZNF219' 'E2F6::SP1' 'SP1::TCF3' 'ETS1::SP1'\n",
      " 'NRF1::SP1']\n",
      "***** MM ../data/processed/fig4_modelling/vocab_sum_final/vocab_instances_df_filt_MM.csv\n",
      "num vocab in expression enrichment:  40\n",
      "['NFE2L1::PRRX1' 'NFE2L1::UBP1' 'NFE2L1::SOX13' 'STAT1::UBP1'\n",
      " 'ARID5B::NFE2L1' 'PRRX1::STAT1' 'SOX13::STAT1' 'NFAT5::NFE2L1'\n",
      " 'NFATC4::NFE2L1' 'HMGA1::NFE2L1' 'EN2::NFE2L1' 'NFE2L1::ZBTB49'\n",
      " 'EGR2::NFE2L1' 'NFE2L1::TBX15' 'EGR4::NFE2L1' 'ARID5B::STAT1'\n",
      " 'NFAT5::STAT1' 'NFATC4::STAT1' 'NFE2L1::SP4' 'SP4::STAT1' 'EN2::STAT1'\n",
      " 'STAT1::TBX15' 'HBP1::STAT1' 'EGR2::STAT1' 'HMGA1::STAT1' 'SOX10::UBP1'\n",
      " 'SOX10::SOX13' 'LEF1::PRRX1' 'LEF1::UBP1' 'NFAT5::SOX10' 'NFATC4::SOX10'\n",
      " 'LEF1::SOX13' 'LEF1::NFAT5' 'EGR2::SOX10' 'SOX10::SP4' 'ARID5B::SOX10'\n",
      " 'HMGA1::SOX10' 'ARID5B::LEF1' 'HMGA1::LEF1' 'NFE2L1::NR3C1']\n",
      "total number of vocab words 239\n"
     ]
    }
   ],
   "source": [
    "num_total_vocab=0\n",
    "for tissue in tissue_to_celltype_mapper.keys():\n",
    "    print('*****', tissue, os.path.join(save_dir, 'vocab_instances_df_filt_'+tissue+'.csv'))\n",
    "    vocab_instances_df_filt = vocab_instances_tissue_dict[tissue]\n",
    "    print('num vocab in expression enrichment: ', vocab_instances_df_filt.vocab_pair.unique().shape[0])\n",
    "    num_total_vocab+=vocab_instances_df_filt.vocab_pair.unique().shape[0]\n",
    "    print(vocab_instances_df_filt.vocab_pair.unique())\n",
    "    vocab_instances_df_filt.to_csv(os.path.join(save_dir, 'vocab_instances_df_filt_'+tissue+'.csv'))\n",
    "print('total number of vocab words',num_total_vocab )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3G get the 20 shortest distance sequences per vocab pair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1087,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "135"
      ]
     },
     "execution_count": 1087,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "num_seq = 20\n",
    "MAX_DIST\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1088,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_padded_length(row, max_dist=MAX_DIST):\n",
    "    # get sequence [start_pad, stop_pad]\n",
    "    dist_to_add = max_dist - row['tot_dist']\n",
    "    dist_to_add_L = int(dist_to_add/2) \n",
    "    dist_to_add_R = int(dist_to_add/2) + (dist_to_add % 2) - 1\n",
    "    start_pad = row['start']-dist_to_add_L\n",
    "    stop_pad = row['stop']+dist_to_add_R\n",
    "    if start_pad<0:\n",
    "        raise\n",
    "#     if (stop_pad- start_pad)!=max_dist:\n",
    "#         raise\n",
    "    \n",
    "    return start_pad, stop_pad\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1089,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Colon (4171, 14)\n",
      "(802, 16) ../data/processed/fig4_modelling/vocab_sum_final/vocab_instance20_df_filt_Colon.csv\n",
      "GDSD6 (2544, 14)\n",
      "(919, 16) ../data/processed/fig4_modelling/vocab_sum_final/vocab_instance20_df_filt_GDSD6.csv\n",
      "GM12878 (1618, 14)\n",
      "(715, 16) ../data/processed/fig4_modelling/vocab_sum_final/vocab_instance20_df_filt_GM12878.csv\n",
      "Melanocytes (2264, 14)\n",
      "(517, 16) ../data/processed/fig4_modelling/vocab_sum_final/vocab_instance20_df_filt_Melanocytes.csv\n",
      "SCC (14364, 14)\n",
      "(853, 16) ../data/processed/fig4_modelling/vocab_sum_final/vocab_instance20_df_filt_SCC.csv\n",
      "MM (2051, 14)\n",
      "(780, 16) ../data/processed/fig4_modelling/vocab_sum_final/vocab_instance20_df_filt_MM.csv\n"
     ]
    }
   ],
   "source": [
    "vocab_instances20_tissue_dict = {}\n",
    "vocab_instances20_alltissue_df = pd.DataFrame()\n",
    "for tissue, df in vocab_instances_tissue_dict.items():\n",
    "    print(tissue, df.shape)\n",
    "    df_short = df.sort_values('btn_dist',ascending=True).groupby('vocab_pair').head(num_seq).sort_values('vocab_pair').drop_duplicates().reset_index(drop=True)\n",
    "    df_short = pd.concat([df_short,pd.DataFrame(df_short.apply(get_padded_length,axis=1).tolist(), index=df_short.index,columns=['start_pad','stop_pad']) ],axis=1)\n",
    "    print(df_short.shape, os.path.join(save_dir, 'vocab_instance20_df_filt_'+tissue+'.csv'))\n",
    "    vocab_instances20_tissue_dict[tissue] = df_short\n",
    "    df_short.to_csv(os.path.join(save_dir, 'vocab_instances20_df_filt_'+tissue+'.csv'))\n",
    "    vocab_instances20_alltissue_df = pd.concat([vocab_instances20_alltissue_df, df_short],axis=0,sort=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1247,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4586, 20)\n"
     ]
    }
   ],
   "source": [
    "vocab_instances20_alltissue_df.reset_index(drop=True, inplace=True)\n",
    "vocab_instances20_alltissue_df[ 'vocab_tissue_count'] = vocab_instances20_alltissue_df.groupby(['vocab_pair','tissue']).cumcount()+1\n",
    "vocab_instances20_alltissue_df['name'] = vocab_instances20_alltissue_df.vocab_pair + '+'+ vocab_instances20_alltissue_df.tissue + '+'+ vocab_instances20_alltissue_df.vocab_tissue_count.map(str)\n",
    "\n",
    "print(vocab_instances20_alltissue_df.shape)\n",
    "vocab_instances20_alltissue_df.to_csv(os.path.join(save_dir, 'vocab_instances20_alltissue_df.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1252,
   "metadata": {},
   "outputs": [],
   "source": [
    "# vocab_instances20_alltissue_df.iloc[4585,:]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### checking some stats about `vocab_instances20_alltissue_df`\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1250,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>vocab_pair</th>\n",
       "      <th>tot_dist</th>\n",
       "      <th>btn_dist</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>vocab1</th>\n",
       "      <th>vocab1_start</th>\n",
       "      <th>vocab1_stop</th>\n",
       "      <th>vocab2</th>\n",
       "      <th>vocab2_start</th>\n",
       "      <th>vocab2_stop</th>\n",
       "      <th>tissue</th>\n",
       "      <th>loop_type</th>\n",
       "      <th>start_pad</th>\n",
       "      <th>stop_pad</th>\n",
       "      <th>vocab_tissue_count</th>\n",
       "      <th>name</th>\n",
       "      <th>sequence</th>\n",
       "      <th>check_rs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>BARX2::FOXF2</td>\n",
       "      <td>28</td>\n",
       "      <td>14</td>\n",
       "      <td>chr1</td>\n",
       "      <td>206848337</td>\n",
       "      <td>206848365</td>\n",
       "      <td>BARX2</td>\n",
       "      <td>206848337</td>\n",
       "      <td>206848341</td>\n",
       "      <td>FOXF2</td>\n",
       "      <td>206848355</td>\n",
       "      <td>206848365</td>\n",
       "      <td>Colon</td>\n",
       "      <td>pro_pro</td>\n",
       "      <td>206848284</td>\n",
       "      <td>206848418</td>\n",
       "      <td>1</td>\n",
       "      <td>BARX2::FOXF2+Colon+1</td>\n",
       "      <td>CTCAGCTCTCCCACAAGATCTCTGGGGAAGCTGGAAGAGGGCCACT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>BARX2::FOXF2</td>\n",
       "      <td>54</td>\n",
       "      <td>40</td>\n",
       "      <td>chr4</td>\n",
       "      <td>38859601</td>\n",
       "      <td>38859655</td>\n",
       "      <td>FOXF2</td>\n",
       "      <td>38859601</td>\n",
       "      <td>38859611</td>\n",
       "      <td>BARX2</td>\n",
       "      <td>38859651</td>\n",
       "      <td>38859655</td>\n",
       "      <td>Colon</td>\n",
       "      <td>loop_loop</td>\n",
       "      <td>38859561</td>\n",
       "      <td>38859695</td>\n",
       "      <td>2</td>\n",
       "      <td>BARX2::FOXF2+Colon+2</td>\n",
       "      <td>GTGGCTCTGAGGGAATTTCAAGTAACCGTGGACTAAAGTCTTAAGA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>BARX2::FOXF2</td>\n",
       "      <td>46</td>\n",
       "      <td>32</td>\n",
       "      <td>chr6</td>\n",
       "      <td>74222776</td>\n",
       "      <td>74222822</td>\n",
       "      <td>FOXF2</td>\n",
       "      <td>74222776</td>\n",
       "      <td>74222786</td>\n",
       "      <td>BARX2</td>\n",
       "      <td>74222818</td>\n",
       "      <td>74222822</td>\n",
       "      <td>Colon</td>\n",
       "      <td>loop_loop</td>\n",
       "      <td>74222732</td>\n",
       "      <td>74222866</td>\n",
       "      <td>3</td>\n",
       "      <td>BARX2::FOXF2+Colon+3</td>\n",
       "      <td>GCCACTGCACTCCAGCCTGGTGACAGAGCGAGACTCTGTCTCAAGA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>BARX2::FOXF2</td>\n",
       "      <td>16</td>\n",
       "      <td>2</td>\n",
       "      <td>chr3</td>\n",
       "      <td>49384674</td>\n",
       "      <td>49384690</td>\n",
       "      <td>FOXF2</td>\n",
       "      <td>49384674</td>\n",
       "      <td>49384684</td>\n",
       "      <td>BARX2</td>\n",
       "      <td>49384686</td>\n",
       "      <td>49384690</td>\n",
       "      <td>Colon</td>\n",
       "      <td>pro_pro</td>\n",
       "      <td>49384615</td>\n",
       "      <td>49384749</td>\n",
       "      <td>4</td>\n",
       "      <td>BARX2::FOXF2+Colon+4</td>\n",
       "      <td>AGTATTATGTGCTCATGGTCCCTACCCTCCAGGGTTCACAGTGTGG...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>BARX2::FOXF2</td>\n",
       "      <td>16</td>\n",
       "      <td>2</td>\n",
       "      <td>chr10</td>\n",
       "      <td>112247520</td>\n",
       "      <td>112247536</td>\n",
       "      <td>BARX2</td>\n",
       "      <td>112247520</td>\n",
       "      <td>112247524</td>\n",
       "      <td>FOXF2</td>\n",
       "      <td>112247526</td>\n",
       "      <td>112247536</td>\n",
       "      <td>Colon</td>\n",
       "      <td>pro_pro</td>\n",
       "      <td>112247461</td>\n",
       "      <td>112247595</td>\n",
       "      <td>5</td>\n",
       "      <td>BARX2::FOXF2+Colon+5</td>\n",
       "      <td>ATGTGGCTGCGGGGCTTTAGGAGCTGACTCTTTGGTCACGGGGAGA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     vocab_pair  tot_dist  btn_dist    chr      start       stop vocab1  \\\n",
       "0  BARX2::FOXF2        28        14   chr1  206848337  206848365  BARX2   \n",
       "1  BARX2::FOXF2        54        40   chr4   38859601   38859655  FOXF2   \n",
       "2  BARX2::FOXF2        46        32   chr6   74222776   74222822  FOXF2   \n",
       "3  BARX2::FOXF2        16         2   chr3   49384674   49384690  FOXF2   \n",
       "4  BARX2::FOXF2        16         2  chr10  112247520  112247536  BARX2   \n",
       "\n",
       "   vocab1_start  vocab1_stop vocab2  vocab2_start  vocab2_stop tissue  \\\n",
       "0     206848337    206848341  FOXF2     206848355    206848365  Colon   \n",
       "1      38859601     38859611  BARX2      38859651     38859655  Colon   \n",
       "2      74222776     74222786  BARX2      74222818     74222822  Colon   \n",
       "3      49384674     49384684  BARX2      49384686     49384690  Colon   \n",
       "4     112247520    112247524  FOXF2     112247526    112247536  Colon   \n",
       "\n",
       "   loop_type  start_pad   stop_pad  vocab_tissue_count                  name  \\\n",
       "0    pro_pro  206848284  206848418                   1  BARX2::FOXF2+Colon+1   \n",
       "1  loop_loop   38859561   38859695                   2  BARX2::FOXF2+Colon+2   \n",
       "2  loop_loop   74222732   74222866                   3  BARX2::FOXF2+Colon+3   \n",
       "3    pro_pro   49384615   49384749                   4  BARX2::FOXF2+Colon+4   \n",
       "4    pro_pro  112247461  112247595                   5  BARX2::FOXF2+Colon+5   \n",
       "\n",
       "                                            sequence  check_rs  \n",
       "0  CTCAGCTCTCCCACAAGATCTCTGGGGAAGCTGGAAGAGGGCCACT...         0  \n",
       "1  GTGGCTCTGAGGGAATTTCAAGTAACCGTGGACTAAAGTCTTAAGA...         0  \n",
       "2  GCCACTGCACTCCAGCCTGGTGACAGAGCGAGACTCTGTCTCAAGA...         0  \n",
       "3  AGTATTATGTGCTCATGGTCCCTACCCTCCAGGGTTCACAGTGTGG...         0  \n",
       "4  ATGTGGCTGCGGGGCTTTAGGAGCTGACTCTTTGGTCACGGGGAGA...         0  "
      ]
     },
     "execution_count": 1250,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_alltissue_df[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1021,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "EGR1::SP3         20\n",
       "KLF4::SP1         20\n",
       "SP1::ZNF770       20\n",
       "TFAP2C::ZNF219    20\n",
       "ARNT::SP1         20\n",
       "                  ..\n",
       "HMGA1::LEF1       13\n",
       "FOXM1::SP2        13\n",
       "FLI1::FOXM1       13\n",
       "CBFB::IRF8        12\n",
       "MAF::RORA         12\n",
       "Name: vocab_pair, Length: 239, dtype: int64"
      ]
     },
     "execution_count": 1021,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_alltissue_df.vocab_pair.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1022,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "      <th>mean</th>\n",
       "      <th>median</th>\n",
       "      <th>std</th>\n",
       "      <th>tissue</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vocab_pair</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>IKZF1::IRF2</td>\n",
       "      <td>24</td>\n",
       "      <td>108</td>\n",
       "      <td>66.230769</td>\n",
       "      <td>79.0</td>\n",
       "      <td>30.199762</td>\n",
       "      <td>GM12878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>RORA::TP63</td>\n",
       "      <td>11</td>\n",
       "      <td>106</td>\n",
       "      <td>64.687500</td>\n",
       "      <td>72.0</td>\n",
       "      <td>29.999375</td>\n",
       "      <td>GDSD6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>IRF8::ZNF816</td>\n",
       "      <td>2</td>\n",
       "      <td>110</td>\n",
       "      <td>64.375000</td>\n",
       "      <td>70.5</td>\n",
       "      <td>35.910769</td>\n",
       "      <td>GM12878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>IKZF1::IRF1</td>\n",
       "      <td>27</td>\n",
       "      <td>94</td>\n",
       "      <td>66.357143</td>\n",
       "      <td>69.5</td>\n",
       "      <td>22.666842</td>\n",
       "      <td>GM12878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>EN2::NFE2L1</td>\n",
       "      <td>1</td>\n",
       "      <td>84</td>\n",
       "      <td>54.550000</td>\n",
       "      <td>66.0</td>\n",
       "      <td>26.937501</td>\n",
       "      <td>MM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>NFE2L1::ZBTB49</td>\n",
       "      <td>1</td>\n",
       "      <td>84</td>\n",
       "      <td>54.550000</td>\n",
       "      <td>66.0</td>\n",
       "      <td>26.937501</td>\n",
       "      <td>MM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>FOSL2::SREBF2</td>\n",
       "      <td>11</td>\n",
       "      <td>113</td>\n",
       "      <td>64.473684</td>\n",
       "      <td>66.0</td>\n",
       "      <td>30.909129</td>\n",
       "      <td>GDSD6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ETS1::IKZF1</td>\n",
       "      <td>21</td>\n",
       "      <td>116</td>\n",
       "      <td>60.650000</td>\n",
       "      <td>65.0</td>\n",
       "      <td>25.344521</td>\n",
       "      <td>GM12878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KLF5::ZNF257</td>\n",
       "      <td>1</td>\n",
       "      <td>88</td>\n",
       "      <td>52.850000</td>\n",
       "      <td>63.5</td>\n",
       "      <td>30.366663</td>\n",
       "      <td>Colon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>JUN::SREBF2</td>\n",
       "      <td>7</td>\n",
       "      <td>113</td>\n",
       "      <td>59.722222</td>\n",
       "      <td>63.0</td>\n",
       "      <td>33.643119</td>\n",
       "      <td>GDSD6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>IRX3::TP63</td>\n",
       "      <td>1</td>\n",
       "      <td>106</td>\n",
       "      <td>56.153846</td>\n",
       "      <td>63.0</td>\n",
       "      <td>40.147325</td>\n",
       "      <td>SCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>EGR2::STAT1</td>\n",
       "      <td>5</td>\n",
       "      <td>103</td>\n",
       "      <td>58.350000</td>\n",
       "      <td>61.5</td>\n",
       "      <td>33.988040</td>\n",
       "      <td>MM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>EGR2::SOX10</td>\n",
       "      <td>1</td>\n",
       "      <td>109</td>\n",
       "      <td>52.866667</td>\n",
       "      <td>60.0</td>\n",
       "      <td>32.793437</td>\n",
       "      <td>MM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>LEF1::NFAT5</td>\n",
       "      <td>1</td>\n",
       "      <td>117</td>\n",
       "      <td>58.400000</td>\n",
       "      <td>59.5</td>\n",
       "      <td>35.405025</td>\n",
       "      <td>MM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>BACH1::KLF5</td>\n",
       "      <td>4</td>\n",
       "      <td>102</td>\n",
       "      <td>50.200000</td>\n",
       "      <td>59.0</td>\n",
       "      <td>28.951190</td>\n",
       "      <td>GDSD6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>NFIC::RARG</td>\n",
       "      <td>3</td>\n",
       "      <td>101</td>\n",
       "      <td>53.857143</td>\n",
       "      <td>56.5</td>\n",
       "      <td>32.210503</td>\n",
       "      <td>GDSD6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>IKZF1::IRF4</td>\n",
       "      <td>4</td>\n",
       "      <td>99</td>\n",
       "      <td>55.350000</td>\n",
       "      <td>56.5</td>\n",
       "      <td>30.150980</td>\n",
       "      <td>GM12878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KLF5::ZFP28</td>\n",
       "      <td>5</td>\n",
       "      <td>103</td>\n",
       "      <td>60.705882</td>\n",
       "      <td>56.0</td>\n",
       "      <td>29.117788</td>\n",
       "      <td>Colon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>IKZF1::IRF8</td>\n",
       "      <td>17</td>\n",
       "      <td>106</td>\n",
       "      <td>57.733333</td>\n",
       "      <td>56.0</td>\n",
       "      <td>31.812995</td>\n",
       "      <td>GM12878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>EBF1::ZNF816</td>\n",
       "      <td>1</td>\n",
       "      <td>102</td>\n",
       "      <td>53.466667</td>\n",
       "      <td>55.0</td>\n",
       "      <td>32.902164</td>\n",
       "      <td>GM12878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>IRF8::ZBTB49</td>\n",
       "      <td>5</td>\n",
       "      <td>110</td>\n",
       "      <td>58.285714</td>\n",
       "      <td>53.5</td>\n",
       "      <td>38.433616</td>\n",
       "      <td>GM12878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>HBP1::RORA</td>\n",
       "      <td>25</td>\n",
       "      <td>108</td>\n",
       "      <td>60.000000</td>\n",
       "      <td>52.0</td>\n",
       "      <td>28.049064</td>\n",
       "      <td>GDSD6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>HMGA1::LEF1</td>\n",
       "      <td>23</td>\n",
       "      <td>86</td>\n",
       "      <td>53.538462</td>\n",
       "      <td>52.0</td>\n",
       "      <td>20.783068</td>\n",
       "      <td>MM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>FOXM1::PATZ1</td>\n",
       "      <td>1</td>\n",
       "      <td>88</td>\n",
       "      <td>49.650000</td>\n",
       "      <td>52.0</td>\n",
       "      <td>22.513797</td>\n",
       "      <td>Colon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>RFX2::THAP1</td>\n",
       "      <td>7</td>\n",
       "      <td>95</td>\n",
       "      <td>50.000000</td>\n",
       "      <td>52.0</td>\n",
       "      <td>33.819373</td>\n",
       "      <td>Colon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KLF4::TFE3</td>\n",
       "      <td>1</td>\n",
       "      <td>113</td>\n",
       "      <td>48.529412</td>\n",
       "      <td>51.0</td>\n",
       "      <td>42.649909</td>\n",
       "      <td>GDSD6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SOX10::SP4</td>\n",
       "      <td>8</td>\n",
       "      <td>99</td>\n",
       "      <td>48.333333</td>\n",
       "      <td>51.0</td>\n",
       "      <td>33.373785</td>\n",
       "      <td>MM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>EN2::STAT1</td>\n",
       "      <td>4</td>\n",
       "      <td>108</td>\n",
       "      <td>50.850000</td>\n",
       "      <td>50.0</td>\n",
       "      <td>31.529894</td>\n",
       "      <td>MM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>FOXM1::KLF15</td>\n",
       "      <td>1</td>\n",
       "      <td>96</td>\n",
       "      <td>48.571429</td>\n",
       "      <td>48.5</td>\n",
       "      <td>27.941462</td>\n",
       "      <td>Colon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>STAT1::TBX15</td>\n",
       "      <td>1</td>\n",
       "      <td>95</td>\n",
       "      <td>47.100000</td>\n",
       "      <td>48.5</td>\n",
       "      <td>31.560969</td>\n",
       "      <td>MM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>IRF8::STAT1</td>\n",
       "      <td>4</td>\n",
       "      <td>91</td>\n",
       "      <td>50.100000</td>\n",
       "      <td>48.5</td>\n",
       "      <td>33.234256</td>\n",
       "      <td>GM12878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>FOXM1::SP4</td>\n",
       "      <td>11</td>\n",
       "      <td>107</td>\n",
       "      <td>53.666667</td>\n",
       "      <td>48.0</td>\n",
       "      <td>29.834464</td>\n",
       "      <td>Colon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>IKZF1::PAX5</td>\n",
       "      <td>5</td>\n",
       "      <td>106</td>\n",
       "      <td>54.000000</td>\n",
       "      <td>48.0</td>\n",
       "      <td>37.307841</td>\n",
       "      <td>GM12878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>FOXM1::SP2</td>\n",
       "      <td>7</td>\n",
       "      <td>110</td>\n",
       "      <td>49.000000</td>\n",
       "      <td>47.0</td>\n",
       "      <td>29.484459</td>\n",
       "      <td>Colon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ASCL2::KLF4</td>\n",
       "      <td>9</td>\n",
       "      <td>97</td>\n",
       "      <td>45.000000</td>\n",
       "      <td>46.5</td>\n",
       "      <td>28.183609</td>\n",
       "      <td>GDSD6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>RORA::SMAD3</td>\n",
       "      <td>1</td>\n",
       "      <td>113</td>\n",
       "      <td>56.823529</td>\n",
       "      <td>46.0</td>\n",
       "      <td>35.560222</td>\n",
       "      <td>GDSD6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>NFIC::RORA</td>\n",
       "      <td>6</td>\n",
       "      <td>92</td>\n",
       "      <td>49.150000</td>\n",
       "      <td>46.0</td>\n",
       "      <td>28.897778</td>\n",
       "      <td>GDSD6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>HMGA1::RORA</td>\n",
       "      <td>4</td>\n",
       "      <td>98</td>\n",
       "      <td>44.500000</td>\n",
       "      <td>45.5</td>\n",
       "      <td>29.421618</td>\n",
       "      <td>GDSD6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>FOS::RORA</td>\n",
       "      <td>1</td>\n",
       "      <td>116</td>\n",
       "      <td>51.100000</td>\n",
       "      <td>44.5</td>\n",
       "      <td>36.898010</td>\n",
       "      <td>GDSD6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KLF5::MAFK</td>\n",
       "      <td>9</td>\n",
       "      <td>94</td>\n",
       "      <td>53.529412</td>\n",
       "      <td>44.0</td>\n",
       "      <td>28.985164</td>\n",
       "      <td>GDSD6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>FOXM1::WT1</td>\n",
       "      <td>12</td>\n",
       "      <td>113</td>\n",
       "      <td>49.210526</td>\n",
       "      <td>44.0</td>\n",
       "      <td>28.481142</td>\n",
       "      <td>Colon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>MSX2::THAP1</td>\n",
       "      <td>9</td>\n",
       "      <td>95</td>\n",
       "      <td>47.466667</td>\n",
       "      <td>43.0</td>\n",
       "      <td>31.656339</td>\n",
       "      <td>Colon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KLF15::ZFP28</td>\n",
       "      <td>3</td>\n",
       "      <td>76</td>\n",
       "      <td>40.600000</td>\n",
       "      <td>42.5</td>\n",
       "      <td>24.917548</td>\n",
       "      <td>Colon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>JUN::RORA</td>\n",
       "      <td>1</td>\n",
       "      <td>115</td>\n",
       "      <td>46.944444</td>\n",
       "      <td>42.5</td>\n",
       "      <td>35.659290</td>\n",
       "      <td>GDSD6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>NFE2L1::RORA</td>\n",
       "      <td>4</td>\n",
       "      <td>86</td>\n",
       "      <td>41.600000</td>\n",
       "      <td>42.5</td>\n",
       "      <td>27.350358</td>\n",
       "      <td>GDSD6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>HMGA1::STAT1</td>\n",
       "      <td>3</td>\n",
       "      <td>67</td>\n",
       "      <td>36.600000</td>\n",
       "      <td>42.5</td>\n",
       "      <td>22.276728</td>\n",
       "      <td>MM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>OSR2::THAP1</td>\n",
       "      <td>3</td>\n",
       "      <td>89</td>\n",
       "      <td>43.300000</td>\n",
       "      <td>39.5</td>\n",
       "      <td>26.870254</td>\n",
       "      <td>Colon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SNAI2::THAP1</td>\n",
       "      <td>7</td>\n",
       "      <td>62</td>\n",
       "      <td>38.150000</td>\n",
       "      <td>38.5</td>\n",
       "      <td>18.316731</td>\n",
       "      <td>Colon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>IRF8::SPIB</td>\n",
       "      <td>1</td>\n",
       "      <td>97</td>\n",
       "      <td>42.812500</td>\n",
       "      <td>38.5</td>\n",
       "      <td>31.471085</td>\n",
       "      <td>GM12878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>EGR2::SNAI2</td>\n",
       "      <td>19</td>\n",
       "      <td>70</td>\n",
       "      <td>42.200000</td>\n",
       "      <td>38.0</td>\n",
       "      <td>15.381465</td>\n",
       "      <td>GDSD6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                min  max       mean  median        std   tissue\n",
       "vocab_pair                                                     \n",
       "IKZF1::IRF2      24  108  66.230769    79.0  30.199762  GM12878\n",
       "RORA::TP63       11  106  64.687500    72.0  29.999375    GDSD6\n",
       "IRF8::ZNF816      2  110  64.375000    70.5  35.910769  GM12878\n",
       "IKZF1::IRF1      27   94  66.357143    69.5  22.666842  GM12878\n",
       "EN2::NFE2L1       1   84  54.550000    66.0  26.937501       MM\n",
       "NFE2L1::ZBTB49    1   84  54.550000    66.0  26.937501       MM\n",
       "FOSL2::SREBF2    11  113  64.473684    66.0  30.909129    GDSD6\n",
       "ETS1::IKZF1      21  116  60.650000    65.0  25.344521  GM12878\n",
       "KLF5::ZNF257      1   88  52.850000    63.5  30.366663    Colon\n",
       "JUN::SREBF2       7  113  59.722222    63.0  33.643119    GDSD6\n",
       "IRX3::TP63        1  106  56.153846    63.0  40.147325      SCC\n",
       "EGR2::STAT1       5  103  58.350000    61.5  33.988040       MM\n",
       "EGR2::SOX10       1  109  52.866667    60.0  32.793437       MM\n",
       "LEF1::NFAT5       1  117  58.400000    59.5  35.405025       MM\n",
       "BACH1::KLF5       4  102  50.200000    59.0  28.951190    GDSD6\n",
       "NFIC::RARG        3  101  53.857143    56.5  32.210503    GDSD6\n",
       "IKZF1::IRF4       4   99  55.350000    56.5  30.150980  GM12878\n",
       "KLF5::ZFP28       5  103  60.705882    56.0  29.117788    Colon\n",
       "IKZF1::IRF8      17  106  57.733333    56.0  31.812995  GM12878\n",
       "EBF1::ZNF816      1  102  53.466667    55.0  32.902164  GM12878\n",
       "IRF8::ZBTB49      5  110  58.285714    53.5  38.433616  GM12878\n",
       "HBP1::RORA       25  108  60.000000    52.0  28.049064    GDSD6\n",
       "HMGA1::LEF1      23   86  53.538462    52.0  20.783068       MM\n",
       "FOXM1::PATZ1      1   88  49.650000    52.0  22.513797    Colon\n",
       "RFX2::THAP1       7   95  50.000000    52.0  33.819373    Colon\n",
       "KLF4::TFE3        1  113  48.529412    51.0  42.649909    GDSD6\n",
       "SOX10::SP4        8   99  48.333333    51.0  33.373785       MM\n",
       "EN2::STAT1        4  108  50.850000    50.0  31.529894       MM\n",
       "FOXM1::KLF15      1   96  48.571429    48.5  27.941462    Colon\n",
       "STAT1::TBX15      1   95  47.100000    48.5  31.560969       MM\n",
       "IRF8::STAT1       4   91  50.100000    48.5  33.234256  GM12878\n",
       "FOXM1::SP4       11  107  53.666667    48.0  29.834464    Colon\n",
       "IKZF1::PAX5       5  106  54.000000    48.0  37.307841  GM12878\n",
       "FOXM1::SP2        7  110  49.000000    47.0  29.484459    Colon\n",
       "ASCL2::KLF4       9   97  45.000000    46.5  28.183609    GDSD6\n",
       "RORA::SMAD3       1  113  56.823529    46.0  35.560222    GDSD6\n",
       "NFIC::RORA        6   92  49.150000    46.0  28.897778    GDSD6\n",
       "HMGA1::RORA       4   98  44.500000    45.5  29.421618    GDSD6\n",
       "FOS::RORA         1  116  51.100000    44.5  36.898010    GDSD6\n",
       "KLF5::MAFK        9   94  53.529412    44.0  28.985164    GDSD6\n",
       "FOXM1::WT1       12  113  49.210526    44.0  28.481142    Colon\n",
       "MSX2::THAP1       9   95  47.466667    43.0  31.656339    Colon\n",
       "KLF15::ZFP28      3   76  40.600000    42.5  24.917548    Colon\n",
       "JUN::RORA         1  115  46.944444    42.5  35.659290    GDSD6\n",
       "NFE2L1::RORA      4   86  41.600000    42.5  27.350358    GDSD6\n",
       "HMGA1::STAT1      3   67  36.600000    42.5  22.276728       MM\n",
       "OSR2::THAP1       3   89  43.300000    39.5  26.870254    Colon\n",
       "SNAI2::THAP1      7   62  38.150000    38.5  18.316731    Colon\n",
       "IRF8::SPIB        1   97  42.812500    38.5  31.471085  GM12878\n",
       "EGR2::SNAI2      19   70  42.200000    38.0  15.381465    GDSD6"
      ]
     },
     "execution_count": 1022,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_alltissue_df_btn_dist_stat = vocab_instances20_alltissue_df.groupby('vocab_pair').agg({'btn_dist':['min','max','mean','median','std'],'tissue':lambda x: list(set(x))[0]})\n",
    "vocab_instances20_alltissue_df_btn_dist_stat.columns = ['min','max','mean','median','std','tissue']\n",
    "vocab_instances20_alltissue_df_btn_dist_stat.sort_values('median',ascending=False)[:50]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1024,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    239.000000\n",
       "mean      48.251046\n",
       "std       39.637058\n",
       "min        1.000000\n",
       "25%       10.500000\n",
       "50%       38.000000\n",
       "75%       90.000000\n",
       "max      117.000000\n",
       "Name: max, dtype: float64"
      ]
     },
     "execution_count": 1024,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_alltissue_df_btn_dist_stat['max'].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1025,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7fcfd1ae1190>"
      ]
     },
     "execution_count": 1025,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXYAAAD7CAYAAAB+B7/XAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAgAElEQVR4nO3de3RU5b038O/ec81kJveZSSCAQiggEsCiRmqD2Bo0F8GWc0rpK21ZTWurZa2sLn2tspaXVmutlVWr7VFWjy5baPH0oDbt2xCrR9uaHCpUCVRALnLJhZlJJrfJ3Gfv949JxgRCZpLMZGb2fD9rZZGdeTLze9jwzZNnnv1sQZZlGUREpBhisgsgIqL4YrATESkMg52ISGEY7ERECsNgJyJSGAY7EZHCMNiJiBRGnewCAKC3dwiSNHPL6QsLjejpcc3Y6yWCEvoAsB+pRAl9AJTRj2h9EEUB+fnZl308JYJdkuQZDfaR10x3SugDwH6kEiX0AVBGP6bTB07FEBEpDIOdiEhhGOxERArDYCciUhgGOxGRwjDYiYgUhsFORKQwKbGOPVUFJcAXCE7YRqdRQ80fj0SUQhjsE/AFgnjvqG3CNtcusUKt418jEaUOjjWJiBSGwU5EpDAMdiIihWGwExEpDIOdiEhhGOxERAoTU7A3NjaiuroaVVVV2LVr1yWPP/vss1i7di3Wr1+P9evXj9uGiIhmRtQF2DabDTt27MDevXuh1WqxadMmXH/99SgrK4u0OXLkCJ5++mmsXLkyocUSEVF0UUfsLS0tqKioQF5eHgwGA9atW4empqYxbY4cOYLnn38edXV1ePTRR+Hz+RJWMBERTSxqsNvtdpjN5sixxWKBzfbJ1ZhDQ0NYsmQJ7r33Xrz66qsYGBjAL37xi8RUS0REUUWdipEkCYIgRI5lWR5znJ2djZ07d0aOt27digceeAANDQ0xF1FYaIy5bbyYzaaobWSnGyajfsI2BoMO5gJDvMqalFj6kA7Yj9ShhD4AyujHdPoQNdiLi4tx4MCByLHD4YDFYokcd3Z2oqWlBRs3bgQQDn61enJ7p/T0uGb05rNmswkOx2DUdm5fEIMu78Rt3D44QqF4lRazWPuQ6tiP1KGEPgDK6Ee0PoiiMOGAOOpUzOrVq9Ha2gqn0wmPx4Pm5mZUVlZGHtfr9fjJT36C8+fPQ5Zl7Nq1C7fccssku0FERPESNditVisaGhqwZcsWbNiwAbW1tSgvL0d9fT0OHz6MgoICPProo/j2t7+NW2+9FbIs4+tf//pM1E5EROMQZFmeuTmQy0jVqZghX2zb9mYnYdteJfy6CbAfqUQJfQCU0Y+ET8UQEVF6YbATESkMg52ISGEY7ERECsNgJyJSGAY7EZHCMNiJiBSGwU5EpDAMdiIihWGwExEpDIOdiEhhGOxERArDYCciUhgGOxGRwjDYiYgUhsFORKQwDHYiIoVhsBMRKQyDnYhIYRjsREQKw2AnIlIYBjsRkcIw2ImIFIbBTkSkMAx2IiKFYbATESkMg52ISGEY7ERECsNgJyJSGAY7EZHCxBTsjY2NqK6uRlVVFXbt2nXZdm+//TZuvvnmuBVHRESTp47WwGazYceOHdi7dy+0Wi02bdqE66+/HmVlZWPadXd348c//nHCCiUiothEHbG3tLSgoqICeXl5MBgMWLduHZqami5pt337dtxzzz0JKZKIiGIXNdjtdjvMZnPk2GKxwGazjWnz8ssv46qrrsLy5cvjXyEREU1K1KkYSZIgCELkWJblMccfffQRmpub8dJLL+HChQtTKqKw0Dil75sOs9kUtY3sdMNk1E/YxmDQwVxgiFdZkxJLH9IB+5E6lNAHQBn9mE4fogZ7cXExDhw4EDl2OBywWCyR46amJjgcDnzxi19EIBCA3W7H5s2bsXv37piL6OlxQZLkSZY+dWazCQ7HYNR2bl8Qgy7vxG3cPjhCoXiVFrNY+5Dq2I/UoYQ+AMroR7Q+iKIw4YA46lTM6tWr0draCqfTCY/Hg+bmZlRWVkYe37ZtG/bt24fXX38dL7zwAiwWy6RCnYiI4itqsFutVjQ0NGDLli3YsGEDamtrUV5ejvr6ehw+fHgmaiQiokmIOhUDAHV1dairqxvztZ07d17SrrS0FG+99VZ8KiMioinhladERArDYCciUhgGOxGRwjDYiYgUhsFORKQwDHYiIoVhsBMRKQyDnYhIYRjsREQKw2AnIlIYBjsRkcIw2ImIFIbBTkSkMAx2IiKFYbATESkMg52ISGEY7ERECsNgJyJSGAY7EZHCMNiJiBSGwU5EpDAMdiIihWGwExEpDIOdiEhhGOxERArDYCciUhgGOxGRwjDYiYgUhsFORKQwDHYiIoWJKdgbGxtRXV2Nqqoq7Nq165LH33jjDdTV1aGmpgb3338//H5/3AslIqLYRA12m82GHTt2YPfu3XjttdewZ88enDx5MvK42+3Go48+ihdffBF/+tOf4PP58Oqrrya0aCIiuryowd7S0oKKigrk5eXBYDBg3bp1aGpqijxuMBjw1ltvoaioCB6PBz09PcjJyUlo0UREdHlRg91ut8NsNkeOLRYLbDbbmDYajQbvvPMObrrpJvT29uLGG2+Mf6VERBQTdbQGkiRBEITIsSzLY45HrFmzBvv378fTTz+Nhx9+GD/96U9jLqKw0Bhz23gxm01R28hON0xG/YRtDAYdzAWGeJU1KbH0IR2wH6lDCX0AlNGP6fQharAXFxfjwIEDkWOHwwGLxRI57uvrw5EjRyKj9Lq6OjQ0NEyqiJ4eFyRJntT3TIfZbILDMRi1ndsXxKDLO3Ebtw+OUChepcUs1j6kOvYjdSihD4Ay+hGtD6IoTDggjjoVs3r1arS2tsLpdMLj8aC5uRmVlZWRx2VZxr333ovOzk4AQFNTE6655prJ9IGIiOIo6ojdarWioaEBW7ZsQSAQwMaNG1FeXo76+nps27YNy5Ytww9+8AN861vfgiAIKCsrwyOPPDITtRMR0TgEWZZnbg7kMlJ1KmbIF8R7R20Ttrl2iRXZuqg/H+NOCb9uAuxHKlFCHwBl9CPhUzGZyuZ04+U/H0Ofy5fsUoiIJmXmh5pp4s/7z+G9Y3aoRAGfXmTGorl5464GIiJKNQz2cfgDIbx3zI6r5xfAOeDDP47aEZJkLL2yINmlERFFxamYcXxwshseXxA3rZyNz316Ngpz9Th7Ib3n7IgoczDYx9Fy5ALyTTosLA1Pv8wuykZPvxc+/8yvVycimiwG+0X6h/w4ctqJG5YWQxTDc+qzirIhA+hyupNbHBFRDBjsF9n/rwuQZBmrry6OfK0oVw+NWkRn91ASKyMiig2D/SIHP3JgntWEWUXZka+JooCSQgM6u4eQAsv+iYgmxGAfRZZlnLe7UDY795LHZhVlw+0Non+INxEhotTGYB+lZ8ALrz+E2ebsSx4bGcFzOoaIUh2DfZQORzi0xwt2Y5YGOdlaBjsRpTwG+yjtDhcAYHbR+HswlBQaYO/1QOI8OxGlMAb7KB3dQyjI0cGgH/+C3MIcPYIhGYNDgRmujIgodgz2UdrtQ5cdrQNAfo4OANA7OPHNN4iIkonBPiwYknDBOYTScebXR+QZtRAEwDnIHR+JKHUx2IfZez0IhuRx3zgdoRJF5GZr0TvAYCei1MVgHzbyxmmpeeIbaxfk6DliJ6KUxmAf1uEYgiCEV75MpMCkg8cXhMcXnKHKiIgmh8E+rN3hgjXfAI1aNWG7T95A5aidiFITg31YR/fEb5yOyDcx2IkotTHYAfj8ITh6PZgdZX4dAPRaNQx6NZwDXPJIRKmJwQ7ggtMNGRizo+NE8k06jtiJKGUx2AF093sAAJa8rJjaF5h06B/yIxSSElkWEdGUMNgBOPrC0ypFefqY2hfk6CHLQJ+LW/gSUephsANw9Htg0KmRrdfE1H7kDVSuZyeiVMRgB9Dd5415tA4ARoMGKlFAv4vBTkSph8GO8By7OTe2+XUAEAUBuUYt30AlopSU8cEuyTIcfV6YY3zjdESeUcc5diJKSRkf7P0uP4IhaVJTMQCQN7y1gNvLvdmJKLVkfLCPLHUsmsRUDADkG7UAgK4ed9xrIiKajowPdkdfONjNkx2xG8MrY7p4D1QiSjExBXtjYyOqq6tRVVWFXbt2XfL4X/7yF6xfvx633347vvOd76C/vz/uhSZK98ga9tzJBbtBr4ZGLaKTI3YiSjFRg91ms2HHjh3YvXs3XnvtNezZswcnT56MPO5yufDwww/jhRdewB/+8AcsWrQIP//5zxNadDw5+j3IM2qj7up4MUEQkGfUoquHI3YiSi1Rg72lpQUVFRXIy8uDwWDAunXr0NTUFHk8EAjgoYcegtVqBQAsWrQIXV1dias4zsJr2Cc3vz4iz6hDV/cQZFmOc1VERFOnjtbAbrfDbDZHji0WC9ra2iLH+fn5uOWWWwAAXq8XL7zwAu68885JFVFYGH1XxXgzm00AgJ5BH5YtKIwcjyY73TAZLz9FU1yUjRPt/dDotcjPmdxUTjyMV3M6Yj9ShxL6ACijH9PpQ9RglyQJgiBEjmVZHnM8YnBwEHfffTcWL16MO+64Y1JF9PS4IEkzN+o1m01wOAYRDEno6fPApFfD4Ri8pJ3bF8Sg6/Lb82ZpwtM3h47bsPSKgoTVO56RPqQ79iN1KKEPgDL6Ea0PoihMOCCOOhVTXFwMh8MROXY4HLBYLGPa2O12bN68GYsWLcJjjz0WS90poWfACxmTX+o4Ind4yWOHg/PsRJQ6ogb76tWr0draCqfTCY/Hg+bmZlRWVkYeD4VCuOuuu3DbbbfhwQcfHHc0n6pGVsRMdqnjiCydGsYsDTqGb4RNRJQKok7FWK1WNDQ0YMuWLQgEAti4cSPKy8tRX1+Pbdu24cKFC/jwww8RCoWwb98+AMDVV1+dFiP3T9awT23EDoRvznHezmAnotQRNdgBoK6uDnV1dWO+tnPnTgDAsmXLcOzYsfhXNgMc/R6oRCFysdFUlFqM+OsHnQiGJKhVGX+9FxGlgIxOop5+Lwpz9BDFqU8flZqzEQxJuMALlYgoRTDYJ3nF6cVKLeF3ps/a0vtdeCJSjowO9u6B6Qe7Nd8ArVrEORvn2YkoNWRssAeCIfS7/Cia5oVFoiig1GLEOY7YiShFZGywOwfCdz+a7ogdAOZZTThnd3FrASJKCRkb7N39U9vVcTxzrUZ4fEE4+i9/lSoR0UzJ2GDvGQiHcGEc9niZaw3v6XDuAqdjiCj5MjbYu/u9EAUB+TlTX8M+otScDVEQuDKGiFJCxgZ7T78H+SYtVOL0/wo0ahVmFRm4MoaIUkIGB7s3LtMwI+ZaTVwZQ0QpIXODfcCLwinu6jieuVYT+of86B30xe05iYimIiODPRiS4Bz0xWWp44iFpbkAgBPtfXF7TiKiqcjIYO/p90KW47PUccRcqxF6rQrHzzHYiSi5MjLY7c7whl3xHLGrRBELS/Nw7Fxv3J6TiGgqMjPYe8PBPt3tBC62aG4eunrcGBjyx/V5iYgmI0ODPXyDjYJ4B/ucPADAR+c5HUNEyZOZwe50I9eohUYd3+7PKzZBp1FxOoaIkiozg73XHfdpGABQq0SUlebiOEfsRJREGRvs8XzjdLTFc/PQ4RjCoJvz7ESUHBkX7JIko7vPk7BgXzQnHwDn2YkoeTIu2J0DXgRDMix58bvqdLQrSkzQaVU4dKonIc9PRBRNxgW7vS+8IiZRwa5WiVj1KTMOHrcjEAwl5DWIiCaSscFuzk9MsANAxdXF8PhC+OAkR+1ENPMyLtgdvR6oVQIKTImZYweAJXPzkWfUovXIhYS9BhHR5WRcsNv7PLAWGCCKQsJeQxQFVCwtxuHTPVwdQ0QzLuOC3dHrQXFhdsJf54alxQhJMv5x1J7w1yIiGi2jgl2WZTj6PSiZgWCfYzGi1GzEu4e7IMtywl+PiGhERgW7yxOAxxdCcVHigx0Abr5mNs5cGMTB444ZeT0iIgBQJ7uAmTSyIiaeI3ZBFDDkC4772KcXW/DmP9vx2zdPYMmVhcjWqSZ8rqAE+ALjP9doOk1GnTYimqSMSgjH8K6OxYWGuD2nLxDCoY8uPyK/+soC7PvHefyp5WP8+9qyKM8VxHtHbVFf89ol1knXSUSZI6apmMbGRlRXV6Oqqgq7du26bLv77rsPe/fujVtx8TYyYrfOwBz7CGuBAVeUmPCXA+fR0T00Y69LRJkrarDbbDbs2LEDu3fvxmuvvYY9e/bg5MmTl7S56667sG/fvoQVGg+OXg/yTTroNBNPicTbqkVmZOnUeOq376OT4U5ECRY12FtaWlBRUYG8vDwYDAasW7cOTU1NY9o0Njbic5/7HG677baEFRoP9j4PzAnaSmAiBr0G2zYuhwzgyd++j3aHa8ZrIKLMEXWO3W63w2w2R44tFgva2trGtPnGN74BADh48OCUiigsNE7p+yarZ8CLaxaF56fNZlPU9rLTDZNx4itUNRp11DYAMH9OPp64+0Zs/4938cOXD+L/3LoYt1cugGrUhVKxvB4AGAw6ALH1IR2wH6lDCX0AlNGP6fQharBLkgRBGBU+sjzmOB56elyQpMSu9fYFQnAO+GDKCnfZ4RiM+j1uXxCDLu+EbQKB6G0AwO32IVunxoN3rsKv9x3Hfzb+C/9z4By+ftsSlFqMMb/eyHOhwBBTH1Kd2WxiP1KEEvoAKKMf0fogisKEA+KoUzHFxcVwOD5Z9eFwOGCxWCZZZvI5EryrY6zyTTp894vLcNf6peju9+KRl97Da387jUBQSmpdRKQcUYN99erVaG1thdPphMfjQXNzMyorK2eitrgaWepoSeCujrESBAHXLbHih9+4HtctseAP757BT377PveVIaK4iBrsVqsVDQ0N2LJlCzZs2IDa2lqUl5ejvr4ehw8fnoka46KzJ7wapbggfmvYJ2PkQqbRH6JKxOaqRfh69RKctQ3iqd3vo3fQl5T6iEg5YrpAqa6uDnV1dWO+tnPnzkvaPfHEE/GpKgE6HEMozNEhS5eca7KiXch0y7Vz8Pe2Luzbfw63VsxFnlE3g9URkZJkzF4x7Q4XZptnZvXNVBTl6nHPxnKoVALePNAOtzf61gJEROPJiGAPhiR09bgx2zxzV5xORWGuHjdfUwpfIIQ3D7bzDVUimpKMCHab042QJKM0hUfsIwpz9bhp5Wz0Dfqw/8Po+8YQEV0sI4J9ZI+W2TO0Xe90zSrKxvKyQpzuHMCpjv5kl0NEaSYjgr3d4YIoCDNyg414uXpBIaz5Wdj/oQ0DQ1wGSUSxy4hg73AMwVqQBY06fborCgJuXF4CURTwt0NdCb8yl4iUI32SbhraHa60mF+/WLZegxuWFqNnwIvDp3uSXQ4RpQnFB7vPH4Kjz5vyK2IuZ16xCVeWmNB2qgc9/dH3kSEiUnywj7xxmo4j9hHXXWWFXqvG3w93IRTiEkgimpjyg3147/N0HbEDgE6jwuqri9Hv8uP9E93JLoeIUpzig73dMQStWkzKDTbiabY5G5+ak4cPz/TiZHtfssshohSm+GA/bx/ErKJsiHHeQz4ZPr3IDJNBg1/vOw63N5DscogoRSk62IMhCac7B1A2OzfZpcSFRi3iM8uK0Tvow7P/dQiyzCWQRHQpRQf7x10D8AclLJqbn+xS4saSb0DdZ67E3z7owP/737PJLoeIUlBy9rCdIcfOheeiF83NS3Il8fX5VaXoG/Jj7zunMbvIiBULi5JdEhGlEEWP2I+f60Wp2QhjlibZpcSVIAj47r+vwNxiE/7j9SM4wouXiGgUxQZ7MCThZHs/Fs9T1mh9hF6rRsO/LUdxgQE/+30bDhyzJ7skIkoRig32053h+fXFCppfv1hOthb3bV6JK0ty8MvXj2DvX09xD3ciUm6wHz/XCwHAp+Yoc8Q+wqDX4HtfWoEblhbjjy1n8chL7+FfHzu5YoYogyn2zdNj5/pQalHe/Pp4dFoVvlF7Fa5bYsXL+47hp3s+QEmhAWuWz8LS+YWYVWiAoIB1/EQUG0UGezAk4VRHPypXzEp2KQkhiALsTjfcvrH3RV1QmovtX70W73/kwN8OdeJ3b50E3jqJHIMGV5TkYI7FGPmw5hsgigKCEuALRL+/qk6jRhrtekyU0RQZ7G2neuAPSlh6RUGyS0kIXyCEo2ftGHSNv9ujIACVK2bha6W5+OhsH0629+G83YUjHzsj+7pr1CJKCrMx25wNAeG7Nhn0l//ncO0SK9Q6Rf5zIVIcRf5Pffv9DuSbdLh6vjKDPVYmgxZqlYDF8/KxeF4+QpKEPpcffYM+OAd86B304dDJbri94RF7QY4OV5bk4MqSnAlDnohSm+L+99r7PPjXx07UfeYKqETOHYymEkUU5uhRmKPHgtnhr5UvLMLbB9vR0T2EczYXDh534J/HHSgpMmD+rFzMsRjT6s5TRKTAYP/rB52AAFQuV+b8erwJgoCCHD0KcvRYNr8Q/S4fTncO4HTnAP7e1gW1SsC8YhNysrVYsaAIosg3YYlSnaKCPRiS8Pe2TqwoK0JBjj7Z5aSlXKMOKz9lxoqFRbD1enC6cwBnuwbx7H8fhsmgwYqyIpQvKMKiuXkZseKIKB0pKtgPHndgwB3ATStnJ7uUtCcIAooLDCguMOC6JRboNGp8eMaJA8ft+FtbFwBgdlE25lqNmGMxYY41vNomx6BNcuVEpJhgH3T78bu3TmBWUTaWXpnZb5rGm1ol4ppFZny2vATBkISPuwZw7GwvTnYM4OjZXrT+yxZpm5OtxaxCA0qKslFSmI2SIgNKCrKh06oibbh0kiixFBHssizjP/90FEOeABr+bbkibqqRagRRwNDwuvlZZiNmmY24eVX4sUG3Hx2OIbQ7hnDkVDccfR6caO9HSPrk6ldjlgZ5Ri3yTDpc8ykzZhVmI9eohc6gg8cXhCgI8PqDcPuCcHuDGPQEMeD2YcgbhNsTCP/pDcA/vGWCIAgw6FQwZmmRk62FOU8Pa4EB+SbdJRdj8QcJZRpFBPtfDrTj0KkebP78Qsy1mpJdjiL5AiEc+sgxYZvKlbOQmx2ed5dkGS53AH0uX2SJZZ/Lh47uIRw57Zz062tUIrQaEWqVCAiALAP+QAiBoDTmB0iWTg1Lnh7m/CxY8rKQn6NHxdJirsGnjBLTv/bGxkb88pe/RDAYxFe/+lV85StfGfP40aNH8eCDD2JoaAirVq3CI488ArU68f+RfIEQfv8/p/DmP9uxoqwIn/t0acJfk2IjCgJyssOj6bnWT74ekmTMs5rg9QfR7/JD1KjQ3++FLMvQ69TI0qlg0GkgigJOdvRBp1FBp1FddjVO+cIi/O/hCxh0+9E76IOjzwNHnxdnbeGbmKtVAg4es2PplQVYPDcfV5SYuAyWFC9q+tpsNuzYsQN79+6FVqvFpk2bcP3116OsrCzS5t5778UPf/hDrFixAg888ABeeeUVbN68OaGFuzwBPPbrg7A53fj8qlJsXLOA+6GkAZUooKQoG9nDI2iz2QSHY/CSdkO+ILr7PVGfTxAEGPRqGPRqWAsMWDwvvJun2xuEo88Dm9ON/iE//vud0wDC++rML8nBFcUmzCs2objAAGuBATqNaqKXiTtJkuH1h+DxBeHxB+HxBeEPSpAlGZIcnl6UZBkCBAgCIIrDfwoCBEGAKIT7PvJ1WQr/liRJ4e+LfD7q60D4eURBgCiO+nz4eXqGAujv90DGyPeOPF/4B/InzylDBiBgpAZEahqpTxiuLxSS4AuEhj8k+PwheP1BeEf96fGN/jz8pz8Ygixj+CNcuyTLwPAvZxqNCK1aBZ1GhFajGvN5rkkPETKydGoYdGpkjfrQqEWoxOFaRSEyYBjZM0+WZUReRg73U0a4X1k6FbK06vAARKuCRi2mbOZEDfaWlhZUVFQgLy+8S+K6devQ1NSEe+65BwDQ0dEBr9eLFStWAAC+8IUv4JlnnplUsE9lbXRIkrCwNBf1tUtQVjr5HRxjeU21SoRBP/GSvljaJOK5snRqhILTf76k1K9RwTc8V97d54l8PpooTq8ug16DorwsLLmiAMvLihAavv/tqc5+nLe58MHJbhwcNbWUpVUj26BBdpYGRn04EDRqESqVCLUoQqUK/5sZCZpP/vOHv1+n18Dj9kMGEAxK8AXDQRYIhuAPfPK5LyDB6wvCFwhF7ZuSaTUqaDXhMNZpVCjM1UOjUUGnVkGnFaFRqxDOzOEfEgh/qFQiQpKMUEiCPxhCICgjEAz/3QZCEgIBCYNuH9ze4R8QCfx7VolC5DdKrVYFrVqEVi2G+6EJB79OrYr8oAxJMmRZhi8oo9/lw+CQH+sr5+OqeeNvLT5RRkXLr6jBbrfbYTabI8cWiwVtbW2XfdxsNsNms2Ey8vOzJ9UeAAoLjfi/V0z9lnCFhcaY2pWWRL8R9vzS2PZ8j6VdPJ8rGa8Z63ONkZc17pfnzortJuSxvua80nysjbkoouSKNaPGE3WyUZKkMb9uyLI85jja40RENLOiBntxcTEcjk9+ZXU4HLBYLJd9vLu7e8zjREQ0s6IG++rVq9Ha2gqn0wmPx4Pm5mZUVlZGHp89ezZ0Oh0OHjwIAHj99dfHPE5ERDNLkGO4h1pjYyOef/55BAIBbNy4EfX19aivr8e2bduwbNkyHDt2DNu3b4fL5cLSpUvxox/9CFotLy0nIkqGmIKdiIjSB6/UICJSGAY7EZHCMNiJiBSGwU5EpDAZFeyNjY2orq5GVVUVdu3alexypuzOO+9ETU0N1q9fj/Xr1+PQoUPJLilmLpcLtbW1aG9vBxDesqKurg5VVVXYsWNHkquL3cX9+P73v4+qqqrIOXnjjTeSXOHEnn32WdTU1KCmpgZPPvkkgPQ8F+P1I93Oxc9+9jNUV1ejpqYGL774IoA4nAs5Q1y4cEFeu3at3NvbKw8NDcl1dXXyiRMnkl3WpEmSJN94441yIBBIdimT9sEHH8i1tbXy0qVL5fPnz8sej0des2aNfO7cOTkQCMhbt26V33777WSXGdXF/ZBlWa6trZVtNluSK4vNu+++K3/pS1+SfT6f7Pf75S1btsiNjY1pdy7G60dzcz367RsAAANBSURBVHNanYv9+/fLmzZtkgOBgOzxeOS1a9fKR48enfa5yJgR++jNzAwGQ2Qzs3Rz+nR4l8KtW7fi9ttvx29+85skVxS7V155BQ899FDkyuS2tjbMmzcPc+bMgVqtRl1dXVqck4v74fF40NnZiQceeAB1dXV45plnIEmXbmyWKsxmM+6//35otVpoNBosWLAAZ86cSbtzMV4/Ojs70+pcXHfddXj55ZehVqvR09ODUCiEgYGBaZ+LjAn28TYzm+xmZalgYGAAN9xwA5577jm89NJL+N3vfod333032WXF5LHHHsOqVasix+l6Ti7uR3d3NyoqKvD444/jlVdewYEDB/D73/8+iRVObOHChZHdWM+cOYM///nPEAQh7c7FeP347Gc/m1bnAgA0Gg2eeeYZ1NTU4IYbbojL/4uMCXalbFa2cuVKPPnkkzCZTCgoKMDGjRvxzjvvJLusKVHKOZkzZw6ee+45WCwWZGVl4c4770yLc3LixAls3boV9913H+bMmZO252J0P+bPn5+W52Lbtm1obW1FV1cXzpw5M+1zkTHBHm0zs3Rx4MABtLa2Ro5lWZ6Ru1UlglLOyfHjx7Fv377IcTqck4MHD+JrX/savve97+GOO+5I23NxcT/S7VycOnUKR48eBQBkZWWhqqoK+/fvn/a5yJhgj7aZWboYHBzEk08+CZ/PB5fLhVdffRW33HJLssuakuXLl+Pjjz/G2bNnEQqF8Mc//jEtz4ksy3j88cfR39+PQCCAPXv2pPQ56erqwt13342nnnoKNTU1ANLzXIzXj3Q7F+3t7di+fTv8fj/8fj/efPNNbNq0adrnInV/lMWZ1WpFQ0MDtmzZEtnMrLy8PNllTdratWtx6NAhbNiwAZIkYfPmzVi5cmWyy5oSnU6HJ554At/97nfh8/mwZs0a3Hrrrckua9IWL16Mb37zm/jyl7+MYDCIqqoq1NbWJrusy/rVr34Fn8+HJ554IvK1TZs2pd25uFw/0ulcrFmzBm1tbdiwYQNUKhWqqqpQU1ODgoKCaZ0LbgJGRKQwGTMVQ0SUKRjsREQKw2AnIlIYBjsRkcIw2ImIFIbBTkSkMAx2IiKFYbATESnM/wdtrx7cV6VyzgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.distplot(vocab_instances20_alltissue_df_btn_dist_stat['min'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1026,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7fcfd1a911d0>"
      ]
     },
     "execution_count": 1026,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYIAAAD7CAYAAABnoJM0AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAgAElEQVR4nO3de1hU57n4/e8MMxwGhvMMKAoe8BQUicFITApJtgUVCQk/sxtjQn7Nrsa8u9WL9y17a0ztL6m5rNmkutukXkl2dvo2r261SQqlVTQxMYnBRiEaQdGIZwGH4aByZoZZ7x82kxDFAR1gDvfnurxkzbPWM8/NGuZe61lrPY9KURQFIYQQXks93A0QQggxvCQRCCGEl5NEIIQQXk4SgRBCeDlJBEII4eUkEQghhJeTRCCEEF5OM9wNuBXNzW3YbAN7/CEiIojGxtZBatHQ8qRYwLPikVhckyfFAgOPR61WERYW2Ge5WyYCm00ZcCL4ZjtP4UmxgGfFI7G4Jk+KBZwbj3QNCSGEl5NEIIQQXk4SgRBCeDlJBEII4eUkEQghhJeTRCCEEF5OEoEQQng5t3yOQHzLaoMui9Xp9fppNWjkMEEIryCJwM11WawcrDI5vd6ZU6LQ+MnHQwhv0K9jvuLiYubPn096ejqbN2++rryqqoqcnBwyMjJYvXo1VmvvI9Rjx44xdepU+3J3dzf5+fnMmzePRx55hFOnTt1mGEIIIW6Vw0RgMpnYsGEDW7ZsobCwkG3btlFdXd1rnfz8fNasWcOuXbtQFIXt27fbyzo6OvjVr36FxWKxv/bOO+8QEBDAzp07ee6551i1apUTQxJCCDEQDhNBaWkpKSkphIaGotPpyMjIoKSkxF5eU1NDZ2cnSUlJAOTk5PQq//Wvf81TTz3Vq869e/fy0EMPATBz5kyampqora11SkBCCCEGxmEncH19PQaDwb5sNBo5cuRIn+UGgwGT6Vqf9Z49e+js7GTu3Lk3rdNgMHDp0iVGjhzZr0ZHRAT1a73vMxj0t7SdK/omFqWpHX2Qv9Pr1+n8MITrnF5vXzxx33gCicV1OTMeh4nAZrOhUqnsy4qi9Fruq9xsNrNp0yb+8Ic/XFfn9+tQFAW1uv+3qDQ2tg545D2DQY/Z3DKgbVzVd2Np77LS0trp9Pdob+/C3NPj9HpvxFP3jbuTWFzXQONRq1U3PYB2+O0bHR2N2Wy2L5vNZoxGY5/lDQ0NGI1G9u7dy+XLl1m8eDHZ2dkAZGdn09raSlRUFPX19ddtI4QQYug5TASzZ89m//79NDU10dHRwe7du0lNTbWXx8TE4OfnR3l5OQBFRUWkpqby6KOP8uGHH1JUVERRUZG9LCgoiLS0NPtrZWVl+Pn59btbSAghhHM5TARRUVHk5eWRm5vLww8/zIIFC0hMTGTJkiVUVFQAUFBQwLp165g7dy7t7e3k5ubetM4nn3yS7u5uMjMzeemll3j55ZedE40QQogBUymK4nbT9sg1gm9jaesavAfKAofogTJP3TfuTmJxXUN+jUAIIYRnk0QghBBeThKBEEJ4OUkEQgjh5SQRCCGEl5NEIIQQXk4SgRBCeDlJBEII4eUkEQghhJeTRCCEEF5OEoEQQng5SQRCCOHlJBEIIYSXk0QghBBeThKBEEJ4OUkEQgjh5fqVCIqLi5k/fz7p6els3rz5uvKqqipycnLIyMhg9erVWK1W4No0lDk5OWRlZbFs2TKuXLkCwIEDB5g1axbZ2dlkZ2ezatUqJ4YkhBBiIBwmApPJxIYNG9iyZQuFhYVs27aN6urqXuvk5+ezZs0adu3ahaIobN++HYBVq1bx8ssvU1xcTHx8PG+99RYAlZWVPP300/b5jNetWzcIoQkhhOgPh4mgtLSUlJQUQkND0el0ZGRkUFJSYi+vqamhs7OTpKQkAHJycuzlO3bsID4+HovFgslkIjg4GICKigr27dtnP1Ooq6sbjNiEEEL0g8NEUF9fj8FgsC8bjUZMJlOf5QaDwV6u1Wo5ceIEaWlpfPHFF2RmZgKg1+t58sknKS4uJi0tjby8PKcFJIQQYmAczk5us9lQqVT2ZUVRei07Kp80aRKlpaVs3bqVvLw8tm7dyosvvmgvX7RoEa+88gotLS3o9fp+NfpmkzDfjMHQv/rdwTexKE3t6IP8nV6/TueHIVzn9Hr74on7xhNILK7LmfE4TATR0dGUlZXZl81mM0ajsVe52Wy2Lzc0NGA0Gunq6uKzzz5jzpw5ADz00EOsX78em83G66+/ztKlS/Hx8bFv992fHWlsbMVmU/q9Plz7pZnNLQPaxlV9N5b2ListrZ1Of4/29i7MPT1Or/dGPHXfuDuJxXUNNB61WnXTA2iHXUOzZ89m//79NDU10dHRwe7du0lNTbWXx8TE4OfnR3l5OQBFRUWkpqai0Wh44YUXqKysBGDnzp3MmDEDtVrNBx98wK5duwAoLCxk+vTp6HRDd/QphBDiWw7PCKKiosjLyyM3NxeLxcLChQtJTExkyZIlLF++nGnTplFQUMDzzz9Pa2srCQkJ5Obm4uPjw4YNG1izZg09PT1ERUXx0ksvAbB+/Xp+8Ytf8NprrxEeHs7LL7886IEKIYS4MZWiKAPrY3EB0jX0bSxtXVYOVpkcbDFwM6dEEejn8DjBKTx137g7icV1DXnXkBBCCM8miUAIIbycJAIhhPBykgiEEMLLSSIQQggvJ4lACCG8nCQCIYTwcpIIhBDCy0kiEEIILyeJQAghvJwkAiGE8HKSCIQQwstJIhBCCC8niUAIIbycJAIhhPBykgiEEMLL9SsRFBcXM3/+fNLT09m8efN15VVVVeTk5JCRkcHq1auxWq0AlJWVkZOTQ1ZWFsuWLePKlSsAXL16laVLlzJv3jwWL17ca85jIYQQQ8thIjCZTGzYsIEtW7ZQWFjItm3bqK6u7rVOfn4+a9asYdeuXSiKwvbt2wFYtWoVL7/8MsXFxcTHx/PWW28BsHHjRpKTk9m5cyePPvqofQpLIYQQQ89hIigtLSUlJYXQ0FB0Oh0ZGRmUlJTYy2tqaujs7CQpKQmAnJwce/mOHTuIj4/HYrFgMpkIDg4GYO/evWRlZQGwYMECPv30UywWi9ODE0II4ZjDRFBfX4/BYLAvG41GTCZTn+UGg8FertVqOXHiBGlpaXzxxRdkZmZet41GoyEoKIimpibnRCSEEGJAHM5ObrPZUKlU9mVFUXotOyqfNGkSpaWlbN26lby8PLZu3XrdeyiKglrd/+vWN5uE+WYMBv0tbeeKvolFaWpHH+Tv9Pp1Oj8M4Tqn19sXT9w3nkBicV3OjMdhIoiOjqasrMy+bDabMRqNvcq/e7G3oaEBo9FIV1cXn332GXPmzAHgoYceYv369cC1s4qGhgaio6OxWq20tbURGhra70Y3NrZisyn9Xh+u/dLM5pYBbeOqvhtLe5eVltZOp79He3sX5p4ep9d7I566b9ydxOK6BhqPWq266QG0w8Pw2bNns3//fpqamujo6GD37t2kpqbay2NiYvDz86O8vByAoqIiUlNT0Wg0vPDCC1RWVgKwc+dOZsyYAUBaWhqFhYXAtesIycnJaLXafgclhBDCeRyeEURFRZGXl0dubi4Wi4WFCxeSmJjIkiVLWL58OdOmTaOgoIDnn3+e1tZWEhISyM3NxcfHhw0bNrBmzRp6enqIioqy3x20YsUKVq5cSWZmJnq9noKCgkEPVAghxI2pFEUZWB+LC5CuoW9jaeuycrDK5GCLgZs5JYpAP4fHCU7hqfvG3UksrmvIu4aEEEJ4NkkEQgjh5SQRCCGEl5NEIIQQXk4SgRBCeDlJBEII4eUkEQghhJeTRCCEEF5OEoEQQng5SQRCCOHlJBEIIYSXk0QghBBeThKBEEJ4OUkEQgjh5SQRCCGEl5NEIIQQXq5fiaC4uJj58+eTnp7O5s2bryuvqqoiJyeHjIwMVq9ejdVqBaC8vJyFCxeSnZ3NU089RU1NDQAHDhxg1qxZZGdnk52dzapVq5wYkhBCiIFwmAhMJhMbNmxgy5YtFBYWsm3bNqqrq3utk5+fz5o1a9i1axeKorB9+3b762vXrqWoqIisrCzWrl0LQGVlJU8//TRFRUUUFRWxbt26QQhNCCFEfzhMBKWlpaSkpBAaGopOpyMjI4OSkhJ7eU1NDZ2dnSQlJQGQk5NDSUkJ3d3drFixgsmTJwMwadIk6urqAKioqGDfvn1kZWWxbNky++ti6FisNto6LbS2W+i29Ax3c4QQw8jhpLT19fUYDAb7stFo5MiRI32WGwwGTCYTvr6+ZGdnA2Cz2Xj11VeZM2cOAHq9nnnz5pGens7//M//kJeXx9atW50WlOhbZ7eVytNNnDh/mZ5/zPusVqkYO0LP5DFhRAT7D3MLhRBDzWEisNlsqFQq+7KiKL2WHZV3d3ezcuVKrFYrzzzzDAAvvviivXzRokW88sortLS0oNfr+9Xom03CfDMGQ//qdwffxKI0taMP6t+X96mLl/nw4AV6emxMiA0jxhCISqXC3NxO1dlmTtVeJWmCgXsSR6DT+WEI1w1mCL144r7xBBKL63JmPA4TQXR0NGVlZfZls9mM0WjsVW42m+3LDQ0N9vK2tjaeffZZQkND2bRpE1qtFpvNxuuvv87SpUvx8fGxb/fdnx1pbGzF9o+j2f4yGPSYzS0D2sZVfTeW9i4rLa2dDrc5b2rhk8O1RIb4M3tqNCFBfvayUZE6EsaE8eXXDRw+acZ8uZ2xUUGoeoamy8hT9427k1hc10DjUatVNz2AdniNYPbs2ezfv5+mpiY6OjrYvXs3qamp9vKYmBj8/PwoLy8HoKioyF6en59PXFwcGzduxNfX9x8NUvPBBx+wa9cuAAoLC5k+fTo63dAdfXqbi+ZWPj1cS0SwP/+UPKpXEviGr9aHlIQoZt1hpLahjVffq6BLrh0I4RUcJoKoqCjy8vLIzc3l4YcfZsGCBSQmJrJkyRIqKioAKCgoYN26dcydO5f29nZyc3M5duwYe/bs4csvv+SRRx4hOzubJUuWALB+/Xr++Mc/kpmZyXvvvWe/m0g4X3unlc++qiNU78ec5FH4am5+5jUpNoy0pJFcMLXwZvGxAZ95CSHcj0pRFLf7S5euoW9jaeuycrDKdMP1FEVh76FaahvayLp3DMGBvv1+j7YOK+99cor0maN57J8mOKXdffHUfePuJBbX5eyuIYfXCIT7Om9q5UJ9KzMmRg4oCQDcPyOGy61d7D54gfiYEJInGx1vJIRwSzLEhIfqtvTwxTET4cF+3DEm/Jbq+NGD8Ywdoef/LTlOc0uXk1sohHAVkgg81PFzzXR293BPQjRqtcrxBjeg8VGzJCsBS4+N//7bMWzu14sohOgHSQQeyGK1cexcM6MMgUSE3N4DYtHhOh57cAJHzzbz8Zc1TmqhEMKVSCLwQCfON9NtsZEYH+GU+tKSRpIwNpz3PjklXURCeCBJBB7GYrVx7GwzIyN1RIYEOKVOlUrFE+kTsfYobN1z0il1CiFchyQCD1N98Qqd3T0kjnfO2cA3osJ0LJgdx8Hj9VScbnRq3UKI4SWJwIMoisLx880YQv0xhjn/Se15s+KICtexeffXWKw2p9cvhBgekgg8iKmpg5Z2CxNHhw5K/VqNmsU/nED95Q72lF8clPcQQgw9SQQe5OuLl/HVqImLHrxRFqeOjWDauAiKS8/S0t49aO8jhBg6kgg8RGe3lfOXWhk3MhiNz+Du1n9+MJ6u7h7+su/soL6PEGJoSCLwEKdrrmJTFCYMUrfQd8VEBpKWNJKPD9VQ19g26O8nhBhckgg8gKIonLx4BUOoP2H664eYHgzZ943Fz1fN9o+qHa8shHBpkgg8QNPVLq60dTM+JmTI3jM40JcF94zhq1ONHD3bNGTvK4RwPkkEHuDspauoVBAXNbRT8c1JHkVkiD/b9lTLvAVCuDFJBG7OpiicrWthZGQgfr79n+7TGbQaHxbeP56L5lb2VdQN6XsLIZxHEoGbO1t3lbZOK2NHDM/E3DMnG4mPCeHPn56ms9s6LG0QQtyefiWC4uJi5s+fT3p6Ops3b76uvKqqipycHDIyMli9ejVW67UvhPLychYuXEh2djZPPfUUNTXXRq+8evUqS5cuZd68eSxevBiz2ezEkLzLlyfMqNUqRhn7nn3oVqjUKtq6rA7/tXf38NAPxnKlrZu/fH7W4fryQLIQrsfhDGUmk4kNGzbw/vvv4+vry2OPPcasWbOIj4+3r5Ofn8/atWtJSkriueeeY/v27Tz++OPk5+fz+9//nsmTJ/Puu++ydu1aNm3axMaNG0lOTuaNN96gsLCQl156iY0bNw5qoJ7IZlM49HUDowyBDuciHqguSw9ffd3/BB0XreeDgxcICtCg89f2ud7MKVFo/GRiPCFcicMzgtLSUlJSUggNDUWn05GRkUFJSYm9vKamhs7OTpKSkgDIycmhpKSE7u5uVqxYweTJkwGYNGkSdXXX+pH37t1LVlYWAAsWLODTTz/FYrE4PThPd+LCZa62dzNmRPBwN4UZEyNRFDh0smG4myKEGCCHh2b19fUYDAb7stFo5MiRI32WGwwGTCYTvr6+ZGdnA2Cz2Xj11VeZM2fOddtoNBqCgoJoamoiKiqqX42+2STMN2MwDE8/+mAwGPS8v+8MWo2ayWPC0Tr5jECr1aAP6v+kNvogfxLjIzl80kzylGgiQ288BLZO54ch/PoB8Txt33gKicV1OTMeh4nAZrOhUn071aGiKL2WHZV3d3ezcuVKrFYrzzzzzA3fQ1EU1Or+X7dubGwd8O2KBoMes7llQNu4KoNBT339VfYfqWVSbCidnRY6ce4ZlcVipaW1c0DbTIoN4djZRj49dJE5yaN6fQ6+0d7ehbmnp9drnrZvJBbX40mxwMDjUatVNz2AdvjtGx0d3etirtlsxmg09lne0NBgL29ra+MnP/kJVquVTZs2odVe6zs2Go00NFzrQrBarbS1tREaOvhDI3iSC/WtNF7tcvq8A7fDT+vD9PGR1DW2U9sgQ08I4S4cJoLZs2ezf/9+mpqa6OjoYPfu3aSmptrLY2Ji8PPzo7y8HICioiJ7eX5+PnFxcWzcuBFfX1/7NmlpaRQWFgKwY8cOkpOT7UlC9M+hkw2ogISxrpMIACbGhqLXaSk7YZaHzIRwEw67hqKiosjLyyM3NxeLxcLChQtJTExkyZIlLF++nGnTplFQUMDzzz9Pa2srCQkJ5ObmcuzYMfbs2UN8fDyPPPIIcO1M4M0332TFihWsXLmSzMxM9Ho9BQUFgx6opzl00sz4mBCCA30drzyEfNQqZkw08MnhWk5evMKkWDnTE8LVqRRFcbvDNm+/RqBofPiXtR/w6P3jSb0zhoNVJqe/x/SJhgHdPvpdiqKw+8AFLrd28/APxvZ64nnmlCgCv3f7qCftG4nFNXlSLDAM1wiE6zlw9BIASRMih7klN6ZSqbj7DiPdlh4OV8vtpEK4OkkEbuiLo5eIDtcxIiJwuJvSpzC9PxNjQ/n6/GWarg7s7iMhxNCSROBmOrutVJ5qZHq8a10kvpGk+Eh8tT4cqKrHDXsghfAakgjczPFzl7H22Jg2zvUTgZ+vDzMmRVLf3EF1zZXhbo4Qog+SCNzMkdONBPj5MGGUe9yNEx8TgjEsgPITZjq6ZHRSIVyRJAI3oigKFacamT7BgFbjHrtOpVKRkhCF1Wqj7Hj9cDdHCHED7vFtIgCobWyn8Wond03u35hMriI0yI+p4yI4U9fC0TMyraUQrkYSgRupONUI4HaJAGDa+HBCg3zZ8sHXtHbISLNCuBJJBG6k4nQjMYZADGE3HtnTlfmo1dybOILWDgv/3+4Tw90cIcR3SCJwEx1dVr6+cJlEN7hbqC8Rwf7MT4njQFU9Xxxz/tPQQohbI4nATZw4f5kem8JUN04EAHNmjmb8yGD+uOs4pub24W6OEAJJBG7j6Jkm/LQ+xMeEDHdTbouPWsUz2QmoVSo2/bkSi7XH8UZCiEElicBNVJ5pZFJsqNvcNnozkSEB/GTBHZyvb2XLhyeHuzlCeD33/1bxAubLHZiaO0gYGz7cTXGa6fGRzE+J45PDtfxt3+nhbo4QXk0SgRs4evbavfdTPSgRAOSkjiMpPpI3Cis4ckpGKRViuEgicANHTzcREexH9A0mfXdnarWKpQ/dwZiRIWwqOsq5S54zXrwQ7qRfiaC4uJj58+eTnp7O5s2bryuvqqoiJyeHjIwMVq9ejdXae0yZjRs38rvf/c6+fODAAWbNmkV2djbZ2dmsWrXqNsPwXD02G8fONZMwNvyGk8G7O39fDWv+ZRaB/hpe2XaYi/Wtw90kIbyOw0RgMpnYsGEDW7ZsobCwkG3btlFdXd1rnfz8fNasWcOuXbtQFIXt27cD0NLSwnPPPcfbb7/da/3KykqefvppioqKKCoqYt26dU4MybOcqWuho8vqcnMTO1NESAD5i+5E46PiP7YeokYmvhdiSDlMBKWlpaSkpBAaGopOpyMjI4OSkhJ7eU1NDZ2dnSQlJQGQk5NjL9+zZw9jxozhxz/+ca86Kyoq2LdvH1lZWSxbtoy6ujpnxuRRjp5pQqWCKXFhw92UQRUVpuPfHp+BWqXi5S1fcrr26nA3SQiv4TAR1NfXYzAY7MtGoxGTydRnucFgsJc//PDDLF26FB+fb+esBdDr9Tz55JMUFxeTlpZGXl7ebQfiqSrPNDJ2RDBBAdrhbsqgiw7X8e+LZ+Cn9eHlLV9y6OStzZkshBgYjaMVbDZbr75pRVF6LTsqv5EXX3zR/vOiRYt45ZVXaGlpQa/X96vRN5uE+WYMhv7V7ypaOyycqb3Ko3MmXtf2b5aVpnb0Qf5Of2+tVjMo9ep0fhhucNH7m3gMBj2/yUvjxbe+4LX3K1g8dwoLH5yAWu34+khLezcdnc6f8yDAX4Ne59vv9d3tc3YzEovrcmY8DhNBdHQ0ZWVl9mWz2YzRaOxVbjZ/e+TW0NDQq/z7bDYbr7/++nVnCt8/a7iZxsZWbLaBTX1oMOgxm93rrpTyE/XYFBhrDOrV9u/G0t5lpaXV+XMCWyyDU297exfmnt5PE99o3/w/j07n7Z1VvLOzikPHTfwk6w6CHXwZt3VZOVjl/DGMZk6JorOtq1/ruuPnrC8Si+saaDxqteqmB9AOu4Zmz57N/v37aWpqoqOjg927d5Oammovj4mJwc/Pj/LycgCKiop6lV/fIDUffPABu3btAqCwsJDp06ej03nWrZHOUHmmCX9fH8aNDB7upgw5P18fnnkogSczJnH8fDO/+K8v+PuxSzL3sRCDwGEiiIqKIi8vj9zcXB5++GEWLFhAYmIiS5YsoaKiAoCCggLWrVvH3LlzaW9vJzc396Z1rl+/nj/+8Y9kZmby3nvvsXbtWudE40EURaHydBNT4sLQ+Hjn4x4qlYoH7oxhzVMziQzx542/HOM/3z1CXaPcVSSEM6kUNzzE8oauIVNTO6ve+DtPpk/kgRmjepV9N5bB6g6ZPtHAV187/2LtzClRBPr17pHsz76x2RQ+LLtA4b4zdFtspN05kqzZYwgN8rOvM5hdQ99vc1/c7XN2MxKL63J211D/Pt1iyFX+Y0pHTxpf6Hao1SrS744lJSGaos/P8MmhWj77qpbZU6PJuDuWERGBw91EIdyWJAIXdfRME4ZQf4xhcu3ku4IDfXkyfRIZM0ez68AF9lXU8dlXdSRNiLzuzEkI0T+SCFyQtcdG1flm7kmIHu6muCxjmI4nMyaRfd9YPiy/yMdfXuTQyQYiQvyZHBvKmBF6fNTeeW1FiIGSROCCqi9eoau7h2nSLeRQcKAvOanjmJ8Sy0eHath94AKfV1yi/ISZCaNCmBgbSqC/5z+MJ8TtkETggipON+KjVjHZw4eVcCZ/Xw1pSTEE+PpQ19jO8fOXqTjdROWZJmKNQUyKCyMqLMAjB+4T4nZJInBBFaebmDAqhIB+3qkivqVSqRgZGcjIyEBa2rv5+sJlTl68wjlTK5Eh/kwbH8EoQ6AkBCG+QzpRXUxzSxcXza1Mc/NJ6l2BXufLXZOMLLx/PCkJUXR29/DxlzX8tfQc5y61yMNpQvyDHHK6mMozjQBMlUTgNBofNRNHhxIfE8KZuqtUnGrkk8O1hAT5ctdEA6OMtzZ2lRCeQhKBi6k83URIkC+jDHJfvLOp1SrGx4QwdmQw5+paOFzdwEdf1jAyMpCZkw2EfOfhNCG8iXQNuZAem42jZ5qYNjZC+rAHkVqlYuzIYB66byzJkwyYL3fwl8/PcrCqnm5Lj+MKhPAwckbgQs7UttDeZWXqOLltdCj4qFXcMTacsSODOXyygapzzZypu8rdU4zEReslGQuvIWcELqTidCMqlQwrMdQC/DTcMzWazNlxBAZo+fSrOvYeqqV9EOY2EMIVSSJwIZVnGhk/MkQegBomEcH+zJsVy4xJBmob2vjLvjNUX7widxcJjyeJwEVcbe/mbF2LdAsNM7VaxdSx4WTdO4ZQvR+llZfYVFhJc0v/JqYRwh1JInARR880oYA8P+AiggN9ybh7NDOnGDl18Qq/+K8v2F8pE+MIzySJwEVUnm4kKEBLXLRnzavqzlQqFVPiwvj3J+5iZGQgb/71GK/9uZKrbd3D3TQhnKpfiaC4uJj58+eTnp7O5s2bryuvqqoiJyeHjIwMVq9ejdXa+yLbxo0b+d3vfmdfvnr1KkuXLmXevHksXry415zH3simKFSeaWLq2HDUcqeKyzGGBbBy8Qz++YF4jpxq4Pn/+oKy4/XD3SwhnMZhIjCZTGzYsIEtW7ZQWFjItm3bqK6u7rVOfn4+a9asYdeuXSiKwvbt2wFoaWnhueee4+233+61/saNG0lOTmbnzp08+uijvPTSS04Myf2cN7XQ0m7xiusDKrWKti5rr3/1Te3XvXYr/wY4ad2AqNUq5s6K5Zf/eyYRIf78vrCSN/5ylNYOy+C9qRBDxOFzBKWlpaSkpBAaGgpARkYGJSUl/PSnPwWgpqaGzs5OkpKSAMjJyeG3v/0tjz/+OHv27GHMmDH8+Mc/7lXn3r177WcWCxYs4MUXX8RisdfDbYAAABlESURBVKDVeufdMhWn/jGsxFjPvz7QZem5bgpMfZA/La2dt1339ImG267DkRhDEKufvIsdfz9H8ednqTrfzI/nTSZxfOSgv7cQg8XhGUF9fT0Gw7d/YEajEZPJ1Ge5wWCwlz/88MMsXboUHx+fPuvUaDQEBQXR1NR0e5G4sYrTTcRF6QkO9B3upoh+0PioeejesTyfm0xQgJaNfzrC2zuq5LkD4bYcnhHYbLZeT1gqitJr2VF5fyiKgnoAs0ndbBLmmzEYXO9C7OWWLk7VXmHRDycNqH3frKs0taMP8nd6u7RazZDW64z3Gqw263R+GMKvnzLUYNAzfUoUW3ad4P2PT1JxpomnsxK4f8Yoj3kq2RX/Zm6VJ8UCzo3HYSKIjo6mrKzMvmw2mzEajb3Kv3uxt6GhoVf5jRiNRhoaGoiOjsZqtdLW1mbveuqPxsZWbAPsEDYY9JjNLQPaZih89lUtigITRgb3u33fjaW9y+qUbpXvs1iGrl5ndQ0NVpvb27sw9/Q9BtH8u0eTEBvKH3ed4DdbvuRvn53mifSJxBjce1RTV/2buRWeFAsMPB61WnXTA2iHh+GzZ89m//79NDU10dHRwe7du0lNTbWXx8TE4OfnR3l5OQBFRUW9ym8kLS2NwsJCAHbs2EFycrLXXh84XN1AeLAfsVHu/aXh7eKi9azOvYufPjqdi+ZW/s/bB9n20UnaOuVisnB9Ds8IoqKiyMvLIzc3F4vFwsKFC0lMTGTJkiUsX76cadOmUVBQwPPPP09raysJCQnk5ubetM4VK1awcuVKMjMz0ev1FBQUOC0gd9Jt6eHomSbuSxzhMV0J3kytUpGRMoYJI/S898lpdh24wGdf1TF3Vixzkkfh7ytjPArXpFLc8FFJT+kaOnyygd++d4T/+0fTB3TH0HdjaeuycrDK5GCLgZs+0XDd3T2DVa8z7xoajDbfnRCN0s/Pm07nR3v7teEoLppb+VvpWSpPN6HXaUmfGcu9iSPQaq6diPtpNWhc+JFOV/ybuVWeFAs4v2tIDlGG0eFqM/6+PkyOlUnqXdmNbnnty/eT2oyJBkYbgzj0dQPvfXKKnX8/x+QxYUwcFcK9iSPRyLzUwgXIp3CY2BSFw9WNTBsXgcbHhQ8LxW0zhAaQfvdo6hrbqDjVxJcnzBypbsB0uYP7k2IID3bunU6ufqYhXI8kgmFyquYKV9u6uXOCPIjkLUZEBDIiIpDGK50cO9vEJ4dq2ftlDbFReibHhmIMC3DKtaKZU6LkTEMMiHxahsnB4/VofNRMj5dE4G0iQvz5wfSRLB4RzJ/3VnPy4hXOXWohNMiXSbGhjBsZYr+OIMRQkEQwDGyKQtnxeqaNCydAjty8Vpjej7smGZkeH8mZuhZOnG/mi2P1fHmigXExwUyKDSU0yG+4mym8gHwLDYNTNVe43NrNzMk3f/BOeAeNj5oJo0KIjwmm4UonJ85f5uSFK5w4f5nocB2TYkMZbQxCrZZbjMXgkEQwDKRbSNyISqXCEBqAITSAuyYZqL54ha8vXOaTw7Xo/DVMHB3KhFEhchYpnE4+UUNMuoVEfwT4aZg2PoKEceFcrG/lxPnLHD7ZwJHqBmKjr11cNoQ65+KyEPJNNMSkW0gMhFqlIjZKT2yUniut3Xx94TLVNVc4W9dCmN6PSbGhjB0RLBeXxW2RRDDEDhyTbiFxa0KCfJk5xUjShEjO1F3lxPnL/P2oifITZuJjQpgUGypDmYtbIolgCFl7bHxRZWLGxEjpFhK3TKtR268X1F/u4MS5yxw/30zVuWZGRuoIDvTlzvhI6TYS/SbfRkPoq+pGWjsszJ46YribIjyASqUiKkxHVJiOji4rJy9c5vj5y7z6XgVjovXMS4njrokGudtIOCSJYAiVVtYREuhLwlgZW0g4V4CfhsT4SBLGhtNjg4++vMimwkqMYQHMnRXLvVOj0Wp8HFckvJIkgiFytb2bI6ca+WHyaHwGMBubEAPh46MmZWoUc+4aRfnXZnb8/Rx/LDlB4Wdn+GHyKB64Mwadv3fO/SH6JolgiHxxzESPTWH2tOjhborwAmq1ipmTjSRPMnD8XDM7vjjPe5+c5m/7z3H/nTH8MHk0YXp5allcI4lgiJRWXCIuSs8oN5++ULgXlUrFlDHhTBkTzrlLLez84hy7Dpznw7IL3JMQzdxZsYyICBzuZoph1q8+iuLiYubPn096ejqbN2++rryqqoqcnBwyMjJYvXo1VqsVgNraWhYvXszcuXN59tlnaWtrA+DAgQPMmjWL7OxssrOzWbVqlRNDcj2na69yztTCfYlykVgMn7hoPcuyp7JuaQo/SBzJ34+ZeP7NL3j1/QpO1V4Z7uaJYeQwEZhMJjZs2MCWLVsoLCxk27ZtVFdX91onPz+fNWvWsGvXLhRFYfv27QC88MILPP7445SUlDB16lR+//vfA1BZWcnTTz9NUVERRUVFrFu3bhBCcx17yi/i7+vD7KnSLSSGnzFMx5MZk/iPZ2eTOTuO4+eaeemP5azf/CVHTjXihpMWitvkMBGUlpaSkpJCaGgoOp2OjIwMSkpK7OU1NTV0dnaSlJQEQE5ODiUlJVgsFg4ePEhGRkav1wEqKirYt28fWVlZLFu2jLq6usGIzSVcbevm4HET904dIc8OCJcSHOhLTup4/uP/ms2PHoyn/nIHG//0Fb/874PsP3qJnh7bcDdRDBGHiaC+vh6DwWBfNhqNmEymPssNBgMmk4nm5maCgoLQaDS9XgfQ6/U8+eSTFBcXk5aWRl5entMCcjWfflWLtUfhwbtihrspQtxQgJ+GjLtjWb/sHp6eP4Uem403i4+xdN2HfFh2gS5Lz3A3UQwyh4eoNput1xOKiqL0Wu6r/PvrAfblF1980f7aokWLeOWVV2hpaUGv1/er0TebhPlmDIb+1e8sPT02Pv2qlqQJBhInO7db6JtYlKZ29EHOneoQQKvVDGm9znivoW5zX/q77mC1V6fzwxCuu6VtH4kOIfuBCRw8don3Pq5my4cnKS49R9Z9Y8m8b5xbD2Ex1H//g82Z8ThMBNHR0ZSVldmXzWYzRqOxV7nZ/O3E3g0NDRiNRsLDw2lpaaGnpwcfHx/7djabjddff52lS5fi4/PtAy7f/dmRxsZWbLaB9WMaDHrM5pYBbXO7DlSZaLjSyWP/NMGp7/3dWNq7rL0mS3cWi2Xo6v3+hO/OrNsZBlLvQGIZrPZ2dHZz9mLXbdURFerPzxffRcXJej48eIEtu0/w7scnSU0cSfrdo4kMCXBSa4fGcPz9D6aBxqNWq256AO0wEcyePZvf/e53NDU1ERAQwO7du/nVr35lL4+JicHPz4/y8nLuuusuioqKSE1NRavVkpyczI4dO8jKyqKwsJDU1FTUajUffPABcXFxzJ8/n8LCQqZPn45Od2tHMK7Kpij8tfQc0eE6kmSAOTGEuiw9fPW12fGKDnyT1GZMMjB2ZDCmpnY+PlTDR1/WMOsOI/NmxTHKKLdDewKHiSAqKoq8vDxyc3OxWCwsXLiQxMRElixZwvLly5k2bRoFBQU8//zztLa2kpCQQG5uLgC//OUvWblyJZs2bWLEiBH85je/AWD9+vX84he/4LXXXiM8PJyXX355cKMcBodPNnDR3MqSBXfIWC/C7YXp/Ui/O5ZH749n98ELfPpVLfuPmkgcH8G8WbFMHB0qg9y5sX7dxpKVlUVWVlav19588037z5MnT+bdd9+9bruYmBjeeeed616fMGECW7duHWhb3YaiKBR/fhZjaAB33yHzDgjPERHiz6I5E8i6dwwffXmRD8susn7LIcaPDGZeShxJEyJRS0JwO3I/4yA4cqqRc6YWfjx/sowrJDxSUICWh+4dS8bdsew7UseuA+d59f0KRkTomHt3LPdMjUbjI599dyGJwMlsNoU/f3aayBB/7kmQB8iE51CpVbR1Wa97PWVqNDPviOLQ12Y+LLvA2zuPU7jvDHOSR3PP1Oh+zZ7mp9Ugk6wNH0kETvbZkVrOm1pZ+tAdckQkPEp/LkI/OCOG2oZ2jpxq5E8fV/PX0jMkjAlnwujQmyaEmVOi0MgDl8NGfvNO1N5p4f1PTzNhVAizpkQNd3OEGHIqlYoYQyAjI3WYmjo4cqqRshNmKk43cceYMCbFheIr8yK4HEkETvSXz8/S2m7h8X+eKHdQCK+mUqmIjtARHaGjvrmDilONHDrZwNEzTUyOC2NKXBh+vpIQXIUkAic5d6mFPeUX+cH0kcRFe9YTjELcDmNYAP+UPIrGK50cOdXIkVONVJ1tZlJsKFPGhMkYXC5A9oATdFt6eKP4KEE6LQvvHz/czRHCJUWE+PPAjBiaW7qoONVI5Zkmqs41M3F0KBNHhxIoCWHYyNVMJ/jT3lPUNbbzL5lTCAqQaQCFuJkwvR+pSSPJvm8scdF6jp9v5v/89wHe2X2CxivOH3JDOCYp+DZ9Vd3AnvKLzEkexdSxEcPdHCHcRkiQL/cljmB6fASmpg4+PVzLp4drmT01msx74jCGedawM65MEsFtOG9q4fW/HGW0MYiFadIlJMSt0Ot8efCu0Tzyg3GUfHGeT76q5fOKS8y6I4rMe+IYGSlTaQ42SQS3qLmli/989wgBfhpWLEzEVyt3QAhxOyJC/FmcPpHM2XHsOnCejw/V8Pejl0iaEMmDd41iSlyYDF8xSCQR3ILLrV38ZvthOrqsrHriLsKDnT+mvBDeKjTIjx89OIH5KXF8UHaBvYdqOXSygahwHQ/cGcO906IJ9Jdrcc4kiWCATM3tvLL1MC3tFn72v6YxWobhFWJQ6HXXptLMmj2GsuNmPjp0ka17TvL+J6e4a5KBlIRo7hgTJuN5OYEkggE4cb6ZTYWV2BTIX3Qn40YGD3eThPAIfY1j9I3ECZEkTojkgqmFzyvq+PJrM/uPmtDrtMyYaCB5ipG4KP11D3LKGEb9I4mgH6w9Nor2nWHH/nMYQgNY8WgiIyLkApYQzjKQyXTGx4QwJlrPRXMbZ+qu8tlXdXxyuJYAPw2jDIGMMgYxIkKHxkctYxj1k/yGbkJRFI6cauTdvaeoaWjjB4kjWDRnAv6+8msTYjj5+KiJi9YTF62ny9LDBVMrF82tnKm7ysmLV/BRq4gK19Hc0kXCmHBC5VbUm5JvtBuw9tg4fLKBD8su8PXFKxjDAvhZzjTunGgY7qYJIb7HT+tD/KgQ4keF0GNTMDW1U2Nuo7axjeLPz1L8+Vm02w4zJlrP+JgQYo1BjI7SEx0eINcX/qFfiaC4uJhNmzZhtVp56qmnWLx4ca/yqqoqVq9eTVtbG8nJybzwwgtoNBpqa2vJz8+nsbGRsWPHUlBQQGBgIFevXuXnP/85Fy5cIDw8nI0bN2IwDO+XrMVq48SFZr6qbqTseD1X2roJD/bjifSJpE4fKUNKC+EGfNQqRkYG2p89mDImnNr6Vi42dfDV1/V8WHYBa48CgFajJuYf6xrDAogK09n/1/l71zGyw2hNJhMbNmzg/fffx9fXl8cee4xZs2YRHx9vXyc/P5+1a9eSlJTEc889x/bt23n88cd54YUXePzxx8nMzOS1117j97//Pfn5+WzcuJHk5GTeeOMNCgsLeemll9i4ceOgBvoNi7WHy63dNFzu4FJzB7UN1/oZz5tasPYoaDVqpo4N5wfTR5I4LkLmGxbCjQUFaLlzooF0gx6zORZrj41Lje1cqG/lfH0L502tVJ1rprTyUq/tAvx8CNP7ExbkS6jejzC9H0EBvgT6awgK0BIYoL32v78Gnb/G7c8sHCaC0tJSUlJSCA0NBSAjI4OSkhJ++tOfAlBTU0NnZydJSUkA5OTk8Nvf/pZHH32UgwcP8tprr9lff+KJJ8jPz2fv3r1s3rwZgAULFvDiiy9isVjQavt3b/CtfDkrisJ/vnuEusa2Xq/7an2IMQRy10QjY2OCiY8JwdcNbjP45neg8VGjG4R7qoey3gA/DT3W238vV/hdDCQWV2jvzXw/lsFq72DWrfFR2/9W1GoVvmofYqP1xEbruZcR9vW6rTaarnTScKWDhiudXGnt5krbtX+m5g6qa66gKH2/j1ajRqvxwU+jRqu99rOvRo1Wo77WBpUKlVqFWnXtrEWtujZUt1r9j3+osCkKNkWhx6aAotCjXJvx0GZT6LbasFh7SIo3kJIQZY+nvxyt6zAR1NfX9+q2MRqNHDlypM9yg8GAyWSiubmZoKAgNBpNr9e/v41GoyEoKIimpiaiovo3mUtY2K3dsbP22XtvaTtXFBHx7fMLo0aEDMp7jBsV5lb1DmbdUu/g1jvYdUPvv5kbGRHlXreDO4pnIBwe+tpstl735iqK0mu5r/Lvrwf0OVmLoiio3fzUSggh3JXDb9/o6GjM5m/v7zWbzRiNxj7LGxoaMBqNhIeH09LSQk9Pz3XbGY1GGhoaALBarbS1tdm7noQQQgwth4lg9uzZ7N+/n6amJjo6Oti9ezepqan28piYGPz8/CgvLwegqKiI1NRUtFotycnJ7NixA4DCwkL7dmlpaRQWFgKwY8cOkpOT+319QAghhHOpFOVml0CuKS4u5vXXX8disbBw4UKWLFnCkiVLWL58OdOmTeP48eM8//zztLa2kpCQwLp16/D19aWmpoaVK1fS2NjIiBEj+M1vfkNISAiXL19m5cqVXLhwAb1eT0FBAaNGjRqKeIUQQnxPvxKBEEIIzyVXaIUQwstJIhBCCC8niUAIIbycJAIhhPByHp0IysvLWbhwIdnZ2Tz11FPU1NQAcPXqVZYuXcq8efNYvHhxr+cgXF1xcTHz588nPT3dPkyHO3n11VfJzMwkMzOTl19+Gbg2jElWVhbp6els2LBhmFs4cOvXr2flypXAtQEYc3JyyMjIYPXq1VitfU+24ko++ugjcnJymDdvHmvXrgXce78UFRXZP2fr168H3G/ftLa2smDBAi5evAj0vT+cEpfiwR544AGlqqpKURRF+dOf/qQsW7ZMURRFeeGFF5TXX39dURRF+fOf/6ysWLFi2No4EJcuXVIeeOABpbm5WWlra1OysrKUkydPDnez+u3zzz9XfvSjHyldXV1Kd3e3kpubqxQXFytpaWnK+fPnFYvFojz99NPK3r17h7up/VZaWqrMmjVL+fd//3dFURQlMzNTOXTokKIoirJq1Spl8+bNw9m8fjl//rxy3333KXV1dUp3d7eyaNEiZe/evW67X9rb25WZM2cqjY2NisViURYuXKh8/vnnbrVvDh8+rCxYsEBJSEhQLly4oHR0dPS5P5wRl8eeEXR3d7NixQomT54MwKRJk6irqwNg7969ZGVlAdcGvfv000+xWCzD1tb++u4AgDqdzj4AoLswGAysXLkSX19ftFot48eP5+zZs8TFxTF69Gg0Gg1ZWVluE9Ply5fZsGEDy5YtA248AKM7xPLBBx8wf/58oqOj0Wq1bNiwgYCAALfdLz09PdhsNjo6OrBarVitVjQajVvtm+3bt/PLX/7SPhrDkSNHbrg/nPWZ89hBt319fcnOzgaujYf06quvMmfOHOD2B70bLo4GAHR1EyZMsP989uxZdu7cyRNPPHFdTN8MTujq1qxZQ15env0Ao68BGF3duXPn0Gq1LFu2jLq6Ou6//34mTJjgtvslKCiIFStWMG/ePAICApg5cyZardat9s1LL73Ua/lGf/smk8lpnzmPSAQ7d+5k3bp1vV4bN24cf/jDH+ju7mblypVYrVaeeeaZG26vuMmgd44GAHQXJ0+e5JlnnuHf/u3f8PHx4ezZs/Yyd4npT3/6EyNGjOCee+7h/fffB9x3//T09FBWVsY777yDTqfj2Wefxd/f3y1jATh+/DjvvfceH3/8MXq9np///Od8/vnnbhsP9P3ZctZnziMSwbx585g3b951r7e1tfHss88SGhrKpk2b7OMZfTPoXXR0tFsNehcdHU1ZWZl9+fsDALqD8vJyli9fznPPPUdmZiYHDhy46aCGrmrHjh2YzWays7O5cuUK7e3tqFSqGw7A6OoiIyO55557CA8PB2DOnDmUlJTg4+NjX8dd9gvAvn37uOeee4iIiACudZe89dZbbrlvvtHX4J99Dfo5UK5/GHwb8vPziYuLY+PGjfj6+tpfd9dB7xwNAOjq6urq+Nd//VcKCgrIzMwEYPr06Zw5c4Zz587R09PDX//6V7eI6e233+avf/0rRUVFLF++nAcffJB169bdcABGV/fAAw+wb98+rl69Sk9PD5999hlz5851y/0CMHnyZEpLS2lvb0dRFD766CPuvvtut9w33+jr76SvQT8HyiPOCG7k2LFj7Nmzh/j4eB555BHg2pnAm2++yYoVK1i5ciWZmZn2Qe/cQVRUFHl5eeTm5toHAExMTBzuZvXbW2+9RVdXF7/+9a/trz322GP8+te/5mc/+xldXV2kpaUxd+7cYWzl7SkoKOg1AGNubu5wN8mh6dOn85Of/ITHH38ci8XCvffey6JFixg3bpxb7pf77ruPY8eOkZOTg1arZdq0aSxdupQf/vCHbrdvvuHn59fn34kzPnMy6JwQQng5j+4aEkII4ZgkAiGE8HKSCIQQwstJIhBCCC8niUAIIbycJAIhhPBykgiEEMLLSSIQQggv9/8DDsyFZM9oc/UAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.distplot(vocab_instances20_alltissue_df_btn_dist_stat['median'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1459,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "count    9172.000000\n",
      "mean       10.907436\n",
      "std         5.349342\n",
      "min         3.000000\n",
      "25%         6.000000\n",
      "50%        10.000000\n",
      "75%        14.000000\n",
      "max        21.000000\n",
      "dtype: float64\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array(['THAP1', 'PATZ1', 'SP2', 'ZFP28', 'SP1', 'ZNF816', 'ZNF324', 'MAZ',\n",
       "       'VEZF1'], dtype=object)"
      ]
     },
     "execution_count": 1459,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# vocab length\n",
    "print(pd.Series(vocab_instances20_alltissue_df.vocab1_stop-vocab_instances20_alltissue_df.vocab1_start).append(pd.Series(vocab_instances20_alltissue_df.vocab2_stop-vocab_instances20_alltissue_df.vocab2_start)).describe())\n",
    "# vocab_instances20_alltissue_df_rs_filt[((vocab_instances20_alltissue_df_rs_filt.vocab1_stop-vocab_instances20_alltissue_df_rs_filt.vocab1_start)>=20)| ((vocab_instances20_alltissue_df_rs_filt.vocab2_stop-vocab_instances20_alltissue_df_rs_filt.vocab2_start)>=20)]#.vocab2.unique()\n",
    "# vocab_instances20_alltissue_df_rs_filt[ ((vocab_instances20_alltissue_df_rs_filt.vocab2_stop-vocab_instances20_alltissue_df_rs_filt.vocab2_start)>=20)].vocab2.unique()\n",
    "\n",
    "\n",
    "# really long tfs\n",
    "vocab_instances20_alltissue_df_rs_filt[ ((vocab_instances20_alltissue_df_rs_filt.vocab1_stop-vocab_instances20_alltissue_df_rs_filt.vocab1_start)>=20)].vocab1.unique()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. Add in positive and negative controls\n",
    "\n",
    "need to run 8F_controls.ipynb first"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1027,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "135"
      ]
     },
     "execution_count": 1027,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MAX_DIST"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1028,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>loc</th>\n",
       "      <th>end</th>\n",
       "      <th>width</th>\n",
       "      <th>strand</th>\n",
       "      <th>TSS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1846266</td>\n",
       "      <td>1846266</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>CALML6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>231376933</td>\n",
       "      <td>231376933</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>C1orf131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>241683061</td>\n",
       "      <td>241683061</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>FH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>120202421</td>\n",
       "      <td>120202421</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>PHGDH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>204135465</td>\n",
       "      <td>204135465</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>REN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>20255</td>\n",
       "      <td>chrX</td>\n",
       "      <td>140084756</td>\n",
       "      <td>140084756</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>SPANXB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>20256</td>\n",
       "      <td>chrX</td>\n",
       "      <td>52683950</td>\n",
       "      <td>52683950</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>SSX7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>20257</td>\n",
       "      <td>chrX</td>\n",
       "      <td>140096761</td>\n",
       "      <td>140096761</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>SPANXB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>20258</td>\n",
       "      <td>chrX</td>\n",
       "      <td>52736239</td>\n",
       "      <td>52736239</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>SSX2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>20259</td>\n",
       "      <td>chrX</td>\n",
       "      <td>147582139</td>\n",
       "      <td>147582139</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>AFF2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20178 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr        loc        end  width strand       TSS\n",
       "0      chr1    1846266    1846266      1      *    CALML6\n",
       "1      chr1  231376933  231376933      1      *  C1orf131\n",
       "2      chr1  241683061  241683061      1      *        FH\n",
       "3      chr1  120202421  120202421      1      *     PHGDH\n",
       "4      chr1  204135465  204135465      1      *       REN\n",
       "...     ...        ...        ...    ...    ...       ...\n",
       "20255  chrX  140084756  140084756      1      *   SPANXB2\n",
       "20256  chrX   52683950   52683950      1      *      SSX7\n",
       "20257  chrX  140096761  140096761      1      *   SPANXB1\n",
       "20258  chrX   52736239   52736239      1      *      SSX2\n",
       "20259  chrX  147582139  147582139      1      *      AFF2\n",
       "\n",
       "[20178 rows x 6 columns]"
      ]
     },
     "execution_count": 1028,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tss_df = pd.read_csv('../data/external/TSS_annon_hg19.csv',index_col=0).drop_duplicates('TSS')\n",
    "tss_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1029,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pos genes (85, 1)\n",
      "neg genes (275, 1)\n"
     ]
    }
   ],
   "source": [
    "# pos_genes_df = pd.read_csv(os.path.join(save_dir, 'pos_genes_df.csv'),index_col=0)#.gene.values#.iloc[:,0]\n",
    "pos_genes_df = pd.read_csv(os.path.join(save_dir, 'all_pos_genes.csv'),header=None, index_col=0).index.values\n",
    "pos_genes_df = pd.DataFrame.from_dict({'gene':pos_genes_df},orient='columns')\n",
    "print('pos genes', pos_genes_df.shape)\n",
    "neg_genes_df = pd.read_csv(os.path.join(save_dir, 'neg_genes_far.csv'),header=None, index_col=0).index.values\n",
    "neg_genes_df = pd.DataFrame.from_dict({'gene':neg_genes_df},orient='columns')\n",
    "print('neg genes', neg_genes_df.shape)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1030,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(85, 4)"
      ]
     },
     "execution_count": 1030,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pos_genes_df = pos_genes_df.merge(tss_df[['TSS','chr','end']], how='left',left_on='gene',right_on='TSS').dropna().drop_duplicates()\n",
    "pos_genes_df['start'] = pos_genes_df.end - MAX_DIST\n",
    "pos_genes_df = pos_genes_df[['chr','start','end','gene']].drop_duplicates().sort_values(['chr','start']).reset_index(drop=True)\n",
    "\n",
    "pos_genes_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1031,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(95, 4)"
      ]
     },
     "execution_count": 1031,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "neg_genes_df = neg_genes_df.merge(tss_df[['TSS','chr','end']], how='left',left_on='gene',right_on='TSS').dropna().drop_duplicates()#.shape\n",
    "neg_genes_df.end = neg_genes_df.end.map(int)\n",
    "neg_genes_df['start'] = neg_genes_df.end - MAX_DIST\n",
    "neg_genes_df = neg_genes_df[['chr','start','end','gene']].drop_duplicates().sort_values(['chr','start']).reset_index(drop=True)\n",
    "neg_genes_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1032,
   "metadata": {},
   "outputs": [],
   "source": [
    "pos_genes_df.to_csv(os.path.join(save_dir, 'pos_genes_position.csv'))\n",
    "neg_genes_df.to_csv(os.path.join(save_dir, 'neg_genes_position.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1232,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BedTool(../data/processed/fig4_modelling/vocab_sum_final/neg_genes_position.bed)>"
      ]
     },
     "execution_count": 1232,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pybedtools.BedTool.from_dataframe(pos_genes_df).sort().saveas(os.path.join(save_dir, 'pos_genes_position.bed'))\n",
    "pybedtools.BedTool.from_dataframe(neg_genes_df).sort().saveas(os.path.join(save_dir, 'neg_genes_position.bed'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. get sequences\n",
    "\n",
    "- vocab sequences in : `../data/processed/fig4_modelling/vocab_sum_final/vocab_instances20_alltissue_df.csv`\n",
    "- pos sequences in: `../data/processed/fig4_modelling/vocab_sum_final/pos_genes_position.csv`\n",
    "- neg sequences in : `../data/processed/fig4_modelling/vocab_sum_final/neg_genes_position.csv`\n",
    "\n",
    "get sequence by url requesting http://genome.ucsc.edu/cgi-bin/das/hg19/dna?segment=chr1:100000,200000\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4A get sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1037,
   "metadata": {},
   "outputs": [],
   "source": [
    "import urllib.request\n",
    "import xmltodict\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1038,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sequence(row,chr_col='chr', start_col='start', stop_col='end'):\n",
    "    url = 'http://genome.ucsc.edu/cgi-bin/das/hg19/dna?segment={}:{},{}'.format(row[chr_col],row[start_col],row[stop_col])#chr4:39355986,39356120'\n",
    "    contents = urllib.request.urlopen(url).read()\n",
    "    sequence = ''.join(xmltodict.parse(contents)['DASDNA']['SEQUENCE']['DNA']['#text'].upper().split('\\n'))\n",
    "    return(sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1171,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #testing\n",
    "# row = vocab_instances20_alltissue_df.loc[0,:]\n",
    "# get_sequence(row,chr_col='chr', start_col='start_pad', stop_col='stop_pad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1172,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4586, 18)\n",
      "CPU times: user 4.72 s, sys: 1.51 s, total: 6.23 s\n",
      "Wall time: 15min 45s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# probably will take ~20 min (12 min 35s for 15instances , 17 min for 20instances)\n",
    "print(vocab_instances20_alltissue_df.shape)\n",
    "vocab_instances20_alltissue_df['sequence'] = vocab_instances20_alltissue_df.apply(lambda x: get_sequence(x, chr_col='chr', start_col='start_pad', stop_col='stop_pad'),axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1173,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(85, 6)\n",
      "CPU times: user 81.4 ms, sys: 28.6 ms, total: 110 ms\n",
      "Wall time: 16.5 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# takes 18.3 s\n",
    "print(pos_genes_df.shape)\n",
    "pos_genes_df['sequence'] = pos_genes_df.apply(lambda x: get_sequence(x, chr_col='chr', start_col='start', stop_col='end'),axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1174,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(95, 6)\n",
      "CPU times: user 94.3 ms, sys: 32 ms, total: 126 ms\n",
      "Wall time: 20.1 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# takes 18.8 s\n",
    "print(neg_genes_df.shape)\n",
    "neg_genes_df['sequence'] = neg_genes_df.apply(lambda x: get_sequence(x, chr_col='chr', start_col='start', stop_col='end'),axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1175,
   "metadata": {},
   "outputs": [],
   "source": [
    "vocab_instances20_alltissue_df.to_csv(os.path.join(save_dir, 'vocab_instances20_alltissue_df_wseq.csv'))\n",
    "pos_genes_df.to_csv(os.path.join(save_dir, 'pos_genes_position_wseq.csv'))\n",
    "neg_genes_df.to_csv(os.path.join(save_dir, 'pos_genes_position_wseq.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4B MPRA CONSTANTS (choosing RS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# define constants\n",
    "LEN_FILLER = 20\n",
    "NUM_OLIGO_PER_MOTIF = 10\n",
    "DESIRED_OLIGO_LENGTH = 150\n",
    "MAX_OLIGO_LENGTH = 230\n",
    "MAX_MOTIF_LENGTH = 63\n",
    "FORWARD_PCR_PRIMER = 'ACTGGCCGCTTCACTG'\n",
    "REVERSE_PCR_PRIMER = 'AGATCGGAAGAGCGTCG'\n",
    "# LINKER_SPACER_v10 = 'CAGTATGCGATGCTCATGATTGTGACTGAAAGTCTATGAGTCCATTGA'\n",
    "RS_ECORI = 'GAATTC' # 5'-3'\n",
    "RS_BAMHI = 'GGATCC'\n",
    "RS_XHOI = 'CTCGAG'\n",
    "RS_XBAI = 'TCTAGA'\n",
    "RS_BSPDI = 'ATCGAT'\n",
    "RS_NHEI = 'GCTAGC'\n",
    "RS_NCOI = 'CCATGG'\n",
    "\n",
    "# CCTAGG formed by RS_XHOI and RS_NHEI, can be cut by AvrII\n",
    "\n",
    "# RS_LIST = [RS_ECORI ,RS_BAMHI , RS_XHOI , RS_XBAI] #original\n",
    "# RS_LIST = [RS_BSPDI ,RS_BAMHI , RS_XHOI, RS_NHEI]\n",
    "RS_LIST = [RS_ECORI ,RS_BAMHI , RS_XHOI , RS_NHEI]\n",
    "\n",
    "RS_XBAI_dam1 = 'GATCTAGA'\n",
    "RS_XBAI_dam2 = 'TCTAGATC'\n",
    "RS_XBAI_dam1_pre = '^GA'\n",
    "RS_XBAI_dam2_suf = 'TC$'\n",
    "LETTERS=['A', 'C', 'G', 'T']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4C check if the mpra restriction enzymes (and then remove sequences that don't pass rs checking\n",
    "\n",
    "revision: replace XbaI with NcoI b/c dam methylation of Xba (NcoI pairs with XhoI, in the second restriction digest)\n",
    "\n",
    "not if the restriction site is NOT the sequence area manually change the sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def check_rs(seq, rs_list = RS_LIST):\n",
    "    rs_sites = 0\n",
    "    for rs in rs_list:\n",
    "        if rs in seq:\n",
    "            rs_sites+=1\n",
    "    return rs_sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1183,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((340, 20), (4586, 20))"
      ]
     },
     "execution_count": 1183,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_alltissue_df['check_rs'] = vocab_instances20_alltissue_df.sequence.apply(check_rs)\n",
    "vocab_instances20_alltissue_df[vocab_instances20_alltissue_df.check_rs>0].shape, vocab_instances20_alltissue_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1184,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NFIC::RARG    10\n",
       "CBFB::IRF8    10\n",
       "Name: vocab_pair, dtype: int64"
      ]
     },
     "execution_count": 1184,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test = vocab_instances20_alltissue_df[vocab_instances20_alltissue_df.check_rs==0].vocab_pair.value_counts()\n",
    "test[test<=10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "507 of the 4691 instances are bad . But only all vocab words have at least 10 genomic instances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1185,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(13, 6) (85, 6)\n",
      "(6, 6) (95, 6)\n"
     ]
    }
   ],
   "source": [
    "# controls\n",
    "pos_genes_df['check_rs'] = pos_genes_df.sequence.apply(check_rs)\n",
    "print(pos_genes_df[pos_genes_df.check_rs>0].shape, pos_genes_df.shape)\n",
    "\n",
    "neg_genes_df['check_rs'] = neg_genes_df.sequence.apply(check_rs)\n",
    "print(neg_genes_df[neg_genes_df.check_rs>0].shape, neg_genes_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1186,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4246, 20)\n",
      "(72, 6)\n",
      "(89, 6)\n"
     ]
    }
   ],
   "source": [
    "# new dataframes\n",
    "vocab_instances20_alltissue_df_rs_filt = vocab_instances20_alltissue_df[vocab_instances20_alltissue_df.check_rs==0]\n",
    "pos_genes_df_rs_filt = pos_genes_df[pos_genes_df.check_rs==0]\n",
    "neg_genes_df_rs_filt = neg_genes_df[neg_genes_df.check_rs==0]\n",
    "print(vocab_instances20_alltissue_df_rs_filt.shape)\n",
    "print(pos_genes_df_rs_filt.shape)\n",
    "print(neg_genes_df_rs_filt.shape)\n",
    "vocab_instances20_alltissue_df_rs_filt.to_csv(os.path.join(save_dir, 'vocab_instances20_alltissue_df_wseq_rsfilt.csv'))\n",
    "pos_genes_df_rs_filt.to_csv(os.path.join(save_dir, 'pos_genes_position_wseq_rsfilt.csv'))\n",
    "neg_genes_df_rs_filt.to_csv(os.path.join(save_dir, 'neg_genes_position_wseq_rsfilt.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1187,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "BATF::KLF15       20\n",
       "NFE2L1::ZBTB49    20\n",
       "IRF8::STAT1       20\n",
       "FOXL1::PRRX1      20\n",
       "SP1::TBX1         20\n",
       "                  ..\n",
       "IRF4::PLAG1       12\n",
       "IKZF1::IRF2       12\n",
       "FOXM1::SP2        11\n",
       "NFIC::RARG        10\n",
       "CBFB::IRF8        10\n",
       "Name: vocab_pair, Length: 239, dtype: int64"
      ]
     },
     "execution_count": 1187,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_alltissue_df_rs_filt[['vocab_pair','chr','start_pad','stop_pad']].vocab_pair.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "all vocab instances have at least 10 genomic sites"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 5. scramble sequences (vocab words)\n",
    "\n",
    "test a max of `MAX_ITER` times, \n",
    "\n",
    "options:\n",
    "- scramble only the motif\n",
    "- scramble the 10 bp window around the motif"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5A get the scrambles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [],
   "source": [
    "vocab_instances20_alltissue_df_rs_filt = pd.read_csv(os.path.join(save_dir, 'vocab_instances20_alltissue_df_wseq_rsfilt.csv'),index_col=0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['vocab_pair', 'tot_dist', 'btn_dist', 'chr', 'start', 'stop', 'vocab1',\n",
       "       'vocab1_start', 'vocab1_stop', 'vocab2', 'vocab2_start', 'vocab2_stop',\n",
       "       'tissue', 'loop_type', 'start_pad', 'stop_pad', 'vocab_tissue_count',\n",
       "       'name', 'sequence', 'check_rs'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 162,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_alltissue_df_rs_filt.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([135])"
      ]
     },
     "execution_count": 163,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_alltissue_df_rs_filt.sequence.apply(len).unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>vocab_pair</th>\n",
       "      <th>tot_dist</th>\n",
       "      <th>btn_dist</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>vocab1</th>\n",
       "      <th>vocab1_start</th>\n",
       "      <th>vocab1_stop</th>\n",
       "      <th>vocab2</th>\n",
       "      <th>vocab2_start</th>\n",
       "      <th>vocab2_stop</th>\n",
       "      <th>tissue</th>\n",
       "      <th>loop_type</th>\n",
       "      <th>start_pad</th>\n",
       "      <th>stop_pad</th>\n",
       "      <th>vocab_tissue_count</th>\n",
       "      <th>name</th>\n",
       "      <th>sequence</th>\n",
       "      <th>check_rs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>BARX2::FOXF2</td>\n",
       "      <td>28</td>\n",
       "      <td>14</td>\n",
       "      <td>chr1</td>\n",
       "      <td>206848337</td>\n",
       "      <td>206848365</td>\n",
       "      <td>BARX2</td>\n",
       "      <td>206848337</td>\n",
       "      <td>206848341</td>\n",
       "      <td>FOXF2</td>\n",
       "      <td>206848355</td>\n",
       "      <td>206848365</td>\n",
       "      <td>Colon</td>\n",
       "      <td>pro_pro</td>\n",
       "      <td>206848284</td>\n",
       "      <td>206848418</td>\n",
       "      <td>1</td>\n",
       "      <td>BARX2::FOXF2+Colon+1</td>\n",
       "      <td>CTCAGCTCTCCCACAAGATCTCTGGGGAAGCTGGAAGAGGGCCACT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>BARX2::FOXF2</td>\n",
       "      <td>54</td>\n",
       "      <td>40</td>\n",
       "      <td>chr4</td>\n",
       "      <td>38859601</td>\n",
       "      <td>38859655</td>\n",
       "      <td>FOXF2</td>\n",
       "      <td>38859601</td>\n",
       "      <td>38859611</td>\n",
       "      <td>BARX2</td>\n",
       "      <td>38859651</td>\n",
       "      <td>38859655</td>\n",
       "      <td>Colon</td>\n",
       "      <td>loop_loop</td>\n",
       "      <td>38859561</td>\n",
       "      <td>38859695</td>\n",
       "      <td>2</td>\n",
       "      <td>BARX2::FOXF2+Colon+2</td>\n",
       "      <td>GTGGCTCTGAGGGAATTTCAAGTAACCGTGGACTAAAGTCTTAAGA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>BARX2::FOXF2</td>\n",
       "      <td>46</td>\n",
       "      <td>32</td>\n",
       "      <td>chr6</td>\n",
       "      <td>74222776</td>\n",
       "      <td>74222822</td>\n",
       "      <td>FOXF2</td>\n",
       "      <td>74222776</td>\n",
       "      <td>74222786</td>\n",
       "      <td>BARX2</td>\n",
       "      <td>74222818</td>\n",
       "      <td>74222822</td>\n",
       "      <td>Colon</td>\n",
       "      <td>loop_loop</td>\n",
       "      <td>74222732</td>\n",
       "      <td>74222866</td>\n",
       "      <td>3</td>\n",
       "      <td>BARX2::FOXF2+Colon+3</td>\n",
       "      <td>GCCACTGCACTCCAGCCTGGTGACAGAGCGAGACTCTGTCTCAAGA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>BARX2::FOXF2</td>\n",
       "      <td>16</td>\n",
       "      <td>2</td>\n",
       "      <td>chr3</td>\n",
       "      <td>49384674</td>\n",
       "      <td>49384690</td>\n",
       "      <td>FOXF2</td>\n",
       "      <td>49384674</td>\n",
       "      <td>49384684</td>\n",
       "      <td>BARX2</td>\n",
       "      <td>49384686</td>\n",
       "      <td>49384690</td>\n",
       "      <td>Colon</td>\n",
       "      <td>pro_pro</td>\n",
       "      <td>49384615</td>\n",
       "      <td>49384749</td>\n",
       "      <td>4</td>\n",
       "      <td>BARX2::FOXF2+Colon+4</td>\n",
       "      <td>AGTATTATGTGCTCATGGTCCCTACCCTCCAGGGTTCACAGTGTGG...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>BARX2::FOXF2</td>\n",
       "      <td>16</td>\n",
       "      <td>2</td>\n",
       "      <td>chr10</td>\n",
       "      <td>112247520</td>\n",
       "      <td>112247536</td>\n",
       "      <td>BARX2</td>\n",
       "      <td>112247520</td>\n",
       "      <td>112247524</td>\n",
       "      <td>FOXF2</td>\n",
       "      <td>112247526</td>\n",
       "      <td>112247536</td>\n",
       "      <td>Colon</td>\n",
       "      <td>pro_pro</td>\n",
       "      <td>112247461</td>\n",
       "      <td>112247595</td>\n",
       "      <td>5</td>\n",
       "      <td>BARX2::FOXF2+Colon+5</td>\n",
       "      <td>ATGTGGCTGCGGGGCTTTAGGAGCTGACTCTTTGGTCACGGGGAGA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     vocab_pair  tot_dist  btn_dist    chr      start       stop vocab1  \\\n",
       "0  BARX2::FOXF2        28        14   chr1  206848337  206848365  BARX2   \n",
       "1  BARX2::FOXF2        54        40   chr4   38859601   38859655  FOXF2   \n",
       "2  BARX2::FOXF2        46        32   chr6   74222776   74222822  FOXF2   \n",
       "3  BARX2::FOXF2        16         2   chr3   49384674   49384690  FOXF2   \n",
       "4  BARX2::FOXF2        16         2  chr10  112247520  112247536  BARX2   \n",
       "\n",
       "   vocab1_start  vocab1_stop vocab2  vocab2_start  vocab2_stop tissue  \\\n",
       "0     206848337    206848341  FOXF2     206848355    206848365  Colon   \n",
       "1      38859601     38859611  BARX2      38859651     38859655  Colon   \n",
       "2      74222776     74222786  BARX2      74222818     74222822  Colon   \n",
       "3      49384674     49384684  BARX2      49384686     49384690  Colon   \n",
       "4     112247520    112247524  FOXF2     112247526    112247536  Colon   \n",
       "\n",
       "   loop_type  start_pad   stop_pad  vocab_tissue_count                  name  \\\n",
       "0    pro_pro  206848284  206848418                   1  BARX2::FOXF2+Colon+1   \n",
       "1  loop_loop   38859561   38859695                   2  BARX2::FOXF2+Colon+2   \n",
       "2  loop_loop   74222732   74222866                   3  BARX2::FOXF2+Colon+3   \n",
       "3    pro_pro   49384615   49384749                   4  BARX2::FOXF2+Colon+4   \n",
       "4    pro_pro  112247461  112247595                   5  BARX2::FOXF2+Colon+5   \n",
       "\n",
       "                                            sequence  check_rs  \n",
       "0  CTCAGCTCTCCCACAAGATCTCTGGGGAAGCTGGAAGAGGGCCACT...         0  \n",
       "1  GTGGCTCTGAGGGAATTTCAAGTAACCGTGGACTAAAGTCTTAAGA...         0  \n",
       "2  GCCACTGCACTCCAGCCTGGTGACAGAGCGAGACTCTGTCTCAAGA...         0  \n",
       "3  AGTATTATGTGCTCATGGTCCCTACCCTCCAGGGTTCACAGTGTGG...         0  \n",
       "4  ATGTGGCTGCGGGGCTTTAGGAGCTGACTCTTTGGTCACGGGGAGA...         0  "
      ]
     },
     "execution_count": 164,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_alltissue_df_rs_filt[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [],
   "source": [
    "# motif_bed = pybedtools.BedTool('../data/raw/merged_motif_scan/Colon_merged.bed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%time\n",
    "# motif_bed.to_dataframe()[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {},
   "outputs": [],
   "source": [
    "row = vocab_instances20_alltissue_df_rs_filt.iloc[0,:]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%time\n",
    "# pybedtools.BedTool.from_dataframe(pd.DataFrame(row[['chr','start_pad','stop_pad']]).T).intersect(motif_bed,wa=True,wb=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "vocab_pair                                                 BARX2::FOXF2\n",
       "tot_dist                                                             28\n",
       "btn_dist                                                             14\n",
       "chr                                                                chr1\n",
       "start                                                         206848337\n",
       "stop                                                          206848365\n",
       "vocab1                                                            BARX2\n",
       "vocab1_start                                                  206848337\n",
       "vocab1_stop                                                   206848341\n",
       "vocab2                                                            FOXF2\n",
       "vocab2_start                                                  206848355\n",
       "vocab2_stop                                                   206848365\n",
       "tissue                                                            Colon\n",
       "loop_type                                                       pro_pro\n",
       "start_pad                                                     206848284\n",
       "stop_pad                                                      206848418\n",
       "vocab_tissue_count                                                    1\n",
       "name                                               BARX2::FOXF2+Colon+1\n",
       "sequence              CTCAGCTCTCCCACAAGATCTCTGGGGAAGCTGGAAGAGGGCCACT...\n",
       "check_rs                                                              0\n",
       "Name: 0, dtype: object"
      ]
     },
     "execution_count": 169,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['../data/raw/merged_motif_scan/COLO-SCR-DMSO_merged.bed',\n",
       " '../data/raw/merged_motif_scan/Icon\\r',\n",
       " '../data/raw/merged_motif_scan/Renal_merged.bed',\n",
       " '../data/raw/merged_motif_scan/Prostate_merged.bed',\n",
       " '../data/raw/merged_motif_scan/Uterine_merged.bed',\n",
       " '../data/raw/merged_motif_scan/GDSD6_merged.bed',\n",
       " '../data/raw/merged_motif_scan/Pancreas_merged.bed',\n",
       " '../data/raw/merged_motif_scan/SCC13-CTRLi_merged.bed',\n",
       " '../data/raw/merged_motif_scan/Melanocytes_merged.bed',\n",
       " '../data/raw/merged_motif_scan/Ovarian_merged.bed',\n",
       " '../data/raw/merged_motif_scan/Bladder_merged.bed',\n",
       " '../data/raw/merged_motif_scan/GDSD3_merged.bed',\n",
       " '../data/raw/merged_motif_scan/Esophageal_merged.bed',\n",
       " '../data/raw/merged_motif_scan/A431-CTRLi_merged.bed',\n",
       " '../data/raw/merged_motif_scan/GM12878_merged.bed',\n",
       " '../data/raw/merged_motif_scan/Astrocytes_merged.bed',\n",
       " '../data/raw/merged_motif_scan/Colon_merged.bed',\n",
       " '../data/raw/merged_motif_scan/WM2664-SCR-DMSO_merged.bed',\n",
       " '../data/raw/merged_motif_scan/GDSD0_merged.bed',\n",
       " '../data/raw/merged_motif_scan/CAL27-CTRLi_merged.bed',\n",
       " '../data/raw/merged_motif_scan/Thyroid_merged.bed',\n",
       " '../data/raw/merged_motif_scan/Airway_merged.bed',\n",
       " '../data/raw/merged_motif_scan/HMEC_merged.bed']"
      ]
     },
     "execution_count": 170,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glob.glob('../data/raw/merged_motif_scan/*')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {},
   "outputs": [],
   "source": [
    "MAX_ITER = 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {},
   "outputs": [],
   "source": [
    "from random import shuffle\n",
    "\n",
    "def mixup(word):\n",
    "    as_list_of_letters = list(word)\n",
    "    shuffle(as_list_of_letters)\n",
    "    return ''.join(as_list_of_letters)\n",
    "\n",
    "def get_scrambles(row, motif_len_scr=10):\n",
    "    # getting vocab locations in the sequence\n",
    "    vocab1_med = int((row['vocab1_start']+row['vocab1_stop'])/2)\n",
    "    vocab1_start_shift = vocab1_med - motif_len_scr - row['start_pad']\n",
    "    vocab1_stop_shift = vocab1_med + motif_len_scr -row['start_pad'] + 1\n",
    "    \n",
    "    vocab2_med = int((row['vocab2_start']+row['vocab2_stop'])/2)\n",
    "    vocab2_start_shift = vocab2_med - motif_len_scr - row['start_pad']\n",
    "    vocab2_stop_shift = vocab2_med + motif_len_scr -row['start_pad'] + 1\n",
    "    \n",
    "    # for overlaps\n",
    "    if vocab1_stop_shift>= vocab2_start_shift:\n",
    "#         print(idx)\n",
    "#         print(row)\n",
    "        \n",
    "        # case 1 vocab words are really close, the padding goes into the vocab word\n",
    "        if (row['vocab1_stop']-row['start_pad'] + 1) >=vocab2_start_shift:\n",
    "            \n",
    "            vocab1_stop_shift = row['vocab1_stop']-row['start_pad'] + 1\n",
    "            vocab2_start_shift = row['vocab1_stop']-row['start_pad'] + 1\n",
    "#             print(vocab1_start_shift,vocab1_stop_shift, vocab2_start_shift, vocab2_stop_shift)\n",
    "#             raise\n",
    "        else: \n",
    "            vocab1_stop_shift = vocab1_stop_shift\n",
    "            vocab2_start_shift = vocab1_stop_shift\n",
    "#             print(vocab1_start_shift,vocab1_stop_shift, vocab2_start_shift, vocab2_stop_shift)\n",
    "#             raise\n",
    "    # for edges:\n",
    "    vocab1_start_shift = max(0,vocab1_start_shift)\n",
    "    vocab2_stop_shift = min(len(row['sequence']),vocab2_stop_shift)\n",
    "    \n",
    "    \n",
    "    # get sequences\n",
    "    vocab1_seq = row['sequence'][vocab1_start_shift:vocab1_stop_shift]\n",
    "    vocab2_seq = row['sequence'][vocab2_start_shift:vocab2_stop_shift]\n",
    "\n",
    "    vocab1_seq_scr = mixup(vocab1_seq)\n",
    "    vocab2_seq_scr = mixup(vocab2_seq)\n",
    "\n",
    "    A_B_seq = row['sequence']\n",
    "    scrA_B_seq = row['sequence'][:vocab1_start_shift] + vocab1_seq_scr +  row['sequence'][vocab1_stop_shift:]\n",
    "    A_scrB_seq = row['sequence'][:vocab2_start_shift] + vocab2_seq_scr +  row['sequence'][vocab2_stop_shift:]\n",
    "    scrA_scrB_seq = row['sequence'][:vocab1_start_shift] + vocab1_seq_scr +  row['sequence'][vocab1_stop_shift:vocab2_start_shift] + vocab2_seq_scr +  row['sequence'][vocab2_stop_shift:]\n",
    "\n",
    "    if len(scrA_scrB_seq)!=135:\n",
    "        \n",
    "        print(row)\n",
    "        print(vocab1_start_shift, vocab1_stop_shift)\n",
    "        print('vocab1_seq_scr',vocab1_seq_scr)\n",
    "        print(vocab2_start_shift, vocab2_stop_shift)\n",
    "        print('vocab2_seq_scr',vocab2_seq_scr)\n",
    "        \n",
    "        raise\n",
    "#     if vocab1_stop_shift< vocab2_start_shift:\n",
    "#         scrA_scrB_seq = row['sequence'][:vocab1_start_shift] + vocab1_seq_scr +  row['sequence'][vocab1_stop_shift:vocab2_start_shift] + vocab2_seq_scr +  row['sequence'][vocab2_stop_shift:]\n",
    "#     else: # if the 10 bp windows for vocab1 and vocab2 overlap\n",
    "#         vocab12_seq = row['sequence'][vocab1_start_shift:vocab2_stop_shift]\n",
    "#         vocab12_seq_scr = mixup(vocab12_seq)\n",
    "#         scrA_scrB_seq = row['sequence'][:vocab1_start_shift] + vocab12_seq_scr +  row['sequence'][vocab2_stop_shift:]\n",
    "        \n",
    "\n",
    "#     # FOR DEBUG\n",
    "#     print(A_B_seq)\n",
    "#     print(scrA_B_seq)\n",
    "#     print(A_scrB_seq)\n",
    "#     print(scrA_scrB_seq)\n",
    "\n",
    "    return A_B_seq, scrA_B_seq, A_scrB_seq, scrA_scrB_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ### tesin\n",
    "# row = vocab_instances20_alltissue_df_rs_filt.iloc[100,:]\n",
    "# print(row)\n",
    "# motif_len_scr=10\n",
    "# vocab1_med = int((row['vocab1_start']+row['vocab1_stop'])/2)\n",
    "# vocab1_start_shift = vocab1_med - motif_len_scr - row['start_pad']\n",
    "# vocab1_stop_shift = vocab1_med + motif_len_scr -row['start_pad'] + 1\n",
    "# vocab2_med = int((row['vocab2_start']+row['vocab2_stop'])/2)\n",
    "# vocab2_start_shift = vocab2_med - motif_len_scr - row['start_pad']\n",
    "# vocab2_stop_shift = vocab2_med + motif_len_scr -row['start_pad'] + 1\n",
    "\n",
    "# print(vocab1_start_shift,vocab1_stop_shift, vocab2_start_shift, vocab2_stop_shift)\n",
    "# get_scrambles(row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4246, 25)\n",
      "0\n",
      "100\n",
      "200\n",
      "300\n",
      "400\n",
      "500\n",
      "600\n",
      "700\n",
      "800\n",
      "900\n",
      "1000\n",
      "1100\n",
      "1200\n",
      "1300\n",
      "1400\n",
      "1500\n",
      "1600\n",
      "1700\n",
      "1800\n",
      "1900\n",
      "2000\n",
      "2100\n",
      "2200\n",
      "2300\n",
      "2400\n",
      "2500\n",
      "2600\n",
      "2700\n",
      "2800\n",
      "2900\n",
      "3000\n",
      "3100\n",
      "3200\n",
      "3300\n",
      "3400\n",
      "3500\n",
      "3600\n",
      "3700\n",
      "3800\n",
      "3900\n",
      "4000\n",
      "4100\n",
      "4200\n"
     ]
    }
   ],
   "source": [
    "## iterate through the rows of the rs_filt vocab dataframe vocab_instances20_alltissue_df_rs_filt\n",
    "print(vocab_instances20_alltissue_df_rs_filt.shape)\n",
    "vocab_instances20_alltissue_df_rs_filt['scr_pass_rs'] = False\n",
    "vocab_instances20_alltissue_df_rs_filt['A_B_seq'] = ''\n",
    "vocab_instances20_alltissue_df_rs_filt['scrA_B_seq'] = ''\n",
    "vocab_instances20_alltissue_df_rs_filt['A_scrB_seq'] = ''\n",
    "vocab_instances20_alltissue_df_rs_filt['scrA_scrB_seq'] = ''\n",
    "\n",
    "for idx, row in vocab_instances20_alltissue_df_rs_filt.iterrows():\n",
    "    ## DEBUG\n",
    "    if (idx%100)==0:\n",
    "        print(idx)\n",
    "    \n",
    "    # set default\n",
    "    scr_pass_rs = False\n",
    "    num_iter = 0\n",
    "    # iterate scrambles until find some that pass the check_rs filter\n",
    "    while (not scr_pass_rs) ^ (num_iter > MAX_ITER):\n",
    "        A_B_seq, scrA_B_seq, A_scrB_seq, scrA_scrB_seq = get_scrambles(row)\n",
    "\n",
    "\n",
    "\n",
    "        # check_rs\n",
    "        A_B_seq_rs = check_rs(A_B_seq)\n",
    "        scrA_B_seq_rs = check_rs(scrA_B_seq)\n",
    "        A_scrB_seq_rs = check_rs(A_scrB_seq)\n",
    "        scrA_scrB_seq_rs = check_rs(scrA_scrB_seq)\n",
    "\n",
    "        if (A_B_seq_rs+scrA_B_seq_rs+A_scrB_seq_rs+scrA_scrB_seq_rs==0):\n",
    "            scr_pass_rs=True\n",
    "        num_iter+=1\n",
    "    \n",
    "    # add information to df\n",
    "    vocab_instances20_alltissue_df_rs_filt.at[idx, 'scr_pass_rs'] = scr_pass_rs\n",
    "    vocab_instances20_alltissue_df_rs_filt.at[idx, 'A_B_seq'] = A_B_seq\n",
    "    vocab_instances20_alltissue_df_rs_filt.at[idx, 'scrA_B_seq'] = scrA_B_seq\n",
    "    vocab_instances20_alltissue_df_rs_filt.at[idx, 'A_scrB_seq'] = A_scrB_seq\n",
    "    vocab_instances20_alltissue_df_rs_filt.at[idx, 'scrA_scrB_seq'] = scrA_scrB_seq\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True    4246\n",
       "Name: scr_pass_rs, dtype: int64"
      ]
     },
     "execution_count": 195,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# manual step\n",
    "vocab_instances20_alltissue_df_rs_filt.scr_pass_rs.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "all sequences and scr configurations pass the rs filter\n",
    "\n",
    "if they don't keep running the scrambling until they all pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>vocab_pair</th>\n",
       "      <th>tot_dist</th>\n",
       "      <th>btn_dist</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>start_pad</th>\n",
       "      <th>stop_pad</th>\n",
       "      <th>vocab1</th>\n",
       "      <th>vocab1_start</th>\n",
       "      <th>...</th>\n",
       "      <th>vocab2_start</th>\n",
       "      <th>vocab2_stop</th>\n",
       "      <th>tissue</th>\n",
       "      <th>loop_type</th>\n",
       "      <th>vocab_tissue_count</th>\n",
       "      <th>name</th>\n",
       "      <th>check_rs</th>\n",
       "      <th>scr_pass_rs</th>\n",
       "      <th>seq_config</th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>BARX2::FOXF2</td>\n",
       "      <td>28</td>\n",
       "      <td>14</td>\n",
       "      <td>chr1</td>\n",
       "      <td>206848337</td>\n",
       "      <td>206848365</td>\n",
       "      <td>206848284</td>\n",
       "      <td>206848418</td>\n",
       "      <td>BARX2</td>\n",
       "      <td>206848337</td>\n",
       "      <td>...</td>\n",
       "      <td>206848355</td>\n",
       "      <td>206848365</td>\n",
       "      <td>Colon</td>\n",
       "      <td>pro_pro</td>\n",
       "      <td>1</td>\n",
       "      <td>BARX2::FOXF2+Colon+1</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>A_B_seq</td>\n",
       "      <td>CTCAGCTCTCCCACAAGATCTCTGGGGAAGCTGGAAGAGGGCCACT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>BARX2::FOXF2</td>\n",
       "      <td>54</td>\n",
       "      <td>40</td>\n",
       "      <td>chr4</td>\n",
       "      <td>38859601</td>\n",
       "      <td>38859655</td>\n",
       "      <td>38859561</td>\n",
       "      <td>38859695</td>\n",
       "      <td>FOXF2</td>\n",
       "      <td>38859601</td>\n",
       "      <td>...</td>\n",
       "      <td>38859651</td>\n",
       "      <td>38859655</td>\n",
       "      <td>Colon</td>\n",
       "      <td>loop_loop</td>\n",
       "      <td>2</td>\n",
       "      <td>BARX2::FOXF2+Colon+2</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>A_B_seq</td>\n",
       "      <td>GTGGCTCTGAGGGAATTTCAAGTAACCGTGGACTAAAGTCTTAAGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>BARX2::FOXF2</td>\n",
       "      <td>46</td>\n",
       "      <td>32</td>\n",
       "      <td>chr6</td>\n",
       "      <td>74222776</td>\n",
       "      <td>74222822</td>\n",
       "      <td>74222732</td>\n",
       "      <td>74222866</td>\n",
       "      <td>FOXF2</td>\n",
       "      <td>74222776</td>\n",
       "      <td>...</td>\n",
       "      <td>74222818</td>\n",
       "      <td>74222822</td>\n",
       "      <td>Colon</td>\n",
       "      <td>loop_loop</td>\n",
       "      <td>3</td>\n",
       "      <td>BARX2::FOXF2+Colon+3</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>A_B_seq</td>\n",
       "      <td>GCCACTGCACTCCAGCCTGGTGACAGAGCGAGACTCTGTCTCAAGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>BARX2::FOXF2</td>\n",
       "      <td>16</td>\n",
       "      <td>2</td>\n",
       "      <td>chr3</td>\n",
       "      <td>49384674</td>\n",
       "      <td>49384690</td>\n",
       "      <td>49384615</td>\n",
       "      <td>49384749</td>\n",
       "      <td>FOXF2</td>\n",
       "      <td>49384674</td>\n",
       "      <td>...</td>\n",
       "      <td>49384686</td>\n",
       "      <td>49384690</td>\n",
       "      <td>Colon</td>\n",
       "      <td>pro_pro</td>\n",
       "      <td>4</td>\n",
       "      <td>BARX2::FOXF2+Colon+4</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>A_B_seq</td>\n",
       "      <td>AGTATTATGTGCTCATGGTCCCTACCCTCCAGGGTTCACAGTGTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>BARX2::FOXF2</td>\n",
       "      <td>16</td>\n",
       "      <td>2</td>\n",
       "      <td>chr10</td>\n",
       "      <td>112247520</td>\n",
       "      <td>112247536</td>\n",
       "      <td>112247461</td>\n",
       "      <td>112247595</td>\n",
       "      <td>BARX2</td>\n",
       "      <td>112247520</td>\n",
       "      <td>...</td>\n",
       "      <td>112247526</td>\n",
       "      <td>112247536</td>\n",
       "      <td>Colon</td>\n",
       "      <td>pro_pro</td>\n",
       "      <td>5</td>\n",
       "      <td>BARX2::FOXF2+Colon+5</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>A_B_seq</td>\n",
       "      <td>ATGTGGCTGCGGGGCTTTAGGAGCTGACTCTTTGGTCACGGGGAGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16979</td>\n",
       "      <td>STAT1::UBP1</td>\n",
       "      <td>28</td>\n",
       "      <td>7</td>\n",
       "      <td>chr2</td>\n",
       "      <td>65648057</td>\n",
       "      <td>65648085</td>\n",
       "      <td>65648004</td>\n",
       "      <td>65648138</td>\n",
       "      <td>UBP1</td>\n",
       "      <td>65648057</td>\n",
       "      <td>...</td>\n",
       "      <td>65648069</td>\n",
       "      <td>65648085</td>\n",
       "      <td>MM</td>\n",
       "      <td>loop_loop</td>\n",
       "      <td>16</td>\n",
       "      <td>STAT1::UBP1+MM+16</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>scrA_scrB_seq</td>\n",
       "      <td>AGAGATTTAAACATCATCATCTTGATTCCCAGTAACAACAATGCTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16980</td>\n",
       "      <td>STAT1::UBP1</td>\n",
       "      <td>32</td>\n",
       "      <td>14</td>\n",
       "      <td>chr1</td>\n",
       "      <td>24106333</td>\n",
       "      <td>24106365</td>\n",
       "      <td>24106282</td>\n",
       "      <td>24106416</td>\n",
       "      <td>UBP1</td>\n",
       "      <td>24106333</td>\n",
       "      <td>...</td>\n",
       "      <td>24106352</td>\n",
       "      <td>24106365</td>\n",
       "      <td>MM</td>\n",
       "      <td>pro_pro</td>\n",
       "      <td>17</td>\n",
       "      <td>STAT1::UBP1+MM+17</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>scrA_scrB_seq</td>\n",
       "      <td>TTGCCCAGGGTCACACAACTAAACAGGTATTCAAACTCAGGTTCCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16981</td>\n",
       "      <td>STAT1::UBP1</td>\n",
       "      <td>35</td>\n",
       "      <td>14</td>\n",
       "      <td>chr7</td>\n",
       "      <td>114551357</td>\n",
       "      <td>114551392</td>\n",
       "      <td>114551307</td>\n",
       "      <td>114551441</td>\n",
       "      <td>UBP1</td>\n",
       "      <td>114551357</td>\n",
       "      <td>...</td>\n",
       "      <td>114551376</td>\n",
       "      <td>114551392</td>\n",
       "      <td>MM</td>\n",
       "      <td>loop_loop</td>\n",
       "      <td>18</td>\n",
       "      <td>STAT1::UBP1+MM+18</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>scrA_scrB_seq</td>\n",
       "      <td>TCATTAGTGGGCTTTCAAAATACCAGCTTCTTGGTTACTGAGACAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16982</td>\n",
       "      <td>STAT1::UBP1</td>\n",
       "      <td>28</td>\n",
       "      <td>7</td>\n",
       "      <td>chr1</td>\n",
       "      <td>120178558</td>\n",
       "      <td>120178586</td>\n",
       "      <td>120178505</td>\n",
       "      <td>120178639</td>\n",
       "      <td>STAT1</td>\n",
       "      <td>120178558</td>\n",
       "      <td>...</td>\n",
       "      <td>120178581</td>\n",
       "      <td>120178586</td>\n",
       "      <td>MM</td>\n",
       "      <td>pro_pro</td>\n",
       "      <td>19</td>\n",
       "      <td>STAT1::UBP1+MM+19</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>scrA_scrB_seq</td>\n",
       "      <td>TCATATCACACTACGAAGAGAGCAACTAGAAGATATTATGCTCCTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16983</td>\n",
       "      <td>STAT1::UBP1</td>\n",
       "      <td>25</td>\n",
       "      <td>7</td>\n",
       "      <td>chr9</td>\n",
       "      <td>132360293</td>\n",
       "      <td>132360318</td>\n",
       "      <td>132360238</td>\n",
       "      <td>132360372</td>\n",
       "      <td>UBP1</td>\n",
       "      <td>132360293</td>\n",
       "      <td>...</td>\n",
       "      <td>132360305</td>\n",
       "      <td>132360318</td>\n",
       "      <td>MM</td>\n",
       "      <td>loop_loop</td>\n",
       "      <td>20</td>\n",
       "      <td>STAT1::UBP1+MM+20</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>scrA_scrB_seq</td>\n",
       "      <td>GCTTCCTGCATCAGCCAAGCCCTGCCCGCAGTGTAACCCGTGGTCC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16984 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         vocab_pair  tot_dist  btn_dist    chr      start       stop  \\\n",
       "0      BARX2::FOXF2        28        14   chr1  206848337  206848365   \n",
       "1      BARX2::FOXF2        54        40   chr4   38859601   38859655   \n",
       "2      BARX2::FOXF2        46        32   chr6   74222776   74222822   \n",
       "3      BARX2::FOXF2        16         2   chr3   49384674   49384690   \n",
       "4      BARX2::FOXF2        16         2  chr10  112247520  112247536   \n",
       "...             ...       ...       ...    ...        ...        ...   \n",
       "16979   STAT1::UBP1        28         7   chr2   65648057   65648085   \n",
       "16980   STAT1::UBP1        32        14   chr1   24106333   24106365   \n",
       "16981   STAT1::UBP1        35        14   chr7  114551357  114551392   \n",
       "16982   STAT1::UBP1        28         7   chr1  120178558  120178586   \n",
       "16983   STAT1::UBP1        25         7   chr9  132360293  132360318   \n",
       "\n",
       "       start_pad   stop_pad vocab1  vocab1_start  ...  vocab2_start  \\\n",
       "0      206848284  206848418  BARX2     206848337  ...     206848355   \n",
       "1       38859561   38859695  FOXF2      38859601  ...      38859651   \n",
       "2       74222732   74222866  FOXF2      74222776  ...      74222818   \n",
       "3       49384615   49384749  FOXF2      49384674  ...      49384686   \n",
       "4      112247461  112247595  BARX2     112247520  ...     112247526   \n",
       "...          ...        ...    ...           ...  ...           ...   \n",
       "16979   65648004   65648138   UBP1      65648057  ...      65648069   \n",
       "16980   24106282   24106416   UBP1      24106333  ...      24106352   \n",
       "16981  114551307  114551441   UBP1     114551357  ...     114551376   \n",
       "16982  120178505  120178639  STAT1     120178558  ...     120178581   \n",
       "16983  132360238  132360372   UBP1     132360293  ...     132360305   \n",
       "\n",
       "      vocab2_stop  tissue  loop_type vocab_tissue_count                  name  \\\n",
       "0       206848365   Colon    pro_pro                  1  BARX2::FOXF2+Colon+1   \n",
       "1        38859655   Colon  loop_loop                  2  BARX2::FOXF2+Colon+2   \n",
       "2        74222822   Colon  loop_loop                  3  BARX2::FOXF2+Colon+3   \n",
       "3        49384690   Colon    pro_pro                  4  BARX2::FOXF2+Colon+4   \n",
       "4       112247536   Colon    pro_pro                  5  BARX2::FOXF2+Colon+5   \n",
       "...           ...     ...        ...                ...                   ...   \n",
       "16979    65648085      MM  loop_loop                 16     STAT1::UBP1+MM+16   \n",
       "16980    24106365      MM    pro_pro                 17     STAT1::UBP1+MM+17   \n",
       "16981   114551392      MM  loop_loop                 18     STAT1::UBP1+MM+18   \n",
       "16982   120178586      MM    pro_pro                 19     STAT1::UBP1+MM+19   \n",
       "16983   132360318      MM  loop_loop                 20     STAT1::UBP1+MM+20   \n",
       "\n",
       "       check_rs scr_pass_rs     seq_config  \\\n",
       "0             0        True        A_B_seq   \n",
       "1             0        True        A_B_seq   \n",
       "2             0        True        A_B_seq   \n",
       "3             0        True        A_B_seq   \n",
       "4             0        True        A_B_seq   \n",
       "...         ...         ...            ...   \n",
       "16979         0        True  scrA_scrB_seq   \n",
       "16980         0        True  scrA_scrB_seq   \n",
       "16981         0        True  scrA_scrB_seq   \n",
       "16982         0        True  scrA_scrB_seq   \n",
       "16983         0        True  scrA_scrB_seq   \n",
       "\n",
       "                                                     seq  \n",
       "0      CTCAGCTCTCCCACAAGATCTCTGGGGAAGCTGGAAGAGGGCCACT...  \n",
       "1      GTGGCTCTGAGGGAATTTCAAGTAACCGTGGACTAAAGTCTTAAGA...  \n",
       "2      GCCACTGCACTCCAGCCTGGTGACAGAGCGAGACTCTGTCTCAAGA...  \n",
       "3      AGTATTATGTGCTCATGGTCCCTACCCTCCAGGGTTCACAGTGTGG...  \n",
       "4      ATGTGGCTGCGGGGCTTTAGGAGCTGACTCTTTGGTCACGGGGAGA...  \n",
       "...                                                  ...  \n",
       "16979  AGAGATTTAAACATCATCATCTTGATTCCCAGTAACAACAATGCTG...  \n",
       "16980  TTGCCCAGGGTCACACAACTAAACAGGTATTCAAACTCAGGTTCCC...  \n",
       "16981  TCATTAGTGGGCTTTCAAAATACCAGCTTCTTGGTTACTGAGACAC...  \n",
       "16982  TCATATCACACTACGAAGAGAGCAACTAGAAGATATTATGCTCCTA...  \n",
       "16983  GCTTCCTGCATCAGCCAAGCCCTGCCCGCAGTGTAACCCGTGGTCC...  \n",
       "\n",
       "[16984 rows x 22 columns]"
      ]
     },
     "execution_count": 196,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_alltissue_df_rs_filt.reset_index(drop=True, inplace=True)\n",
    "vocab_instances20_alltissue_df_rs_filt[ 'vocab_tissue_count'] = vocab_instances20_alltissue_df_rs_filt.groupby(['vocab_pair','tissue']).cumcount()+1\n",
    "vocab_instances20_alltissue_df_rs_filt['name'] = vocab_instances20_alltissue_df_rs_filt.vocab_pair + '+'+ vocab_instances20_alltissue_df_rs_filt.tissue + '+'+ vocab_instances20_alltissue_df_rs_filt.vocab_tissue_count.map(str)\n",
    "vocab_instances20_seqs = pd.melt(vocab_instances20_alltissue_df_rs_filt.drop('sequence',axis=1),\n",
    "         id_vars=['vocab_pair', 'tot_dist', 'btn_dist', 'chr', 'start', 'stop', 'start_pad', 'stop_pad',\n",
    "                  'vocab1','vocab1_start', 'vocab1_stop', 'vocab2', 'vocab2_start', 'vocab2_stop',\n",
    "                  'tissue', 'loop_type',  'vocab_tissue_count','name','check_rs','scr_pass_rs'], \n",
    "            var_name='seq_config',\n",
    "             value_name=\"seq\")\n",
    "\n",
    "vocab_instances20_seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([135])"
      ]
     },
     "execution_count": 197,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_seqs.seq.apply(len).unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(16984, 22) (4246, 25)\n"
     ]
    }
   ],
   "source": [
    "# save the vocab info\n",
    "print(vocab_instances20_seqs.shape, vocab_instances20_alltissue_df_rs_filt.shape)\n",
    "vocab_instances20_seqs.to_csv(os.path.join(save_dir, 'vocab_instances20_seq_wscr.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['vocab_pair', 'tot_dist', 'btn_dist', 'chr', 'start', 'stop',\n",
       "       'start_pad', 'stop_pad', 'vocab1', 'vocab1_start', 'vocab1_stop',\n",
       "       'vocab2', 'vocab2_start', 'vocab2_stop', 'tissue', 'loop_type',\n",
       "       'vocab_tissue_count', 'name', 'check_rs', 'scr_pass_rs', 'seq_config',\n",
       "       'seq'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 199,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_seqs.columns\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5B only pick 10 best sequences (vocab) add the controls for final df\n",
    "\n",
    "\n",
    "Input DFs:\n",
    "- `vocab_instances20_seqs`: `vocab_instances20_seq_wscr.csv`\n",
    "- `pos_genes_df_rs_filt`: `pos_genes_position_wseq_rsfilt.csv`\n",
    "- `neg_genes_df_rs_filt`: `neg_genes_position_wseq_rsfilt.csv`\n",
    "\n",
    "Output DF: \n",
    "- `mpra_lib_seq`: 'mpra_lib_seq.csv`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(239, 2)"
      ]
     },
     "execution_count": 200,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_seqs[['vocab_pair','tissue']].drop_duplicates().shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "still have 239 vocab words"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(239,)"
      ]
     },
     "execution_count": 201,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_seqs[vocab_instances20_seqs.vocab_tissue_count==10].vocab_pair.unique().shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "all 239 vocab words have at least 10 genomic instances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>vocab_pair</th>\n",
       "      <th>tot_dist</th>\n",
       "      <th>btn_dist</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>start_pad</th>\n",
       "      <th>stop_pad</th>\n",
       "      <th>vocab1</th>\n",
       "      <th>vocab1_start</th>\n",
       "      <th>...</th>\n",
       "      <th>vocab2_stop</th>\n",
       "      <th>tissue</th>\n",
       "      <th>loop_type</th>\n",
       "      <th>vocab_tissue_count</th>\n",
       "      <th>name</th>\n",
       "      <th>check_rs</th>\n",
       "      <th>scr_pass_rs</th>\n",
       "      <th>seq_config</th>\n",
       "      <th>seq</th>\n",
       "      <th>name_config</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>4413</td>\n",
       "      <td>FOXM1::SP2</td>\n",
       "      <td>131</td>\n",
       "      <td>110</td>\n",
       "      <td>chr16</td>\n",
       "      <td>68268437</td>\n",
       "      <td>68268568</td>\n",
       "      <td>68268435</td>\n",
       "      <td>68268569</td>\n",
       "      <td>SP2</td>\n",
       "      <td>68268437</td>\n",
       "      <td>...</td>\n",
       "      <td>68268568</td>\n",
       "      <td>Colon</td>\n",
       "      <td>loop_loop</td>\n",
       "      <td>8</td>\n",
       "      <td>FOXM1::SP2+Colon+8</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>scrA_B_seq</td>\n",
       "      <td>GCAGCCCCGCCTCCAGGTCTGACCAACCGAGGGGCAGAGAAACTCC...</td>\n",
       "      <td>FOXM1::SP2+Colon+8+scrA_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4425</td>\n",
       "      <td>FOXM1::SP4</td>\n",
       "      <td>131</td>\n",
       "      <td>107</td>\n",
       "      <td>chr16</td>\n",
       "      <td>68268437</td>\n",
       "      <td>68268568</td>\n",
       "      <td>68268435</td>\n",
       "      <td>68268569</td>\n",
       "      <td>SP4</td>\n",
       "      <td>68268437</td>\n",
       "      <td>...</td>\n",
       "      <td>68268568</td>\n",
       "      <td>Colon</td>\n",
       "      <td>loop_loop</td>\n",
       "      <td>9</td>\n",
       "      <td>FOXM1::SP4+Colon+9</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>scrA_B_seq</td>\n",
       "      <td>GCAGCCCCGCCTCCAGGTCTGACCAACCGAGGGGCAGAGAAACTCC...</td>\n",
       "      <td>FOXM1::SP4+Colon+9+scrA_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4624</td>\n",
       "      <td>KLF5::RUNX3</td>\n",
       "      <td>123</td>\n",
       "      <td>105</td>\n",
       "      <td>chr1</td>\n",
       "      <td>6259618</td>\n",
       "      <td>6259741</td>\n",
       "      <td>6259612</td>\n",
       "      <td>6259746</td>\n",
       "      <td>RUNX3</td>\n",
       "      <td>6259618</td>\n",
       "      <td>...</td>\n",
       "      <td>6259741</td>\n",
       "      <td>Colon</td>\n",
       "      <td>loop_loop</td>\n",
       "      <td>9</td>\n",
       "      <td>KLF5::RUNX3+Colon+9</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>scrA_B_seq</td>\n",
       "      <td>CCATACTAACCACAGGAGCCATGGCGGCAGCGGAGTTAGAAAGGGA...</td>\n",
       "      <td>KLF5::RUNX3+Colon+9+scrA_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4625</td>\n",
       "      <td>KLF5::RUNX3</td>\n",
       "      <td>123</td>\n",
       "      <td>105</td>\n",
       "      <td>chr1</td>\n",
       "      <td>6259618</td>\n",
       "      <td>6259741</td>\n",
       "      <td>6259612</td>\n",
       "      <td>6259746</td>\n",
       "      <td>RUNX3</td>\n",
       "      <td>6259618</td>\n",
       "      <td>...</td>\n",
       "      <td>6259741</td>\n",
       "      <td>Colon</td>\n",
       "      <td>pro_pro</td>\n",
       "      <td>10</td>\n",
       "      <td>KLF5::RUNX3+Colon+10</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>scrA_B_seq</td>\n",
       "      <td>CCATACTAACCACAGGAGCCATGGCGGCAGCGGAGTTAGAAAGGGA...</td>\n",
       "      <td>KLF5::RUNX3+Colon+10+scrA_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4734</td>\n",
       "      <td>MSX2::THAP1</td>\n",
       "      <td>121</td>\n",
       "      <td>95</td>\n",
       "      <td>chr11</td>\n",
       "      <td>102177000</td>\n",
       "      <td>102177121</td>\n",
       "      <td>102176993</td>\n",
       "      <td>102177127</td>\n",
       "      <td>MSX2</td>\n",
       "      <td>102177000</td>\n",
       "      <td>...</td>\n",
       "      <td>102177121</td>\n",
       "      <td>Colon</td>\n",
       "      <td>loop_loop</td>\n",
       "      <td>10</td>\n",
       "      <td>MSX2::THAP1+Colon+10</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>scrA_B_seq</td>\n",
       "      <td>AATGGTGAAATTAGTAATCATCCCAGCTGGTATAATACTGAATTGT...</td>\n",
       "      <td>MSX2::THAP1+Colon+10+scrA_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>14903</td>\n",
       "      <td>IRF8::ZNF816</td>\n",
       "      <td>128</td>\n",
       "      <td>106</td>\n",
       "      <td>chr9</td>\n",
       "      <td>134604524</td>\n",
       "      <td>134604652</td>\n",
       "      <td>134604521</td>\n",
       "      <td>134604655</td>\n",
       "      <td>ZNF816</td>\n",
       "      <td>134604524</td>\n",
       "      <td>...</td>\n",
       "      <td>134604652</td>\n",
       "      <td>GM12878</td>\n",
       "      <td>loop_loop</td>\n",
       "      <td>9</td>\n",
       "      <td>IRF8::ZNF816+GM12878+9</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>scrA_scrB_seq</td>\n",
       "      <td>CCCAGGGGCCATGAGGCAGGTCTCCCACCTCTCAGCTCAGGCTCAT...</td>\n",
       "      <td>IRF8::ZNF816+GM12878+9+scrA_scrB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>14904</td>\n",
       "      <td>IRF8::ZNF816</td>\n",
       "      <td>134</td>\n",
       "      <td>101</td>\n",
       "      <td>chr22</td>\n",
       "      <td>39337645</td>\n",
       "      <td>39337779</td>\n",
       "      <td>39337645</td>\n",
       "      <td>39337779</td>\n",
       "      <td>IRF8</td>\n",
       "      <td>39337645</td>\n",
       "      <td>...</td>\n",
       "      <td>39337779</td>\n",
       "      <td>GM12878</td>\n",
       "      <td>loop_loop</td>\n",
       "      <td>10</td>\n",
       "      <td>IRF8::ZNF816+GM12878+10</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>scrA_scrB_seq</td>\n",
       "      <td>TGGTTTTAGTTCCTTGGGTGGTTTCCTCGGGGAACCACTATTTCTT...</td>\n",
       "      <td>IRF8::ZNF816+GM12878+10+scrA_scrB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>15290</td>\n",
       "      <td>MITF::ZNF589</td>\n",
       "      <td>123</td>\n",
       "      <td>110</td>\n",
       "      <td>chr11</td>\n",
       "      <td>62446468</td>\n",
       "      <td>62446591</td>\n",
       "      <td>62446462</td>\n",
       "      <td>62446596</td>\n",
       "      <td>ZNF589</td>\n",
       "      <td>62446468</td>\n",
       "      <td>...</td>\n",
       "      <td>62446591</td>\n",
       "      <td>Melanocytes</td>\n",
       "      <td>loop_loop</td>\n",
       "      <td>1</td>\n",
       "      <td>MITF::ZNF589+Melanocytes+1</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>scrA_scrB_seq</td>\n",
       "      <td>GCGAGGCAACCCGCCCTCGACACCCGCCGACGGGCGCTCGCTCTCT...</td>\n",
       "      <td>MITF::ZNF589+Melanocytes+1+scrA_scrB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16351</td>\n",
       "      <td>EGR2::SOX10</td>\n",
       "      <td>127</td>\n",
       "      <td>109</td>\n",
       "      <td>chr9</td>\n",
       "      <td>38381691</td>\n",
       "      <td>38381818</td>\n",
       "      <td>38381687</td>\n",
       "      <td>38381821</td>\n",
       "      <td>EGR2</td>\n",
       "      <td>38381691</td>\n",
       "      <td>...</td>\n",
       "      <td>38381818</td>\n",
       "      <td>MM</td>\n",
       "      <td>loop_loop</td>\n",
       "      <td>8</td>\n",
       "      <td>EGR2::SOX10+MM+8</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>scrA_scrB_seq</td>\n",
       "      <td>AGGTGCCACCTCCACTCTGTCCTCACGGCCCTTGGCCTGGGACCCC...</td>\n",
       "      <td>EGR2::SOX10+MM+8+scrA_scrB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16522</td>\n",
       "      <td>LEF1::NFAT5</td>\n",
       "      <td>132</td>\n",
       "      <td>117</td>\n",
       "      <td>chr7</td>\n",
       "      <td>16783057</td>\n",
       "      <td>16783189</td>\n",
       "      <td>16783056</td>\n",
       "      <td>16783190</td>\n",
       "      <td>NFAT5</td>\n",
       "      <td>16783057</td>\n",
       "      <td>...</td>\n",
       "      <td>16783189</td>\n",
       "      <td>MM</td>\n",
       "      <td>pro_pro</td>\n",
       "      <td>1</td>\n",
       "      <td>LEF1::NFAT5+MM+1</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>scrA_scrB_seq</td>\n",
       "      <td>TATTTCCAAAAACTCTGTGTCTTTGTCGTAGGTTGTATCTAGTAGA...</td>\n",
       "      <td>LEF1::NFAT5+MM+1+scrA_scrB</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>74 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         vocab_pair  tot_dist  btn_dist    chr      start       stop  \\\n",
       "4413     FOXM1::SP2       131       110  chr16   68268437   68268568   \n",
       "4425     FOXM1::SP4       131       107  chr16   68268437   68268568   \n",
       "4624    KLF5::RUNX3       123       105   chr1    6259618    6259741   \n",
       "4625    KLF5::RUNX3       123       105   chr1    6259618    6259741   \n",
       "4734    MSX2::THAP1       121        95  chr11  102177000  102177121   \n",
       "...             ...       ...       ...    ...        ...        ...   \n",
       "14903  IRF8::ZNF816       128       106   chr9  134604524  134604652   \n",
       "14904  IRF8::ZNF816       134       101  chr22   39337645   39337779   \n",
       "15290  MITF::ZNF589       123       110  chr11   62446468   62446591   \n",
       "16351   EGR2::SOX10       127       109   chr9   38381691   38381818   \n",
       "16522   LEF1::NFAT5       132       117   chr7   16783057   16783189   \n",
       "\n",
       "       start_pad   stop_pad  vocab1  vocab1_start  ...  vocab2_stop  \\\n",
       "4413    68268435   68268569     SP2      68268437  ...     68268568   \n",
       "4425    68268435   68268569     SP4      68268437  ...     68268568   \n",
       "4624     6259612    6259746   RUNX3       6259618  ...      6259741   \n",
       "4625     6259612    6259746   RUNX3       6259618  ...      6259741   \n",
       "4734   102176993  102177127    MSX2     102177000  ...    102177121   \n",
       "...          ...        ...     ...           ...  ...          ...   \n",
       "14903  134604521  134604655  ZNF816     134604524  ...    134604652   \n",
       "14904   39337645   39337779    IRF8      39337645  ...     39337779   \n",
       "15290   62446462   62446596  ZNF589      62446468  ...     62446591   \n",
       "16351   38381687   38381821    EGR2      38381691  ...     38381818   \n",
       "16522   16783056   16783190   NFAT5      16783057  ...     16783189   \n",
       "\n",
       "            tissue  loop_type  vocab_tissue_count                        name  \\\n",
       "4413         Colon  loop_loop                   8          FOXM1::SP2+Colon+8   \n",
       "4425         Colon  loop_loop                   9          FOXM1::SP4+Colon+9   \n",
       "4624         Colon  loop_loop                   9         KLF5::RUNX3+Colon+9   \n",
       "4625         Colon    pro_pro                  10        KLF5::RUNX3+Colon+10   \n",
       "4734         Colon  loop_loop                  10        MSX2::THAP1+Colon+10   \n",
       "...            ...        ...                 ...                         ...   \n",
       "14903      GM12878  loop_loop                   9      IRF8::ZNF816+GM12878+9   \n",
       "14904      GM12878  loop_loop                  10     IRF8::ZNF816+GM12878+10   \n",
       "15290  Melanocytes  loop_loop                   1  MITF::ZNF589+Melanocytes+1   \n",
       "16351           MM  loop_loop                   8            EGR2::SOX10+MM+8   \n",
       "16522           MM    pro_pro                   1            LEF1::NFAT5+MM+1   \n",
       "\n",
       "      check_rs  scr_pass_rs     seq_config  \\\n",
       "4413         0         True     scrA_B_seq   \n",
       "4425         0         True     scrA_B_seq   \n",
       "4624         0         True     scrA_B_seq   \n",
       "4625         0         True     scrA_B_seq   \n",
       "4734         0         True     scrA_B_seq   \n",
       "...        ...          ...            ...   \n",
       "14903        0         True  scrA_scrB_seq   \n",
       "14904        0         True  scrA_scrB_seq   \n",
       "15290        0         True  scrA_scrB_seq   \n",
       "16351        0         True  scrA_scrB_seq   \n",
       "16522        0         True  scrA_scrB_seq   \n",
       "\n",
       "                                                     seq  \\\n",
       "4413   GCAGCCCCGCCTCCAGGTCTGACCAACCGAGGGGCAGAGAAACTCC...   \n",
       "4425   GCAGCCCCGCCTCCAGGTCTGACCAACCGAGGGGCAGAGAAACTCC...   \n",
       "4624   CCATACTAACCACAGGAGCCATGGCGGCAGCGGAGTTAGAAAGGGA...   \n",
       "4625   CCATACTAACCACAGGAGCCATGGCGGCAGCGGAGTTAGAAAGGGA...   \n",
       "4734   AATGGTGAAATTAGTAATCATCCCAGCTGGTATAATACTGAATTGT...   \n",
       "...                                                  ...   \n",
       "14903  CCCAGGGGCCATGAGGCAGGTCTCCCACCTCTCAGCTCAGGCTCAT...   \n",
       "14904  TGGTTTTAGTTCCTTGGGTGGTTTCCTCGGGGAACCACTATTTCTT...   \n",
       "15290  GCGAGGCAACCCGCCCTCGACACCCGCCGACGGGCGCTCGCTCTCT...   \n",
       "16351  AGGTGCCACCTCCACTCTGTCCTCACGGCCCTTGGCCTGGGACCCC...   \n",
       "16522  TATTTCCAAAAACTCTGTGTCTTTGTCGTAGGTTGTATCTAGTAGA...   \n",
       "\n",
       "                                name_config  \n",
       "4413              FOXM1::SP2+Colon+8+scrA_B  \n",
       "4425              FOXM1::SP4+Colon+9+scrA_B  \n",
       "4624             KLF5::RUNX3+Colon+9+scrA_B  \n",
       "4625            KLF5::RUNX3+Colon+10+scrA_B  \n",
       "4734            MSX2::THAP1+Colon+10+scrA_B  \n",
       "...                                     ...  \n",
       "14903      IRF8::ZNF816+GM12878+9+scrA_scrB  \n",
       "14904     IRF8::ZNF816+GM12878+10+scrA_scrB  \n",
       "15290  MITF::ZNF589+Melanocytes+1+scrA_scrB  \n",
       "16351            EGR2::SOX10+MM+8+scrA_scrB  \n",
       "16522            LEF1::NFAT5+MM+1+scrA_scrB  \n",
       "\n",
       "[74 rows x 23 columns]"
      ]
     },
     "execution_count": 202,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_seqs_sel[vocab_instances20_seqs_sel.seq.apply(len)!=135]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "all vocab genomic i"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>vocab_pair</th>\n",
       "      <th>tot_dist</th>\n",
       "      <th>btn_dist</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>start_pad</th>\n",
       "      <th>stop_pad</th>\n",
       "      <th>vocab1</th>\n",
       "      <th>vocab1_start</th>\n",
       "      <th>...</th>\n",
       "      <th>vocab2_stop</th>\n",
       "      <th>tissue</th>\n",
       "      <th>loop_type</th>\n",
       "      <th>vocab_tissue_count</th>\n",
       "      <th>name</th>\n",
       "      <th>check_rs</th>\n",
       "      <th>scr_pass_rs</th>\n",
       "      <th>seq_config</th>\n",
       "      <th>seq</th>\n",
       "      <th>name_config</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>BARX2::FOXF2</td>\n",
       "      <td>28</td>\n",
       "      <td>14</td>\n",
       "      <td>chr1</td>\n",
       "      <td>206848337</td>\n",
       "      <td>206848365</td>\n",
       "      <td>206848284</td>\n",
       "      <td>206848418</td>\n",
       "      <td>BARX2</td>\n",
       "      <td>206848337</td>\n",
       "      <td>...</td>\n",
       "      <td>206848365</td>\n",
       "      <td>Colon</td>\n",
       "      <td>pro_pro</td>\n",
       "      <td>1</td>\n",
       "      <td>BARX2::FOXF2+Colon+1</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>A_B_seq</td>\n",
       "      <td>CTCAGCTCTCCCACAAGATCTCTGGGGAAGCTGGAAGAGGGCCACT...</td>\n",
       "      <td>BARX2::FOXF2+Colon+1+A_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>BARX2::FOXF2</td>\n",
       "      <td>54</td>\n",
       "      <td>40</td>\n",
       "      <td>chr4</td>\n",
       "      <td>38859601</td>\n",
       "      <td>38859655</td>\n",
       "      <td>38859561</td>\n",
       "      <td>38859695</td>\n",
       "      <td>FOXF2</td>\n",
       "      <td>38859601</td>\n",
       "      <td>...</td>\n",
       "      <td>38859655</td>\n",
       "      <td>Colon</td>\n",
       "      <td>loop_loop</td>\n",
       "      <td>2</td>\n",
       "      <td>BARX2::FOXF2+Colon+2</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>A_B_seq</td>\n",
       "      <td>GTGGCTCTGAGGGAATTTCAAGTAACCGTGGACTAAAGTCTTAAGA...</td>\n",
       "      <td>BARX2::FOXF2+Colon+2+A_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>BARX2::FOXF2</td>\n",
       "      <td>46</td>\n",
       "      <td>32</td>\n",
       "      <td>chr6</td>\n",
       "      <td>74222776</td>\n",
       "      <td>74222822</td>\n",
       "      <td>74222732</td>\n",
       "      <td>74222866</td>\n",
       "      <td>FOXF2</td>\n",
       "      <td>74222776</td>\n",
       "      <td>...</td>\n",
       "      <td>74222822</td>\n",
       "      <td>Colon</td>\n",
       "      <td>loop_loop</td>\n",
       "      <td>3</td>\n",
       "      <td>BARX2::FOXF2+Colon+3</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>A_B_seq</td>\n",
       "      <td>GCCACTGCACTCCAGCCTGGTGACAGAGCGAGACTCTGTCTCAAGA...</td>\n",
       "      <td>BARX2::FOXF2+Colon+3+A_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>BARX2::FOXF2</td>\n",
       "      <td>16</td>\n",
       "      <td>2</td>\n",
       "      <td>chr3</td>\n",
       "      <td>49384674</td>\n",
       "      <td>49384690</td>\n",
       "      <td>49384615</td>\n",
       "      <td>49384749</td>\n",
       "      <td>FOXF2</td>\n",
       "      <td>49384674</td>\n",
       "      <td>...</td>\n",
       "      <td>49384690</td>\n",
       "      <td>Colon</td>\n",
       "      <td>pro_pro</td>\n",
       "      <td>4</td>\n",
       "      <td>BARX2::FOXF2+Colon+4</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>A_B_seq</td>\n",
       "      <td>AGTATTATGTGCTCATGGTCCCTACCCTCCAGGGTTCACAGTGTGG...</td>\n",
       "      <td>BARX2::FOXF2+Colon+4+A_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>BARX2::FOXF2</td>\n",
       "      <td>16</td>\n",
       "      <td>2</td>\n",
       "      <td>chr10</td>\n",
       "      <td>112247520</td>\n",
       "      <td>112247536</td>\n",
       "      <td>112247461</td>\n",
       "      <td>112247595</td>\n",
       "      <td>BARX2</td>\n",
       "      <td>112247520</td>\n",
       "      <td>...</td>\n",
       "      <td>112247536</td>\n",
       "      <td>Colon</td>\n",
       "      <td>pro_pro</td>\n",
       "      <td>5</td>\n",
       "      <td>BARX2::FOXF2+Colon+5</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>A_B_seq</td>\n",
       "      <td>ATGTGGCTGCGGGGCTTTAGGAGCTGACTCTTTGGTCACGGGGAGA...</td>\n",
       "      <td>BARX2::FOXF2+Colon+5+A_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>15125</td>\n",
       "      <td>FOXL1::PRRX1</td>\n",
       "      <td>33</td>\n",
       "      <td>13</td>\n",
       "      <td>chr17</td>\n",
       "      <td>45908368</td>\n",
       "      <td>45908401</td>\n",
       "      <td>45908317</td>\n",
       "      <td>45908451</td>\n",
       "      <td>PRRX1</td>\n",
       "      <td>45908368</td>\n",
       "      <td>...</td>\n",
       "      <td>45908401</td>\n",
       "      <td>Melanocytes</td>\n",
       "      <td>pro_pro</td>\n",
       "      <td>6</td>\n",
       "      <td>FOXL1::PRRX1+Melanocytes+6</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>scrA_scrB_seq</td>\n",
       "      <td>TTTAAATGAGCTAACACATATAAAAATAGTTATCACAGTGCCCTAT...</td>\n",
       "      <td>FOXL1::PRRX1+Melanocytes+6+scrA_scrB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>15126</td>\n",
       "      <td>FOXL1::PRRX1</td>\n",
       "      <td>69</td>\n",
       "      <td>49</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11322685</td>\n",
       "      <td>11322754</td>\n",
       "      <td>11322652</td>\n",
       "      <td>11322786</td>\n",
       "      <td>PRRX1</td>\n",
       "      <td>11322685</td>\n",
       "      <td>...</td>\n",
       "      <td>11322754</td>\n",
       "      <td>Melanocytes</td>\n",
       "      <td>pro_pro</td>\n",
       "      <td>7</td>\n",
       "      <td>FOXL1::PRRX1+Melanocytes+7</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>scrA_scrB_seq</td>\n",
       "      <td>TAGGGAGAAGATGGACTGAAAAACAACGGCGATAAAGAAAATGTGA...</td>\n",
       "      <td>FOXL1::PRRX1+Melanocytes+7+scrA_scrB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>15127</td>\n",
       "      <td>FOXL1::PRRX1</td>\n",
       "      <td>22</td>\n",
       "      <td>2</td>\n",
       "      <td>chr16</td>\n",
       "      <td>84210876</td>\n",
       "      <td>84210898</td>\n",
       "      <td>84210820</td>\n",
       "      <td>84210954</td>\n",
       "      <td>PRRX1</td>\n",
       "      <td>84210876</td>\n",
       "      <td>...</td>\n",
       "      <td>84210898</td>\n",
       "      <td>Melanocytes</td>\n",
       "      <td>pro_pro</td>\n",
       "      <td>8</td>\n",
       "      <td>FOXL1::PRRX1+Melanocytes+8</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>scrA_scrB_seq</td>\n",
       "      <td>CAGTCTGAGCTGGAGCCCGTGCAGCCTCGGGCGGGGGCAGGCCTGG...</td>\n",
       "      <td>FOXL1::PRRX1+Melanocytes+8+scrA_scrB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>15128</td>\n",
       "      <td>FOXL1::PRRX1</td>\n",
       "      <td>29</td>\n",
       "      <td>9</td>\n",
       "      <td>chrX</td>\n",
       "      <td>102820628</td>\n",
       "      <td>102820657</td>\n",
       "      <td>102820575</td>\n",
       "      <td>102820709</td>\n",
       "      <td>PRRX1</td>\n",
       "      <td>102820628</td>\n",
       "      <td>...</td>\n",
       "      <td>102820657</td>\n",
       "      <td>Melanocytes</td>\n",
       "      <td>pro_pro</td>\n",
       "      <td>9</td>\n",
       "      <td>FOXL1::PRRX1+Melanocytes+9</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>scrA_scrB_seq</td>\n",
       "      <td>TAAGAAGTCAGTGCTGTACTACAGCCACAGTTTTTTAAAATTAACT...</td>\n",
       "      <td>FOXL1::PRRX1+Melanocytes+9+scrA_scrB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>15129</td>\n",
       "      <td>FOXL1::PRRX1</td>\n",
       "      <td>54</td>\n",
       "      <td>34</td>\n",
       "      <td>chr1</td>\n",
       "      <td>33886937</td>\n",
       "      <td>33886991</td>\n",
       "      <td>33886897</td>\n",
       "      <td>33887031</td>\n",
       "      <td>PRRX1</td>\n",
       "      <td>33886937</td>\n",
       "      <td>...</td>\n",
       "      <td>33886991</td>\n",
       "      <td>Melanocytes</td>\n",
       "      <td>loop_loop</td>\n",
       "      <td>10</td>\n",
       "      <td>FOXL1::PRRX1+Melanocytes+10</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>scrA_scrB_seq</td>\n",
       "      <td>GTAAGCTAGTAAAGGACAGGAAGTAGGGCAAATGCCAGTGGGAAGT...</td>\n",
       "      <td>FOXL1::PRRX1+Melanocytes+10+scrA_scrB</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>480 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         vocab_pair  tot_dist  btn_dist    chr      start       stop  \\\n",
       "0      BARX2::FOXF2        28        14   chr1  206848337  206848365   \n",
       "1      BARX2::FOXF2        54        40   chr4   38859601   38859655   \n",
       "2      BARX2::FOXF2        46        32   chr6   74222776   74222822   \n",
       "3      BARX2::FOXF2        16         2   chr3   49384674   49384690   \n",
       "4      BARX2::FOXF2        16         2  chr10  112247520  112247536   \n",
       "...             ...       ...       ...    ...        ...        ...   \n",
       "15125  FOXL1::PRRX1        33        13  chr17   45908368   45908401   \n",
       "15126  FOXL1::PRRX1        69        49   chr1   11322685   11322754   \n",
       "15127  FOXL1::PRRX1        22         2  chr16   84210876   84210898   \n",
       "15128  FOXL1::PRRX1        29         9   chrX  102820628  102820657   \n",
       "15129  FOXL1::PRRX1        54        34   chr1   33886937   33886991   \n",
       "\n",
       "       start_pad   stop_pad vocab1  vocab1_start  ...  vocab2_stop  \\\n",
       "0      206848284  206848418  BARX2     206848337  ...    206848365   \n",
       "1       38859561   38859695  FOXF2      38859601  ...     38859655   \n",
       "2       74222732   74222866  FOXF2      74222776  ...     74222822   \n",
       "3       49384615   49384749  FOXF2      49384674  ...     49384690   \n",
       "4      112247461  112247595  BARX2     112247520  ...    112247536   \n",
       "...          ...        ...    ...           ...  ...          ...   \n",
       "15125   45908317   45908451  PRRX1      45908368  ...     45908401   \n",
       "15126   11322652   11322786  PRRX1      11322685  ...     11322754   \n",
       "15127   84210820   84210954  PRRX1      84210876  ...     84210898   \n",
       "15128  102820575  102820709  PRRX1     102820628  ...    102820657   \n",
       "15129   33886897   33887031  PRRX1      33886937  ...     33886991   \n",
       "\n",
       "            tissue  loop_type  vocab_tissue_count  \\\n",
       "0            Colon    pro_pro                   1   \n",
       "1            Colon  loop_loop                   2   \n",
       "2            Colon  loop_loop                   3   \n",
       "3            Colon    pro_pro                   4   \n",
       "4            Colon    pro_pro                   5   \n",
       "...            ...        ...                 ...   \n",
       "15125  Melanocytes    pro_pro                   6   \n",
       "15126  Melanocytes    pro_pro                   7   \n",
       "15127  Melanocytes    pro_pro                   8   \n",
       "15128  Melanocytes    pro_pro                   9   \n",
       "15129  Melanocytes  loop_loop                  10   \n",
       "\n",
       "                              name check_rs  scr_pass_rs     seq_config  \\\n",
       "0             BARX2::FOXF2+Colon+1        0         True        A_B_seq   \n",
       "1             BARX2::FOXF2+Colon+2        0         True        A_B_seq   \n",
       "2             BARX2::FOXF2+Colon+3        0         True        A_B_seq   \n",
       "3             BARX2::FOXF2+Colon+4        0         True        A_B_seq   \n",
       "4             BARX2::FOXF2+Colon+5        0         True        A_B_seq   \n",
       "...                            ...      ...          ...            ...   \n",
       "15125   FOXL1::PRRX1+Melanocytes+6        0         True  scrA_scrB_seq   \n",
       "15126   FOXL1::PRRX1+Melanocytes+7        0         True  scrA_scrB_seq   \n",
       "15127   FOXL1::PRRX1+Melanocytes+8        0         True  scrA_scrB_seq   \n",
       "15128   FOXL1::PRRX1+Melanocytes+9        0         True  scrA_scrB_seq   \n",
       "15129  FOXL1::PRRX1+Melanocytes+10        0         True  scrA_scrB_seq   \n",
       "\n",
       "                                                     seq  \\\n",
       "0      CTCAGCTCTCCCACAAGATCTCTGGGGAAGCTGGAAGAGGGCCACT...   \n",
       "1      GTGGCTCTGAGGGAATTTCAAGTAACCGTGGACTAAAGTCTTAAGA...   \n",
       "2      GCCACTGCACTCCAGCCTGGTGACAGAGCGAGACTCTGTCTCAAGA...   \n",
       "3      AGTATTATGTGCTCATGGTCCCTACCCTCCAGGGTTCACAGTGTGG...   \n",
       "4      ATGTGGCTGCGGGGCTTTAGGAGCTGACTCTTTGGTCACGGGGAGA...   \n",
       "...                                                  ...   \n",
       "15125  TTTAAATGAGCTAACACATATAAAAATAGTTATCACAGTGCCCTAT...   \n",
       "15126  TAGGGAGAAGATGGACTGAAAAACAACGGCGATAAAGAAAATGTGA...   \n",
       "15127  CAGTCTGAGCTGGAGCCCGTGCAGCCTCGGGCGGGGGCAGGCCTGG...   \n",
       "15128  TAAGAAGTCAGTGCTGTACTACAGCCACAGTTTTTTAAAATTAACT...   \n",
       "15129  GTAAGCTAGTAAAGGACAGGAAGTAGGGCAAATGCCAGTGGGAAGT...   \n",
       "\n",
       "                                 name_config  \n",
       "0                   BARX2::FOXF2+Colon+1+A_B  \n",
       "1                   BARX2::FOXF2+Colon+2+A_B  \n",
       "2                   BARX2::FOXF2+Colon+3+A_B  \n",
       "3                   BARX2::FOXF2+Colon+4+A_B  \n",
       "4                   BARX2::FOXF2+Colon+5+A_B  \n",
       "...                                      ...  \n",
       "15125   FOXL1::PRRX1+Melanocytes+6+scrA_scrB  \n",
       "15126   FOXL1::PRRX1+Melanocytes+7+scrA_scrB  \n",
       "15127   FOXL1::PRRX1+Melanocytes+8+scrA_scrB  \n",
       "15128   FOXL1::PRRX1+Melanocytes+9+scrA_scrB  \n",
       "15129  FOXL1::PRRX1+Melanocytes+10+scrA_scrB  \n",
       "\n",
       "[480 rows x 23 columns]"
      ]
     },
     "execution_count": 203,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_seqs_sel[vocab_instances20_seqs_sel.vocab_pair.str.contains('FOX')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### saving selection of vocab words\n",
    "how `name_config` works:\n",
    "\n",
    "`df['vocab_pair','tissue','vocab_tissue_count','seq_config'] = df.name_config.str.split('+',expand=True)`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for rerunnin \n",
    "pos_genes_df_rs_filt = pd.read_csv(os.path.join(save_dir, 'pos_genes_position_wseq_rsfilt.csv'),index_col=0)\n",
    "neg_genes_df_rs_filt = pd.read_csv(os.path.join(save_dir, 'neg_genes_position_wseq_rsfilt.csv'),index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(9560, 23) (16984, 22)\n"
     ]
    }
   ],
   "source": [
    "vocab_instances20_seqs_sel = vocab_instances20_seqs[vocab_instances20_seqs.vocab_tissue_count<=10]\n",
    "vocab_instances20_seqs_sel['name_config'] = vocab_instances20_seqs_sel.name + '+' + vocab_instances20_seqs_sel.seq_config.str.split('_seq',expand=True).iloc[:,0]\n",
    "print(vocab_instances20_seqs_sel.shape, vocab_instances20_seqs.shape)\n",
    "vocab_instances20_seqs_sel.to_csv(os.path.join(save_dir, 'vocab_instances20_seq_wscr_sel.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['vocab_pair', 'tot_dist', 'btn_dist', 'chr', 'start', 'stop',\n",
       "       'start_pad', 'stop_pad', 'vocab1', 'vocab1_start', 'vocab1_stop',\n",
       "       'vocab2', 'vocab2_start', 'vocab2_stop', 'tissue', 'loop_type',\n",
       "       'vocab_tissue_count', 'name', 'check_rs', 'scr_pass_rs', 'seq_config',\n",
       "       'seq', 'name_config'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 212,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_seqs_sel.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {},
   "outputs": [],
   "source": [
    "vocab_instances20_seqs_sel.to_csv(os.path.join(save_dir, 'vocab_instances20_seqs_sel.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  This is separate from the ipykernel package so we can avoid doing imports until\n"
     ]
    }
   ],
   "source": [
    "vocab_instances20_seqs_selcols = vocab_instances20_seqs_sel[['chr','start_pad','stop_pad','name_config','seq']]\n",
    "vocab_instances20_seqs_selcols.columns = ['chr','start','stop','name','seq']\n",
    "vocab_instances20_seqs_selcols['seq_type'] = 'vocab'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {},
   "outputs": [],
   "source": [
    "pos_genes_df_seqs_sel_cols = pos_genes_df_rs_filt[['chr','start','end','gene','sequence']]\n",
    "pos_genes_df_seqs_sel_cols.columns = ['chr','start','stop','name','seq']\n",
    "pos_genes_df_seqs_sel_cols['seq_type'] = 'pos'\n",
    "neg_genes_df_seqs_sel_cols = neg_genes_df_rs_filt[['chr','start','end','gene','sequence']]\n",
    "neg_genes_df_seqs_sel_cols.columns = ['chr','start','stop','name','seq']\n",
    "neg_genes_df_seqs_sel_cols['seq_type'] = 'neg'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>name</th>\n",
       "      <th>seq</th>\n",
       "      <th>seq_type</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>BARX2::FOXF2+Colon+1+A_B</td>\n",
       "      <td>CTCAGCTCTCCCACAAGATCTCTGGGGAAGCTGGAAGAGGGCCACT...</td>\n",
       "      <td>vocab</td>\n",
       "      <td>206848284</td>\n",
       "      <td>206848418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>chr4</td>\n",
       "      <td>BARX2::FOXF2+Colon+2+A_B</td>\n",
       "      <td>GTGGCTCTGAGGGAATTTCAAGTAACCGTGGACTAAAGTCTTAAGA...</td>\n",
       "      <td>vocab</td>\n",
       "      <td>38859561</td>\n",
       "      <td>38859695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>chr6</td>\n",
       "      <td>BARX2::FOXF2+Colon+3+A_B</td>\n",
       "      <td>GCCACTGCACTCCAGCCTGGTGACAGAGCGAGACTCTGTCTCAAGA...</td>\n",
       "      <td>vocab</td>\n",
       "      <td>74222732</td>\n",
       "      <td>74222866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>chr3</td>\n",
       "      <td>BARX2::FOXF2+Colon+4+A_B</td>\n",
       "      <td>AGTATTATGTGCTCATGGTCCCTACCCTCCAGGGTTCACAGTGTGG...</td>\n",
       "      <td>vocab</td>\n",
       "      <td>49384615</td>\n",
       "      <td>49384749</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>chr10</td>\n",
       "      <td>BARX2::FOXF2+Colon+5+A_B</td>\n",
       "      <td>ATGTGGCTGCGGGGCTTTAGGAGCTGACTCTTTGGTCACGGGGAGA...</td>\n",
       "      <td>vocab</td>\n",
       "      <td>112247461</td>\n",
       "      <td>112247595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>89</td>\n",
       "      <td>chrX</td>\n",
       "      <td>RHOXF2B</td>\n",
       "      <td>ACCACGCTCTCCTAGGCAGTGCTAGAACTGTGAGTCGCCGCCTGCA...</td>\n",
       "      <td>neg</td>\n",
       "      <td>119211572</td>\n",
       "      <td>119211707</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>90</td>\n",
       "      <td>chrX</td>\n",
       "      <td>CT47A12</td>\n",
       "      <td>AGTCTCTCTCTTTATTGAGGGAATAAGAGTCTGTCTCAGACGACAC...</td>\n",
       "      <td>neg</td>\n",
       "      <td>120066016</td>\n",
       "      <td>120066151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>91</td>\n",
       "      <td>chrX</td>\n",
       "      <td>CT47A6</td>\n",
       "      <td>AGTCTCTCTCTTTATTGAGGGAATAAGAGTCTGTCTCAGACGACAC...</td>\n",
       "      <td>neg</td>\n",
       "      <td>120095202</td>\n",
       "      <td>120095337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>92</td>\n",
       "      <td>chrX</td>\n",
       "      <td>SPANXN4</td>\n",
       "      <td>TGTCAGAGGTGGTTCTTTGTGATGCCAAAGCCACCCTAGGACTACC...</td>\n",
       "      <td>neg</td>\n",
       "      <td>142113569</td>\n",
       "      <td>142113704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>93</td>\n",
       "      <td>chrX</td>\n",
       "      <td>SPANXN2</td>\n",
       "      <td>TGTTCTTTTACCCAATGTCCATGTAGTTCCTTTAAGTTTCCTATTT...</td>\n",
       "      <td>neg</td>\n",
       "      <td>142804381</td>\n",
       "      <td>142804516</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9721 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chr                      name  \\\n",
       "0    chr1  BARX2::FOXF2+Colon+1+A_B   \n",
       "1    chr4  BARX2::FOXF2+Colon+2+A_B   \n",
       "2    chr6  BARX2::FOXF2+Colon+3+A_B   \n",
       "3    chr3  BARX2::FOXF2+Colon+4+A_B   \n",
       "4   chr10  BARX2::FOXF2+Colon+5+A_B   \n",
       "..    ...                       ...   \n",
       "89   chrX                   RHOXF2B   \n",
       "90   chrX                   CT47A12   \n",
       "91   chrX                    CT47A6   \n",
       "92   chrX                   SPANXN4   \n",
       "93   chrX                   SPANXN2   \n",
       "\n",
       "                                                  seq seq_type      start  \\\n",
       "0   CTCAGCTCTCCCACAAGATCTCTGGGGAAGCTGGAAGAGGGCCACT...    vocab  206848284   \n",
       "1   GTGGCTCTGAGGGAATTTCAAGTAACCGTGGACTAAAGTCTTAAGA...    vocab   38859561   \n",
       "2   GCCACTGCACTCCAGCCTGGTGACAGAGCGAGACTCTGTCTCAAGA...    vocab   74222732   \n",
       "3   AGTATTATGTGCTCATGGTCCCTACCCTCCAGGGTTCACAGTGTGG...    vocab   49384615   \n",
       "4   ATGTGGCTGCGGGGCTTTAGGAGCTGACTCTTTGGTCACGGGGAGA...    vocab  112247461   \n",
       "..                                                ...      ...        ...   \n",
       "89  ACCACGCTCTCCTAGGCAGTGCTAGAACTGTGAGTCGCCGCCTGCA...      neg  119211572   \n",
       "90  AGTCTCTCTCTTTATTGAGGGAATAAGAGTCTGTCTCAGACGACAC...      neg  120066016   \n",
       "91  AGTCTCTCTCTTTATTGAGGGAATAAGAGTCTGTCTCAGACGACAC...      neg  120095202   \n",
       "92  TGTCAGAGGTGGTTCTTTGTGATGCCAAAGCCACCCTAGGACTACC...      neg  142113569   \n",
       "93  TGTTCTTTTACCCAATGTCCATGTAGTTCCTTTAAGTTTCCTATTT...      neg  142804381   \n",
       "\n",
       "         stop  \n",
       "0   206848418  \n",
       "1    38859695  \n",
       "2    74222866  \n",
       "3    49384749  \n",
       "4   112247595  \n",
       "..        ...  \n",
       "89  119211707  \n",
       "90  120066151  \n",
       "91  120095337  \n",
       "92  142113704  \n",
       "93  142804516  \n",
       "\n",
       "[9721 rows x 6 columns]"
      ]
     },
     "execution_count": 216,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mpra_lib_seq = pd.concat([vocab_instances20_seqs_selcols,pos_genes_df_seqs_sel_cols, neg_genes_df_seqs_sel_cols ],axis=0,sort=True)\n",
    "mpra_lib_seq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### some stats then saving"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "vocab    9560\n",
       "neg        89\n",
       "pos        72\n",
       "Name: seq_type, dtype: int64"
      ]
     },
     "execution_count": 217,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mpra_lib_seq.seq_type.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GDSD6          49.0\n",
       "SCC            43.0\n",
       "Colon          42.0\n",
       "MM             40.0\n",
       "GM12878        39.0\n",
       "Melanocytes    26.0\n",
       "Name: tissue, dtype: float64"
      ]
     },
     "execution_count": 218,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_seqs_sel.tissue.value_counts()/40"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'irf_vocab' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-220-c1f9fcc06457>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mvocab_instances20_seqs_sel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvocab_pair\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mvocab_instances20_seqs_sel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvocab_pair\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mirf_vocab\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'irf_vocab' is not defined"
     ]
    }
   ],
   "source": [
    "vocab_instances20_seqs_sel.vocab_pair[vocab_instances20_seqs_sel.vocab_pair.isin(irf_vocab)].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1426,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(19,)"
      ]
     },
     "execution_count": 1426,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# number of vocab words that have irf vocab setup\n",
    "vocab_instances20_seqs_sel.vocab_pair[vocab_instances20_seqs_sel.vocab_pair.isin(irf_vocab)].unique().shape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "metadata": {},
   "outputs": [],
   "source": [
    "mpra_lib_seq.to_csv(os.path.join(save_dir, 'mpra_lib_seq.csv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 6. add MPRA adaptors and barcodes\n",
    "\n",
    "done in 8E1R (based on `make_library_psych-postcadd-01119.ipynb` from the mpra library\n",
    "\n",
    "note that the rs sites chosen from section 4B are: `[RS_ECORI ,RS_BAMHI , RS_XHOI , RS_NHEI]`\n",
    "\n",
    "input is `../data/processed/fig4_modelling/vocab_sum_final/mpra_lib_seq.csv`\n",
    "\n",
    "output is `../data/processed/fig4_modelling/vocab_sum_final/mpra_oligo_df_filt.csv`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# define constants\n",
    "LEN_FILLER = 20\n",
    "NUM_OLIGO_PER_MOTIF = 10\n",
    "DESIRED_OLIGO_LENGTH = 135 #(changed from 150 )\n",
    "MAX_OLIGO_LENGTH = 230\n",
    "MAX_MOTIF_LENGTH = 63\n",
    "FORWARD_PCR_PRIMER = 'ACTGGCCGCTTCACTG'\n",
    "REVERSE_PCR_PRIMER = 'AGATCGGAAGAGCGTCG'\n",
    "# LINKER_SPACER_v10 = 'CAGTATGCGATGCTCATGATTGTGACTGAAAGTCTATGAGTCCATTGA'\n",
    "RS_ECORI = 'GAATTC' # 5'-3'\n",
    "RS_BAMHI = 'GGATCC'\n",
    "RS_XHOI = 'CTCGAG'\n",
    "RS_NHEI = 'GCTAGC'\n",
    "\n",
    "\n",
    "RS_LIST = [RS_ECORI ,RS_BAMHI , RS_XHOI , RS_NHEI]\n",
    "\n",
    "LETTERS=['A', 'C', 'G', 'T']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 7. MPRA final check \n",
    "\n",
    "- checks to make sure all sequences have 10 barcodes\n",
    "- all rs sites are passed (done in R notebook)\n",
    "- that there are still 239 vocab words with 40 instance (4 config of 10 each)\n",
    "\n",
    "- `../data/processed/fig4_modelling/vocab_sum_final/mpra_lib_seq.csv` - includes information on all the sequences used\n",
    "- `../data/processed/fig4_modelling/vocab_sum_final/mpra_oligo_df_filt.csv` - oligo information with a lot of columns (RS site check results etc.(\n",
    "- `../data/processed/fig4_modelling/vocab_sum_final/mpra_oligo_df_final.txt` - with columns: name, seq_type, id, oligo, seq, filler, barcode\n",
    "- `../data/processed/fig4_modelling/vocab_sum_final/mpra_oligo_df_final_SUBMIT.txt` - with columns: id, oligo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1436,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['../data/processed/fig4_modelling/vocab_sum_final/']"
      ]
     },
     "execution_count": 1436,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glob.glob(os.path.join(save_dir))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(97210, 21)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>seq</th>\n",
       "      <th>seq_type</th>\n",
       "      <th>count</th>\n",
       "      <th>id</th>\n",
       "      <th>pass_restrict_seq</th>\n",
       "      <th>pass_restrict_fwd_seq</th>\n",
       "      <th>pass_restrict_seq_xho</th>\n",
       "      <th>barcode</th>\n",
       "      <th>len_seq</th>\n",
       "      <th>...</th>\n",
       "      <th>1:n()</th>\n",
       "      <th>filler</th>\n",
       "      <th>filler_check</th>\n",
       "      <th>oligo</th>\n",
       "      <th>len_oligo</th>\n",
       "      <th>pass_restrict_ECORI</th>\n",
       "      <th>pass_restrict_BAMHI</th>\n",
       "      <th>pass_restrict_XHOI</th>\n",
       "      <th>pass_restrict_NHEI</th>\n",
       "      <th>pass_all</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>ABCF3</td>\n",
       "      <td>ACCTTGAGCTCCCAGCCTTTGCCCCACCCCTCGCTCCCGGAACTCC...</td>\n",
       "      <td>pos</td>\n",
       "      <td>1</td>\n",
       "      <td>ABCF3.1</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>GACCGGAGTAGGTTAGGTTA</td>\n",
       "      <td>136</td>\n",
       "      <td>...</td>\n",
       "      <td>8</td>\n",
       "      <td>CCGCCATTGTCTCGGGCGAA</td>\n",
       "      <td>True</td>\n",
       "      <td>ACTGGCCGCTTCACTGACCTTGAGCTCCCAGCCTTTGCCCCACCCC...</td>\n",
       "      <td>221</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>ABCF3</td>\n",
       "      <td>ACCTTGAGCTCCCAGCCTTTGCCCCACCCCTCGCTCCCGGAACTCC...</td>\n",
       "      <td>pos</td>\n",
       "      <td>2</td>\n",
       "      <td>ABCF3.2</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>GCTGGCGAATATAAGAGTAT</td>\n",
       "      <td>136</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>GCCCTATTATATGTGCGTCC</td>\n",
       "      <td>True</td>\n",
       "      <td>ACTGGCCGCTTCACTGACCTTGAGCTCCCAGCCTTTGCCCCACCCC...</td>\n",
       "      <td>221</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>ABCF3</td>\n",
       "      <td>ACCTTGAGCTCCCAGCCTTTGCCCCACCCCTCGCTCCCGGAACTCC...</td>\n",
       "      <td>pos</td>\n",
       "      <td>3</td>\n",
       "      <td>ABCF3.3</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>TGAGATCCAACCAAGGTTGC</td>\n",
       "      <td>136</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>ACAAGGTAGGAGGGGCAGAG</td>\n",
       "      <td>True</td>\n",
       "      <td>ACTGGCCGCTTCACTGACCTTGAGCTCCCAGCCTTTGCCCCACCCC...</td>\n",
       "      <td>221</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>ABCF3</td>\n",
       "      <td>ACCTTGAGCTCCCAGCCTTTGCCCCACCCCTCGCTCCCGGAACTCC...</td>\n",
       "      <td>pos</td>\n",
       "      <td>4</td>\n",
       "      <td>ABCF3.4</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>AAGATTAATCGGAGGAGTTG</td>\n",
       "      <td>136</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>CCTAAAATATGGACTTACCG</td>\n",
       "      <td>True</td>\n",
       "      <td>ACTGGCCGCTTCACTGACCTTGAGCTCCCAGCCTTTGCCCCACCCC...</td>\n",
       "      <td>221</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>ABCF3</td>\n",
       "      <td>ACCTTGAGCTCCCAGCCTTTGCCCCACCCCTCGCTCCCGGAACTCC...</td>\n",
       "      <td>pos</td>\n",
       "      <td>5</td>\n",
       "      <td>ABCF3.5</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>GAAGTACTCTCCTGCGAGGA</td>\n",
       "      <td>136</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>AGGACCTCCGTTGGCAATTC</td>\n",
       "      <td>True</td>\n",
       "      <td>ACTGGCCGCTTCACTGACCTTGAGCTCCCAGCCTTTGCCCCACCCC...</td>\n",
       "      <td>221</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    name                                                seq seq_type  count  \\\n",
       "1  ABCF3  ACCTTGAGCTCCCAGCCTTTGCCCCACCCCTCGCTCCCGGAACTCC...      pos      1   \n",
       "2  ABCF3  ACCTTGAGCTCCCAGCCTTTGCCCCACCCCTCGCTCCCGGAACTCC...      pos      2   \n",
       "3  ABCF3  ACCTTGAGCTCCCAGCCTTTGCCCCACCCCTCGCTCCCGGAACTCC...      pos      3   \n",
       "4  ABCF3  ACCTTGAGCTCCCAGCCTTTGCCCCACCCCTCGCTCCCGGAACTCC...      pos      4   \n",
       "5  ABCF3  ACCTTGAGCTCCCAGCCTTTGCCCCACCCCTCGCTCCCGGAACTCC...      pos      5   \n",
       "\n",
       "        id  pass_restrict_seq  pass_restrict_fwd_seq  pass_restrict_seq_xho  \\\n",
       "1  ABCF3.1               True                   True                   True   \n",
       "2  ABCF3.2               True                   True                   True   \n",
       "3  ABCF3.3               True                   True                   True   \n",
       "4  ABCF3.4               True                   True                   True   \n",
       "5  ABCF3.5               True                   True                   True   \n",
       "\n",
       "                barcode  len_seq  ...  1:n()                filler  \\\n",
       "1  GACCGGAGTAGGTTAGGTTA      136  ...      8  CCGCCATTGTCTCGGGCGAA   \n",
       "2  GCTGGCGAATATAAGAGTAT      136  ...     14  GCCCTATTATATGTGCGTCC   \n",
       "3  TGAGATCCAACCAAGGTTGC      136  ...      5  ACAAGGTAGGAGGGGCAGAG   \n",
       "4  AAGATTAATCGGAGGAGTTG      136  ...      1  CCTAAAATATGGACTTACCG   \n",
       "5  GAAGTACTCTCCTGCGAGGA      136  ...      4  AGGACCTCCGTTGGCAATTC   \n",
       "\n",
       "  filler_check                                              oligo len_oligo  \\\n",
       "1         True  ACTGGCCGCTTCACTGACCTTGAGCTCCCAGCCTTTGCCCCACCCC...       221   \n",
       "2         True  ACTGGCCGCTTCACTGACCTTGAGCTCCCAGCCTTTGCCCCACCCC...       221   \n",
       "3         True  ACTGGCCGCTTCACTGACCTTGAGCTCCCAGCCTTTGCCCCACCCC...       221   \n",
       "4         True  ACTGGCCGCTTCACTGACCTTGAGCTCCCAGCCTTTGCCCCACCCC...       221   \n",
       "5         True  ACTGGCCGCTTCACTGACCTTGAGCTCCCAGCCTTTGCCCCACCCC...       221   \n",
       "\n",
       "   pass_restrict_ECORI  pass_restrict_BAMHI  pass_restrict_XHOI  \\\n",
       "1                 True                 True                True   \n",
       "2                 True                 True                True   \n",
       "3                 True                 True                True   \n",
       "4                 True                 True                True   \n",
       "5                 True                 True                True   \n",
       "\n",
       "   pass_restrict_NHEI  pass_all  \n",
       "1                True      True  \n",
       "2                True      True  \n",
       "3                True      True  \n",
       "4                True      True  \n",
       "5                True      True  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 226,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oligo_df_filt = pd.read_csv(os.path.join(save_dir,'mpra_oligo_df_filt.csv'),index_col=0)\n",
    "print(oligo_df_filt.shape)\n",
    "oligo_df_filt[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Series([], Name: name, dtype: int64)"
      ]
     },
     "execution_count": 227,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq_counts = oligo_df_filt.name.value_counts()\n",
    "seq_counts[seq_counts!=10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " all sequences have 10 barcodes\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "rs sites checks (done in R notebook)\n",
    "- no EcoRI \n",
    "- no BamHI sites\n",
    "- exactly 1 XhoI site\n",
    "- exactly 1 NheI site\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/pandas/core/frame.py:3494: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self[k1] = value[k2]\n"
     ]
    }
   ],
   "source": [
    "vocab_oligo_df =  oligo_df_filt[oligo_df_filt.seq_type=='vocab']\n",
    "    \n",
    "vocab_oligo_df[['vocab_pair','tissue','vocab_tissue_count','seq_config']]= vocab_oligo_df.name.str.split('+',expand=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Series([], Name: vocab_pair, dtype: int64)"
      ]
     },
     "execution_count": 229,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_oligo_counts = vocab_oligo_df.vocab_pair.value_counts()\n",
    "vocab_oligo_counts[vocab_oligo_counts!=400]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " all vocab types have 4 configs x 10 genomic instances x 10 barcodes = 400 oligos\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(239,)"
      ]
     },
     "execution_count": 230,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_oligo_counts.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "there are 239 vocab words"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7B motif scan\n",
    "\n",
    "### 1. make fasta \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(9560, 2)\n"
     ]
    }
   ],
   "source": [
    "fasta_file = os.path.join(save_dir, 'mpra_vocab_seq2.fasta')\n",
    "print(vocab_oligo_df[['name','seq']].drop_duplicates().shape)\n",
    "with open(fasta_file, 'w') as f:\n",
    "    for idx, row in vocab_oligo_df[['name','seq']].drop_duplicates().iterrows():\n",
    "        f.write('>'+row['name']+'\\n')\n",
    "        f.write(row['seq']+'\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 321,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>ABCF3</td>\n",
       "      <td>ACCTTGAGCTCCCAGCCTTTGCCCCACCCCTCGCTCCCGGAACTCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11</td>\n",
       "      <td>ADM</td>\n",
       "      <td>GCTCCTCTCGGCTTTGCTCCGTCCCTGCCCCGGCCCCTCCGCGGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>21</td>\n",
       "      <td>ALDH16A1</td>\n",
       "      <td>ATAAATCCGGAGCCCTCCTAACTATTCGCAAAGTACCCACATCGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>31</td>\n",
       "      <td>ANAPC15</td>\n",
       "      <td>CTTACCGCGCCGGGAGGCTGCTGCCGGCGGCGACCCGGAAGCGCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>41</td>\n",
       "      <td>ANXA2</td>\n",
       "      <td>CCCAGCAGGGCTCTGACCTACCTGGGCAGCCTCATTCCCCCACATC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>97161</td>\n",
       "      <td>ZNF580</td>\n",
       "      <td>GAGTAGCAAGGTGCGCGACGCAGGAACCTTTCTCTTGTGGGTGCGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>97171</td>\n",
       "      <td>ZNF581</td>\n",
       "      <td>GGGAGAAGGGGCTATTTCCCATGGGAAGACAAAGTGGGATGAATAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>97181</td>\n",
       "      <td>ZNF679</td>\n",
       "      <td>CAGAAATATGCCACAATCCCCAATGGAGGCCAAGACAAGGCAAGTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>97191</td>\n",
       "      <td>ZNF839</td>\n",
       "      <td>AACTGCGGAATCTTTTGGACTGGCCCAGAGGTTATCCAAGGCTACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>97201</td>\n",
       "      <td>ZNF99</td>\n",
       "      <td>AGCCCCTTCCCCTTCTCAGGATATCGGACCCGGCACTCTCACCATT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>161 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           name                                                seq\n",
       "1         ABCF3  ACCTTGAGCTCCCAGCCTTTGCCCCACCCCTCGCTCCCGGAACTCC...\n",
       "11          ADM  GCTCCTCTCGGCTTTGCTCCGTCCCTGCCCCGGCCCCTCCGCGGGA...\n",
       "21     ALDH16A1  ATAAATCCGGAGCCCTCCTAACTATTCGCAAAGTACCCACATCGCC...\n",
       "31      ANAPC15  CTTACCGCGCCGGGAGGCTGCTGCCGGCGGCGACCCGGAAGCGCTT...\n",
       "41        ANXA2  CCCAGCAGGGCTCTGACCTACCTGGGCAGCCTCATTCCCCCACATC...\n",
       "...         ...                                                ...\n",
       "97161    ZNF580  GAGTAGCAAGGTGCGCGACGCAGGAACCTTTCTCTTGTGGGTGCGT...\n",
       "97171    ZNF581  GGGAGAAGGGGCTATTTCCCATGGGAAGACAAAGTGGGATGAATAG...\n",
       "97181    ZNF679  CAGAAATATGCCACAATCCCCAATGGAGGCCAAGACAAGGCAAGTC...\n",
       "97191    ZNF839  AACTGCGGAATCTTTTGGACTGGCCCAGAGGTTATCCAAGGCTACC...\n",
       "97201     ZNF99  AGCCCCTTCCCCTTCTCAGGATATCGGACCCGGCACTCTCACCATT...\n",
       "\n",
       "[161 rows x 2 columns]"
      ]
     },
     "execution_count": 321,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 323,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(161, 2)\n"
     ]
    }
   ],
   "source": [
    "fasta_control_file = os.path.join(save_dir, 'mpra_control_seq.fasta')\n",
    "control_oligo_df = oligo_df_filt[oligo_df_filt.seq_type.isin(['pos','neg'])]\n",
    "print(control_oligo_df[['name','seq']].drop_duplicates().shape)\n",
    "with open(fasta_control_file, 'w') as f:\n",
    "    for idx, row in control_oligo_df[['name','seq']].drop_duplicates().iterrows():\n",
    "        f.write('>'+row['name']+'\\n')\n",
    "        f.write(row['seq']+'\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 339,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(89, 2)\n"
     ]
    }
   ],
   "source": [
    "fasta_neg_file = os.path.join(save_dir, 'mpra_neg_control_seq.fasta')\n",
    "neg_control_oligo_df = oligo_df_filt[oligo_df_filt.seq_type.isin(['neg'])].sort_values('name')\n",
    "print(neg_control_oligo_df[['name','seq']].drop_duplicates().shape)\n",
    "with open(fasta_neg_file, 'w') as f:\n",
    "    for idx, row in neg_control_oligo_df[['name','seq']].drop_duplicates().iterrows():\n",
    "        f.write('>'+row['name']+'\\n')\n",
    "        f.write(row['seq']+'\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 341,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(72, 2)\n"
     ]
    }
   ],
   "source": [
    "fasta_pos_file = os.path.join(save_dir, 'mpra_pos_control_seq.fasta')\n",
    "pos_control_oligo_df = oligo_df_filt[oligo_df_filt.seq_type.isin(['pos'])].sort_values('name')\n",
    "print(pos_control_oligo_df[['name','seq']].drop_duplicates().shape)\n",
    "with open(fasta_pos_file, 'w') as f:\n",
    "    for idx, row in pos_control_oligo_df[['name','seq']].drop_duplicates().iterrows():\n",
    "        f.write('>'+row['name']+'\\n')\n",
    "        f.write(row['seq']+'\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### scanning try 1: RSAT FAILED\n",
    "\n",
    "\n",
    "http://rsat.sb-roscoff.fr/matrix-scan.cgi\n",
    "\n",
    "\n",
    "\n",
    "Server command\n",
    "$RSAT/perl-scripts/matrix-scan -v 1   -quick -matrix_format transfac -m $RSAT/public_html/tmp/apache/2020/09/02/matrix-scan_2020-09-02.015859_5S2uJF.matrix -pseudo 1 -decimals 1 -2str -origin end -bginput -markov 1 -bg_pseudo 0.01 -return limits -return sites -lth score 1 -i $RSAT/public_html/tmp/apache/2020/09/02/tmp_sequence_2020-09-02.015859_VpAOG3.fasta -seq_format fasta -n score \n",
    "\n",
    "[RSAT] matrix-scan 2020-09-02.020844\n",
    "\n",
    "\n",
    "Command: nice -n 19 $RSAT/perl-scripts/matrix-scan -v 1   -quick -matrix_format transfac -m $RSAT/public_html/tmp/apache/2020/09/02/matrix-scan_2020-09-02.020844_AC8494.matrix -pseudo 1 -decimals 1 -2str -origin end -bginput -markov 1 -bg_pseudo 0.01 -return limits -return sites -lth score 1 -i $RSAT/public_html/tmp/apache/2020/09/02/tmp_sequence_2020-09-02.020844_IedJeb.fasta -seq_format fasta -n score\n",
    "\n",
    "Your task has been submitted to the RSAT server.\n",
    "\n",
    "After completion of the task, an email will be sent to mguo123@stanford.edu\n",
    "\n",
    "The result will become available at\n",
    "\thttp://rsat.sb-roscoff.fr/tmp/apache/2020/09/02/matrix-scan_2020-09-02.020844_AC8494.ft\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### scanning try 2: meme suite failed\n",
    "\n",
    "also try http://meme-suite.org/info/status?service=MAST&id=appMAST_5.1.11599005652624-545759054\n",
    "    \n",
    "    Status Messages\n",
    "Parsing arguments\n",
    "Arguments ok\n",
    "Starting mast\n",
    "mast -oc . -nostatus -minseqs 9560 -ev 10.0 HOCOMOCOv11_core_HUMAN_mono_meme_format.meme sequences.fa\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "try: \n",
    "http://meme-suite.org/info/status?service=FIMO&id=appFIMO_5.1.11599763721265-808903115\n",
    "`fimo --oc . --verbosity 1 --thresh 1.0E-4 motifs.meme mpra_vocab_seq2.fasta`\n",
    "\n",
    "enter `../data/external/HOCOMOCOv11_core_HUMAN_mono_jaspar_format.txt`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### scanning try 3 MOODS package\n",
    "https://github.com/jhkorhonen/MOODS/wiki/Getting-started\n",
    "Matrices. The matrices are expected in the JASPAR .pfm format or the .adm format for first-order models. A JASPAR .pfm file consist of four rows, specifying the counts or frequencies for nucleotides A, C, G and T, respectively:\n",
    "\n",
    "#### DESCRIPTION\n",
    "\n",
    "MOODS 1.9.1 introduced a Python command-line script moods-dna.py for standalone PWM analysis using the MOODS algorithms. This section explains the basics of using `moods-dna.py`\n",
    "\n",
    "#### Input matrics (motifs)\n",
    "pfms are in `/Users/mguo123/Google Drive/1_khavari/omics_project-LD/pan_omics/data/external/hocomoco_human_trim`\n",
    "Note: the extensions are pwm but are pfms\n",
    "\n",
    "#### Input sequences\n",
    "Sequences. The sequences can be given either as plain text files, or as fasta files possibly containing multiple sequences. Newlines and leading and trailing whitespace will be ignored, but other characters not encoding nucleotides will be treated as non-matchable positions (i.e., they should not appear outside fasta headers). IUPAC nucleotide ambiguity codes (WSMKRYBDHV) will be treated as coding SNPs, and MOODS will match the matrices versus all possible combinations.\n",
    "\n",
    "#### Output (csv)\n",
    "The comma-separated fields are:\n",
    "\n",
    "- Sequence name (either file name or fasta header).\n",
    "- Matrix file name.\n",
    "- Hit position (first position of sequence is position 0).\n",
    "- Indicates whether the match is versus the input strand (+) or the reverse complement (-).\n",
    "- Match score.\n",
    "- Hit site in the original sequence (note that if the match is -, then the match is versus the reverse complement of this).\n",
    "- Hit sequence with SNPs applied, i.e. shows how the ambiguity codes are resolved to get this score (there can be multiple hits at the same position with different \"real\" hit sequences).\n",
    "\n",
    "\n",
    "#### Installation \n",
    "\n",
    "follow: https://pypi.org/project/MOODS-python/\n",
    "\n",
    "this doesn't work with subprocess\n",
    "```\n",
    "pip install --user MOODS-python\n",
    "PATH=\"$PATH:`python -m site --user-base`/bin\"\n",
    "```\n",
    "\n",
    "#### Basic usage\n",
    "The basic usage pattern of `moods-dna.py` is the following:\n",
    "\n",
    "```\n",
    "moods-dna.py --help\n",
    "usage: moods-dna.py [-h] [-v] [-m M [M ...]] [-S M [M ...]] [-s S [S ...]]\n",
    "                    [-p p] [-t T] [-B n] [-o outfile] [--sep S] [-R]\n",
    "                    [--no-snps] [--batch] [--bg pA pC pG pT] [--ps p]\n",
    "                    [--log-base x] [--lo-bg pA pC pG pT]\n",
    "                    [--threshold-precision x]\n",
    "\n",
    "matches position weight matrices against DNA sequences\n",
    "\n",
    "optional arguments:\n",
    "  -h, --help            show this help message and exit\n",
    "  -v, --verbosity       verbose (-vv, -vvv for more)\n",
    "\n",
    "input files (at least one matrix and sequence file required):\n",
    "  -m M [M ...], --matrices M [M ...]\n",
    "                        matrix files (count/frequency, will be converted to\n",
    "                        PWM before matching)\n",
    "  -S M [M ...], --score-matrices M [M ...]\n",
    "                        matrix files (PWM/other score matrix, will be matched\n",
    "                        directly)\n",
    "  -s S [S ...], --sequences S [S ...]\n",
    "                        sequence files\n",
    "\n",
    "threshold selection (exactly one required):\n",
    "  -p p, --p-value p     compute threshold from p-value\n",
    "  -t T, --threshold T   use specified absolute threshold\n",
    "  -B n, --best-hits n   return at least the specified amount of best matches\n",
    "\n",
    "output (optional):\n",
    "  -o outfile, --outfile outfile\n",
    "                        output to file instead of standard output\n",
    "  --sep S               set field separator in out (default \",\")\n",
    "\n",
    "search and model behaviour (optional):\n",
    "  -R, --no-rc           disable matching versus reverse complement strand\n",
    "  --no-snps             ignore IUPAC symbols coding multiple nucleotides\n",
    "  --batch               do not recompute thresholds from p-values for each\n",
    "                        sequence separately (recommended when dealing with\n",
    "                        lots of short sequences)\n",
    "  --bg pA pC pG pT      background distribution for computing thresholds from\n",
    "                        p-value with --batch (default is 0.25 for all alleles)\n",
    "  --ps p                total pseudocount added to each matrix column in log-\n",
    "                        odds conversion (default = 0.01)\n",
    "  --log-base x          logarithm base for log-odds conversion (default\n",
    "                        natural logarithm)\n",
    "  --lo-bg pA pC pG pT   background distribution for log-odds conversion\n",
    "                        (default is 0.25 for all alleles)\n",
    "  --threshold-precision x\n",
    "                        specify the precision used for computing the\n",
    "                        thresholds from p-values (default = 2000.0)\n",
    "                        \n",
    "                        \n",
    "moods-dna.py -m example-data/matrices/*.{pfm,adm} -s example-data/seq/chr1-5k-55k.fa -p 0.0001\n",
    "moods-dna.py -m ../data/external/hocomoco_human_trim/*.pwm -s ../data/processed/fig4_modelling/vocab_sum_final/mpra_vocab_seq.fasta -p 0.0001\n",
    "\n",
    "moods-dna.py -m ../data/external/hocomoco_human_trim/FOXM1_HUMAN.H11MO.0.A.pwm.trim.pwm -s ../data/processed/fig4_modelling/vocab_sum_final/mpra_vocab_seq.fasta -p 0.0001\n",
    "\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1445,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # !pip install --user MOODS-python\n",
    "# !PATH=\"$PATH:`python -m site --user-base`/bin\"\n",
    "# glob.glob(os.path.join(save_dir,\"*fasta\"))\n",
    "import subprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 325,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "123"
      ]
     },
     "execution_count": 325,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# num actual tfs used \n",
    "list_tfs = sorted(set(list(vocab_instances20_alltissue_df_rs_filt.vocab1.unique()) + list(vocab_instances20_alltissue_df_rs_filt.vocab2.unique())))\n",
    "len(list_tfs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### for vocab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1687,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make directory\n",
    "motif_path = os.path.join(save_dir, 'motif_scan3') #../data/processed/fig4_modelling/vocab_sum_final/motif_scan3\n",
    "\n",
    "if not os.path.exists(motif_path):\n",
    "    os.makedirs(motif_path)\n",
    "print(motif_path)\n",
    "\n",
    "input_fasta = '../data/processed/fig4_modelling/vocab_sum_final/mpra_vocab_seq2.fasta'\n",
    "# run_file = 'run_8E1_make_mpra_7_motif_scan.sh'\n",
    "run_file = 'run_8E1_make_mpra_7_motif_scan_vocab_filt3.sh'\n",
    "\n",
    "with open(run_file, 'w') as f:\n",
    "# print(len(glob.glob('../data/external/hocomoco_human_trim/*pwm')))\n",
    "    for pwm_file in sorted(glob.glob('../data/external/hocomoco_human_trim/*pwm')):\n",
    "        pwm = os.path.basename(pwm_file).split('.pwm')[0]\n",
    "        tf = tf_id_abbr_to_name_dict[pwm]\n",
    "        if tf in list_tfs:\n",
    "            output_file = os.path.join(motif_path, pwm+'_scan.csv')\n",
    "#             cmd = 'moods-dna.py -m {} -s {} -p 0.00005 -o {}'.format(pwm_file,input_fasta, output_file) \n",
    "#             cmd = 'moods-dna.py -m {} -s {} -p 0.001 -o {}'.format(pwm_file,input_fasta, output_file) #motif_scan and motif_scan2 use this, the difference is in scrambling methods (see final analysis step)\n",
    "            cmd = 'moods-dna.py -m {} -s {} -p 0.05 -o {}'.format(pwm_file,input_fasta, output_file) #motif_scan3 uses this, difference from motif_scan2 is the pvalue\n",
    "            f.write(cmd+'\\n')\n",
    "        \n",
    "# testing\n",
    "# cmd = 'moods-dna.py -m {} -s {} -p 0.00005 -o {}'.format(pwm_file,input_fasta, output_file)\n",
    "cmd = 'moods-dna.py -m {} -s {} -p 0.05 -o {}'.format(pwm_file,input_fasta, output_file)\n",
    "print(cmd)\n",
    "#moods-dna.py -m ../data/external/hocomoco_human_trim/ZSC31_HUMAN.H11MO.0.C.pwm.trim.pwm -s ../data/processed/fig4_modelling/vocab_sum_final/mpra_vocab_seq2.fasta -p 0.05 -o ../data/processed/fig4_modelling/vocab_sum_final/motif_scan3/ZN816_HUMAN.H11MO.1.C_scan.csv\n",
    "\n",
    "# subprocess.Popen(cmd.split())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1688,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "moods-dna.py -m ../data/external/hocomoco_human_trim/ZSC31_HUMAN.H11MO.0.C.pwm.trim.pwm -s ../data/processed/fig4_modelling/vocab_sum_final/mpra_vocab_seq2.fasta -p 0.05 -o ../data/processed/fig4_modelling/vocab_sum_final/motif_scan3/ZN816_HUMAN.H11MO.1.C_scan.csv\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### for controls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 338,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../data/processed/fig4_modelling/vocab_sum_final/motif_scan_control_pval001\n",
      "moods-dna.py -m ../data/external/hocomoco_human_trim/ZSC31_HUMAN.H11MO.0.C.pwm.trim.pwm -s ../data/processed/fig4_modelling/vocab_sum_final/mpra_control_seq.fasta -p 0.05 -o ../data/processed/fig4_modelling/vocab_sum_final/motif_scan_control_pval001/ZSC31_HUMAN.H11MO.0.C_scan.csv\n"
     ]
    }
   ],
   "source": [
    "# make directory\n",
    "# motif_path = os.path.join(save_dir, 'motif_scan_control')\n",
    "motif_path = os.path.join(save_dir, 'motif_scan_control_pval001')\n",
    "\n",
    "if not os.path.exists(motif_path):\n",
    "    os.makedirs(motif_path)\n",
    "print(motif_path)\n",
    "input_fasta = '../data/processed/fig4_modelling/vocab_sum_final/mpra_control_seq.fasta'\n",
    "run_file = 'run_8E1_make_mpra_7_motif_scan_control2.sh'\n",
    "\n",
    "with open(run_file, 'w') as f:\n",
    "# print(len(glob.glob('../data/external/hocomoco_human_trim/*pwm')))\n",
    "    for pwm_file in sorted(glob.glob('../data/external/hocomoco_human_trim/*pwm')):\n",
    "        pwm = os.path.basename(pwm_file).split('.pwm')[0]\n",
    "        tf = tf_id_abbr_to_name_dict[pwm]\n",
    "#         if tf in list_tfs:\n",
    "        output_file = os.path.join(motif_path, pwm+'_scan.csv')\n",
    "#         cmd = 'moods-dna.py -m {} -s {} -p 0.05 -o {}'.format(pwm_file,input_fasta, output_file) #motif_scan_control\n",
    "        cmd = 'moods-dna.py -m {} -s {} -p 0.001 -o {}'.format(pwm_file,input_fasta, output_file) #motif_scan_control_pval001 \n",
    "        f.write(cmd+'\\n')\n",
    "# testing\n",
    "# cmd = 'moods-dna.py -m {} -s {} -p 0.00005 -o {}'.format(pwm_file,input_fasta, output_file)\n",
    "cmd = 'moods-dna.py -m {} -s {} -p 0.05 -o {}'.format(pwm_file,input_fasta, output_file)\n",
    "print(cmd)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "in cmd line in this folder run\n",
    "`bash run_8E1_make_mpra_7_motif_scan_vocab_filt.sh`\n",
    "`bash run_8E1_make_mpra_7_motif_scan_vocab_filt2.sh`\n",
    "`bash run_8E1_make_mpra_7_motif_scan_vocab_filt3.sh`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "in sherlock\n",
    "```\n",
    "cd /oak/stanford/groups/khavari/users/mguo123/pan_omics/notebooks\n",
    "\n",
    "sbatch run_8E1_make_mpra_7_motif_scan_vocab_filt.sbatch\n",
    "\n",
    "\n",
    "\n",
    "#!/bin/bash\n",
    "#run_8E1_make_mpra_7_motif_scan.sbatch\n",
    "#09/01/2020\n",
    "#SBATCH -J motif_filt# A single job name for the array\n",
    "#SBATCH -p rbaltman,khavari # Partition\n",
    "#SBATCH -n 1 # one core\n",
    "#SBATCH --mem=64000\n",
    "#SBATCH --time=48:00:00\n",
    "#SBATCH -o run_8E1_filt.out # Standard output\n",
    "#SBATCH -e run_8E1_filt.err # Standard error\n",
    "#SBATCH --mail-type=FAIL,END # notifications for job done & fail\n",
    "#SBATCH --mail-user=mguo123@stanford.edu\n",
    "\n",
    "cd /oak/stanford/groups/khavari/users/mguo123/pan_omics/notebooks\n",
    "\n",
    "bash run_8E1_make_mpra_7_motif_scan_vocab_filt.sh\n",
    "\n",
    "```\n",
    "\n",
    "#### header of the `*_scan.csv`\n",
    "   `ARID5B::KLF5+GDSD6+4+A_B,SRBP2_HUMAN.H11MO.0.B.pwm.trim.pwm,73,-,5.35564936116,CCACCCCGC,`\n",
    "cvs headers (8 cols)\n",
    "- vocabulary config name (from fasta)\n",
    "- pwm filename\n",
    "- position in the sequence (index 0)\n",
    "- strand: Indicates whether the match is versus the input strand (+) or the reverse complement (-).\n",
    "- score (bit score)\n",
    "- sequence substring from what we are scanning (original sequence) (note that if the match is -, then the match is versus the reverse complement of this).\n",
    "- (NOT USED)Hit sequence with SNPs applied, i.e. shows how the ambiguity codes are resolved to get this score (there can be multiple hits at the same position with different \"real\" hit sequences).\n",
    "\n",
    "\n",
    "#### more on the score\n",
    "    https://www.cs.helsinki.fi/group/pssmfind/doc/python/MOODS.html\n",
    "      The input PWM matrix transformed to log-odds scores.\n",
    "      The score for nucleotide N and position i is\n",
    "\n",
    "         log ( matrix[N][i] + ps * bg[N] ) / C[i]\n",
    "           - log ( bg[N] ),\n",
    "\n",
    "      where C[i] is the sum of terms\n",
    "\n",
    "         matrix[N][i] + ps * bg[N]\n",
    "\n",
    "      for all nucleotides N.\n",
    "\n",
    "    so all the matches to the PWM transformed log-odds score are summed to get the match score\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1455,
   "metadata": {},
   "outputs": [],
   "source": [
    "# x= vocab_instances20_alltissue_df_rs_filt[vocab_instances20_alltissue_df_rs_filt.name=='BARX2::FOXF2+Colon+1']\n",
    "# x.start_pad - x.start    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# vocab_oligo_df[vocab_oligo_df.name=='FLI1::FOXM1+Colon+9+A_B']\n",
    "# vocab_instances20_alltissue_df_rs_filt[vocab_instances20_alltissue_df_rs_filt.name=='FLI1::FOXM1+Colon+9']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1454,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>vocab_pair</th>\n",
       "      <th>tot_dist</th>\n",
       "      <th>btn_dist</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>vocab1</th>\n",
       "      <th>vocab1_start</th>\n",
       "      <th>vocab1_stop</th>\n",
       "      <th>vocab2</th>\n",
       "      <th>...</th>\n",
       "      <th>stop_pad</th>\n",
       "      <th>vocab_tissue_count</th>\n",
       "      <th>name</th>\n",
       "      <th>sequence</th>\n",
       "      <th>check_rs</th>\n",
       "      <th>scr_pass_rs</th>\n",
       "      <th>A_B_seq</th>\n",
       "      <th>scrA_B_seq</th>\n",
       "      <th>A_scrB_seq</th>\n",
       "      <th>scrA_scrB_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [vocab_pair, tot_dist, btn_dist, chr, start, stop, vocab1, vocab1_start, vocab1_stop, vocab2, vocab2_start, vocab2_stop, tissue, loop_type, start_pad, stop_pad, vocab_tissue_count, name, sequence, check_rs, scr_pass_rs, A_B_seq, scrA_B_seq, A_scrB_seq, scrA_scrB_seq]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 25 columns]"
      ]
     },
     "execution_count": 1454,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_alltissue_df_rs_filt[vocab_instances20_alltissue_df_rs_filt.name.str.contains('COT1')][:60]#.vocab_pair.unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7C read in motif scan data \n",
    "\n",
    "cases for the motif in the \n",
    "1. scrambling removes motif \n",
    "2. there's a dummy motif somewhere else in the genomic instance (????? How would you deal with these “noisy” cases?\n",
    "    2A. is the dummy motif is present in all 4 cases, (not changed by scrambling)\n",
    "    2B. the dummy motif is present in only some cases, so probably changed by scrambling the other site?\n",
    "3. scrambling reintroduces the motif at a (hopefully) lower strength, and sometimes with a shifted start position  ??? Potential to rescramble (but sometimes motifs have too many repetitive nucleotides so this might not work…\n",
    "    3A. the scrambling introduces the motif at a lower strength\n",
    "    3B. the scrambling does not lower the score of the motif\n",
    "4. we don't detect the motif in any config in the genomic instance at a certain p value threshold (this usually isn’t the case, if we raise the p-value threshold enough; but the motif will be detected in a lot more genomic instances). This is sort of the sensitivity vs specificity tradeoff.\n",
    "\n",
    "\n",
    "2 parts\n",
    "1. map the vocab word locations to motifs and match to case above (get the score and the case) - exact match\n",
    "2. determine vocab words gained lost between the motif configurations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 7C1: get vocab_instance_df set up with locations and vocab predictions given config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>vocab_pair</th>\n",
       "      <th>tot_dist</th>\n",
       "      <th>btn_dist</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>start_pad</th>\n",
       "      <th>stop_pad</th>\n",
       "      <th>vocab1</th>\n",
       "      <th>vocab1_start</th>\n",
       "      <th>...</th>\n",
       "      <th>vocab2_stop</th>\n",
       "      <th>tissue</th>\n",
       "      <th>loop_type</th>\n",
       "      <th>vocab_tissue_count</th>\n",
       "      <th>name</th>\n",
       "      <th>check_rs</th>\n",
       "      <th>scr_pass_rs</th>\n",
       "      <th>seq_config</th>\n",
       "      <th>seq</th>\n",
       "      <th>name_config</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>BARX2::FOXF2</td>\n",
       "      <td>28</td>\n",
       "      <td>14</td>\n",
       "      <td>chr1</td>\n",
       "      <td>206848337</td>\n",
       "      <td>206848365</td>\n",
       "      <td>206848284</td>\n",
       "      <td>206848418</td>\n",
       "      <td>BARX2</td>\n",
       "      <td>206848337</td>\n",
       "      <td>...</td>\n",
       "      <td>206848365</td>\n",
       "      <td>Colon</td>\n",
       "      <td>pro_pro</td>\n",
       "      <td>1</td>\n",
       "      <td>BARX2::FOXF2+Colon+1</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>A_B_seq</td>\n",
       "      <td>CTCAGCTCTCCCACAAGATCTCTGGGGAAGCTGGAAGAGGGCCACT...</td>\n",
       "      <td>BARX2::FOXF2+Colon+1+A_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>BARX2::FOXF2</td>\n",
       "      <td>54</td>\n",
       "      <td>40</td>\n",
       "      <td>chr4</td>\n",
       "      <td>38859601</td>\n",
       "      <td>38859655</td>\n",
       "      <td>38859561</td>\n",
       "      <td>38859695</td>\n",
       "      <td>FOXF2</td>\n",
       "      <td>38859601</td>\n",
       "      <td>...</td>\n",
       "      <td>38859655</td>\n",
       "      <td>Colon</td>\n",
       "      <td>loop_loop</td>\n",
       "      <td>2</td>\n",
       "      <td>BARX2::FOXF2+Colon+2</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>A_B_seq</td>\n",
       "      <td>GTGGCTCTGAGGGAATTTCAAGTAACCGTGGACTAAAGTCTTAAGA...</td>\n",
       "      <td>BARX2::FOXF2+Colon+2+A_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>BARX2::FOXF2</td>\n",
       "      <td>46</td>\n",
       "      <td>32</td>\n",
       "      <td>chr6</td>\n",
       "      <td>74222776</td>\n",
       "      <td>74222822</td>\n",
       "      <td>74222732</td>\n",
       "      <td>74222866</td>\n",
       "      <td>FOXF2</td>\n",
       "      <td>74222776</td>\n",
       "      <td>...</td>\n",
       "      <td>74222822</td>\n",
       "      <td>Colon</td>\n",
       "      <td>loop_loop</td>\n",
       "      <td>3</td>\n",
       "      <td>BARX2::FOXF2+Colon+3</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>A_B_seq</td>\n",
       "      <td>GCCACTGCACTCCAGCCTGGTGACAGAGCGAGACTCTGTCTCAAGA...</td>\n",
       "      <td>BARX2::FOXF2+Colon+3+A_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>BARX2::FOXF2</td>\n",
       "      <td>16</td>\n",
       "      <td>2</td>\n",
       "      <td>chr3</td>\n",
       "      <td>49384674</td>\n",
       "      <td>49384690</td>\n",
       "      <td>49384615</td>\n",
       "      <td>49384749</td>\n",
       "      <td>FOXF2</td>\n",
       "      <td>49384674</td>\n",
       "      <td>...</td>\n",
       "      <td>49384690</td>\n",
       "      <td>Colon</td>\n",
       "      <td>pro_pro</td>\n",
       "      <td>4</td>\n",
       "      <td>BARX2::FOXF2+Colon+4</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>A_B_seq</td>\n",
       "      <td>AGTATTATGTGCTCATGGTCCCTACCCTCCAGGGTTCACAGTGTGG...</td>\n",
       "      <td>BARX2::FOXF2+Colon+4+A_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>BARX2::FOXF2</td>\n",
       "      <td>16</td>\n",
       "      <td>2</td>\n",
       "      <td>chr10</td>\n",
       "      <td>112247520</td>\n",
       "      <td>112247536</td>\n",
       "      <td>112247461</td>\n",
       "      <td>112247595</td>\n",
       "      <td>BARX2</td>\n",
       "      <td>112247520</td>\n",
       "      <td>...</td>\n",
       "      <td>112247536</td>\n",
       "      <td>Colon</td>\n",
       "      <td>pro_pro</td>\n",
       "      <td>5</td>\n",
       "      <td>BARX2::FOXF2+Colon+5</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>A_B_seq</td>\n",
       "      <td>ATGTGGCTGCGGGGCTTTAGGAGCTGACTCTTTGGTCACGGGGAGA...</td>\n",
       "      <td>BARX2::FOXF2+Colon+5+A_B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     vocab_pair  tot_dist  btn_dist    chr      start       stop  start_pad  \\\n",
       "0  BARX2::FOXF2        28        14   chr1  206848337  206848365  206848284   \n",
       "1  BARX2::FOXF2        54        40   chr4   38859601   38859655   38859561   \n",
       "2  BARX2::FOXF2        46        32   chr6   74222776   74222822   74222732   \n",
       "3  BARX2::FOXF2        16         2   chr3   49384674   49384690   49384615   \n",
       "4  BARX2::FOXF2        16         2  chr10  112247520  112247536  112247461   \n",
       "\n",
       "    stop_pad vocab1  vocab1_start  ...  vocab2_stop tissue  loop_type  \\\n",
       "0  206848418  BARX2     206848337  ...    206848365  Colon    pro_pro   \n",
       "1   38859695  FOXF2      38859601  ...     38859655  Colon  loop_loop   \n",
       "2   74222866  FOXF2      74222776  ...     74222822  Colon  loop_loop   \n",
       "3   49384749  FOXF2      49384674  ...     49384690  Colon    pro_pro   \n",
       "4  112247595  BARX2     112247520  ...    112247536  Colon    pro_pro   \n",
       "\n",
       "   vocab_tissue_count                  name check_rs  scr_pass_rs seq_config  \\\n",
       "0                   1  BARX2::FOXF2+Colon+1        0         True    A_B_seq   \n",
       "1                   2  BARX2::FOXF2+Colon+2        0         True    A_B_seq   \n",
       "2                   3  BARX2::FOXF2+Colon+3        0         True    A_B_seq   \n",
       "3                   4  BARX2::FOXF2+Colon+4        0         True    A_B_seq   \n",
       "4                   5  BARX2::FOXF2+Colon+5        0         True    A_B_seq   \n",
       "\n",
       "                                                 seq               name_config  \n",
       "0  CTCAGCTCTCCCACAAGATCTCTGGGGAAGCTGGAAGAGGGCCACT...  BARX2::FOXF2+Colon+1+A_B  \n",
       "1  GTGGCTCTGAGGGAATTTCAAGTAACCGTGGACTAAAGTCTTAAGA...  BARX2::FOXF2+Colon+2+A_B  \n",
       "2  GCCACTGCACTCCAGCCTGGTGACAGAGCGAGACTCTGTCTCAAGA...  BARX2::FOXF2+Colon+3+A_B  \n",
       "3  AGTATTATGTGCTCATGGTCCCTACCCTCCAGGGTTCACAGTGTGG...  BARX2::FOXF2+Colon+4+A_B  \n",
       "4  ATGTGGCTGCGGGGCTTTAGGAGCTGACTCTTTGGTCACGGGGAGA...  BARX2::FOXF2+Colon+5+A_B  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# vocab_oligo_df.columns\n",
    "vocab_instances20_seqs_sel = pd.read_csv(os.path.join(save_dir, 'vocab_instances20_seq_wscr_sel.csv'),index_col=0)\n",
    "vocab_instances20_seqs_sel[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def vocab_present_func(row):\n",
    "    # determine if there should be a vocab present (based on configuration)\n",
    "    return row['vocab_order'] in row['config'].split('_')\n",
    "#     vocab_instances_df.vocab_order.apply(lambda x: xisin(vocab_instances_df.config.str.split('_'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  This is separate from the ipykernel package so we can avoid doing imports until\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  import sys\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>name_config</th>\n",
       "      <th>start_pad</th>\n",
       "      <th>stop_pad</th>\n",
       "      <th>vocab</th>\n",
       "      <th>vocab_order</th>\n",
       "      <th>vocab_start</th>\n",
       "      <th>vocab_stop</th>\n",
       "      <th>instance</th>\n",
       "      <th>config</th>\n",
       "      <th>vocab_present_pred</th>\n",
       "      <th>vocab_loc_start</th>\n",
       "      <th>vocab_loc_stop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>744</td>\n",
       "      <td>chr18</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+1+A_B</td>\n",
       "      <td>61243074</td>\n",
       "      <td>61243208</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>A</td>\n",
       "      <td>61243131</td>\n",
       "      <td>61243135</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+1</td>\n",
       "      <td>A_B</td>\n",
       "      <td>True</td>\n",
       "      <td>57</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>744</td>\n",
       "      <td>chr18</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+1+A_B</td>\n",
       "      <td>61243074</td>\n",
       "      <td>61243208</td>\n",
       "      <td>KLF5</td>\n",
       "      <td>B</td>\n",
       "      <td>61243141</td>\n",
       "      <td>61243152</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+1</td>\n",
       "      <td>A_B</td>\n",
       "      <td>True</td>\n",
       "      <td>67</td>\n",
       "      <td>78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9236</td>\n",
       "      <td>chr18</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+1+A_scrB</td>\n",
       "      <td>61243074</td>\n",
       "      <td>61243208</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>A</td>\n",
       "      <td>61243131</td>\n",
       "      <td>61243135</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+1</td>\n",
       "      <td>A_scrB</td>\n",
       "      <td>True</td>\n",
       "      <td>57</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9236</td>\n",
       "      <td>chr18</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+1+A_scrB</td>\n",
       "      <td>61243074</td>\n",
       "      <td>61243208</td>\n",
       "      <td>KLF5</td>\n",
       "      <td>B</td>\n",
       "      <td>61243141</td>\n",
       "      <td>61243152</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+1</td>\n",
       "      <td>A_scrB</td>\n",
       "      <td>False</td>\n",
       "      <td>67</td>\n",
       "      <td>78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4990</td>\n",
       "      <td>chr18</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+1+scrA_B</td>\n",
       "      <td>61243074</td>\n",
       "      <td>61243208</td>\n",
       "      <td>KLF5</td>\n",
       "      <td>B</td>\n",
       "      <td>61243141</td>\n",
       "      <td>61243152</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+1</td>\n",
       "      <td>scrA_B</td>\n",
       "      <td>True</td>\n",
       "      <td>67</td>\n",
       "      <td>78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11996</td>\n",
       "      <td>chr7</td>\n",
       "      <td>ZFX::E2F6+SCC+9+A_scrB</td>\n",
       "      <td>44153033</td>\n",
       "      <td>44153167</td>\n",
       "      <td>E2F6</td>\n",
       "      <td>A</td>\n",
       "      <td>44153086</td>\n",
       "      <td>44153095</td>\n",
       "      <td>ZFX::E2F6+SCC+9</td>\n",
       "      <td>A_scrB</td>\n",
       "      <td>True</td>\n",
       "      <td>53</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7750</td>\n",
       "      <td>chr7</td>\n",
       "      <td>ZFX::E2F6+SCC+9+scrA_B</td>\n",
       "      <td>44153033</td>\n",
       "      <td>44153167</td>\n",
       "      <td>E2F6</td>\n",
       "      <td>A</td>\n",
       "      <td>44153086</td>\n",
       "      <td>44153095</td>\n",
       "      <td>ZFX::E2F6+SCC+9</td>\n",
       "      <td>scrA_B</td>\n",
       "      <td>False</td>\n",
       "      <td>53</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7750</td>\n",
       "      <td>chr7</td>\n",
       "      <td>ZFX::E2F6+SCC+9+scrA_B</td>\n",
       "      <td>44153033</td>\n",
       "      <td>44153167</td>\n",
       "      <td>ZFX</td>\n",
       "      <td>B</td>\n",
       "      <td>44153101</td>\n",
       "      <td>44153114</td>\n",
       "      <td>ZFX::E2F6+SCC+9</td>\n",
       "      <td>scrA_B</td>\n",
       "      <td>True</td>\n",
       "      <td>68</td>\n",
       "      <td>81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16242</td>\n",
       "      <td>chr7</td>\n",
       "      <td>ZFX::E2F6+SCC+9+scrA_scrB</td>\n",
       "      <td>44153033</td>\n",
       "      <td>44153167</td>\n",
       "      <td>ZFX</td>\n",
       "      <td>B</td>\n",
       "      <td>44153101</td>\n",
       "      <td>44153114</td>\n",
       "      <td>ZFX::E2F6+SCC+9</td>\n",
       "      <td>scrA_scrB</td>\n",
       "      <td>False</td>\n",
       "      <td>68</td>\n",
       "      <td>81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16242</td>\n",
       "      <td>chr7</td>\n",
       "      <td>ZFX::E2F6+SCC+9+scrA_scrB</td>\n",
       "      <td>44153033</td>\n",
       "      <td>44153167</td>\n",
       "      <td>E2F6</td>\n",
       "      <td>A</td>\n",
       "      <td>44153086</td>\n",
       "      <td>44153095</td>\n",
       "      <td>ZFX::E2F6+SCC+9</td>\n",
       "      <td>scrA_scrB</td>\n",
       "      <td>False</td>\n",
       "      <td>53</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19120 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr                  name_config  start_pad  stop_pad   vocab  \\\n",
       "744    chr18     ARID5B::KLF5+GDSD6+1+A_B   61243074  61243208  ARID5B   \n",
       "744    chr18     ARID5B::KLF5+GDSD6+1+A_B   61243074  61243208    KLF5   \n",
       "9236   chr18  ARID5B::KLF5+GDSD6+1+A_scrB   61243074  61243208  ARID5B   \n",
       "9236   chr18  ARID5B::KLF5+GDSD6+1+A_scrB   61243074  61243208    KLF5   \n",
       "4990   chr18  ARID5B::KLF5+GDSD6+1+scrA_B   61243074  61243208    KLF5   \n",
       "...      ...                          ...        ...       ...     ...   \n",
       "11996   chr7       ZFX::E2F6+SCC+9+A_scrB   44153033  44153167    E2F6   \n",
       "7750    chr7       ZFX::E2F6+SCC+9+scrA_B   44153033  44153167    E2F6   \n",
       "7750    chr7       ZFX::E2F6+SCC+9+scrA_B   44153033  44153167     ZFX   \n",
       "16242   chr7    ZFX::E2F6+SCC+9+scrA_scrB   44153033  44153167     ZFX   \n",
       "16242   chr7    ZFX::E2F6+SCC+9+scrA_scrB   44153033  44153167    E2F6   \n",
       "\n",
       "      vocab_order  vocab_start  vocab_stop              instance     config  \\\n",
       "744             A     61243131    61243135  ARID5B::KLF5+GDSD6+1        A_B   \n",
       "744             B     61243141    61243152  ARID5B::KLF5+GDSD6+1        A_B   \n",
       "9236            A     61243131    61243135  ARID5B::KLF5+GDSD6+1     A_scrB   \n",
       "9236            B     61243141    61243152  ARID5B::KLF5+GDSD6+1     A_scrB   \n",
       "4990            B     61243141    61243152  ARID5B::KLF5+GDSD6+1     scrA_B   \n",
       "...           ...          ...         ...                   ...        ...   \n",
       "11996           A     44153086    44153095       ZFX::E2F6+SCC+9     A_scrB   \n",
       "7750            A     44153086    44153095       ZFX::E2F6+SCC+9     scrA_B   \n",
       "7750            B     44153101    44153114       ZFX::E2F6+SCC+9     scrA_B   \n",
       "16242           B     44153101    44153114       ZFX::E2F6+SCC+9  scrA_scrB   \n",
       "16242           A     44153086    44153095       ZFX::E2F6+SCC+9  scrA_scrB   \n",
       "\n",
       "       vocab_present_pred  vocab_loc_start  vocab_loc_stop  \n",
       "744                  True               57              61  \n",
       "744                  True               67              78  \n",
       "9236                 True               57              61  \n",
       "9236                False               67              78  \n",
       "4990                 True               67              78  \n",
       "...                   ...              ...             ...  \n",
       "11996                True               53              62  \n",
       "7750                False               53              62  \n",
       "7750                 True               68              81  \n",
       "16242               False               68              81  \n",
       "16242               False               53              62  \n",
       "\n",
       "[19120 rows x 13 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances_df = vocab_instances20_seqs_sel[['name_config','chr','start_pad','stop_pad','vocab1','vocab1_start','vocab1_stop']]\n",
    "vocab_instances_df.columns = ['name_config','chr','start_pad','stop_pad','vocab','vocab_start','vocab_stop']\n",
    "vocab_instances_df['vocab_order'] = 'A'\n",
    "\n",
    "vocab_instances_df2 = vocab_instances20_seqs_sel[['name_config','chr','start_pad','stop_pad','vocab2','vocab2_start','vocab2_stop']]\n",
    "vocab_instances_df2.columns = ['name_config','chr','start_pad','stop_pad','vocab','vocab_start','vocab_stop']\n",
    "vocab_instances_df2['vocab_order'] = 'B'\n",
    "vocab_instances_df = pd.concat([vocab_instances_df,vocab_instances_df2],axis=0,sort=True)\n",
    "vocab_instances_df['instance'] = vocab_instances_df.name_config.apply(lambda x: '+'.join(x.split('+')[:-1]))\n",
    "vocab_instances_df['config'] = vocab_instances_df.name_config.str.split('+',expand=True).iloc[:,-1]\n",
    "vocab_instances_df['vocab_present_pred'] = vocab_instances_df.apply(vocab_present_func,axis=1)\n",
    "vocab_instances_df['vocab_loc_start'] = vocab_instances_df.vocab_start - vocab_instances_df.start_pad\n",
    "vocab_instances_df['vocab_loc_stop'] = vocab_instances_df.vocab_stop - vocab_instances_df.start_pad\n",
    "vocab_instances_df.sort_values('name_config',inplace=True)\n",
    "vocab_instances_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2390,)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances_df.instance.unique().shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 7C2: add in motif_scan information per case"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 328,
   "metadata": {},
   "outputs": [],
   "source": [
    "# helper functions\n",
    "def label_config(row):\n",
    "    if row['vocab_present_actual']: \n",
    "        if row['vocab_present_pred']:\n",
    "            return 'TP' # good, Case 1\n",
    "        else:\n",
    "            return 'FP' # noisy, Case 3\n",
    "    else:\n",
    "        if row['vocab_present_pred']:\n",
    "            return 'FN' # not detected (Case 4)\n",
    "        else:\n",
    "            return 'TN' # good, Case 1\n",
    "def get_case_instance(row):\n",
    "    row_counts = row.value_counts()\n",
    "    # case 1: all the configs that should have the vocab have the vocab and those that don't do not\n",
    "    if (row_counts.get('TP',0) +row_counts.get('TN',0) )==4:\n",
    "        return 'case 1'\n",
    "    # case 4: the pvalue threshold was too stringent\n",
    "    elif (row_counts.get('FN',0) +row_counts.get('TN',0) )==4:\n",
    "        return 'case 4'\n",
    "    # case 3: too noisy\n",
    "    elif (row_counts.get('TP',0) +row_counts.get('FP',0) )==4:\n",
    "        return 'case 3'\n",
    "    else:\n",
    "        return 'other'\n",
    "\n",
    "def analyze_motif_scan(motif_scan_file, tf_id_abbr_to_name_dict=tf_id_abbr_to_name_dict):\n",
    "    # read in motif scan file\n",
    "    motif_scan_results = pd.read_csv(motif_scan_file,names=['name_config','pwm_file','pos','strand','score','seq','seq_actual'])\n",
    "    motif = tf_id_abbr_to_name_dict[os.path.basename(motif_scan_file).split('_scan')[0]]\n",
    "    motif_scan_results['motif'] = motif_scan_results.pwm_file.str.split('.pwm',expand=True).iloc[:,0].map(tf_id_abbr_to_name_dict)\n",
    "\n",
    "    # merge in genomic location information\n",
    "    vocab_motif_scan = vocab_instances_df[vocab_instances_df.vocab==motif].merge( motif_scan_results[['name_config','pos','score','motif']], how='left',left_on=['name_config','vocab'], right_on=['name_config','motif'])\n",
    "    vocab_motif_scan.pos.fillna(-1.0, inplace=True)\n",
    "    vocab_motif_scan.score.fillna(0, inplace=True)\n",
    "    vocab_motif_scan.motif.fillna('', inplace=True)\n",
    "    vocab_motif_scan['vocab_present_actual'] = abs(vocab_motif_scan.vocab_loc_start - vocab_motif_scan.pos ) <=3\n",
    "    vocab_motif_scan['off_target']  =  ~( ( vocab_motif_scan.vocab_present_actual) | (vocab_motif_scan.pos<0))\n",
    "\n",
    "    #### case 2: off-target effects\n",
    "    vocab_motif_scan_off_target = vocab_motif_scan[vocab_motif_scan.off_target]\n",
    "    vocab_motif_scan_off_target.sort_values('score',ascending=False,inplace=True)\n",
    "    vocab_motif_scan_off_target.drop_duplicates('name_config',inplace=True)\n",
    "    vocab_motif_scan_off_target.sort_values('name_config',inplace=True)\n",
    "    vocab_motif_scan_off_target['pos_close'] = vocab_motif_scan_off_target.pos - (vocab_motif_scan_off_target.pos%2)\n",
    "    off_target_counts_df = pd.DataFrame(vocab_motif_scan_off_target.groupby('instance').pos_close.value_counts())\n",
    "    off_target_counts_df.columns = ['count_pos']\n",
    "    off_target_counts_df['motif'] = motif\n",
    "    off_target_counts_df['case'] = off_target_counts_df.count_pos.apply(lambda x: 'case 2A' if x==4 else 'case 2B')\n",
    "    off_target_counts_df.reset_index(inplace=True)\n",
    "\n",
    "    ##### Case 1,3,4\n",
    "    # - case 1: good\n",
    "    # - case 3: FP (too noisy, scrambling might not be good), case 4: FN (pval, threshold too low )\n",
    "\n",
    "    # only include the highest scoring \n",
    "    vocab_motif_scan_on_target = vocab_motif_scan[~vocab_motif_scan.off_target]\n",
    "    vocab_motif_scan_on_target.sort_values('score',ascending=False,inplace=True)\n",
    "    vocab_motif_scan_on_target.drop_duplicates('name_config',inplace=True)\n",
    "    vocab_motif_scan_on_target.sort_values('name_config',inplace=True)\n",
    "\n",
    "    # fill in all possible instances in the library\n",
    "    vocab_motif_scan_on_target = vocab_instances_df[vocab_instances_df.vocab==motif].merge(vocab_motif_scan_on_target,how='left',on=list(vocab_instances_df.columns.values))\n",
    "    vocab_motif_scan_on_target.pos.fillna(-1.0, inplace=True)\n",
    "    vocab_motif_scan_on_target.score.fillna(0, inplace=True)\n",
    "    vocab_motif_scan_on_target.motif.fillna('', inplace=True)\n",
    "    vocab_motif_scan_on_target.vocab_present_actual.fillna(False, inplace=True)\n",
    "    vocab_motif_scan_on_target.off_target.fillna(False, inplace=True)\n",
    "    vocab_motif_scan_on_target['eval'] = vocab_motif_scan_on_target.apply(label_config, axis=1)\n",
    "\n",
    "    # create data frame thats has rows (instances) and columns are the eval (TP,FP, etc) and score per config\n",
    "    vocab_motif_scan_on_target_config = vocab_motif_scan_on_target.pivot(index='instance', columns='config', values=['eval','score'])\n",
    "    vocab_motif_scan_on_target_config.columns = ['_'.join(list(x)) for x in vocab_motif_scan_on_target_config.columns.values]\n",
    "    vocab_motif_scan_on_target_config['motif'] = motif\n",
    "    vocab_motif_scan_on_target_config['case'] = vocab_motif_scan_on_target_config.apply(get_case_instance,axis=1)\n",
    "\n",
    "    # dict of case_type: # instances that fall in type\n",
    "    on_target_config_cases = vocab_motif_scan_on_target_config.case.value_counts()\n",
    "\n",
    "    return motif, off_target_counts_df, vocab_motif_scan_on_target_config, on_target_config_cases\n",
    "\n",
    "def analyze_motif_scan_wrap(motif_scan_files, save_prefix=None):\n",
    "    print('# motif files: ', len(motif_scan_files))\n",
    "    # overall dataframe setup\n",
    "    off_target_counts_df_all = pd.DataFrame()\n",
    "    on_target_cases_df_all = pd.DataFrame()\n",
    "    on_target_config_cases_dict = {}\n",
    "    \n",
    "    #get scan results from file\n",
    "    for idx,motif_scan_file in enumerate(motif_scan_files):\n",
    "        ## DEBUG\n",
    "    #     print(idx,motif_scan_file)\n",
    "        motif, off_target_counts_df, vocab_motif_scan_on_target_config, on_target_config_cases = analyze_motif_scan(motif_scan_file)\n",
    "        # ADD TO RUNNING LIST (ACROSS MOTIFS)\n",
    "        off_target_counts_df_all = pd.concat([off_target_counts_df_all, off_target_counts_df],axis=0,sort=True)\n",
    "        on_target_cases_df_all = pd.concat([on_target_cases_df_all, vocab_motif_scan_on_target_config],axis=0,sort=True)\n",
    "        on_target_config_cases_dict[motif] = on_target_config_cases\n",
    "\n",
    "    # convert dict to dataframe (row: motif, columns case type)\n",
    "    on_target_config_cases_df = pd.DataFrame.from_dict(on_target_config_cases_dict,orient='index').fillna(0)\n",
    "\n",
    "    ## TODO: add in save functionality\n",
    "    if save_prefix is not None:\n",
    "        off_target_counts_df_all.to_csv(save_prefix+'off_target_counts_df_all.csv')\n",
    "        on_target_cases_df_all.to_csv(save_prefix+'motif_scan_'+'on_target_cases_df_all.csv')\n",
    "        on_target_config_cases_df.to_csv(save_prefix+ 'motif_scan_'+'on_target_config_cases_df.csv')\n",
    "        print('saved', save_prefix+'*.csv')\n",
    "    return off_target_counts_df_all, on_target_cases_df_all, on_target_config_cases_df\n",
    "\n",
    "def refine_case3(row):\n",
    "    if row['case']!='case 3':\n",
    "#         print(row['case'])\n",
    "        return row['case']\n",
    "    else:\n",
    "        # only remodify case 3:\n",
    "        if row['score_A_B']> min(row['score_A_scrB'],row['score_scrA_B'] ):\n",
    "    #         print('here')\n",
    "            return 'case 3A' # this is the good case where scrambling reduces the score of the motif\n",
    "        else:\n",
    "            return 'case 3B'\n",
    "        # case 3, where the TP are have higher score than where the true negatives are\n",
    "\n",
    "def get_offtarget_num(off_target_counts_df, vocab_instances_df=vocab_instances20_seqs_sel):\n",
    "    off_target_res_case_2A = off_target_counts_df[off_target_counts_df.case=='case 2A']\n",
    "    print(off_target_res_case_2A.shape)\n",
    "    off_target_res_case_2A['off_target_pos_str'] = off_target_res_case_2A.pos_close.map(int).map(str)\n",
    "    off_target_res_case_2A = off_target_res_case_2A.groupby(['instance','motif']).agg({'off_target_pos_str':'|'.join}).reset_index()\n",
    "    off_target_res_case_2A.columns = ['off_target_name','off_target_motif','off_target_pos_str']\n",
    "    print(off_target_res_case_2A.shape) # groupbed by instance\n",
    "    vocab_instances20_seqs_sel_offtarget = vocab_instances_df.merge(off_target_res_case_2A, how='left',left_on='name',right_on='off_target_name')\n",
    "    vocab_instances20_seqs_sel_offtarget.off_target_name.fillna('',inplace=True)\n",
    "    vocab_instances20_seqs_sel_offtarget.off_target_motif.fillna('',inplace=True)\n",
    "    vocab_instances20_seqs_sel_offtarget.off_target_pos_str.fillna('',inplace=True)\n",
    "    vocab_instances20_seqs_sel_offtarget = pd.concat([vocab_instances20_seqs_sel_offtarget,vocab_instances20_seqs_sel_offtarget.apply(get_vocab_loc,axis=1)],axis=1)\n",
    "    \n",
    "#     return vocab_instances20_seqs_sel_offtarget\n",
    "    vocab_instances20_seqs_sel_offtarget = vocab_instances20_seqs_sel_offtarget.groupby(['vocab_pair', 'tot_dist', 'btn_dist', 'chr', 'start', 'stop',\n",
    "           'start_pad', 'stop_pad', 'vocab1', 'vocab1_start', 'vocab1_stop',\n",
    "           'vocab2', 'vocab2_start', 'vocab2_stop', 'tissue', 'loop_type', 'name', 'check_rs', 'scr_pass_rs', \n",
    "           'vocab_tissue_count', 'seq_config', 'seq','name_config', 'off_target_name','vocab1_start_shift', 'vocab1_stop_shift',\n",
    "           'vocab2_start_shift', 'vocab2_stop_shift']).agg({'off_target_motif':'::'.join,'off_target_pos_str':'::'.join}).reset_index()\n",
    "\n",
    "    vocab_instances20_seqs_sel_offtarget = vocab_instances20_seqs_sel_offtarget.groupby(['vocab_pair', 'tot_dist', 'btn_dist', 'chr', 'start', 'stop',\n",
    "           'start_pad', 'stop_pad', 'vocab1', 'vocab1_start', 'vocab1_stop',\n",
    "           'vocab2', 'vocab2_start', 'vocab2_stop', 'tissue', 'loop_type', 'name',\n",
    "           'check_rs', 'scr_pass_rs', 'vocab_tissue_count',  'vocab1_start_shift',\n",
    "           'vocab1_stop_shift', 'vocab2_start_shift', 'vocab2_stop_shift','off_target_name',\n",
    "           'off_target_motif', 'off_target_pos_str']).agg({'seq_config':list,'seq':list, 'name_config':list}).reset_index()\n",
    "    # \n",
    "\n",
    "\n",
    "    off_target_num_df = pd.DataFrame(vocab_instances20_seqs_sel_offtarget.set_index('name').off_target_pos_str.apply(off_target_pos_str_find_num)).reset_index()\n",
    "    off_target_num_df['instance'] = off_target_num_df.name.apply(lambda x: '+'.join(x.split('+')[:-1]))\n",
    "    off_target_num_df_counts = off_target_num_df[off_target_num_df.off_target_pos_str==0].instance.value_counts().value_counts().sort_index()\n",
    "    print(off_target_num_df_counts/off_target_num_df_counts.sum())\n",
    "    return off_target_num_df[off_target_num_df.off_target_pos_str==0].instance.value_counts().describe()\n",
    "\n",
    "\n",
    "def get_vocab_loc(row, motif_len_scr=10):\n",
    "    # getting vocab locations in the sequence\n",
    "    vocab1_med = int((row['vocab1_start']+row['vocab1_stop'])/2)\n",
    "    vocab1_start_shift = vocab1_med - motif_len_scr - row['start_pad']\n",
    "    vocab1_stop_shift = vocab1_med + motif_len_scr -row['start_pad'] + 1\n",
    "    \n",
    "    vocab2_med = int((row['vocab2_start']+row['vocab2_stop'])/2)\n",
    "    vocab2_start_shift = vocab2_med - motif_len_scr - row['start_pad']\n",
    "    vocab2_stop_shift = vocab2_med + motif_len_scr -row['start_pad'] + 1\n",
    "    \n",
    "    # for overlaps\n",
    "    if vocab1_stop_shift>= vocab2_start_shift:\n",
    "#         print(idx)\n",
    "#         print(row)\n",
    "        \n",
    "        # case 1 vocab words are really close, the padding goes into the vocab word\n",
    "        if (row['vocab1_stop']-row['start_pad'] + 1) >=vocab2_start_shift:\n",
    "            \n",
    "            vocab1_stop_shift = row['vocab1_stop']-row['start_pad'] + 1\n",
    "            vocab2_start_shift = row['vocab1_stop']-row['start_pad'] + 1\n",
    "#             print(vocab1_start_shift,vocab1_stop_shift, vocab2_start_shift, vocab2_stop_shift)\n",
    "#             raise\n",
    "        else: \n",
    "            vocab1_stop_shift = vocab1_stop_shift\n",
    "            vocab2_start_shift = vocab1_stop_shift\n",
    "#             print(vocab1_start_shift,vocab1_stop_shift, vocab2_start_shift, vocab2_stop_shift)\n",
    "#             raise\n",
    "\n",
    "    # for edges:\n",
    "    vocab1_start_shift = max(0,vocab1_start_shift)\n",
    "    vocab2_stop_shift = min(len(row['seq']),vocab2_stop_shift)\n",
    "    \n",
    "    return pd.Series({'vocab1_start_shift': vocab1_start_shift,\n",
    "                      'vocab1_stop_shift':vocab1_stop_shift, \n",
    "                      'vocab2_start_shift':vocab2_start_shift, \n",
    "                      'vocab2_stop_shift':vocab2_stop_shift})\n",
    "\n",
    "\n",
    "def get_vocab_to_loc_dict(row):\n",
    "    return {row['vocab1']:[row['vocab1_start_shift'],row['vocab1_stop_shift']],\n",
    "           row['vocab2']:[row['vocab2_start_shift'],row['vocab2_stop_shift']]}\n",
    "\n",
    "\n",
    "def off_target_pos_str_find_num(pos_str):\n",
    "    pos_arr_per_vocab_word = pos_str.split('::')\n",
    "    if len(pos_arr_per_vocab_word)>0:\n",
    "        arr = []\n",
    "        for pos in pos_arr_per_vocab_word:\n",
    "            for x in pos.split(':'):\n",
    "                try:\n",
    "                    arr.append(int(x))\n",
    "                except:\n",
    "                    continue\n",
    "        return len(set(arr))\n",
    "    return 0\n",
    "\n",
    "def on_target_new_cases(row):\n",
    "    motif_arr = row['motif'].split('|')\n",
    "    case_mod_arr = row['case_mod'].split('|')\n",
    "    num_motifs = len(set(motif_arr))\n",
    "    if num_motifs>2:\n",
    "        return '|'.join(sorted(set(motif_arr))), 'more than 2 motifs'\n",
    "    elif num_motifs<2:\n",
    "        return '|'.join(sorted(set(motif_arr))), 'less than 2 motifs'    \n",
    "    else:\n",
    "        motif_to_case = {}\n",
    "        for motif, case in zip(motif_arr, case_mod_arr):\n",
    "            if motif in motif_to_case:\n",
    "                if (motif_to_case[motif]!='case 3A') & (motif_to_case[motif]!='case 1'):\n",
    "                    motif_to_case[motif]=case\n",
    "            else:\n",
    "                motif_to_case[motif]=case\n",
    "        motif_new = '|'.join(list(motif_to_case.keys()))\n",
    "        case_new = '|'.join(list(motif_to_case.values()))\n",
    "        return motif_new, case_new\n",
    "    \n",
    "def on_target_new_case_good(case_new):\n",
    "    if '|' not in case_new:\n",
    "        return False # not two motifs\n",
    "    else:\n",
    "        case_status = True\n",
    "        for case in case_new.split('|'):\n",
    "            if  (case!='case 3A') & (case!='case 1') :\n",
    "                case_status = False\n",
    "        return case_status\n",
    "    \n",
    "def get_on_target_good_instance_num(on_target_cases_df):\n",
    "    on_target_counts = on_target_cases_df.reset_index()[['instance','motif','case_mod']].drop_duplicates().groupby('instance').agg({\n",
    "    'motif':'|'.join, 'case_mod':'|'.join\n",
    "    }).reset_index()\n",
    "    on_target_counts['name'] = on_target_counts.instance.apply(lambda x: '+'.join(x.split('+')[:-1]))\n",
    "    on_target_counts = pd.concat([on_target_counts, on_target_counts.apply(on_target_new_cases,axis=1).apply(lambda s: pd.Series({'motif_new':s[0], 'case_new':s[1]}))],axis=1)\n",
    "    on_target_counts['case_good_bool'] = on_target_counts.case_new.apply(on_target_new_case_good)\n",
    "    return on_target_counts[on_target_counts.case_good_bool==True].name.value_counts().describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 257,
   "metadata": {},
   "outputs": [],
   "source": [
    "# testing\n",
    "    # read in motif scan file\n",
    "motif_scan_results = pd.read_csv(motif_scan_files1[0],names=['name_config','pwm_file','pos','strand','score','seq','seq_actual'])\n",
    "motif = tf_id_abbr_to_name_dict[os.path.basename(motif_scan_files1[0]).split('_scan')[0]]\n",
    "motif_scan_results['motif'] = motif_scan_results.pwm_file.str.split('.pwm',expand=True).iloc[:,0].map(tf_id_abbr_to_name_dict)\n",
    "\n",
    "# merge in genomic location information\n",
    "vocab_motif_scan = vocab_instances_df[vocab_instances_df.vocab==motif].merge( motif_scan_results[['name_config','pos','score','motif']], how='left',left_on=['name_config','vocab'], right_on=['name_config','motif'])\n",
    "vocab_motif_scan.pos.fillna(-1.0, inplace=True)\n",
    "vocab_motif_scan.score.fillna(0, inplace=True)\n",
    "vocab_motif_scan.motif.fillna('', inplace=True)\n",
    "vocab_motif_scan['vocab_present_actual'] = abs(vocab_motif_scan.vocab_loc_start - vocab_motif_scan.pos ) <=3\n",
    "vocab_motif_scan['off_target']  =  ~( ( vocab_motif_scan.vocab_present_actual) | (vocab_motif_scan.pos<0))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 258,
   "metadata": {},
   "outputs": [],
   "source": [
    "# vocab_motif_scan_off_target"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### version 1: analyze the original scrambled version (only scramble the motifs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# motif files:  158\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:43: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:44: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:45: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:46: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:59: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:60: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:61: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "saved ../data/processed/fig4_modelling/vocab_sum_final/motif_scan_*.csv\n",
      "CPU times: user 14.1 s, sys: 242 ms, total: 14.3 s\n",
      "Wall time: 14.7 s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "case 1    3093.0\n",
       "case 4    1077.0\n",
       "case 3     535.0\n",
       "other       75.0\n",
       "dtype: float64"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "# takes ~ 15s\n",
    "\n",
    "motif_scan_files1 = sorted(glob.glob(os.path.join(save_dir, 'motif_scan','*csv')))\n",
    "off_target_counts_df_all_1, on_target_cases_df_all_1, on_target_config_cases_df_1 = analyze_motif_scan_wrap(motif_scan_files1, save_prefix=os.path.join(save_dir, 'motif_scan_'))\n",
    "on_target_config_cases_df_1.sum(axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 260,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "case 1    3625\n",
       "case 4    1722\n",
       "case 3     611\n",
       "other      122\n",
       "Name: case, dtype: int64"
      ]
     },
     "execution_count": 260,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "on_target_cases_df_all_1.case.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### version 2: analyze the scrambled version where we scramble a fixed 20bp length, p val <0.001"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 278,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# motif files:  158\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:43: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:44: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:45: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:46: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:59: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:60: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:61: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "saved ../data/processed/fig4_modelling/vocab_sum_final/motif_scan2_*.csv\n",
      "CPU times: user 13.9 s, sys: 235 ms, total: 14.1 s\n",
      "Wall time: 14.1 s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "case 1    3329.0\n",
       "case 4    1085.0\n",
       "other       93.0\n",
       "case 3     273.0\n",
       "dtype: float64"
      ]
     },
     "execution_count": 278,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "# takes ~ 15s\n",
    "motif_scan_files2 = sorted(glob.glob(os.path.join(save_dir, 'motif_scan2','*csv')))\n",
    "off_target_counts_df_all_2, on_target_cases_df_all_2, on_target_config_cases_df_2 = analyze_motif_scan_wrap(motif_scan_files2, save_prefix=os.path.join(save_dir, 'motif_scan2_'))\n",
    "on_target_config_cases_df_2.sum(axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 282,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "case 1    3875\n",
       "case 4    1740\n",
       "case 3     320\n",
       "other      145\n",
       "Name: case, dtype: int64"
      ]
     },
     "execution_count": 282,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "on_target_cases_df_all_2.case.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 283,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "case 1     3875\n",
       "case 4     1740\n",
       "case 3A     165\n",
       "case 3B     155\n",
       "other       145\n",
       "Name: case_mod, dtype: int64"
      ]
     },
     "execution_count": 283,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "on_target_cases_df_all_2['case_mod'] = on_target_cases_df_all_2.apply(refine_case3, axis=1)\n",
    "on_target_cases_df_all_2.case_mod.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 329,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    234.000000\n",
       "mean       6.820513\n",
       "std        2.845418\n",
       "min        1.000000\n",
       "25%        5.000000\n",
       "50%        8.000000\n",
       "75%        9.000000\n",
       "max       10.000000\n",
       "Name: name, dtype: float64"
      ]
     },
     "execution_count": 329,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_on_target_good_instance_num(on_target_cases_df_all_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 301,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(6080,)"
      ]
     },
     "execution_count": 301,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "on_target_cases_df_all_2.case_mod.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 285,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4780.0"
      ]
     },
     "execution_count": 285,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "on_target_config_cases_df_2.sum(axis=0).sum() #double the number of instances b/c of the two vocab words per instance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 286,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "FN|FN|FP|FP    31\n",
       "FN|FP|FN|FP    26\n",
       "TP|TP|FP|TN    16\n",
       "TP|TN|FN|TN    13\n",
       "TP|TP|TN|FP    11\n",
       "TP|FN|TN|TN     9\n",
       "FN|TP|TN|TN     8\n",
       "TP|FP|TP|TN     7\n",
       "FN|TN|TP|TN     5\n",
       "TP|TN|TP|FP     5\n",
       "FN|FN|FP|TN     5\n",
       "FN|TN|FN|FP     3\n",
       "FN|FN|TN|FP     2\n",
       "TP|TN|FN|FP     1\n",
       "FN|FP|FN|TN     1\n",
       "TP|FP|FN|TN     1\n",
       "TP|FN|FP|TN     1\n",
       "dtype: int64"
      ]
     },
     "execution_count": 286,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "on_target_cases_df_all_2[on_target_cases_df_all_2.case=='other'][['eval_A_B',\n",
    "                                                                  'eval_A_scrB',\t\n",
    "                                                                  'eval_scrA_B',\t\n",
    "                                                                  'eval_scrA_scrB']].apply('|'.join,axis=1).value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 288,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2840, 3)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "case 2B    1703\n",
       "case 2A    1137\n",
       "Name: case, dtype: int64"
      ]
     },
     "execution_count": 288,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(off_target_counts_df_all_2[['instance','motif','case']].drop_duplicates().shape)\n",
    "off_target_counts_df_all_2[['instance','motif','case']].drop_duplicates()['case'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 309,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5    57\n",
       "4    51\n",
       "3    38\n",
       "6    32\n",
       "2    21\n",
       "1    13\n",
       "7    12\n",
       "9     3\n",
       "8     3\n",
       "Name: instance, dtype: int64"
      ]
     },
     "execution_count": 309,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "off_target_counts_df_all_2[off_target_counts_df_all_2.case=='case 2A'].instance.drop_duplicates().apply(lambda x: '+'.join(x.split('+')[:-1])).value_counts().value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 315,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1248, 5)\n",
      "(1137, 3)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:127: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1     0.004184\n",
      "2     0.012552\n",
      "3     0.033473\n",
      "4     0.100418\n",
      "5     0.221757\n",
      "6     0.188285\n",
      "7     0.205021\n",
      "8     0.125523\n",
      "9     0.066946\n",
      "10    0.041841\n",
      "Name: instance, dtype: float64\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "count    239.000000\n",
       "mean       6.230126\n",
       "std        1.780311\n",
       "min        1.000000\n",
       "25%        5.000000\n",
       "50%        6.000000\n",
       "75%        7.000000\n",
       "max       10.000000\n",
       "Name: instance, dtype: float64"
      ]
     },
     "execution_count": 315,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_offtarget_num(off_target_counts_df_all_2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### version 3: analyze the scrambled version where we scramble a fixed 20bp length and raise the p val threshold to 0.05"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "we find that that the 20bp scrambling words better (less case 3s, more case 1s, and slight increase in case 4). that's why we're trying raising the p value threshold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 289,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# motif files:  158\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:43: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:44: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:45: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:46: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:59: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:60: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:61: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "saved ../data/processed/fig4_modelling/vocab_sum_final/motif_scan3_*.csv\n",
      "CPU times: user 1min 53s, sys: 4.16 s, total: 1min 58s\n",
      "Wall time: 1min 58s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "case 3    2886.0\n",
       "case 1    1627.0\n",
       "other      195.0\n",
       "case 4      72.0\n",
       "dtype: float64"
      ]
     },
     "execution_count": 289,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "# takes 30 s \n",
    "\n",
    "motif_scan_files3 = sorted(glob.glob(os.path.join(save_dir, 'motif_scan3','*csv')))\n",
    "off_target_counts_df_all_3, on_target_cases_df_all_3, on_target_config_cases_df_3 = analyze_motif_scan_wrap(motif_scan_files3, save_prefix=os.path.join(save_dir, 'motif_scan3_'))\n",
    "on_target_config_cases_df_3.sum(axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 291,
   "metadata": {},
   "outputs": [],
   "source": [
    "on_target_cases_df_all_3['case_mod'] = on_target_cases_df_all_3.apply(refine_case3, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 292,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "case 3    3625\n",
       "case 1    1989\n",
       "other      313\n",
       "case 4     153\n",
       "Name: case, dtype: int64"
      ]
     },
     "execution_count": 292,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# row = on_target_cases_df_all_3[on_target_cases_df_all_3.case=='case 3'].iloc[0,:]\n",
    "on_target_cases_df_all_3.case.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 293,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "case 3A    3216\n",
       "case 1     1989\n",
       "case 3B     409\n",
       "other       313\n",
       "case 4      153\n",
       "Name: case_mod, dtype: int64"
      ]
     },
     "execution_count": 293,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "on_target_cases_df_all_3.case_mod.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 333,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6080"
      ]
     },
     "execution_count": 333,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "on_target_cases_df_all_3.case_mod.value_counts().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 335,
   "metadata": {},
   "outputs": [],
   "source": [
    "on_target_cases_df_all_3.to_csv(os.path.join(save_dir, 'on_target_cases_df_all_3_mod.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 330,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    239.000000\n",
       "mean       8.736402\n",
       "std        1.369526\n",
       "min        5.000000\n",
       "25%        8.000000\n",
       "50%        9.000000\n",
       "75%       10.000000\n",
       "max       10.000000\n",
       "Name: name, dtype: float64"
      ]
     },
     "execution_count": 330,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_on_target_good_instance_num(on_target_cases_df_all_3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 294,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6080"
      ]
     },
     "execution_count": 294,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "on_target_cases_df_all_3.case_mod.value_counts().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 295,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4780.0"
      ]
     },
     "execution_count": 295,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "on_target_config_cases_df_3.sum(axis=0).sum() #double the number of instances b/c of the two vocab words per instance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 296,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "FN|FP|FN|FP    87\n",
       "FN|FN|FP|FP    52\n",
       "TP|FP|TP|TN    41\n",
       "TP|TN|TP|FP    39\n",
       "TP|TP|FP|TN    31\n",
       "TP|TP|TN|FP    23\n",
       "FN|FP|TP|FP     5\n",
       "FN|TN|FN|FP     5\n",
       "FN|TN|TP|TN     4\n",
       "FN|FP|FN|TN     4\n",
       "TP|FP|FN|FP     4\n",
       "TP|TN|FN|TN     3\n",
       "FN|TP|TN|FP     3\n",
       "FN|TP|TN|TN     2\n",
       "TP|FP|FN|TN     2\n",
       "FN|TP|FP|FP     2\n",
       "FN|FN|TN|FP     1\n",
       "FN|TN|TP|FP     1\n",
       "TP|FN|FP|TN     1\n",
       "TP|FN|FP|FP     1\n",
       "FN|FN|FP|TN     1\n",
       "TP|FN|TN|TN     1\n",
       "dtype: int64"
      ]
     },
     "execution_count": 296,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "on_target_cases_df_all_3[on_target_cases_df_all_3.case=='other'][['eval_A_B',\n",
    "                                                                  'eval_A_scrB',\t\n",
    "                                                                  'eval_scrA_B',\t\n",
    "                                                                  'eval_scrA_scrB']].apply('|'.join,axis=1).value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 297,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5227, 3)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "case 2B    3050\n",
       "case 2A    2177\n",
       "Name: case, dtype: int64"
      ]
     },
     "execution_count": 297,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(off_target_counts_df_all_3[['instance','motif','case']].drop_duplicates().shape)\n",
    "off_target_counts_df_all_3[['instance','motif','case']].drop_duplicates()['case'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 298,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4764, 2)"
      ]
     },
     "execution_count": 298,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "off_target_counts_df_all_3.reset_index()[['instance','motif']].drop_duplicates().shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 299,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>case</th>\n",
       "      <th>count_pos</th>\n",
       "      <th>instance</th>\n",
       "      <th>motif</th>\n",
       "      <th>pos_close</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>case 2A</td>\n",
       "      <td>4</td>\n",
       "      <td>KLF15::TFAP2B+Melanocytes+1</td>\n",
       "      <td>TFAP2B</td>\n",
       "      <td>42.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>case 2A</td>\n",
       "      <td>4</td>\n",
       "      <td>KLF15::TFAP2B+Melanocytes+10</td>\n",
       "      <td>TFAP2B</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>case 2A</td>\n",
       "      <td>4</td>\n",
       "      <td>KLF15::TFAP2B+Melanocytes+2</td>\n",
       "      <td>TFAP2B</td>\n",
       "      <td>82.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>case 2A</td>\n",
       "      <td>4</td>\n",
       "      <td>KLF15::TFAP2B+Melanocytes+3</td>\n",
       "      <td>TFAP2B</td>\n",
       "      <td>108.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>12</td>\n",
       "      <td>case 2A</td>\n",
       "      <td>4</td>\n",
       "      <td>KLF15::TFAP2B+Melanocytes+7</td>\n",
       "      <td>TFAP2B</td>\n",
       "      <td>94.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>28</td>\n",
       "      <td>case 2A</td>\n",
       "      <td>4</td>\n",
       "      <td>IRF4::ZNF816+GM12878+4</td>\n",
       "      <td>ZNF816</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>29</td>\n",
       "      <td>case 2A</td>\n",
       "      <td>4</td>\n",
       "      <td>IRF4::ZNF816+GM12878+5</td>\n",
       "      <td>ZNF816</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>30</td>\n",
       "      <td>case 2A</td>\n",
       "      <td>4</td>\n",
       "      <td>IRF4::ZNF816+GM12878+6</td>\n",
       "      <td>ZNF816</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>31</td>\n",
       "      <td>case 2A</td>\n",
       "      <td>4</td>\n",
       "      <td>IRF4::ZNF816+GM12878+7</td>\n",
       "      <td>ZNF816</td>\n",
       "      <td>104.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>51</td>\n",
       "      <td>case 2A</td>\n",
       "      <td>4</td>\n",
       "      <td>IRF8::ZNF816+GM12878+7</td>\n",
       "      <td>ZNF816</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2446 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       case  count_pos                      instance   motif  pos_close\n",
       "0   case 2A          4   KLF15::TFAP2B+Melanocytes+1  TFAP2B       42.0\n",
       "1   case 2A          4  KLF15::TFAP2B+Melanocytes+10  TFAP2B        4.0\n",
       "2   case 2A          4   KLF15::TFAP2B+Melanocytes+2  TFAP2B       82.0\n",
       "3   case 2A          4   KLF15::TFAP2B+Melanocytes+3  TFAP2B      108.0\n",
       "12  case 2A          4   KLF15::TFAP2B+Melanocytes+7  TFAP2B       94.0\n",
       "..      ...        ...                           ...     ...        ...\n",
       "28  case 2A          4        IRF4::ZNF816+GM12878+4  ZNF816        6.0\n",
       "29  case 2A          4        IRF4::ZNF816+GM12878+5  ZNF816        4.0\n",
       "30  case 2A          4        IRF4::ZNF816+GM12878+6  ZNF816        6.0\n",
       "31  case 2A          4        IRF4::ZNF816+GM12878+7  ZNF816      104.0\n",
       "51  case 2A          4        IRF8::ZNF816+GM12878+7  ZNF816        6.0\n",
       "\n",
       "[2446 rows x 5 columns]"
      ]
     },
     "execution_count": 299,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "off_target_counts_df_all_3[off_target_counts_df_all_3.case=='case 2A']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 308,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7     59\n",
       "8     46\n",
       "6     42\n",
       "9     35\n",
       "5     18\n",
       "4     15\n",
       "10    13\n",
       "3      5\n",
       "2      4\n",
       "1      2\n",
       "Name: instance, dtype: int64"
      ]
     },
     "execution_count": 308,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "off_target_counts_df_all_3[off_target_counts_df_all_3.case=='case 2A'].instance.drop_duplicates().apply(lambda x: '+'.join(x.split('+')[:-1])).value_counts().value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 316,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2446, 5)\n",
      "(2177, 3)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:127: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1    0.085470\n",
      "2    0.200855\n",
      "3    0.213675\n",
      "4    0.217949\n",
      "5    0.123932\n",
      "6    0.085470\n",
      "7    0.042735\n",
      "8    0.021368\n",
      "9    0.008547\n",
      "Name: instance, dtype: float64\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "count    234.000000\n",
       "mean       3.679487\n",
       "std        1.761488\n",
       "min        1.000000\n",
       "25%        2.000000\n",
       "50%        3.500000\n",
       "75%        5.000000\n",
       "max        9.000000\n",
       "Name: instance, dtype: float64"
      ]
     },
     "execution_count": 316,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_offtarget_num(off_target_counts_df_all_3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## AME motif enrichment for vocab words\n",
    "\n",
    "`ame --verbose 1 --oc . --scoring avg --method fisher --hit-lo-fraction 0.25 --evalue-report-threshold 10.0 --control --shuffle-- --kmer 2 mpra_vocab_seq2.fasta motifs.meme\n",
    "`\n",
    "\n",
    "AME motif enrichment neg control\n",
    "\n",
    "motifs: '../data/external/hocomoco_human_trim_jaspar_format.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 342,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "enriched motifs\n",
      "437\n",
      "['AHR', 'AR', 'ARID3A', 'ARNT', 'ARNT2', 'ARNTL', 'ASCL1', 'ASCL2', 'ATF1', 'ATF2', 'ATF3', 'ATF4', 'ATF6', 'ATF7', 'ATOH1', 'BACH1', 'BACH2', 'BARHL2', 'BATF', 'BATF3', 'BCL11A', 'BCL6', 'BHLHA15', 'BHLHE23', 'BHLHE40', 'BHLHE41', 'BPTF', 'BSX', 'CBFB', 'CDC5L', 'CDX1', 'CDX2', 'CEBPA', 'CEBPB', 'CEBPZ', 'CLOCK', 'CPEB1', 'CREB1', 'CREB3', 'CREB3L1', 'CREB3L2', 'CREB5', 'CREM', 'CTCF', 'CTCFL', 'DLX2', 'DLX4', 'DLX5', 'DLX6', 'DMBX1', 'DMRT1', 'E2F1', 'E2F3', 'E2F4', 'E2F6', 'E2F7', 'E2F8', 'E4F1', 'EGR1', 'EGR2', 'EGR3', 'EGR4', 'EHF', 'ELF1', 'ELF2', 'ELF3', 'ELF5', 'ELK1', 'ELK3', 'ELK4', 'EN1', 'EPAS1', 'ERG', 'ESR2', 'ESRRA', 'ESRRB', 'ESX1', 'ETS1', 'ETS2', 'ETV1', 'ETV2', 'ETV3', 'ETV4', 'ETV5', 'ETV6', 'ETV7', 'FEV', 'FEZF1', 'FIGLA', 'FLI1', 'FOS', 'FOSB', 'FOSL1', 'FOSL2', 'FOXA1', 'FOXA2', 'FOXA3', 'FOXB1', 'FOXC1', 'FOXC2', 'FOXD1', 'FOXD2', 'FOXD3', 'FOXF1', 'FOXF2', 'FOXG1', 'FOXH1', 'FOXI1', 'FOXJ2', 'FOXJ3', 'FOXK1', 'FOXL1', 'FOXM1', 'FOXO1', 'FOXO3', 'FOXO4', 'FOXO6', 'FOXP1', 'FOXP2', 'FOXP3', 'FOXQ1', 'FUBP1', 'GABPA', 'GATA3', 'GBX1', 'GBX2', 'GLI3', 'GSX1', 'GSX2', 'HAND1', 'HEY1', 'HIF1A', 'HMGA1', 'HMX1', 'HMX2', 'HNF1A', 'HNF4A', 'HNF4G', 'HOXA10', 'HOXA11', 'HOXA13', 'HOXB13', 'HOXB3', 'HOXB4', 'HOXB6', 'HOXC10', 'HOXC11', 'HOXC12', 'HOXC13', 'HOXC6', 'HOXC9', 'HOXD10', 'HOXD11', 'HOXD12', 'HOXD13', 'HOXD4', 'HOXD9', 'HSF1', 'HSFY1; HSFY2', 'ID4', 'IKZF1', 'IRF1', 'IRF2', 'IRF3', 'IRF4', 'IRF5', 'IRF6', 'IRF7', 'IRF8', 'IRF9', 'IRX2', 'JDP2', 'JUN', 'JUNB', 'JUND', 'KLF1', 'KLF12', 'KLF13', 'KLF14', 'KLF15', 'KLF16', 'KLF3', 'KLF4', 'KLF5', 'KLF6', 'KLF9', 'LEF1', 'LHX3', 'LHX4', 'LHX9', 'LMX1A', 'LMX1B', 'LYL1', 'MAF', 'MAFA', 'MAFB', 'MAFF', 'MAFG', 'MAFK', 'MAX', 'MAZ', 'MEF2A', 'MEF2B', 'MEF2C', 'MEF2D', 'MEIS1', 'MEIS2', 'MEIS3', 'MEOX2', 'MESP1', 'MITF', 'MLX', 'MNX1', 'MSX1', 'MSX2', 'MXI1', 'MYC', 'MYCN', 'MYF6', 'MYNN', 'MYOD1', 'MYOG', 'NANOG', 'NEUROD1', 'NEUROD2', 'NEUROG2', 'NFAT5', 'NFATC1', 'NFATC2', 'NFATC3', 'NFATC4', 'NFE2', 'NFE2L1', 'NFE2L2', 'NFYA', 'NFYB', 'NFYC', 'NHLH1', 'NKX2-2', 'NKX2-5', 'NKX6-1', 'NKX6-2', 'NOTO', 'NR1H3', 'NR2C1', 'NR2E1', 'NR2E3', 'NR2F1', 'NR2F2', 'NR2F6', 'NR3C1', 'NR4A1', 'NR4A2', 'NR4A3', 'NRF1', 'NRL', 'OLIG1', 'OLIG2', 'OSR2', 'OVOL1', 'PATZ1', 'PAX4', 'PAX5', 'PBX1', 'PBX3', 'PGR', 'PLAG1', 'POU1F1', 'POU2F1', 'POU2F2', 'POU2F3', 'POU3F2', 'POU3F3', 'POU3F4', 'POU4F2', 'POU5F1', 'POU5F1B', 'POU6F1', 'PPARA', 'PPARD', 'PPARG', 'PRDM1', 'PRDM6', 'PROX1', 'PRRX2', 'PTF1A', 'PURA', 'RARA', 'RARB', 'RARG', 'RBPJ', 'RFX1', 'RFX2', 'RFX3', 'RFX4', 'RFX5', 'RORA', 'RREB1', 'RUNX1', 'RUNX2', 'RUNX3', 'RXRA', 'RXRB', 'RXRG', 'SALL4', 'SHOX', 'SIX1', 'SIX2', 'SMAD1', 'SMAD2', 'SMAD3', 'SMARCA1', 'SMARCA5', 'SNAI1', 'SNAI2', 'SOX10', 'SOX11', 'SOX15', 'SOX17', 'SOX2', 'SOX21', 'SOX3', 'SOX4', 'SOX8', 'SOX9', 'SP1', 'SP2', 'SP3', 'SP4', 'SPDEF', 'SPI1', 'SPIB', 'SREBF1', 'SREBF2', 'SRF', 'SRY', 'STAT1', 'STAT2', 'STAT3', 'STAT4', 'STAT5A', 'STAT5B', 'STAT6', 'TAF1', 'TAL1', 'TBP', 'TBX1', 'TBX15', 'TBX20', 'TBX21', 'TBX3', 'TBX4', 'TBX5', 'TBXT', 'TCF12', 'TCF3', 'TCF4', 'TCF7', 'TEAD1', 'TEAD2', 'TEAD3', 'TEAD4', 'TEF', 'TFAP4', 'TFDP1', 'TFE3', 'TFEB', 'TGIF1', 'TGIF2', 'TGIF2LX', 'THAP1', 'THAP11', 'THRA', 'THRB', 'TP63', 'TWIST1', 'UBP1', 'USF1', 'USF2', 'VENTX', 'VEZF1', 'WT1', 'XBP1', 'YY1', 'ZBTB14', 'ZBTB17', 'ZBTB18', 'ZBTB33', 'ZBTB7B', 'ZEB1', 'ZFHX3', 'ZFP28', 'ZFP64', 'ZFP82', 'ZFX', 'ZIC1', 'ZIC3', 'ZIM3', 'ZNF121', 'ZNF143', 'ZNF148', 'ZNF18', 'ZNF214', 'ZNF219', 'ZNF263', 'ZNF274', 'ZNF281', 'ZNF282', 'ZNF322', 'ZNF324', 'ZNF329', 'ZNF331', 'ZNF333', 'ZNF335', 'ZNF341', 'ZNF350', 'ZNF354A', 'ZNF384', 'ZNF394', 'ZNF41', 'ZNF418', 'ZNF436', 'ZNF449', 'ZNF467', 'ZNF528', 'ZNF554', 'ZNF563', 'ZNF586', 'ZNF708', 'ZNF713', 'ZNF76', 'ZNF768', 'ZNF770', 'ZNF784', 'ZNF816', 'ZSCAN22', 'ZSCAN31']\n",
      "enriched motifs with tested vocab pair\n",
      "(490, 19) (131, 19)\n",
      "106\n",
      "['ARNT', 'ASCL2', 'BACH1', 'BATF', 'CBFB', 'E2F1', 'E2F3', 'E2F4', 'E2F6', 'E2F7', 'EGR1', 'EGR2', 'EGR3', 'EGR4', 'ETS1', 'ETV2', 'ETV5', 'FLI1', 'FOS', 'FOSL2', 'FOXD3', 'FOXF2', 'FOXL1', 'FOXM1', 'HMGA1', 'ID4', 'IKZF1', 'IRF1', 'IRF2', 'IRF3', 'IRF4', 'IRF8', 'JUN', 'JUNB', 'JUND', 'KLF15', 'KLF16', 'KLF4', 'KLF5', 'KLF6', 'LEF1', 'MAF', 'MAFB', 'MAFF', 'MAFG', 'MAFK', 'MAZ', 'MESP1', 'MITF', 'MSX2', 'NFAT5', 'NFATC1', 'NFATC4', 'NFE2L1', 'NFYB', 'NR2F1', 'NR3C1', 'NRF1', 'OSR2', 'PATZ1', 'PAX5', 'PBX1', 'PLAG1', 'PRDM1', 'PURA', 'RARA', 'RARG', 'RFX2', 'RORA', 'RUNX3', 'SALL4', 'SMAD3', 'SNAI2', 'SOX10', 'SP1', 'SP2', 'SP3', 'SP4', 'SPIB', 'SREBF2', 'STAT1', 'STAT2', 'STAT3', 'TBX1', 'TBX15', 'TCF3', 'TFDP1', 'TFE3', 'TGIF1', 'THAP1', 'THRB', 'TP63', 'UBP1', 'VEZF1', 'WT1', 'XBP1', 'ZBTB14', 'ZBTB17', 'ZFP28', 'ZFX', 'ZNF219', 'ZNF324', 'ZNF394', 'ZNF563', 'ZNF770', 'ZNF816']\n"
     ]
    }
   ],
   "source": [
    "#read in motif enrichment results\n",
    "ame_result_file = os.path.join(save_dir, 'motif_scan_vocab_ame.tsv')\n",
    "ame_result = pd.read_csv(ame_result_file, sep='\\t')\n",
    "ame_result['motif_DB'] = 'hocomoco_human_trim'\n",
    "ame_result['motif'] = ame_result.motif_ID.map(tf_id_abbr_to_name_dict)\n",
    "ame_result = ame_result[~ame_result.motif.isna()]\n",
    "ame_result['in_vocab'] = ame_result.motif.isin(list_tfs)\n",
    "\n",
    "print('enriched motifs')\n",
    "enriched_motifs = sorted(ame_result.motif.unique())\n",
    "print(len(enriched_motifs))\n",
    "print(enriched_motifs)\n",
    "\n",
    "print('enriched motifs with tested vocab pair')\n",
    "ame_result_filt = ame_result[ame_result.in_vocab]\n",
    "print(ame_result.shape, ame_result_filt.shape)\n",
    "enriched_motifs_vocab = sorted(ame_result_filt.motif.unique())\n",
    "print(len(enriched_motifs_vocab))\n",
    "print(enriched_motifs_vocab)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
