{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "09/04/2020\n",
    "\n",
    "post scan scrambling the the homotype motifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 242,
   "metadata": {},
   "outputs": [],
   "source": [
    "# basic packages\n",
    "import re\n",
    "import os, glob\n",
    "import pandas as pd\n",
    "import numpy as np; np.random.seed(0)\n",
    "import itertools\n",
    "from collections import Counter, defaultdict\n",
    "import time\n",
    "# Import tools needed for visualization\n",
    "import seaborn as sns; sns.set()\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import pybedtools\n",
    "from random import shuffle\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_dir = '../data/processed/fig4_modelling/vocab_sum_final/'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_dir = '../data/processed/fig4_modelling/vocab_sum_final_post_scan/'\n",
    "if not os.path.exists(save_dir):\n",
    "    os.makedirs(save_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 0. Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 0A Global Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "metadata": {},
   "outputs": [],
   "source": [
    "tissue_to_celltype_mapper = {\n",
    "    'Colon':['Colon'],\n",
    "    'GDSD6':['GDSD6'],\n",
    "    'GM12878':['GM12878'],\n",
    "    'Melanocytes':['Melanocytes'],\n",
    "    'SCC':['A431-CTRLi','CAL27-CTRLi','SCC13-CTRLi'],\n",
    "    'MM':['COLO-SCR-DMSO','WM-SCR-DMSO']\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%time\n",
    "# # import\n",
    "# data_all = pd.read_csv('/Users/mguo123/Google Drive/1_khavari/omics_project-LD/pan_omics/data/processed/tissue_crms/all_count_comb_overall.csv',index_col=0,header=0)\n",
    "# data_all = data_all[data_all.tissue.isin(['Colon','GDSD6','GM12878','Melanocytes','A431-CTRLi','CAL27-CTRLi','SCC13-CTRLi','COLO_SCR_DMSO','WM_SCR_DMSO'])]\n",
    "# data_all = data_all[data_all.iloc[:,2:].sum(axis=1)>1e-1]\n",
    "\n",
    "# # expression labels\n",
    "# exp_label = list(np.log10(data_all.exp.values+1e-2))\n",
    "# labels_all  = np.array(np.array(exp_label)>THRES)\n",
    "\n",
    "\n",
    "# tissues_label  = data_all.tissue.values#np.array((data_all.exp>THRES).values)\n",
    "# tissue_num_labels =  data_all.tissue.map(tissues_dict).values\n",
    "# is_cancer_label = np.array(['cancer' if x in cancer_tissues  else 'normal' for x in tissues_label ])\n",
    "# tissues_label[tissues_label=='COLO_SCR_DMSO'] = 'COLO-SCR-DMSO'\n",
    "# tissues_label[tissues_label=='WM_SCR_DMSO'] = 'WM-SCR-DMSO'\n",
    "\n",
    "# genes_all = data_all.index.values\n",
    "# gene_to_num_dict = dict(zip(np.unique(genes_all),range(len(np.unique(genes_all)))))\n",
    "# genes_num_all = np.vectorize(gene_to_num_dict.get)(genes_all)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>loc</th>\n",
       "      <th>end</th>\n",
       "      <th>width</th>\n",
       "      <th>strand</th>\n",
       "      <th>TSS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1846266</td>\n",
       "      <td>1846266</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>CALML6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>231376933</td>\n",
       "      <td>231376933</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>C1orf131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>241683061</td>\n",
       "      <td>241683061</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>FH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>120202421</td>\n",
       "      <td>120202421</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>PHGDH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>204135465</td>\n",
       "      <td>204135465</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>REN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>20255</td>\n",
       "      <td>chrX</td>\n",
       "      <td>140084756</td>\n",
       "      <td>140084756</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>SPANXB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>20256</td>\n",
       "      <td>chrX</td>\n",
       "      <td>52683950</td>\n",
       "      <td>52683950</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>SSX7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>20257</td>\n",
       "      <td>chrX</td>\n",
       "      <td>140096761</td>\n",
       "      <td>140096761</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>SPANXB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>20258</td>\n",
       "      <td>chrX</td>\n",
       "      <td>52736239</td>\n",
       "      <td>52736239</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>SSX2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>20259</td>\n",
       "      <td>chrX</td>\n",
       "      <td>147582139</td>\n",
       "      <td>147582139</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>AFF2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20178 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr        loc        end  width strand       TSS\n",
       "0      chr1    1846266    1846266      1      *    CALML6\n",
       "1      chr1  231376933  231376933      1      *  C1orf131\n",
       "2      chr1  241683061  241683061      1      *        FH\n",
       "3      chr1  120202421  120202421      1      *     PHGDH\n",
       "4      chr1  204135465  204135465      1      *       REN\n",
       "...     ...        ...        ...    ...    ...       ...\n",
       "20255  chrX  140084756  140084756      1      *   SPANXB2\n",
       "20256  chrX   52683950   52683950      1      *      SSX7\n",
       "20257  chrX  140096761  140096761      1      *   SPANXB1\n",
       "20258  chrX   52736239   52736239      1      *      SSX2\n",
       "20259  chrX  147582139  147582139      1      *      AFF2\n",
       "\n",
       "[20178 rows x 6 columns]"
      ]
     },
     "execution_count": 247,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tss_df = pd.read_csv('../data/external/TSS_annon_hg19.csv',index_col=0).drop_duplicates('TSS')\n",
    "tss_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%time\n",
    "# tissue_to_foot_df = {}\n",
    "# foot_tissues = ['Colon','GDSD6','GM12878','Melanocytes','A431-CTRLi','CAL27-CTRLi','SCC13-CTRLi','COLO-SCR-DMSO','WM-SCR-DMSO']\n",
    "# for tissue in foot_tissues:\n",
    "#     foot_df = pd.read_csv(os.path.join('../data/interim/annon/promoter_motif_scan/','promoter_'+tissue+'_annon.bed'),sep='\\t',header=None)\n",
    "#     foot_df.columns = ['chr','start','stop','TSS','chr_m','start_m','stop_m','id_trim','score','strand']\n",
    "#     tissue_to_foot_df[tissue] = foot_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 249,
   "metadata": {},
   "outputs": [],
   "source": [
    "tf_annon_df = pd.read_csv('../data/external/HOCOMOCOv11_annotation.csv',index_col=0)\n",
    "tf_annon_df['id_trim'] = tf_annon_df['id'] + '.pwm.trim'\n",
    "tf_name_to_id_dict = pd.Series(tf_annon_df.id_trim.values, index=tf_annon_df.tf.values).to_dict()\n",
    "tf_id_to_name_dict = pd.Series(tf_annon_df.tf.values, index=tf_annon_df.id_trim.values).to_dict()\n",
    "tf_name_to_id_abbr_dict = pd.Series(tf_annon_df.id.values, index=tf_annon_df.tf.values).to_dict()\n",
    "tf_id_abbr_to_name_dict = pd.Series(tf_annon_df.tf.values, index=tf_annon_df.id.values).to_dict()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 250,
   "metadata": {},
   "outputs": [],
   "source": [
    "MAX_DIST=135\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# define constants\n",
    "LEN_FILLER = 20\n",
    "NUM_OLIGO_PER_MOTIF = 10\n",
    "DESIRED_OLIGO_LENGTH = 221#150\n",
    "MAX_OLIGO_LENGTH = 230\n",
    "MAX_MOTIF_LENGTH = 63\n",
    "FORWARD_PCR_PRIMER = 'ACTGGCCGCTTCACTG'\n",
    "REVERSE_PCR_PRIMER = 'AGATCGGAAGAGCGTCG'\n",
    "# LINKER_SPACER_v10 = 'CAGTATGCGATGCTCATGATTGTGACTGAAAGTCTATGAGTCCATTGA'\n",
    "RS_ECORI = 'GAATTC' # 5'-3'\n",
    "RS_BAMHI = 'GGATCC'\n",
    "RS_XHOI = 'CTCGAG'\n",
    "RS_XBAI = 'TCTAGA'\n",
    "RS_BSPDI = 'ATCGAT'\n",
    "RS_NHEI = 'GCTAGC'\n",
    "RS_NCOI = 'CCATGG'\n",
    "\n",
    "# CCTAGG formed by RS_XHOI and RS_NHEI, can be cut by AvrII\n",
    "\n",
    "# RS_LIST = [RS_ECORI ,RS_BAMHI , RS_XHOI , RS_XBAI] #original\n",
    "# RS_LIST = [RS_BSPDI ,RS_BAMHI , RS_XHOI, RS_NHEI]\n",
    "RS_LIST = [RS_ECORI ,RS_BAMHI , RS_XHOI , RS_NHEI]\n",
    "\n",
    "RS_XBAI_dam1 = 'GATCTAGA'\n",
    "RS_XBAI_dam2 = 'TCTAGATC'\n",
    "RS_XBAI_dam1_pre = '^GA'\n",
    "RS_XBAI_dam2_suf = 'TC$'\n",
    "LETTERS=['A', 'C', 'G', 'T']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 0B helper functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 473,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def check_rs(seq, rs_list = RS_LIST):\n",
    "    rs_sites = 0\n",
    "    for rs in rs_list:\n",
    "        if rs in seq:\n",
    "            rs_sites+=1\n",
    "    return rs_sites\n",
    "\n",
    "\n",
    "\n",
    "def ranges(nums):\n",
    "    # get consecutive integers in a list\n",
    "    nums = sorted(set(nums))\n",
    "    gaps = [[s, e] for s, e in zip(nums, nums[1:]) if s+1 < e]\n",
    "    edges = iter(nums[:1] + sum(gaps, []) + nums[-1:])\n",
    "    return list(zip(edges, edges))\n",
    "\n",
    "def mixup(word):\n",
    "    as_list_of_letters = list(word)\n",
    "    shuffle(as_list_of_letters)\n",
    "    return ''.join(as_list_of_letters)\n",
    "\n",
    "\n",
    "def get_vocab_loc(row, motif_len_scr=10):\n",
    "    # getting vocab locations in the sequence\n",
    "    vocab1_med = int((row['vocab1_start']+row['vocab1_stop'])/2)\n",
    "    vocab1_start_shift = vocab1_med - motif_len_scr - row['start_pad']\n",
    "    vocab1_stop_shift = vocab1_med + motif_len_scr -row['start_pad'] + 1\n",
    "    \n",
    "    vocab2_med = int((row['vocab2_start']+row['vocab2_stop'])/2)\n",
    "    vocab2_start_shift = vocab2_med - motif_len_scr - row['start_pad']\n",
    "    vocab2_stop_shift = vocab2_med + motif_len_scr -row['start_pad'] + 1\n",
    "    \n",
    "    # for overlaps\n",
    "    if vocab1_stop_shift>= vocab2_start_shift:\n",
    "#         print(idx)\n",
    "#         print(row)\n",
    "        \n",
    "        # case 1 vocab words are really close, the padding goes into the vocab word\n",
    "        if (row['vocab1_stop']-row['start_pad'] + 1) >=vocab2_start_shift:\n",
    "            \n",
    "            vocab1_stop_shift = row['vocab1_stop']-row['start_pad'] + 1\n",
    "            vocab2_start_shift = row['vocab1_stop']-row['start_pad'] + 1\n",
    "#             print(vocab1_start_shift,vocab1_stop_shift, vocab2_start_shift, vocab2_stop_shift)\n",
    "#             raise\n",
    "        else: \n",
    "            vocab1_stop_shift = vocab1_stop_shift\n",
    "            vocab2_start_shift = vocab1_stop_shift\n",
    "#             print(vocab1_start_shift,vocab1_stop_shift, vocab2_start_shift, vocab2_stop_shift)\n",
    "#             raise\n",
    "\n",
    "    # for edges:\n",
    "    vocab1_start_shift = max(0,vocab1_start_shift)\n",
    "    vocab2_stop_shift = min(len(row['seq']),vocab2_stop_shift)\n",
    "    \n",
    "    return pd.Series({'vocab1_start_shift': vocab1_start_shift,\n",
    "                      'vocab1_stop_shift':vocab1_stop_shift, \n",
    "                      'vocab2_start_shift':vocab2_start_shift, \n",
    "                      'vocab2_stop_shift':vocab2_stop_shift})\n",
    "\n",
    "\n",
    "def get_vocab_to_loc_dict(row):\n",
    "    return {row['vocab1']:[row['vocab1_start_shift'],row['vocab1_stop_shift']],\n",
    "           row['vocab2']:[row['vocab2_start_shift'],row['vocab2_stop_shift']]}\n",
    "\n",
    "\n",
    "\n",
    "def scramble_off_target(row,motif_len_scr=10, verbose=False):\n",
    "    # get places that we can't scramble\n",
    "    # don't change anything if no homotypic sequences\n",
    "    if row['off_target_name']=='':\n",
    "        return row['seq']\n",
    "\n",
    "    # get vocab lengths\n",
    "    vocab_word_to_loc = get_vocab_to_loc_dict(row)\n",
    "    if verbose:\n",
    "        print('vocab_word_to_loc', vocab_word_to_loc)\n",
    "    # start mask defaults\n",
    "    mask = np.zeros(len(row['seq'][0]),dtype=int)\n",
    "\n",
    "    for motif,pos_str in zip(row['off_target_motif'].split('::'),row['off_target_pos_str'].split('::')):\n",
    "        pos_arr = [int(x) for x in pos_str.split('|')]\n",
    "#         print(motif, pos_arr, vocab_word_to_loc[motif])\n",
    "\n",
    "        v_start, v_stop = vocab_word_to_loc[motif]\n",
    "        v_len = v_stop - v_start\n",
    "        for pos in pos_arr:\n",
    "#             ### OPTION 1 (do +/- 10 bp)\n",
    "#             pos_med = int(pos+v_len/2)\n",
    "#             pos_start = pos_med - motif_len_scr\n",
    "#             pos_stop = pos_med + motif_len_scr + 1\n",
    "            \n",
    "            ### option 2 (just scr motif area)\n",
    "            pos_start = max(0,pos)\n",
    "            pos_stop = min(len(row['seq'][0]),pos_start + v_len)\n",
    "            \n",
    "            mask[pos_start:pos_stop] = 1\n",
    "\n",
    "    # only considering masking non vocab areas\n",
    "    mask[row['vocab1_start_shift']:row['vocab1_stop_shift']] = 2\n",
    "    mask[row['vocab2_start_shift']:row['vocab2_stop_shift']] = 2\n",
    "    if verbose:\n",
    "        print('mask',mask)\n",
    "    mask_indices = np.where(mask==1)[0]\n",
    "    \n",
    "    scr_ranges = ranges(mask_indices)\n",
    "    if verbose:\n",
    "        print('scr_ranges',scr_ranges)\n",
    "        \n",
    "    new_seqs = row['seq'].copy()\n",
    "    if len(new_seqs)!=4:\n",
    "        raise ValueError('should have 4 sequences')\n",
    "#     print(len(new_seqs))\n",
    "    if len(scr_ranges)>0:\n",
    "        # iterate through scr ranges\n",
    "        for (scr_start, scr_stop) in scr_ranges:\n",
    "            # always take the first sequence (the A_B_seq config)\n",
    "            seq_scr_ori = new_seqs[0][scr_start:scr_stop]\n",
    "            seq_scr = mixup(seq_scr_ori)\n",
    "            for seq_idx in range(len(new_seqs)):\n",
    "                new_seqs[seq_idx] = new_seqs[seq_idx][:scr_start] + seq_scr + new_seqs[seq_idx][scr_stop:]\n",
    "                \n",
    "    # check if the first sequence length is the same\n",
    "    if len(new_seqs[0])!=len(row['seq'][0]):\n",
    "        print(new_seqs[0])\n",
    "        print(row['seq'][0])\n",
    "        raise ValueError('seq lengths are not the same post scrambling')\n",
    "\n",
    "    return new_seqs\n",
    "\n",
    "\n",
    "def off_target_pos_str_find_num(pos_str):\n",
    "    pos_arr_per_vocab_word = pos_str.split('::')\n",
    "    if len(pos_arr_per_vocab_word)>0:\n",
    "        arr = []\n",
    "        for pos in pos_arr_per_vocab_word:\n",
    "            for x in pos.split(':'):\n",
    "                try:\n",
    "                    arr.append(int(x))\n",
    "                except:\n",
    "                    continue\n",
    "        return len(set(arr))\n",
    "    return 0\n",
    "\n",
    "#     vocab1_seq = row['sequence'][vocab1_start_shift:vocab1_stop_shift]\n",
    "#     vocab2_seq = row['sequence'][vocab2_start_shift:vocab2_stop_shift]\n",
    "\n",
    "#     vocab1_seq_scr = mixup(vocab1_seq)\n",
    "#     vocab2_seq_scr = mixup(vocab2_seq)\n",
    "\n",
    "#     A_B_seq = row['sequence']\n",
    "#     scrA_B_seq = row['sequence'][:vocab1_start_shift] + vocab1_seq_scr +  row['sequence'][vocab1_stop_shift:]\n",
    "#     A_scrB_seq = row['sequence'][:vocab2_start_shift] + vocab2_seq_scr +  row['sequence'][vocab2_stop_shift:]\n",
    "#     scrA_scrB_seq = row['sequence'][:vocab1_start_shift] + vocab1_seq_scr +  row['sequence'][vocab1_stop_shift:vocab2_start_shift] + vocab2_seq_scr +  row['sequence'][vocab2_stop_shift:]\n",
    "\n",
    "#     # FOR DEBUG\n",
    "#     print(A_B_seq)\n",
    "#     print(scrA_B_seq)\n",
    "#     print(A_scrB_seq)\n",
    "#     print(scrA_scrB_seq)\n",
    "\n",
    "#     return A_B_seq, scrA_B_seq, A_scrB_seq, scrA_scrB_seq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Read in df for building library (vocab and add in homotypic locations\n",
    "\n",
    "#### RECALL CASES\n",
    "\n",
    "\n",
    "cases for the motif in the \n",
    "1. scrambling removes motif \n",
    "2. there's a dummy motif somewhere else in the genomic instance (????? How would you deal with these “noisy” cases?\n",
    "    2A. is the dummy motif is present in all 4 cases, (not changed by scrambling)\n",
    "    2B. the dummy motif is present in only some cases, so probably changed by scrambling the other site?\n",
    "3. scrambling reintroduces the motif at a (hopefully) lower strength, and sometimes with a shifted start position  ??? Potential to rescramble (but sometimes motifs have too many repetitive nucleotides so this might not work…\n",
    "    3A. the scrambling introduces the motif at a lower strength\n",
    "    3B. the scrambling does not lower the score of the motif\n",
    "4. we don't detect the motif in any config in the genomic instance at a certain p value threshold (this usually isn’t the case, if we raise the p-value threshold enough; but the motif will be detected in a lot more genomic instances). This is sort of the sensitivity vs specificity tradeoff.\n",
    "\n",
    "\n",
    "todo\n",
    "for case 2A (homotype motifs) scramble those)\n",
    "\n",
    "1. map the vocab word locations to motifs and match to case above (get the score and the case) - exact match\n",
    "2. determine vocab words gained lost between the motif configurations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 471,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "vocab_instances20_seqs_sel = pd.read_csv(os.path.join(input_dir, 'vocab_instances20_seqs_sel.csv'),index_col=0)\n",
    "pos_genes_df_rs_filt = pd.read_csv(os.path.join(input_dir, 'pos_genes_position_wseq_rsfilt.csv'),index_col=0)\n",
    "neg_genes_df_rs_filt = pd.read_csv(os.path.join(input_dir, 'neg_genes_position_wseq_rsfilt.csv'),index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 441,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['vocab_pair', 'tot_dist', 'btn_dist', 'chr', 'start', 'stop',\n",
       "       'start_pad', 'stop_pad', 'vocab1', 'vocab1_start', 'vocab1_stop',\n",
       "       'vocab2', 'vocab2_start', 'vocab2_stop', 'tissue', 'loop_type',\n",
       "       'vocab_tissue_count', 'name', 'check_rs', 'scr_pass_rs', 'seq_config',\n",
       "       'seq', 'name_config'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 441,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_seqs_sel.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 442,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([135])"
      ]
     },
     "execution_count": 442,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_seqs_sel.seq.apply(len).unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 443,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>vocab_pair</th>\n",
       "      <th>tot_dist</th>\n",
       "      <th>btn_dist</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>start_pad</th>\n",
       "      <th>stop_pad</th>\n",
       "      <th>vocab1</th>\n",
       "      <th>vocab1_start</th>\n",
       "      <th>...</th>\n",
       "      <th>vocab2_stop</th>\n",
       "      <th>tissue</th>\n",
       "      <th>loop_type</th>\n",
       "      <th>vocab_tissue_count</th>\n",
       "      <th>name</th>\n",
       "      <th>check_rs</th>\n",
       "      <th>scr_pass_rs</th>\n",
       "      <th>seq_config</th>\n",
       "      <th>seq</th>\n",
       "      <th>name_config</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>BARX2::FOXF2</td>\n",
       "      <td>28</td>\n",
       "      <td>14</td>\n",
       "      <td>chr1</td>\n",
       "      <td>206848337</td>\n",
       "      <td>206848365</td>\n",
       "      <td>206848284</td>\n",
       "      <td>206848418</td>\n",
       "      <td>BARX2</td>\n",
       "      <td>206848337</td>\n",
       "      <td>...</td>\n",
       "      <td>206848365</td>\n",
       "      <td>Colon</td>\n",
       "      <td>pro_pro</td>\n",
       "      <td>1</td>\n",
       "      <td>BARX2::FOXF2+Colon+1</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>A_B_seq</td>\n",
       "      <td>CTCAGCTCTCCCACAAGATCTCTGGGGAAGCTGGAAGAGGGCCACT...</td>\n",
       "      <td>BARX2::FOXF2+Colon+1+A_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>BARX2::FOXF2</td>\n",
       "      <td>54</td>\n",
       "      <td>40</td>\n",
       "      <td>chr4</td>\n",
       "      <td>38859601</td>\n",
       "      <td>38859655</td>\n",
       "      <td>38859561</td>\n",
       "      <td>38859695</td>\n",
       "      <td>FOXF2</td>\n",
       "      <td>38859601</td>\n",
       "      <td>...</td>\n",
       "      <td>38859655</td>\n",
       "      <td>Colon</td>\n",
       "      <td>loop_loop</td>\n",
       "      <td>2</td>\n",
       "      <td>BARX2::FOXF2+Colon+2</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>A_B_seq</td>\n",
       "      <td>GTGGCTCTGAGGGAATTTCAAGTAACCGTGGACTAAAGTCTTAAGA...</td>\n",
       "      <td>BARX2::FOXF2+Colon+2+A_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>BARX2::FOXF2</td>\n",
       "      <td>46</td>\n",
       "      <td>32</td>\n",
       "      <td>chr6</td>\n",
       "      <td>74222776</td>\n",
       "      <td>74222822</td>\n",
       "      <td>74222732</td>\n",
       "      <td>74222866</td>\n",
       "      <td>FOXF2</td>\n",
       "      <td>74222776</td>\n",
       "      <td>...</td>\n",
       "      <td>74222822</td>\n",
       "      <td>Colon</td>\n",
       "      <td>loop_loop</td>\n",
       "      <td>3</td>\n",
       "      <td>BARX2::FOXF2+Colon+3</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>A_B_seq</td>\n",
       "      <td>GCCACTGCACTCCAGCCTGGTGACAGAGCGAGACTCTGTCTCAAGA...</td>\n",
       "      <td>BARX2::FOXF2+Colon+3+A_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>BARX2::FOXF2</td>\n",
       "      <td>16</td>\n",
       "      <td>2</td>\n",
       "      <td>chr3</td>\n",
       "      <td>49384674</td>\n",
       "      <td>49384690</td>\n",
       "      <td>49384615</td>\n",
       "      <td>49384749</td>\n",
       "      <td>FOXF2</td>\n",
       "      <td>49384674</td>\n",
       "      <td>...</td>\n",
       "      <td>49384690</td>\n",
       "      <td>Colon</td>\n",
       "      <td>pro_pro</td>\n",
       "      <td>4</td>\n",
       "      <td>BARX2::FOXF2+Colon+4</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>A_B_seq</td>\n",
       "      <td>AGTATTATGTGCTCATGGTCCCTACCCTCCAGGGTTCACAGTGTGG...</td>\n",
       "      <td>BARX2::FOXF2+Colon+4+A_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>BARX2::FOXF2</td>\n",
       "      <td>16</td>\n",
       "      <td>2</td>\n",
       "      <td>chr10</td>\n",
       "      <td>112247520</td>\n",
       "      <td>112247536</td>\n",
       "      <td>112247461</td>\n",
       "      <td>112247595</td>\n",
       "      <td>BARX2</td>\n",
       "      <td>112247520</td>\n",
       "      <td>...</td>\n",
       "      <td>112247536</td>\n",
       "      <td>Colon</td>\n",
       "      <td>pro_pro</td>\n",
       "      <td>5</td>\n",
       "      <td>BARX2::FOXF2+Colon+5</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>A_B_seq</td>\n",
       "      <td>ATGTGGCTGCGGGGCTTTAGGAGCTGACTCTTTGGTCACGGGGAGA...</td>\n",
       "      <td>BARX2::FOXF2+Colon+5+A_B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     vocab_pair  tot_dist  btn_dist    chr      start       stop  start_pad  \\\n",
       "0  BARX2::FOXF2        28        14   chr1  206848337  206848365  206848284   \n",
       "1  BARX2::FOXF2        54        40   chr4   38859601   38859655   38859561   \n",
       "2  BARX2::FOXF2        46        32   chr6   74222776   74222822   74222732   \n",
       "3  BARX2::FOXF2        16         2   chr3   49384674   49384690   49384615   \n",
       "4  BARX2::FOXF2        16         2  chr10  112247520  112247536  112247461   \n",
       "\n",
       "    stop_pad vocab1  vocab1_start  ...  vocab2_stop tissue  loop_type  \\\n",
       "0  206848418  BARX2     206848337  ...    206848365  Colon    pro_pro   \n",
       "1   38859695  FOXF2      38859601  ...     38859655  Colon  loop_loop   \n",
       "2   74222866  FOXF2      74222776  ...     74222822  Colon  loop_loop   \n",
       "3   49384749  FOXF2      49384674  ...     49384690  Colon    pro_pro   \n",
       "4  112247595  BARX2     112247520  ...    112247536  Colon    pro_pro   \n",
       "\n",
       "   vocab_tissue_count                  name check_rs  scr_pass_rs seq_config  \\\n",
       "0                   1  BARX2::FOXF2+Colon+1        0         True    A_B_seq   \n",
       "1                   2  BARX2::FOXF2+Colon+2        0         True    A_B_seq   \n",
       "2                   3  BARX2::FOXF2+Colon+3        0         True    A_B_seq   \n",
       "3                   4  BARX2::FOXF2+Colon+4        0         True    A_B_seq   \n",
       "4                   5  BARX2::FOXF2+Colon+5        0         True    A_B_seq   \n",
       "\n",
       "                                                 seq               name_config  \n",
       "0  CTCAGCTCTCCCACAAGATCTCTGGGGAAGCTGGAAGAGGGCCACT...  BARX2::FOXF2+Colon+1+A_B  \n",
       "1  GTGGCTCTGAGGGAATTTCAAGTAACCGTGGACTAAAGTCTTAAGA...  BARX2::FOXF2+Colon+2+A_B  \n",
       "2  GCCACTGCACTCCAGCCTGGTGACAGAGCGAGACTCTGTCTCAAGA...  BARX2::FOXF2+Colon+3+A_B  \n",
       "3  AGTATTATGTGCTCATGGTCCCTACCCTCCAGGGTTCACAGTGTGG...  BARX2::FOXF2+Colon+4+A_B  \n",
       "4  ATGTGGCTGCGGGGCTTTAGGAGCTGACTCTTTGGTCACGGGGAGA...  BARX2::FOXF2+Colon+5+A_B  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 443,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_seqs_sel[:5]#[['vocab_pair','chr','start_pad','stop_pad']].vocab_pair.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 444,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['../data/processed/fig4_modelling/vocab_sum_final/motif_scan2_off_target_counts_df_all.csv',\n",
       " '../data/processed/fig4_modelling/vocab_sum_final/motif_scan2_motif_scan_on_target_cases_df_all.csv',\n",
       " '../data/processed/fig4_modelling/vocab_sum_final/motif_scan2_motif_scan_on_target_config_cases_df.csv']"
      ]
     },
     "execution_count": 444,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# motif scaning results \n",
    "glob.glob(os.path.join(input_dir, 'motif_scan2_*.csv'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 445,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1248, 5)\n",
      "(1137, 3)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  after removing the cwd from sys.path.\n"
     ]
    }
   ],
   "source": [
    "off_target_res = pd.read_csv(os.path.join(input_dir, 'motif_scan2_off_target_counts_df_all.csv'),index_col=0)\n",
    "off_target_res_case_2A = off_target_res[off_target_res.case=='case 2A']\n",
    "print(off_target_res_case_2A.shape)\n",
    "off_target_res_case_2A['off_target_pos_str'] = off_target_res_case_2A.pos_close.map(int).map(str)\n",
    "off_target_res_case_2A = off_target_res_case_2A.groupby(['instance','motif']).agg({'off_target_pos_str':'|'.join}).reset_index()\n",
    "off_target_res_case_2A.columns = ['off_target_name','off_target_motif','off_target_pos_str']\n",
    "print(off_target_res_case_2A.shape) # groupbed by instance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 446,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0        BARX2::FOXF2+Colon+1\n",
       "1        BARX2::FOXF2+Colon+2\n",
       "2        BARX2::FOXF2+Colon+3\n",
       "3        BARX2::FOXF2+Colon+4\n",
       "4        BARX2::FOXF2+Colon+5\n",
       "                 ...         \n",
       "16969        STAT1::UBP1+MM+6\n",
       "16970        STAT1::UBP1+MM+7\n",
       "16971        STAT1::UBP1+MM+8\n",
       "16972        STAT1::UBP1+MM+9\n",
       "16973       STAT1::UBP1+MM+10\n",
       "Name: name, Length: 9560, dtype: object"
      ]
     },
     "execution_count": 446,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_seqs_sel.name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 447,
   "metadata": {},
   "outputs": [],
   "source": [
    "vocab_instances20_seqs_sel_offtarget = vocab_instances20_seqs_sel.merge(off_target_res_case_2A, how='left',left_on='name',right_on='off_target_name')\n",
    "vocab_instances20_seqs_sel_offtarget.off_target_name.fillna('',inplace=True)\n",
    "vocab_instances20_seqs_sel_offtarget.off_target_motif.fillna('',inplace=True)\n",
    "vocab_instances20_seqs_sel_offtarget.off_target_pos_str.fillna('',inplace=True)\n",
    "vocab_instances20_seqs_sel_offtarget = pd.concat([vocab_instances20_seqs_sel_offtarget,vocab_instances20_seqs_sel_offtarget.apply(get_vocab_loc,axis=1)],axis=1)\n",
    "vocab_instances20_seqs_sel_offtarget = vocab_instances20_seqs_sel_offtarget.groupby(['vocab_pair', 'tot_dist', 'btn_dist', 'chr', 'start', 'stop',\n",
    "       'start_pad', 'stop_pad', 'vocab1', 'vocab1_start', 'vocab1_stop',\n",
    "       'vocab2', 'vocab2_start', 'vocab2_stop', 'tissue', 'loop_type', 'name', 'check_rs', 'scr_pass_rs', \n",
    "       'vocab_tissue_count', 'seq_config', 'seq','name_config', 'off_target_name','vocab1_start_shift', 'vocab1_stop_shift',\n",
    "       'vocab2_start_shift', 'vocab2_stop_shift']).agg({'off_target_motif':'::'.join,'off_target_pos_str':'::'.join}).reset_index()\n",
    "\n",
    "vocab_instances20_seqs_sel_offtarget = vocab_instances20_seqs_sel_offtarget.groupby(['vocab_pair', 'tot_dist', 'btn_dist', 'chr', 'start', 'stop',\n",
    "       'start_pad', 'stop_pad', 'vocab1', 'vocab1_start', 'vocab1_stop',\n",
    "       'vocab2', 'vocab2_start', 'vocab2_stop', 'tissue', 'loop_type', 'name',\n",
    "       'check_rs', 'scr_pass_rs', 'vocab_tissue_count',  'vocab1_start_shift',\n",
    "       'vocab1_stop_shift', 'vocab2_start_shift', 'vocab2_stop_shift','off_target_name',\n",
    "       'off_target_motif', 'off_target_pos_str']).agg({'seq_config':list,'seq':list, 'name_config':list}).reset_index()\n",
    "# vocab_instances20_seqs_sel_offtarget['vocab1_loc_start'] = vocab_instances20_seqs_sel_offtarget.vocab1_start - vocab_instances20_seqs_sel_offtarget.start_pad\n",
    "# vocab_instances20_seqs_sel_offtarget['vocab1_loc_stop'] = vocab_instances20_seqs_sel_offtarget.vocab1_stop - vocab_instances20_seqs_sel_offtarget.start_pad\n",
    "# vocab_instances20_seqs_sel_offtarget['vocab2_loc_start'] = vocab_instances20_seqs_sel_offtarget.vocab2_start - vocab_instances20_seqs_sel_offtarget.start_pad\n",
    "# vocab_instances20_seqs_sel_offtarget['vocab2_loc_stop'] = vocab_instances20_seqs_sel_offtarget.vocab2_stop - vocab_instances20_seqs_sel_offtarget.start_pad\n",
    "# vocab_instances20_seqs_sel_offtarget = vocab_instances20_seqs_sel_offtarget[['']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 448,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['vocab_pair', 'tot_dist', 'btn_dist', 'chr', 'start', 'stop',\n",
       "       'start_pad', 'stop_pad', 'vocab1', 'vocab1_start', 'vocab1_stop',\n",
       "       'vocab2', 'vocab2_start', 'vocab2_stop', 'tissue', 'loop_type', 'name',\n",
       "       'check_rs', 'scr_pass_rs', 'vocab_tissue_count', 'vocab1_start_shift',\n",
       "       'vocab1_stop_shift', 'vocab2_start_shift', 'vocab2_stop_shift',\n",
       "       'off_target_name', 'off_target_motif', 'off_target_pos_str',\n",
       "       'seq_config', 'seq', 'name_config'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 448,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_seqs_sel_offtarget.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 449,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['A_B_seq|A_scrB_seq|scrA_B_seq|scrA_scrB_seq'], dtype=object)"
      ]
     },
     "execution_count": 449,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_seqs_sel_offtarget.seq_config.apply('|'.join).unique()\n",
    "# this is the order for all configurations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 450,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([4])"
      ]
     },
     "execution_count": 450,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_seqs_sel_offtarget.seq.apply(len).unique()\n",
    "### all instances have 4 configurations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 451,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[135]\n",
      "[135]\n",
      "[135]\n",
      "[135]\n"
     ]
    }
   ],
   "source": [
    "print(vocab_instances20_seqs_sel_offtarget.seq.apply(lambda x: len(x[0])).unique())\n",
    "print(vocab_instances20_seqs_sel_offtarget.seq.apply(lambda x: len(x[1])).unique())\n",
    "print(vocab_instances20_seqs_sel_offtarget.seq.apply(lambda x: len(x[2])).unique())\n",
    "print(vocab_instances20_seqs_sel_offtarget.seq.apply(lambda x: len(x[3])).unique())\n",
    "### all sequences are 135 bp long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 452,
   "metadata": {},
   "outputs": [],
   "source": [
    "# row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 453,
   "metadata": {},
   "outputs": [],
   "source": [
    "# row['seq']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 454,
   "metadata": {},
   "outputs": [],
   "source": [
    "# scramble_off_target(row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 273,
   "metadata": {},
   "outputs": [],
   "source": [
    "# row['new_seqs'] = scramble_off_target(row)\n",
    "# # dict(zip(row['name_config'],row['new_seqs']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. scramble homotype off-target sequences (vocab words)\n",
    "\n",
    "test a max of `MAX_ITER` times, \n",
    "\n",
    "options:\n",
    "- scramble only non motif areas\n",
    "- scramble the 10 bp window around the motif"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2A get the scrambles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 457,
   "metadata": {},
   "outputs": [],
   "source": [
    "MAX_ITER = 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 458,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>vocab_pair</th>\n",
       "      <th>tot_dist</th>\n",
       "      <th>btn_dist</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>start_pad</th>\n",
       "      <th>stop_pad</th>\n",
       "      <th>vocab1</th>\n",
       "      <th>vocab1_start</th>\n",
       "      <th>...</th>\n",
       "      <th>vocab1_start_shift</th>\n",
       "      <th>vocab1_stop_shift</th>\n",
       "      <th>vocab2_start_shift</th>\n",
       "      <th>vocab2_stop_shift</th>\n",
       "      <th>off_target_name</th>\n",
       "      <th>off_target_motif</th>\n",
       "      <th>off_target_pos_str</th>\n",
       "      <th>seq_config</th>\n",
       "      <th>seq</th>\n",
       "      <th>name_config</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>ARID5B::KLF5</td>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "      <td>chr14</td>\n",
       "      <td>69253439</td>\n",
       "      <td>69253456</td>\n",
       "      <td>69253380</td>\n",
       "      <td>69253514</td>\n",
       "      <td>KLF5</td>\n",
       "      <td>69253439</td>\n",
       "      <td>...</td>\n",
       "      <td>54</td>\n",
       "      <td>71</td>\n",
       "      <td>71</td>\n",
       "      <td>85</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+8</td>\n",
       "      <td>ARID5B::KLF5</td>\n",
       "      <td>48::16</td>\n",
       "      <td>[A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]</td>\n",
       "      <td>[ACTCCCAAAACTTCCTGGGTGTAAAGGAACCAAAGGCAAGTGAGC...</td>\n",
       "      <td>[ARID5B::KLF5+GDSD6+8+A_B, ARID5B::KLF5+GDSD6+...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>ARID5B::KLF5</td>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "      <td>chr6</td>\n",
       "      <td>13477440</td>\n",
       "      <td>13477457</td>\n",
       "      <td>13477381</td>\n",
       "      <td>13477515</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>13477440</td>\n",
       "      <td>...</td>\n",
       "      <td>51</td>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "      <td>81</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+5</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>28</td>\n",
       "      <td>[A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]</td>\n",
       "      <td>[CAGGTGGCCTTGATCCAGGCAAAATCTGTATTAGTCAGGGTTCAC...</td>\n",
       "      <td>[ARID5B::KLF5+GDSD6+5+A_B, ARID5B::KLF5+GDSD6+...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>ARID5B::KLF5</td>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "      <td>chr6</td>\n",
       "      <td>13477440</td>\n",
       "      <td>13477457</td>\n",
       "      <td>13477381</td>\n",
       "      <td>13477515</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>13477440</td>\n",
       "      <td>...</td>\n",
       "      <td>51</td>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "      <td>81</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+6</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>28</td>\n",
       "      <td>[A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]</td>\n",
       "      <td>[CAGGTGGCCTTGATCCAGGCAAAATCTGTATTAGTCAGGGTTCAC...</td>\n",
       "      <td>[ARID5B::KLF5+GDSD6+6+A_B, ARID5B::KLF5+GDSD6+...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>ARID5B::KLF5</td>\n",
       "      <td>18</td>\n",
       "      <td>3</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7145778</td>\n",
       "      <td>7145796</td>\n",
       "      <td>7145720</td>\n",
       "      <td>7145854</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>7145778</td>\n",
       "      <td>...</td>\n",
       "      <td>50</td>\n",
       "      <td>63</td>\n",
       "      <td>63</td>\n",
       "      <td>81</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>[A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]</td>\n",
       "      <td>[GACGGCGGCGACGCGCGGGCGGATTCAGCGGAGCGATCCACGAAT...</td>\n",
       "      <td>[ARID5B::KLF5+GDSD6+10+A_B, ARID5B::KLF5+GDSD6...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>ARID5B::KLF5</td>\n",
       "      <td>18</td>\n",
       "      <td>3</td>\n",
       "      <td>chr19</td>\n",
       "      <td>14616989</td>\n",
       "      <td>14617007</td>\n",
       "      <td>14616931</td>\n",
       "      <td>14617065</td>\n",
       "      <td>KLF5</td>\n",
       "      <td>14616989</td>\n",
       "      <td>...</td>\n",
       "      <td>53</td>\n",
       "      <td>70</td>\n",
       "      <td>70</td>\n",
       "      <td>85</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+7</td>\n",
       "      <td>KLF5</td>\n",
       "      <td>26</td>\n",
       "      <td>[A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]</td>\n",
       "      <td>[TCTTCCCCTCCACGCTGCGGGCTATGGGAAGTGGCTCGGTCAGTT...</td>\n",
       "      <td>[ARID5B::KLF5+GDSD6+7+A_B, ARID5B::KLF5+GDSD6+...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     vocab_pair  tot_dist  btn_dist    chr     start      stop  start_pad  \\\n",
       "0  ARID5B::KLF5        17         2  chr14  69253439  69253456   69253380   \n",
       "1  ARID5B::KLF5        17         2   chr6  13477440  13477457   13477381   \n",
       "2  ARID5B::KLF5        17         2   chr6  13477440  13477457   13477381   \n",
       "3  ARID5B::KLF5        18         3  chr17   7145778   7145796    7145720   \n",
       "4  ARID5B::KLF5        18         3  chr19  14616989  14617007   14616931   \n",
       "\n",
       "   stop_pad  vocab1  vocab1_start  ...  vocab1_start_shift vocab1_stop_shift  \\\n",
       "0  69253514    KLF5      69253439  ...                  54                71   \n",
       "1  13477515  ARID5B      13477440  ...                  51                64   \n",
       "2  13477515  ARID5B      13477440  ...                  51                64   \n",
       "3   7145854  ARID5B       7145778  ...                  50                63   \n",
       "4  14617065    KLF5      14616989  ...                  53                70   \n",
       "\n",
       "   vocab2_start_shift  vocab2_stop_shift       off_target_name  \\\n",
       "0                  71                 85  ARID5B::KLF5+GDSD6+8   \n",
       "1                  64                 81  ARID5B::KLF5+GDSD6+5   \n",
       "2                  64                 81  ARID5B::KLF5+GDSD6+6   \n",
       "3                  63                 81                         \n",
       "4                  70                 85  ARID5B::KLF5+GDSD6+7   \n",
       "\n",
       "  off_target_motif off_target_pos_str  \\\n",
       "0     ARID5B::KLF5             48::16   \n",
       "1           ARID5B                 28   \n",
       "2           ARID5B                 28   \n",
       "3                                       \n",
       "4             KLF5                 26   \n",
       "\n",
       "                                         seq_config  \\\n",
       "0  [A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]   \n",
       "1  [A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]   \n",
       "2  [A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]   \n",
       "3  [A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]   \n",
       "4  [A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]   \n",
       "\n",
       "                                                 seq  \\\n",
       "0  [ACTCCCAAAACTTCCTGGGTGTAAAGGAACCAAAGGCAAGTGAGC...   \n",
       "1  [CAGGTGGCCTTGATCCAGGCAAAATCTGTATTAGTCAGGGTTCAC...   \n",
       "2  [CAGGTGGCCTTGATCCAGGCAAAATCTGTATTAGTCAGGGTTCAC...   \n",
       "3  [GACGGCGGCGACGCGCGGGCGGATTCAGCGGAGCGATCCACGAAT...   \n",
       "4  [TCTTCCCCTCCACGCTGCGGGCTATGGGAAGTGGCTCGGTCAGTT...   \n",
       "\n",
       "                                         name_config  \n",
       "0  [ARID5B::KLF5+GDSD6+8+A_B, ARID5B::KLF5+GDSD6+...  \n",
       "1  [ARID5B::KLF5+GDSD6+5+A_B, ARID5B::KLF5+GDSD6+...  \n",
       "2  [ARID5B::KLF5+GDSD6+6+A_B, ARID5B::KLF5+GDSD6+...  \n",
       "3  [ARID5B::KLF5+GDSD6+10+A_B, ARID5B::KLF5+GDSD6...  \n",
       "4  [ARID5B::KLF5+GDSD6+7+A_B, ARID5B::KLF5+GDSD6+...  \n",
       "\n",
       "[5 rows x 30 columns]"
      ]
     },
     "execution_count": 458,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_seqs_sel_offtarget[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 459,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['vocab_pair', 'tot_dist', 'btn_dist', 'chr', 'start', 'stop',\n",
       "       'start_pad', 'stop_pad', 'vocab1', 'vocab1_start', 'vocab1_stop',\n",
       "       'vocab2', 'vocab2_start', 'vocab2_stop', 'tissue', 'loop_type', 'name',\n",
       "       'check_rs', 'scr_pass_rs', 'vocab_tissue_count', 'vocab1_start_shift',\n",
       "       'vocab1_stop_shift', 'vocab2_start_shift', 'vocab2_stop_shift',\n",
       "       'off_target_name', 'off_target_motif', 'off_target_pos_str',\n",
       "       'seq_config', 'seq', 'name_config'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 459,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_seqs_sel_offtarget.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 474,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "vocab_pair                                                 ARID5B::KLF5\n",
      "tot_dist                                                             17\n",
      "btn_dist                                                              2\n",
      "chr                                                               chr14\n",
      "start                                                          69253439\n",
      "stop                                                           69253456\n",
      "start_pad                                                      69253380\n",
      "stop_pad                                                       69253514\n",
      "vocab1                                                             KLF5\n",
      "vocab1_start                                                   69253439\n",
      "vocab1_stop                                                    69253450\n",
      "vocab2                                                           ARID5B\n",
      "vocab2_start                                                   69253452\n",
      "vocab2_stop                                                    69253456\n",
      "tissue                                                            GDSD6\n",
      "loop_type                                                     loop_loop\n",
      "name                                               ARID5B::KLF5+GDSD6+8\n",
      "check_rs                                                              0\n",
      "scr_pass_rs                                                        True\n",
      "vocab_tissue_count                                                    8\n",
      "vocab1_start_shift                                                   54\n",
      "vocab1_stop_shift                                                    71\n",
      "vocab2_start_shift                                                   71\n",
      "vocab2_stop_shift                                                    85\n",
      "off_target_name                                    ARID5B::KLF5+GDSD6+8\n",
      "off_target_motif                                           ARID5B::KLF5\n",
      "off_target_pos_str                                               48::16\n",
      "seq_config             [A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]\n",
      "seq                   [ACTCCCAAAACTTCCTGGGTGTAAAGGAACCAAAGGCAAGTGAGC...\n",
      "name_config           [ARID5B::KLF5+GDSD6+8+A_B, ARID5B::KLF5+GDSD6+...\n",
      "Name: 0, dtype: object\n",
      "vocab_word_to_loc {'KLF5': [54, 71], 'ARID5B': [71, 85]}\n",
      "mask [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0 0 0\n",
      " 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2\n",
      " 2 2 2 2 2 2 2 2 2 2 2 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0\n",
      " 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]\n",
      "scr_ranges [(16, 32), (48, 53)]\n",
      "***\n",
      "ACTCCCAAAACTTCCTGGGTGTAAAGGAACCAAAGGCAAGTGAGCTGCAAATAGCAATCTCTGCCTCTCCCCAAATAAATCCTCCCAAGGCTCACACTCAAGATTTGTGAAAGTTCCAACTCCTCCGGCGGGTGA\n",
      "ACTCCCAAAACTTCCTGGATACAACAGGTGAGAAGGCAAGTGAGCTGCAAAATGCAATCTCTGCCTCTCCCCAAATAAATCCTCCCAAGGCTCACACTCAAGATTTGTGAAAGTTCCAACTCCTCCGGCGGGTGA\n",
      "***\n",
      "ACTCCCAAAACTTCCTGGGTGTAAAGGAACCAAAGGCAAGTGAGCTGCAAATAGCAATCTCTGCCTCTCCCCATATTACCAAACCCAAGGCTCACACTCAAGATTTGTGAAAGTTCCAACTCCTCCGGCGGGTGA\n",
      "ACTCCCAAAACTTCCTGGATACAACAGGTGAGAAGGCAAGTGAGCTGCAAAATGCAATCTCTGCCTCTCCCCATATTACCAAACCCAAGGCTCACACTCAAGATTTGTGAAAGTTCCAACTCCTCCGGCGGGTGA\n",
      "***\n",
      "ACTCCCAAAACTTCCTGGGTGTAAAGGAACCAAAGGCAAGTGAGCTGCAAATAGTCTCTGATCCACCCCTCCAAATAAATCCTCCCAAGGCTCACACTCAAGATTTGTGAAAGTTCCAACTCCTCCGGCGGGTGA\n",
      "ACTCCCAAAACTTCCTGGATACAACAGGTGAGAAGGCAAGTGAGCTGCAAAATGTCTCTGATCCACCCCTCCAAATAAATCCTCCCAAGGCTCACACTCAAGATTTGTGAAAGTTCCAACTCCTCCGGCGGGTGA\n",
      "***\n",
      "ACTCCCAAAACTTCCTGGGTGTAAAGGAACCAAAGGCAAGTGAGCTGCAAATAGTCTCTGATCCACCCCTCCATATTACCAAACCCAAGGCTCACACTCAAGATTTGTGAAAGTTCCAACTCCTCCGGCGGGTGA\n",
      "ACTCCCAAAACTTCCTGGATACAACAGGTGAGAAGGCAAGTGAGCTGCAAAATGTCTCTGATCCACCCCTCCATATTACCAAACCCAAGGCTCACACTCAAGATTTGTGAAAGTTCCAACTCCTCCGGCGGGTGA\n"
     ]
    }
   ],
   "source": [
    "#testing scrambling \n",
    "row= vocab_instances20_seqs_sel_offtarget.iloc[0,:]\n",
    "print(row)\n",
    "##\n",
    "for x,y in zip(row['seq'],scramble_off_target(row,verbose=True)):\n",
    "    print('***')\n",
    "    print(x)\n",
    "    print(y)\n",
    "    ###\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 475,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2390, 30)\n",
      "0\n",
      "100\n",
      "200\n",
      "300\n",
      "400\n",
      "500\n",
      "600\n",
      "700\n",
      "800\n",
      "900\n",
      "1000\n",
      "1100\n",
      "1200\n",
      "1300\n",
      "1400\n",
      "1500\n",
      "1600\n",
      "1700\n",
      "1800\n",
      "1900\n",
      "2000\n",
      "2100\n",
      "2200\n",
      "2300\n"
     ]
    }
   ],
   "source": [
    "## iterate through the rows of the rs_filt vocab dataframe vocab_instances20_alltissue_df_rs_filt\n",
    "print(vocab_instances20_seqs_sel_offtarget.shape)\n",
    "vocab_instances20_seqs_sel_offtarget['scr_pass_rs'] = False\n",
    "vocab_instances20_seqs_sel_offtarget['A_B_seq'] = ''\n",
    "vocab_instances20_seqs_sel_offtarget['A_scrB_seq'] = ''\n",
    "vocab_instances20_seqs_sel_offtarget['scrA_B_seq'] = ''\n",
    "vocab_instances20_seqs_sel_offtarget['scrA_scrB_seq'] = ''\n",
    "\n",
    "for idx, row in vocab_instances20_seqs_sel_offtarget.iterrows():\n",
    "    ## DEBUG\n",
    "    if (idx%100)==0:\n",
    "        print(idx)\n",
    "    \n",
    "    # set default\n",
    "    scr_pass_rs = False\n",
    "    num_iter = 0\n",
    "    # iterate scrambles until find some that pass the check_rs filter\n",
    "    while (not scr_pass_rs) ^ (num_iter > MAX_ITER):\n",
    "        # array(['A_B_seq|A_scrB_seq|scrA_B_seq|scrA_scrB_seq'], dtype=object), is order of results\n",
    "        A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq = scramble_off_target(row)\n",
    "\n",
    "\n",
    "\n",
    "        # check_rs\n",
    "        A_B_seq_rs = check_rs(A_B_seq)\n",
    "        A_scrB_seq_rs = check_rs(A_scrB_seq)\n",
    "        scrA_B_seq_rs = check_rs(scrA_B_seq)\n",
    "        scrA_scrB_seq_rs = check_rs(scrA_scrB_seq)\n",
    "\n",
    "        if (A_B_seq_rs+A_scrB_seq_rs+scrA_B_seq_rs+scrA_scrB_seq_rs==0):\n",
    "            scr_pass_rs=True\n",
    "        num_iter+=1\n",
    "    \n",
    "    # add information to df\n",
    "    vocab_instances20_seqs_sel_offtarget.at[idx, 'scr_pass_rs'] = scr_pass_rs\n",
    "    vocab_instances20_seqs_sel_offtarget.at[idx, 'A_B_seq'] = A_B_seq\n",
    "    vocab_instances20_seqs_sel_offtarget.at[idx, 'A_scrB_seq'] = A_scrB_seq\n",
    "    vocab_instances20_seqs_sel_offtarget.at[idx, 'scrA_B_seq'] = scrA_B_seq\n",
    "    vocab_instances20_seqs_sel_offtarget.at[idx, 'scrA_scrB_seq'] = scrA_scrB_seq\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2B some gut checks and saving"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 476,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True    2390\n",
       "Name: scr_pass_rs, dtype: int64"
      ]
     },
     "execution_count": 476,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# manual step\n",
    "vocab_instances20_seqs_sel_offtarget.scr_pass_rs.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "all sequences and scr configurations pass the rs filter\n",
    "\n",
    "if they don't keep running the scrambling until they all pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 477,
   "metadata": {},
   "outputs": [],
   "source": [
    "vocab_instances20_seqs_sel_offtarget.to_csv(os.path.join(save_dir, 'vocab_instances20_seqs_sel_offtarget_all_col.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 478,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['vocab_pair', 'tot_dist', 'btn_dist', 'chr', 'start', 'stop',\n",
      "       'start_pad', 'stop_pad', 'vocab1', 'vocab1_start', 'vocab1_stop',\n",
      "       'vocab2', 'vocab2_start', 'vocab2_stop', 'tissue', 'loop_type', 'name',\n",
      "       'check_rs', 'scr_pass_rs', 'vocab_tissue_count', 'vocab1_start_shift',\n",
      "       'vocab1_stop_shift', 'vocab2_start_shift', 'vocab2_stop_shift',\n",
      "       'off_target_name', 'off_target_motif', 'off_target_pos_str',\n",
      "       'seq_config', 'seq', 'name_config', 'A_B_seq', 'A_scrB_seq',\n",
      "       'scrA_B_seq', 'scrA_scrB_seq'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>vocab_pair</th>\n",
       "      <th>tot_dist</th>\n",
       "      <th>btn_dist</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>start_pad</th>\n",
       "      <th>stop_pad</th>\n",
       "      <th>vocab1</th>\n",
       "      <th>vocab1_start</th>\n",
       "      <th>...</th>\n",
       "      <th>off_target_name</th>\n",
       "      <th>off_target_motif</th>\n",
       "      <th>off_target_pos_str</th>\n",
       "      <th>seq_config</th>\n",
       "      <th>seq</th>\n",
       "      <th>name_config</th>\n",
       "      <th>A_B_seq</th>\n",
       "      <th>A_scrB_seq</th>\n",
       "      <th>scrA_B_seq</th>\n",
       "      <th>scrA_scrB_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>ARID5B::KLF5</td>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "      <td>chr14</td>\n",
       "      <td>69253439</td>\n",
       "      <td>69253456</td>\n",
       "      <td>69253380</td>\n",
       "      <td>69253514</td>\n",
       "      <td>KLF5</td>\n",
       "      <td>69253439</td>\n",
       "      <td>...</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+8</td>\n",
       "      <td>ARID5B::KLF5</td>\n",
       "      <td>48::16</td>\n",
       "      <td>[A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]</td>\n",
       "      <td>[ACTCCCAAAACTTCCTGGGTGTAAAGGAACCAAAGGCAAGTGAGC...</td>\n",
       "      <td>[ARID5B::KLF5+GDSD6+8+A_B, ARID5B::KLF5+GDSD6+...</td>\n",
       "      <td>ACTCCCAAAACTTCCTAACAGAAGAGCGGTTGAAGGCAAGTGAGCT...</td>\n",
       "      <td>ACTCCCAAAACTTCCTAACAGAAGAGCGGTTGAAGGCAAGTGAGCT...</td>\n",
       "      <td>ACTCCCAAAACTTCCTAACAGAAGAGCGGTTGAAGGCAAGTGAGCT...</td>\n",
       "      <td>ACTCCCAAAACTTCCTAACAGAAGAGCGGTTGAAGGCAAGTGAGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>ARID5B::KLF5</td>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "      <td>chr6</td>\n",
       "      <td>13477440</td>\n",
       "      <td>13477457</td>\n",
       "      <td>13477381</td>\n",
       "      <td>13477515</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>13477440</td>\n",
       "      <td>...</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+5</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>28</td>\n",
       "      <td>[A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]</td>\n",
       "      <td>[CAGGTGGCCTTGATCCAGGCAAAATCTGTATTAGTCAGGGTTCAC...</td>\n",
       "      <td>[ARID5B::KLF5+GDSD6+5+A_B, ARID5B::KLF5+GDSD6+...</td>\n",
       "      <td>CAGGTGGCCTTGATCCAGGCAAAATCTGGTTAAGATTCGGTTCACC...</td>\n",
       "      <td>CAGGTGGCCTTGATCCAGGCAAAATCTGGTTAAGATTCGGTTCACC...</td>\n",
       "      <td>CAGGTGGCCTTGATCCAGGCAAAATCTGGTTAAGATTCGGTTCACC...</td>\n",
       "      <td>CAGGTGGCCTTGATCCAGGCAAAATCTGGTTAAGATTCGGTTCACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>ARID5B::KLF5</td>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "      <td>chr6</td>\n",
       "      <td>13477440</td>\n",
       "      <td>13477457</td>\n",
       "      <td>13477381</td>\n",
       "      <td>13477515</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>13477440</td>\n",
       "      <td>...</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+6</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>28</td>\n",
       "      <td>[A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]</td>\n",
       "      <td>[CAGGTGGCCTTGATCCAGGCAAAATCTGTATTAGTCAGGGTTCAC...</td>\n",
       "      <td>[ARID5B::KLF5+GDSD6+6+A_B, ARID5B::KLF5+GDSD6+...</td>\n",
       "      <td>CAGGTGGCCTTGATCCAGGCAAAATCTGAGTGGTAGTTCATTCACC...</td>\n",
       "      <td>CAGGTGGCCTTGATCCAGGCAAAATCTGAGTGGTAGTTCATTCACC...</td>\n",
       "      <td>CAGGTGGCCTTGATCCAGGCAAAATCTGAGTGGTAGTTCATTCACC...</td>\n",
       "      <td>CAGGTGGCCTTGATCCAGGCAAAATCTGAGTGGTAGTTCATTCACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>ARID5B::KLF5</td>\n",
       "      <td>18</td>\n",
       "      <td>3</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7145778</td>\n",
       "      <td>7145796</td>\n",
       "      <td>7145720</td>\n",
       "      <td>7145854</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>7145778</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>[A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]</td>\n",
       "      <td>[GACGGCGGCGACGCGCGGGCGGATTCAGCGGAGCGATCCACGAAT...</td>\n",
       "      <td>[ARID5B::KLF5+GDSD6+10+A_B, ARID5B::KLF5+GDSD6...</td>\n",
       "      <td>GACGGCGGCGACGCGCGGGCGGATTCAGCGGAGCGATCCACGAATT...</td>\n",
       "      <td>GACGGCGGCGACGCGCGGGCGGATTCAGCGGAGCGATCCACGAATT...</td>\n",
       "      <td>GACGGCGGCGACGCGCGGGCGGATTCAGCGGAGCGATCCACGAATT...</td>\n",
       "      <td>GACGGCGGCGACGCGCGGGCGGATTCAGCGGAGCGATCCACGAATT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>ARID5B::KLF5</td>\n",
       "      <td>18</td>\n",
       "      <td>3</td>\n",
       "      <td>chr19</td>\n",
       "      <td>14616989</td>\n",
       "      <td>14617007</td>\n",
       "      <td>14616931</td>\n",
       "      <td>14617065</td>\n",
       "      <td>KLF5</td>\n",
       "      <td>14616989</td>\n",
       "      <td>...</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+7</td>\n",
       "      <td>KLF5</td>\n",
       "      <td>26</td>\n",
       "      <td>[A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]</td>\n",
       "      <td>[TCTTCCCCTCCACGCTGCGGGCTATGGGAAGTGGCTCGGTCAGTT...</td>\n",
       "      <td>[ARID5B::KLF5+GDSD6+7+A_B, ARID5B::KLF5+GDSD6+...</td>\n",
       "      <td>TCTTCCCCTCCACGCTGCGGGCTATGGGTACGCAGGGTATGCGTTT...</td>\n",
       "      <td>TCTTCCCCTCCACGCTGCGGGCTATGGGTACGCAGGGTATGCGTTT...</td>\n",
       "      <td>TCTTCCCCTCCACGCTGCGGGCTATGGGTACGCAGGGTATGCGTTT...</td>\n",
       "      <td>TCTTCCCCTCCACGCTGCGGGCTATGGGTACGCAGGGTATGCGTTT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     vocab_pair  tot_dist  btn_dist    chr     start      stop  start_pad  \\\n",
       "0  ARID5B::KLF5        17         2  chr14  69253439  69253456   69253380   \n",
       "1  ARID5B::KLF5        17         2   chr6  13477440  13477457   13477381   \n",
       "2  ARID5B::KLF5        17         2   chr6  13477440  13477457   13477381   \n",
       "3  ARID5B::KLF5        18         3  chr17   7145778   7145796    7145720   \n",
       "4  ARID5B::KLF5        18         3  chr19  14616989  14617007   14616931   \n",
       "\n",
       "   stop_pad  vocab1  vocab1_start  ...       off_target_name off_target_motif  \\\n",
       "0  69253514    KLF5      69253439  ...  ARID5B::KLF5+GDSD6+8     ARID5B::KLF5   \n",
       "1  13477515  ARID5B      13477440  ...  ARID5B::KLF5+GDSD6+5           ARID5B   \n",
       "2  13477515  ARID5B      13477440  ...  ARID5B::KLF5+GDSD6+6           ARID5B   \n",
       "3   7145854  ARID5B       7145778  ...                                          \n",
       "4  14617065    KLF5      14616989  ...  ARID5B::KLF5+GDSD6+7             KLF5   \n",
       "\n",
       "   off_target_pos_str                                        seq_config  \\\n",
       "0              48::16  [A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]   \n",
       "1                  28  [A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]   \n",
       "2                  28  [A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]   \n",
       "3                      [A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]   \n",
       "4                  26  [A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]   \n",
       "\n",
       "                                                 seq  \\\n",
       "0  [ACTCCCAAAACTTCCTGGGTGTAAAGGAACCAAAGGCAAGTGAGC...   \n",
       "1  [CAGGTGGCCTTGATCCAGGCAAAATCTGTATTAGTCAGGGTTCAC...   \n",
       "2  [CAGGTGGCCTTGATCCAGGCAAAATCTGTATTAGTCAGGGTTCAC...   \n",
       "3  [GACGGCGGCGACGCGCGGGCGGATTCAGCGGAGCGATCCACGAAT...   \n",
       "4  [TCTTCCCCTCCACGCTGCGGGCTATGGGAAGTGGCTCGGTCAGTT...   \n",
       "\n",
       "                                         name_config  \\\n",
       "0  [ARID5B::KLF5+GDSD6+8+A_B, ARID5B::KLF5+GDSD6+...   \n",
       "1  [ARID5B::KLF5+GDSD6+5+A_B, ARID5B::KLF5+GDSD6+...   \n",
       "2  [ARID5B::KLF5+GDSD6+6+A_B, ARID5B::KLF5+GDSD6+...   \n",
       "3  [ARID5B::KLF5+GDSD6+10+A_B, ARID5B::KLF5+GDSD6...   \n",
       "4  [ARID5B::KLF5+GDSD6+7+A_B, ARID5B::KLF5+GDSD6+...   \n",
       "\n",
       "                                             A_B_seq  \\\n",
       "0  ACTCCCAAAACTTCCTAACAGAAGAGCGGTTGAAGGCAAGTGAGCT...   \n",
       "1  CAGGTGGCCTTGATCCAGGCAAAATCTGGTTAAGATTCGGTTCACC...   \n",
       "2  CAGGTGGCCTTGATCCAGGCAAAATCTGAGTGGTAGTTCATTCACC...   \n",
       "3  GACGGCGGCGACGCGCGGGCGGATTCAGCGGAGCGATCCACGAATT...   \n",
       "4  TCTTCCCCTCCACGCTGCGGGCTATGGGTACGCAGGGTATGCGTTT...   \n",
       "\n",
       "                                          A_scrB_seq  \\\n",
       "0  ACTCCCAAAACTTCCTAACAGAAGAGCGGTTGAAGGCAAGTGAGCT...   \n",
       "1  CAGGTGGCCTTGATCCAGGCAAAATCTGGTTAAGATTCGGTTCACC...   \n",
       "2  CAGGTGGCCTTGATCCAGGCAAAATCTGAGTGGTAGTTCATTCACC...   \n",
       "3  GACGGCGGCGACGCGCGGGCGGATTCAGCGGAGCGATCCACGAATT...   \n",
       "4  TCTTCCCCTCCACGCTGCGGGCTATGGGTACGCAGGGTATGCGTTT...   \n",
       "\n",
       "                                          scrA_B_seq  \\\n",
       "0  ACTCCCAAAACTTCCTAACAGAAGAGCGGTTGAAGGCAAGTGAGCT...   \n",
       "1  CAGGTGGCCTTGATCCAGGCAAAATCTGGTTAAGATTCGGTTCACC...   \n",
       "2  CAGGTGGCCTTGATCCAGGCAAAATCTGAGTGGTAGTTCATTCACC...   \n",
       "3  GACGGCGGCGACGCGCGGGCGGATTCAGCGGAGCGATCCACGAATT...   \n",
       "4  TCTTCCCCTCCACGCTGCGGGCTATGGGTACGCAGGGTATGCGTTT...   \n",
       "\n",
       "                                       scrA_scrB_seq  \n",
       "0  ACTCCCAAAACTTCCTAACAGAAGAGCGGTTGAAGGCAAGTGAGCT...  \n",
       "1  CAGGTGGCCTTGATCCAGGCAAAATCTGGTTAAGATTCGGTTCACC...  \n",
       "2  CAGGTGGCCTTGATCCAGGCAAAATCTGAGTGGTAGTTCATTCACC...  \n",
       "3  GACGGCGGCGACGCGCGGGCGGATTCAGCGGAGCGATCCACGAATT...  \n",
       "4  TCTTCCCCTCCACGCTGCGGGCTATGGGTACGCAGGGTATGCGTTT...  \n",
       "\n",
       "[5 rows x 34 columns]"
      ]
     },
     "execution_count": 478,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(vocab_instances20_seqs_sel_offtarget.columns)\n",
    "vocab_instances20_seqs_sel_offtarget[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 479,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[135]\n",
      "[135]\n",
      "[135]\n",
      "[135]\n"
     ]
    }
   ],
   "source": [
    "print(vocab_instances20_seqs_sel_offtarget.A_B_seq.apply(len).unique())\n",
    "print(vocab_instances20_seqs_sel_offtarget.A_scrB_seq.apply(len).unique())\n",
    "print(vocab_instances20_seqs_sel_offtarget.scrA_B_seq.apply(len).unique())\n",
    "print(vocab_instances20_seqs_sel_offtarget.scrA_scrB_seq.apply(len).unique())\n",
    "\n",
    "# all sequences are 135 bp in length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 480,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "vocab_pair                                                 ARID5B::KLF5\n",
      "tot_dist                                                             17\n",
      "btn_dist                                                              2\n",
      "chr                                                               chr14\n",
      "start                                                          69253439\n",
      "stop                                                           69253456\n",
      "start_pad                                                      69253380\n",
      "stop_pad                                                       69253514\n",
      "vocab1                                                             KLF5\n",
      "vocab1_start                                                   69253439\n",
      "vocab1_stop                                                    69253450\n",
      "vocab2                                                           ARID5B\n",
      "vocab2_start                                                   69253452\n",
      "vocab2_stop                                                    69253456\n",
      "tissue                                                            GDSD6\n",
      "loop_type                                                     loop_loop\n",
      "name                                               ARID5B::KLF5+GDSD6+8\n",
      "check_rs                                                              0\n",
      "scr_pass_rs                                                        True\n",
      "vocab_tissue_count                                                    8\n",
      "vocab1_start_shift                                                   54\n",
      "vocab1_stop_shift                                                    71\n",
      "vocab2_start_shift                                                   71\n",
      "vocab2_stop_shift                                                    85\n",
      "off_target_name                                    ARID5B::KLF5+GDSD6+8\n",
      "off_target_motif                                           ARID5B::KLF5\n",
      "off_target_pos_str                                               48::16\n",
      "seq_config             [A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]\n",
      "seq                   [ACTCCCAAAACTTCCTGGGTGTAAAGGAACCAAAGGCAAGTGAGC...\n",
      "name_config           [ARID5B::KLF5+GDSD6+8+A_B, ARID5B::KLF5+GDSD6+...\n",
      "A_B_seq               ACTCCCAAAACTTCCTAACAGAAGAGCGGTTGAAGGCAAGTGAGCT...\n",
      "A_scrB_seq            ACTCCCAAAACTTCCTAACAGAAGAGCGGTTGAAGGCAAGTGAGCT...\n",
      "scrA_B_seq            ACTCCCAAAACTTCCTAACAGAAGAGCGGTTGAAGGCAAGTGAGCT...\n",
      "scrA_scrB_seq         ACTCCCAAAACTTCCTAACAGAAGAGCGGTTGAAGGCAAGTGAGCT...\n",
      "Name: 0, dtype: object\n",
      "ACTCCCAAAACTTCCTGGGTGTAAAGGAACCAAAGGCAAGTGAGCTGCAAATAGCAATCTCTGCCTCTCCCCAAATAAATCCTCCCAAGGCTCACACTCAAGATTTGTGAAAGTTCCAACTCCTCCGGCGGGTGA\n",
      "ACTCCCAAAACTTCCTAACAGAAGAGCGGTTGAAGGCAAGTGAGCTGCAAATAGCAATCTCTGCCTCTCCCCAAATAAATCCTCCCAAGGCTCACACTCAAGATTTGTGAAAGTTCCAACTCCTCCGGCGGGTGA\n",
      "ACTCCCAAAACTTCCTAACAGAAGAGCGGTTGAAGGCAAGTGAGCTGCAAATAGCAATCTCTGCCTCTCCCCATATTACCAAACCCAAGGCTCACACTCAAGATTTGTGAAAGTTCCAACTCCTCCGGCGGGTGA\n",
      "ACTCCCAAAACTTCCTAACAGAAGAGCGGTTGAAGGCAAGTGAGCTGCAAATAGTCTCTGATCCACCCCTCCAAATAAATCCTCCCAAGGCTCACACTCAAGATTTGTGAAAGTTCCAACTCCTCCGGCGGGTGA\n",
      "ACTCCCAAAACTTCCTAACAGAAGAGCGGTTGAAGGCAAGTGAGCTGCAAATAGTCTCTGATCCACCCCTCCATATTACCAAACCCAAGGCTCACACTCAAGATTTGTGAAAGTTCCAACTCCTCCGGCGGGTGA\n"
     ]
    }
   ],
   "source": [
    "row = vocab_instances20_seqs_sel_offtarget.iloc[0,:]\n",
    "print(row)\n",
    "print(row['seq'][0])\n",
    "print(row['A_B_seq'])\n",
    "print(row['A_scrB_seq'])\n",
    "print(row['scrA_B_seq'])\n",
    "print(row['scrA_scrB_seq'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 530,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 532,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>vocab_pair</th>\n",
       "      <th>tot_dist</th>\n",
       "      <th>btn_dist</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>start_pad</th>\n",
       "      <th>stop_pad</th>\n",
       "      <th>vocab1</th>\n",
       "      <th>vocab1_start</th>\n",
       "      <th>...</th>\n",
       "      <th>off_target_name</th>\n",
       "      <th>off_target_motif</th>\n",
       "      <th>off_target_pos_str</th>\n",
       "      <th>seq_config</th>\n",
       "      <th>seq</th>\n",
       "      <th>name_config</th>\n",
       "      <th>A_B_seq</th>\n",
       "      <th>A_scrB_seq</th>\n",
       "      <th>scrA_B_seq</th>\n",
       "      <th>scrA_scrB_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>ARID5B::KLF5</td>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "      <td>chr14</td>\n",
       "      <td>69253439</td>\n",
       "      <td>69253456</td>\n",
       "      <td>69253380</td>\n",
       "      <td>69253514</td>\n",
       "      <td>KLF5</td>\n",
       "      <td>69253439</td>\n",
       "      <td>...</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+8</td>\n",
       "      <td>ARID5B::KLF5</td>\n",
       "      <td>48::16</td>\n",
       "      <td>[A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]</td>\n",
       "      <td>[ACTCCCAAAACTTCCTGGGTGTAAAGGAACCAAAGGCAAGTGAGC...</td>\n",
       "      <td>[ARID5B::KLF5+GDSD6+8+A_B, ARID5B::KLF5+GDSD6+...</td>\n",
       "      <td>ACTCCCAAAACTTCCTAACAGAAGAGCGGTTGAAGGCAAGTGAGCT...</td>\n",
       "      <td>ACTCCCAAAACTTCCTAACAGAAGAGCGGTTGAAGGCAAGTGAGCT...</td>\n",
       "      <td>ACTCCCAAAACTTCCTAACAGAAGAGCGGTTGAAGGCAAGTGAGCT...</td>\n",
       "      <td>ACTCCCAAAACTTCCTAACAGAAGAGCGGTTGAAGGCAAGTGAGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>ARID5B::KLF5</td>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "      <td>chr6</td>\n",
       "      <td>13477440</td>\n",
       "      <td>13477457</td>\n",
       "      <td>13477381</td>\n",
       "      <td>13477515</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>13477440</td>\n",
       "      <td>...</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+5</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>28</td>\n",
       "      <td>[A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]</td>\n",
       "      <td>[CAGGTGGCCTTGATCCAGGCAAAATCTGTATTAGTCAGGGTTCAC...</td>\n",
       "      <td>[ARID5B::KLF5+GDSD6+5+A_B, ARID5B::KLF5+GDSD6+...</td>\n",
       "      <td>CAGGTGGCCTTGATCCAGGCAAAATCTGGTTAAGATTCGGTTCACC...</td>\n",
       "      <td>CAGGTGGCCTTGATCCAGGCAAAATCTGGTTAAGATTCGGTTCACC...</td>\n",
       "      <td>CAGGTGGCCTTGATCCAGGCAAAATCTGGTTAAGATTCGGTTCACC...</td>\n",
       "      <td>CAGGTGGCCTTGATCCAGGCAAAATCTGGTTAAGATTCGGTTCACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>ARID5B::KLF5</td>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "      <td>chr6</td>\n",
       "      <td>13477440</td>\n",
       "      <td>13477457</td>\n",
       "      <td>13477381</td>\n",
       "      <td>13477515</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>13477440</td>\n",
       "      <td>...</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+6</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>28</td>\n",
       "      <td>[A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]</td>\n",
       "      <td>[CAGGTGGCCTTGATCCAGGCAAAATCTGTATTAGTCAGGGTTCAC...</td>\n",
       "      <td>[ARID5B::KLF5+GDSD6+6+A_B, ARID5B::KLF5+GDSD6+...</td>\n",
       "      <td>CAGGTGGCCTTGATCCAGGCAAAATCTGAGTGGTAGTTCATTCACC...</td>\n",
       "      <td>CAGGTGGCCTTGATCCAGGCAAAATCTGAGTGGTAGTTCATTCACC...</td>\n",
       "      <td>CAGGTGGCCTTGATCCAGGCAAAATCTGAGTGGTAGTTCATTCACC...</td>\n",
       "      <td>CAGGTGGCCTTGATCCAGGCAAAATCTGAGTGGTAGTTCATTCACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>ARID5B::KLF5</td>\n",
       "      <td>18</td>\n",
       "      <td>3</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7145778</td>\n",
       "      <td>7145796</td>\n",
       "      <td>7145720</td>\n",
       "      <td>7145854</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>7145778</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>[A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]</td>\n",
       "      <td>[GACGGCGGCGACGCGCGGGCGGATTCAGCGGAGCGATCCACGAAT...</td>\n",
       "      <td>[ARID5B::KLF5+GDSD6+10+A_B, ARID5B::KLF5+GDSD6...</td>\n",
       "      <td>GACGGCGGCGACGCGCGGGCGGATTCAGCGGAGCGATCCACGAATT...</td>\n",
       "      <td>GACGGCGGCGACGCGCGGGCGGATTCAGCGGAGCGATCCACGAATT...</td>\n",
       "      <td>GACGGCGGCGACGCGCGGGCGGATTCAGCGGAGCGATCCACGAATT...</td>\n",
       "      <td>GACGGCGGCGACGCGCGGGCGGATTCAGCGGAGCGATCCACGAATT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>ARID5B::KLF5</td>\n",
       "      <td>18</td>\n",
       "      <td>3</td>\n",
       "      <td>chr19</td>\n",
       "      <td>14616989</td>\n",
       "      <td>14617007</td>\n",
       "      <td>14616931</td>\n",
       "      <td>14617065</td>\n",
       "      <td>KLF5</td>\n",
       "      <td>14616989</td>\n",
       "      <td>...</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+7</td>\n",
       "      <td>KLF5</td>\n",
       "      <td>26</td>\n",
       "      <td>[A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]</td>\n",
       "      <td>[TCTTCCCCTCCACGCTGCGGGCTATGGGAAGTGGCTCGGTCAGTT...</td>\n",
       "      <td>[ARID5B::KLF5+GDSD6+7+A_B, ARID5B::KLF5+GDSD6+...</td>\n",
       "      <td>TCTTCCCCTCCACGCTGCGGGCTATGGGTACGCAGGGTATGCGTTT...</td>\n",
       "      <td>TCTTCCCCTCCACGCTGCGGGCTATGGGTACGCAGGGTATGCGTTT...</td>\n",
       "      <td>TCTTCCCCTCCACGCTGCGGGCTATGGGTACGCAGGGTATGCGTTT...</td>\n",
       "      <td>TCTTCCCCTCCACGCTGCGGGCTATGGGTACGCAGGGTATGCGTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2385</td>\n",
       "      <td>ZFX::E2F6</td>\n",
       "      <td>26</td>\n",
       "      <td>4</td>\n",
       "      <td>chr20</td>\n",
       "      <td>2632854</td>\n",
       "      <td>2632880</td>\n",
       "      <td>2632800</td>\n",
       "      <td>2632934</td>\n",
       "      <td>ZFX</td>\n",
       "      <td>2632854</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>[A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]</td>\n",
       "      <td>[CCGAGTAGGGGACGGGAACGCCTGGGTCCCGGGGACGTGGGGCGC...</td>\n",
       "      <td>[ZFX::E2F6+SCC+8+A_B, ZFX::E2F6+SCC+8+A_scrB, ...</td>\n",
       "      <td>CCGAGTAGGGGACGGGAACGCCTGGGTCCCGGGGACGTGGGGCGCA...</td>\n",
       "      <td>CCGAGTAGGGGACGGGAACGCCTGGGTCCCGGGGACGTGGGGCGCA...</td>\n",
       "      <td>CCGAGTAGGGGACGGGAACGCCTGGGTCCCGGGGACGTGGGGCGCA...</td>\n",
       "      <td>CCGAGTAGGGGACGGGAACGCCTGGGTCCCGGGGACGTGGGGCGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2386</td>\n",
       "      <td>ZFX::E2F6</td>\n",
       "      <td>28</td>\n",
       "      <td>6</td>\n",
       "      <td>chr2</td>\n",
       "      <td>85811652</td>\n",
       "      <td>85811680</td>\n",
       "      <td>85811599</td>\n",
       "      <td>85811733</td>\n",
       "      <td>ZFX</td>\n",
       "      <td>85811652</td>\n",
       "      <td>...</td>\n",
       "      <td>ZFX::E2F6+SCC+1</td>\n",
       "      <td>ZFX</td>\n",
       "      <td>104|104</td>\n",
       "      <td>[A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]</td>\n",
       "      <td>[GCGGCGGCAGCAGCGATGGTGAGGGCCCAGGCGGGGCCGGCCAGC...</td>\n",
       "      <td>[ZFX::E2F6+SCC+1+A_B, ZFX::E2F6+SCC+1+A_scrB, ...</td>\n",
       "      <td>GCGGCGGCAGCAGCGATGGTGAGGGCCCAGGCGGGGCCGGCCAGCC...</td>\n",
       "      <td>GCGGCGGCAGCAGCGATGGTGAGGGCCCAGGCGGGGCCGGCCAGCC...</td>\n",
       "      <td>GCGGCGGCAGCAGCGATGGTGAGGGCCCAGGCGGGGCCGGCCAGCC...</td>\n",
       "      <td>GCGGCGGCAGCAGCGATGGTGAGGGCCCAGGCGGGGCCGGCCAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2387</td>\n",
       "      <td>ZFX::E2F6</td>\n",
       "      <td>28</td>\n",
       "      <td>6</td>\n",
       "      <td>chr2</td>\n",
       "      <td>220083840</td>\n",
       "      <td>220083868</td>\n",
       "      <td>220083787</td>\n",
       "      <td>220083921</td>\n",
       "      <td>E2F6</td>\n",
       "      <td>220083840</td>\n",
       "      <td>...</td>\n",
       "      <td>ZFX::E2F6+SCC+2</td>\n",
       "      <td>E2F6</td>\n",
       "      <td>110</td>\n",
       "      <td>[A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]</td>\n",
       "      <td>[GTGCAGGCAACGAATGCAAATTAGAGAAGCAGAGCCGGCTTGGTC...</td>\n",
       "      <td>[ZFX::E2F6+SCC+2+A_B, ZFX::E2F6+SCC+2+A_scrB, ...</td>\n",
       "      <td>GTGCAGGCAACGAATGCAAATTAGAGAAGCAGAGCCGGCTTGGTCA...</td>\n",
       "      <td>GTGCAGGCAACGAATGCAAATTAGAGAAGCAGAGCCGGCTTGGTCA...</td>\n",
       "      <td>GTGCAGGCAACGAATGCAAATTAGAGAAGCAGAGCCGGCTTGGTCA...</td>\n",
       "      <td>GTGCAGGCAACGAATGCAAATTAGAGAAGCAGAGCCGGCTTGGTCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2388</td>\n",
       "      <td>ZFX::E2F6</td>\n",
       "      <td>28</td>\n",
       "      <td>6</td>\n",
       "      <td>chr7</td>\n",
       "      <td>44153086</td>\n",
       "      <td>44153114</td>\n",
       "      <td>44153033</td>\n",
       "      <td>44153167</td>\n",
       "      <td>E2F6</td>\n",
       "      <td>44153086</td>\n",
       "      <td>...</td>\n",
       "      <td>ZFX::E2F6+SCC+9</td>\n",
       "      <td>ZFX</td>\n",
       "      <td>108</td>\n",
       "      <td>[A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]</td>\n",
       "      <td>[AGGTAGAGTCTGGGGAGCCTGGGGGCGAAATTCAGAGAGGGAGGG...</td>\n",
       "      <td>[ZFX::E2F6+SCC+9+A_B, ZFX::E2F6+SCC+9+A_scrB, ...</td>\n",
       "      <td>AGGTAGAGTCTGGGGAGCCTGGGGGCGAAATTCAGAGAGGGAGGGC...</td>\n",
       "      <td>AGGTAGAGTCTGGGGAGCCTGGGGGCGAAATTCAGAGAGGGAGGGC...</td>\n",
       "      <td>AGGTAGAGTCTGGGGAGCCTGGGGGCGAAATTCAGAGAGGGAGGGC...</td>\n",
       "      <td>AGGTAGAGTCTGGGGAGCCTGGGGGCGAAATTCAGAGAGGGAGGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2389</td>\n",
       "      <td>ZFX::E2F6</td>\n",
       "      <td>30</td>\n",
       "      <td>8</td>\n",
       "      <td>chr1</td>\n",
       "      <td>937343</td>\n",
       "      <td>937373</td>\n",
       "      <td>937291</td>\n",
       "      <td>937425</td>\n",
       "      <td>ZFX</td>\n",
       "      <td>937343</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>[A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]</td>\n",
       "      <td>[CGGCCCCGTAGCCCAGGAGGTGGTGTGGATGGTGTCCCCGTCCCG...</td>\n",
       "      <td>[ZFX::E2F6+SCC+10+A_B, ZFX::E2F6+SCC+10+A_scrB...</td>\n",
       "      <td>CGGCCCCGTAGCCCAGGAGGTGGTGTGGATGGTGTCCCCGTCCCGC...</td>\n",
       "      <td>CGGCCCCGTAGCCCAGGAGGTGGTGTGGATGGTGTCCCCGTCCCGC...</td>\n",
       "      <td>CGGCCCCGTAGCCCAGGAGGTGGTGTGGATGGTGTCCCCGTCCCGC...</td>\n",
       "      <td>CGGCCCCGTAGCCCAGGAGGTGGTGTGGATGGTGTCCCCGTCCCGC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2390 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        vocab_pair  tot_dist  btn_dist    chr      start       stop  \\\n",
       "0     ARID5B::KLF5        17         2  chr14   69253439   69253456   \n",
       "1     ARID5B::KLF5        17         2   chr6   13477440   13477457   \n",
       "2     ARID5B::KLF5        17         2   chr6   13477440   13477457   \n",
       "3     ARID5B::KLF5        18         3  chr17    7145778    7145796   \n",
       "4     ARID5B::KLF5        18         3  chr19   14616989   14617007   \n",
       "...            ...       ...       ...    ...        ...        ...   \n",
       "2385     ZFX::E2F6        26         4  chr20    2632854    2632880   \n",
       "2386     ZFX::E2F6        28         6   chr2   85811652   85811680   \n",
       "2387     ZFX::E2F6        28         6   chr2  220083840  220083868   \n",
       "2388     ZFX::E2F6        28         6   chr7   44153086   44153114   \n",
       "2389     ZFX::E2F6        30         8   chr1     937343     937373   \n",
       "\n",
       "      start_pad   stop_pad  vocab1  vocab1_start  ...       off_target_name  \\\n",
       "0      69253380   69253514    KLF5      69253439  ...  ARID5B::KLF5+GDSD6+8   \n",
       "1      13477381   13477515  ARID5B      13477440  ...  ARID5B::KLF5+GDSD6+5   \n",
       "2      13477381   13477515  ARID5B      13477440  ...  ARID5B::KLF5+GDSD6+6   \n",
       "3       7145720    7145854  ARID5B       7145778  ...                         \n",
       "4      14616931   14617065    KLF5      14616989  ...  ARID5B::KLF5+GDSD6+7   \n",
       "...         ...        ...     ...           ...  ...                   ...   \n",
       "2385    2632800    2632934     ZFX       2632854  ...                         \n",
       "2386   85811599   85811733     ZFX      85811652  ...       ZFX::E2F6+SCC+1   \n",
       "2387  220083787  220083921    E2F6     220083840  ...       ZFX::E2F6+SCC+2   \n",
       "2388   44153033   44153167    E2F6      44153086  ...       ZFX::E2F6+SCC+9   \n",
       "2389     937291     937425     ZFX        937343  ...                         \n",
       "\n",
       "     off_target_motif  off_target_pos_str  \\\n",
       "0        ARID5B::KLF5              48::16   \n",
       "1              ARID5B                  28   \n",
       "2              ARID5B                  28   \n",
       "3                                           \n",
       "4                KLF5                  26   \n",
       "...               ...                 ...   \n",
       "2385                                        \n",
       "2386              ZFX             104|104   \n",
       "2387             E2F6                 110   \n",
       "2388              ZFX                 108   \n",
       "2389                                        \n",
       "\n",
       "                                            seq_config  \\\n",
       "0     [A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]   \n",
       "1     [A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]   \n",
       "2     [A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]   \n",
       "3     [A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]   \n",
       "4     [A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]   \n",
       "...                                                ...   \n",
       "2385  [A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]   \n",
       "2386  [A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]   \n",
       "2387  [A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]   \n",
       "2388  [A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]   \n",
       "2389  [A_B_seq, A_scrB_seq, scrA_B_seq, scrA_scrB_seq]   \n",
       "\n",
       "                                                    seq  \\\n",
       "0     [ACTCCCAAAACTTCCTGGGTGTAAAGGAACCAAAGGCAAGTGAGC...   \n",
       "1     [CAGGTGGCCTTGATCCAGGCAAAATCTGTATTAGTCAGGGTTCAC...   \n",
       "2     [CAGGTGGCCTTGATCCAGGCAAAATCTGTATTAGTCAGGGTTCAC...   \n",
       "3     [GACGGCGGCGACGCGCGGGCGGATTCAGCGGAGCGATCCACGAAT...   \n",
       "4     [TCTTCCCCTCCACGCTGCGGGCTATGGGAAGTGGCTCGGTCAGTT...   \n",
       "...                                                 ...   \n",
       "2385  [CCGAGTAGGGGACGGGAACGCCTGGGTCCCGGGGACGTGGGGCGC...   \n",
       "2386  [GCGGCGGCAGCAGCGATGGTGAGGGCCCAGGCGGGGCCGGCCAGC...   \n",
       "2387  [GTGCAGGCAACGAATGCAAATTAGAGAAGCAGAGCCGGCTTGGTC...   \n",
       "2388  [AGGTAGAGTCTGGGGAGCCTGGGGGCGAAATTCAGAGAGGGAGGG...   \n",
       "2389  [CGGCCCCGTAGCCCAGGAGGTGGTGTGGATGGTGTCCCCGTCCCG...   \n",
       "\n",
       "                                            name_config  \\\n",
       "0     [ARID5B::KLF5+GDSD6+8+A_B, ARID5B::KLF5+GDSD6+...   \n",
       "1     [ARID5B::KLF5+GDSD6+5+A_B, ARID5B::KLF5+GDSD6+...   \n",
       "2     [ARID5B::KLF5+GDSD6+6+A_B, ARID5B::KLF5+GDSD6+...   \n",
       "3     [ARID5B::KLF5+GDSD6+10+A_B, ARID5B::KLF5+GDSD6...   \n",
       "4     [ARID5B::KLF5+GDSD6+7+A_B, ARID5B::KLF5+GDSD6+...   \n",
       "...                                                 ...   \n",
       "2385  [ZFX::E2F6+SCC+8+A_B, ZFX::E2F6+SCC+8+A_scrB, ...   \n",
       "2386  [ZFX::E2F6+SCC+1+A_B, ZFX::E2F6+SCC+1+A_scrB, ...   \n",
       "2387  [ZFX::E2F6+SCC+2+A_B, ZFX::E2F6+SCC+2+A_scrB, ...   \n",
       "2388  [ZFX::E2F6+SCC+9+A_B, ZFX::E2F6+SCC+9+A_scrB, ...   \n",
       "2389  [ZFX::E2F6+SCC+10+A_B, ZFX::E2F6+SCC+10+A_scrB...   \n",
       "\n",
       "                                                A_B_seq  \\\n",
       "0     ACTCCCAAAACTTCCTAACAGAAGAGCGGTTGAAGGCAAGTGAGCT...   \n",
       "1     CAGGTGGCCTTGATCCAGGCAAAATCTGGTTAAGATTCGGTTCACC...   \n",
       "2     CAGGTGGCCTTGATCCAGGCAAAATCTGAGTGGTAGTTCATTCACC...   \n",
       "3     GACGGCGGCGACGCGCGGGCGGATTCAGCGGAGCGATCCACGAATT...   \n",
       "4     TCTTCCCCTCCACGCTGCGGGCTATGGGTACGCAGGGTATGCGTTT...   \n",
       "...                                                 ...   \n",
       "2385  CCGAGTAGGGGACGGGAACGCCTGGGTCCCGGGGACGTGGGGCGCA...   \n",
       "2386  GCGGCGGCAGCAGCGATGGTGAGGGCCCAGGCGGGGCCGGCCAGCC...   \n",
       "2387  GTGCAGGCAACGAATGCAAATTAGAGAAGCAGAGCCGGCTTGGTCA...   \n",
       "2388  AGGTAGAGTCTGGGGAGCCTGGGGGCGAAATTCAGAGAGGGAGGGC...   \n",
       "2389  CGGCCCCGTAGCCCAGGAGGTGGTGTGGATGGTGTCCCCGTCCCGC...   \n",
       "\n",
       "                                             A_scrB_seq  \\\n",
       "0     ACTCCCAAAACTTCCTAACAGAAGAGCGGTTGAAGGCAAGTGAGCT...   \n",
       "1     CAGGTGGCCTTGATCCAGGCAAAATCTGGTTAAGATTCGGTTCACC...   \n",
       "2     CAGGTGGCCTTGATCCAGGCAAAATCTGAGTGGTAGTTCATTCACC...   \n",
       "3     GACGGCGGCGACGCGCGGGCGGATTCAGCGGAGCGATCCACGAATT...   \n",
       "4     TCTTCCCCTCCACGCTGCGGGCTATGGGTACGCAGGGTATGCGTTT...   \n",
       "...                                                 ...   \n",
       "2385  CCGAGTAGGGGACGGGAACGCCTGGGTCCCGGGGACGTGGGGCGCA...   \n",
       "2386  GCGGCGGCAGCAGCGATGGTGAGGGCCCAGGCGGGGCCGGCCAGCC...   \n",
       "2387  GTGCAGGCAACGAATGCAAATTAGAGAAGCAGAGCCGGCTTGGTCA...   \n",
       "2388  AGGTAGAGTCTGGGGAGCCTGGGGGCGAAATTCAGAGAGGGAGGGC...   \n",
       "2389  CGGCCCCGTAGCCCAGGAGGTGGTGTGGATGGTGTCCCCGTCCCGC...   \n",
       "\n",
       "                                             scrA_B_seq  \\\n",
       "0     ACTCCCAAAACTTCCTAACAGAAGAGCGGTTGAAGGCAAGTGAGCT...   \n",
       "1     CAGGTGGCCTTGATCCAGGCAAAATCTGGTTAAGATTCGGTTCACC...   \n",
       "2     CAGGTGGCCTTGATCCAGGCAAAATCTGAGTGGTAGTTCATTCACC...   \n",
       "3     GACGGCGGCGACGCGCGGGCGGATTCAGCGGAGCGATCCACGAATT...   \n",
       "4     TCTTCCCCTCCACGCTGCGGGCTATGGGTACGCAGGGTATGCGTTT...   \n",
       "...                                                 ...   \n",
       "2385  CCGAGTAGGGGACGGGAACGCCTGGGTCCCGGGGACGTGGGGCGCA...   \n",
       "2386  GCGGCGGCAGCAGCGATGGTGAGGGCCCAGGCGGGGCCGGCCAGCC...   \n",
       "2387  GTGCAGGCAACGAATGCAAATTAGAGAAGCAGAGCCGGCTTGGTCA...   \n",
       "2388  AGGTAGAGTCTGGGGAGCCTGGGGGCGAAATTCAGAGAGGGAGGGC...   \n",
       "2389  CGGCCCCGTAGCCCAGGAGGTGGTGTGGATGGTGTCCCCGTCCCGC...   \n",
       "\n",
       "                                          scrA_scrB_seq  \n",
       "0     ACTCCCAAAACTTCCTAACAGAAGAGCGGTTGAAGGCAAGTGAGCT...  \n",
       "1     CAGGTGGCCTTGATCCAGGCAAAATCTGGTTAAGATTCGGTTCACC...  \n",
       "2     CAGGTGGCCTTGATCCAGGCAAAATCTGAGTGGTAGTTCATTCACC...  \n",
       "3     GACGGCGGCGACGCGCGGGCGGATTCAGCGGAGCGATCCACGAATT...  \n",
       "4     TCTTCCCCTCCACGCTGCGGGCTATGGGTACGCAGGGTATGCGTTT...  \n",
       "...                                                 ...  \n",
       "2385  CCGAGTAGGGGACGGGAACGCCTGGGTCCCGGGGACGTGGGGCGCA...  \n",
       "2386  GCGGCGGCAGCAGCGATGGTGAGGGCCCAGGCGGGGCCGGCCAGCC...  \n",
       "2387  GTGCAGGCAACGAATGCAAATTAGAGAAGCAGAGCCGGCTTGGTCA...  \n",
       "2388  AGGTAGAGTCTGGGGAGCCTGGGGGCGAAATTCAGAGAGGGAGGGC...  \n",
       "2389  CGGCCCCGTAGCCCAGGAGGTGGTGTGGATGGTGTCCCCGTCCCGC...  \n",
       "\n",
       "[2390 rows x 34 columns]"
      ]
     },
     "execution_count": 532,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_seqs_sel_offtarget"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 549,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1     0.004184\n",
      "2     0.012552\n",
      "3     0.033473\n",
      "4     0.100418\n",
      "5     0.221757\n",
      "6     0.188285\n",
      "7     0.205021\n",
      "8     0.125523\n",
      "9     0.066946\n",
      "10    0.041841\n",
      "Name: instance, dtype: float64\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "count    239.000000\n",
       "mean       6.230126\n",
       "std        1.780311\n",
       "min        1.000000\n",
       "25%        5.000000\n",
       "50%        6.000000\n",
       "75%        7.000000\n",
       "max       10.000000\n",
       "Name: instance, dtype: float64"
      ]
     },
     "execution_count": 549,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "off_target_num_df = pd.DataFrame(vocab_instances20_seqs_sel_offtarget.set_index('name').off_target_pos_str.apply(off_target_pos_str_find_num)).reset_index()\n",
    "off_target_num_df['instance'] = off_target_num_df.name.apply(lambda x: '+'.join(x.split('+')[:-1]))\n",
    "off_target_num_df_counts = off_target_num_df[off_target_num_df.off_target_pos_str==0].instance.value_counts().value_counts().sort_index()\n",
    "print(off_target_num_df_counts/off_target_num_df_counts.sum())\n",
    "off_target_num_df[off_target_num_df.off_target_pos_str==0].instance.value_counts().describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 288,
   "metadata": {},
   "outputs": [],
   "source": [
    "# vocab_instances20_seqs_sel_offtarget[vocab_instances20_seqs_sel_offtarget.scrA_B_seq.apply(len)==249]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 483,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>vocab_pair</th>\n",
       "      <th>tot_dist</th>\n",
       "      <th>btn_dist</th>\n",
       "      <th>chr</th>\n",
       "      <th>start_pad</th>\n",
       "      <th>stop_pad</th>\n",
       "      <th>vocab1</th>\n",
       "      <th>vocab1_start_shift</th>\n",
       "      <th>vocab1_stop_shift</th>\n",
       "      <th>vocab2</th>\n",
       "      <th>...</th>\n",
       "      <th>vocab2_stop_shift</th>\n",
       "      <th>tissue</th>\n",
       "      <th>loop_type</th>\n",
       "      <th>vocab_tissue_count</th>\n",
       "      <th>name</th>\n",
       "      <th>check_rs</th>\n",
       "      <th>scr_pass_rs</th>\n",
       "      <th>seq_config</th>\n",
       "      <th>seq</th>\n",
       "      <th>name_config</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>ARID5B::KLF5</td>\n",
       "      <td>21</td>\n",
       "      <td>6</td>\n",
       "      <td>chr18</td>\n",
       "      <td>61243074</td>\n",
       "      <td>61243208</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>49</td>\n",
       "      <td>62</td>\n",
       "      <td>KLF5</td>\n",
       "      <td>...</td>\n",
       "      <td>83</td>\n",
       "      <td>GDSD6</td>\n",
       "      <td>loop_loop</td>\n",
       "      <td>1</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+1</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>A_B_seq</td>\n",
       "      <td>AATCCAGGTTTTAAGGATTACACCATAGCACTCCTAAAGATACATT...</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+1+A_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2397</td>\n",
       "      <td>ARID5B::KLF5</td>\n",
       "      <td>21</td>\n",
       "      <td>6</td>\n",
       "      <td>chr18</td>\n",
       "      <td>61243074</td>\n",
       "      <td>61243208</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>49</td>\n",
       "      <td>62</td>\n",
       "      <td>KLF5</td>\n",
       "      <td>...</td>\n",
       "      <td>83</td>\n",
       "      <td>GDSD6</td>\n",
       "      <td>loop_loop</td>\n",
       "      <td>1</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+1</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>A_scrB_seq</td>\n",
       "      <td>AATCCAGGTTTTAAGGATTACACCATAGCACTCCTAAAGATACATT...</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+1+A_scrB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4787</td>\n",
       "      <td>ARID5B::KLF5</td>\n",
       "      <td>21</td>\n",
       "      <td>6</td>\n",
       "      <td>chr18</td>\n",
       "      <td>61243074</td>\n",
       "      <td>61243208</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>49</td>\n",
       "      <td>62</td>\n",
       "      <td>KLF5</td>\n",
       "      <td>...</td>\n",
       "      <td>83</td>\n",
       "      <td>GDSD6</td>\n",
       "      <td>loop_loop</td>\n",
       "      <td>1</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+1</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>scrA_B_seq</td>\n",
       "      <td>AATCCAGGTTTTAAGGATTACACCATAGCACTCCTAAAGATACATT...</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+1+scrA_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7177</td>\n",
       "      <td>ARID5B::KLF5</td>\n",
       "      <td>21</td>\n",
       "      <td>6</td>\n",
       "      <td>chr18</td>\n",
       "      <td>61243074</td>\n",
       "      <td>61243208</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>49</td>\n",
       "      <td>62</td>\n",
       "      <td>KLF5</td>\n",
       "      <td>...</td>\n",
       "      <td>83</td>\n",
       "      <td>GDSD6</td>\n",
       "      <td>loop_loop</td>\n",
       "      <td>1</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+1</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>scrA_scrB_seq</td>\n",
       "      <td>AATCCAGGTTTTAAGGATTACACCATAGCACTCCTAAAGATACATT...</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+1+scrA_scrB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>ARID5B::KLF5</td>\n",
       "      <td>18</td>\n",
       "      <td>3</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7145720</td>\n",
       "      <td>7145854</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>50</td>\n",
       "      <td>63</td>\n",
       "      <td>KLF5</td>\n",
       "      <td>...</td>\n",
       "      <td>81</td>\n",
       "      <td>GDSD6</td>\n",
       "      <td>loop_loop</td>\n",
       "      <td>10</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+10</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>A_B_seq</td>\n",
       "      <td>GACGGCGGCGACGCGCGGGCGGATTCAGCGGAGCGATCCACGAATT...</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+10+A_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9555</td>\n",
       "      <td>ZFX::E2F6</td>\n",
       "      <td>26</td>\n",
       "      <td>4</td>\n",
       "      <td>chr20</td>\n",
       "      <td>2632800</td>\n",
       "      <td>2632934</td>\n",
       "      <td>ZFX</td>\n",
       "      <td>50</td>\n",
       "      <td>68</td>\n",
       "      <td>E2F6</td>\n",
       "      <td>...</td>\n",
       "      <td>86</td>\n",
       "      <td>SCC</td>\n",
       "      <td>loop_loop</td>\n",
       "      <td>8</td>\n",
       "      <td>ZFX::E2F6+SCC+8</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>scrA_scrB_seq</td>\n",
       "      <td>CCGAGTAGGGGACGGGAACGCCTGGGTCCCGGGGACGTGGGGCGCA...</td>\n",
       "      <td>ZFX::E2F6+SCC+8+scrA_scrB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2388</td>\n",
       "      <td>ZFX::E2F6</td>\n",
       "      <td>28</td>\n",
       "      <td>6</td>\n",
       "      <td>chr7</td>\n",
       "      <td>44153033</td>\n",
       "      <td>44153167</td>\n",
       "      <td>E2F6</td>\n",
       "      <td>47</td>\n",
       "      <td>68</td>\n",
       "      <td>ZFX</td>\n",
       "      <td>...</td>\n",
       "      <td>85</td>\n",
       "      <td>SCC</td>\n",
       "      <td>pro_pro</td>\n",
       "      <td>9</td>\n",
       "      <td>ZFX::E2F6+SCC+9</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>A_B_seq</td>\n",
       "      <td>AGGTAGAGTCTGGGGAGCCTGGGGGCGAAATTCAGAGAGGGAGGGC...</td>\n",
       "      <td>ZFX::E2F6+SCC+9+A_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4778</td>\n",
       "      <td>ZFX::E2F6</td>\n",
       "      <td>28</td>\n",
       "      <td>6</td>\n",
       "      <td>chr7</td>\n",
       "      <td>44153033</td>\n",
       "      <td>44153167</td>\n",
       "      <td>E2F6</td>\n",
       "      <td>47</td>\n",
       "      <td>68</td>\n",
       "      <td>ZFX</td>\n",
       "      <td>...</td>\n",
       "      <td>85</td>\n",
       "      <td>SCC</td>\n",
       "      <td>pro_pro</td>\n",
       "      <td>9</td>\n",
       "      <td>ZFX::E2F6+SCC+9</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>A_scrB_seq</td>\n",
       "      <td>AGGTAGAGTCTGGGGAGCCTGGGGGCGAAATTCAGAGAGGGAGGGC...</td>\n",
       "      <td>ZFX::E2F6+SCC+9+A_scrB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7168</td>\n",
       "      <td>ZFX::E2F6</td>\n",
       "      <td>28</td>\n",
       "      <td>6</td>\n",
       "      <td>chr7</td>\n",
       "      <td>44153033</td>\n",
       "      <td>44153167</td>\n",
       "      <td>E2F6</td>\n",
       "      <td>47</td>\n",
       "      <td>68</td>\n",
       "      <td>ZFX</td>\n",
       "      <td>...</td>\n",
       "      <td>85</td>\n",
       "      <td>SCC</td>\n",
       "      <td>pro_pro</td>\n",
       "      <td>9</td>\n",
       "      <td>ZFX::E2F6+SCC+9</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>scrA_B_seq</td>\n",
       "      <td>AGGTAGAGTCTGGGGAGCCTGGGGGCGAAATTCAGAGAGGGAGGGC...</td>\n",
       "      <td>ZFX::E2F6+SCC+9+scrA_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9558</td>\n",
       "      <td>ZFX::E2F6</td>\n",
       "      <td>28</td>\n",
       "      <td>6</td>\n",
       "      <td>chr7</td>\n",
       "      <td>44153033</td>\n",
       "      <td>44153167</td>\n",
       "      <td>E2F6</td>\n",
       "      <td>47</td>\n",
       "      <td>68</td>\n",
       "      <td>ZFX</td>\n",
       "      <td>...</td>\n",
       "      <td>85</td>\n",
       "      <td>SCC</td>\n",
       "      <td>pro_pro</td>\n",
       "      <td>9</td>\n",
       "      <td>ZFX::E2F6+SCC+9</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>scrA_scrB_seq</td>\n",
       "      <td>AGGTAGAGTCTGGGGAGCCTGGGGGCGAAATTCAGAGAGGGAGGGC...</td>\n",
       "      <td>ZFX::E2F6+SCC+9+scrA_scrB</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9560 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        vocab_pair  tot_dist  btn_dist    chr  start_pad  stop_pad  vocab1  \\\n",
       "7     ARID5B::KLF5        21         6  chr18   61243074  61243208  ARID5B   \n",
       "2397  ARID5B::KLF5        21         6  chr18   61243074  61243208  ARID5B   \n",
       "4787  ARID5B::KLF5        21         6  chr18   61243074  61243208  ARID5B   \n",
       "7177  ARID5B::KLF5        21         6  chr18   61243074  61243208  ARID5B   \n",
       "3     ARID5B::KLF5        18         3  chr17    7145720   7145854  ARID5B   \n",
       "...            ...       ...       ...    ...        ...       ...     ...   \n",
       "9555     ZFX::E2F6        26         4  chr20    2632800   2632934     ZFX   \n",
       "2388     ZFX::E2F6        28         6   chr7   44153033  44153167    E2F6   \n",
       "4778     ZFX::E2F6        28         6   chr7   44153033  44153167    E2F6   \n",
       "7168     ZFX::E2F6        28         6   chr7   44153033  44153167    E2F6   \n",
       "9558     ZFX::E2F6        28         6   chr7   44153033  44153167    E2F6   \n",
       "\n",
       "      vocab1_start_shift  vocab1_stop_shift vocab2  ...  vocab2_stop_shift  \\\n",
       "7                     49                 62   KLF5  ...                 83   \n",
       "2397                  49                 62   KLF5  ...                 83   \n",
       "4787                  49                 62   KLF5  ...                 83   \n",
       "7177                  49                 62   KLF5  ...                 83   \n",
       "3                     50                 63   KLF5  ...                 81   \n",
       "...                  ...                ...    ...  ...                ...   \n",
       "9555                  50                 68   E2F6  ...                 86   \n",
       "2388                  47                 68    ZFX  ...                 85   \n",
       "4778                  47                 68    ZFX  ...                 85   \n",
       "7168                  47                 68    ZFX  ...                 85   \n",
       "9558                  47                 68    ZFX  ...                 85   \n",
       "\n",
       "      tissue  loop_type vocab_tissue_count                   name check_rs  \\\n",
       "7      GDSD6  loop_loop                  1   ARID5B::KLF5+GDSD6+1        0   \n",
       "2397   GDSD6  loop_loop                  1   ARID5B::KLF5+GDSD6+1        0   \n",
       "4787   GDSD6  loop_loop                  1   ARID5B::KLF5+GDSD6+1        0   \n",
       "7177   GDSD6  loop_loop                  1   ARID5B::KLF5+GDSD6+1        0   \n",
       "3      GDSD6  loop_loop                 10  ARID5B::KLF5+GDSD6+10        0   \n",
       "...      ...        ...                ...                    ...      ...   \n",
       "9555     SCC  loop_loop                  8        ZFX::E2F6+SCC+8        0   \n",
       "2388     SCC    pro_pro                  9        ZFX::E2F6+SCC+9        0   \n",
       "4778     SCC    pro_pro                  9        ZFX::E2F6+SCC+9        0   \n",
       "7168     SCC    pro_pro                  9        ZFX::E2F6+SCC+9        0   \n",
       "9558     SCC    pro_pro                  9        ZFX::E2F6+SCC+9        0   \n",
       "\n",
       "      scr_pass_rs     seq_config  \\\n",
       "7            True        A_B_seq   \n",
       "2397         True     A_scrB_seq   \n",
       "4787         True     scrA_B_seq   \n",
       "7177         True  scrA_scrB_seq   \n",
       "3            True        A_B_seq   \n",
       "...           ...            ...   \n",
       "9555         True  scrA_scrB_seq   \n",
       "2388         True        A_B_seq   \n",
       "4778         True     A_scrB_seq   \n",
       "7168         True     scrA_B_seq   \n",
       "9558         True  scrA_scrB_seq   \n",
       "\n",
       "                                                    seq  \\\n",
       "7     AATCCAGGTTTTAAGGATTACACCATAGCACTCCTAAAGATACATT...   \n",
       "2397  AATCCAGGTTTTAAGGATTACACCATAGCACTCCTAAAGATACATT...   \n",
       "4787  AATCCAGGTTTTAAGGATTACACCATAGCACTCCTAAAGATACATT...   \n",
       "7177  AATCCAGGTTTTAAGGATTACACCATAGCACTCCTAAAGATACATT...   \n",
       "3     GACGGCGGCGACGCGCGGGCGGATTCAGCGGAGCGATCCACGAATT...   \n",
       "...                                                 ...   \n",
       "9555  CCGAGTAGGGGACGGGAACGCCTGGGTCCCGGGGACGTGGGGCGCA...   \n",
       "2388  AGGTAGAGTCTGGGGAGCCTGGGGGCGAAATTCAGAGAGGGAGGGC...   \n",
       "4778  AGGTAGAGTCTGGGGAGCCTGGGGGCGAAATTCAGAGAGGGAGGGC...   \n",
       "7168  AGGTAGAGTCTGGGGAGCCTGGGGGCGAAATTCAGAGAGGGAGGGC...   \n",
       "9558  AGGTAGAGTCTGGGGAGCCTGGGGGCGAAATTCAGAGAGGGAGGGC...   \n",
       "\n",
       "                         name_config  \n",
       "7           ARID5B::KLF5+GDSD6+1+A_B  \n",
       "2397     ARID5B::KLF5+GDSD6+1+A_scrB  \n",
       "4787     ARID5B::KLF5+GDSD6+1+scrA_B  \n",
       "7177  ARID5B::KLF5+GDSD6+1+scrA_scrB  \n",
       "3          ARID5B::KLF5+GDSD6+10+A_B  \n",
       "...                              ...  \n",
       "9555       ZFX::E2F6+SCC+8+scrA_scrB  \n",
       "2388             ZFX::E2F6+SCC+9+A_B  \n",
       "4778          ZFX::E2F6+SCC+9+A_scrB  \n",
       "7168          ZFX::E2F6+SCC+9+scrA_B  \n",
       "9558       ZFX::E2F6+SCC+9+scrA_scrB  \n",
       "\n",
       "[9560 rows x 21 columns]"
      ]
     },
     "execution_count": 483,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_seqs_sel_offtarget.reset_index(drop=True, inplace=True)\n",
    "# vocab_instances20_seqs_sel_offtarget[ 'vocab_tissue_count'] = vocab_instances20_seqs_sel_offtarget.groupby(['vocab_pair','tissue']).cumcount()+1\n",
    "# vocab_instances20_seqs_sel_offtarget['name'] = vocab_instances20_seqs_sel_offtarget.vocab_pair + '+'+ vocab_instances20_seqs_sel_offtarget.tissue + '+'+ vocab_instances20_seqs_sel_offtarget.vocab_tissue_count.map(str)\n",
    "vocab_instances20_seqs = pd.melt(vocab_instances20_seqs_sel_offtarget.drop(['start','stop', 'vocab1_start', 'vocab1_stop',\n",
    "                                                                            'vocab2_start', 'vocab2_stop','off_target_name', 'off_target_motif', 'off_target_pos_str',\n",
    "                                                                            'seq_config','seq','name_config'],axis=1),\n",
    "         id_vars=['vocab_pair', 'tot_dist', 'btn_dist', 'chr', 'start_pad', 'stop_pad',\n",
    "                  'vocab1','vocab1_start_shift','vocab1_stop_shift','vocab2', 'vocab2_start_shift', 'vocab2_stop_shift',\n",
    "                  'tissue', 'loop_type',  'vocab_tissue_count','name','check_rs','scr_pass_rs'], \n",
    "            var_name='seq_config',\n",
    "             value_name=\"seq\")\n",
    "vocab_instances20_seqs['name_config'] = vocab_instances20_seqs.name + '+' + vocab_instances20_seqs.seq_config.str.split('_seq',expand=True).iloc[:,0]\n",
    "\n",
    "vocab_instances20_seqs.sort_values(['name_config'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 499,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "135    9560\n",
       "Name: seq, dtype: int64"
      ]
     },
     "execution_count": 499,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_seqs.seq.str.len().value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 484,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(9560,)"
      ]
     },
     "execution_count": 484,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_seqs.name_config.unique().shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 485,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(9560, 2)"
      ]
     },
     "execution_count": 485,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_seqs[['name','seq_config']].drop_duplicates().shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 486,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['A_B_seq', 'scrA_B_seq', 'A_scrB_seq', 'scrA_scrB_seq'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 486,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_seqs_sel.seq_config.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 487,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([135])"
      ]
     },
     "execution_count": 487,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_seqs.seq.apply(len).unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 488,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(9560, 21) (4246, 20)\n"
     ]
    }
   ],
   "source": [
    "# save the vocab info\n",
    "print(vocab_instances20_seqs.shape, vocab_instances20_alltissue_df_rs_filt.shape)\n",
    "vocab_instances20_seqs.to_csv(os.path.join(save_dir, 'vocab_instances20_seq_wscr.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 498,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "135    9560\n",
       "Name: seq, dtype: int64"
      ]
     },
     "execution_count": 498,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 489,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['vocab_pair', 'tot_dist', 'btn_dist', 'chr', 'start_pad', 'stop_pad',\n",
       "       'vocab1', 'vocab1_start_shift', 'vocab1_stop_shift', 'vocab2',\n",
       "       'vocab2_start_shift', 'vocab2_stop_shift', 'tissue', 'loop_type',\n",
       "       'vocab_tissue_count', 'name', 'check_rs', 'scr_pass_rs', 'seq_config',\n",
       "       'seq', 'name_config'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 489,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_seqs.columns\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. make post scan mpra library\n",
    "\n",
    "Input DFs:\n",
    "- `vocab_instances20_seqs`: `vocab_instances20_seq_wscr.csv`\n",
    "- `pos_genes_df_rs_filt`: `pos_genes_position_wseq_rsfilt.csv`\n",
    "- `neg_genes_df_rs_filt`: `neg_genes_position_wseq_rsfilt.csv`\n",
    "\n",
    "Output DF: \n",
    "- `mpra_lib_seq`: 'mpra_lib_seq.csv`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 490,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(239, 2)"
      ]
     },
     "execution_count": 490,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_seqs[['vocab_pair','tissue']].drop_duplicates().shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "still have 239 vocab words"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 491,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(239,)"
      ]
     },
     "execution_count": 491,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_seqs[vocab_instances20_seqs.vocab_tissue_count==10].vocab_pair.unique().shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "all 239 vocab words have at least 10 genomic instances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 493,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>vocab_pair</th>\n",
       "      <th>tot_dist</th>\n",
       "      <th>btn_dist</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>start_pad</th>\n",
       "      <th>stop_pad</th>\n",
       "      <th>vocab1</th>\n",
       "      <th>vocab1_start</th>\n",
       "      <th>...</th>\n",
       "      <th>vocab2_stop</th>\n",
       "      <th>tissue</th>\n",
       "      <th>loop_type</th>\n",
       "      <th>vocab_tissue_count</th>\n",
       "      <th>name</th>\n",
       "      <th>check_rs</th>\n",
       "      <th>scr_pass_rs</th>\n",
       "      <th>seq_config</th>\n",
       "      <th>seq</th>\n",
       "      <th>name_config</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [vocab_pair, tot_dist, btn_dist, chr, start, stop, start_pad, stop_pad, vocab1, vocab1_start, vocab1_stop, vocab2, vocab2_start, vocab2_stop, tissue, loop_type, vocab_tissue_count, name, check_rs, scr_pass_rs, seq_config, seq, name_config]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 23 columns]"
      ]
     },
     "execution_count": 493,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_seqs_sel[vocab_instances20_seqs_sel.seq.apply(len)!=135]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "all vocab genomic i"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### saving selection of vocab words\n",
    "how `name_config` works:\n",
    "\n",
    "`df['vocab_pair','tissue','vocab_tissue_count','seq_config'] = df.name_config.str.split('+',expand=True)`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 596,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(9560, 2)"
      ]
     },
     "execution_count": 596,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_seqs[['name','seq_config']].drop_duplicates().shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 597,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SNAI2::WT1+Colon+7             4\n",
       "SP4::THAP1+Colon+8             4\n",
       "NFE2L1::TBX15+MM+6             4\n",
       "NFE2L1::NR3C1+MM+8             4\n",
       "ARID5B::STAT1+MM+2             4\n",
       "                              ..\n",
       "KLF15::ZNF219+Melanocytes+2    4\n",
       "RORA::TP63+GDSD6+2             4\n",
       "EGR2::TP63+GDSD6+4             4\n",
       "EN2::PRRX1+Melanocytes+10      4\n",
       "KLF15::KLF5+Colon+9            4\n",
       "Name: name, Length: 2390, dtype: int64"
      ]
     },
     "execution_count": 597,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_seqs.name.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 598,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['vocab_pair', 'tot_dist', 'btn_dist', 'chr', 'start_pad', 'stop_pad',\n",
       "       'vocab1', 'vocab1_start_shift', 'vocab1_stop_shift', 'vocab2',\n",
       "       'vocab2_start_shift', 'vocab2_stop_shift', 'tissue', 'loop_type',\n",
       "       'vocab_tissue_count', 'name', 'check_rs', 'scr_pass_rs', 'seq_config',\n",
       "       'seq', 'name_config'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 598,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_seqs.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 601,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>vocab_pair</th>\n",
       "      <th>tot_dist</th>\n",
       "      <th>btn_dist</th>\n",
       "      <th>chr</th>\n",
       "      <th>start_pad</th>\n",
       "      <th>stop_pad</th>\n",
       "      <th>vocab1</th>\n",
       "      <th>vocab1_start_shift</th>\n",
       "      <th>vocab1_stop_shift</th>\n",
       "      <th>vocab2</th>\n",
       "      <th>...</th>\n",
       "      <th>vocab2_stop_shift</th>\n",
       "      <th>tissue</th>\n",
       "      <th>loop_type</th>\n",
       "      <th>vocab_tissue_count</th>\n",
       "      <th>name</th>\n",
       "      <th>check_rs</th>\n",
       "      <th>scr_pass_rs</th>\n",
       "      <th>seq_config</th>\n",
       "      <th>seq</th>\n",
       "      <th>name_config</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>ARID5B::KLF5</td>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "      <td>chr14</td>\n",
       "      <td>69253380</td>\n",
       "      <td>69253514</td>\n",
       "      <td>KLF5</td>\n",
       "      <td>54</td>\n",
       "      <td>71</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>...</td>\n",
       "      <td>85</td>\n",
       "      <td>GDSD6</td>\n",
       "      <td>loop_loop</td>\n",
       "      <td>8</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+8</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>A_B_seq</td>\n",
       "      <td>ACTCCCAAAACTTCCTAACAGAAGAGCGGTTGAAGGCAAGTGAGCT...</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+8+A_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>ARID5B::KLF5</td>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "      <td>chr6</td>\n",
       "      <td>13477381</td>\n",
       "      <td>13477515</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>51</td>\n",
       "      <td>64</td>\n",
       "      <td>KLF5</td>\n",
       "      <td>...</td>\n",
       "      <td>81</td>\n",
       "      <td>GDSD6</td>\n",
       "      <td>loop_loop</td>\n",
       "      <td>5</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+5</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>A_B_seq</td>\n",
       "      <td>CAGGTGGCCTTGATCCAGGCAAAATCTGGTTAAGATTCGGTTCACC...</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+5+A_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>ARID5B::KLF5</td>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "      <td>chr6</td>\n",
       "      <td>13477381</td>\n",
       "      <td>13477515</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>51</td>\n",
       "      <td>64</td>\n",
       "      <td>KLF5</td>\n",
       "      <td>...</td>\n",
       "      <td>81</td>\n",
       "      <td>GDSD6</td>\n",
       "      <td>pro_pro</td>\n",
       "      <td>6</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+6</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>A_B_seq</td>\n",
       "      <td>CAGGTGGCCTTGATCCAGGCAAAATCTGAGTGGTAGTTCATTCACC...</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+6+A_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>ARID5B::KLF5</td>\n",
       "      <td>18</td>\n",
       "      <td>3</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7145720</td>\n",
       "      <td>7145854</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>50</td>\n",
       "      <td>63</td>\n",
       "      <td>KLF5</td>\n",
       "      <td>...</td>\n",
       "      <td>81</td>\n",
       "      <td>GDSD6</td>\n",
       "      <td>loop_loop</td>\n",
       "      <td>10</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+10</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>A_B_seq</td>\n",
       "      <td>GACGGCGGCGACGCGCGGGCGGATTCAGCGGAGCGATCCACGAATT...</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+10+A_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>ARID5B::KLF5</td>\n",
       "      <td>18</td>\n",
       "      <td>3</td>\n",
       "      <td>chr19</td>\n",
       "      <td>14616931</td>\n",
       "      <td>14617065</td>\n",
       "      <td>KLF5</td>\n",
       "      <td>53</td>\n",
       "      <td>70</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>...</td>\n",
       "      <td>85</td>\n",
       "      <td>GDSD6</td>\n",
       "      <td>loop_loop</td>\n",
       "      <td>7</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+7</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>A_B_seq</td>\n",
       "      <td>TCTTCCCCTCCACGCTGCGGGCTATGGGTACGCAGGGTATGCGTTT...</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+7+A_B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     vocab_pair  tot_dist  btn_dist    chr  start_pad  stop_pad  vocab1  \\\n",
       "0  ARID5B::KLF5        17         2  chr14   69253380  69253514    KLF5   \n",
       "1  ARID5B::KLF5        17         2   chr6   13477381  13477515  ARID5B   \n",
       "2  ARID5B::KLF5        17         2   chr6   13477381  13477515  ARID5B   \n",
       "3  ARID5B::KLF5        18         3  chr17    7145720   7145854  ARID5B   \n",
       "4  ARID5B::KLF5        18         3  chr19   14616931  14617065    KLF5   \n",
       "\n",
       "   vocab1_start_shift  vocab1_stop_shift  vocab2  ...  vocab2_stop_shift  \\\n",
       "0                  54                 71  ARID5B  ...                 85   \n",
       "1                  51                 64    KLF5  ...                 81   \n",
       "2                  51                 64    KLF5  ...                 81   \n",
       "3                  50                 63    KLF5  ...                 81   \n",
       "4                  53                 70  ARID5B  ...                 85   \n",
       "\n",
       "   tissue  loop_type vocab_tissue_count                   name check_rs  \\\n",
       "0   GDSD6  loop_loop                  8   ARID5B::KLF5+GDSD6+8        0   \n",
       "1   GDSD6  loop_loop                  5   ARID5B::KLF5+GDSD6+5        0   \n",
       "2   GDSD6    pro_pro                  6   ARID5B::KLF5+GDSD6+6        0   \n",
       "3   GDSD6  loop_loop                 10  ARID5B::KLF5+GDSD6+10        0   \n",
       "4   GDSD6  loop_loop                  7   ARID5B::KLF5+GDSD6+7        0   \n",
       "\n",
       "   scr_pass_rs  seq_config                                                seq  \\\n",
       "0         True     A_B_seq  ACTCCCAAAACTTCCTAACAGAAGAGCGGTTGAAGGCAAGTGAGCT...   \n",
       "1         True     A_B_seq  CAGGTGGCCTTGATCCAGGCAAAATCTGGTTAAGATTCGGTTCACC...   \n",
       "2         True     A_B_seq  CAGGTGGCCTTGATCCAGGCAAAATCTGAGTGGTAGTTCATTCACC...   \n",
       "3         True     A_B_seq  GACGGCGGCGACGCGCGGGCGGATTCAGCGGAGCGATCCACGAATT...   \n",
       "4         True     A_B_seq  TCTTCCCCTCCACGCTGCGGGCTATGGGTACGCAGGGTATGCGTTT...   \n",
       "\n",
       "                 name_config  \n",
       "0   ARID5B::KLF5+GDSD6+8+A_B  \n",
       "1   ARID5B::KLF5+GDSD6+5+A_B  \n",
       "2   ARID5B::KLF5+GDSD6+6+A_B  \n",
       "3  ARID5B::KLF5+GDSD6+10+A_B  \n",
       "4   ARID5B::KLF5+GDSD6+7+A_B  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 601,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_seqs[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 602,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(9560,)"
      ]
     },
     "execution_count": 602,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_seqs.name_config.unique().shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 603,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  This is separate from the ipykernel package so we can avoid doing imports until\n"
     ]
    }
   ],
   "source": [
    "vocab_instances20_seqs_selcols = vocab_instances20_seqs[['chr','start_pad','stop_pad','name_config','seq']]\n",
    "vocab_instances20_seqs_selcols.columns = ['chr','start','stop','name','seq']\n",
    "vocab_instances20_seqs_selcols['seq_type'] = 'vocab'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 604,
   "metadata": {},
   "outputs": [],
   "source": [
    "pos_genes_df_seqs_sel_cols = pos_genes_df_rs_filt[['chr','start','end','gene','sequence']]\n",
    "pos_genes_df_seqs_sel_cols.columns = ['chr','start','stop','name','seq']\n",
    "pos_genes_df_seqs_sel_cols['seq_type'] = 'pos'\n",
    "neg_genes_df_seqs_sel_cols = neg_genes_df_rs_filt[['chr','start','end','gene','sequence']]\n",
    "neg_genes_df_seqs_sel_cols.columns = ['chr','start','stop','name','seq']\n",
    "neg_genes_df_seqs_sel_cols['seq_type'] = 'neg'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 605,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>name</th>\n",
       "      <th>seq</th>\n",
       "      <th>seq_type</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>chr14</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+8+A_B</td>\n",
       "      <td>ACTCCCAAAACTTCCTAACAGAAGAGCGGTTGAAGGCAAGTGAGCT...</td>\n",
       "      <td>vocab</td>\n",
       "      <td>69253380</td>\n",
       "      <td>69253514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>chr6</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+5+A_B</td>\n",
       "      <td>CAGGTGGCCTTGATCCAGGCAAAATCTGGTTAAGATTCGGTTCACC...</td>\n",
       "      <td>vocab</td>\n",
       "      <td>13477381</td>\n",
       "      <td>13477515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>chr6</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+6+A_B</td>\n",
       "      <td>CAGGTGGCCTTGATCCAGGCAAAATCTGAGTGGTAGTTCATTCACC...</td>\n",
       "      <td>vocab</td>\n",
       "      <td>13477381</td>\n",
       "      <td>13477515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>chr17</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+10+A_B</td>\n",
       "      <td>GACGGCGGCGACGCGCGGGCGGATTCAGCGGAGCGATCCACGAATT...</td>\n",
       "      <td>vocab</td>\n",
       "      <td>7145720</td>\n",
       "      <td>7145854</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>chr19</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+7+A_B</td>\n",
       "      <td>TCTTCCCCTCCACGCTGCGGGCTATGGGTACGCAGGGTATGCGTTT...</td>\n",
       "      <td>vocab</td>\n",
       "      <td>14616931</td>\n",
       "      <td>14617065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>89</td>\n",
       "      <td>chrX</td>\n",
       "      <td>RHOXF2B</td>\n",
       "      <td>ACCACGCTCTCCTAGGCAGTGCTAGAACTGTGAGTCGCCGCCTGCA...</td>\n",
       "      <td>neg</td>\n",
       "      <td>119211572</td>\n",
       "      <td>119211707</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>90</td>\n",
       "      <td>chrX</td>\n",
       "      <td>CT47A12</td>\n",
       "      <td>AGTCTCTCTCTTTATTGAGGGAATAAGAGTCTGTCTCAGACGACAC...</td>\n",
       "      <td>neg</td>\n",
       "      <td>120066016</td>\n",
       "      <td>120066151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>91</td>\n",
       "      <td>chrX</td>\n",
       "      <td>CT47A6</td>\n",
       "      <td>AGTCTCTCTCTTTATTGAGGGAATAAGAGTCTGTCTCAGACGACAC...</td>\n",
       "      <td>neg</td>\n",
       "      <td>120095202</td>\n",
       "      <td>120095337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>92</td>\n",
       "      <td>chrX</td>\n",
       "      <td>SPANXN4</td>\n",
       "      <td>TGTCAGAGGTGGTTCTTTGTGATGCCAAAGCCACCCTAGGACTACC...</td>\n",
       "      <td>neg</td>\n",
       "      <td>142113569</td>\n",
       "      <td>142113704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>93</td>\n",
       "      <td>chrX</td>\n",
       "      <td>SPANXN2</td>\n",
       "      <td>TGTTCTTTTACCCAATGTCCATGTAGTTCCTTTAAGTTTCCTATTT...</td>\n",
       "      <td>neg</td>\n",
       "      <td>142804381</td>\n",
       "      <td>142804516</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9721 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chr                       name  \\\n",
       "0   chr14   ARID5B::KLF5+GDSD6+8+A_B   \n",
       "1    chr6   ARID5B::KLF5+GDSD6+5+A_B   \n",
       "2    chr6   ARID5B::KLF5+GDSD6+6+A_B   \n",
       "3   chr17  ARID5B::KLF5+GDSD6+10+A_B   \n",
       "4   chr19   ARID5B::KLF5+GDSD6+7+A_B   \n",
       "..    ...                        ...   \n",
       "89   chrX                    RHOXF2B   \n",
       "90   chrX                    CT47A12   \n",
       "91   chrX                     CT47A6   \n",
       "92   chrX                    SPANXN4   \n",
       "93   chrX                    SPANXN2   \n",
       "\n",
       "                                                  seq seq_type      start  \\\n",
       "0   ACTCCCAAAACTTCCTAACAGAAGAGCGGTTGAAGGCAAGTGAGCT...    vocab   69253380   \n",
       "1   CAGGTGGCCTTGATCCAGGCAAAATCTGGTTAAGATTCGGTTCACC...    vocab   13477381   \n",
       "2   CAGGTGGCCTTGATCCAGGCAAAATCTGAGTGGTAGTTCATTCACC...    vocab   13477381   \n",
       "3   GACGGCGGCGACGCGCGGGCGGATTCAGCGGAGCGATCCACGAATT...    vocab    7145720   \n",
       "4   TCTTCCCCTCCACGCTGCGGGCTATGGGTACGCAGGGTATGCGTTT...    vocab   14616931   \n",
       "..                                                ...      ...        ...   \n",
       "89  ACCACGCTCTCCTAGGCAGTGCTAGAACTGTGAGTCGCCGCCTGCA...      neg  119211572   \n",
       "90  AGTCTCTCTCTTTATTGAGGGAATAAGAGTCTGTCTCAGACGACAC...      neg  120066016   \n",
       "91  AGTCTCTCTCTTTATTGAGGGAATAAGAGTCTGTCTCAGACGACAC...      neg  120095202   \n",
       "92  TGTCAGAGGTGGTTCTTTGTGATGCCAAAGCCACCCTAGGACTACC...      neg  142113569   \n",
       "93  TGTTCTTTTACCCAATGTCCATGTAGTTCCTTTAAGTTTCCTATTT...      neg  142804381   \n",
       "\n",
       "         stop  \n",
       "0    69253514  \n",
       "1    13477515  \n",
       "2    13477515  \n",
       "3     7145854  \n",
       "4    14617065  \n",
       "..        ...  \n",
       "89  119211707  \n",
       "90  120066151  \n",
       "91  120095337  \n",
       "92  142113704  \n",
       "93  142804516  \n",
       "\n",
       "[9721 rows x 6 columns]"
      ]
     },
     "execution_count": 605,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mpra_lib_seq = pd.concat([vocab_instances20_seqs_selcols,pos_genes_df_seqs_sel_cols, neg_genes_df_seqs_sel_cols ],axis=0,sort=True)\n",
    "mpra_lib_seq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### some stats then saving"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 606,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "vocab    9560\n",
       "neg        89\n",
       "pos        72\n",
       "Name: seq_type, dtype: int64"
      ]
     },
     "execution_count": 606,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mpra_lib_seq.seq_type.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 607,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GDSD6          49.0\n",
       "SCC            43.0\n",
       "Colon          42.0\n",
       "MM             40.0\n",
       "GM12878        39.0\n",
       "Melanocytes    26.0\n",
       "Name: tissue, dtype: float64"
      ]
     },
     "execution_count": 607,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances20_seqs_sel.tissue.value_counts()/40"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 608,
   "metadata": {},
   "outputs": [],
   "source": [
    "# vocab_instances20_seqs_sel.vocab_pair[vocab_instances20_seqs_sel.vocab_pair.isin(irf_vocab)].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 609,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # number of vocab words that have irf vocab setup\n",
    "# vocab_instances20_seqs_sel.vocab_pair[vocab_instances20_seqs_sel.vocab_pair.isin(irf_vocab)].unique().shape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 610,
   "metadata": {},
   "outputs": [],
   "source": [
    "mpra_lib_seq.to_csv(os.path.join(save_dir, 'mpra_lib_seq.csv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. add MPRA adaptors and barcodes\n",
    "\n",
    "done in 8E1R (based on `make_library_psych-postcadd-01119.ipynb` from the mpra library\n",
    "\n",
    "note that the rs sites chosen from section 4B are: `[RS_ECORI ,RS_BAMHI , RS_XHOI , RS_NHEI]`\n",
    "\n",
    "input is `../data/processed/fig4_modelling/vocab_sum_final_post_scan/mpra_lib_seq.csv`\n",
    "\n",
    "output is `../data/processed/fig4_modelling/vocab_sum_final_post_scan/mpra_oligo_df_filt.csv`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 611,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# define constants\n",
    "LEN_FILLER = 20\n",
    "NUM_OLIGO_PER_MOTIF = 10\n",
    "DESIRED_OLIGO_LENGTH = 135 #(changed from 150 )\n",
    "MAX_OLIGO_LENGTH = 230\n",
    "MAX_MOTIF_LENGTH = 63\n",
    "FORWARD_PCR_PRIMER = 'ACTGGCCGCTTCACTG'\n",
    "REVERSE_PCR_PRIMER = 'AGATCGGAAGAGCGTCG'\n",
    "# LINKER_SPACER_v10 = 'CAGTATGCGATGCTCATGATTGTGACTGAAAGTCTATGAGTCCATTGA'\n",
    "RS_ECORI = 'GAATTC' # 5'-3'\n",
    "RS_BAMHI = 'GGATCC'\n",
    "RS_XHOI = 'CTCGAG'\n",
    "RS_NHEI = 'GCTAGC'\n",
    "\n",
    "\n",
    "RS_LIST = [RS_ECORI ,RS_BAMHI , RS_XHOI , RS_NHEI]\n",
    "\n",
    "LETTERS=['A', 'C', 'G', 'T']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 7. MPRA final check \n",
    "\n",
    "- checks to make sure all sequences have 10 barcodes\n",
    "- all rs sites are passed (done in R notebook)\n",
    "- that there are still 239 vocab words with 40 instance (4 config of 10 each)\n",
    "\n",
    "- `../data/processed/fig4_modelling/vocab_sum_final_post_scan/mpra_lib_seq.csv` - includes information on all the sequences used\n",
    "- `../data/processed/fig4_modelling/vocab_sum_final_post_scan/mpra_oligo_df_filt.csv` - oligo information with a lot of columns (RS site check results etc.(\n",
    "- `../data/processed/fig4_modelling/vocab_sum_final_post_scan/mpra_oligo_df_final.txt` - with columns: name, seq_type, id, oligo, seq, filler, barcode\n",
    "- `../data/processed/fig4_modelling/vocab_sum_final_post_scan/mpra_oligo_df_final_SUBMIT.txt` - with columns: id, oligo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 612,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['../data/processed/fig4_modelling/vocab_sum_final_post_scan/']"
      ]
     },
     "execution_count": 612,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glob.glob(os.path.join(save_dir))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 613,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(97210, 21)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>seq</th>\n",
       "      <th>seq_type</th>\n",
       "      <th>count</th>\n",
       "      <th>id</th>\n",
       "      <th>pass_restrict_seq</th>\n",
       "      <th>pass_restrict_fwd_seq</th>\n",
       "      <th>pass_restrict_seq_xho</th>\n",
       "      <th>barcode</th>\n",
       "      <th>len_seq</th>\n",
       "      <th>...</th>\n",
       "      <th>1:n()</th>\n",
       "      <th>filler</th>\n",
       "      <th>filler_check</th>\n",
       "      <th>oligo</th>\n",
       "      <th>len_oligo</th>\n",
       "      <th>pass_restrict_ECORI</th>\n",
       "      <th>pass_restrict_BAMHI</th>\n",
       "      <th>pass_restrict_XHOI</th>\n",
       "      <th>pass_restrict_NHEI</th>\n",
       "      <th>pass_all</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>ABCF3</td>\n",
       "      <td>ACCTTGAGCTCCCAGCCTTTGCCCCACCCCTCGCTCCCGGAACTCC...</td>\n",
       "      <td>pos</td>\n",
       "      <td>1</td>\n",
       "      <td>ABCF3.1</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>AGTTGGAATGAAGGCCTCGA</td>\n",
       "      <td>136</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>TCCTTACAAAAGAACGGCCG</td>\n",
       "      <td>True</td>\n",
       "      <td>ACTGGCCGCTTCACTGACCTTGAGCTCCCAGCCTTTGCCCCACCCC...</td>\n",
       "      <td>221</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>ABCF3</td>\n",
       "      <td>ACCTTGAGCTCCCAGCCTTTGCCCCACCCCTCGCTCCCGGAACTCC...</td>\n",
       "      <td>pos</td>\n",
       "      <td>2</td>\n",
       "      <td>ABCF3.2</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>GTCATAGTCAGAGAAGCTCT</td>\n",
       "      <td>136</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>TCTCTTGACGCGATGCGAGC</td>\n",
       "      <td>True</td>\n",
       "      <td>ACTGGCCGCTTCACTGACCTTGAGCTCCCAGCCTTTGCCCCACCCC...</td>\n",
       "      <td>221</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>ABCF3</td>\n",
       "      <td>ACCTTGAGCTCCCAGCCTTTGCCCCACCCCTCGCTCCCGGAACTCC...</td>\n",
       "      <td>pos</td>\n",
       "      <td>3</td>\n",
       "      <td>ABCF3.3</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>GATACTTCATGCAATATGCT</td>\n",
       "      <td>136</td>\n",
       "      <td>...</td>\n",
       "      <td>19</td>\n",
       "      <td>TGCTATGGGGGATAGATCTT</td>\n",
       "      <td>True</td>\n",
       "      <td>ACTGGCCGCTTCACTGACCTTGAGCTCCCAGCCTTTGCCCCACCCC...</td>\n",
       "      <td>221</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>ABCF3</td>\n",
       "      <td>ACCTTGAGCTCCCAGCCTTTGCCCCACCCCTCGCTCCCGGAACTCC...</td>\n",
       "      <td>pos</td>\n",
       "      <td>4</td>\n",
       "      <td>ABCF3.4</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CGATTGCAGGAGAATAGTCC</td>\n",
       "      <td>136</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ACCGTGCCAGGGCTCACCGT</td>\n",
       "      <td>True</td>\n",
       "      <td>ACTGGCCGCTTCACTGACCTTGAGCTCCCAGCCTTTGCCCCACCCC...</td>\n",
       "      <td>221</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>ABCF3</td>\n",
       "      <td>ACCTTGAGCTCCCAGCCTTTGCCCCACCCCTCGCTCCCGGAACTCC...</td>\n",
       "      <td>pos</td>\n",
       "      <td>5</td>\n",
       "      <td>ABCF3.5</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>TGAGAACCAGTTCCAACGCA</td>\n",
       "      <td>136</td>\n",
       "      <td>...</td>\n",
       "      <td>11</td>\n",
       "      <td>AGACTGGTAGTGAACCTGAT</td>\n",
       "      <td>True</td>\n",
       "      <td>ACTGGCCGCTTCACTGACCTTGAGCTCCCAGCCTTTGCCCCACCCC...</td>\n",
       "      <td>221</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    name                                                seq seq_type  count  \\\n",
       "1  ABCF3  ACCTTGAGCTCCCAGCCTTTGCCCCACCCCTCGCTCCCGGAACTCC...      pos      1   \n",
       "2  ABCF3  ACCTTGAGCTCCCAGCCTTTGCCCCACCCCTCGCTCCCGGAACTCC...      pos      2   \n",
       "3  ABCF3  ACCTTGAGCTCCCAGCCTTTGCCCCACCCCTCGCTCCCGGAACTCC...      pos      3   \n",
       "4  ABCF3  ACCTTGAGCTCCCAGCCTTTGCCCCACCCCTCGCTCCCGGAACTCC...      pos      4   \n",
       "5  ABCF3  ACCTTGAGCTCCCAGCCTTTGCCCCACCCCTCGCTCCCGGAACTCC...      pos      5   \n",
       "\n",
       "        id  pass_restrict_seq  pass_restrict_fwd_seq  pass_restrict_seq_xho  \\\n",
       "1  ABCF3.1               True                   True                   True   \n",
       "2  ABCF3.2               True                   True                   True   \n",
       "3  ABCF3.3               True                   True                   True   \n",
       "4  ABCF3.4               True                   True                   True   \n",
       "5  ABCF3.5               True                   True                   True   \n",
       "\n",
       "                barcode  len_seq  ...  1:n()                filler  \\\n",
       "1  AGTTGGAATGAAGGCCTCGA      136  ...      4  TCCTTACAAAAGAACGGCCG   \n",
       "2  GTCATAGTCAGAGAAGCTCT      136  ...      3  TCTCTTGACGCGATGCGAGC   \n",
       "3  GATACTTCATGCAATATGCT      136  ...     19  TGCTATGGGGGATAGATCTT   \n",
       "4  CGATTGCAGGAGAATAGTCC      136  ...      2  ACCGTGCCAGGGCTCACCGT   \n",
       "5  TGAGAACCAGTTCCAACGCA      136  ...     11  AGACTGGTAGTGAACCTGAT   \n",
       "\n",
       "  filler_check                                              oligo len_oligo  \\\n",
       "1         True  ACTGGCCGCTTCACTGACCTTGAGCTCCCAGCCTTTGCCCCACCCC...       221   \n",
       "2         True  ACTGGCCGCTTCACTGACCTTGAGCTCCCAGCCTTTGCCCCACCCC...       221   \n",
       "3         True  ACTGGCCGCTTCACTGACCTTGAGCTCCCAGCCTTTGCCCCACCCC...       221   \n",
       "4         True  ACTGGCCGCTTCACTGACCTTGAGCTCCCAGCCTTTGCCCCACCCC...       221   \n",
       "5         True  ACTGGCCGCTTCACTGACCTTGAGCTCCCAGCCTTTGCCCCACCCC...       221   \n",
       "\n",
       "   pass_restrict_ECORI  pass_restrict_BAMHI  pass_restrict_XHOI  \\\n",
       "1                 True                 True                True   \n",
       "2                 True                 True                True   \n",
       "3                 True                 True                True   \n",
       "4                 True                 True                True   \n",
       "5                 True                 True                True   \n",
       "\n",
       "   pass_restrict_NHEI  pass_all  \n",
       "1                True      True  \n",
       "2                True      True  \n",
       "3                True      True  \n",
       "4                True      True  \n",
       "5                True      True  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 613,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oligo_df_filt = pd.read_csv(os.path.join(save_dir,'mpra_oligo_df_filt.csv'),index_col=0)\n",
    "print(oligo_df_filt.shape)\n",
    "oligo_df_filt[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 614,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Series([], Name: name, dtype: int64)"
      ]
     },
     "execution_count": 614,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq_counts = oligo_df_filt.name.value_counts()\n",
    "seq_counts[seq_counts!=10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " all sequences have 10 barcodes\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "rs sites checks (done in R notebook)\n",
    "- no EcoRI \n",
    "- no BamHI sites\n",
    "- exactly 1 XhoI site\n",
    "- exactly 1 NheI site\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 615,
   "metadata": {},
   "outputs": [],
   "source": [
    "vocab_oligo_df =  oligo_df_filt[oligo_df_filt.seq_type=='vocab']\n",
    "    \n",
    "vocab_oligo_df[['vocab_pair','tissue','vocab_tissue_count','seq_config']]= vocab_oligo_df.name.str.split('+',expand=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 616,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Series([], Name: vocab_pair, dtype: int64)"
      ]
     },
     "execution_count": 616,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_oligo_counts = vocab_oligo_df.vocab_pair.value_counts()\n",
    "vocab_oligo_counts[vocab_oligo_counts!=400]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " all vocab types have 4 configs x 10 genomic instances x 10 barcodes = 400 oligos\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 617,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(239,)"
      ]
     },
     "execution_count": 617,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_oligo_counts.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "there are 239 vocab words"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7B motif scan\n",
    "\n",
    "### 1. make fasta \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 619,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(9560, 2)\n"
     ]
    }
   ],
   "source": [
    "fasta_file = os.path.join(save_dir, 'mpra_vocab_seq5.fasta')\n",
    "print(vocab_oligo_df[['name','seq']].drop_duplicates().shape)\n",
    "with open(fasta_file, 'w') as f:\n",
    "    for idx, row in vocab_oligo_df[['name','seq']].drop_duplicates().iterrows():\n",
    "        f.write('>'+row['name']+'\\n')\n",
    "        f.write(row['seq']+'\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_control_file = os.path.join(save_dir, 'mpra_control_seq.fasta')\n",
    "control_oligo_df = oligo_df_filt[oligo_df_filt.seq_type.isin(['pos','neg'])]\n",
    "print(control_oligo_df[['name','seq']].drop_duplicates().shape)\n",
    "with open(fasta_control_file, 'w') as f:\n",
    "    for idx, row in control_oligo_df[['name','seq']].drop_duplicates().iterrows():\n",
    "        f.write('>'+row['name']+'\\n')\n",
    "        f.write(row['seq']+'\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### scanning try 3 MOODS package\n",
    "https://github.com/jhkorhonen/MOODS/wiki/Getting-started\n",
    "Matrices. The matrices are expected in the JASPAR .pfm format or the .adm format for first-order models. A JASPAR .pfm file consist of four rows, specifying the counts or frequencies for nucleotides A, C, G and T, respectively:\n",
    "\n",
    "#### DESCRIPTION\n",
    "\n",
    "MOODS 1.9.1 introduced a Python command-line script moods-dna.py for standalone PWM analysis using the MOODS algorithms. This section explains the basics of using `moods-dna.py`\n",
    "\n",
    "#### Input matrics (motifs)\n",
    "pfms are in `/Users/mguo123/Google Drive/1_khavari/omics_project-LD/pan_omics/data/external/hocomoco_human_trim`\n",
    "Note: the extensions are pwm but are pfms\n",
    "\n",
    "#### Input sequences\n",
    "Sequences. The sequences can be given either as plain text files, or as fasta files possibly containing multiple sequences. Newlines and leading and trailing whitespace will be ignored, but other characters not encoding nucleotides will be treated as non-matchable positions (i.e., they should not appear outside fasta headers). IUPAC nucleotide ambiguity codes (WSMKRYBDHV) will be treated as coding SNPs, and MOODS will match the matrices versus all possible combinations.\n",
    "\n",
    "#### Output (csv)\n",
    "The comma-separated fields are:\n",
    "\n",
    "- Sequence name (either file name or fasta header).\n",
    "- Matrix file name.\n",
    "- Hit position (first position of sequence is position 0).\n",
    "- Indicates whether the match is versus the input strand (+) or the reverse complement (-).\n",
    "- Match score.\n",
    "- Hit site in the original sequence (note that if the match is -, then the match is versus the reverse complement of this).\n",
    "- Hit sequence with SNPs applied, i.e. shows how the ambiguity codes are resolved to get this score (there can be multiple hits at the same position with different \"real\" hit sequences).\n",
    "\n",
    "\n",
    "#### Installation \n",
    "\n",
    "follow: https://pypi.org/project/MOODS-python/\n",
    "\n",
    "this doesn't work with subprocess\n",
    "```\n",
    "pip install --user MOODS-python\n",
    "PATH=\"$PATH:`python -m site --user-base`/bin\"\n",
    "```\n",
    "\n",
    "#### Basic usage\n",
    "The basic usage pattern of `moods-dna.py` is the following:\n",
    "\n",
    "```\n",
    "moods-dna.py --help\n",
    "usage: moods-dna.py [-h] [-v] [-m M [M ...]] [-S M [M ...]] [-s S [S ...]]\n",
    "                    [-p p] [-t T] [-B n] [-o outfile] [--sep S] [-R]\n",
    "                    [--no-snps] [--batch] [--bg pA pC pG pT] [--ps p]\n",
    "                    [--log-base x] [--lo-bg pA pC pG pT]\n",
    "                    [--threshold-precision x]\n",
    "\n",
    "matches position weight matrices against DNA sequences\n",
    "\n",
    "optional arguments:\n",
    "  -h, --help            show this help message and exit\n",
    "  -v, --verbosity       verbose (-vv, -vvv for more)\n",
    "\n",
    "input files (at least one matrix and sequence file required):\n",
    "  -m M [M ...], --matrices M [M ...]\n",
    "                        matrix files (count/frequency, will be converted to\n",
    "                        PWM before matching)\n",
    "  -S M [M ...], --score-matrices M [M ...]\n",
    "                        matrix files (PWM/other score matrix, will be matched\n",
    "                        directly)\n",
    "  -s S [S ...], --sequences S [S ...]\n",
    "                        sequence files\n",
    "\n",
    "threshold selection (exactly one required):\n",
    "  -p p, --p-value p     compute threshold from p-value\n",
    "  -t T, --threshold T   use specified absolute threshold\n",
    "  -B n, --best-hits n   return at least the specified amount of best matches\n",
    "\n",
    "output (optional):\n",
    "  -o outfile, --outfile outfile\n",
    "                        output to file instead of standard output\n",
    "  --sep S               set field separator in out (default \",\")\n",
    "\n",
    "search and model behaviour (optional):\n",
    "  -R, --no-rc           disable matching versus reverse complement strand\n",
    "  --no-snps             ignore IUPAC symbols coding multiple nucleotides\n",
    "  --batch               do not recompute thresholds from p-values for each\n",
    "                        sequence separately (recommended when dealing with\n",
    "                        lots of short sequences)\n",
    "  --bg pA pC pG pT      background distribution for computing thresholds from\n",
    "                        p-value with --batch (default is 0.25 for all alleles)\n",
    "  --ps p                total pseudocount added to each matrix column in log-\n",
    "                        odds conversion (default = 0.01)\n",
    "  --log-base x          logarithm base for log-odds conversion (default\n",
    "                        natural logarithm)\n",
    "  --lo-bg pA pC pG pT   background distribution for log-odds conversion\n",
    "                        (default is 0.25 for all alleles)\n",
    "  --threshold-precision x\n",
    "                        specify the precision used for computing the\n",
    "                        thresholds from p-values (default = 2000.0)\n",
    "                        \n",
    "                        \n",
    "moods-dna.py -m example-data/matrices/*.{pfm,adm} -s example-data/seq/chr1-5k-55k.fa -p 0.0001\n",
    "moods-dna.py -m ../data/external/hocomoco_human_trim/*.pwm -s ../data/processed/fig4_modelling/vocab_sum_final/mpra_vocab_seq.fasta -p 0.0001\n",
    "\n",
    "moods-dna.py -m ../data/external/hocomoco_human_trim/FOXM1_HUMAN.H11MO.0.A.pwm.trim.pwm -s ../data/processed/fig4_modelling/vocab_sum_final/mpra_vocab_seq.fasta -p 0.0001\n",
    "\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 510,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # !pip install --user MOODS-python\n",
    "# !PATH=\"$PATH:`python -m site --user-base`/bin\"\n",
    "# glob.glob(os.path.join(save_dir,\"*fasta\"))\n",
    "import subprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 511,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "123"
      ]
     },
     "execution_count": 511,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# num actual tfs used \n",
    "list_tfs = sorted(set(list(vocab_instances20_seqs.vocab1.unique()) + list(vocab_instances20_seqs.vocab2.unique())))\n",
    "len(list_tfs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 523,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make directory\n",
    "motif_path = os.path.join(save_dir, 'motif_scan4')\n",
    "if not os.path.exists(motif_path):\n",
    "    os.makedirs(motif_path)\n",
    "print(motif_path)\n",
    "# input_fasta = '../data/processed/fig4_modelling/vocab_sum_final_post_scan/mpra_vocab_seq5.fasta'\n",
    "input_fasta = '../data/processed/fig4_modelling/vocab_sum_final_post_scan/mpra_vocab_seq4.fasta'\n",
    "# run_file = 'run_8E1_make_mpra_7_motif_scan.sh'\n",
    "run_file = 'run_8F1_make_mpra_7_motif_scan_vocab_filt5.sh'\n",
    "run_file = 'run_8F1_make_mpra_7_motif_scan_vocab_filt4.sh'\n",
    "\n",
    "with open(run_file, 'w') as f:\n",
    "# print(len(glob.glob('../data/external/hocomoco_human_trim/*pwm')))\n",
    "    for pwm_file in sorted(glob.glob('../data/external/hocomoco_human_trim/*pwm')):\n",
    "        pwm = os.path.basename(pwm_file).split('.pwm')[0]\n",
    "        tf = tf_id_abbr_to_name_dict[pwm]\n",
    "        if tf in list_tfs:\n",
    "            output_file = os.path.join(motif_path, pwm+'_scan.csv')\n",
    "            cmd = 'moods-dna.py -m {} -s {} -p 0.05 -o {}'.format(pwm_file,input_fasta, output_file) #motif_scan3 uses this, difference from motif_scan4 is the pvalue            \n",
    "#             cmd = 'moods-dna.py -m {} -s {} -p 0.001 -o {}'.format(pwm_file,input_fasta, output_file) #motif_scan5 \n",
    "            f.write(cmd+'\\n')\n",
    "# testing\n",
    "# cmd = 'moods-dna.py -m {} -s {} -p 0.00005 -o {}'.format(pwm_file,input_fasta, output_file)\n",
    "cmd = 'moods-dna.py -m {} -s {} -p 0.05 -o {}'.format(pwm_file,input_fasta, output_file)\n",
    "print(cmd)    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### controls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 622,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../data/processed/fig4_modelling/vocab_sum_final_post_scan/motif_scan_control\n",
      "moods-dna.py -m ../data/external/hocomoco_human_trim/ZSC31_HUMAN.H11MO.0.C.pwm.trim.pwm -s ../data/processed/fig4_modelling/vocab_sum_final_post_scan/mpra_control_seq.fasta -p 0.05 -o ../data/processed/fig4_modelling/vocab_sum_final_post_scan/motif_scan_control/ZN816_HUMAN.H11MO.1.C_scan.csv\n"
     ]
    }
   ],
   "source": [
    "# make directory\n",
    "motif_path = os.path.join(save_dir, 'motif_scan_control')\n",
    "if not os.path.exists(motif_path):\n",
    "    os.makedirs(motif_path)\n",
    "print(motif_path)\n",
    "input_fasta = '../data/processed/fig4_modelling/vocab_sum_final_post_scan/mpra_control_seq.fasta'\n",
    "run_file = 'run_8F1_make_mpra_7_motif_scan_control.sh'\n",
    "\n",
    "with open(run_file, 'w') as f:\n",
    "# print(len(glob.glob('../data/external/hocomoco_human_trim/*pwm')))\n",
    "    for pwm_file in sorted(glob.glob('../data/external/hocomoco_human_trim/*pwm')):\n",
    "        pwm = os.path.basename(pwm_file).split('.pwm')[0]\n",
    "        tf = tf_id_abbr_to_name_dict[pwm]\n",
    "        if tf in list_tfs:\n",
    "            output_file = os.path.join(motif_path, pwm+'_scan.csv')\n",
    "            cmd = 'moods-dna.py -m {} -s {} -p 0.05 -o {}'.format(pwm_file,input_fasta, output_file) #motif_scan3 uses this, difference from motif_scan4 is the pvalue            \n",
    "#             cmd = 'moods-dna.py -m {} -s {} -p 0.001 -o {}'.format(pwm_file,input_fasta, output_file) #motif_scan5 \n",
    "            f.write(cmd+'\\n')\n",
    "# testing\n",
    "# cmd = 'moods-dna.py -m {} -s {} -p 0.00005 -o {}'.format(pwm_file,input_fasta, output_file)\n",
    "cmd = 'moods-dna.py -m {} -s {} -p 0.05 -o {}'.format(pwm_file,input_fasta, output_file)\n",
    "print(cmd)    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "in cmd line in this folder run\n",
    "- `bash run_8E1_make_mpra_7_motif_scan_vocab_filt4.sh`\n",
    "- `bash run_8E1_make_mpra_7_motif_scan_vocab_filt5.sh`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "in sherlock\n",
    "```\n",
    "cd /oak/stanford/groups/khavari/users/mguo123/pan_omics/notebooks\n",
    "\n",
    "sbatch run_8F1_make_mpra_7_motif_scan_vocab_filt4.sbatch\n",
    "\n",
    "\n",
    "\n",
    "#!/bin/bash\n",
    "#run_8F1_make_mpra_7_motif_scan4.sbatch\n",
    "#09/01/2020\n",
    "#SBATCH -J motif_filt# A single job name for the array\n",
    "#SBATCH -p rbaltman,khavari # Partition\n",
    "#SBATCH -n 1 # one core\n",
    "#SBATCH --mem=64000\n",
    "#SBATCH --time=48:00:00\n",
    "#SBATCH -o run_8E1_filt.out # Standard output\n",
    "#SBATCH -e run_8E1_filt.err # Standard error\n",
    "#SBATCH --mail-type=FAIL,END # notifications for job done & fail\n",
    "#SBATCH --mail-user=mguo123@stanford.edu\n",
    "\n",
    "cd /oak/stanford/groups/khavari/users/mguo123/pan_omics/notebooks\n",
    "PATH=\"$PATH:`python -m site --user-base`/bin\"\n",
    "\n",
    "bash run_8F1_make_mpra_7_motif_scan_vocab_filt4.sh\n",
    "\n",
    "```\n",
    "\n",
    "#### header of the `*_scan.csv`\n",
    "   `ARID5B::KLF5+GDSD6+4+A_B,SRBP2_HUMAN.H11MO.0.B.pwm.trim.pwm,73,-,5.35564936116,CCACCCCGC,`\n",
    "cvs headers (8 cols)\n",
    "- vocabulary config name (from fasta)\n",
    "- pwm filename\n",
    "- position in the sequence (index 0)\n",
    "- strand: Indicates whether the match is versus the input strand (+) or the reverse complement (-).\n",
    "- score (bit score)\n",
    "- sequence substring from what we are scanning (original sequence) (note that if the match is -, then the match is versus the reverse complement of this).\n",
    "- (NOT USED)Hit sequence with SNPs applied, i.e. shows how the ambiguity codes are resolved to get this score (there can be multiple hits at the same position with different \"real\" hit sequences).\n",
    "\n",
    "\n",
    "#### more on the score\n",
    "    https://www.cs.helsinki.fi/group/pssmfind/doc/python/MOODS.html\n",
    "      The input PWM matrix transformed to log-odds scores.\n",
    "      The score for nucleotide N and position i is\n",
    "\n",
    "         log ( matrix[N][i] + ps * bg[N] ) / C[i]\n",
    "           - log ( bg[N] ),\n",
    "\n",
    "      where C[i] is the sum of terms\n",
    "\n",
    "         matrix[N][i] + ps * bg[N]\n",
    "\n",
    "      for all nucleotides N.\n",
    "\n",
    "    so all the matches to the PWM transformed log-odds score are summed to get the match score\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7C read in motif scan data \n",
    "\n",
    "cases for the motif in the \n",
    "1. scrambling removes motif \n",
    "2. there's a dummy motif somewhere else in the genomic instance (????? How would you deal with these “noisy” cases?\n",
    "    2A. is the dummy motif is present in all 4 cases, (not changed by scrambling)\n",
    "    2B. the dummy motif is present in only some cases, so probably changed by scrambling the other site?\n",
    "3. scrambling reintroduces the motif at a (hopefully) lower strength, and sometimes with a shifted start position  ??? Potential to rescramble (but sometimes motifs have too many repetitive nucleotides so this might not work…\n",
    "    3A. the scrambling introduces the motif at a lower strength\n",
    "    3B. the scrambling does not lower the score of the motif\n",
    "4. we don't detect the motif in any config in the genomic instance at a certain p value threshold (this usually isn’t the case, if we raise the p-value threshold enough; but the motif will be detected in a lot more genomic instances). This is sort of the sensitivity vs specificity tradeoff.\n",
    "\n",
    "\n",
    "2 parts\n",
    "1. map the vocab word locations to motifs and match to case above (get the score and the case) - exact match\n",
    "2. determine vocab words gained lost between the motif configurations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 7C1: get vocab_instance_df set up with locations and vocab predictions given config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 395,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>vocab_pair</th>\n",
       "      <th>tot_dist</th>\n",
       "      <th>btn_dist</th>\n",
       "      <th>chr</th>\n",
       "      <th>start_pad</th>\n",
       "      <th>stop_pad</th>\n",
       "      <th>vocab1</th>\n",
       "      <th>vocab1_start_shift</th>\n",
       "      <th>vocab1_stop_shift</th>\n",
       "      <th>vocab2</th>\n",
       "      <th>...</th>\n",
       "      <th>vocab2_stop_shift</th>\n",
       "      <th>tissue</th>\n",
       "      <th>loop_type</th>\n",
       "      <th>vocab_tissue_count</th>\n",
       "      <th>name</th>\n",
       "      <th>check_rs</th>\n",
       "      <th>scr_pass_rs</th>\n",
       "      <th>seq_config</th>\n",
       "      <th>seq</th>\n",
       "      <th>name_config</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>ARID5B::KLF5</td>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "      <td>chr14</td>\n",
       "      <td>69253380</td>\n",
       "      <td>69253514</td>\n",
       "      <td>KLF5</td>\n",
       "      <td>54</td>\n",
       "      <td>71</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>...</td>\n",
       "      <td>85</td>\n",
       "      <td>GDSD6</td>\n",
       "      <td>loop_loop</td>\n",
       "      <td>8</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+8</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>A_B_seq</td>\n",
       "      <td>ACTCCCAAAACTTCAAGCTCGAAAACAGGTTGAGGGCAAGTGAGCA...</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+8+A_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>ARID5B::KLF5</td>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "      <td>chr6</td>\n",
       "      <td>13477381</td>\n",
       "      <td>13477515</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>51</td>\n",
       "      <td>64</td>\n",
       "      <td>KLF5</td>\n",
       "      <td>...</td>\n",
       "      <td>81</td>\n",
       "      <td>GDSD6</td>\n",
       "      <td>loop_loop</td>\n",
       "      <td>5</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+5</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>A_B_seq</td>\n",
       "      <td>CAGGTGGCCTTGATCCAGGCAAAAATTGCGACTGTTTGATATCGCC...</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+5+A_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>ARID5B::KLF5</td>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "      <td>chr6</td>\n",
       "      <td>13477381</td>\n",
       "      <td>13477515</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>51</td>\n",
       "      <td>64</td>\n",
       "      <td>KLF5</td>\n",
       "      <td>...</td>\n",
       "      <td>81</td>\n",
       "      <td>GDSD6</td>\n",
       "      <td>pro_pro</td>\n",
       "      <td>6</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+6</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>A_B_seq</td>\n",
       "      <td>CAGGTGGCCTTGATCCAGGCAAAACTGGGAGATTGTTATCTTACCC...</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+6+A_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>ARID5B::KLF5</td>\n",
       "      <td>18</td>\n",
       "      <td>3</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7145720</td>\n",
       "      <td>7145854</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>50</td>\n",
       "      <td>63</td>\n",
       "      <td>KLF5</td>\n",
       "      <td>...</td>\n",
       "      <td>81</td>\n",
       "      <td>GDSD6</td>\n",
       "      <td>loop_loop</td>\n",
       "      <td>10</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+10</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>A_B_seq</td>\n",
       "      <td>GACGGCGGCGACGCGCGGGCGGATTCAGCGGAGCGATCCACGAATT...</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+10+A_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>ARID5B::KLF5</td>\n",
       "      <td>18</td>\n",
       "      <td>3</td>\n",
       "      <td>chr19</td>\n",
       "      <td>14616931</td>\n",
       "      <td>14617065</td>\n",
       "      <td>KLF5</td>\n",
       "      <td>53</td>\n",
       "      <td>70</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>...</td>\n",
       "      <td>85</td>\n",
       "      <td>GDSD6</td>\n",
       "      <td>loop_loop</td>\n",
       "      <td>7</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+7</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>A_B_seq</td>\n",
       "      <td>TCTTCCCCTCCACGCTGCGGGCTAGGATTCGGATGTGGCCTGGATT...</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+7+A_B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     vocab_pair  tot_dist  btn_dist    chr  start_pad  stop_pad  vocab1  \\\n",
       "0  ARID5B::KLF5        17         2  chr14   69253380  69253514    KLF5   \n",
       "1  ARID5B::KLF5        17         2   chr6   13477381  13477515  ARID5B   \n",
       "2  ARID5B::KLF5        17         2   chr6   13477381  13477515  ARID5B   \n",
       "3  ARID5B::KLF5        18         3  chr17    7145720   7145854  ARID5B   \n",
       "4  ARID5B::KLF5        18         3  chr19   14616931  14617065    KLF5   \n",
       "\n",
       "   vocab1_start_shift  vocab1_stop_shift  vocab2  ...  vocab2_stop_shift  \\\n",
       "0                  54                 71  ARID5B  ...                 85   \n",
       "1                  51                 64    KLF5  ...                 81   \n",
       "2                  51                 64    KLF5  ...                 81   \n",
       "3                  50                 63    KLF5  ...                 81   \n",
       "4                  53                 70  ARID5B  ...                 85   \n",
       "\n",
       "   tissue  loop_type vocab_tissue_count                   name check_rs  \\\n",
       "0   GDSD6  loop_loop                  8   ARID5B::KLF5+GDSD6+8        0   \n",
       "1   GDSD6  loop_loop                  5   ARID5B::KLF5+GDSD6+5        0   \n",
       "2   GDSD6    pro_pro                  6   ARID5B::KLF5+GDSD6+6        0   \n",
       "3   GDSD6  loop_loop                 10  ARID5B::KLF5+GDSD6+10        0   \n",
       "4   GDSD6  loop_loop                  7   ARID5B::KLF5+GDSD6+7        0   \n",
       "\n",
       "   scr_pass_rs  seq_config                                                seq  \\\n",
       "0         True     A_B_seq  ACTCCCAAAACTTCAAGCTCGAAAACAGGTTGAGGGCAAGTGAGCA...   \n",
       "1         True     A_B_seq  CAGGTGGCCTTGATCCAGGCAAAAATTGCGACTGTTTGATATCGCC...   \n",
       "2         True     A_B_seq  CAGGTGGCCTTGATCCAGGCAAAACTGGGAGATTGTTATCTTACCC...   \n",
       "3         True     A_B_seq  GACGGCGGCGACGCGCGGGCGGATTCAGCGGAGCGATCCACGAATT...   \n",
       "4         True     A_B_seq  TCTTCCCCTCCACGCTGCGGGCTAGGATTCGGATGTGGCCTGGATT...   \n",
       "\n",
       "                 name_config  \n",
       "0   ARID5B::KLF5+GDSD6+8+A_B  \n",
       "1   ARID5B::KLF5+GDSD6+5+A_B  \n",
       "2   ARID5B::KLF5+GDSD6+6+A_B  \n",
       "3  ARID5B::KLF5+GDSD6+10+A_B  \n",
       "4   ARID5B::KLF5+GDSD6+7+A_B  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 395,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# vocab_oligo_df.columns\n",
    "vocab_instances20_seqs_sel = pd.read_csv(os.path.join(save_dir, 'vocab_instances20_seq_wscr.csv'),index_col=0)\n",
    "vocab_instances20_seqs_sel[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 396,
   "metadata": {},
   "outputs": [],
   "source": [
    "def vocab_present_func(row):\n",
    "    # determine if there should be a vocab present (based on configuration)\n",
    "    return row['vocab_order'] in row['config'].split('_')\n",
    "#     vocab_instances_df.vocab_order.apply(lambda x: xisin(vocab_instances_df.config.str.split('_'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 397,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  This is separate from the ipykernel package so we can avoid doing imports until\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  import sys\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>name_config</th>\n",
       "      <th>start_pad</th>\n",
       "      <th>stop_pad</th>\n",
       "      <th>vocab</th>\n",
       "      <th>vocab_order</th>\n",
       "      <th>vocab_start</th>\n",
       "      <th>vocab_stop</th>\n",
       "      <th>instance</th>\n",
       "      <th>config</th>\n",
       "      <th>vocab_present_pred</th>\n",
       "      <th>vocab_loc_start</th>\n",
       "      <th>vocab_loc_stop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>chr18</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+1+A_B</td>\n",
       "      <td>61243074</td>\n",
       "      <td>61243208</td>\n",
       "      <td>KLF5</td>\n",
       "      <td>B</td>\n",
       "      <td>62</td>\n",
       "      <td>83</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+1</td>\n",
       "      <td>A_B</td>\n",
       "      <td>True</td>\n",
       "      <td>62</td>\n",
       "      <td>83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>chr18</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+1+A_B</td>\n",
       "      <td>61243074</td>\n",
       "      <td>61243208</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>A</td>\n",
       "      <td>49</td>\n",
       "      <td>62</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+1</td>\n",
       "      <td>A_B</td>\n",
       "      <td>True</td>\n",
       "      <td>49</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2397</td>\n",
       "      <td>chr18</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+1+A_scrB</td>\n",
       "      <td>61243074</td>\n",
       "      <td>61243208</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>A</td>\n",
       "      <td>49</td>\n",
       "      <td>62</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+1</td>\n",
       "      <td>A_scrB</td>\n",
       "      <td>True</td>\n",
       "      <td>49</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2397</td>\n",
       "      <td>chr18</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+1+A_scrB</td>\n",
       "      <td>61243074</td>\n",
       "      <td>61243208</td>\n",
       "      <td>KLF5</td>\n",
       "      <td>B</td>\n",
       "      <td>62</td>\n",
       "      <td>83</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+1</td>\n",
       "      <td>A_scrB</td>\n",
       "      <td>False</td>\n",
       "      <td>62</td>\n",
       "      <td>83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4787</td>\n",
       "      <td>chr18</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+1+scrA_B</td>\n",
       "      <td>61243074</td>\n",
       "      <td>61243208</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>A</td>\n",
       "      <td>49</td>\n",
       "      <td>62</td>\n",
       "      <td>ARID5B::KLF5+GDSD6+1</td>\n",
       "      <td>scrA_B</td>\n",
       "      <td>False</td>\n",
       "      <td>49</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4778</td>\n",
       "      <td>chr7</td>\n",
       "      <td>ZFX::E2F6+SCC+9+A_scrB</td>\n",
       "      <td>44153033</td>\n",
       "      <td>44153167</td>\n",
       "      <td>E2F6</td>\n",
       "      <td>A</td>\n",
       "      <td>47</td>\n",
       "      <td>68</td>\n",
       "      <td>ZFX::E2F6+SCC+9</td>\n",
       "      <td>A_scrB</td>\n",
       "      <td>True</td>\n",
       "      <td>47</td>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7168</td>\n",
       "      <td>chr7</td>\n",
       "      <td>ZFX::E2F6+SCC+9+scrA_B</td>\n",
       "      <td>44153033</td>\n",
       "      <td>44153167</td>\n",
       "      <td>ZFX</td>\n",
       "      <td>B</td>\n",
       "      <td>68</td>\n",
       "      <td>85</td>\n",
       "      <td>ZFX::E2F6+SCC+9</td>\n",
       "      <td>scrA_B</td>\n",
       "      <td>True</td>\n",
       "      <td>68</td>\n",
       "      <td>85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7168</td>\n",
       "      <td>chr7</td>\n",
       "      <td>ZFX::E2F6+SCC+9+scrA_B</td>\n",
       "      <td>44153033</td>\n",
       "      <td>44153167</td>\n",
       "      <td>E2F6</td>\n",
       "      <td>A</td>\n",
       "      <td>47</td>\n",
       "      <td>68</td>\n",
       "      <td>ZFX::E2F6+SCC+9</td>\n",
       "      <td>scrA_B</td>\n",
       "      <td>False</td>\n",
       "      <td>47</td>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9558</td>\n",
       "      <td>chr7</td>\n",
       "      <td>ZFX::E2F6+SCC+9+scrA_scrB</td>\n",
       "      <td>44153033</td>\n",
       "      <td>44153167</td>\n",
       "      <td>E2F6</td>\n",
       "      <td>A</td>\n",
       "      <td>47</td>\n",
       "      <td>68</td>\n",
       "      <td>ZFX::E2F6+SCC+9</td>\n",
       "      <td>scrA_scrB</td>\n",
       "      <td>False</td>\n",
       "      <td>47</td>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9558</td>\n",
       "      <td>chr7</td>\n",
       "      <td>ZFX::E2F6+SCC+9+scrA_scrB</td>\n",
       "      <td>44153033</td>\n",
       "      <td>44153167</td>\n",
       "      <td>ZFX</td>\n",
       "      <td>B</td>\n",
       "      <td>68</td>\n",
       "      <td>85</td>\n",
       "      <td>ZFX::E2F6+SCC+9</td>\n",
       "      <td>scrA_scrB</td>\n",
       "      <td>False</td>\n",
       "      <td>68</td>\n",
       "      <td>85</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19120 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr                  name_config  start_pad  stop_pad   vocab  \\\n",
       "7     chr18     ARID5B::KLF5+GDSD6+1+A_B   61243074  61243208    KLF5   \n",
       "7     chr18     ARID5B::KLF5+GDSD6+1+A_B   61243074  61243208  ARID5B   \n",
       "2397  chr18  ARID5B::KLF5+GDSD6+1+A_scrB   61243074  61243208  ARID5B   \n",
       "2397  chr18  ARID5B::KLF5+GDSD6+1+A_scrB   61243074  61243208    KLF5   \n",
       "4787  chr18  ARID5B::KLF5+GDSD6+1+scrA_B   61243074  61243208  ARID5B   \n",
       "...     ...                          ...        ...       ...     ...   \n",
       "4778   chr7       ZFX::E2F6+SCC+9+A_scrB   44153033  44153167    E2F6   \n",
       "7168   chr7       ZFX::E2F6+SCC+9+scrA_B   44153033  44153167     ZFX   \n",
       "7168   chr7       ZFX::E2F6+SCC+9+scrA_B   44153033  44153167    E2F6   \n",
       "9558   chr7    ZFX::E2F6+SCC+9+scrA_scrB   44153033  44153167    E2F6   \n",
       "9558   chr7    ZFX::E2F6+SCC+9+scrA_scrB   44153033  44153167     ZFX   \n",
       "\n",
       "     vocab_order  vocab_start  vocab_stop              instance     config  \\\n",
       "7              B           62          83  ARID5B::KLF5+GDSD6+1        A_B   \n",
       "7              A           49          62  ARID5B::KLF5+GDSD6+1        A_B   \n",
       "2397           A           49          62  ARID5B::KLF5+GDSD6+1     A_scrB   \n",
       "2397           B           62          83  ARID5B::KLF5+GDSD6+1     A_scrB   \n",
       "4787           A           49          62  ARID5B::KLF5+GDSD6+1     scrA_B   \n",
       "...          ...          ...         ...                   ...        ...   \n",
       "4778           A           47          68       ZFX::E2F6+SCC+9     A_scrB   \n",
       "7168           B           68          85       ZFX::E2F6+SCC+9     scrA_B   \n",
       "7168           A           47          68       ZFX::E2F6+SCC+9     scrA_B   \n",
       "9558           A           47          68       ZFX::E2F6+SCC+9  scrA_scrB   \n",
       "9558           B           68          85       ZFX::E2F6+SCC+9  scrA_scrB   \n",
       "\n",
       "      vocab_present_pred  vocab_loc_start  vocab_loc_stop  \n",
       "7                   True               62              83  \n",
       "7                   True               49              62  \n",
       "2397                True               49              62  \n",
       "2397               False               62              83  \n",
       "4787               False               49              62  \n",
       "...                  ...              ...             ...  \n",
       "4778                True               47              68  \n",
       "7168                True               68              85  \n",
       "7168               False               47              68  \n",
       "9558               False               47              68  \n",
       "9558               False               68              85  \n",
       "\n",
       "[19120 rows x 13 columns]"
      ]
     },
     "execution_count": 397,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances_df = vocab_instances20_seqs_sel[['name_config','chr','start_pad','stop_pad','vocab1','vocab1_start_shift','vocab1_stop_shift']]\n",
    "vocab_instances_df.columns = ['name_config','chr','start_pad','stop_pad','vocab','vocab_start','vocab_stop']\n",
    "vocab_instances_df['vocab_order'] = 'A'\n",
    "\n",
    "vocab_instances_df2 = vocab_instances20_seqs_sel[['name_config','chr','start_pad','stop_pad','vocab2','vocab2_start_shift','vocab2_stop_shift']]\n",
    "vocab_instances_df2.columns = ['name_config','chr','start_pad','stop_pad','vocab','vocab_start','vocab_stop']\n",
    "vocab_instances_df2['vocab_order'] = 'B'\n",
    "vocab_instances_df = pd.concat([vocab_instances_df,vocab_instances_df2],axis=0,sort=True)\n",
    "vocab_instances_df['instance'] = vocab_instances_df.name_config.apply(lambda x: '+'.join(x.split('+')[:-1]))\n",
    "vocab_instances_df['config'] = vocab_instances_df.name_config.str.split('+',expand=True).iloc[:,-1]\n",
    "vocab_instances_df['vocab_present_pred'] = vocab_instances_df.apply(vocab_present_func,axis=1)\n",
    "vocab_instances_df['vocab_loc_start'] = vocab_instances_df.vocab_start# - vocab_instances_df.start_pad\n",
    "vocab_instances_df['vocab_loc_stop'] = vocab_instances_df.vocab_stop #- vocab_instances_df.start_pad\n",
    "vocab_instances_df.sort_values('name_config',inplace=True)\n",
    "vocab_instances_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 398,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2390,)"
      ]
     },
     "execution_count": 398,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_instances_df.instance.unique().shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 7C2: add in motif_scan information per case"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 585,
   "metadata": {},
   "outputs": [],
   "source": [
    "# helper functions\n",
    "def label_config(row):\n",
    "    if row['vocab_present_actual']: \n",
    "        if row['vocab_present_pred']:\n",
    "            return 'TP' # good, Case 1\n",
    "        else:\n",
    "            return 'FP' # noisy, Case 3\n",
    "    else:\n",
    "        if row['vocab_present_pred']:\n",
    "            return 'FN' # not detected (Case 4)\n",
    "        else:\n",
    "            return 'TN' # good, Case 1\n",
    "def get_case_instance(row):\n",
    "    row_counts = row.value_counts()\n",
    "    # case 1: all the configs that should have the vocab have the vocab and those that don't do not\n",
    "    if (row_counts.get('TP',0) +row_counts.get('TN',0) )==4:\n",
    "        return 'case 1'\n",
    "    # case 4: the pvalue threshold was too stringent\n",
    "    elif (row_counts.get('FN',0) +row_counts.get('TN',0) )==4:\n",
    "        return 'case 4'\n",
    "    # case 3: too noisy\n",
    "    elif (row_counts.get('TP',0) +row_counts.get('FP',0) )==4:\n",
    "        return 'case 3'\n",
    "    else:\n",
    "        return 'other'\n",
    "\n",
    "def analyze_motif_scan(motif_scan_file, tf_id_abbr_to_name_dict=tf_id_abbr_to_name_dict):\n",
    "    # read in motif scan file\n",
    "    motif_scan_results = pd.read_csv(motif_scan_file,names=['name_config','pwm_file','pos','strand','score','seq','seq_actual'])\n",
    "    motif = tf_id_abbr_to_name_dict[os.path.basename(motif_scan_file).split('_scan')[0]]\n",
    "    motif_scan_results['motif'] = motif_scan_results.pwm_file.str.split('.pwm',expand=True).iloc[:,0].map(tf_id_abbr_to_name_dict)\n",
    "\n",
    "    # merge in genomic location information\n",
    "    vocab_motif_scan = vocab_instances_df[vocab_instances_df.vocab==motif].merge( motif_scan_results[['name_config','pos','score','motif']], how='left',left_on=['name_config','vocab'], right_on=['name_config','motif'])\n",
    "    vocab_motif_scan.pos.fillna(-1.0, inplace=True)\n",
    "    vocab_motif_scan.score.fillna(0, inplace=True)\n",
    "    vocab_motif_scan.motif.fillna('', inplace=True)\n",
    "    vocab_motif_scan['vocab_present_actual'] = abs(vocab_motif_scan.vocab_loc_start - vocab_motif_scan.pos ) <=3\n",
    "    vocab_motif_scan['off_target']  =  ~( ( vocab_motif_scan.vocab_present_actual) | (vocab_motif_scan.pos<0))\n",
    "\n",
    "    #### case 2: off-target effects\n",
    "    vocab_motif_scan_off_target = vocab_motif_scan[vocab_motif_scan.off_target]\n",
    "    vocab_motif_scan_off_target.sort_values('score',ascending=False,inplace=True)\n",
    "    vocab_motif_scan_off_target.drop_duplicates('name_config',inplace=True)\n",
    "    vocab_motif_scan_off_target.sort_values('name_config',inplace=True)\n",
    "    vocab_motif_scan_off_target['pos_close'] = vocab_motif_scan_off_target.pos - (vocab_motif_scan_off_target.pos%2)\n",
    "    off_target_counts_df = pd.DataFrame(vocab_motif_scan_off_target.groupby('instance').pos_close.value_counts())\n",
    "    off_target_counts_df.columns = ['count_pos']\n",
    "    off_target_counts_df['motif'] = motif\n",
    "    off_target_counts_df['case'] = off_target_counts_df.count_pos.apply(lambda x: 'case 2A' if x==4 else 'case 2B')\n",
    "    off_target_counts_df.reset_index(inplace=True)\n",
    "\n",
    "    ##### Case 1,3,4\n",
    "    # - case 1: good\n",
    "    # - case 3: FP (too noisy, scrambling might not be good), case 4: FN (pval, threshold too low )\n",
    "\n",
    "    # only include the highest scoring \n",
    "    vocab_motif_scan_on_target = vocab_motif_scan[~vocab_motif_scan.off_target]\n",
    "    vocab_motif_scan_on_target.sort_values('score',ascending=False,inplace=True)\n",
    "    vocab_motif_scan_on_target.drop_duplicates('name_config',inplace=True)\n",
    "    vocab_motif_scan_on_target.sort_values('name_config',inplace=True)\n",
    "\n",
    "    # fill in all possible instances in the library\n",
    "    vocab_motif_scan_on_target = vocab_instances_df[vocab_instances_df.vocab==motif].merge(vocab_motif_scan_on_target,how='left',on=list(vocab_instances_df.columns.values))\n",
    "    vocab_motif_scan_on_target.pos.fillna(-1.0, inplace=True)\n",
    "    vocab_motif_scan_on_target.score.fillna(0, inplace=True)\n",
    "    vocab_motif_scan_on_target.motif.fillna('', inplace=True)\n",
    "    vocab_motif_scan_on_target.vocab_present_actual.fillna(False, inplace=True)\n",
    "    vocab_motif_scan_on_target.off_target.fillna(False, inplace=True)\n",
    "    vocab_motif_scan_on_target['eval'] = vocab_motif_scan_on_target.apply(label_config, axis=1)\n",
    "\n",
    "    # create data frame thats has rows (instances) and columns are the eval (TP,FP, etc) and score per config\n",
    "    vocab_motif_scan_on_target_config = vocab_motif_scan_on_target.pivot(index='instance', columns='config', values=['eval','score'])\n",
    "    vocab_motif_scan_on_target_config.columns = ['_'.join(list(x)) for x in vocab_motif_scan_on_target_config.columns.values]\n",
    "    vocab_motif_scan_on_target_config['motif'] = motif\n",
    "    vocab_motif_scan_on_target_config['case'] = vocab_motif_scan_on_target_config.apply(get_case_instance,axis=1)\n",
    "\n",
    "    # dict of case_type: # instances that fall in type\n",
    "    on_target_config_cases = vocab_motif_scan_on_target_config.case.value_counts()\n",
    "    \n",
    "    return motif, off_target_counts_df, vocab_motif_scan_on_target_config, on_target_config_cases\n",
    "\n",
    "def analyze_motif_scan_wrap(motif_scan_files, save_prefix=None):\n",
    "    print('# motif files: ', len(motif_scan_files))\n",
    "    # overall dataframe setup\n",
    "    off_target_counts_df_all = pd.DataFrame()\n",
    "    on_target_cases_df_all = pd.DataFrame()\n",
    "    on_target_config_cases_dict = {}\n",
    "    \n",
    "    #get scan results from file\n",
    "    for idx,motif_scan_file in enumerate(motif_scan_files):\n",
    "        ## DEBUG\n",
    "    #     print(idx,motif_scan_file)\n",
    "        motif, off_target_counts_df, vocab_motif_scan_on_target_config, on_target_config_cases = analyze_motif_scan(motif_scan_file)\n",
    "        # ADD TO RUNNING LIST (ACROSS MOTIFS)\n",
    "        off_target_counts_df_all = pd.concat([off_target_counts_df_all, off_target_counts_df],axis=0,sort=True)\n",
    "        on_target_cases_df_all = pd.concat([on_target_cases_df_all, vocab_motif_scan_on_target_config],axis=0,sort=True)\n",
    "        on_target_config_cases_dict[motif] = on_target_config_cases\n",
    "\n",
    "    # convert dict to dataframe (row: motif, columns case type)\n",
    "    on_target_config_cases_df = pd.DataFrame.from_dict(on_target_config_cases_dict,orient='index').fillna(0)\n",
    "\n",
    "    ## TODO: add in save functionality\n",
    "    if save_prefix is not None:\n",
    "        off_target_counts_df_all.to_csv(save_prefix+'off_target_counts_df_all.csv')\n",
    "        on_target_cases_df_all.to_csv(save_prefix+'motif_scan_'+'on_target_cases_df_all.csv')\n",
    "        on_target_config_cases_df.to_csv(save_prefix+ 'motif_scan_'+'on_target_config_cases_df.csv')\n",
    "        print('saved', save_prefix+'*.csv')\n",
    "    return off_target_counts_df_all, on_target_cases_df_all, on_target_config_cases_df\n",
    "\n",
    "def refine_case3(row):\n",
    "    if row['case']!='case 3':\n",
    "#         print(row['case'])\n",
    "        return row['case']\n",
    "    else:\n",
    "        # only remodify case 3:\n",
    "        if row['score_A_B']> min(row['score_A_scrB'],row['score_scrA_B'] ):\n",
    "    #         print('here')\n",
    "            return 'case 3A' # this is the good case where scrambling reduces the score of the motif\n",
    "        else:\n",
    "            return 'case 3B'\n",
    "        # case 3, where the TP are have higher score than where the true negatives are\n",
    "\n",
    "def get_offtarget_num(off_target_counts_df= off_target_counts_df_all_5, vocab_instances_df=vocab_instances20_seqs_sel):\n",
    "    off_target_res_case_2A = off_target_counts_df[off_target_counts_df.case=='case 2A']\n",
    "    print(off_target_res_case_2A.shape)\n",
    "    off_target_res_case_2A['off_target_pos_str'] = off_target_res_case_2A.pos_close.map(int).map(str)\n",
    "    off_target_res_case_2A = off_target_res_case_2A.groupby(['instance','motif']).agg({'off_target_pos_str':'|'.join}).reset_index()\n",
    "    off_target_res_case_2A.columns = ['off_target_name','off_target_motif','off_target_pos_str']\n",
    "    print(off_target_res_case_2A.shape) # groupbed by instance\n",
    "    vocab_instances20_seqs_sel_offtarget = vocab_instances_df.merge(off_target_res_case_2A, how='left',left_on='name',right_on='off_target_name')\n",
    "    vocab_instances20_seqs_sel_offtarget.off_target_name.fillna('',inplace=True)\n",
    "    vocab_instances20_seqs_sel_offtarget.off_target_motif.fillna('',inplace=True)\n",
    "    vocab_instances20_seqs_sel_offtarget.off_target_pos_str.fillna('',inplace=True)\n",
    "    vocab_instances20_seqs_sel_offtarget = pd.concat([vocab_instances20_seqs_sel_offtarget,vocab_instances20_seqs_sel_offtarget.apply(get_vocab_loc,axis=1)],axis=1)\n",
    "    \n",
    "#     return vocab_instances20_seqs_sel_offtarget\n",
    "    vocab_instances20_seqs_sel_offtarget = vocab_instances20_seqs_sel_offtarget.groupby(['vocab_pair', 'tot_dist', 'btn_dist', 'chr', 'start', 'stop',\n",
    "           'start_pad', 'stop_pad', 'vocab1', 'vocab1_start', 'vocab1_stop',\n",
    "           'vocab2', 'vocab2_start', 'vocab2_stop', 'tissue', 'loop_type', 'name', 'check_rs', 'scr_pass_rs', \n",
    "           'vocab_tissue_count', 'seq_config', 'seq','name_config', 'off_target_name','vocab1_start_shift', 'vocab1_stop_shift',\n",
    "           'vocab2_start_shift', 'vocab2_stop_shift']).agg({'off_target_motif':'::'.join,'off_target_pos_str':'::'.join}).reset_index()\n",
    "\n",
    "    vocab_instances20_seqs_sel_offtarget = vocab_instances20_seqs_sel_offtarget.groupby(['vocab_pair', 'tot_dist', 'btn_dist', 'chr', 'start', 'stop',\n",
    "           'start_pad', 'stop_pad', 'vocab1', 'vocab1_start', 'vocab1_stop',\n",
    "           'vocab2', 'vocab2_start', 'vocab2_stop', 'tissue', 'loop_type', 'name',\n",
    "           'check_rs', 'scr_pass_rs', 'vocab_tissue_count',  'vocab1_start_shift',\n",
    "           'vocab1_stop_shift', 'vocab2_start_shift', 'vocab2_stop_shift','off_target_name',\n",
    "           'off_target_motif', 'off_target_pos_str']).agg({'seq_config':list,'seq':list, 'name_config':list}).reset_index()\n",
    "    # \n",
    "\n",
    "\n",
    "    off_target_num_df = pd.DataFrame(vocab_instances20_seqs_sel_offtarget.set_index('name').off_target_pos_str.apply(off_target_pos_str_find_num)).reset_index()\n",
    "    off_target_num_df['instance'] = off_target_num_df.name.apply(lambda x: '+'.join(x.split('+')[:-1]))\n",
    "    off_target_num_df_counts = off_target_num_df[off_target_num_df.off_target_pos_str==0].instance.value_counts().value_counts().sort_index()\n",
    "    print(off_target_num_df_counts/off_target_num_df_counts.sum())\n",
    "    return off_target_num_df[off_target_num_df.off_target_pos_str==0].instance.value_counts().describe()\n",
    "\n",
    "\n",
    "\n",
    "def off_target_pos_str_find_num(pos_str):\n",
    "    pos_arr_per_vocab_word = pos_str.split('::')\n",
    "    if len(pos_arr_per_vocab_word)>0:\n",
    "        arr = []\n",
    "        for pos in pos_arr_per_vocab_word:\n",
    "            for x in pos.split(':'):\n",
    "                try:\n",
    "                    arr.append(int(x))\n",
    "                except:\n",
    "                    continue\n",
    "        return len(set(arr))\n",
    "    return 0\n",
    "\n",
    "def on_target_new_cases(row):\n",
    "    motif_arr = row['motif'].split('|')\n",
    "    case_mod_arr = row['case_mod'].split('|')\n",
    "    num_motifs = len(set(motif_arr))\n",
    "    if num_motifs>2:\n",
    "        return '|'.join(sorted(set(motif_arr))), 'more than 2 motifs'\n",
    "    elif num_motifs<2:\n",
    "        return '|'.join(sorted(set(motif_arr))), 'less than 2 motifs'    \n",
    "    else:\n",
    "        motif_to_case = {}\n",
    "        for motif, case in zip(motif_arr, case_mod_arr):\n",
    "            if motif in motif_to_case:\n",
    "                if (motif_to_case[motif]!='case 3A') & (motif_to_case[motif]!='case 1'):\n",
    "                    motif_to_case[motif]=case\n",
    "            else:\n",
    "                motif_to_case[motif]=case\n",
    "        motif_new = '|'.join(list(motif_to_case.keys()))\n",
    "        case_new = '|'.join(list(motif_to_case.values()))\n",
    "        return motif_new, case_new\n",
    "    \n",
    "def on_target_new_case_good(case_new):\n",
    "    if '|' not in case_new:\n",
    "        return False # not two motifs\n",
    "    else:\n",
    "        case_status = True\n",
    "        for case in case_new.split('|'):\n",
    "            if  (case!='case 3A') & (case!='case 1') :\n",
    "                case_status = False\n",
    "        return case_status\n",
    "    \n",
    "def get_on_target_good_instance_num(on_target_cases_df):\n",
    "    on_target_counts = on_target_cases_df.reset_index()[['instance','motif','case_mod']].drop_duplicates().groupby('instance').agg({\n",
    "    'motif':'|'.join, 'case_mod':'|'.join\n",
    "    }).reset_index()\n",
    "    on_target_counts['name'] = on_target_counts.instance.apply(lambda x: '+'.join(x.split('+')[:-1]))\n",
    "    on_target_counts = pd.concat([on_target_counts, on_target_counts.apply(on_target_new_cases,axis=1).apply(lambda s: pd.Series({'motif_new':s[0], 'case_new':s[1]}))],axis=1)\n",
    "    on_target_counts['case_good_bool'] = on_target_counts.case_new.apply(on_target_new_case_good)\n",
    "    return on_target_counts[on_target_counts.case_good_bool==True].name.value_counts().describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### version 1: homotypic scrambling with pval <0.05"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 684,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# motif files:  158\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:43: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:44: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:45: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:46: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:59: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:60: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:61: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "saved ../data/processed/fig4_modelling/vocab_sum_final_post_scan/motif_scan4_*.csv\n",
      "CPU times: user 1min 58s, sys: 3.53 s, total: 2min 2s\n",
      "Wall time: 2min 2s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "case 3    2153.0\n",
       "case 1     986.0\n",
       "other      898.0\n",
       "case 4     743.0\n",
       "dtype: float64"
      ]
     },
     "execution_count": 684,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "# takes ~ 15s\n",
    "\n",
    "motif_scan_files4 = sorted(glob.glob(os.path.join(save_dir, 'motif_scan4','*csv')))\n",
    "off_target_counts_df_all_4, on_target_cases_df_all_4, on_target_config_cases_df_4 = analyze_motif_scan_wrap(motif_scan_files4, save_prefix=os.path.join(save_dir, 'motif_scan4_'))\n",
    "on_target_config_cases_df_4.sum(axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 685,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "case 3    2852\n",
       "case 1    1324\n",
       "other     1043\n",
       "case 4     861\n",
       "Name: case, dtype: int64"
      ]
     },
     "execution_count": 685,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "on_target_cases_df_all_4.case.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 686,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "case 3A    1905\n",
       "case 1     1324\n",
       "other      1043\n",
       "case 3B     947\n",
       "case 4      861\n",
       "Name: case_mod, dtype: int64"
      ]
     },
     "execution_count": 686,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "on_target_cases_df_all_4['case_mod'] = on_target_cases_df_all_4.apply(refine_case3, axis=1)\n",
    "on_target_cases_df_all_4.case_mod.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 687,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "FN|FN|FP|FP    467\n",
       "FN|FP|FN|FP    359\n",
       "TP|FP|TP|TN     55\n",
       "TP|TN|TP|FP     52\n",
       "TP|FP|FN|FP     17\n",
       "FN|TN|TP|TN     15\n",
       "FN|TN|FN|FP     14\n",
       "FN|FP|FN|TN     13\n",
       "TP|TN|FN|TN      9\n",
       "FN|FP|TP|FP      8\n",
       "TP|TP|FP|TN      6\n",
       "FN|TN|TP|FP      6\n",
       "TP|TP|TN|FP      5\n",
       "TP|FN|FP|FP      4\n",
       "FN|FN|TN|FP      3\n",
       "TP|FP|FN|TN      3\n",
       "FN|TP|FP|FP      2\n",
       "FN|TP|TN|TN      2\n",
       "TP|FN|FP|TN      1\n",
       "FN|FN|FP|TN      1\n",
       "FN|TP|FP|TN      1\n",
       "dtype: int64"
      ]
     },
     "execution_count": 687,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "on_target_cases_df_all_4[on_target_cases_df_all_4.case=='other'][['eval_A_B',\n",
    "                                                                  'eval_A_scrB',\t\n",
    "                                                                  'eval_scrA_B',\t\n",
    "                                                                  'eval_scrA_scrB']].apply('|'.join,axis=1).value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 688,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    208.000000\n",
       "mean       3.495192\n",
       "std        2.206704\n",
       "min        1.000000\n",
       "25%        2.000000\n",
       "50%        3.000000\n",
       "75%        4.000000\n",
       "max       10.000000\n",
       "Name: name, dtype: float64"
      ]
     },
     "execution_count": 688,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_on_target_good_instance_num(on_target_cases_df_all_4)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 689,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4780.0"
      ]
     },
     "execution_count": 689,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "on_target_config_cases_df_4.sum(axis=0).sum() #double the number of instances b/c of the two vocab words per instance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 690,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4777, 2)"
      ]
     },
     "execution_count": 690,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "off_target_counts_df_all_4.reset_index()[['instance','motif']].drop_duplicates().shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 691,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(813, 2)"
      ]
     },
     "execution_count": 691,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "off_target_counts_df_all_4[off_target_counts_df_all_4.case=='case 2A'][['instance','motif']].drop_duplicates().shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 692,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5141, 3)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "case 2B    4328\n",
       "case 2A     813\n",
       "Name: case, dtype: int64"
      ]
     },
     "execution_count": 692,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(off_target_counts_df_all_4[['instance','motif','case']].drop_duplicates().shape)\n",
    "off_target_counts_df_all_4[['instance','motif','case']].drop_duplicates()['case'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 693,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(888, 5)\n",
      "(813, 3)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:127: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1     0.004184\n",
      "2     0.008368\n",
      "3     0.033473\n",
      "4     0.050209\n",
      "5     0.104603\n",
      "6     0.150628\n",
      "7     0.163180\n",
      "8     0.209205\n",
      "9     0.175732\n",
      "10    0.100418\n",
      "Name: instance, dtype: float64\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "count    239.000000\n",
       "mean       7.150628\n",
       "std        1.930056\n",
       "min        1.000000\n",
       "25%        6.000000\n",
       "50%        7.000000\n",
       "75%        9.000000\n",
       "max       10.000000\n",
       "Name: instance, dtype: float64"
      ]
     },
     "execution_count": 693,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_offtarget_num(off_target_counts_df_all_4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### version 2: homotypic scrambling with pval <0.001"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 623,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# motif files:  158\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:43: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:44: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:45: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:46: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:59: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:60: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:61: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "saved ../data/processed/fig4_modelling/vocab_sum_final_post_scan/motif_scan5_*.csv\n",
      "CPU times: user 14.1 s, sys: 147 ms, total: 14.3 s\n",
      "Wall time: 14.3 s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "case 4    3559.0\n",
       "case 1     817.0\n",
       "other      277.0\n",
       "case 3     127.0\n",
       "dtype: float64"
      ]
     },
     "execution_count": 623,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "# takes ~ 15s\n",
    "motif_scan_files5 = sorted(glob.glob(os.path.join(save_dir, 'motif_scan5','*csv')))\n",
    "off_target_counts_df_all_5, on_target_cases_df_all_5, on_target_config_cases_df_5 = analyze_motif_scan_wrap(motif_scan_files5, save_prefix=os.path.join(save_dir, 'motif_scan5_'))\n",
    "on_target_config_cases_df_5.sum(axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 624,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "case 4    4413\n",
       "case 1    1152\n",
       "other      333\n",
       "case 3     182\n",
       "Name: case, dtype: int64"
      ]
     },
     "execution_count": 624,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "on_target_cases_df_all_5.case.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 625,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "case 4     4413\n",
       "case 1     1152\n",
       "other       333\n",
       "case 3A     116\n",
       "case 3B      66\n",
       "Name: case_mod, dtype: int64"
      ]
     },
     "execution_count": 625,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "on_target_cases_df_all_5['case_mod'] = on_target_cases_df_all_5.apply(refine_case3, axis=1)\n",
    "on_target_cases_df_all_5.case_mod.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 683,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    74.000000\n",
       "mean      1.905405\n",
       "std       1.206931\n",
       "min       1.000000\n",
       "25%       1.000000\n",
       "50%       2.000000\n",
       "75%       2.000000\n",
       "max       7.000000\n",
       "Name: name, dtype: float64"
      ]
     },
     "execution_count": 683,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_on_target_good_instance_num(on_target_cases_df_all_5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 626,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4780.0"
      ]
     },
     "execution_count": 626,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "on_target_config_cases_df_5.sum(axis=0).sum() #double the number of instances b/c of the two vocab words per instance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 627,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3871, 2)"
      ]
     },
     "execution_count": 627,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "off_target_counts_df_all_5.reset_index()[['instance','motif']].drop_duplicates().shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 628,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3931, 3)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "case 2B    3687\n",
       "case 2A     244\n",
       "Name: case, dtype: int64"
      ]
     },
     "execution_count": 628,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(off_target_counts_df_all_5[['instance','motif','case']].drop_duplicates().shape)\n",
    "off_target_counts_df_all_5[['instance','motif','case']].drop_duplicates()['case'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 654,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(258, 5)\n",
      "(244, 3)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:127: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4     0.004184\n",
      "6     0.008368\n",
      "7     0.054393\n",
      "8     0.179916\n",
      "9     0.347280\n",
      "10    0.405858\n",
      "Name: instance, dtype: float64\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "count    239.000000\n",
       "mean       9.071130\n",
       "std        0.991118\n",
       "min        4.000000\n",
       "25%        9.000000\n",
       "50%        9.000000\n",
       "75%       10.000000\n",
       "max       10.000000\n",
       "Name: instance, dtype: float64"
      ]
     },
     "execution_count": 654,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_offtarget_num(off_target_counts_df_all_5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 681,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
