{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 9E_vocab_compare_gene\n",
    "\n",
    "6/17/21\n",
    "\n",
    "Can you see if there is any target gene overlap between any of the 'AND' vocabs for MM vs MC and for SCC vs KCD6? Maybe can scrounge up a  v4C vignette showing how cancer vs normal use different TFs to regulate the same genes? Engh?\n",
    "\n",
    "And then if there ARE any overlapping target genes, would also need the genomic locations of the vocabs\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pybedtools\n",
    "import glob, os\n",
    "import pandas as pd\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>713908</td>\n",
       "      <td>714326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>762050</td>\n",
       "      <td>763512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>781110</td>\n",
       "      <td>781315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>793350</td>\n",
       "      <td>794448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>840069</td>\n",
       "      <td>840282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>67456</td>\n",
       "      <td>chrY</td>\n",
       "      <td>21154493</td>\n",
       "      <td>21155089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>67457</td>\n",
       "      <td>chrY</td>\n",
       "      <td>23153060</td>\n",
       "      <td>23153234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>67458</td>\n",
       "      <td>chrY</td>\n",
       "      <td>26416783</td>\n",
       "      <td>26417235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>67459</td>\n",
       "      <td>chrY</td>\n",
       "      <td>27545204</td>\n",
       "      <td>27545651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>67460</td>\n",
       "      <td>chrY</td>\n",
       "      <td>59029537</td>\n",
       "      <td>59030204</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>67461 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom     start       end\n",
       "0      chr1    713908    714326\n",
       "1      chr1    762050    763512\n",
       "2      chr1    781110    781315\n",
       "3      chr1    793350    794448\n",
       "4      chr1    840069    840282\n",
       "...     ...       ...       ...\n",
       "67456  chrY  21154493  21155089\n",
       "67457  chrY  23153060  23153234\n",
       "67458  chrY  26416783  26417235\n",
       "67459  chrY  27545204  27545651\n",
       "67460  chrY  59029537  59030204\n",
       "\n",
       "[67461 rows x 3 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bed_file = '../data/interim/merged/atac/Colon_merged.bed'\n",
    "\n",
    "# pandas\n",
    "# pd.read_table(bed_file, names = ['chr','start','stop','name','chrx','startx','stopy','rowname','distance'])\n",
    "\n",
    "pybedtools.BedTool(bed_file).merge().sort().to_dataframe()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_dir = '../data/processed/fig7_vocab_go/'\n",
    "vocab_geneset_dir = '../data/processed/fig7_vocab_go/vocab_geneset_all_tissue'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# tfs \n",
    "tf_annon_df = pd.read_csv('../data/external/HOCOMOCOv11_annotation.csv',index_col=0)\n",
    "tf_annon_df['id_trim'] = tf_annon_df['id'] + '.pwm.trim'\n",
    "tf_name_to_id_dict = pd.Series(tf_annon_df.id_trim.values, index=tf_annon_df.tf.values).to_dict()\n",
    "tf_id_to_name_dict = pd.Series(tf_annon_df.tf.values, index=tf_annon_df.id_trim.values).to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. cancer+normal using the same vocab for the same target gene?\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(35, 9)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Index(['vocab', 'tissue', 'mpra_tissue', 'class', 'instance', 'count',\n",
       "       'consensus_class', 'tf1', 'tf2'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get mpra vocabs + configs\n",
    "new_mpra_df = pd.read_csv('../data/processed/mpra/results_df_piv_match_count.csv',index_col=0)\n",
    "# filter for skin stuff\n",
    "new_mpra_df = new_mpra_df[new_mpra_df.tissue.isin(['GDSD6','SCC','Melanocytes','MM'])]\n",
    "# filter for AND VOCABs\n",
    "new_mpra_df = new_mpra_df[new_mpra_df.consensus_class=='AND']\n",
    "print(new_mpra_df.shape)\n",
    "\n",
    "# new_mpra_df = new_mpra_df[new_mpra_df.tissue.isin(['MM','SCC'])][['vocab','tissue']].rename(columns={\"vocab\": \"vocab_pair\"}).drop_duplicates()\n",
    "new_mpra_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ARID5B::LEF1 MM\n",
      "num normal genes 13\n",
      "num cancer genes 464\n",
      "num intersect genes 10\n",
      "1\n",
      "ARNT::SP1 SCC\n",
      "num normal genes 91\n",
      "num cancer genes 828\n",
      "num intersect genes 71\n",
      "5\n",
      "BACH1::KLF5 GDSD6\n",
      "num normal genes 100\n",
      "num cancer genes 988\n",
      "num intersect genes 80\n",
      "8\n",
      "E2F6::SP1 SCC\n",
      "num normal genes 274\n",
      "num cancer genes 1385\n",
      "num intersect genes 240\n",
      "18\n",
      "EGR1::SP3 SCC\n",
      "num normal genes 327\n",
      "num cancer genes 1432\n",
      "num intersect genes 281\n",
      "25\n",
      "EGR2::FOSL2 GDSD6\n",
      "num normal genes 255\n",
      "num cancer genes 1164\n",
      "num intersect genes 202\n",
      "26\n",
      "EGR2::NFE2L1 MM\n",
      "num normal genes 0\n",
      "num cancer genes 610\n",
      "num intersect genes 0\n",
      "31\n",
      "EGR2::RARG GDSD6\n",
      "num normal genes 196\n",
      "num cancer genes 1103\n",
      "num intersect genes 155\n",
      "32\n",
      "EGR2::RORA GDSD6\n",
      "num normal genes 205\n",
      "num cancer genes 0\n",
      "num intersect genes 0\n",
      "33\n",
      "EN2::STAT1 MM\n",
      "num normal genes 3\n",
      "num cancer genes 459\n",
      "num intersect genes 2\n",
      "43\n",
      "ETV2::PRRX1 Melanocytes\n",
      "num normal genes 45\n",
      "num cancer genes 553\n",
      "num intersect genes 33\n",
      "47\n",
      "FOXD3::PRRX1 Melanocytes\n",
      "num normal genes 21\n",
      "num cancer genes 366\n",
      "num intersect genes 14\n",
      "54\n",
      "HBP1::STAT1 MM\n",
      "num normal genes 12\n",
      "num cancer genes 467\n",
      "num intersect genes 6\n",
      "66\n",
      "HMGA1::KLF5 GDSD6\n",
      "num normal genes 284\n",
      "num cancer genes 1444\n",
      "num intersect genes 230\n",
      "67\n",
      "JUNB::RORA GDSD6\n",
      "num normal genes 120\n",
      "num cancer genes 491\n",
      "num intersect genes 66\n",
      "103\n",
      "KLF15::ZNF589 Melanocytes\n",
      "num normal genes 101\n",
      "num cancer genes 0\n",
      "num intersect genes 0\n",
      "117\n",
      "KLF16::TFAP2C SCC\n",
      "num normal genes 191\n",
      "num cancer genes 1220\n",
      "num intersect genes 156\n",
      "120\n",
      "KLF4::KLF5 SCC\n",
      "num normal genes 432\n",
      "num cancer genes 1509\n",
      "num intersect genes 355\n",
      "121\n",
      "KLF5::MAFG GDSD6\n",
      "num normal genes 159\n",
      "num cancer genes 1131\n",
      "num intersect genes 127\n",
      "128\n",
      "KLF5::MAFK GDSD6\n",
      "num normal genes 144\n",
      "num cancer genes 1094\n",
      "num intersect genes 117\n",
      "129\n",
      "KLF5::NFYB Melanocytes\n",
      "num normal genes 29\n",
      "num cancer genes 173\n",
      "num intersect genes 21\n",
      "131\n",
      "MAF::RORA GDSD6\n",
      "num normal genes 109\n",
      "num cancer genes 0\n",
      "num intersect genes 0\n",
      "149\n",
      "MAZ::SP1 SCC\n",
      "num normal genes 473\n",
      "num cancer genes 1583\n",
      "num intersect genes 404\n",
      "153\n",
      "MITF::ZNF589 Melanocytes\n",
      "num normal genes 14\n",
      "num cancer genes 0\n",
      "num intersect genes 0\n",
      "156\n",
      "NFATC4::STAT1 MM\n",
      "num normal genes 26\n",
      "num cancer genes 615\n",
      "num intersect genes 17\n",
      "165\n",
      "NFE2L1::NR3C1 MM\n",
      "num normal genes 7\n",
      "num cancer genes 522\n",
      "num intersect genes 4\n",
      "167\n",
      "NFE2L1::TBX15 MM\n",
      "num normal genes 0\n",
      "num cancer genes 580\n",
      "num intersect genes 0\n",
      "172\n",
      "NFE2L1::ZBTB49 MM\n",
      "num normal genes 15\n",
      "num cancer genes 599\n",
      "num intersect genes 11\n",
      "174\n",
      "PRDM1::RORA GDSD6\n",
      "num normal genes 91\n",
      "num cancer genes 405\n",
      "num intersect genes 43\n",
      "185\n",
      "SOX10::SOX13 MM\n",
      "num normal genes 18\n",
      "num cancer genes 627\n",
      "num intersect genes 14\n",
      "203\n",
      "SOX10::SP4 MM\n",
      "num normal genes 0\n",
      "num cancer genes 542\n",
      "num intersect genes 0\n",
      "204\n",
      "SP1::TFDP1 SCC\n",
      "num normal genes 179\n",
      "num cancer genes 1189\n",
      "num intersect genes 158\n",
      "214\n",
      "SP1::ZBTB14 SCC\n",
      "num normal genes 109\n",
      "num cancer genes 1018\n",
      "num intersect genes 95\n",
      "217\n",
      "SP1::ZBTB17 SCC\n",
      "num normal genes 423\n",
      "num cancer genes 1539\n",
      "num intersect genes 359\n",
      "218\n",
      "SP1::ZNF770 SCC\n",
      "num normal genes 384\n",
      "num cancer genes 1523\n",
      "num intersect genes 326\n",
      "220\n"
     ]
    }
   ],
   "source": [
    "new_mpra_df['tissues_to_check'] = ''\n",
    "new_mpra_df['num_overlap_genes'] = 0\n",
    "new_mpra_df['num_normal_genes'] = 0\n",
    "new_mpra_df['num_cancer_genes'] = 0\n",
    "new_mpra_df['overlap_genes'] = ''\n",
    "new_mpra_df['normal_genes'] = ''\n",
    "new_mpra_df['cancer_genes'] = ''\n",
    "# new_mpra_df['Melanocytes_genes'] = ''\n",
    "# new_mpra_df['MM_genes'] = ''\n",
    "for idx, row in new_mpra_df.iterrows():\n",
    "    print(row['vocab'],row.tissue)\n",
    "    tf1,tf2 = row['vocab'].split('::')\n",
    "    \n",
    "    # get tissues to check\n",
    "    if row.tissue in ['GDSD6',  'SCC']:\n",
    "        tissues_to_check = ['GDSD6',  'SCC']\n",
    "    else:\n",
    "        tissues_to_check = ['Melanocytes','MM']\n",
    "    new_mpra_df.at[idx, 'tissues_to_check'] = '|'.join(tissues_to_check )\n",
    "    \n",
    "    genesets = [set(),set()]\n",
    "    for i_mpra,mpra_tissue in enumerate(tissues_to_check):\n",
    "        geneset_file = os.path.join(vocab_geneset_dir,row['tissue']+'_'+tf1 + '-' + tf2+'_'+mpra_tissue+'.txt' )\n",
    "        if os.path.exists(geneset_file):\n",
    "            genesets[i_mpra] = set(pd.read_csv(geneset_file,header=None).iloc[:,0].values)\n",
    "            \n",
    "        \n",
    "    print('num normal genes',len(genesets[0]))\n",
    "    print('num cancer genes',len(genesets[1]))\n",
    "\n",
    "    intersect_genes = genesets[0].intersection(genesets[1])\n",
    "#     if len(intersect_genes)==0:\n",
    "#         print('no overlapping genes')\n",
    "#         continue\n",
    "    print('num intersect genes',len(intersect_genes))\n",
    "\n",
    "    new_mpra_df.at[idx, 'num_overlap_genes'] = len(intersect_genes)\n",
    "    new_mpra_df.at[idx, 'num_normal_genes'] = len(genesets[0])\n",
    "    new_mpra_df.at[idx, 'num_cancer_genes'] = len(genesets[1])\n",
    "    new_mpra_df.at[idx, 'overlap_genes'] = '|'.join(sorted(intersect_genes))\n",
    "    new_mpra_df.at[idx, 'normal_genes'] = '|'.join(sorted(genesets[0]))\n",
    "    new_mpra_df.at[idx, 'cancer_genes'] = '|'.join(sorted(genesets[1]))\n",
    "    print(idx)\n",
    "#     new_mpra_df.at[idx, tissues_to_check[0]+'_genes'] = '|'.join(sorted(genesets[0]))\n",
    "#     new_mpra_df.at[idx, tissues_to_check[1]+'_genes'] = '|'.join(sorted(genesets[1]))\n",
    "   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>vocab</th>\n",
       "      <th>tissue</th>\n",
       "      <th>mpra_tissue</th>\n",
       "      <th>class</th>\n",
       "      <th>instance</th>\n",
       "      <th>count</th>\n",
       "      <th>consensus_class</th>\n",
       "      <th>tf1</th>\n",
       "      <th>tf2</th>\n",
       "      <th>tissues_to_check</th>\n",
       "      <th>num_overlap_genes</th>\n",
       "      <th>num_normal_genes</th>\n",
       "      <th>num_cancer_genes</th>\n",
       "      <th>overlap_genes</th>\n",
       "      <th>normal_genes</th>\n",
       "      <th>cancer_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>ARID5B::LEF1</td>\n",
       "      <td>MM</td>\n",
       "      <td>COLO</td>\n",
       "      <td>other|LEADING_B|other|OR|OR|AND|other|other|AN...</td>\n",
       "      <td>1|10|2|3|4|5|6|7|8|9</td>\n",
       "      <td>10</td>\n",
       "      <td>AND</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>LEF1</td>\n",
       "      <td>Melanocytes|MM</td>\n",
       "      <td>10</td>\n",
       "      <td>13</td>\n",
       "      <td>464</td>\n",
       "      <td>CLP1|DUSP7|FTL|MRPL10|MRPL34|N4BP1|OCEL1|PKD1|...</td>\n",
       "      <td>CLP1|CORO1C|DUSP7|FTL|MRPL10|MRPL34|N4BP1|NFE2...</td>\n",
       "      <td>ABCC5|ABCF3|ABRACL|ABTB2|ACADS|ACOT9|ACTG1|ACT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>SCC</td>\n",
       "      <td>A431</td>\n",
       "      <td>AND|AND|OR|LEADING_A|AND|LEADING_A|LEADING_B|o...</td>\n",
       "      <td>1|10|2|3|4|5|6|7|8|9</td>\n",
       "      <td>10</td>\n",
       "      <td>AND</td>\n",
       "      <td>ARNT</td>\n",
       "      <td>SP1</td>\n",
       "      <td>GDSD6|SCC</td>\n",
       "      <td>71</td>\n",
       "      <td>91</td>\n",
       "      <td>828</td>\n",
       "      <td>AHNAK|AP1S1|AP3D1|ARHGEF3|ARID3B|CA11|CCDC51|C...</td>\n",
       "      <td>AHNAK|ALOX15B|AP1S1|AP3D1|ARHGEF3|ARID3B|CA11|...</td>\n",
       "      <td>AAMP|AARS2|ABCA7|ABCB6|ABCB8|ABCF3|ABHD10|ABTB...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>BACH1::KLF5</td>\n",
       "      <td>GDSD6</td>\n",
       "      <td>KCD6</td>\n",
       "      <td>AND|LEADING_B|OR|other|OR|other|other|LEADING_...</td>\n",
       "      <td>1|10|2|3|4|5|6|7|8|9</td>\n",
       "      <td>10</td>\n",
       "      <td>AND</td>\n",
       "      <td>BACH1</td>\n",
       "      <td>KLF5</td>\n",
       "      <td>GDSD6|SCC</td>\n",
       "      <td>80</td>\n",
       "      <td>100</td>\n",
       "      <td>988</td>\n",
       "      <td>ABCB8|AHNAK2|ANKRD11|AP3D1|BCL2L1|CALML5|CBY1|...</td>\n",
       "      <td>ABCB8|AHNAK2|ALOX15B|ANKRD11|AP3D1|ATP12A|BACH...</td>\n",
       "      <td>AARS2|ABCA7|ABCB8|ABCC5|ABCF3|ABHD10|ACADVL|AC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18</td>\n",
       "      <td>E2F6::SP1</td>\n",
       "      <td>SCC</td>\n",
       "      <td>A431</td>\n",
       "      <td>other|AND|other|XOR|OR|AND|other|other|other|OR</td>\n",
       "      <td>1|10|2|3|4|5|6|7|8|9</td>\n",
       "      <td>10</td>\n",
       "      <td>AND</td>\n",
       "      <td>E2F6</td>\n",
       "      <td>SP1</td>\n",
       "      <td>GDSD6|SCC</td>\n",
       "      <td>240</td>\n",
       "      <td>274</td>\n",
       "      <td>1385</td>\n",
       "      <td>ABCB6|ACTG1|ACTN4|ADAM15|AHNAK2|ALDH16A1|AP2S1...</td>\n",
       "      <td>ABCB6|ACTG1|ACTN4|ADAM15|AHNAK2|AKR1B10|ALDH16...</td>\n",
       "      <td>AARS2|ABCA7|ABCB6|ABCB8|ABCC5|ABCF3|ABHD10|ABL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>25</td>\n",
       "      <td>EGR1::SP3</td>\n",
       "      <td>SCC</td>\n",
       "      <td>A431</td>\n",
       "      <td>other|other|other|AND|LEADING_A|AND|other|OR|o...</td>\n",
       "      <td>1|10|2|3|4|5|6|7|8|9</td>\n",
       "      <td>10</td>\n",
       "      <td>AND</td>\n",
       "      <td>EGR1</td>\n",
       "      <td>SP3</td>\n",
       "      <td>GDSD6|SCC</td>\n",
       "      <td>281</td>\n",
       "      <td>327</td>\n",
       "      <td>1432</td>\n",
       "      <td>AAMP|ABCA7|ABCB6|ABCB8|ABCF3|ACTG1|ACTN4|ADAM1...</td>\n",
       "      <td>AAMP|ABCA7|ABCB6|ABCB8|ABCF3|ACTG1|ACTN4|ADAM1...</td>\n",
       "      <td>AAMP|AARS2|ABCA7|ABCB6|ABCB8|ABCC5|ABCF3|ABHD1...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           vocab tissue mpra_tissue  \\\n",
       "1   ARID5B::LEF1     MM        COLO   \n",
       "5      ARNT::SP1    SCC        A431   \n",
       "8    BACH1::KLF5  GDSD6        KCD6   \n",
       "18     E2F6::SP1    SCC        A431   \n",
       "25     EGR1::SP3    SCC        A431   \n",
       "\n",
       "                                                class              instance  \\\n",
       "1   other|LEADING_B|other|OR|OR|AND|other|other|AN...  1|10|2|3|4|5|6|7|8|9   \n",
       "5   AND|AND|OR|LEADING_A|AND|LEADING_A|LEADING_B|o...  1|10|2|3|4|5|6|7|8|9   \n",
       "8   AND|LEADING_B|OR|other|OR|other|other|LEADING_...  1|10|2|3|4|5|6|7|8|9   \n",
       "18    other|AND|other|XOR|OR|AND|other|other|other|OR  1|10|2|3|4|5|6|7|8|9   \n",
       "25  other|other|other|AND|LEADING_A|AND|other|OR|o...  1|10|2|3|4|5|6|7|8|9   \n",
       "\n",
       "    count consensus_class     tf1   tf2 tissues_to_check  num_overlap_genes  \\\n",
       "1      10             AND  ARID5B  LEF1   Melanocytes|MM                 10   \n",
       "5      10             AND    ARNT   SP1        GDSD6|SCC                 71   \n",
       "8      10             AND   BACH1  KLF5        GDSD6|SCC                 80   \n",
       "18     10             AND    E2F6   SP1        GDSD6|SCC                240   \n",
       "25     10             AND    EGR1   SP3        GDSD6|SCC                281   \n",
       "\n",
       "    num_normal_genes  num_cancer_genes  \\\n",
       "1                 13               464   \n",
       "5                 91               828   \n",
       "8                100               988   \n",
       "18               274              1385   \n",
       "25               327              1432   \n",
       "\n",
       "                                        overlap_genes  \\\n",
       "1   CLP1|DUSP7|FTL|MRPL10|MRPL34|N4BP1|OCEL1|PKD1|...   \n",
       "5   AHNAK|AP1S1|AP3D1|ARHGEF3|ARID3B|CA11|CCDC51|C...   \n",
       "8   ABCB8|AHNAK2|ANKRD11|AP3D1|BCL2L1|CALML5|CBY1|...   \n",
       "18  ABCB6|ACTG1|ACTN4|ADAM15|AHNAK2|ALDH16A1|AP2S1...   \n",
       "25  AAMP|ABCA7|ABCB6|ABCB8|ABCF3|ACTG1|ACTN4|ADAM1...   \n",
       "\n",
       "                                         normal_genes  \\\n",
       "1   CLP1|CORO1C|DUSP7|FTL|MRPL10|MRPL34|N4BP1|NFE2...   \n",
       "5   AHNAK|ALOX15B|AP1S1|AP3D1|ARHGEF3|ARID3B|CA11|...   \n",
       "8   ABCB8|AHNAK2|ALOX15B|ANKRD11|AP3D1|ATP12A|BACH...   \n",
       "18  ABCB6|ACTG1|ACTN4|ADAM15|AHNAK2|AKR1B10|ALDH16...   \n",
       "25  AAMP|ABCA7|ABCB6|ABCB8|ABCF3|ACTG1|ACTN4|ADAM1...   \n",
       "\n",
       "                                         cancer_genes  \n",
       "1   ABCC5|ABCF3|ABRACL|ABTB2|ACADS|ACOT9|ACTG1|ACT...  \n",
       "5   AAMP|AARS2|ABCA7|ABCB6|ABCB8|ABCF3|ABHD10|ABTB...  \n",
       "8   AARS2|ABCA7|ABCB8|ABCC5|ABCF3|ABHD10|ACADVL|AC...  \n",
       "18  AARS2|ABCA7|ABCB6|ABCB8|ABCC5|ABCF3|ABHD10|ABL...  \n",
       "25  AAMP|AARS2|ABCA7|ABCB6|ABCB8|ABCC5|ABCF3|ABHD1...  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_mpra_df[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_mpra_df.to_csv(os.path.join(save_dir, 'target_gene_overlap_normal_cancer.csv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. get genomic instances for above"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "tissue_mapper = {\n",
    "    'Colon':'Colon', 'GDSD6':'GDSD6', 'GM12878':'GM12878', 'Melanocytes':'Melanocytes','Prostate':'Prostate',\n",
    "    'SCC':'A431-CTRLi', 'MM':'COLO_SCR_DMSO'\n",
    "}\n",
    "# tissue_mapper = {\n",
    "#     'Colon':['Colon'], 'GDSD6':['GDSD6'], 'GM12878':['GM12878'], 'Melanocytes':['Melanocytes'],\n",
    "#     'SCC':['A431-CTRLi','CAL27-CTRLi','SCC13-CTRLi'], 'MM':['COLO_SCR_DMSO','WM_SCR_DMSO']\n",
    "# }\n",
    "\n",
    "file_pro_pro_mapper = {\n",
    "    'GDSD6':'../data/processed/fig4_modelling/tf_tf_pairs/expr_GDSD6_pro_pro_vocab_info.csv',\n",
    "    'Melanocytes':'../data/processed/fig4_modelling/tf_tf_pairs/expr_Melanocytes_pro_pro_vocab_info.csv',\n",
    "    'Prostate':'../data/processed/fig4_modelling/tf_tf_pairs/expr_Prostate_pro_pro_vocab_info.csv',\n",
    "    'SCC':'../data/processed/fig4_modelling/tf_tf_pairs_scc/expr_A431-CTRLi_pro_pro_vocab_info.csv',\n",
    "    'MM':'../data/processed/fig4_modelling/tf_tf_pairs_mm/expr_COLO-SCR-DMSO_pro_pro_vocab_info.csv',\n",
    "}\n",
    "\n",
    "file_loop_loop_mapper = {\n",
    "    'GDSD6':'../data/processed/fig4_modelling/tf_tf_pairs/expr_GDSD6_loop_loop_vocab_info.csv',\n",
    "    'Melanocytes':'../data/processed/fig4_modelling/tf_tf_pairs/expr_Melanocytes_loop_loop_vocab_info.csv',\n",
    "    'Prostate':'../data/processed/fig4_modelling/tf_tf_pairs/expr_Prostate_loop_loop_vocab_info.csv',\n",
    "    'SCC':'../data/processed/fig4_modelling/tf_tf_pairs_scc/expr_A431-CTRLi_loop_loop_vocab_info.csv',\n",
    "    'MM':'../data/processed/fig4_modelling/tf_tf_pairs_mm/expr_COLO-SCR-DMSO_loop_loop_vocab_info.csv',\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_loop_counter(loop_df):\n",
    "    loop_loop_dict = defaultdict(list)\n",
    "    for anc_a,anc_b in zip(loop_df.source,loop_df.target):\n",
    "        loop_loop_dict[anc_a].append(anc_b) \n",
    "        loop_loop_dict[anc_b].append(anc_a)\n",
    "    return loop_loop_dict\n",
    "\n",
    "def get_genomic_instances_pro_loop_single_gene(vocab, tissue, gene,\n",
    "                                               anc_prom,loop_df,anc_foot,\n",
    "                                               verbose=False):\n",
    "    vocab_check = dict(zip(vocab.split('::'),vocab.split('::')[::-1]))\n",
    "\n",
    "    if verbose:\n",
    "        print('=======================')\n",
    "        print( vocab, tissue)\n",
    "        print(gene)\n",
    "    \n",
    "    # get anchors\n",
    "    anc_prom_sel = anc_prom[anc_prom.name==gene]\n",
    "    if verbose:\n",
    "        print('anc_prom', anc_prom_sel.shape)\n",
    "    if anc_prom_sel.shape[0]==0:\n",
    "        if verbose:\n",
    "            print(gene,'NO INSTANCES FOUND - no anchor promoters')\n",
    "        return None\n",
    "    \n",
    "    # get anchors that loop to promoter anchors\n",
    "    anc_prom_arr = anc_prom_sel.thickEnd.unique()\n",
    "    loop_df_sel = loop_df[loop_df.source.isin(anc_prom_arr)|loop_df.target.isin(anc_prom_arr)]\n",
    "    loop_loop_dict = make_loop_counter(loop_df_sel)\n",
    "#     for anc in loop_loop_dict:\n",
    "#         print(anc, len(loop_loop_dict[anc]))\n",
    "    all_anc = sorted(set(anc_prom.thickEnd.to_list() +loop_df_sel.source.to_list() + loop_df_sel.target.to_list() ))\n",
    "    anc_loop_arr = sorted(set(all_anc) - set(anc_prom_arr))\n",
    "    if verbose:\n",
    "        print('num all anchors', len(all_anc))#all_anc)\n",
    "        print('num promoter anchors', len(anc_prom_arr))\n",
    "        print('num looped anchors', len(anc_loop_arr))\n",
    "  \n",
    "    # get footprinting anchor data\n",
    "    anc_foot_sel = anc_foot[anc_foot.tf.isin(vocab_check.keys())]\n",
    "    anc_foot_pro = anc_foot_sel[anc_foot_sel.name.isin(anc_prom_arr)]\n",
    "    anc_prom_arr_to_idx = dict(zip(anc_prom_arr, ['pro_'+str(idx) for idx in range(len(anc_prom_arr))]))\n",
    "    anc_foot_pro['type']= anc_foot_pro['name'].map(anc_prom_arr_to_idx)\n",
    "    if anc_foot_pro.shape[0]==0:\n",
    "        if verbose:\n",
    "            print(gene, 'NO INSTANCES FOUND - no promoter footprints')\n",
    "        return None\n",
    "\n",
    "    anc_foot_loop = pd.DataFrame()\n",
    "    for idx, row  in anc_foot_pro[['name','tf','type']].drop_duplicates().iterrows():\n",
    "        anc_pro = row['name']\n",
    "        anc_pro_tf = row.tf\n",
    "        anc_loop_tf = vocab_check[anc_pro_tf]\n",
    "        anc_loop_arr_test = loop_loop_dict[anc_pro]\n",
    "        anc_foot1 = anc_foot_sel[anc_foot_sel.name.isin(anc_loop_arr_test) & (anc_foot.tf==anc_loop_tf)]\n",
    "        anc_foot1['type']= row.type +'+loop'\n",
    "        anc_foot_loop = pd.concat([anc_foot_loop, anc_foot1])\n",
    "\n",
    "    if anc_foot_loop.shape[0]==0:\n",
    "        if verbose:\n",
    "            print(gene, 'NO INSTANCES FOUND - no looped footprints')\n",
    "        return None\n",
    "        \n",
    "    if verbose:\n",
    "        print('number of promoter footprinted anchors with vocab',anc_foot_pro.shape)\n",
    "        print('number of looped footprinted anchors with vocab',anc_foot_pro.shape)\n",
    "\n",
    "    anc_foot_pro['name'] = gene\n",
    "    genomic_instances = pd.concat([anc_foot_pro,anc_foot_loop])\n",
    "    genomic_instances.columns = ['chr', 'start', 'stop', 'TSS', 'chr_m', 'start_m', 'stop_m', 'id_trim',\n",
    "       'score', 'strand', 'qual','tf','type']\n",
    "    genomic_instances.drop('qual',axis=1,inplace=True)\n",
    "    genomic_instances['vocab'] = vocab\n",
    "    genomic_instances['tissue'] = tissue\n",
    "    genomic_instances['config'] = 'pro_loop'\n",
    "#     genomic_instances['gene'] = gene\n",
    "    \n",
    "    if verbose:\n",
    "        print('*** genomic instances per gene',genomic_instances.shape)\n",
    "#         display(genomic_instances)\n",
    "    return genomic_instances\n",
    "\n",
    "def get_genomic_instances_pro_loop(vocab, tissue, list_genes=None,verbose=False, \n",
    "                          tf_id_to_name_dict=tf_id_to_name_dict,\n",
    "                              save_dir=None):\n",
    "    # get genes with this vocab (any configuraton)\n",
    "    if verbose:\n",
    "        print('=======================')\n",
    "        print( vocab, tissue)\n",
    "        print(list_genes)\n",
    "        print(len(list_genes))\n",
    "    # getting files\n",
    "    anc_prom = pybedtools.BedTool('../data/interim/annon/promoter_anchors/promoter_'+tissue+'_annon.bed').to_dataframe()\n",
    "    loop_df = pd.read_csv('../data/interim/merged/loops/'+tissue+'.loops.csv',index_col=0)\n",
    "    anc_foot = pybedtools.BedTool('../data/interim/annon/anchor_footprinting/'+tissue+'_annon.bed').to_dataframe()\n",
    "    anc_foot['tf'] = anc_foot.thickEnd.map(tf_id_to_name_dict)\n",
    "\n",
    "    genomic_instances = pd.DataFrame()\n",
    "    for gene in list_genes:\n",
    "        genomic_instances1 = get_genomic_instances_pro_loop_single_gene(vocab, tissue, gene,anc_prom,loop_df,anc_foot,verbose=verbose)\n",
    "        if genomic_instances1 is not None:\n",
    "#             genomic_instances1['gene'] = gene\n",
    "            print(gene, 'pro-loop',genomic_instances1.shape)\n",
    "            if save_dir is not None:\n",
    "                genomic_instances1.to_csv()\n",
    "            genomic_instances = pd.concat([genomic_instances, genomic_instances1])\n",
    "    \n",
    "    if verbose:\n",
    "        print('*** genomic instances',genomic_instances.shape)\n",
    "        display(genomic_instances)\n",
    "    if save_dir is not None:\n",
    "        genomic_instances.to_csv(os.path.join(save_dir, '_'.join(vocab.split('::'))+'_genomic_instances.csv'))\n",
    "    return genomic_instances\n",
    "\n",
    "\n",
    "# def get_genomic_instances_pro_loop(vocab, tissue, list_genes=None,verbose=False, \n",
    "#                           tf_id_to_name_dict=tf_id_to_name_dict,\n",
    "#                           save_dir=None):\n",
    "#     # get genes with this vocab (any configuraton)\n",
    "#     if verbose:\n",
    "#         print('=======================')\n",
    "#         print( vocab, tissue)\n",
    "#         print(list_genes)\n",
    "#         print(len(list_genes))\n",
    "    \n",
    "#     # get anchors\n",
    "#     anc_prom = pybedtools.BedTool('../data/interim/annon/promoter_anchors/promoter_'+tissue+'_annon.bed').to_dataframe()\n",
    "#     if list_genes is not None:\n",
    "#         anc_prom = anc_prom[anc_prom.name.isin(list_genes)]\n",
    "#     if verbose:\n",
    "#         print('anc_prom', anc_prom.shape)\n",
    "    \n",
    "#     # get anchors that loop to promoter anchors\n",
    "#     anc_prom_arr = anc_prom.thickEnd.unique()\n",
    "#     if verbose:\n",
    "#         print('num promoter anchors', len(anc_prom_arr))\n",
    "    \n",
    "#     # get loops\n",
    "#     loop_df = pd.read_csv('../data/interim/merged/loops/'+tissue+'.loops.csv',index_col=0)\n",
    "#     loop_df = loop_df[loop_df.source.isin(anc_prom_arr)|loop_df.target.isin(anc_prom_arr)]\n",
    "#     all_anc = sorted(set(anc_prom.thickEnd.to_list() +loop_df.source.to_list() + loop_df.target.to_list() ))\n",
    "#     if verbose:\n",
    "#         print('num all anchors', len(all_anc))#all_anc)\n",
    "#         display(loop_df)\n",
    "    \n",
    "#     # get footprinting anchor data\n",
    "#     anc_foot = pybedtools.BedTool('../data/interim/annon/anchor_footprinting/'+tissue+'_annon.bed').to_dataframe()\n",
    "#     anc_foot['tf'] = anc_foot.thickEnd.map(tf_id_to_name_dict)\n",
    "#     anc_foot = anc_foot[anc_foot.tf.isin(vocab.split('::'))]\n",
    "#     anc_foot = anc_foot[anc_foot.name.isin(all_anc)]\n",
    "#     if verbose:\n",
    "#         print('number of footprinted anchors with vocab',anc_foot.shape)\n",
    "    \n",
    "#     # get promoter footprint data\n",
    "#     pro_anc_foot = pd.DataFrame()\n",
    "#     pro_anc_foot_df = pybedtools.BedTool('../data/interim/annon/promoter_footprinting/promoter_'+tissue+'_annon.bed').to_dataframe()\n",
    "#     pro_anc_foot_df['tf'] = pro_anc_foot_df.thickEnd.map(tf_id_to_name_dict)\n",
    "#     if list_genes is not None:\n",
    "#         # add the footprint info if it is in a targetted gene or if the tf is in a looped anchor region of interest\n",
    "#         for gene in list_genes:\n",
    "#             pro_anc_foot1 = pro_anc_foot_df[(pro_anc_foot_df.name==gene)|(pro_anc_foot_df.tf.isin(vocab.split('::')))]# not perfect will catch later???\n",
    "            \n",
    "#             pro_anc_foot1['anchor'] = pro_anc_foot_df.score + \"_\" + pro_anc_foot_df.strand.map(str)+\"_\"+ pro_anc_foot_df.thickStart.map(str)\n",
    "#             pro_anc_foot1 = pro_anc_foot1[pro_anc_foot1.anchor.isin(all_anc)]\n",
    "#             pro_anc_foot1.drop(columns=['anchor'],inplace=True)\n",
    "#             pro_anc_foot = pd.concat([pro_anc_foot, pro_anc_foot1])\n",
    "#     else:\n",
    "#         pro_anc_foot = pro_anc_foot_df[pro_anc_foot_df.tf.isin(vocab.split('::'))]\n",
    "    \n",
    "#     if pro_anc_foot.shape[0]==0:\n",
    "#         genomic_instances = pd.DataFrame(columns=['chr', 'start', 'stop', 'TSS', 'chr_m', 'start_m', 'stop_m', 'id_trim',\n",
    "#        'score', 'strand', 'tf'])\n",
    "#     else:\n",
    "#         genomic_instances = pd.concat([anc_foot,pro_anc_foot])\n",
    "#         genomic_instances.columns = ['chr', 'start', 'stop', 'TSS', 'chr_m', 'start_m', 'stop_m', 'id_trim',\n",
    "#            'score', 'strand', 'qual','tf']\n",
    "#         genomic_instances.drop('qual',axis=1,inplace=True)\n",
    "#         genomic_instances['vocab'] = vocab\n",
    "#         genomic_instances['tissue'] = tissue\n",
    "#         genomic_instances['config'] = 'pro_loop'\n",
    "    \n",
    "#     if verbose:\n",
    "#         print('*** genomic instances',genomic_instances.shape)\n",
    "#         display(genomic_instances)\n",
    "#     if save_dir is not None:\n",
    "#         genomic_instances.to_csv(os.path.join(save_dir, '_'.join(vocab.split('::'))+'_genomic_instances.csv'))\n",
    "#     return genomic_instances\n",
    "\n",
    "\n",
    "def get_genomic_instances_all(vocab, tissue, list_genes,verbose=False, \n",
    "                          tf_id_to_name_dict=tf_id_to_name_dict,\n",
    "                          save_prefix=None,rewrite=True):\n",
    "    #pro-pro\n",
    "    print('(((())))')\n",
    "    print(vocab, tissue, len(list_genes))\n",
    "    if (not rewrite) and save_prefix is not None:\n",
    "        save_file = save_prefix + 'genomic_instances_'+'_'.join(vocab.split('::'))+\"_\"+tissue+'.csv'\n",
    "        if os.path.exists(save_file):\n",
    "            print('loading from file')\n",
    "            genomic_instances_all = pd.read_csv(save_file,index_col=0)\n",
    "            print('overall instances',genomic_instances_all.shape)\n",
    "            return genomic_instances_all\n",
    "\n",
    "\n",
    "    print('(((())))')\n",
    "    print(vocab, tissue, len(list_genes))\n",
    "    pro_pro_file = file_pro_pro_mapper[tissue]\n",
    "    genomic_instances_pro_pro = pd.read_csv(pro_pro_file,index_col=0)\n",
    "    genomic_instances_pro_pro = genomic_instances_pro_pro[genomic_instances_pro_pro.TSS.isin(list_genes)]\n",
    "    genomic_instances_pro_pro = genomic_instances_pro_pro[genomic_instances_pro_pro.vocab==vocab]\n",
    "    genomic_instances_pro_pro.drop('genes',axis=1,inplace=True)\n",
    "    genomic_instances_pro_pro['config'] = 'pro_pro'\n",
    "    genomic_instances_pro_pro['type']  = '-'\n",
    "    print('pro-pro instances',genomic_instances_pro_pro.shape)\n",
    "    \n",
    "    # pro-loop\n",
    "\n",
    "    genomic_instances_pro_loop = get_genomic_instances_pro_loop(vocab, tissue_mapper[tissue], list_genes,verbose=verbose, \n",
    "                          tf_id_to_name_dict=tf_id_to_name_dict,\n",
    "                          save_dir=None)\n",
    "    print('pro-loop instances',genomic_instances_pro_loop.shape)\n",
    "    \n",
    "    #loop-loop\n",
    "    loop_loop_file = file_loop_loop_mapper[tissue]\n",
    "    genomic_instances_loop_loop = pd.read_csv(loop_loop_file,index_col=0)\n",
    "    genomic_instances_loop_loop = genomic_instances_loop_loop[genomic_instances_loop_loop.TSS.isin(list_genes)]\n",
    "    genomic_instances_loop_loop = genomic_instances_loop_loop[genomic_instances_loop_loop.vocab==vocab]\n",
    "    genomic_instances_loop_loop.drop('genes',axis=1,inplace=True)\n",
    "    genomic_instances_loop_loop['config'] = 'loop_loop'\n",
    "    genomic_instances_loop_loop['type']  = '-'\n",
    "    print('loop-loop instances',genomic_instances_loop_loop.shape)\n",
    "\n",
    "    genomic_instances_all = pd.concat([genomic_instances_pro_pro, genomic_instances_pro_loop, genomic_instances_loop_loop],ignore_index=True)\n",
    "    print('overall instances',genomic_instances_all.shape)\n",
    "    if save_prefix is not None:\n",
    "        genomic_instances_all.to_csv(save_prefix + 'genomic_instances_'+'_'.join(vocab.split('::'))+\"_\"+tissue+'.csv')\n",
    "\n",
    "    return genomic_instances_all\n",
    "\n",
    "def get_target_genes(vocab,tissue, verbose=False):\n",
    "    df_all_instances = get_genomic_instances_pro_loop(vocab, tissue,verbose=verbose)\n",
    "    df = df_all_instances[~df_all_instances.TSS.str.startswith('chr')]\n",
    "    print(len(df.TSS.unique()), 'target genes found for', vocab, 'in tissue: ', tissue)\n",
    "    print(sorted(df.TSS.unique()))\n",
    "    return df_all_instances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%time\n",
    "# for idx, row in new_mpra_df.iterrows():\n",
    "#     for tissue in row.tissues_to_check.split('|'):\n",
    "#         get_genomic_instances_all(row.vocab, tissue, row.overlap_genes.split('|'),\n",
    "#                                   verbose=False, \n",
    "#                                   save_prefix=save_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3 Different vocabs same gene\n",
    "\n",
    " I was more interested in the overlap between the different vocabs but targeting the same gene in their respective tissues, for example if there is overlap between ARNT::SP1 target genes in SCC with BACH1::KLF5 target genes in GDSD6. Essentially wanting to get at cancer coopts a new vocab to regulate the same target genes that are regulated in normal by a different vocab.\n",
    " \n",
    " \n",
    " update 7/23/21\n",
    "- do this for all skin type vocabs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10\n",
      "10\n",
      "5\n",
      "10\n"
     ]
    }
   ],
   "source": [
    "# get mpra vocabs + configs\n",
    "vocab_df = pd.read_csv('../data/processed/mpra/results_df_piv_match_count.csv',index_col=0)\n",
    "# filter for skin stuff\n",
    "vocab_df = vocab_df[vocab_df.tissue.isin(['GDSD6','SCC','Melanocytes','MM'])]\n",
    "\n",
    "# get target genes for each AND vocab word in the correct tissue\n",
    "vocab_to_target_genes = {}\n",
    "for idx, row in vocab_df.iterrows():\n",
    "    tf1,tf2 = row['vocab'].split('::')\n",
    "    geneset_file = os.path.join('../data/processed/fig7_vocab_go/vocab_genesets/',row['tissue']+'_'+tf1 + '-' + tf2+'.txt' )\n",
    "    geneset = set(pd.read_csv(geneset_file,header=None).iloc[:,0].values)\n",
    "    vocab_to_target_genes[row.vocab] = geneset\n",
    "#     print(row.vocab, len(geneset))\n",
    "\n",
    "    \n",
    "# filter for AND VOCABs\n",
    "vocab_df = vocab_df[vocab_df.consensus_class=='AND']\n",
    "\n",
    "# get vocab words for each tissue\n",
    "GDSD6_AND_vocabs = vocab_df[vocab_df.tissue=='GDSD6'].vocab.values\n",
    "SCC_AND_vocabs = vocab_df[vocab_df.tissue=='SCC'].vocab.values\n",
    "Melanocytes_AND_vocabs = vocab_df[vocab_df.tissue=='Melanocytes'].vocab.values\n",
    "MM_AND_vocabs = vocab_df[vocab_df.tissue=='MM'].vocab.values\n",
    "AND_vocabs = [GDSD6_AND_vocabs,SCC_AND_vocabs, Melanocytes_AND_vocabs, MM_AND_vocabs ]\n",
    "for v_list in AND_vocabs:\n",
    "    print(len(v_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {},
   "outputs": [],
   "source": [
    "intersect_genes_MC_MM = pd.DataFrame(index=Melanocytes_AND_vocabs, columns = MM_AND_vocabs)\n",
    "for MC_vocab in Melanocytes_AND_vocabs:\n",
    "    for MM_vocab in MM_AND_vocabs:\n",
    "        intersect_genes = vocab_to_target_genes[MC_vocab].intersection(vocab_to_target_genes[MM_vocab])\n",
    "        intersect_genes_MC_MM.at[MC_vocab, MM_vocab] = '|'.join(sorted(intersect_genes))\n",
    "intersect_genes_MC_MM_long = intersect_genes_MC_MM.reset_index().melt('index').drop_duplicates()\n",
    "intersect_genes_MC_MM_long.columns = ['MC_vocab','MM_vocab','intersect_genes']\n",
    "intersect_genes_MC_MM_long['num_genes'] = intersect_genes_MC_MM_long.intersect_genes.apply(lambda str_v: len(str_v.split('|')))\n",
    "intersect_genes_MC_MM_long.to_csv(os.path.join(save_dir, 'intersect_genes_MC_MM_long.csv'))\n",
    "# intersect_genes_MC_MM_long # size 5x9 = 45"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>KC_vocab</th>\n",
       "      <th>SCC_vocab</th>\n",
       "      <th>intersect_genes</th>\n",
       "      <th>num_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>BACH1::KLF5</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>ABCB8|AHNAK2|ANKRD11|AP3D1|CALML5|CELSR1|CHRNB...</td>\n",
       "      <td>69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>EGR2::FOSL2</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>ABCB8|ABCF3|ABHD10|ACTR3|ADAM15|AHNAK|AHNAK2|A...</td>\n",
       "      <td>150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>EGR2::RARG</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>ABCB8|ABCF3|ABHD10|ACADVL|AHNAK2|AP1S1|AP2S1|A...</td>\n",
       "      <td>126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>EGR2::RORA</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>ABCA7|ABCB6|ABCF3|ACADVL|ACTR3|ADAM15|AHNAK2|A...</td>\n",
       "      <td>137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>HMGA1::KLF5</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>AAMP|ABCA7|ABCB8|ABHD10|ACADVL|ACTG1|ADAMTS1|A...</td>\n",
       "      <td>160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>76</td>\n",
       "      <td>HMGA1::KLF5</td>\n",
       "      <td>SP1::ZBTB17</td>\n",
       "      <td>AAMP|ABCA7|ABCB8|ABHD10|ACADVL|ACTG1|ACTN4|ADA...</td>\n",
       "      <td>232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>77</td>\n",
       "      <td>KLF5::MAFG</td>\n",
       "      <td>SP1::ZBTB17</td>\n",
       "      <td>ABCB8|ACTN4|ADAM15|ADAMTS1|AHNAK2|AKR1B1|ANKRD...</td>\n",
       "      <td>133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>78</td>\n",
       "      <td>KLF5::MAFK</td>\n",
       "      <td>SP1::ZBTB17</td>\n",
       "      <td>ABCB6|ABCB8|ACADVL|ADAM15|ADAMTS1|AHNAK2|AKR1B...</td>\n",
       "      <td>120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>79</td>\n",
       "      <td>MAF::RORA</td>\n",
       "      <td>SP1::ZBTB17</td>\n",
       "      <td>ACADVL|ACTN4|AHNAK2|ANKZF1|AP3D1|ARHGDIA|ATG9A...</td>\n",
       "      <td>95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>80</td>\n",
       "      <td>PRDM1::RORA</td>\n",
       "      <td>SP1::ZBTB17</td>\n",
       "      <td>ABCB6|ADAM15|AHNAK2|AP3D1|APOBEC3C|ARID3B|ARTN...</td>\n",
       "      <td>74</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>81 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       KC_vocab    SCC_vocab  \\\n",
       "0   BACH1::KLF5    ARNT::SP1   \n",
       "1   EGR2::FOSL2    ARNT::SP1   \n",
       "2    EGR2::RARG    ARNT::SP1   \n",
       "3    EGR2::RORA    ARNT::SP1   \n",
       "4   HMGA1::KLF5    ARNT::SP1   \n",
       "..          ...          ...   \n",
       "76  HMGA1::KLF5  SP1::ZBTB17   \n",
       "77   KLF5::MAFG  SP1::ZBTB17   \n",
       "78   KLF5::MAFK  SP1::ZBTB17   \n",
       "79    MAF::RORA  SP1::ZBTB17   \n",
       "80  PRDM1::RORA  SP1::ZBTB17   \n",
       "\n",
       "                                      intersect_genes  num_genes  \n",
       "0   ABCB8|AHNAK2|ANKRD11|AP3D1|CALML5|CELSR1|CHRNB...         69  \n",
       "1   ABCB8|ABCF3|ABHD10|ACTR3|ADAM15|AHNAK|AHNAK2|A...        150  \n",
       "2   ABCB8|ABCF3|ABHD10|ACADVL|AHNAK2|AP1S1|AP2S1|A...        126  \n",
       "3   ABCA7|ABCB6|ABCF3|ACADVL|ACTR3|ADAM15|AHNAK2|A...        137  \n",
       "4   AAMP|ABCA7|ABCB8|ABHD10|ACADVL|ACTG1|ADAMTS1|A...        160  \n",
       "..                                                ...        ...  \n",
       "76  AAMP|ABCA7|ABCB8|ABHD10|ACADVL|ACTG1|ACTN4|ADA...        232  \n",
       "77  ABCB8|ACTN4|ADAM15|ADAMTS1|AHNAK2|AKR1B1|ANKRD...        133  \n",
       "78  ABCB6|ABCB8|ACADVL|ADAM15|ADAMTS1|AHNAK2|AKR1B...        120  \n",
       "79  ACADVL|ACTN4|AHNAK2|ANKZF1|AP3D1|ARHGDIA|ATG9A...         95  \n",
       "80  ABCB6|ADAM15|AHNAK2|AP3D1|APOBEC3C|ARID3B|ARTN...         74  \n",
       "\n",
       "[81 rows x 4 columns]"
      ]
     },
     "execution_count": 171,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intersect_genes_KC_SCC = pd.DataFrame(index=GDSD6_AND_vocabs, columns = SCC_AND_vocabs)\n",
    "for KC_vocab in GDSD6_AND_vocabs:\n",
    "    for SCC_vocab in SCC_AND_vocabs:\n",
    "        intersect_genes = vocab_to_target_genes[KC_vocab].intersection(vocab_to_target_genes[SCC_vocab])\n",
    "        intersect_genes_KC_SCC.at[KC_vocab, SCC_vocab] = '|'.join(sorted(intersect_genes))\n",
    "intersect_genes_KC_SCC_long = intersect_genes_KC_SCC.reset_index().melt('index').drop_duplicates()\n",
    "intersect_genes_KC_SCC_long.columns = ['KC_vocab','SCC_vocab','intersect_genes']\n",
    "intersect_genes_KC_SCC_long['num_genes'] = intersect_genes_KC_SCC_long.intersect_genes.apply(lambda str_v: len(str_v.split('|')))\n",
    "intersect_genes_KC_SCC_long.to_csv(os.path.join(save_dir, 'intersect_genes_KC_SCC_long.csv'))\n",
    "intersect_genes_KC_SCC_long #size 9x9 = 81"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 +++++++++++++++\n",
      "(((())))\n",
      "ETV2::PRRX1 Melanocytes 29\n",
      "pro-pro instances (42, 14)\n",
      "pro-loop instances (257, 14)\n",
      "loop-loop instances (284, 14)\n",
      "overall instances (583, 14)\n",
      "(((())))\n",
      "EGR2::NFE2L1 MM 29\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (223, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (223, 14)\n",
      "1 +++++++++++++++\n",
      "(((())))\n",
      "FOXD3::PRRX1 Melanocytes 16\n",
      "pro-pro instances (138, 14)\n",
      "pro-loop instances (214, 14)\n",
      "loop-loop instances (248, 14)\n",
      "overall instances (600, 14)\n",
      "(((())))\n",
      "EGR2::NFE2L1 MM 16\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (114, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (114, 14)\n",
      "2 +++++++++++++++\n",
      "(((())))\n",
      "KLF15::ZNF589 Melanocytes 70\n",
      "pro-pro instances (210, 14)\n",
      "pro-loop instances (483, 14)\n",
      "loop-loop instances (674, 14)\n",
      "overall instances (1367, 14)\n",
      "(((())))\n",
      "EGR2::NFE2L1 MM 70\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (505, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (505, 14)\n",
      "3 +++++++++++++++\n",
      "(((())))\n",
      "KLF5::NFYB Melanocytes 22\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (107, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (107, 14)\n",
      "(((())))\n",
      "EGR2::NFE2L1 MM 22\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (225, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (225, 14)\n",
      "4 +++++++++++++++\n",
      "(((())))\n",
      "MITF::ZNF589 Melanocytes 9\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (56, 14)\n",
      "loop-loop instances (30, 14)\n",
      "overall instances (86, 14)\n",
      "(((())))\n",
      "EGR2::NFE2L1 MM 9\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (106, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (106, 14)\n",
      "5 +++++++++++++++\n",
      "(((())))\n",
      "ETV2::PRRX1 Melanocytes 19\n",
      "pro-pro instances (79, 14)\n",
      "pro-loop instances (179, 14)\n",
      "loop-loop instances (240, 14)\n",
      "overall instances (498, 14)\n",
      "(((())))\n",
      "EN2::STAT1 MM 19\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (59, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (59, 14)\n",
      "6 +++++++++++++++\n",
      "(((())))\n",
      "FOXD3::PRRX1 Melanocytes 11\n",
      "pro-pro instances (117, 14)\n",
      "pro-loop instances (180, 14)\n",
      "loop-loop instances (182, 14)\n",
      "overall instances (479, 14)\n",
      "(((())))\n",
      "EN2::STAT1 MM 11\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (44, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (44, 14)\n",
      "7 +++++++++++++++\n",
      "(((())))\n",
      "KLF15::ZNF589 Melanocytes 38\n",
      "pro-pro instances (134, 14)\n",
      "pro-loop instances (301, 14)\n",
      "loop-loop instances (355, 14)\n",
      "overall instances (790, 14)\n",
      "(((())))\n",
      "EN2::STAT1 MM 38\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (98, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (98, 14)\n",
      "8 +++++++++++++++\n",
      "(((())))\n",
      "KLF5::NFYB Melanocytes 18\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (89, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (89, 14)\n",
      "(((())))\n",
      "EN2::STAT1 MM 18\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (56, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (56, 14)\n",
      "9 +++++++++++++++\n",
      "(((())))\n",
      "MITF::ZNF589 Melanocytes 5\n",
      "pro-pro instances (4, 14)\n",
      "pro-loop instances (34, 14)\n",
      "loop-loop instances (13, 14)\n",
      "overall instances (51, 14)\n",
      "(((())))\n",
      "EN2::STAT1 MM 5\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (26, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (26, 14)\n",
      "10 +++++++++++++++\n",
      "(((())))\n",
      "ETV2::PRRX1 Melanocytes 18\n",
      "pro-pro instances (65, 14)\n",
      "pro-loop instances (172, 14)\n",
      "loop-loop instances (216, 14)\n",
      "overall instances (453, 14)\n",
      "(((())))\n",
      "HBP1::STAT1 MM 18\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (55, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (55, 14)\n",
      "11 +++++++++++++++\n",
      "(((())))\n",
      "FOXD3::PRRX1 Melanocytes 11\n",
      "pro-pro instances (117, 14)\n",
      "pro-loop instances (177, 14)\n",
      "loop-loop instances (182, 14)\n",
      "overall instances (476, 14)\n",
      "(((())))\n",
      "HBP1::STAT1 MM 11\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (40, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (40, 14)\n",
      "12 +++++++++++++++\n",
      "(((())))\n",
      "KLF15::ZNF589 Melanocytes 41\n",
      "pro-pro instances (185, 14)\n",
      "pro-loop instances (310, 14)\n",
      "loop-loop instances (404, 14)\n",
      "overall instances (899, 14)\n",
      "(((())))\n",
      "HBP1::STAT1 MM 41\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (90, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (90, 14)\n",
      "13 +++++++++++++++\n",
      "(((())))\n",
      "KLF5::NFYB Melanocytes 16\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (77, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (77, 14)\n",
      "(((())))\n",
      "HBP1::STAT1 MM 16\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (53, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (53, 14)\n",
      "14 +++++++++++++++\n",
      "(((())))\n",
      "MITF::ZNF589 Melanocytes 6\n",
      "pro-pro instances (4, 14)\n",
      "pro-loop instances (37, 14)\n",
      "loop-loop instances (13, 14)\n",
      "overall instances (54, 14)\n",
      "(((())))\n",
      "HBP1::STAT1 MM 6\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (24, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (24, 14)\n",
      "15 +++++++++++++++\n",
      "(((())))\n",
      "ETV2::PRRX1 Melanocytes 22\n",
      "pro-pro instances (79, 14)\n",
      "pro-loop instances (204, 14)\n",
      "loop-loop instances (246, 14)\n",
      "overall instances (529, 14)\n",
      "(((())))\n",
      "NFATC4::STAT1 MM 22\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (113, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (113, 14)\n",
      "16 +++++++++++++++\n",
      "(((())))\n",
      "FOXD3::PRRX1 Melanocytes 12\n",
      "pro-pro instances (117, 14)\n",
      "pro-loop instances (182, 14)\n",
      "loop-loop instances (182, 14)\n",
      "overall instances (481, 14)\n",
      "(((())))\n",
      "NFATC4::STAT1 MM 12\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (77, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (77, 14)\n",
      "17 +++++++++++++++\n",
      "(((())))\n",
      "KLF15::ZNF589 Melanocytes 47\n",
      "pro-pro instances (200, 14)\n",
      "pro-loop instances (346, 14)\n",
      "loop-loop instances (459, 14)\n",
      "overall instances (1005, 14)\n",
      "(((())))\n",
      "NFATC4::STAT1 MM 47\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (244, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (244, 14)\n",
      "18 +++++++++++++++\n",
      "(((())))\n",
      "KLF5::NFYB Melanocytes 18\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (91, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (91, 14)\n",
      "(((())))\n",
      "NFATC4::STAT1 MM 18\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (131, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (131, 14)\n",
      "19 +++++++++++++++\n",
      "(((())))\n",
      "MITF::ZNF589 Melanocytes 7\n",
      "pro-pro instances (4, 14)\n",
      "pro-loop instances (40, 14)\n",
      "loop-loop instances (20, 14)\n",
      "overall instances (64, 14)\n",
      "(((())))\n",
      "NFATC4::STAT1 MM 7\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (71, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (71, 14)\n",
      "20 +++++++++++++++\n",
      "(((())))\n",
      "ETV2::PRRX1 Melanocytes 21\n",
      "pro-pro instances (63, 14)\n",
      "pro-loop instances (195, 14)\n",
      "loop-loop instances (261, 14)\n",
      "overall instances (519, 14)\n",
      "(((())))\n",
      "NFE2L1::NR3C1 MM 21\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (63, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (63, 14)\n",
      "21 +++++++++++++++\n",
      "(((())))\n",
      "FOXD3::PRRX1 Melanocytes 11\n",
      "pro-pro instances (122, 14)\n",
      "pro-loop instances (187, 14)\n",
      "loop-loop instances (212, 14)\n",
      "overall instances (521, 14)\n",
      "(((())))\n",
      "NFE2L1::NR3C1 MM 11\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (50, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (50, 14)\n",
      "22 +++++++++++++++\n",
      "(((())))\n",
      "KLF15::ZNF589 Melanocytes 44\n",
      "pro-pro instances (141, 14)\n",
      "pro-loop instances (309, 14)\n",
      "loop-loop instances (466, 14)\n",
      "overall instances (916, 14)\n",
      "(((())))\n",
      "NFE2L1::NR3C1 MM 44\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (106, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (106, 14)\n",
      "23 +++++++++++++++\n",
      "(((())))\n",
      "KLF5::NFYB Melanocytes 16\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (89, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (89, 14)\n",
      "(((())))\n",
      "NFE2L1::NR3C1 MM 16\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (55, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (55, 14)\n",
      "24 +++++++++++++++\n",
      "(((())))\n",
      "MITF::ZNF589 Melanocytes 8\n",
      "pro-pro instances (4, 14)\n",
      "pro-loop instances (38, 14)\n",
      "loop-loop instances (35, 14)\n",
      "overall instances (77, 14)\n",
      "(((())))\n",
      "NFE2L1::NR3C1 MM 8\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (37, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (37, 14)\n",
      "25 +++++++++++++++\n",
      "(((())))\n",
      "ETV2::PRRX1 Melanocytes 30\n",
      "pro-pro instances (53, 14)\n",
      "pro-loop instances (276, 14)\n",
      "loop-loop instances (302, 14)\n",
      "overall instances (631, 14)\n",
      "(((())))\n",
      "NFE2L1::TBX15 MM 30\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (211, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (211, 14)\n",
      "26 +++++++++++++++\n",
      "(((())))\n",
      "FOXD3::PRRX1 Melanocytes 16\n",
      "pro-pro instances (138, 14)\n",
      "pro-loop instances (214, 14)\n",
      "loop-loop instances (248, 14)\n",
      "overall instances (600, 14)\n",
      "(((())))\n",
      "NFE2L1::TBX15 MM 16\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (108, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (108, 14)\n",
      "27 +++++++++++++++\n",
      "(((())))\n",
      "KLF15::ZNF589 Melanocytes 71\n",
      "pro-pro instances (257, 14)\n",
      "pro-loop instances (500, 14)\n",
      "loop-loop instances (731, 14)\n",
      "overall instances (1488, 14)\n",
      "(((())))\n",
      "NFE2L1::TBX15 MM 71\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (578, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (578, 14)\n",
      "28 +++++++++++++++\n",
      "(((())))\n",
      "KLF5::NFYB Melanocytes 22\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (107, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (107, 14)\n",
      "(((())))\n",
      "NFE2L1::TBX15 MM 22\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (256, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (256, 14)\n",
      "29 +++++++++++++++\n",
      "(((())))\n",
      "MITF::ZNF589 Melanocytes 9\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (56, 14)\n",
      "loop-loop instances (30, 14)\n",
      "overall instances (86, 14)\n",
      "(((())))\n",
      "NFE2L1::TBX15 MM 9\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (104, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (104, 14)\n",
      "30 +++++++++++++++\n",
      "(((())))\n",
      "ETV2::PRRX1 Melanocytes 25\n",
      "pro-pro instances (93, 14)\n",
      "pro-loop instances (260, 14)\n",
      "loop-loop instances (290, 14)\n",
      "overall instances (643, 14)\n",
      "(((())))\n",
      "NFE2L1::ZBTB49 MM 25\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (92, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (92, 14)\n",
      "31 +++++++++++++++\n",
      "(((())))\n",
      "FOXD3::PRRX1 Melanocytes 12\n",
      "pro-pro instances (138, 14)\n",
      "pro-loop instances (191, 14)\n",
      "loop-loop instances (226, 14)\n",
      "overall instances (555, 14)\n",
      "(((())))\n",
      "NFE2L1::ZBTB49 MM 12\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (59, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (59, 14)\n",
      "32 +++++++++++++++\n",
      "(((())))\n",
      "KLF15::ZNF589 Melanocytes 59\n",
      "pro-pro instances (140, 14)\n",
      "pro-loop instances (427, 14)\n",
      "loop-loop instances (568, 14)\n",
      "overall instances (1135, 14)\n",
      "(((())))\n",
      "NFE2L1::ZBTB49 MM 59\n",
      "pro-pro instances (3, 14)\n",
      "pro-loop instances (160, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (163, 14)\n",
      "33 +++++++++++++++\n",
      "(((())))\n",
      "KLF5::NFYB Melanocytes 21\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (97, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (97, 14)\n",
      "(((())))\n",
      "NFE2L1::ZBTB49 MM 21\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (72, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (72, 14)\n",
      "34 +++++++++++++++\n",
      "(((())))\n",
      "MITF::ZNF589 Melanocytes 7\n",
      "pro-pro instances (4, 14)\n",
      "pro-loop instances (49, 14)\n",
      "loop-loop instances (14, 14)\n",
      "overall instances (67, 14)\n",
      "(((())))\n",
      "NFE2L1::ZBTB49 MM 7\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (37, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (37, 14)\n",
      "35 +++++++++++++++\n",
      "(((())))\n",
      "ETV2::PRRX1 Melanocytes 27\n",
      "pro-pro instances (65, 14)\n",
      "pro-loop instances (240, 14)\n",
      "loop-loop instances (282, 14)\n",
      "overall instances (587, 14)\n",
      "(((())))\n",
      "SOX10::SOX13 MM 27\n",
      "pro-pro instances (11, 14)\n",
      "pro-loop instances (245, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (256, 14)\n",
      "36 +++++++++++++++\n",
      "(((())))\n",
      "FOXD3::PRRX1 Melanocytes 14\n",
      "pro-pro instances (138, 14)\n",
      "pro-loop instances (205, 14)\n",
      "loop-loop instances (222, 14)\n",
      "overall instances (565, 14)\n",
      "(((())))\n",
      "SOX10::SOX13 MM 14\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (166, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (166, 14)\n",
      "37 +++++++++++++++\n",
      "(((())))\n",
      "KLF15::ZNF589 Melanocytes 61\n",
      "pro-pro instances (171, 14)\n",
      "pro-loop instances (424, 14)\n",
      "loop-loop instances (613, 14)\n",
      "overall instances (1208, 14)\n",
      "(((())))\n",
      "SOX10::SOX13 MM 61\n",
      "pro-pro instances (15, 14)\n",
      "pro-loop instances (473, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (488, 14)\n",
      "38 +++++++++++++++\n",
      "(((())))\n",
      "KLF5::NFYB Melanocytes 24\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (111, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (111, 14)\n",
      "(((())))\n",
      "SOX10::SOX13 MM 24\n",
      "pro-pro instances (5, 14)\n",
      "pro-loop instances (238, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (243, 14)\n",
      "39 +++++++++++++++\n",
      "(((())))\n",
      "MITF::ZNF589 Melanocytes 9\n",
      "pro-pro instances (4, 14)\n",
      "pro-loop instances (53, 14)\n",
      "loop-loop instances (24, 14)\n",
      "overall instances (81, 14)\n",
      "(((())))\n",
      "SOX10::SOX13 MM 9\n",
      "pro-pro instances (5, 14)\n",
      "pro-loop instances (132, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (137, 14)\n",
      "40 +++++++++++++++\n",
      "(((())))\n",
      "ETV2::PRRX1 Melanocytes 26\n",
      "pro-pro instances (39, 14)\n",
      "pro-loop instances (229, 14)\n",
      "loop-loop instances (256, 14)\n",
      "overall instances (524, 14)\n",
      "(((())))\n",
      "SOX10::SP4 MM 26\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (293, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (293, 14)\n",
      "41 +++++++++++++++\n",
      "(((())))\n",
      "FOXD3::PRRX1 Melanocytes 14\n",
      "pro-pro instances (138, 14)\n",
      "pro-loop instances (205, 14)\n",
      "loop-loop instances (222, 14)\n",
      "overall instances (565, 14)\n",
      "(((())))\n",
      "SOX10::SP4 MM 14\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (143, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (143, 14)\n",
      "42 +++++++++++++++\n",
      "(((())))\n",
      "KLF15::ZNF589 Melanocytes 60\n",
      "pro-pro instances (171, 14)\n",
      "pro-loop instances (419, 14)\n",
      "loop-loop instances (612, 14)\n",
      "overall instances (1202, 14)\n",
      "(((())))\n",
      "SOX10::SP4 MM 60\n",
      "pro-pro instances (22, 14)\n",
      "pro-loop instances (693, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (715, 14)\n",
      "43 +++++++++++++++\n",
      "(((())))\n",
      "KLF5::NFYB Melanocytes 24\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (111, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (111, 14)\n",
      "(((())))\n",
      "SOX10::SP4 MM 24\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (310, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (310, 14)\n",
      "44 +++++++++++++++\n",
      "(((())))\n",
      "MITF::ZNF589 Melanocytes 9\n",
      "pro-pro instances (4, 14)\n",
      "pro-loop instances (53, 14)\n",
      "loop-loop instances (24, 14)\n",
      "overall instances (81, 14)\n",
      "(((())))\n",
      "SOX10::SP4 MM 9\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (136, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (136, 14)\n",
      "CPU times: user 26min 58s, sys: 1min 20s, total: 28min 18s\n",
      "Wall time: 25min 10s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# takes awhile\n",
    "for idx, row in intersect_genes_MC_MM_long.iterrows():\n",
    "    print(idx, '+++++++++++++++')\n",
    "    get_genomic_instances_all(row.MC_vocab, 'Melanocytes', row.intersect_genes.split('|'),\n",
    "                              verbose=False, \n",
    "                              save_prefix='../data/processed/fig7_vocab_go/genomic_instance_overlap/'+str(idx)+'_')\n",
    "    get_genomic_instances_all(row.MM_vocab, 'MM', row.intersect_genes.split('|'),\n",
    "                              verbose=False, \n",
    "                              save_prefix='../data/processed/fig7_vocab_go/genomic_instance_overlap/'+str(idx)+'_')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 +++++++++++++++\n",
      "(((())))\n",
      "BACH1::KLF5 GDSD6 69\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (313, 14)\n",
      "loop-loop instances (134, 14)\n",
      "overall instances (447, 14)\n",
      "(((())))\n",
      "ARNT::SP1 SCC 69\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (1644, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (1644, 14)\n",
      "1 +++++++++++++++\n",
      "(((())))\n",
      "EGR2::FOSL2 GDSD6 150\n",
      "pro-pro instances (14, 14)\n",
      "pro-loop instances (1051, 14)\n",
      "loop-loop instances (558, 14)\n",
      "overall instances (1623, 14)\n",
      "(((())))\n",
      "ARNT::SP1 SCC 150\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (2943, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (2943, 14)\n",
      "2 +++++++++++++++\n",
      "(((())))\n",
      "EGR2::RARG GDSD6 126\n",
      "pro-pro instances (12, 14)\n",
      "pro-loop instances (873, 14)\n",
      "loop-loop instances (486, 14)\n",
      "overall instances (1371, 14)\n",
      "(((())))\n",
      "ARNT::SP1 SCC 126\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (2760, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (2760, 14)\n",
      "3 +++++++++++++++\n",
      "(((())))\n",
      "EGR2::RORA GDSD6 137\n",
      "pro-pro instances (120, 14)\n",
      "pro-loop instances (953, 14)\n",
      "loop-loop instances (580, 14)\n",
      "overall instances (1653, 14)\n",
      "(((())))\n",
      "ARNT::SP1 SCC 137\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (2892, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (2892, 14)\n",
      "4 +++++++++++++++\n",
      "(((())))\n",
      "HMGA1::KLF5 GDSD6 160\n",
      "pro-pro instances (20, 14)\n",
      "pro-loop instances (857, 14)\n",
      "loop-loop instances (336, 14)\n",
      "overall instances (1213, 14)\n",
      "(((())))\n",
      "ARNT::SP1 SCC 160\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (3440, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (3440, 14)\n",
      "5 +++++++++++++++\n",
      "(((())))\n",
      "KLF5::MAFG GDSD6 102\n",
      "pro-pro instances (17, 14)\n",
      "pro-loop instances (475, 14)\n",
      "loop-loop instances (206, 14)\n",
      "overall instances (698, 14)\n",
      "(((())))\n",
      "ARNT::SP1 SCC 102\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (2407, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (2407, 14)\n",
      "6 +++++++++++++++\n",
      "(((())))\n",
      "KLF5::MAFK GDSD6 93\n",
      "pro-pro instances (14, 14)\n",
      "pro-loop instances (437, 14)\n",
      "loop-loop instances (180, 14)\n",
      "overall instances (631, 14)\n",
      "(((())))\n",
      "ARNT::SP1 SCC 93\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (1913, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (1913, 14)\n",
      "7 +++++++++++++++\n",
      "(((())))\n",
      "MAF::RORA GDSD6 78\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (240, 14)\n",
      "loop-loop instances (90, 14)\n",
      "overall instances (330, 14)\n",
      "(((())))\n",
      "ARNT::SP1 SCC 78\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (1930, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (1930, 14)\n",
      "8 +++++++++++++++\n",
      "(((())))\n",
      "PRDM1::RORA GDSD6 58\n",
      "pro-pro instances (2, 14)\n",
      "pro-loop instances (172, 14)\n",
      "loop-loop instances (53, 14)\n",
      "overall instances (227, 14)\n",
      "(((())))\n",
      "ARNT::SP1 SCC 58\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (1627, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (1627, 14)\n",
      "9 +++++++++++++++\n",
      "(((())))\n",
      "BACH1::KLF5 GDSD6 83\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (362, 14)\n",
      "loop-loop instances (156, 14)\n",
      "overall instances (518, 14)\n",
      "(((())))\n",
      "E2F6::SP1 SCC 83\n",
      "pro-pro instances (35, 14)\n",
      "pro-loop instances (2073, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (2108, 14)\n",
      "10 +++++++++++++++\n",
      "(((())))\n",
      "EGR2::FOSL2 GDSD6 210\n",
      "pro-pro instances (31, 14)\n",
      "pro-loop instances (1337, 14)\n",
      "loop-loop instances (693, 14)\n",
      "overall instances (2061, 14)\n",
      "(((())))\n",
      "E2F6::SP1 SCC 210\n",
      "pro-pro instances (86, 14)\n",
      "pro-loop instances (4089, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (4175, 14)\n",
      "11 +++++++++++++++\n",
      "(((())))\n",
      "EGR2::RARG GDSD6 165\n",
      "pro-pro instances (29, 14)\n",
      "pro-loop instances (1041, 14)\n",
      "loop-loop instances (555, 14)\n",
      "overall instances (1625, 14)\n",
      "(((())))\n",
      "E2F6::SP1 SCC 165\n",
      "pro-pro instances (74, 14)\n",
      "pro-loop instances (3709, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (3783, 14)\n",
      "12 +++++++++++++++\n",
      "(((())))\n",
      "EGR2::RORA GDSD6 174\n",
      "pro-pro instances (127, 14)\n",
      "pro-loop instances (1137, 14)\n",
      "loop-loop instances (710, 14)\n",
      "overall instances (1974, 14)\n",
      "(((())))\n",
      "E2F6::SP1 SCC 174\n",
      "pro-pro instances (67, 14)\n",
      "pro-loop instances (3914, 14)\n",
      "loop-loop instances (5, 14)\n",
      "overall instances (3986, 14)\n",
      "13 +++++++++++++++\n",
      "(((())))\n",
      "HMGA1::KLF5 GDSD6 226\n",
      "pro-pro instances (28, 14)\n",
      "pro-loop instances (1134, 14)\n",
      "loop-loop instances (445, 14)\n",
      "overall instances (1607, 14)\n",
      "(((())))\n",
      "E2F6::SP1 SCC 226\n",
      "pro-pro instances (74, 14)\n",
      "pro-loop instances (4683, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (4757, 14)\n",
      "14 +++++++++++++++\n",
      "(((())))\n",
      "KLF5::MAFG GDSD6 132\n",
      "pro-pro instances (32, 14)\n",
      "pro-loop instances (593, 14)\n",
      "loop-loop instances (248, 14)\n",
      "overall instances (873, 14)\n",
      "(((())))\n",
      "E2F6::SP1 SCC 132\n",
      "pro-pro instances (25, 14)\n",
      "pro-loop instances (3270, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (3295, 14)\n",
      "15 +++++++++++++++\n",
      "(((())))\n",
      "KLF5::MAFK GDSD6 120\n",
      "pro-pro instances (14, 14)\n",
      "pro-loop instances (523, 14)\n",
      "loop-loop instances (210, 14)\n",
      "overall instances (747, 14)\n",
      "(((())))\n",
      "E2F6::SP1 SCC 120\n",
      "pro-pro instances (25, 14)\n",
      "pro-loop instances (2549, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (2574, 14)\n",
      "16 +++++++++++++++\n",
      "(((())))\n",
      "MAF::RORA GDSD6 94\n",
      "pro-pro instances (4, 14)\n",
      "pro-loop instances (287, 14)\n",
      "loop-loop instances (109, 14)\n",
      "overall instances (400, 14)\n",
      "(((())))\n",
      "E2F6::SP1 SCC 94\n",
      "pro-pro instances (67, 14)\n",
      "pro-loop instances (2435, 14)\n",
      "loop-loop instances (5, 14)\n",
      "overall instances (2507, 14)\n",
      "17 +++++++++++++++\n",
      "(((())))\n",
      "PRDM1::RORA GDSD6 71\n",
      "pro-pro instances (2, 14)\n",
      "pro-loop instances (199, 14)\n",
      "loop-loop instances (63, 14)\n",
      "overall instances (264, 14)\n",
      "(((())))\n",
      "E2F6::SP1 SCC 71\n",
      "pro-pro instances (67, 14)\n",
      "pro-loop instances (2066, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (2133, 14)\n",
      "18 +++++++++++++++\n",
      "(((())))\n",
      "BACH1::KLF5 GDSD6 80\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (352, 14)\n",
      "loop-loop instances (157, 14)\n",
      "overall instances (509, 14)\n",
      "(((())))\n",
      "EGR1::SP3 SCC 80\n",
      "pro-pro instances (11, 14)\n",
      "pro-loop instances (1389, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (1400, 14)\n",
      "19 +++++++++++++++\n",
      "(((())))\n",
      "EGR2::FOSL2 GDSD6 210\n",
      "pro-pro instances (27, 14)\n",
      "pro-loop instances (1332, 14)\n",
      "loop-loop instances (704, 14)\n",
      "overall instances (2063, 14)\n",
      "(((())))\n",
      "EGR1::SP3 SCC 210\n",
      "pro-pro instances (54, 14)\n",
      "pro-loop instances (2853, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (2907, 14)\n",
      "20 +++++++++++++++\n",
      "(((())))\n",
      "EGR2::RARG GDSD6 165\n",
      "pro-pro instances (29, 14)\n",
      "pro-loop instances (1047, 14)\n",
      "loop-loop instances (574, 14)\n",
      "overall instances (1650, 14)\n",
      "(((())))\n",
      "EGR1::SP3 SCC 165\n",
      "pro-pro instances (22, 14)\n",
      "pro-loop instances (2496, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (2518, 14)\n",
      "21 +++++++++++++++\n",
      "(((())))\n",
      "EGR2::RORA GDSD6 174\n",
      "pro-pro instances (127, 14)\n",
      "pro-loop instances (1142, 14)\n",
      "loop-loop instances (715, 14)\n",
      "overall instances (1984, 14)\n",
      "(((())))\n",
      "EGR1::SP3 SCC 174\n",
      "pro-pro instances (58, 14)\n",
      "pro-loop instances (2637, 14)\n",
      "loop-loop instances (1, 14)\n",
      "overall instances (2696, 14)\n",
      "22 +++++++++++++++\n",
      "(((())))\n",
      "HMGA1::KLF5 GDSD6 223\n",
      "pro-pro instances (28, 14)\n",
      "pro-loop instances (1123, 14)\n",
      "loop-loop instances (453, 14)\n",
      "overall instances (1604, 14)\n",
      "(((())))\n",
      "EGR1::SP3 SCC 223\n",
      "pro-pro instances (95, 14)\n",
      "pro-loop instances (3246, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (3341, 14)\n",
      "23 +++++++++++++++\n",
      "(((())))\n",
      "KLF5::MAFG GDSD6 129\n",
      "pro-pro instances (32, 14)\n",
      "pro-loop instances (581, 14)\n",
      "loop-loop instances (251, 14)\n",
      "overall instances (864, 14)\n",
      "(((())))\n",
      "EGR1::SP3 SCC 129\n",
      "pro-pro instances (64, 14)\n",
      "pro-loop instances (2233, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (2297, 14)\n",
      "24 +++++++++++++++\n",
      "(((())))\n",
      "KLF5::MAFK GDSD6 116\n",
      "pro-pro instances (14, 14)\n",
      "pro-loop instances (510, 14)\n",
      "loop-loop instances (209, 14)\n",
      "overall instances (733, 14)\n",
      "(((())))\n",
      "EGR1::SP3 SCC 116\n",
      "pro-pro instances (26, 14)\n",
      "pro-loop instances (1758, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (1784, 14)\n",
      "25 +++++++++++++++\n",
      "(((())))\n",
      "MAF::RORA GDSD6 93\n",
      "pro-pro instances (4, 14)\n",
      "pro-loop instances (287, 14)\n",
      "loop-loop instances (109, 14)\n",
      "overall instances (400, 14)\n",
      "(((())))\n",
      "EGR1::SP3 SCC 93\n",
      "pro-pro instances (22, 14)\n",
      "pro-loop instances (1610, 14)\n",
      "loop-loop instances (1, 14)\n",
      "overall instances (1633, 14)\n",
      "26 +++++++++++++++\n",
      "(((())))\n",
      "PRDM1::RORA GDSD6 70\n",
      "pro-pro instances (2, 14)\n",
      "pro-loop instances (193, 14)\n",
      "loop-loop instances (67, 14)\n",
      "overall instances (262, 14)\n",
      "(((())))\n",
      "EGR1::SP3 SCC 70\n",
      "pro-pro instances (18, 14)\n",
      "pro-loop instances (1350, 14)\n",
      "loop-loop instances (3, 14)\n",
      "overall instances (1371, 14)\n",
      "27 +++++++++++++++\n",
      "(((())))\n",
      "BACH1::KLF5 GDSD6 77\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (339, 14)\n",
      "loop-loop instances (151, 14)\n",
      "overall instances (490, 14)\n",
      "(((())))\n",
      "KLF16::TFAP2C SCC 77\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (1322, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (1322, 14)\n",
      "28 +++++++++++++++\n",
      "(((())))\n",
      "EGR2::FOSL2 GDSD6 204\n",
      "pro-pro instances (31, 14)\n",
      "pro-loop instances (1304, 14)\n",
      "loop-loop instances (689, 14)\n",
      "overall instances (2024, 14)\n",
      "(((())))\n",
      "KLF16::TFAP2C SCC 204\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (2796, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (2796, 14)\n",
      "29 +++++++++++++++\n",
      "(((())))\n",
      "EGR2::RARG GDSD6 158\n",
      "pro-pro instances (29, 14)\n",
      "pro-loop instances (1004, 14)\n",
      "loop-loop instances (559, 14)\n",
      "overall instances (1592, 14)\n",
      "(((())))\n",
      "KLF16::TFAP2C SCC 158\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (2418, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (2418, 14)\n",
      "30 +++++++++++++++\n",
      "(((())))\n",
      "EGR2::RORA GDSD6 164\n",
      "pro-pro instances (127, 14)\n",
      "pro-loop instances (1090, 14)\n",
      "loop-loop instances (699, 14)\n",
      "overall instances (1916, 14)\n",
      "(((())))\n",
      "KLF16::TFAP2C SCC 164\n",
      "pro-pro instances (46, 14)\n",
      "pro-loop instances (2564, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (2610, 14)\n",
      "31 +++++++++++++++\n",
      "(((())))\n",
      "HMGA1::KLF5 GDSD6 214\n",
      "pro-pro instances (28, 14)\n",
      "pro-loop instances (1086, 14)\n",
      "loop-loop instances (440, 14)\n",
      "overall instances (1554, 14)\n",
      "(((())))\n",
      "KLF16::TFAP2C SCC 214\n",
      "pro-pro instances (62, 14)\n",
      "pro-loop instances (3181, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (3243, 14)\n",
      "32 +++++++++++++++\n",
      "(((())))\n",
      "KLF5::MAFG GDSD6 131\n",
      "pro-pro instances (32, 14)\n",
      "pro-loop instances (583, 14)\n",
      "loop-loop instances (251, 14)\n",
      "overall instances (866, 14)\n",
      "(((())))\n",
      "KLF16::TFAP2C SCC 131\n",
      "pro-pro instances (46, 14)\n",
      "pro-loop instances (2210, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (2256, 14)\n",
      "33 +++++++++++++++\n",
      "(((())))\n",
      "KLF5::MAFK GDSD6 115\n",
      "pro-pro instances (14, 14)\n",
      "pro-loop instances (502, 14)\n",
      "loop-loop instances (209, 14)\n",
      "overall instances (725, 14)\n",
      "(((())))\n",
      "KLF16::TFAP2C SCC 115\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (1718, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (1718, 14)\n",
      "34 +++++++++++++++\n",
      "(((())))\n",
      "MAF::RORA GDSD6 90\n",
      "pro-pro instances (4, 14)\n",
      "pro-loop instances (277, 14)\n",
      "loop-loop instances (108, 14)\n",
      "overall instances (389, 14)\n",
      "(((())))\n",
      "KLF16::TFAP2C SCC 90\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (1573, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (1573, 14)\n",
      "35 +++++++++++++++\n",
      "(((())))\n",
      "PRDM1::RORA GDSD6 69\n",
      "pro-pro instances (2, 14)\n",
      "pro-loop instances (192, 14)\n",
      "loop-loop instances (66, 14)\n",
      "overall instances (260, 14)\n",
      "(((())))\n",
      "KLF16::TFAP2C SCC 69\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (1351, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (1351, 14)\n",
      "36 +++++++++++++++\n",
      "(((())))\n",
      "BACH1::KLF5 GDSD6 85\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (364, 14)\n",
      "loop-loop instances (159, 14)\n",
      "overall instances (523, 14)\n",
      "(((())))\n",
      "KLF4::KLF5 SCC 85\n",
      "pro-pro instances (8, 14)\n",
      "pro-loop instances (1471, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (1479, 14)\n",
      "37 +++++++++++++++\n",
      "(((())))\n",
      "EGR2::FOSL2 GDSD6 215\n",
      "pro-pro instances (31, 14)\n",
      "pro-loop instances (1360, 14)\n",
      "loop-loop instances (714, 14)\n",
      "overall instances (2105, 14)\n",
      "(((())))\n",
      "KLF4::KLF5 SCC 215\n",
      "pro-pro instances (90, 14)\n",
      "pro-loop instances (2913, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (3003, 14)\n",
      "38 +++++++++++++++\n",
      "(((())))\n",
      "EGR2::RARG GDSD6 171\n",
      "pro-pro instances (29, 14)\n",
      "pro-loop instances (1074, 14)\n",
      "loop-loop instances (576, 14)\n",
      "overall instances (1679, 14)\n",
      "(((())))\n",
      "KLF4::KLF5 SCC 171\n",
      "pro-pro instances (92, 14)\n",
      "pro-loop instances (2590, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (2682, 14)\n",
      "39 +++++++++++++++\n",
      "(((())))\n",
      "EGR2::RORA GDSD6 173\n",
      "pro-pro instances (127, 14)\n",
      "pro-loop instances (1138, 14)\n",
      "loop-loop instances (715, 14)\n",
      "overall instances (1980, 14)\n",
      "(((())))\n",
      "KLF4::KLF5 SCC 173\n",
      "pro-pro instances (62, 14)\n",
      "pro-loop instances (2633, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (2695, 14)\n",
      "40 +++++++++++++++\n",
      "(((())))\n",
      "HMGA1::KLF5 GDSD6 229\n",
      "pro-pro instances (28, 14)\n",
      "pro-loop instances (1149, 14)\n",
      "loop-loop instances (463, 14)\n",
      "overall instances (1640, 14)\n",
      "(((())))\n",
      "KLF4::KLF5 SCC 229\n",
      "pro-pro instances (110, 14)\n",
      "pro-loop instances (3409, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (3519, 14)\n",
      "41 +++++++++++++++\n",
      "(((())))\n",
      "KLF5::MAFG GDSD6 132\n",
      "pro-pro instances (32, 14)\n",
      "pro-loop instances (592, 14)\n",
      "loop-loop instances (252, 14)\n",
      "overall instances (876, 14)\n",
      "(((())))\n",
      "KLF4::KLF5 SCC 132\n",
      "pro-pro instances (65, 14)\n",
      "pro-loop instances (2345, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (2410, 14)\n",
      "42 +++++++++++++++\n",
      "(((())))\n",
      "KLF5::MAFK GDSD6 121\n",
      "pro-pro instances (14, 14)\n",
      "pro-loop instances (522, 14)\n",
      "loop-loop instances (214, 14)\n",
      "overall instances (750, 14)\n",
      "(((())))\n",
      "KLF4::KLF5 SCC 121\n",
      "pro-pro instances (30, 14)\n",
      "pro-loop instances (1859, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (1889, 14)\n",
      "43 +++++++++++++++\n",
      "(((())))\n",
      "MAF::RORA GDSD6 93\n",
      "pro-pro instances (4, 14)\n",
      "pro-loop instances (284, 14)\n",
      "loop-loop instances (107, 14)\n",
      "overall instances (395, 14)\n",
      "(((())))\n",
      "KLF4::KLF5 SCC 93\n",
      "pro-pro instances (24, 14)\n",
      "pro-loop instances (1711, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (1735, 14)\n",
      "44 +++++++++++++++\n",
      "(((())))\n",
      "PRDM1::RORA GDSD6 71\n",
      "pro-pro instances (2, 14)\n",
      "pro-loop instances (198, 14)\n",
      "loop-loop instances (66, 14)\n",
      "overall instances (266, 14)\n",
      "(((())))\n",
      "KLF4::KLF5 SCC 71\n",
      "pro-pro instances (28, 14)\n",
      "pro-loop instances (1388, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (1416, 14)\n",
      "45 +++++++++++++++\n",
      "(((())))\n",
      "BACH1::KLF5 GDSD6 86\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (367, 14)\n",
      "loop-loop instances (159, 14)\n",
      "overall instances (526, 14)\n",
      "(((())))\n",
      "MAZ::SP1 SCC 86\n",
      "pro-pro instances (115, 14)\n",
      "pro-loop instances (4313, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (4428, 14)\n",
      "46 +++++++++++++++\n",
      "(((())))\n",
      "EGR2::FOSL2 GDSD6 221\n",
      "pro-pro instances (31, 14)\n",
      "pro-loop instances (1383, 14)\n",
      "loop-loop instances (716, 14)\n",
      "overall instances (2130, 14)\n",
      "(((())))\n",
      "MAZ::SP1 SCC 221\n",
      "pro-pro instances (625, 14)\n",
      "pro-loop instances (8778, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (9403, 14)\n",
      "47 +++++++++++++++\n",
      "(((())))\n",
      "EGR2::RARG GDSD6 169\n",
      "pro-pro instances (29, 14)\n",
      "pro-loop instances (1064, 14)\n",
      "loop-loop instances (575, 14)\n",
      "overall instances (1668, 14)\n",
      "(((())))\n",
      "MAZ::SP1 SCC 169\n",
      "pro-pro instances (599, 14)\n",
      "pro-loop instances (7760, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (8359, 14)\n",
      "48 +++++++++++++++\n",
      "(((())))\n",
      "EGR2::RORA GDSD6 177\n",
      "pro-pro instances (127, 14)\n",
      "pro-loop instances (1152, 14)\n",
      "loop-loop instances (722, 14)\n",
      "overall instances (2001, 14)\n",
      "(((())))\n",
      "MAZ::SP1 SCC 177\n",
      "pro-pro instances (547, 14)\n",
      "pro-loop instances (8231, 14)\n",
      "loop-loop instances (8, 14)\n",
      "overall instances (8786, 14)\n",
      "49 +++++++++++++++\n",
      "(((())))\n",
      "HMGA1::KLF5 GDSD6 234\n",
      "pro-pro instances (28, 14)\n",
      "pro-loop instances (1163, 14)\n",
      "loop-loop instances (468, 14)\n",
      "overall instances (1659, 14)\n",
      "(((())))\n",
      "MAZ::SP1 SCC 234\n",
      "pro-pro instances (830, 14)\n",
      "pro-loop instances (10020, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (10850, 14)\n",
      "50 +++++++++++++++\n",
      "(((())))\n",
      "KLF5::MAFG GDSD6 134\n",
      "pro-pro instances (32, 14)\n",
      "pro-loop instances (595, 14)\n",
      "loop-loop instances (253, 14)\n",
      "overall instances (880, 14)\n",
      "(((())))\n",
      "MAZ::SP1 SCC 134\n",
      "pro-pro instances (466, 14)\n",
      "pro-loop instances (6863, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (7329, 14)\n",
      "51 +++++++++++++++\n",
      "(((())))\n",
      "KLF5::MAFK GDSD6 122\n",
      "pro-pro instances (14, 14)\n",
      "pro-loop instances (525, 14)\n",
      "loop-loop instances (215, 14)\n",
      "overall instances (754, 14)\n",
      "(((())))\n",
      "MAZ::SP1 SCC 122\n",
      "pro-pro instances (137, 14)\n",
      "pro-loop instances (5404, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (5541, 14)\n",
      "52 +++++++++++++++\n",
      "(((())))\n",
      "MAF::RORA GDSD6 95\n",
      "pro-pro instances (4, 14)\n",
      "pro-loop instances (291, 14)\n",
      "loop-loop instances (109, 14)\n",
      "overall instances (404, 14)\n",
      "(((())))\n",
      "MAZ::SP1 SCC 95\n",
      "pro-pro instances (205, 14)\n",
      "pro-loop instances (5073, 14)\n",
      "loop-loop instances (8, 14)\n",
      "overall instances (5286, 14)\n",
      "53 +++++++++++++++\n",
      "(((())))\n",
      "PRDM1::RORA GDSD6 74\n",
      "pro-pro instances (2, 14)\n",
      "pro-loop instances (206, 14)\n",
      "loop-loop instances (67, 14)\n",
      "overall instances (275, 14)\n",
      "(((())))\n",
      "MAZ::SP1 SCC 74\n",
      "pro-pro instances (201, 14)\n",
      "pro-loop instances (4393, 14)\n",
      "loop-loop instances (9, 14)\n",
      "overall instances (4603, 14)\n",
      "54 +++++++++++++++\n",
      "(((())))\n",
      "BACH1::KLF5 GDSD6 75\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (316, 14)\n",
      "loop-loop instances (153, 14)\n",
      "overall instances (469, 14)\n",
      "(((())))\n",
      "SP1::TFDP1 SCC 75\n",
      "pro-pro instances (37, 14)\n",
      "pro-loop instances (1829, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (1866, 14)\n",
      "55 +++++++++++++++\n",
      "(((())))\n",
      "EGR2::FOSL2 GDSD6 192\n",
      "pro-pro instances (31, 14)\n",
      "pro-loop instances (1197, 14)\n",
      "loop-loop instances (649, 14)\n",
      "overall instances (1877, 14)\n",
      "(((())))\n",
      "SP1::TFDP1 SCC 192\n",
      "pro-pro instances (75, 14)\n",
      "pro-loop instances (3520, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (3595, 14)\n",
      "56 +++++++++++++++\n",
      "(((())))\n",
      "EGR2::RARG GDSD6 152\n",
      "pro-pro instances (29, 14)\n",
      "pro-loop instances (944, 14)\n",
      "loop-loop instances (538, 14)\n",
      "overall instances (1511, 14)\n",
      "(((())))\n",
      "SP1::TFDP1 SCC 152\n",
      "pro-pro instances (108, 14)\n",
      "pro-loop instances (3297, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (3405, 14)\n",
      "57 +++++++++++++++\n",
      "(((())))\n",
      "EGR2::RORA GDSD6 157\n",
      "pro-pro instances (127, 14)\n",
      "pro-loop instances (1044, 14)\n",
      "loop-loop instances (685, 14)\n",
      "overall instances (1856, 14)\n",
      "(((())))\n",
      "SP1::TFDP1 SCC 157\n",
      "pro-pro instances (71, 14)\n",
      "pro-loop instances (3414, 14)\n",
      "loop-loop instances (4, 14)\n",
      "overall instances (3489, 14)\n",
      "58 +++++++++++++++\n",
      "(((())))\n",
      "HMGA1::KLF5 GDSD6 201\n",
      "pro-pro instances (28, 14)\n",
      "pro-loop instances (1002, 14)\n",
      "loop-loop instances (413, 14)\n",
      "overall instances (1443, 14)\n",
      "(((())))\n",
      "SP1::TFDP1 SCC 201\n",
      "pro-pro instances (31, 14)\n",
      "pro-loop instances (4106, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (4137, 14)\n",
      "59 +++++++++++++++\n",
      "(((())))\n",
      "KLF5::MAFG GDSD6 122\n",
      "pro-pro instances (32, 14)\n",
      "pro-loop instances (535, 14)\n",
      "loop-loop instances (240, 14)\n",
      "overall instances (807, 14)\n",
      "(((())))\n",
      "SP1::TFDP1 SCC 122\n",
      "pro-pro instances (45, 14)\n",
      "pro-loop instances (2895, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (2940, 14)\n",
      "60 +++++++++++++++\n",
      "(((())))\n",
      "KLF5::MAFK GDSD6 105\n",
      "pro-pro instances (14, 14)\n",
      "pro-loop instances (462, 14)\n",
      "loop-loop instances (187, 14)\n",
      "overall instances (663, 14)\n",
      "(((())))\n",
      "SP1::TFDP1 SCC 105\n",
      "pro-pro instances (38, 14)\n",
      "pro-loop instances (2177, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (2215, 14)\n",
      "61 +++++++++++++++\n",
      "(((())))\n",
      "MAF::RORA GDSD6 82\n",
      "pro-pro instances (4, 14)\n",
      "pro-loop instances (235, 14)\n",
      "loop-loop instances (95, 14)\n",
      "overall instances (334, 14)\n",
      "(((())))\n",
      "SP1::TFDP1 SCC 82\n",
      "pro-pro instances (38, 14)\n",
      "pro-loop instances (2092, 14)\n",
      "loop-loop instances (4, 14)\n",
      "overall instances (2134, 14)\n",
      "62 +++++++++++++++\n",
      "(((())))\n",
      "PRDM1::RORA GDSD6 63\n",
      "pro-pro instances (2, 14)\n",
      "pro-loop instances (175, 14)\n",
      "loop-loop instances (59, 14)\n",
      "overall instances (236, 14)\n",
      "(((())))\n",
      "SP1::TFDP1 SCC 63\n",
      "pro-pro instances (57, 14)\n",
      "pro-loop instances (1796, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (1853, 14)\n",
      "63 +++++++++++++++\n",
      "(((())))\n",
      "BACH1::KLF5 GDSD6 72\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (301, 14)\n",
      "loop-loop instances (142, 14)\n",
      "overall instances (443, 14)\n",
      "(((())))\n",
      "SP1::ZBTB14 SCC 72\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (1746, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (1746, 14)\n",
      "64 +++++++++++++++\n",
      "(((())))\n",
      "EGR2::FOSL2 GDSD6 174\n",
      "pro-pro instances (31, 14)\n",
      "pro-loop instances (1125, 14)\n",
      "loop-loop instances (596, 14)\n",
      "overall instances (1752, 14)\n",
      "(((())))\n",
      "SP1::ZBTB14 SCC 174\n",
      "pro-pro instances (25, 14)\n",
      "pro-loop instances (3360, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (3385, 14)\n",
      "65 +++++++++++++++\n",
      "(((())))\n",
      "EGR2::RARG GDSD6 135\n",
      "pro-pro instances (23, 14)\n",
      "pro-loop instances (888, 14)\n",
      "loop-loop instances (464, 14)\n",
      "overall instances (1375, 14)\n",
      "(((())))\n",
      "SP1::ZBTB14 SCC 135\n",
      "pro-pro instances (32, 14)\n",
      "pro-loop instances (2993, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (3025, 14)\n",
      "66 +++++++++++++++\n",
      "(((())))\n",
      "EGR2::RORA GDSD6 146\n",
      "pro-pro instances (85, 14)\n",
      "pro-loop instances (993, 14)\n",
      "loop-loop instances (605, 14)\n",
      "overall instances (1683, 14)\n",
      "(((())))\n",
      "SP1::ZBTB14 SCC 146\n",
      "pro-pro instances (79, 14)\n",
      "pro-loop instances (3187, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (3266, 14)\n",
      "67 +++++++++++++++\n",
      "(((())))\n",
      "HMGA1::KLF5 GDSD6 183\n",
      "pro-pro instances (28, 14)\n",
      "pro-loop instances (910, 14)\n",
      "loop-loop instances (382, 14)\n",
      "overall instances (1320, 14)\n",
      "(((())))\n",
      "SP1::ZBTB14 SCC 183\n",
      "pro-pro instances (129, 14)\n",
      "pro-loop instances (3714, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (3843, 14)\n",
      "68 +++++++++++++++\n",
      "(((())))\n",
      "KLF5::MAFG GDSD6 111\n",
      "pro-pro instances (32, 14)\n",
      "pro-loop instances (494, 14)\n",
      "loop-loop instances (222, 14)\n",
      "overall instances (748, 14)\n",
      "(((())))\n",
      "SP1::ZBTB14 SCC 111\n",
      "pro-pro instances (102, 14)\n",
      "pro-loop instances (2748, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (2850, 14)\n",
      "69 +++++++++++++++\n",
      "(((())))\n",
      "KLF5::MAFK GDSD6 102\n",
      "pro-pro instances (14, 14)\n",
      "pro-loop instances (442, 14)\n",
      "loop-loop instances (183, 14)\n",
      "overall instances (639, 14)\n",
      "(((())))\n",
      "SP1::ZBTB14 SCC 102\n",
      "pro-pro instances (9, 14)\n",
      "pro-loop instances (2138, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (2147, 14)\n",
      "70 +++++++++++++++\n",
      "(((())))\n",
      "MAF::RORA GDSD6 77\n",
      "pro-pro instances (4, 14)\n",
      "pro-loop instances (231, 14)\n",
      "loop-loop instances (95, 14)\n",
      "overall instances (330, 14)\n",
      "(((())))\n",
      "SP1::ZBTB14 SCC 77\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (1979, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (1979, 14)\n",
      "71 +++++++++++++++\n",
      "(((())))\n",
      "PRDM1::RORA GDSD6 61\n",
      "pro-pro instances (2, 14)\n",
      "pro-loop instances (174, 14)\n",
      "loop-loop instances (60, 14)\n",
      "overall instances (236, 14)\n",
      "(((())))\n",
      "SP1::ZBTB14 SCC 61\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (1648, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (1648, 14)\n",
      "72 +++++++++++++++\n",
      "(((())))\n",
      "BACH1::KLF5 GDSD6 84\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (363, 14)\n",
      "loop-loop instances (157, 14)\n",
      "overall instances (520, 14)\n",
      "(((())))\n",
      "SP1::ZBTB17 SCC 84\n",
      "pro-pro instances (71, 14)\n",
      "pro-loop instances (2710, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (2781, 14)\n",
      "73 +++++++++++++++\n",
      "(((())))\n",
      "EGR2::FOSL2 GDSD6 219\n",
      "pro-pro instances (31, 14)\n",
      "pro-loop instances (1378, 14)\n",
      "loop-loop instances (715, 14)\n",
      "overall instances (2124, 14)\n",
      "(((())))\n",
      "SP1::ZBTB17 SCC 219\n",
      "pro-pro instances (227, 14)\n",
      "pro-loop instances (5511, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (5738, 14)\n",
      "74 +++++++++++++++\n",
      "(((())))\n",
      "EGR2::RARG GDSD6 168\n",
      "pro-pro instances (29, 14)\n",
      "pro-loop instances (1062, 14)\n",
      "loop-loop instances (575, 14)\n",
      "overall instances (1666, 14)\n",
      "(((())))\n",
      "SP1::ZBTB17 SCC 168\n",
      "pro-pro instances (333, 14)\n",
      "pro-loop instances (4889, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (5222, 14)\n",
      "75 +++++++++++++++\n",
      "(((())))\n",
      "EGR2::RORA GDSD6 177\n",
      "pro-pro instances (127, 14)\n",
      "pro-loop instances (1150, 14)\n",
      "loop-loop instances (722, 14)\n",
      "overall instances (1999, 14)\n",
      "(((())))\n",
      "SP1::ZBTB17 SCC 177\n",
      "pro-pro instances (311, 14)\n",
      "pro-loop instances (5152, 14)\n",
      "loop-loop instances (3, 14)\n",
      "overall instances (5466, 14)\n",
      "76 +++++++++++++++\n",
      "(((())))\n",
      "HMGA1::KLF5 GDSD6 232\n",
      "pro-pro instances (28, 14)\n",
      "pro-loop instances (1158, 14)\n",
      "loop-loop instances (463, 14)\n",
      "overall instances (1649, 14)\n",
      "(((())))\n",
      "SP1::ZBTB17 SCC 232\n",
      "pro-pro instances (408, 14)\n",
      "pro-loop instances (6254, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (6662, 14)\n",
      "77 +++++++++++++++\n",
      "(((())))\n",
      "KLF5::MAFG GDSD6 133\n",
      "pro-pro instances (32, 14)\n",
      "pro-loop instances (593, 14)\n",
      "loop-loop instances (253, 14)\n",
      "overall instances (878, 14)\n",
      "(((())))\n",
      "SP1::ZBTB17 SCC 133\n",
      "pro-pro instances (251, 14)\n",
      "pro-loop instances (4294, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (4545, 14)\n",
      "78 +++++++++++++++\n",
      "(((())))\n",
      "KLF5::MAFK GDSD6 120\n",
      "pro-pro instances (14, 14)\n",
      "pro-loop instances (520, 14)\n",
      "loop-loop instances (211, 14)\n",
      "overall instances (745, 14)\n",
      "(((())))\n",
      "SP1::ZBTB17 SCC 120\n",
      "pro-pro instances (69, 14)\n",
      "pro-loop instances (3377, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (3446, 14)\n",
      "79 +++++++++++++++\n",
      "(((())))\n",
      "MAF::RORA GDSD6 95\n",
      "pro-pro instances (4, 14)\n",
      "pro-loop instances (291, 14)\n",
      "loop-loop instances (109, 14)\n",
      "overall instances (404, 14)\n",
      "(((())))\n",
      "SP1::ZBTB17 SCC 95\n",
      "pro-pro instances (113, 14)\n",
      "pro-loop instances (3194, 14)\n",
      "loop-loop instances (3, 14)\n",
      "overall instances (3310, 14)\n",
      "80 +++++++++++++++\n",
      "(((())))\n",
      "PRDM1::RORA GDSD6 74\n",
      "pro-pro instances (2, 14)\n",
      "pro-loop instances (206, 14)\n",
      "loop-loop instances (67, 14)\n",
      "overall instances (275, 14)\n",
      "(((())))\n",
      "SP1::ZBTB17 SCC 74\n",
      "pro-pro instances (108, 14)\n",
      "pro-loop instances (2757, 14)\n",
      "loop-loop instances (10, 14)\n",
      "overall instances (2875, 14)\n",
      "CPU times: user 28min 18s, sys: 1min 20s, total: 29min 38s\n",
      "Wall time: 20min 24s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# takes awhile\n",
    "for idx, row in intersect_genes_KC_SCC_long.iterrows():\n",
    "    print(idx, '+++++++++++++++')\n",
    "    get_genomic_instances_all(row.KC_vocab, 'GDSD6', row.intersect_genes.split('|'),\n",
    "                              verbose=False, \n",
    "                              save_prefix='../data/processed/fig7_vocab_go/genomic_instance_overlap/'+str(idx)+'_')\n",
    "    get_genomic_instances_all(row.SCC_vocab, 'SCC', row.intersect_genes.split('|'),\n",
    "                              verbose=False, \n",
    "                              save_prefix='../data/processed/fig7_vocab_go/genomic_instance_overlap/'+str(idx)+'_')\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# update 7/24/2021 - see if we can see how many genomic instances we can get for each vocab word in the normal and cancer form \n",
    "\n",
    "use folder `genomic_instance_normal_cancer_comp`\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ARID5B::LEF1 MM\n",
      "(((())))\n",
      "ARID5B::LEF1 Melanocytes 13\n",
      "loading from file\n",
      "overall instances (4024, 14)\n",
      "(((())))\n",
      "ARID5B::LEF1 MM 464\n",
      "loading from file\n",
      "overall instances (574639, 14)\n",
      "vocab                     ARID5B::LEF1\n",
      "tissues_to_check        Melanocytes|MM\n",
      "num_normal_instances              4024\n",
      "num_cancer_instances            574639\n",
      "Name: 1, dtype: object\n",
      "ARNT::SP1 SCC\n",
      "(((())))\n",
      "ARNT::SP1 GDSD6 91\n",
      "loading from file\n",
      "overall instances (104879, 14)\n",
      "(((())))\n",
      "ARNT::SP1 SCC 828\n",
      "loading from file\n",
      "overall instances (897199, 14)\n",
      "vocab                   ARNT::SP1\n",
      "tissues_to_check        GDSD6|SCC\n",
      "num_normal_instances       104879\n",
      "num_cancer_instances       897199\n",
      "Name: 5, dtype: object\n",
      "BACH1::KLF5 GDSD6\n",
      "(((())))\n",
      "BACH1::KLF5 GDSD6 100\n",
      "loading from file\n",
      "overall instances (58058, 14)\n",
      "(((())))\n",
      "BACH1::KLF5 SCC 988\n",
      "loading from file\n",
      "overall instances (572001, 14)\n",
      "vocab                   BACH1::KLF5\n",
      "tissues_to_check          GDSD6|SCC\n",
      "num_normal_instances          58058\n",
      "num_cancer_instances         572001\n",
      "Name: 8, dtype: object\n",
      "E2F6::SP1 SCC\n",
      "(((())))\n",
      "E2F6::SP1 GDSD6 274\n",
      "loading from file\n",
      "overall instances (360936, 14)\n",
      "(((())))\n",
      "E2F6::SP1 SCC 1385\n",
      "loading from file\n",
      "overall instances (1776878, 14)\n",
      "vocab                     E2F6::SP1\n",
      "tissues_to_check          GDSD6|SCC\n",
      "num_normal_instances         360936\n",
      "num_cancer_instances    1.77688e+06\n",
      "Name: 18, dtype: object\n",
      "EGR1::SP3 SCC\n",
      "(((())))\n",
      "EGR1::SP3 GDSD6 327\n",
      "loading from file\n",
      "overall instances (319286, 14)\n",
      "(((())))\n",
      "EGR1::SP3 SCC 1432\n",
      "loading from file\n",
      "overall instances (1334645, 14)\n",
      "vocab                     EGR1::SP3\n",
      "tissues_to_check          GDSD6|SCC\n",
      "num_normal_instances         319286\n",
      "num_cancer_instances    1.33464e+06\n",
      "Name: 25, dtype: object\n",
      "EGR2::FOSL2 GDSD6\n",
      "(((())))\n",
      "EGR2::FOSL2 GDSD6 255\n",
      "loading from file\n",
      "overall instances (260414, 14)\n",
      "(((())))\n",
      "EGR2::FOSL2 SCC 1164\n",
      "loading from file\n",
      "overall instances (1184169, 14)\n",
      "vocab                   EGR2::FOSL2\n",
      "tissues_to_check          GDSD6|SCC\n",
      "num_normal_instances         260414\n",
      "num_cancer_instances    1.18417e+06\n",
      "Name: 26, dtype: object\n",
      "EGR2::NFE2L1 MM\n",
      "(((())))\n",
      "EGR2::NFE2L1 MM 610\n",
      "loading from file\n",
      "overall instances (565239, 14)\n",
      "vocab                     EGR2::NFE2L1\n",
      "tissues_to_check        Melanocytes|MM\n",
      "num_normal_instances                 0\n",
      "num_cancer_instances            565239\n",
      "Name: 31, dtype: object\n",
      "EGR2::RARG GDSD6\n",
      "(((())))\n",
      "EGR2::RARG GDSD6 196\n",
      "loading from file\n",
      "overall instances (185825, 14)\n",
      "(((())))\n",
      "EGR2::RARG SCC 1103\n",
      "loading from file\n",
      "overall instances (1081153, 14)\n",
      "vocab                    EGR2::RARG\n",
      "tissues_to_check          GDSD6|SCC\n",
      "num_normal_instances         185825\n",
      "num_cancer_instances    1.08115e+06\n",
      "Name: 32, dtype: object\n",
      "EGR2::RORA GDSD6\n",
      "(((())))\n",
      "EGR2::RORA GDSD6 205\n",
      "loading from file\n",
      "overall instances (190733, 14)\n",
      "vocab                   EGR2::RORA\n",
      "tissues_to_check         GDSD6|SCC\n",
      "num_normal_instances        190733\n",
      "num_cancer_instances             0\n",
      "Name: 33, dtype: object\n",
      "EN2::STAT1 MM\n",
      "(((())))\n",
      "EN2::STAT1 Melanocytes 3\n",
      "loading from file\n",
      "overall instances (455, 14)\n",
      "(((())))\n",
      "EN2::STAT1 MM 459\n",
      "loading from file\n",
      "overall instances (204327, 14)\n",
      "vocab                       EN2::STAT1\n",
      "tissues_to_check        Melanocytes|MM\n",
      "num_normal_instances               455\n",
      "num_cancer_instances            204327\n",
      "Name: 43, dtype: object\n",
      "ETV2::PRRX1 Melanocytes\n",
      "(((())))\n",
      "ETV2::PRRX1 Melanocytes 45\n",
      "loading from file\n",
      "overall instances (59667, 14)\n",
      "(((())))\n",
      "ETV2::PRRX1 MM 553\n",
      "loading from file\n",
      "overall instances (2419451, 14)\n",
      "vocab                      ETV2::PRRX1\n",
      "tissues_to_check        Melanocytes|MM\n",
      "num_normal_instances             59667\n",
      "num_cancer_instances       2.41945e+06\n",
      "Name: 47, dtype: object\n",
      "FOXD3::PRRX1 Melanocytes\n",
      "(((())))\n",
      "FOXD3::PRRX1 Melanocytes 21\n",
      "loading from file\n",
      "overall instances (26272, 14)\n",
      "(((())))\n",
      "FOXD3::PRRX1 MM 366\n",
      "loading from file\n",
      "overall instances (1580530, 14)\n",
      "vocab                     FOXD3::PRRX1\n",
      "tissues_to_check        Melanocytes|MM\n",
      "num_normal_instances             26272\n",
      "num_cancer_instances       1.58053e+06\n",
      "Name: 54, dtype: object\n",
      "HBP1::STAT1 MM\n",
      "(((())))\n",
      "HBP1::STAT1 Melanocytes 12\n",
      "loading from file\n",
      "overall instances (2085, 14)\n",
      "(((())))\n",
      "HBP1::STAT1 MM 467\n",
      "loading from file\n",
      "overall instances (192619, 14)\n",
      "vocab                      HBP1::STAT1\n",
      "tissues_to_check        Melanocytes|MM\n",
      "num_normal_instances              2085\n",
      "num_cancer_instances            192619\n",
      "Name: 66, dtype: object\n",
      "HMGA1::KLF5 GDSD6\n",
      "(((())))\n",
      "HMGA1::KLF5 GDSD6 284\n",
      "loading from file\n",
      "overall instances (226721, 14)\n",
      "(((())))\n",
      "HMGA1::KLF5 SCC 1444\n",
      "loading from file\n",
      "overall instances (1255611, 14)\n",
      "vocab                   HMGA1::KLF5\n",
      "tissues_to_check          GDSD6|SCC\n",
      "num_normal_instances         226721\n",
      "num_cancer_instances    1.25561e+06\n",
      "Name: 67, dtype: object\n",
      "JUNB::RORA GDSD6\n",
      "(((())))\n",
      "JUNB::RORA GDSD6 120\n",
      "loading from file\n",
      "overall instances (34386, 14)\n",
      "(((())))\n",
      "JUNB::RORA SCC 491\n",
      "loading from file\n",
      "overall instances (150544, 14)\n",
      "vocab                   JUNB::RORA\n",
      "tissues_to_check         GDSD6|SCC\n",
      "num_normal_instances         34386\n",
      "num_cancer_instances        150544\n",
      "Name: 103, dtype: object\n",
      "KLF15::ZNF589 Melanocytes\n",
      "(((())))\n",
      "KLF15::ZNF589 Melanocytes 101\n",
      "loading from file\n",
      "overall instances (95646, 14)\n",
      "vocab                    KLF15::ZNF589\n",
      "tissues_to_check        Melanocytes|MM\n",
      "num_normal_instances             95646\n",
      "num_cancer_instances                 0\n",
      "Name: 117, dtype: object\n",
      "KLF16::TFAP2C SCC\n",
      "(((())))\n",
      "KLF16::TFAP2C GDSD6 191\n",
      "loading from file\n",
      "overall instances (152861, 14)\n",
      "(((())))\n",
      "KLF16::TFAP2C SCC 1220\n",
      "loading from file\n",
      "overall instances (1042609, 14)\n",
      "vocab                   KLF16::TFAP2C\n",
      "tissues_to_check            GDSD6|SCC\n",
      "num_normal_instances           152861\n",
      "num_cancer_instances      1.04261e+06\n",
      "Name: 120, dtype: object\n",
      "KLF4::KLF5 SCC\n",
      "(((())))\n",
      "KLF4::KLF5 GDSD6 432\n",
      "loading from file\n",
      "overall instances (413687, 14)\n",
      "(((())))\n",
      "KLF4::KLF5 SCC 1509\n",
      "loading from file\n",
      "overall instances (1378127, 14)\n",
      "vocab                    KLF4::KLF5\n",
      "tissues_to_check          GDSD6|SCC\n",
      "num_normal_instances         413687\n",
      "num_cancer_instances    1.37813e+06\n",
      "Name: 121, dtype: object\n",
      "KLF5::MAFG GDSD6\n",
      "(((())))\n",
      "KLF5::MAFG GDSD6 159\n",
      "loading from file\n",
      "overall instances (94830, 14)\n",
      "(((())))\n",
      "KLF5::MAFG SCC 1131\n",
      "loading from file\n",
      "overall instances (697623, 14)\n",
      "vocab                   KLF5::MAFG\n",
      "tissues_to_check         GDSD6|SCC\n",
      "num_normal_instances         94830\n",
      "num_cancer_instances        697623\n",
      "Name: 128, dtype: object\n",
      "KLF5::MAFK GDSD6\n",
      "(((())))\n",
      "KLF5::MAFK GDSD6 144\n",
      "loading from file\n",
      "overall instances (87643, 14)\n",
      "(((())))\n",
      "KLF5::MAFK SCC 1094\n",
      "loading from file\n",
      "overall instances (668822, 14)\n",
      "vocab                   KLF5::MAFK\n",
      "tissues_to_check         GDSD6|SCC\n",
      "num_normal_instances         87643\n",
      "num_cancer_instances        668822\n",
      "Name: 129, dtype: object\n",
      "KLF5::NFYB Melanocytes\n",
      "(((())))\n",
      "KLF5::NFYB Melanocytes 29\n",
      "loading from file\n",
      "overall instances (11886, 14)\n",
      "(((())))\n",
      "KLF5::NFYB MM 173\n",
      "loading from file\n",
      "overall instances (90189, 14)\n",
      "vocab                       KLF5::NFYB\n",
      "tissues_to_check        Melanocytes|MM\n",
      "num_normal_instances             11886\n",
      "num_cancer_instances             90189\n",
      "Name: 131, dtype: object\n",
      "MAF::RORA GDSD6\n",
      "(((())))\n",
      "MAF::RORA GDSD6 109\n",
      "loading from file\n",
      "overall instances (31265, 14)\n",
      "vocab                   MAF::RORA\n",
      "tissues_to_check        GDSD6|SCC\n",
      "num_normal_instances        31265\n",
      "num_cancer_instances            0\n",
      "Name: 149, dtype: object\n",
      "MAZ::SP1 SCC\n",
      "(((())))\n",
      "MAZ::SP1 GDSD6 473\n",
      "loading from file\n",
      "overall instances (1231874, 14)\n",
      "(((())))\n",
      "MAZ::SP1 SCC 1583\n",
      "(((())))\n",
      "MAZ::SP1 SCC 1583\n",
      "pro-pro instances (2876, 14)\n",
      "pro-loop instances (3961156, 14)\n",
      "loop-loop instances (133, 14)\n",
      "overall instances (3964165, 14)\n",
      "vocab                      MAZ::SP1\n",
      "tissues_to_check          GDSD6|SCC\n",
      "num_normal_instances    1.23187e+06\n",
      "num_cancer_instances    3.96416e+06\n",
      "Name: 153, dtype: object\n",
      "MITF::ZNF589 Melanocytes\n",
      "(((())))\n",
      "MITF::ZNF589 Melanocytes 14\n",
      "(((())))\n",
      "MITF::ZNF589 Melanocytes 14\n",
      "pro-pro instances (11, 14)\n",
      "pro-loop instances (7016, 14)\n",
      "loop-loop instances (49, 14)\n",
      "overall instances (7076, 14)\n",
      "vocab                     MITF::ZNF589\n",
      "tissues_to_check        Melanocytes|MM\n",
      "num_normal_instances              7076\n",
      "num_cancer_instances                 0\n",
      "Name: 156, dtype: object\n",
      "NFATC4::STAT1 MM\n",
      "(((())))\n",
      "NFATC4::STAT1 Melanocytes 26\n",
      "(((())))\n",
      "NFATC4::STAT1 Melanocytes 26\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (8537, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (8537, 14)\n",
      "(((())))\n",
      "NFATC4::STAT1 MM 615\n",
      "(((())))\n",
      "NFATC4::STAT1 MM 615\n",
      "pro-pro instances (102, 14)\n",
      "pro-loop instances (502155, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (502257, 14)\n",
      "vocab                    NFATC4::STAT1\n",
      "tissues_to_check        Melanocytes|MM\n",
      "num_normal_instances              8537\n",
      "num_cancer_instances            502257\n",
      "Name: 165, dtype: object\n",
      "NFE2L1::NR3C1 MM\n",
      "(((())))\n",
      "NFE2L1::NR3C1 Melanocytes 7\n",
      "(((())))\n",
      "NFE2L1::NR3C1 Melanocytes 7\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (1078, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (1078, 14)\n",
      "(((())))\n",
      "NFE2L1::NR3C1 MM 522\n",
      "(((())))\n",
      "NFE2L1::NR3C1 MM 522\n",
      "pro-pro instances (37, 14)\n",
      "pro-loop instances (223100, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (223137, 14)\n",
      "vocab                    NFE2L1::NR3C1\n",
      "tissues_to_check        Melanocytes|MM\n",
      "num_normal_instances              1078\n",
      "num_cancer_instances            223137\n",
      "Name: 167, dtype: object\n",
      "NFE2L1::TBX15 MM\n",
      "(((())))\n",
      "NFE2L1::TBX15 MM 580\n",
      "(((())))\n",
      "NFE2L1::TBX15 MM 580\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (553463, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (553463, 14)\n",
      "vocab                    NFE2L1::TBX15\n",
      "tissues_to_check        Melanocytes|MM\n",
      "num_normal_instances                 0\n",
      "num_cancer_instances            553463\n",
      "Name: 172, dtype: object\n",
      "NFE2L1::ZBTB49 MM\n",
      "(((())))\n",
      "NFE2L1::ZBTB49 Melanocytes 15\n",
      "(((())))\n",
      "NFE2L1::ZBTB49 Melanocytes 15\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (2932, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (2932, 14)\n",
      "(((())))\n",
      "NFE2L1::ZBTB49 MM 599\n",
      "(((())))\n",
      "NFE2L1::ZBTB49 MM 599\n",
      "pro-pro instances (57, 14)\n",
      "pro-loop instances (304848, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (304905, 14)\n",
      "vocab                   NFE2L1::ZBTB49\n",
      "tissues_to_check        Melanocytes|MM\n",
      "num_normal_instances              2932\n",
      "num_cancer_instances            304905\n",
      "Name: 174, dtype: object\n",
      "PRDM1::RORA GDSD6\n",
      "(((())))\n",
      "PRDM1::RORA GDSD6 91\n",
      "(((())))\n",
      "PRDM1::RORA GDSD6 91\n",
      "pro-pro instances (2, 14)\n",
      "pro-loop instances (19995, 14)\n",
      "loop-loop instances (83, 14)\n",
      "overall instances (20080, 14)\n",
      "(((())))\n",
      "PRDM1::RORA SCC 405\n",
      "(((())))\n",
      "PRDM1::RORA SCC 405\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (104785, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (104785, 14)\n",
      "vocab                   PRDM1::RORA\n",
      "tissues_to_check          GDSD6|SCC\n",
      "num_normal_instances          20080\n",
      "num_cancer_instances         104785\n",
      "Name: 185, dtype: object\n",
      "SOX10::SOX13 MM\n",
      "(((())))\n",
      "SOX10::SOX13 Melanocytes 18\n",
      "(((())))\n",
      "SOX10::SOX13 Melanocytes 18\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (8425, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (8425, 14)\n",
      "(((())))\n",
      "SOX10::SOX13 MM 627\n",
      "(((())))\n",
      "SOX10::SOX13 MM 627\n",
      "pro-pro instances (233, 14)\n",
      "pro-loop instances (995914, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (996147, 14)\n",
      "vocab                     SOX10::SOX13\n",
      "tissues_to_check        Melanocytes|MM\n",
      "num_normal_instances              8425\n",
      "num_cancer_instances            996147\n",
      "Name: 203, dtype: object\n",
      "SOX10::SP4 MM\n",
      "(((())))\n",
      "SOX10::SP4 MM 542\n",
      "(((())))\n",
      "SOX10::SP4 MM 542\n",
      "pro-pro instances (69, 14)\n",
      "pro-loop instances (679415, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (679484, 14)\n",
      "vocab                       SOX10::SP4\n",
      "tissues_to_check        Melanocytes|MM\n",
      "num_normal_instances                 0\n",
      "num_cancer_instances            679484\n",
      "Name: 204, dtype: object\n",
      "SP1::TFDP1 SCC\n",
      "(((())))\n",
      "SP1::TFDP1 GDSD6 179\n",
      "(((())))\n",
      "SP1::TFDP1 GDSD6 179\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (219031, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (219031, 14)\n",
      "(((())))\n",
      "SP1::TFDP1 SCC 1189\n",
      "(((())))\n",
      "SP1::TFDP1 SCC 1189\n",
      "pro-pro instances (436, 14)\n",
      "pro-loop instances (1381047, 14)\n",
      "loop-loop instances (73, 14)\n",
      "overall instances (1381556, 14)\n",
      "vocab                    SP1::TFDP1\n",
      "tissues_to_check          GDSD6|SCC\n",
      "num_normal_instances         219031\n",
      "num_cancer_instances    1.38156e+06\n",
      "Name: 214, dtype: object\n",
      "SP1::ZBTB14 SCC\n",
      "(((())))\n",
      "SP1::ZBTB14 GDSD6 109\n",
      "(((())))\n",
      "SP1::ZBTB14 GDSD6 109\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (130290, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (130290, 14)\n",
      "(((())))\n",
      "SP1::ZBTB14 SCC 1018\n",
      "(((())))\n",
      "SP1::ZBTB14 SCC 1018\n",
      "pro-pro instances (376, 14)\n",
      "pro-loop instances (1155858, 14)\n",
      "loop-loop instances (59, 14)\n",
      "overall instances (1156293, 14)\n",
      "vocab                   SP1::ZBTB14\n",
      "tissues_to_check          GDSD6|SCC\n",
      "num_normal_instances         130290\n",
      "num_cancer_instances    1.15629e+06\n",
      "Name: 217, dtype: object\n",
      "SP1::ZBTB17 SCC\n",
      "(((())))\n",
      "SP1::ZBTB17 GDSD6 423\n",
      "(((())))\n",
      "SP1::ZBTB17 GDSD6 423\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (712295, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (712295, 14)\n",
      "(((())))\n",
      "SP1::ZBTB17 SCC 1539\n",
      "(((())))\n",
      "SP1::ZBTB17 SCC 1539\n",
      "pro-pro instances (1417, 14)\n",
      "pro-loop instances (2547350, 14)\n",
      "loop-loop instances (87, 14)\n",
      "overall instances (2548854, 14)\n",
      "vocab                   SP1::ZBTB17\n",
      "tissues_to_check          GDSD6|SCC\n",
      "num_normal_instances         712295\n",
      "num_cancer_instances    2.54885e+06\n",
      "Name: 218, dtype: object\n",
      "SP1::ZNF770 SCC\n",
      "(((())))\n",
      "SP1::ZNF770 GDSD6 384\n",
      "(((())))\n",
      "SP1::ZNF770 GDSD6 384\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (612135, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (612135, 14)\n",
      "(((())))\n",
      "SP1::ZNF770 SCC 1523\n",
      "(((())))\n",
      "SP1::ZNF770 SCC 1523\n",
      "pro-pro instances (1215, 14)\n",
      "pro-loop instances (2293713, 14)\n",
      "loop-loop instances (99, 14)\n",
      "overall instances (2295027, 14)\n",
      "vocab                   SP1::ZNF770\n",
      "tissues_to_check          GDSD6|SCC\n",
      "num_normal_instances         612135\n",
      "num_cancer_instances    2.29503e+06\n",
      "Name: 220, dtype: object\n"
     ]
    }
   ],
   "source": [
    "# takes awhile\n",
    "genomic_instance_comp_df = pd.DataFrame()\n",
    "genomic_instance_comp_df['vocab'] = ''\n",
    "genomic_instance_comp_df['tissues_to_check'] = ''\n",
    "genomic_instance_comp_df['num_normal_instances'] = 0\n",
    "genomic_instance_comp_df['num_cancer_instances'] = 0\n",
    "for idx, row in new_mpra_df.iterrows():\n",
    "    # get row\n",
    "    print(row['vocab'],row.tissue)\n",
    "    tf1,tf2 = row['vocab'].split('::')\n",
    "    genomic_instance_comp_df.at[idx,'vocab'] = row['vocab']\n",
    "    \n",
    "    # get tissues to check\n",
    "    if row.tissue in ['GDSD6',  'SCC']:\n",
    "        tissues_to_check = ['GDSD6',  'SCC']\n",
    "    else:\n",
    "        tissues_to_check = ['Melanocytes','MM']\n",
    "    genomic_instance_comp_df.at[idx, 'tissues_to_check'] = '|'.join(tissues_to_check )\n",
    "  \n",
    "    # get genomic instances\n",
    "    num_instances = [0,0] # for normal, and cancer\n",
    "    \n",
    "    for i_mpra,mpra_tissue in enumerate(tissues_to_check):\n",
    "        geneset_file = os.path.join(vocab_geneset_dir,row['tissue']+'_'+tf1 + '-' + tf2+'_'+mpra_tissue+'.txt' )\n",
    "        if os.path.exists(geneset_file):\n",
    "            geneset = set(pd.read_csv(geneset_file,header=None).iloc[:,0].values)\n",
    "            if len(geneset)>0:\n",
    "                instances = get_genomic_instances_all(row.vocab, mpra_tissue, sorted(geneset),\n",
    "                                          verbose=False, rewrite=False,\n",
    "                                          save_prefix='../data/processed/fig7_vocab_go/genomic_instance_normal_cancer_comp/'+str(idx)+'_')\n",
    "                num_instances[i_mpra] = len(instances)\n",
    "    genomic_instance_comp_df.at[idx,'num_normal_instances'] = num_instances[0]\n",
    "    genomic_instance_comp_df.at[idx,'num_cancer_instances'] = num_instances[1]\n",
    "    print(genomic_instance_comp_df.loc[idx,:])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>vocab</th>\n",
       "      <th>tissues_to_check</th>\n",
       "      <th>num_normal_instances</th>\n",
       "      <th>num_cancer_instances</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>ARID5B::LEF1</td>\n",
       "      <td>Melanocytes|MM</td>\n",
       "      <td>4024.0</td>\n",
       "      <td>574639.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>GDSD6|SCC</td>\n",
       "      <td>104879.0</td>\n",
       "      <td>897199.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>BACH1::KLF5</td>\n",
       "      <td>GDSD6|SCC</td>\n",
       "      <td>58058.0</td>\n",
       "      <td>572001.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18</td>\n",
       "      <td>E2F6::SP1</td>\n",
       "      <td>GDSD6|SCC</td>\n",
       "      <td>360936.0</td>\n",
       "      <td>1776878.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>25</td>\n",
       "      <td>EGR1::SP3</td>\n",
       "      <td>GDSD6|SCC</td>\n",
       "      <td>319286.0</td>\n",
       "      <td>1334645.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>26</td>\n",
       "      <td>EGR2::FOSL2</td>\n",
       "      <td>GDSD6|SCC</td>\n",
       "      <td>260414.0</td>\n",
       "      <td>1184169.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>31</td>\n",
       "      <td>EGR2::NFE2L1</td>\n",
       "      <td>Melanocytes|MM</td>\n",
       "      <td>0.0</td>\n",
       "      <td>565239.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>32</td>\n",
       "      <td>EGR2::RARG</td>\n",
       "      <td>GDSD6|SCC</td>\n",
       "      <td>185825.0</td>\n",
       "      <td>1081153.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>33</td>\n",
       "      <td>EGR2::RORA</td>\n",
       "      <td>GDSD6|SCC</td>\n",
       "      <td>190733.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>43</td>\n",
       "      <td>EN2::STAT1</td>\n",
       "      <td>Melanocytes|MM</td>\n",
       "      <td>455.0</td>\n",
       "      <td>204327.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>47</td>\n",
       "      <td>ETV2::PRRX1</td>\n",
       "      <td>Melanocytes|MM</td>\n",
       "      <td>59667.0</td>\n",
       "      <td>2419451.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>54</td>\n",
       "      <td>FOXD3::PRRX1</td>\n",
       "      <td>Melanocytes|MM</td>\n",
       "      <td>26272.0</td>\n",
       "      <td>1580530.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>66</td>\n",
       "      <td>HBP1::STAT1</td>\n",
       "      <td>Melanocytes|MM</td>\n",
       "      <td>2085.0</td>\n",
       "      <td>192619.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>67</td>\n",
       "      <td>HMGA1::KLF5</td>\n",
       "      <td>GDSD6|SCC</td>\n",
       "      <td>226721.0</td>\n",
       "      <td>1255611.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>103</td>\n",
       "      <td>JUNB::RORA</td>\n",
       "      <td>GDSD6|SCC</td>\n",
       "      <td>34386.0</td>\n",
       "      <td>150544.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>117</td>\n",
       "      <td>KLF15::ZNF589</td>\n",
       "      <td>Melanocytes|MM</td>\n",
       "      <td>95646.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>120</td>\n",
       "      <td>KLF16::TFAP2C</td>\n",
       "      <td>GDSD6|SCC</td>\n",
       "      <td>152861.0</td>\n",
       "      <td>1042609.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>121</td>\n",
       "      <td>KLF4::KLF5</td>\n",
       "      <td>GDSD6|SCC</td>\n",
       "      <td>413687.0</td>\n",
       "      <td>1378127.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>128</td>\n",
       "      <td>KLF5::MAFG</td>\n",
       "      <td>GDSD6|SCC</td>\n",
       "      <td>94830.0</td>\n",
       "      <td>697623.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>129</td>\n",
       "      <td>KLF5::MAFK</td>\n",
       "      <td>GDSD6|SCC</td>\n",
       "      <td>87643.0</td>\n",
       "      <td>668822.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>131</td>\n",
       "      <td>KLF5::NFYB</td>\n",
       "      <td>Melanocytes|MM</td>\n",
       "      <td>11886.0</td>\n",
       "      <td>90189.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>149</td>\n",
       "      <td>MAF::RORA</td>\n",
       "      <td>GDSD6|SCC</td>\n",
       "      <td>31265.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>153</td>\n",
       "      <td>MAZ::SP1</td>\n",
       "      <td>GDSD6|SCC</td>\n",
       "      <td>1231874.0</td>\n",
       "      <td>3964165.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>156</td>\n",
       "      <td>MITF::ZNF589</td>\n",
       "      <td>Melanocytes|MM</td>\n",
       "      <td>7076.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>165</td>\n",
       "      <td>NFATC4::STAT1</td>\n",
       "      <td>Melanocytes|MM</td>\n",
       "      <td>8537.0</td>\n",
       "      <td>502257.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>167</td>\n",
       "      <td>NFE2L1::NR3C1</td>\n",
       "      <td>Melanocytes|MM</td>\n",
       "      <td>1078.0</td>\n",
       "      <td>223137.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>172</td>\n",
       "      <td>NFE2L1::TBX15</td>\n",
       "      <td>Melanocytes|MM</td>\n",
       "      <td>0.0</td>\n",
       "      <td>553463.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>174</td>\n",
       "      <td>NFE2L1::ZBTB49</td>\n",
       "      <td>Melanocytes|MM</td>\n",
       "      <td>2932.0</td>\n",
       "      <td>304905.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>185</td>\n",
       "      <td>PRDM1::RORA</td>\n",
       "      <td>GDSD6|SCC</td>\n",
       "      <td>20080.0</td>\n",
       "      <td>104785.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>203</td>\n",
       "      <td>SOX10::SOX13</td>\n",
       "      <td>Melanocytes|MM</td>\n",
       "      <td>8425.0</td>\n",
       "      <td>996147.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>204</td>\n",
       "      <td>SOX10::SP4</td>\n",
       "      <td>Melanocytes|MM</td>\n",
       "      <td>0.0</td>\n",
       "      <td>679484.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>214</td>\n",
       "      <td>SP1::TFDP1</td>\n",
       "      <td>GDSD6|SCC</td>\n",
       "      <td>219031.0</td>\n",
       "      <td>1381556.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>217</td>\n",
       "      <td>SP1::ZBTB14</td>\n",
       "      <td>GDSD6|SCC</td>\n",
       "      <td>130290.0</td>\n",
       "      <td>1156293.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>218</td>\n",
       "      <td>SP1::ZBTB17</td>\n",
       "      <td>GDSD6|SCC</td>\n",
       "      <td>712295.0</td>\n",
       "      <td>2548854.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>220</td>\n",
       "      <td>SP1::ZNF770</td>\n",
       "      <td>GDSD6|SCC</td>\n",
       "      <td>612135.0</td>\n",
       "      <td>2295027.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              vocab tissues_to_check  num_normal_instances  \\\n",
       "1      ARID5B::LEF1   Melanocytes|MM                4024.0   \n",
       "5         ARNT::SP1        GDSD6|SCC              104879.0   \n",
       "8       BACH1::KLF5        GDSD6|SCC               58058.0   \n",
       "18        E2F6::SP1        GDSD6|SCC              360936.0   \n",
       "25        EGR1::SP3        GDSD6|SCC              319286.0   \n",
       "26      EGR2::FOSL2        GDSD6|SCC              260414.0   \n",
       "31     EGR2::NFE2L1   Melanocytes|MM                   0.0   \n",
       "32       EGR2::RARG        GDSD6|SCC              185825.0   \n",
       "33       EGR2::RORA        GDSD6|SCC              190733.0   \n",
       "43       EN2::STAT1   Melanocytes|MM                 455.0   \n",
       "47      ETV2::PRRX1   Melanocytes|MM               59667.0   \n",
       "54     FOXD3::PRRX1   Melanocytes|MM               26272.0   \n",
       "66      HBP1::STAT1   Melanocytes|MM                2085.0   \n",
       "67      HMGA1::KLF5        GDSD6|SCC              226721.0   \n",
       "103      JUNB::RORA        GDSD6|SCC               34386.0   \n",
       "117   KLF15::ZNF589   Melanocytes|MM               95646.0   \n",
       "120   KLF16::TFAP2C        GDSD6|SCC              152861.0   \n",
       "121      KLF4::KLF5        GDSD6|SCC              413687.0   \n",
       "128      KLF5::MAFG        GDSD6|SCC               94830.0   \n",
       "129      KLF5::MAFK        GDSD6|SCC               87643.0   \n",
       "131      KLF5::NFYB   Melanocytes|MM               11886.0   \n",
       "149       MAF::RORA        GDSD6|SCC               31265.0   \n",
       "153        MAZ::SP1        GDSD6|SCC             1231874.0   \n",
       "156    MITF::ZNF589   Melanocytes|MM                7076.0   \n",
       "165   NFATC4::STAT1   Melanocytes|MM                8537.0   \n",
       "167   NFE2L1::NR3C1   Melanocytes|MM                1078.0   \n",
       "172   NFE2L1::TBX15   Melanocytes|MM                   0.0   \n",
       "174  NFE2L1::ZBTB49   Melanocytes|MM                2932.0   \n",
       "185     PRDM1::RORA        GDSD6|SCC               20080.0   \n",
       "203    SOX10::SOX13   Melanocytes|MM                8425.0   \n",
       "204      SOX10::SP4   Melanocytes|MM                   0.0   \n",
       "214      SP1::TFDP1        GDSD6|SCC              219031.0   \n",
       "217     SP1::ZBTB14        GDSD6|SCC              130290.0   \n",
       "218     SP1::ZBTB17        GDSD6|SCC              712295.0   \n",
       "220     SP1::ZNF770        GDSD6|SCC              612135.0   \n",
       "\n",
       "     num_cancer_instances  \n",
       "1                574639.0  \n",
       "5                897199.0  \n",
       "8                572001.0  \n",
       "18              1776878.0  \n",
       "25              1334645.0  \n",
       "26              1184169.0  \n",
       "31               565239.0  \n",
       "32              1081153.0  \n",
       "33                    0.0  \n",
       "43               204327.0  \n",
       "47              2419451.0  \n",
       "54              1580530.0  \n",
       "66               192619.0  \n",
       "67              1255611.0  \n",
       "103              150544.0  \n",
       "117                   0.0  \n",
       "120             1042609.0  \n",
       "121             1378127.0  \n",
       "128              697623.0  \n",
       "129              668822.0  \n",
       "131               90189.0  \n",
       "149                   0.0  \n",
       "153             3964165.0  \n",
       "156                   0.0  \n",
       "165              502257.0  \n",
       "167              223137.0  \n",
       "172              553463.0  \n",
       "174              304905.0  \n",
       "185              104785.0  \n",
       "203              996147.0  \n",
       "204              679484.0  \n",
       "214             1381556.0  \n",
       "217             1156293.0  \n",
       "218             2548854.0  \n",
       "220             2295027.0  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genomic_instance_comp_df.to_csv(os.path.join(save_dir, 'genomic_instance_comp_df.csv'))\n",
    "genomic_instance_comp_df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# misc stuff\n",
    "Can you see if chr22:38572526 is in our TF footprint data for MC or any of the melanoma cell lines?\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['../data/interim/merged/footprinting/COLO_SCR_PLX_merged.bed',\n",
       " '../data/interim/merged/footprinting/COLO_shMITF_PLX_merged.bed',\n",
       " '../data/interim/merged/footprinting/COLO_shMITF_DMSO_merged.bed',\n",
       " '../data/interim/merged/footprinting/WM_SCR_DMSO_merged.bed',\n",
       " '../data/interim/merged/footprinting/Melanocytes_merged.bed',\n",
       " '../data/interim/merged/footprinting/SKMEL5_SCR_DMSO_merged.bed',\n",
       " '../data/interim/merged/footprinting/COLO_SCR_DMSO_merged.bed']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mc_foot_bed = []\n",
    "for file in glob.glob('../data/interim/merged/footprinting/*'):\n",
    "    if os.path.basename(file).startswith('COLO'):\n",
    "        mc_foot_bed.append(file)\n",
    "    elif os.path.basename(file).startswith('WM_'):\n",
    "        mc_foot_bed.append(file)\n",
    "    elif os.path.basename(file).startswith('SKMEL'):\n",
    "        mc_foot_bed.append(file)\n",
    "    elif os.path.basename(file).startswith('Melanoc'):\n",
    "        mc_foot_bed.append(file)\n",
    "mc_foot_bed        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "COLO_SCR_PLX_merged.bed\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>thickStart</th>\n",
       "      <th>tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [chrom, start, end, name, score, strand, thickStart, tissue]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "COLO_shMITF_PLX_merged.bed\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>thickStart</th>\n",
       "      <th>tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [chrom, start, end, name, score, strand, thickStart, tissue]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "COLO_shMITF_DMSO_merged.bed\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-10-a99c5cfaa30d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mfoot_bed\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mmc_foot_bed\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbasename\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfoot_bed\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m     \u001b[0mfoot_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpybedtools\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mBedTool\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfoot_bed\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_dataframe\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m     \u001b[0mfoot_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfoot_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfoot_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mchrom\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;34m'chr22'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m     \u001b[0mfoot_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfoot_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfoot_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstart\u001b[0m\u001b[0;34m<=\u001b[0m\u001b[0;36m38572526\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m&\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfoot_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mend\u001b[0m\u001b[0;34m>=\u001b[0m\u001b[0;36m38572526\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/pybedtools/bedtool.py\u001b[0m in \u001b[0;36mto_dataframe\u001b[0;34m(self, disable_auto_names, *args, **kwargs)\u001b[0m\n\u001b[1;32m   3686\u001b[0m             \u001b[0mkwargs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"names\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_names\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3687\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3688\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mpandas\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfn\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"\\t\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3689\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3690\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mtail\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlines\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mas_string\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36mparser_f\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, dialect, error_bad_lines, warn_bad_lines, delim_whitespace, low_memory, memory_map, float_precision)\u001b[0m\n\u001b[1;32m    683\u001b[0m         )\n\u001b[1;32m    684\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 685\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0m_read\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    686\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    687\u001b[0m     \u001b[0mparser_f\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    461\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    462\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 463\u001b[0;31m         \u001b[0mdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparser\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnrows\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    464\u001b[0m     \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    465\u001b[0m         \u001b[0mparser\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36mread\u001b[0;34m(self, nrows)\u001b[0m\n\u001b[1;32m   1167\u001b[0m             \u001b[0mnew_rows\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1168\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1169\u001b[0;31m         \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mDataFrame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcol_dict\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolumns\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1170\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1171\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_currow\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0mnew_rows\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, data, index, columns, dtype, copy)\u001b[0m\n\u001b[1;32m    409\u001b[0m             )\n\u001b[1;32m    410\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdict\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 411\u001b[0;31m             \u001b[0mmgr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0minit_dict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolumns\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    412\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mma\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mMaskedArray\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    413\u001b[0m             \u001b[0;32mimport\u001b[0m \u001b[0mnumpy\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mma\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmrecords\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mmrecords\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/pandas/core/internals/construction.py\u001b[0m in \u001b[0;36minit_dict\u001b[0;34m(data, index, columns, dtype)\u001b[0m\n\u001b[1;32m    255\u001b[0m             \u001b[0marr\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mis_datetime64tz_dtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marr\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0marr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0marr\u001b[0m \u001b[0;32min\u001b[0m \u001b[0marrays\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    256\u001b[0m         ]\n\u001b[0;32m--> 257\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0marrays_to_mgr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marrays\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata_names\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolumns\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    258\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    259\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/pandas/core/internals/construction.py\u001b[0m in \u001b[0;36marrays_to_mgr\u001b[0;34m(arrays, arr_names, index, columns, dtype)\u001b[0m\n\u001b[1;32m     85\u001b[0m     \u001b[0maxes\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mensure_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     86\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 87\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mcreate_block_manager_from_arrays\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marrays\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0marr_names\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxes\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     88\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     89\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36mcreate_block_manager_from_arrays\u001b[0;34m(arrays, names, axes)\u001b[0m\n\u001b[1;32m   1692\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1693\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1694\u001b[0;31m         \u001b[0mblocks\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mform_blocks\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marrays\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnames\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxes\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1695\u001b[0m         \u001b[0mmgr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mBlockManager\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mblocks\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxes\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1696\u001b[0m         \u001b[0mmgr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_consolidate_inplace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36mform_blocks\u001b[0;34m(arrays, names, axes)\u001b[0m\n\u001b[1;32m   1781\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1782\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mitems_dict\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"ObjectBlock\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1783\u001b[0;31m         \u001b[0mobject_blocks\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_simple_blockify\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mitems_dict\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"ObjectBlock\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobject_\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1784\u001b[0m         \u001b[0mblocks\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mextend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobject_blocks\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1785\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36m_simple_blockify\u001b[0;34m(tuples, dtype)\u001b[0m\n\u001b[1;32m   1825\u001b[0m     \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcoerce\u001b[0m \u001b[0mto\u001b[0m \u001b[0mthis\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1826\u001b[0m     \"\"\"\n\u001b[0;32m-> 1827\u001b[0;31m     \u001b[0mvalues\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mplacement\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_stack_arrays\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtuples\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1828\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1829\u001b[0m     \u001b[0;31m# CHECK DTYPE?\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36m_stack_arrays\u001b[0;34m(tuples, dtype)\u001b[0m\n\u001b[1;32m   1874\u001b[0m     \u001b[0mstacked\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mempty\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1875\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0marr\u001b[0m \u001b[0;32min\u001b[0m \u001b[0menumerate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marrays\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1876\u001b[0;31m         \u001b[0mstacked\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_asarray_compat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1877\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1878\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mstacked\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mplacement\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "results_df = pd.DataFrame()\n",
    "for foot_bed in mc_foot_bed:\n",
    "    print(os.path.basename(foot_bed))\n",
    "    foot_df = pybedtools.BedTool(foot_bed).to_dataframe()\n",
    "    foot_df = foot_df[(foot_df.chrom=='chr22')]\n",
    "    foot_df = foot_df[(foot_df.start<=38572526)&(foot_df.end>=38572526)]\n",
    "    foot_df['tissue'] = os.path.basename(foot_bed).split('_merged')[0]\n",
    "    results_df = pd.concat([results_df, foot_df])\n",
    "    display(foot_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# update 06/24/2021\n",
    "Also, since we redid the Fig 3 TF expression/GOTerms figure POU2F2 is no longer a good vignette since it doesn't come up as an enriched motif in KCs :/, can you get me the GDSD6 and Prostate anchor and loop instances of ATF4, SMAD4, and CEBPG? Esp if any of those have FLG as a target gene (GDSD6) or UGDH as a target gene (prostate)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(((())))\n",
      "ATF4::ATF4 GDSD6 1\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (0, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (0, 14)\n",
      "(((())))\n",
      "ATF4::SMAD4 GDSD6 1\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (1, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (1, 14)\n",
      "(((())))\n",
      "ATF4::CEBPG GDSD6 1\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (0, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (0, 14)\n",
      "(((())))\n",
      "SMAD4::ATF4 GDSD6 1\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (1, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (1, 14)\n",
      "(((())))\n",
      "SMAD4::SMAD4 GDSD6 1\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (1, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (1, 14)\n",
      "(((())))\n",
      "SMAD4::CEBPG GDSD6 1\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (1, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (1, 14)\n",
      "(((())))\n",
      "CEBPG::ATF4 GDSD6 1\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (0, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (0, 14)\n",
      "(((())))\n",
      "CEBPG::SMAD4 GDSD6 1\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (1, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (1, 14)\n",
      "(((())))\n",
      "CEBPG::CEBPG GDSD6 1\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (0, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (0, 14)\n"
     ]
    }
   ],
   "source": [
    "gds_flg_df = pd.DataFrame()\n",
    "for tf1 in [ 'ATF4', 'SMAD4',  'CEBPG']:\n",
    "    for tf2 in [ 'ATF4', 'SMAD4',  'CEBPG']:\n",
    "        vocab = tf1+'::'+tf2\n",
    "        df =get_genomic_instances_all(vocab, 'GDSD6', ['FLG'])\n",
    "        gds_flg_df = pd.concat([gds_flg_df, df])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>TSS</th>\n",
       "      <th>chr_m</th>\n",
       "      <th>start_m</th>\n",
       "      <th>stop_m</th>\n",
       "      <th>id_trim</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>tf</th>\n",
       "      <th>vocab</th>\n",
       "      <th>tissue</th>\n",
       "      <th>config</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>152415000</td>\n",
       "      <td>152420000</td>\n",
       "      <td>chr1_152415000_152420000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>152415595</td>\n",
       "      <td>152415606</td>\n",
       "      <td>SMAD4_HUMAN.H11MO.0.B.pwm.trim</td>\n",
       "      <td>10.83232</td>\n",
       "      <td>+</td>\n",
       "      <td>SMAD4</td>\n",
       "      <td>ATF4::SMAD4</td>\n",
       "      <td>GDSD6</td>\n",
       "      <td>pro_loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>152415000</td>\n",
       "      <td>152420000</td>\n",
       "      <td>chr1_152415000_152420000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>152415595</td>\n",
       "      <td>152415606</td>\n",
       "      <td>SMAD4_HUMAN.H11MO.0.B.pwm.trim</td>\n",
       "      <td>10.83232</td>\n",
       "      <td>+</td>\n",
       "      <td>SMAD4</td>\n",
       "      <td>SMAD4::ATF4</td>\n",
       "      <td>GDSD6</td>\n",
       "      <td>pro_loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>152415000</td>\n",
       "      <td>152420000</td>\n",
       "      <td>chr1_152415000_152420000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>152415595</td>\n",
       "      <td>152415606</td>\n",
       "      <td>SMAD4_HUMAN.H11MO.0.B.pwm.trim</td>\n",
       "      <td>10.83232</td>\n",
       "      <td>+</td>\n",
       "      <td>SMAD4</td>\n",
       "      <td>SMAD4::SMAD4</td>\n",
       "      <td>GDSD6</td>\n",
       "      <td>pro_loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>152415000</td>\n",
       "      <td>152420000</td>\n",
       "      <td>chr1_152415000_152420000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>152415595</td>\n",
       "      <td>152415606</td>\n",
       "      <td>SMAD4_HUMAN.H11MO.0.B.pwm.trim</td>\n",
       "      <td>10.83232</td>\n",
       "      <td>+</td>\n",
       "      <td>SMAD4</td>\n",
       "      <td>SMAD4::CEBPG</td>\n",
       "      <td>GDSD6</td>\n",
       "      <td>pro_loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>152415000</td>\n",
       "      <td>152420000</td>\n",
       "      <td>chr1_152415000_152420000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>152415595</td>\n",
       "      <td>152415606</td>\n",
       "      <td>SMAD4_HUMAN.H11MO.0.B.pwm.trim</td>\n",
       "      <td>10.83232</td>\n",
       "      <td>+</td>\n",
       "      <td>SMAD4</td>\n",
       "      <td>CEBPG::SMAD4</td>\n",
       "      <td>GDSD6</td>\n",
       "      <td>pro_loop</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr      start       stop                       TSS chr_m    start_m  \\\n",
       "0  chr1  152415000  152420000  chr1_152415000_152420000  chr1  152415595   \n",
       "0  chr1  152415000  152420000  chr1_152415000_152420000  chr1  152415595   \n",
       "0  chr1  152415000  152420000  chr1_152415000_152420000  chr1  152415595   \n",
       "0  chr1  152415000  152420000  chr1_152415000_152420000  chr1  152415595   \n",
       "0  chr1  152415000  152420000  chr1_152415000_152420000  chr1  152415595   \n",
       "\n",
       "      stop_m                         id_trim     score strand     tf  \\\n",
       "0  152415606  SMAD4_HUMAN.H11MO.0.B.pwm.trim  10.83232      +  SMAD4   \n",
       "0  152415606  SMAD4_HUMAN.H11MO.0.B.pwm.trim  10.83232      +  SMAD4   \n",
       "0  152415606  SMAD4_HUMAN.H11MO.0.B.pwm.trim  10.83232      +  SMAD4   \n",
       "0  152415606  SMAD4_HUMAN.H11MO.0.B.pwm.trim  10.83232      +  SMAD4   \n",
       "0  152415606  SMAD4_HUMAN.H11MO.0.B.pwm.trim  10.83232      +  SMAD4   \n",
       "\n",
       "          vocab tissue    config  \n",
       "0   ATF4::SMAD4  GDSD6  pro_loop  \n",
       "0   SMAD4::ATF4  GDSD6  pro_loop  \n",
       "0  SMAD4::SMAD4  GDSD6  pro_loop  \n",
       "0  SMAD4::CEBPG  GDSD6  pro_loop  \n",
       "0  CEBPG::SMAD4  GDSD6  pro_loop  "
      ]
     },
     "execution_count": 204,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gds_flg_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(((())))\n",
      "ATF4::ATF4 Prostate 1\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (0, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (0, 14)\n",
      "(((())))\n",
      "ATF4::SMAD4 Prostate 1\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (0, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (0, 14)\n",
      "(((())))\n",
      "ATF4::CEBPG Prostate 1\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (0, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (0, 14)\n",
      "(((())))\n",
      "SMAD4::ATF4 Prostate 1\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (0, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (0, 14)\n",
      "(((())))\n",
      "SMAD4::SMAD4 Prostate 1\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (0, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (0, 14)\n",
      "(((())))\n",
      "SMAD4::CEBPG Prostate 1\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (0, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (0, 14)\n",
      "(((())))\n",
      "CEBPG::ATF4 Prostate 1\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (0, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (0, 14)\n",
      "(((())))\n",
      "CEBPG::SMAD4 Prostate 1\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (0, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (0, 14)\n",
      "(((())))\n",
      "CEBPG::CEBPG Prostate 1\n",
      "pro-pro instances (0, 14)\n",
      "pro-loop instances (0, 14)\n",
      "loop-loop instances (0, 14)\n",
      "overall instances (0, 14)\n"
     ]
    }
   ],
   "source": [
    "pros_ugdh_df = pd.DataFrame()\n",
    "for tf1 in [ 'ATF4', 'SMAD4',  'CEBPG']:\n",
    "    for tf2 in [ 'ATF4', 'SMAD4',  'CEBPG']:\n",
    "        vocab = tf1+'::'+tf2\n",
    "        df =get_genomic_instances_all(vocab, 'Prostate', ['UGDH'])\n",
    "        pros_ugdh_df = pd.concat([pros_ugdh_df, df])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>TSS</th>\n",
       "      <th>chr_m</th>\n",
       "      <th>start_m</th>\n",
       "      <th>stop_m</th>\n",
       "      <th>id_trim</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>tf</th>\n",
       "      <th>vocab</th>\n",
       "      <th>tissue</th>\n",
       "      <th>config</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [chr, start, stop, TSS, chr_m, start_m, stop_m, id_trim, score, strand, tf, vocab, tissue, config]\n",
       "Index: []"
      ]
     },
     "execution_count": 208,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pros_ugdh_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then can I also get the list of SMAD4 footprints in anchors/loops and the target genes identified in Prostate? I can try to find a good gene candidate to show"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25 target genes found for SMAD4 in tissue:  Prostate\n",
      "['AC013449.1', 'ARID3B', 'BMF', 'C10orf55', 'CLCNKA', 'EGFL8', 'FAM114A1', 'GAL3ST1', 'HIST2H3A', 'HSPB7', 'ICAM1', 'KIRREL2', 'LMTK3', 'MGAT4B', 'MMP14', 'NBPF24', 'RP11-293M10.1', 'RP11-474G23.1', 'RUNX3', 'S1PR4', 'SLC4A9', 'SQSTM1', 'TLX2', 'TSC22D4', 'TUSC2']\n"
     ]
    }
   ],
   "source": [
    "smad4_pros = get_target_genes('SMAD4', 'Prostate')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "metadata": {},
   "outputs": [],
   "source": [
    "smad4_pros.to_csv('../data/processed/fig7_vocab_go/genomic_instance_overlap/SMAD4_prostate_manual.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "465       -68\n",
       "8813    -4221\n",
       "8909    -2202\n",
       "11977   -2637\n",
       "14158   -4483\n",
       "         ... \n",
       "39492   -2190\n",
       "40354    -637\n",
       "40535   -1201\n",
       "41312   -1259\n",
       "43324    -316\n",
       "Length: 288, dtype: int64"
      ]
     },
     "execution_count": 234,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "smad4_pros.start_m - smad4_pros.stop#.TSS.unique().shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Can I get the gene lists for TF footprints of ATF4 and CEBPG in KCD6 and Prostate pro-loops? \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "38 target genes found for ATF4 in tissue:  GDSD6\n",
      "['AL160175.1', 'ARHGDIA', 'ATAT1', 'C4BPB', 'CCDC9', 'CCL17', 'CSTA', 'DYNC1LI2', 'FAM114A1', 'GDF15', 'GPR87', 'GSS', 'HAPLN2', 'ID1', 'IGFBP5', 'IL36G', 'ING2', 'KIAA0754', 'LSM10', 'MYH15', 'NFATC4', 'PARP2', 'PDP1', 'PIP5KL1', 'PRDM10', 'PYGB', 'RAET1E', 'REN', 'RHOH', 'RNASE7', 'RP11-762I7.5', 'S100A6', 'SESN2', 'SH3RF2', 'SLC25A14', 'SLC35A4', 'SLC9A1', 'STRA6']\n",
      "13 target genes found for ATF4 in tissue:  Prostate\n",
      "['ATAT1', 'CTU1', 'DYNC1LI2', 'GDF15', 'LRRC27', 'MRPS21', 'NDUFA4L2', 'OTOP1', 'PARP2', 'PKIG', 'RP11-529K1.3', 'STK32C', 'STRA6']\n"
     ]
    }
   ],
   "source": [
    "atf4_kcd6 = get_target_genes('ATF4', 'GDSD6')\n",
    "atf4_pros = get_target_genes('ATF4', 'Prostate')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44 target genes found for CEBPG in tissue:  GDSD6\n",
      "['AL160175.1', 'ANKRD11', 'ARHGDIA', 'ATAT1', 'C12orf45', 'C4BPB', 'CCDC9', 'CCL17', 'CSTA', 'CYGB', 'DYNC1LI2', 'FAM114A1', 'GDF15', 'GPR87', 'GSS', 'HAPLN2', 'ID1', 'IGFBP5', 'IL36G', 'ING2', 'KIAA0754', 'KRT33A', 'LSM10', 'NFATC4', 'PARP2', 'PDP1', 'PIP5KL1', 'PRDM10', 'PYGB', 'RAET1E', 'REN', 'RHOH', 'RNASE7', 'RP11-468E2.2', 'RP11-762I7.5', 'S100A6', 'SERPINB13', 'SESN2', 'SH3RF2', 'SLC25A14', 'SLC35A4', 'SLC9A1', 'STRA6', 'TOP1']\n",
      "14 target genes found for CEBPG in tissue:  Prostate\n",
      "['ATAT1', 'CTU1', 'CYGB', 'DYNC1LI2', 'GDF15', 'LRRC27', 'MRPS21', 'NDUFA4L2', 'NFIX', 'PARP2', 'PKIG', 'RP11-529K1.3', 'STK32C', 'STRA6']\n"
     ]
    }
   ],
   "source": [
    "cebpg_kcd6 = get_target_genes('CEBPG', 'GDSD6')\n",
    "cebpg_pros = get_target_genes('CEBPG', 'Prostate')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "atf4_kcd6.to_csv('../data/processed/fig7_vocab_go/genomic_instance_overlap/ATF4_kcd6_manual.csv')\n",
    "atf4_pros.to_csv('../data/processed/fig7_vocab_go/genomic_instance_overlap/ATF4_prostate_manual.csv')\n",
    "cebpg_kcd6.to_csv('../data/processed/fig7_vocab_go/genomic_instance_overlap/CEBPG_kcd6_manual.csv')\n",
    "cebpg_pros.to_csv('../data/processed/fig7_vocab_go/genomic_instance_overlap/CEBPG_prostate_manual.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>TSS</th>\n",
       "      <th>chr_m</th>\n",
       "      <th>start_m</th>\n",
       "      <th>stop_m</th>\n",
       "      <th>id_trim</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>tf</th>\n",
       "      <th>vocab</th>\n",
       "      <th>tissue</th>\n",
       "      <th>config</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>29</td>\n",
       "      <td>chr1</td>\n",
       "      <td>845000</td>\n",
       "      <td>850000</td>\n",
       "      <td>chr1_845000_850000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>846718</td>\n",
       "      <td>846728</td>\n",
       "      <td>CEBPG_HUMAN.H11MO.0.B.pwm.trim</td>\n",
       "      <td>12.215445</td>\n",
       "      <td>+</td>\n",
       "      <td>CEBPG</td>\n",
       "      <td>CEBPG</td>\n",
       "      <td>GDSD6</td>\n",
       "      <td>pro_loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3662</td>\n",
       "      <td>chr1</td>\n",
       "      <td>4760000</td>\n",
       "      <td>4765000</td>\n",
       "      <td>chr1_4760000_4765000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>4763236</td>\n",
       "      <td>4763246</td>\n",
       "      <td>CEBPG_HUMAN.H11MO.0.B.pwm.trim</td>\n",
       "      <td>10.263728</td>\n",
       "      <td>+</td>\n",
       "      <td>CEBPG</td>\n",
       "      <td>CEBPG</td>\n",
       "      <td>GDSD6</td>\n",
       "      <td>pro_loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6574</td>\n",
       "      <td>chr1</td>\n",
       "      <td>8270000</td>\n",
       "      <td>8275000</td>\n",
       "      <td>chr1_8270000_8275000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>8272237</td>\n",
       "      <td>8272247</td>\n",
       "      <td>CEBPG_HUMAN.H11MO.0.B.pwm.trim</td>\n",
       "      <td>15.931376</td>\n",
       "      <td>+</td>\n",
       "      <td>CEBPG</td>\n",
       "      <td>CEBPG</td>\n",
       "      <td>GDSD6</td>\n",
       "      <td>pro_loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18248</td>\n",
       "      <td>chr1</td>\n",
       "      <td>17515000</td>\n",
       "      <td>17520000</td>\n",
       "      <td>chr1_17515000_17520000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>17519080</td>\n",
       "      <td>17519090</td>\n",
       "      <td>CEBPG_HUMAN.H11MO.0.B.pwm.trim</td>\n",
       "      <td>13.381499</td>\n",
       "      <td>+</td>\n",
       "      <td>CEBPG</td>\n",
       "      <td>CEBPG</td>\n",
       "      <td>GDSD6</td>\n",
       "      <td>pro_loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>19855</td>\n",
       "      <td>chr1</td>\n",
       "      <td>19325000</td>\n",
       "      <td>19330000</td>\n",
       "      <td>chr1_19325000_19330000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>19325089</td>\n",
       "      <td>19325099</td>\n",
       "      <td>CEBPG_HUMAN.H11MO.0.B.pwm.trim</td>\n",
       "      <td>12.599845</td>\n",
       "      <td>+</td>\n",
       "      <td>CEBPG</td>\n",
       "      <td>CEBPG</td>\n",
       "      <td>GDSD6</td>\n",
       "      <td>pro_loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>69604</td>\n",
       "      <td>chr6</td>\n",
       "      <td>30582619</td>\n",
       "      <td>30585119</td>\n",
       "      <td>ATAT1</td>\n",
       "      <td>chr6</td>\n",
       "      <td>30582811</td>\n",
       "      <td>30582821</td>\n",
       "      <td>CEBPG_HUMAN.H11MO.0.B.pwm.trim</td>\n",
       "      <td>13.230063</td>\n",
       "      <td>+</td>\n",
       "      <td>CEBPG</td>\n",
       "      <td>CEBPG</td>\n",
       "      <td>GDSD6</td>\n",
       "      <td>pro_loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>72029</td>\n",
       "      <td>chr6</td>\n",
       "      <td>150207238</td>\n",
       "      <td>150209738</td>\n",
       "      <td>RAET1E</td>\n",
       "      <td>chr6</td>\n",
       "      <td>150208740</td>\n",
       "      <td>150208750</td>\n",
       "      <td>CEBPG_HUMAN.H11MO.0.B.pwm.trim</td>\n",
       "      <td>12.604143</td>\n",
       "      <td>+</td>\n",
       "      <td>CEBPG</td>\n",
       "      <td>CEBPG</td>\n",
       "      <td>GDSD6</td>\n",
       "      <td>pro_loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>76362</td>\n",
       "      <td>chr8</td>\n",
       "      <td>94858035</td>\n",
       "      <td>94860535</td>\n",
       "      <td>PDP1</td>\n",
       "      <td>chr8</td>\n",
       "      <td>94858942</td>\n",
       "      <td>94858952</td>\n",
       "      <td>CEBPG_HUMAN.H11MO.0.B.pwm.trim</td>\n",
       "      <td>11.196498</td>\n",
       "      <td>-</td>\n",
       "      <td>CEBPG</td>\n",
       "      <td>CEBPG</td>\n",
       "      <td>GDSD6</td>\n",
       "      <td>pro_loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>78485</td>\n",
       "      <td>chr9</td>\n",
       "      <td>130681076</td>\n",
       "      <td>130683576</td>\n",
       "      <td>PIP5KL1</td>\n",
       "      <td>chr9</td>\n",
       "      <td>130681864</td>\n",
       "      <td>130681874</td>\n",
       "      <td>CEBPG_HUMAN.H11MO.0.B.pwm.trim</td>\n",
       "      <td>10.880806</td>\n",
       "      <td>-</td>\n",
       "      <td>CEBPG</td>\n",
       "      <td>CEBPG</td>\n",
       "      <td>GDSD6</td>\n",
       "      <td>pro_loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>81099</td>\n",
       "      <td>chrX</td>\n",
       "      <td>129461874</td>\n",
       "      <td>129464374</td>\n",
       "      <td>SLC25A14</td>\n",
       "      <td>chrX</td>\n",
       "      <td>129464365</td>\n",
       "      <td>129464375</td>\n",
       "      <td>CEBPG_HUMAN.H11MO.0.B.pwm.trim</td>\n",
       "      <td>10.570578</td>\n",
       "      <td>-</td>\n",
       "      <td>CEBPG</td>\n",
       "      <td>CEBPG</td>\n",
       "      <td>GDSD6</td>\n",
       "      <td>pro_loop</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>714 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr      start       stop                     TSS chr_m    start_m  \\\n",
       "29     chr1     845000     850000      chr1_845000_850000  chr1     846718   \n",
       "3662   chr1    4760000    4765000    chr1_4760000_4765000  chr1    4763236   \n",
       "6574   chr1    8270000    8275000    chr1_8270000_8275000  chr1    8272237   \n",
       "18248  chr1   17515000   17520000  chr1_17515000_17520000  chr1   17519080   \n",
       "19855  chr1   19325000   19330000  chr1_19325000_19330000  chr1   19325089   \n",
       "...     ...        ...        ...                     ...   ...        ...   \n",
       "69604  chr6   30582619   30585119                   ATAT1  chr6   30582811   \n",
       "72029  chr6  150207238  150209738                  RAET1E  chr6  150208740   \n",
       "76362  chr8   94858035   94860535                    PDP1  chr8   94858942   \n",
       "78485  chr9  130681076  130683576                 PIP5KL1  chr9  130681864   \n",
       "81099  chrX  129461874  129464374                SLC25A14  chrX  129464365   \n",
       "\n",
       "          stop_m                         id_trim      score strand     tf  \\\n",
       "29        846728  CEBPG_HUMAN.H11MO.0.B.pwm.trim  12.215445      +  CEBPG   \n",
       "3662     4763246  CEBPG_HUMAN.H11MO.0.B.pwm.trim  10.263728      +  CEBPG   \n",
       "6574     8272247  CEBPG_HUMAN.H11MO.0.B.pwm.trim  15.931376      +  CEBPG   \n",
       "18248   17519090  CEBPG_HUMAN.H11MO.0.B.pwm.trim  13.381499      +  CEBPG   \n",
       "19855   19325099  CEBPG_HUMAN.H11MO.0.B.pwm.trim  12.599845      +  CEBPG   \n",
       "...          ...                             ...        ...    ...    ...   \n",
       "69604   30582821  CEBPG_HUMAN.H11MO.0.B.pwm.trim  13.230063      +  CEBPG   \n",
       "72029  150208750  CEBPG_HUMAN.H11MO.0.B.pwm.trim  12.604143      +  CEBPG   \n",
       "76362   94858952  CEBPG_HUMAN.H11MO.0.B.pwm.trim  11.196498      -  CEBPG   \n",
       "78485  130681874  CEBPG_HUMAN.H11MO.0.B.pwm.trim  10.880806      -  CEBPG   \n",
       "81099  129464375  CEBPG_HUMAN.H11MO.0.B.pwm.trim  10.570578      -  CEBPG   \n",
       "\n",
       "       vocab tissue    config  \n",
       "29     CEBPG  GDSD6  pro_loop  \n",
       "3662   CEBPG  GDSD6  pro_loop  \n",
       "6574   CEBPG  GDSD6  pro_loop  \n",
       "18248  CEBPG  GDSD6  pro_loop  \n",
       "19855  CEBPG  GDSD6  pro_loop  \n",
       "...      ...    ...       ...  \n",
       "69604  CEBPG  GDSD6  pro_loop  \n",
       "72029  CEBPG  GDSD6  pro_loop  \n",
       "76362  CEBPG  GDSD6  pro_loop  \n",
       "78485  CEBPG  GDSD6  pro_loop  \n",
       "81099  CEBPG  GDSD6  pro_loop  \n",
       "\n",
       "[714 rows x 14 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cebpg_kcd6"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "# update 07/15/2021\n",
    "Can you send me the KLF5 TF footprint motif loci + target genes for colon and KC?\n",
    "\n",
    "Also maybe PATZ1 colon vs renal\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "343 target genes found for KLF5 in tissue:  GDSD6\n",
      "['ABCA7', 'ABCF3', 'AC003101.1', 'AC009041.2', 'AC018512.1', 'AC024580.1', 'AC068533.7', 'AC104841.2', 'AC105020.1', 'ACLY', 'ACPT', 'ACTN4', 'ACVR1B', 'ACVRL1', 'ADAM15', 'ADM2', 'ADRA1B', 'AHNAK2', 'AL118506.1', 'AL136376.1', 'AL136531.1', 'ALKBH7', 'AMT', 'ANGPTL2', 'ANKHD1', 'ANKHD1-EIF4EBP3', 'ANKK1', 'ANKZF1', 'ANXA8L1', 'AOC2', 'AP000349.1', 'AP2S1', 'ARAP1', 'ARFRP1', 'ARHGEF12', 'ARMC12', 'ARSG', 'ATG9A', 'ATN1', 'ATP12A', 'ATP5B', 'BAI2', 'BANF2', 'BHLHE41', 'BIRC3', 'C10orf55', 'C11orf84', 'C11orf91', 'C17orf61-PLSCR3', 'C19orf45', 'C1orf134', 'C1orf210', 'C1orf216', 'C1orf56', 'C20orf112', 'C22orf15', 'C2CD4D', 'C4A', 'C6orf223', 'C9orf114', 'C9orf173', 'C9orf69', 'CA11', 'CALML5', 'CARD9', 'CCDC120', 'CCDC51', 'CCDC69', 'CCDC78', 'CCNI2', 'CCT3', 'CDC42SE1', 'CELSR2', 'CEMP1', 'CES5A', 'CHMP2A', 'CHMP4C', 'CISH', 'CLDN6', 'CLK3', 'CNIH2', 'CNTFR', 'COL6A1', 'COPZ2', 'COX7A1', 'CRTC2', 'CRYAB', 'CTD-2267D19.3', 'CTDSP1', 'CTSL', 'CUX1', 'DDX5', 'DLG3', 'DLK1', 'DMKN', 'DMWD', 'DNAAF3', 'DNAL1', 'DNPEP', 'DPM2', 'DRAXIN', 'DUOX1', 'DUOXA1', 'E2F1', 'EHD1', 'EIF4E1B', 'EMD', 'EML3', 'EPHA2', 'ESPN', 'ESYT2', 'FAM110C', 'FAM78A', 'FAM83F', 'FAM84B', 'FDXACB1', 'FKBP10', 'FLAD1', 'FOLR3', 'FOSL1', 'FOXH1', 'FZR1', 'GABPB2', 'GEMIN2', 'GFOD1', 'GGNBP2', 'GIPC3', 'GIT1', 'GNB3', 'GNB5', 'GPHA2', 'GPR107', 'GPSM3', 'GRB10', 'GSTM2', 'HAGHL', 'HCFC1R1', 'HECA', 'HES2', 'HIST1H3B', 'HIST2H3A', 'HMOX2', 'HOXD1', 'HPN', 'IDH1', 'IDH3B', 'IER5L', 'IGFBP5', 'IZUMO1', 'JPH2', 'KCNJ9', 'KCTD18', 'KIAA0754', 'KLHL21', 'KLHL30', 'KM-PA-2', 'LDHA', 'LEPREL4', 'LMF1', 'LMTK3', 'LPAR2', 'LPCAT1', 'LRRC25', 'LSM10', 'LTBR', 'LUC7L3', 'LY6G6C', 'MAD2L2', 'MAF1', 'MAFG', 'MAN2B2', 'MAP2K3', 'MAPKAPK3', 'MCAM', 'MCHR1', 'MED22', 'MESDC1', 'MGAT1', 'MGAT4B', 'MICAL1', 'MICAL3', 'MIF4GD', 'MPHOSPH9', 'MROH1', 'MRPL45', 'MRPS21', 'MS4A10', 'MSL1', 'MT3', 'NANOS3', 'NEFH', 'NHSL1', 'NLGN2', 'NPY5R', 'NR2C2AP', 'NTMT1', 'NUBP2', 'NUPR1L', 'OGFR', 'OPN1MW', 'OPN1MW2', 'OSBPL2', 'OSBPL3', 'OSR1', 'PABPN1L', 'PAFAH1B3', 'PAGR1', 'PCDHB1', 'PDZD4', 'PEBP4', 'PHF13', 'PHGR1', 'PIANP', 'PIGT', 'PIKFYVE', 'PLEKHG2', 'PLK5', 'POLR2L', 'POU2F2', 'PPIB', 'PPM1J', 'PPP1R14B', 'PPP1R16A', 'PPP4C', 'PPP6R1', 'PRCC', 'PREB', 'PRG3', 'PRICKLE3', 'PRKCG', 'PRMT1', 'PROC', 'PRR19', 'PRSS36', 'PSMD8', 'PTMS', 'PTRH1', 'PYCRL', 'QPCTL', 'RAB3A', 'RAPGEF3', 'RARG', 'REC8', 'RHCG', 'RHOV', 'RIN1', 'RINT1', 'RMND5B', 'RNASE13', 'RNF224', 'ROM1', 'RP1-130H16.18', 'RP11-1407O15.2', 'RP11-293M10.1', 'RP11-347C12.3', 'RP11-474G23.1', 'RP11-51F16.8', 'RP13-512J5.1', 'RPL10', 'RPL24', 'RPL3', 'RPL7A', 'RPS9', 'RTN3', 'S100A11', 'SALL2', 'SCN4A', 'SCRT1', 'SDR39U1', 'SEC31A', 'SELO', 'SEMA7A', 'SERPINB13', 'SERPINF2', 'SERTAD3', 'SFN', 'SH3BP1', 'SIAH2', 'SIGIRR', 'SIRT4', 'SLC10A5', 'SLC16A3', 'SLC23A3', 'SLC30A2', 'SLC9A6', 'SNCB', 'SNRPD2', 'SOX8', 'SPDYE2B', 'SQSTM1', 'SRPK2', 'SRSF2', 'SSC5D', 'SSRP1', 'STARD10', 'STARD3', 'STK19', 'SURF1', 'SYNGR3', 'TAOK2', 'TBC1D10C', 'TCHP', 'TEX28', 'THBD', 'THOC6', 'TKTL1', 'TLX1', 'TMA7', 'TMC4', 'TMEM145', 'TMEM256', 'TMEM256-PLSCR3', 'TMEM40', 'TMEM88', 'TNIP1', 'TNNI3', 'TNS4', 'TPM3', 'TRAPPC8', 'TRIM17', 'TRIM72', 'TRPM2', 'TSFM', 'TSPAN31', 'TSPAN4', 'TSSK2', 'TTC39A', 'TUSC2', 'U2AF2', 'UBA7', 'ULK1', 'VASP', 'VGF', 'VPS11', 'WBSCR27', 'WDR64', 'WDR74', 'WNT4', 'ZFAND1', 'ZGPAT', 'ZMYND19', 'ZNF580', 'ZNF581']\n",
      "368 target genes found for KLF5 in tissue:  Colon\n",
      "['ABCA7', 'AC009041.2', 'AC015660.1', 'AC024580.1', 'AC068533.7', 'AC104841.2', 'AC132872.2', 'ACSS2', 'ACTN4', 'ACVRL1', 'ADM', 'ADM2', 'ADRA1B', 'AFAP1', 'AHNAK2', 'AKT1S1', 'AL136376.1', 'AL136531.1', 'AL589739.1', 'AMOTL2', 'AMT', 'ANGPTL2', 'ANKZF1', 'AP000349.1', 'ARAP1', 'ARFRP1', 'ARHGDIA', 'ARHGEF12', 'ARHGEF28', 'ARL14', 'ARMC12', 'ARMC2', 'ARSG', 'ATG9A', 'ATP12A', 'ATP2B3', 'ATP5B', 'BANF2', 'BHLHA9', 'BHLHE41', 'BTBD18', 'C11orf84', 'C11orf91', 'C15orf59', 'C16orf3', 'C19orf43', 'C1orf100', 'C1orf172', 'C1orf200', 'C1orf216', 'C20orf112', 'C22orf15', 'C2CD4D', 'C3orf52', 'C9orf114', 'C9orf173', 'CACNG6', 'CALML5', 'CAMK2N1', 'CARD9', 'CAV1', 'CCDC24', 'CCDC51', 'CCDC69', 'CCDC78', 'CCL27', 'CDC42SE1', 'CEACAM16', 'CELSR2', 'CEMP1', 'CFC1', 'CHMP1A', 'CHMP2A', 'CIRBP', 'CISH', 'CLDN6', 'CLDN7', 'CLEC17A', 'CLK3', 'COPZ2', 'COX7A1', 'CPNE7', 'CRABP2', 'CRCT1', 'CRTC2', 'CTD-3214H19.4', 'CTDSP1', 'CTNND1', 'CTSL', 'CYB561D2', 'DCAKD', 'DCTPP1', 'DDX47', 'DEAF1', 'DENND2D', 'DLG3', 'DNAAF3', 'DNAJC22', 'DNPEP', 'DPM3', 'DYNC1LI2', 'E2F1', 'E4F1', 'EFCAB4B', 'EFNA3', 'EFNA4', 'EMD', 'EML3', 'EN1', 'EPHA2', 'ERCC1', 'ERP27', 'ESPN', 'ESYT2', 'EXOC3L4', 'EXT1', 'FAM110C', 'FAM71F1', 'FAM84B', 'FKBP10', 'FLAD1', 'FLII', 'FOSL1', 'GABPB2', 'GABRP', 'GEMIN2', 'GFAP', 'GGNBP2', 'GGT7', 'GIPC3', 'GJA4', 'GLCCI1', 'GNB5', 'GPATCH4', 'GPHA2', 'GPR107', 'GPR78', 'GPSM3', 'GSDMA', 'GSS', 'GSTM2', 'GSTT1', 'GZF1', 'H3F3B', 'HAGHL', 'HAMP', 'HCFC1R1', 'HECA', 'HES2', 'HIP1', 'HIST2H3A', 'HMOX2', 'HOXB8', 'HPN', 'ICAM1', 'IDH1', 'IDH3B', 'IER5L', 'IFI30', 'IFT122', 'IKZF4', 'IMPDH2', 'INPP5D', 'ITGA3', 'ITPRIP', 'JPH2', 'KCNJ10', 'KCNJ9', 'KIAA0754', 'KLK8', 'LAIR1', 'LDHA', 'LEPREL4', 'LLGL2', 'LMF1', 'LMTK3', 'LRRC25', 'LSM10', 'LSMEM2', 'LUC7L3', 'MADD', 'MAFG', 'MAPKAPK3', 'MB', 'MBD4', 'MDFI', 'MESDC1', 'MGAT4B', 'MICAL1', 'MICAL3', 'MMP25', 'MMRN2', 'MR1', 'MROH1', 'MRPS21', 'MSH5', 'MSH5-SAPCD1', 'MT1G', 'MT3', 'MYADML2', 'MYRF', 'NANOS3', 'NEFH', 'NFKBIE', 'NHSL1', 'NPRL2', 'NR2C2AP', 'NR2F6', 'NRN1L', 'NTMT1', 'NUBP2', 'NUPR1L', 'OMP', 'OPN1MW', 'OPN1MW2', 'OSBPL3', 'OSR1', 'PABPN1L', 'PALLD', 'PCSK4', 'PET100', 'PFN4', 'PHGR1', 'PIGR', 'PIGT', 'PIKFYVE', 'PIM3', 'PKMYT1', 'PLD2', 'PLEKHG2', 'PLK5', 'PNLDC1', 'POLM', 'POP7', 'POU2F2', 'PPIB', 'PPP4C', 'PPP6R1', 'PRAP1', 'PRDX2', 'PREB', 'PRKCG', 'PRMT1', 'PROC', 'PRPF6', 'PRR25', 'PRSS33', 'PSMD8', 'PSME2', 'PTCH1', 'PTMS', 'RAB3A', 'RABGGTA', 'RAPGEF3', 'RARRES1', 'RASGEF1A', 'RASL10A', 'REC8', 'REEP6', 'RGL4', 'RGS11', 'RHCG', 'RINT1', 'RNASE13', 'RNF224', 'RNF31', 'RP1-130H16.18', 'RP11-108K14.8', 'RP11-347C12.3', 'RP11-51F16.8', 'RP11-794P6.2', 'RPL10', 'RPL22', 'RPL24', 'RPL32', 'RPS3', 'S100A1', 'S100A11', 'SALL2', 'SAMD10', 'SCN4A', 'SCNM1', 'SDR39U1', 'SELO', 'SERPINF2', 'SERTAD3', 'SFN', 'SH2B2', 'SHANK2', 'SLC10A5', 'SLC16A3', 'SLC23A3', 'SLC26A9', 'SLC30A2', 'SLC4A9', 'SLC50A1', 'SLC6A12', 'SLC9A6', 'SLMO1', 'SLMO2', 'SMARCD1', 'SNRNP70', 'SOX8', 'SOX9', 'SPATA20', 'SPATA24', 'SPATA33', 'SPDYE2', 'SPDYE2B', 'SPRED2', 'SQSTM1', 'SREBF1', 'SRPK2', 'SRSF2', 'SSC5D', 'SSRP1', 'SSSCA1', 'STARD10', 'STAT5A', 'STMN3', 'SURF1', 'SYNGR3', 'TAB1', 'TBC1D17', 'TBC1D2', 'TEX28', 'THOC6', 'TKTL1', 'TMA7', 'TMC4', 'TMEM145', 'TMEM214', 'TMEM238', 'TMEM88', 'TNFAIP2', 'TNFAIP8L2', 'TNFRSF10C', 'TNIP1', 'TNNI3', 'TRIM31', 'TRIM36', 'TRIM72', 'TRPM2', 'TSFM', 'TSKU', 'TTC39A', 'TTC39C', 'TUBB1', 'U2AF2', 'UBA7', 'VGF', 'VPS37C', 'WBSCR27', 'WDR64', 'WNT5B', 'XAB2', 'XIRP1', 'XRCC3', 'XXYLT1', 'XXcos-LUCA11.5', 'ZDHHC8', 'ZFHX3', 'ZFYVE21', 'ZGPAT', 'ZMYND19', 'ZNF696', 'ZNF771', 'ZNF774']\n"
     ]
    }
   ],
   "source": [
    "KLF5_kcd6 = get_target_genes('KLF5', 'GDSD6')\n",
    "KLF5_colon = get_target_genes('KLF5', 'Colon')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "558 target genes found for PATZ1 in tissue:  Colon\n",
      "['ABCA7', 'AC009041.2', 'AC010327.2', 'AC010646.3', 'AC068533.7', 'AC104841.2', 'AC105020.1', 'AC127496.1', 'AC132216.1', 'ACADVL', 'ACLY', 'ACSL1', 'ACSS2', 'ACTN4', 'ACTR3C', 'ACVRL1', 'ADAM15', 'ADAP1', 'ADM', 'ADM5', 'ADRA1B', 'AFAP1', 'AGXT', 'AHNAK', 'AHNAK2', 'AKT1S1', 'AL031320.1', 'AL136376.1', 'AL136531.1', 'AL589739.1', 'AL645608.1', 'AMOTL2', 'AMT', 'ANKRD66', 'ANKZF1', 'ANXA2', 'AOC2', 'AP000349.1', 'AP000769.1', 'AP1G2', 'APH1A', 'APOBEC3C', 'APOBEC3D', 'ARAP1', 'ARFRP1', 'ARHGDIA', 'ARHGEF12', 'ARHGEF28', 'ARL14', 'ARMC12', 'ATG9A', 'ATN1', 'ATP2A2', 'ATP2B3', 'ATP5B', 'ATP5H', 'B3GAT3', 'B4GALNT1', 'B4GALT3', 'B9D2', 'BANF2', 'BCO2', 'BHLHE41', 'BMF', 'BOLA2B', 'BRMS1L', 'BTBD18', 'C11orf91', 'C16orf3', 'C17orf96', 'C19orf24', 'C19orf43', 'C19orf54', 'C1QB', 'C1QTNF6', 'C1orf100', 'C1orf116', 'C1orf134', 'C1orf172', 'C1orf216', 'C1orf85', 'C20orf112', 'C22orf15', 'C22orf26', 'C22orf29', 'C22orf34', 'C2CD4D', 'C3orf52', 'C4A', 'C8orf82', 'C9orf16', 'C9orf50', 'C9orf78', 'CABIN1', 'CACNA2D4', 'CACNG6', 'CAMK2N1', 'CARD9', 'CAV1', 'CCDC120', 'CCDC24', 'CCDC51', 'CCDC69', 'CCDC78', 'CCL27', 'CCNI2', 'CDK4', 'CDX2', 'CEACAM16', 'CEMP1', 'CES5A', 'CFC1', 'CHCHD1', 'CHMP1A', 'CHMP2A', 'CIB4', 'CIRBP', 'CISH', 'CLCNKA', 'CLDN7', 'CLEC18C', 'CLN5', 'CLP1', 'CLPS', 'CNOT6', 'CNTD2', 'COG2', 'COL6A1', 'CPLX4', 'CRABP2', 'CRCT1', 'CRTC2', 'CSF2', 'CST6', 'CTD-2207O23.10', 'CTD-2545G14.7', 'CTDSP1', 'CTNND1', 'CTRC', 'CUEDC2', 'CUL9', 'CUX2', 'CXorf65', 'CYB561', 'CYGB', 'DAB2IP', 'DCAKD', 'DDB1', 'DDX47', 'DDX5', 'DDX54', 'DEGS2', 'DKC1', 'DLG3', 'DLX4', 'DMPK', 'DNAJC22', 'DNAJC28', 'DNPEP', 'DPM2', 'DRAXIN', 'DSCR3', 'DUOX1', 'DUOXA1', 'DUSP5', 'DUSP7', 'DVL1', 'DYNC1LI2', 'E2F1', 'E4F1', 'ECE1', 'EFNA3', 'EFNA4', 'EGFL8', 'EHD1', 'EIF1', 'EIF4E1B', 'EIF5A', 'EML3', 'EPHA2', 'ERBB2', 'ERCC1', 'ERP27', 'FAM127A', 'FAM129B', 'FAM129C', 'FAM196B', 'FAM73B', 'FAM84B', 'FBXL15', 'FBXL16', 'FBXO36', 'FDXR', 'FIBCD1', 'FLAD1', 'FOSL1', 'FOXA1', 'FOXD1', 'FOXH1', 'FOXK1', 'GALE', 'GCDH', 'GEMIN2', 'GGT7', 'GHDC', 'GJA4', 'GJC2', 'GKAP1', 'GNB1L', 'GNB5', 'GPATCH4', 'GPHA2', 'GPR107', 'GPR142', 'GPSM3', 'GRM4', 'GSDMA', 'GSDMD', 'GSS', 'HAGHL', 'HAMP', 'HIC1', 'HIST2H3A', 'HLCS', 'HNRNPF', 'HOGA1', 'HOXB8', 'HOXC4', 'HOXC9', 'HOXD10', 'HOXD4', 'HPN', 'HSPB9', 'IDH1', 'IDH3B', 'IER5L', 'IFI30', 'IFITM5', 'IFT122', 'IGIP', 'IKZF4', 'IL2RA', 'IMPDH2', 'INMT-FAM188B', 'INPP5D', 'ITPRIP', 'JPH2', 'JPH4', 'KCNJ9', 'KCNK12', 'KDM4B', 'KIAA0754', 'KIRREL2', 'KLHL21', 'KLHL35', 'KRT7', 'KRT81', 'LHPP', 'LHX5', 'LLGL2', 'LMF1', 'LMTK3', 'LRRC25', 'LRRC46', 'LRRC4B', 'LSM10', 'LTB4R2', 'LTBR', 'LUC7L3', 'MAD2L2', 'MADD', 'MAN2B2', 'MAPKAPK3', 'MARCH9', 'MB', 'MBD4', 'MED11', 'MESDC1', 'MGAT4B', 'MICAL1', 'MLLT11', 'MLX', 'MMP14', 'MMP25', 'MMRN2', 'MNAT1', 'MOCS1', 'MORF4L1', 'MPDU1', 'MPHOSPH9', 'MR1', 'MROH1', 'MRPL10', 'MRPL45', 'MRPS21', 'MSL1', 'MT1G', 'MT3', 'MVD', 'MXRA8', 'MYADML2', 'MYOF', 'N4BP1', 'NANOS3', 'NAT14', 'NCCRP1', 'NFKBIE', 'NISCH', 'NLGN2', 'NLRX1', 'NPLOC4', 'NQO2', 'NR2C2AP', 'NR2F6', 'NRN1L', 'NTMT1', 'NTN3', 'NUBP2', 'NUGGC', 'NUTM2A', 'ODF1', 'OSBPL2', 'OSBPL3', 'OSBPL7', 'OTUB1', 'PABPN1L', 'PAGR1', 'PALLD', 'PCDH1', 'PCSK4', 'PDE6C', 'PDE6G', 'PDE6H', 'PDZD7', 'PEG10', 'PGPEP1', 'PHF13', 'PHGR1', 'PI4K2A', 'PIDD', 'PIGR', 'PIKFYVE', 'PIM3', 'PKMYT1', 'PLA2G15', 'PLA2G1B', 'PLA2G4D', 'PLB1', 'PLEKHD1', 'PLEKHM3', 'PLK5', 'PMPCA', 'PNLDC1', 'PNPLA6', 'POLA2', 'POLD4', 'PPIB', 'PPIL2', 'PPM1J', 'PPP1R16A', 'PPP1R9B', 'PPP4C', 'PPP6R1', 'PRAP1', 'PRDX2', 'PRKCG', 'PRMT1', 'PROC', 'PRPF6', 'PRSS33', 'PRSS36', 'PSMC3', 'PSMC3IP', 'PSMC6', 'PSMD8', 'PTCH1', 'PTMS', 'RAB17', 'RAB1B', 'RAD9A', 'RAP1GAP', 'RAPGEF3', 'RASGEF1A', 'RASL10A', 'RBM6', 'REEP4', 'REEP6', 'RELA', 'RELB', 'RGL4', 'RGS11', 'RHCG', 'RHEBL1', 'RHOV', 'RINT1', 'RITA1', 'RNASE13', 'RNF165', 'RNF181', 'RNF223', 'RNF224', 'RP1-130H16.18', 'RP1-4G17.5', 'RP11-108K14.8', 'RP11-201K10.3', 'RP11-293M10.1', 'RP11-444E17.6', 'RP11-51F16.8', 'RP11-542C16.2', 'RP11-762I7.5', 'RP11-794P6.2', 'RPL10', 'RPL18', 'RPL24', 'RPLP2', 'RPS9', 'RTBDN', 'RUSC1', 'S100A1', 'S100A11', 'SALL2', 'SAMD10', 'SCAF8', 'SCN4A', 'SCNM1', 'SDCCAG3', 'SDR39U1', 'SELPLG', 'SEMA7A', 'SEPT1', 'SERPINB6', 'SERPINE1', 'SERTAD3', 'SERTAD4', 'SFTA2', 'SGCE', 'SHANK2', 'SIRT7', 'SIX1', 'SLC10A3', 'SLC15A2', 'SLC26A9', 'SLC30A2', 'SLC4A10', 'SLC4A9', 'SLC50A1', 'SLC52A2', 'SLC6A11', 'SLC6A12', 'SLMO2', 'SLX1A', 'SMCO1', 'SNCB', 'SNX21', 'SOX15', 'SOX18', 'SOX8', 'SOX9', 'SPATA20', 'SPATA33', 'SPDYE2', 'SPDYE2B', 'SPHK2', 'SPRED2', 'SPSB3', 'SQSTM1', 'SREBF1', 'SRPK2', 'SRPK3', 'SRSF2', 'SSBP4', 'SSRP1', 'SSSCA1', 'ST3GAL2', 'STARD10', 'STAT5A', 'STIP1', 'STK40', 'STMN3', 'SULT1A3', 'SYCN', 'SYNGAP1', 'SYNGR3', 'SYT15', 'TAOK2', 'TAP1', 'TBC1D17', 'TBC1D2', 'TCAP', 'TCEA2', 'TEAD3', 'TECR', 'TGFBI', 'TGFBR3L', 'THAP3', 'TINF2', 'TLCD1', 'TLE2', 'TLL2', 'TM4SF1', 'TMA7', 'TMC4', 'TMEM145', 'TMEM14E', 'TMEM238', 'TMEM249', 'TMEM88', 'TMPO', 'TNFAIP8L2', 'TNFRSF13C', 'TNFRSF6B', 'TNFRSF8', 'TNNC2', 'TNNI2', 'TNNI3', 'TOP1', 'TPD52L2', 'TRIM17', 'TRIM31', 'TRIM8', 'TRIOBP', 'TRPM2', 'TRPV4', 'TSPAN4', 'TTC39B', 'TTC39C', 'TTLL10', 'TUBA1B', 'TUBB1', 'TUSC5', 'UBA5', 'UBA7', 'USP20', 'VAMP2', 'VASP', 'VGF', 'VPS37C', 'WBP2', 'WBSCR27', 'WDR1', 'WDR61', 'WDR64', 'WNT5B', 'WSB2', 'XIRP1', 'XXYLT1', 'YPEL4', 'ZDHHC8', 'ZFHX3', 'ZFP36L1', 'ZGPAT', 'ZMYND19', 'ZNF185', 'ZNF444', 'ZNF512B', 'ZNF821', 'ZNF844', 'ZP4']\n",
      "557 target genes found for PATZ1 in tissue:  Renal\n",
      "['ABCA7', 'ABCF3', 'AC009041.2', 'AC010327.2', 'AC010646.3', 'AC018470.1', 'AC018867.1', 'AC068533.7', 'AC104841.2', 'AC105020.1', 'AC114783.1', 'AC127496.1', 'AC132216.1', 'ACADVL', 'ACSL1', 'ACSS2', 'ACTG1', 'ACTN4', 'ACVRL1', 'ADAM15', 'ADAP1', 'ADM', 'ADM2', 'ADRA1B', 'AFAP1', 'AGXT', 'AHNAK', 'AHNAK2', 'AKR1B1', 'AKT1S1', 'AL109659.1', 'AL136376.1', 'AL136531.1', 'AL158091.1', 'AL590714.1', 'AL645608.1', 'ALKBH7', 'ANKRD1', 'ANKRD66', 'ANKZF1', 'ANXA2', 'AOC2', 'AP000349.1', 'AP1G2', 'APLP1', 'APOBEC3C', 'APOBEC3D', 'APOLD1', 'ARFRP1', 'ARHGDIA', 'ARHGEF12', 'ARL14', 'ARMC12', 'ARMC2', 'ARSG', 'ATG9A', 'ATP2A2', 'ATP2B3', 'ATP5B', 'AURKAIP1', 'B3GAT3', 'B4GALNT1', 'B4GALT3', 'B9D2', 'BANF2', 'BCO2', 'BFSP1', 'BHLHE41', 'BIN3', 'BMF', 'BOLA2B', 'BRMS1L', 'BTBD18', 'C10orf129', 'C11orf91', 'C16orf3', 'C17orf96', 'C19orf24', 'C19orf25', 'C19orf48', 'C19orf67', 'C1QB', 'C1QTNF5', 'C1QTNF6', 'C1orf100', 'C1orf116', 'C1orf134', 'C1orf172', 'C1orf210', 'C1orf216', 'C1orf85', 'C20orf112', 'C22orf15', 'C22orf26', 'C22orf29', 'C2CD4D', 'C3orf52', 'C4A', 'C8orf82', 'C9orf50', 'CACNA1I', 'CACNG6', 'CALML5', 'CAMK2N1', 'CAPN8', 'CAPN9', 'CARD9', 'CAV1', 'CCDC24', 'CCDC51', 'CCDC69', 'CCDC78', 'CCNI2', 'CCR10', 'CDK4', 'CDX2', 'CEBPA', 'CEMP1', 'CES5A', 'CHCHD1', 'CHD9', 'CHMP1A', 'CHMP2A', 'CIDEB', 'CISH', 'CLDN23', 'CLDN7', 'CLN5', 'CLP1', 'COL6A1', 'CPLX4', 'CRABP2', 'CRYGN', 'CSF2', 'CST6', 'CTDSP1', 'CTNND1', 'CTRC', 'CTSE', 'CUEDC2', 'CUX2', 'CXXC11', 'CXorf65', 'CYB561', 'CYGB', 'DCAKD', 'DDX47', 'DDX5', 'DEFB134', 'DEGS2', 'DKC1', 'DLG3', 'DLL1', 'DLX4', 'DMPK', 'DNAJC22', 'DNAJC28', 'DNAJC5G', 'DNPEP', 'DRAXIN', 'DSCR3', 'DUSP7', 'DYNC1LI2', 'E4F1', 'ECE1', 'EFNA3', 'EFNA4', 'EGFL8', 'EHD1', 'EIF1', 'EIF4E1B', 'ELP5', 'EML3', 'EPHA2', 'ERBB2', 'ERP27', 'FAM120B', 'FAM127A', 'FAM129B', 'FAM129C', 'FAM196B', 'FAM213B', 'FAM73B', 'FAU', 'FBXL15', 'FBXL16', 'FIBCD1', 'FLAD1', 'FOSL1', 'FOXA1', 'FOXD1', 'FOXG1', 'FOXH1', 'GABRP', 'GADD45G', 'GAL3ST4', 'GALE', 'GATA2', 'GCDH', 'GEMIN2', 'GFER', 'GGT7', 'GHDC', 'GJA4', 'GJC2', 'GKAP1', 'GNB1L', 'GPAT2', 'GPATCH3', 'GPATCH4', 'GPHN', 'GPR107', 'GPR37L1', 'GRM4', 'GSDMA', 'GSDMD', 'GSS', 'GSTA4', 'GSTM2', 'H2AFX', 'HAGHL', 'HAMP', 'HIC1', 'HIST1H3B', 'HIST2H2AA3', 'HIST2H3A', 'HNRNPF', 'HNRNPH1', 'HOGA1', 'HOXB8', 'HOXC4', 'HOXD10', 'HPN', 'HSPB9', 'HUG1', 'ICAM1', 'IDH1', 'IER5L', 'IFI30', 'IFITM5', 'IFT122', 'IGIP', 'IKZF4', 'IL2RA', 'IMPDH2', 'INF2', 'INMT-FAM188B', 'ITGA3', 'ITPRIP', 'JPH2', 'JPH4', 'KCNA10', 'KCNJ9', 'KCNK12', 'KDM4B', 'KIAA1875', 'KIRREL2', 'KLF6', 'KLHL21', 'KMT2E', 'KRT81', 'LAMC3', 'LAPTM5', 'LHPP', 'LILRA6', 'LILRB3', 'LMF1', 'LMTK3', 'LRP5', 'LRRC25', 'LRRC46', 'LRRC4B', 'LRTM2', 'LSM10', 'LTB4R', 'LTB4R2', 'LTBR', 'LUZP1', 'MAD2L2', 'MADD', 'MAN2B2', 'MAPKAPK3', 'MARCH9', 'MB', 'MBD4', 'MESDC1', 'MFRP', 'MFSD2A', 'MGAT4B', 'MICAL1', 'MICALL2', 'MLLT11', 'MLPH', 'MLX', 'MMP14', 'MMRN2', 'MNAT1', 'MOCS1', 'MORF4L1', 'MPHOSPH9', 'MR1', 'MRPL10', 'MRPL45', 'MRPL52', 'MS4A10', 'MSL1', 'MT3', 'MVD', 'MYADML2', 'N4BP1', 'NAT14', 'NCCRP1', 'NEUROD2', 'NISCH', 'NLGN2', 'NOXO1', 'NPAS4', 'NPFFR1', 'NPLOC4', 'NRN1L', 'NTHL1', 'NTMT1', 'NUBP2', 'NUDC', 'OBP2A', 'ODF1', 'OMP', 'OSBPL3', 'OSBPL7', 'OSM', 'OTUB1', 'PABPN1L', 'PAGR1', 'PALLD', 'PCDH1', 'PCDHGB6', 'PCSK4', 'PDCD1LG2', 'PDE6G', 'PDE6H', 'PEX16', 'PGPEP1', 'PHF13', 'PI4K2A', 'PIDD', 'PIKFYVE', 'PIM3', 'PITX3', 'PKMYT1', 'PLA2G1B', 'PLA2G2C', 'PLB1', 'PLCH2', 'PLEKHG2', 'PLEKHM3', 'PLK5', 'PML', 'PMPCA', 'PNLDC1', 'POLA2', 'POLR2I', 'PPIL2', 'PPM1J', 'PPP1R9B', 'PPP4C', 'PPP6R1', 'PRAP1', 'PRCC', 'PRDX2', 'PRMT1', 'PROB1', 'PROC', 'PRPF6', 'PRSS33', 'PSMC3', 'PSMC3IP', 'PSMC6', 'PSMD8', 'PTCH1', 'PTMS', 'PYCRL', 'RAB17', 'RAD9A', 'RAP1GAP', 'RAPGEF3', 'RASGEF1A', 'RASL10A', 'REEP6', 'RELA', 'RGS11', 'RHEBL1', 'RHOV', 'RINT1', 'RNASE13', 'RNF165', 'RNF19B', 'RNF223', 'RNF224', 'RP1-130H16.18', 'RP1-4G17.5', 'RP11-108K14.8', 'RP11-201K10.3', 'RP11-293M10.1', 'RP11-444E17.6', 'RP11-51F16.8', 'RP11-794P6.2', 'RPL10', 'RPL18', 'RPL24', 'RPLP2', 'RPS9', 'RUSC1', 'S100A1', 'S100A11', 'SALL2', 'SAMD10', 'SCAF8', 'SCN4A', 'SCN4B', 'SDCCAG3', 'SDIM1', 'SDR39U1', 'SELPLG', 'SEMA7A', 'SEPT1', 'SEPT3', 'SERPINE1', 'SERTAD3', 'SETD7', 'SF1', 'SF3B4', 'SGCA', 'SHANK2', 'SIPA1', 'SIRT7', 'SIX1', 'SLC10A3', 'SLC14A1', 'SLC16A3', 'SLC22A18AS', 'SLC26A9', 'SLC2A6', 'SLC30A2', 'SLC30A3', 'SLC30A8', 'SLC35B2', 'SLC4A9', 'SLC50A1', 'SLC52A2', 'SLC6A11', 'SLC7A7', 'SLX1A', 'SMAD7', 'SNCB', 'SNRNP70', 'SNX21', 'SOX18', 'SOX8', 'SOX9', 'SP2', 'SPATA33', 'SPDYE2', 'SPDYE2B', 'SPHK2', 'SPSB3', 'SPTBN2', 'SQSTM1', 'SREBF1', 'SRPK2', 'SRSF2', 'SRSF5', 'SSBP4', 'SSRP1', 'SSSCA1', 'ST3GAL2', 'STMN3', 'STOML1', 'STRA6', 'SULT1A3', 'SYCN', 'SYNGR3', 'TACC3', 'TAOK2', 'TAP1', 'TBC1D17', 'TBC1D2', 'TBCB', 'TCEA2', 'TEAD3', 'TECR', 'TGFBI', 'TGFBR3L', 'THAP3', 'TINF2', 'TLCD1', 'TLE2', 'TLL2', 'TLR9', 'TM4SF1', 'TMA7', 'TMEM129', 'TMEM145', 'TMEM180', 'TMEM249', 'TMEM88', 'TMPO', 'TNFAIP8L3', 'TNFRSF13C', 'TNFRSF6B', 'TNFRSF8', 'TNNC2', 'TNNI3', 'TNS3', 'TOP1', 'TPPP', 'TPST2', 'TRADD', 'TRIM17', 'TRIM36', 'TRIM8', 'TRIOBP', 'TRMT61A', 'TRPC6', 'TSC2', 'TSFM', 'TTC39B', 'TTLL10', 'TUBA1B', 'TUBB1', 'TUSC5', 'U2AF2', 'U82695.9', 'UBA7', 'VGF', 'WBP2', 'WBSCR27', 'WDR1', 'WDR64', 'WDR74', 'WNT16', 'WNT5B', 'WNT6', 'XXYLT1', 'YPEL4', 'ZDHHC8', 'ZFHX3', 'ZFP36L1', 'ZGPAT', 'ZIC2', 'ZMYND19', 'ZNF185', 'ZNF345', 'ZNF512B', 'ZNF629', 'ZNF821', 'ZNF844', 'ZP4']\n"
     ]
    }
   ],
   "source": [
    "PATZ1_colon = get_target_genes('PATZ1', 'Colon')\n",
    "PATZ1_renal = get_target_genes('PATZ1', 'Renal')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "KLF5_kcd6.to_csv('../data/processed/fig7_vocab_go/genomic_instance_overlap/KLF5_kcd6_manual.csv')\n",
    "KLF5_colon.to_csv('../data/processed/fig7_vocab_go/genomic_instance_overlap/KLF5_colon_manual.csv')\n",
    "PATZ1_colon.to_csv('../data/processed/fig7_vocab_go/genomic_instance_overlap/PATZ1_colon_manual.csv')\n",
    "PATZ1_renal.to_csv('../data/processed/fig7_vocab_go/genomic_instance_overlap/PATZ1_renal_manual.csv')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# update 10/13/2021\n",
    "\n",
    "Also, can you find the genomic instances (chr location) of ARNT+SP1 and the TF footprint nucleotides associated with then in A431s looped to or in the promoter of the CYP1B1 gene?\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vocab = 'ARNT::SP1'\n",
    "df =get_genomic_instances_all(vocab, 'SCC', ['CYP1B1'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('../data/processed/fig7_vocab_go/genomic_instance_overlap/ARNT::SP1_a431_instances_all.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "# tissue='A431-CTRLi'\n",
    "# list_genes = ['CYP1B1']\n",
    "# anc_prom = pybedtools.BedTool('../data/interim/annon/promoter_anchors/promoter_'+tissue+'_annon.bed').to_dataframe()\n",
    "# anc_prom = anc_prom[anc_prom.name.isin(list_genes)]\n",
    "# anc_prom_arr = anc_prom.thickEnd.unique()\n",
    "    \n",
    "# loop_df = pd.read_csv('../data/interim/merged/loops/'+tissue+'.loops.csv',index_col=0)\n",
    "# loop_df = loop_df[loop_df.source.isin(anc_prom_arr)|loop_df.target.isin(anc_prom_arr)]\n",
    "# all_anc = sorted(set(anc_prom.thickEnd.to_list() +loop_df.source.to_list() + loop_df.target.to_list() ))\n",
    "\n",
    "# anc_foot = pybedtools.BedTool('../data/interim/annon/anchor_footprinting/'+tissue+'_annon.bed').to_dataframe()\n",
    "# anc_foot['tf'] = anc_foot.thickEnd.map(tf_id_to_name_dict)\n",
    "# anc_foot = anc_foot[anc_foot.tf.isin(vocab.split('::'))]\n",
    "# anc_foot = anc_foot[anc_foot.name.isin(all_anc)]\n",
    "\n",
    "\n",
    "# pro_anc_foot_df = pybedtools.BedTool('../data/interim/annon/promoter_footprinting/promoter_'+tissue+'_annon.bed').to_dataframe()\n",
    "# pro_anc_foot_df['tf'] = pro_anc_foot_df.thickEnd.map(tf_id_to_name_dict)\n",
    "# pro_anc_foot = pd.DataFrame()\n",
    "\n",
    "# for gene in list_genes:\n",
    "#     pro_anc_foot1 = pro_anc_foot_df[(pro_anc_foot_df.name==gene)|(pro_anc_foot_df.tf.isin(vocab.split('::')))]# not perfect will catch later???\n",
    "#     pro_anc_foot1['anchor'] = pro_anc_foot_df.score + \"_\" + pro_anc_foot_df.strand.map(str)+\"_\"+ pro_anc_foot_df.thickStart.map(str)\n",
    "#     pro_anc_foot1 = pro_anc_foot1[pro_anc_foot1.anchor.isin(all_anc)]\n",
    "#     pro_anc_foot1.drop(columns=['anchor'],inplace=True)\n",
    "#     pro_anc_foot = pd.concat([pro_anc_foot, pro_anc_foot1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ARNT::SP1'"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "tissue = 'A431-CTRLi'\n",
    "anc_prom = pybedtools.BedTool('../data/interim/annon/promoter_anchors/promoter_'+tissue+'_annon.bed').to_dataframe()\n",
    "loop_df = pd.read_csv('../data/interim/merged/loops/'+tissue+'.loops.csv',index_col=0)\n",
    "anc_foot = pybedtools.BedTool('../data/interim/annon/anchor_footprinting/'+tissue+'_annon.bed').to_dataframe()\n",
    "anc_foot['tf'] = anc_foot.thickEnd.map(tf_id_to_name_dict)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So ARNT-SP1, SP1-ZBTB14, and SP1-ZNF770 genomic loci + associated target genes in SCC would be great (edited) \n",
    "\n",
    "Laura Donohue  11 minutes ago\n",
    "Second tier would be EGR2-NFE2L1, NFE2L1-NR3C1, and SOX10-SOX13 in Melanoma\n",
    "Third tier is EGR2-FOSL2, EGR2-RORA, HMGA1-KLF5, KLF5-MAFK, and PRDM1-RORA in KC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(((())))\n",
      "ARNT::SP1 SCC 828\n",
      "(((())))\n",
      "ARNT::SP1 SCC 828\n",
      "pro-pro instances (59, 15)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:37: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:49: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:50: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:62: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pro-loop instances (2237, 15)\n",
      "loop-loop instances (0, 15)\n",
      "overall instances (2296, 15)\n",
      "(((())))\n",
      "SP1::ZBTB14 SCC 1018\n",
      "(((())))\n",
      "SP1::ZBTB14 SCC 1018\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:224: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pro-pro instances (376, 15)\n",
      "pro-loop instances (3546, 15)\n",
      "loop-loop instances (59, 15)\n",
      "overall instances (3981, 15)\n",
      "(((())))\n",
      "SP1::ZNF770 SCC 1523\n",
      "(((())))\n",
      "SP1::ZNF770 SCC 1523\n",
      "pro-pro instances (1215, 15)\n",
      "pro-loop instances (15836, 15)\n",
      "loop-loop instances (99, 15)\n",
      "overall instances (17150, 15)\n",
      "CPU times: user 3h 5min 50s, sys: 3min 4s, total: 3h 8min 55s\n",
      "Wall time: 1h 7min 41s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# vocab = 'ARNT::SP1'\n",
    "for vocab in ['ARNT::SP1','SP1::ZBTB14','SP1::ZNF770']:\n",
    "    list_genes = vocab_to_target_genes[vocab]\n",
    "    df =get_genomic_instances_all(vocab, 'SCC', list_genes)\n",
    "    df.to_csv('../data/processed/fig7_vocab_go/genomic_instance_overlap/'+vocab+'_a431_instances_all.csv')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=======================\n",
      "ARNT::SP1 A431-CTRLi\n",
      "AARS2\n",
      "anc_prom (2, 9)\n",
      "num all anchors 28601\n",
      "num promoter anchors 2\n",
      "num looped anchors 28599\n",
      "NO INSTANCES FOUND - no promoter footprints\n",
      "=======================\n",
      "ARNT::SP1 A431-CTRLi\n",
      "ABCA7\n",
      "anc_prom (3, 9)\n",
      "num all anchors 28616\n",
      "num promoter anchors 3\n",
      "num looped anchors 28613\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:35: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr19_1025000_1030000 34\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:47: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NO INSTANCES FOUND - no looped footprints\n",
      "=======================\n",
      "ARNT::SP1 A431-CTRLi\n",
      "ACBD4\n",
      "anc_prom (3, 9)\n",
      "num all anchors 28649\n",
      "num promoter anchors 3\n",
      "num looped anchors 28646\n",
      "chr17_43200000_43205000 81\n",
      "NO INSTANCES FOUND - no looped footprints\n",
      "=======================\n",
      "ARNT::SP1 A431-CTRLi\n",
      "ACER2\n",
      "anc_prom (2, 9)\n",
      "num all anchors 28620\n",
      "num promoter anchors 2\n",
      "num looped anchors 28618\n",
      "chr9_19395000_19400000 20\n",
      "NO INSTANCES FOUND - no looped footprints\n",
      "=======================\n",
      "ARNT::SP1 A431-CTRLi\n",
      "ACLY\n",
      "anc_prom (3, 9)\n",
      "num all anchors 28614\n",
      "num promoter anchors 3\n",
      "num looped anchors 28611\n",
      "chr17_40070000_40075000 13\n",
      "chr17_40075000_40080000 32\n",
      "NO INSTANCES FOUND - no looped footprints\n",
      "=======================\n",
      "ARNT::SP1 A431-CTRLi\n",
      "ACTG1\n",
      "anc_prom (3, 9)\n",
      "num all anchors 28703\n",
      "num promoter anchors 3\n",
      "num looped anchors 28700\n",
      "chr17_79480000_79485000 164\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:48: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr17_79485000_79490000 10\n",
      "number of promoter footprinted anchors with vocab (10, 13)\n",
      "number of looped footprinted anchors with vocab (10, 13)\n",
      "*** genomic instances (15, 15)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>TSS</th>\n",
       "      <th>chr_m</th>\n",
       "      <th>start_m</th>\n",
       "      <th>stop_m</th>\n",
       "      <th>id_trim</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>tf</th>\n",
       "      <th>type</th>\n",
       "      <th>vocab</th>\n",
       "      <th>tissue</th>\n",
       "      <th>config</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>1025338</td>\n",
       "      <td>chr17</td>\n",
       "      <td>79480000</td>\n",
       "      <td>79485000</td>\n",
       "      <td>chr17_79480000_79485000</td>\n",
       "      <td>chr17</td>\n",
       "      <td>79480116</td>\n",
       "      <td>79480137</td>\n",
       "      <td>SP1_HUMAN.H11MO.0.A.pwm.trim</td>\n",
       "      <td>14.500829</td>\n",
       "      <td>-</td>\n",
       "      <td>SP1</td>\n",
       "      <td>pro_1</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>A431-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1025346</td>\n",
       "      <td>chr17</td>\n",
       "      <td>79480000</td>\n",
       "      <td>79485000</td>\n",
       "      <td>chr17_79480000_79485000</td>\n",
       "      <td>chr17</td>\n",
       "      <td>79480126</td>\n",
       "      <td>79480136</td>\n",
       "      <td>SP1_HUMAN.H11MO.1.A.pwm.trim</td>\n",
       "      <td>13.382828</td>\n",
       "      <td>-</td>\n",
       "      <td>SP1</td>\n",
       "      <td>pro_1</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>A431-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1025365</td>\n",
       "      <td>chr17</td>\n",
       "      <td>79480000</td>\n",
       "      <td>79485000</td>\n",
       "      <td>chr17_79480000_79485000</td>\n",
       "      <td>chr17</td>\n",
       "      <td>79480189</td>\n",
       "      <td>79480199</td>\n",
       "      <td>SP1_HUMAN.H11MO.1.A.pwm.trim</td>\n",
       "      <td>18.026685</td>\n",
       "      <td>+</td>\n",
       "      <td>SP1</td>\n",
       "      <td>pro_1</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>A431-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1025374</td>\n",
       "      <td>chr17</td>\n",
       "      <td>79480000</td>\n",
       "      <td>79485000</td>\n",
       "      <td>chr17_79480000_79485000</td>\n",
       "      <td>chr17</td>\n",
       "      <td>79480193</td>\n",
       "      <td>79480214</td>\n",
       "      <td>SP1_HUMAN.H11MO.0.A.pwm.trim</td>\n",
       "      <td>7.210513</td>\n",
       "      <td>+</td>\n",
       "      <td>SP1</td>\n",
       "      <td>pro_1</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>A431-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1025393</td>\n",
       "      <td>chr17</td>\n",
       "      <td>79480000</td>\n",
       "      <td>79485000</td>\n",
       "      <td>chr17_79480000_79485000</td>\n",
       "      <td>chr17</td>\n",
       "      <td>79480579</td>\n",
       "      <td>79480600</td>\n",
       "      <td>SP1_HUMAN.H11MO.0.A.pwm.trim</td>\n",
       "      <td>10.309607</td>\n",
       "      <td>-</td>\n",
       "      <td>SP1</td>\n",
       "      <td>pro_1</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>A431-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1025404</td>\n",
       "      <td>chr17</td>\n",
       "      <td>79480000</td>\n",
       "      <td>79485000</td>\n",
       "      <td>chr17_79480000_79485000</td>\n",
       "      <td>chr17</td>\n",
       "      <td>79480589</td>\n",
       "      <td>79480599</td>\n",
       "      <td>SP1_HUMAN.H11MO.1.A.pwm.trim</td>\n",
       "      <td>14.347540</td>\n",
       "      <td>-</td>\n",
       "      <td>SP1</td>\n",
       "      <td>pro_1</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>A431-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1025523</td>\n",
       "      <td>chr17</td>\n",
       "      <td>79480000</td>\n",
       "      <td>79485000</td>\n",
       "      <td>chr17_79480000_79485000</td>\n",
       "      <td>chr17</td>\n",
       "      <td>79482054</td>\n",
       "      <td>79482064</td>\n",
       "      <td>SP1_HUMAN.H11MO.1.A.pwm.trim</td>\n",
       "      <td>11.402480</td>\n",
       "      <td>+</td>\n",
       "      <td>SP1</td>\n",
       "      <td>pro_1</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>A431-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1025556</td>\n",
       "      <td>chr17</td>\n",
       "      <td>79485000</td>\n",
       "      <td>79490000</td>\n",
       "      <td>chr17_79485000_79490000</td>\n",
       "      <td>chr17</td>\n",
       "      <td>79486071</td>\n",
       "      <td>79486081</td>\n",
       "      <td>SP1_HUMAN.H11MO.1.A.pwm.trim</td>\n",
       "      <td>13.620915</td>\n",
       "      <td>-</td>\n",
       "      <td>SP1</td>\n",
       "      <td>pro_2</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>A431-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1025587</td>\n",
       "      <td>chr17</td>\n",
       "      <td>79485000</td>\n",
       "      <td>79490000</td>\n",
       "      <td>chr17_79485000_79490000</td>\n",
       "      <td>chr17</td>\n",
       "      <td>79486628</td>\n",
       "      <td>79486649</td>\n",
       "      <td>SP1_HUMAN.H11MO.0.A.pwm.trim</td>\n",
       "      <td>6.760533</td>\n",
       "      <td>-</td>\n",
       "      <td>SP1</td>\n",
       "      <td>pro_2</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>A431-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1025616</td>\n",
       "      <td>chr17</td>\n",
       "      <td>79485000</td>\n",
       "      <td>79490000</td>\n",
       "      <td>chr17_79485000_79490000</td>\n",
       "      <td>chr17</td>\n",
       "      <td>79486648</td>\n",
       "      <td>79486658</td>\n",
       "      <td>SP1_HUMAN.H11MO.1.A.pwm.trim</td>\n",
       "      <td>15.451729</td>\n",
       "      <td>+</td>\n",
       "      <td>SP1</td>\n",
       "      <td>pro_2</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>A431-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1022252</td>\n",
       "      <td>chr17</td>\n",
       "      <td>78385000</td>\n",
       "      <td>78390000</td>\n",
       "      <td>chr17_78385000_78390000</td>\n",
       "      <td>chr17</td>\n",
       "      <td>78388964</td>\n",
       "      <td>78388970</td>\n",
       "      <td>ARNT_HUMAN.H11MO.0.B.pwm.trim</td>\n",
       "      <td>10.397177</td>\n",
       "      <td>+</td>\n",
       "      <td>ARNT</td>\n",
       "      <td>pro_1+loop</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>A431-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1024289</td>\n",
       "      <td>chr17</td>\n",
       "      <td>79090000</td>\n",
       "      <td>79095000</td>\n",
       "      <td>chr17_79090000_79095000</td>\n",
       "      <td>chr17</td>\n",
       "      <td>79093208</td>\n",
       "      <td>79093214</td>\n",
       "      <td>ARNT_HUMAN.H11MO.0.B.pwm.trim</td>\n",
       "      <td>10.397177</td>\n",
       "      <td>+</td>\n",
       "      <td>ARNT</td>\n",
       "      <td>pro_1+loop</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>A431-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1026790</td>\n",
       "      <td>chr17</td>\n",
       "      <td>79865000</td>\n",
       "      <td>79870000</td>\n",
       "      <td>chr17_79865000_79870000</td>\n",
       "      <td>chr17</td>\n",
       "      <td>79868575</td>\n",
       "      <td>79868581</td>\n",
       "      <td>ARNT_HUMAN.H11MO.0.B.pwm.trim</td>\n",
       "      <td>10.397177</td>\n",
       "      <td>-</td>\n",
       "      <td>ARNT</td>\n",
       "      <td>pro_1+loop</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>A431-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1028492</td>\n",
       "      <td>chr17</td>\n",
       "      <td>80170000</td>\n",
       "      <td>80175000</td>\n",
       "      <td>chr17_80170000_80175000</td>\n",
       "      <td>chr17</td>\n",
       "      <td>80171372</td>\n",
       "      <td>80171378</td>\n",
       "      <td>ARNT_HUMAN.H11MO.0.B.pwm.trim</td>\n",
       "      <td>10.397177</td>\n",
       "      <td>-</td>\n",
       "      <td>ARNT</td>\n",
       "      <td>pro_1+loop</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>A431-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1029136</td>\n",
       "      <td>chr17</td>\n",
       "      <td>80250000</td>\n",
       "      <td>80255000</td>\n",
       "      <td>chr17_80250000_80255000</td>\n",
       "      <td>chr17</td>\n",
       "      <td>80250724</td>\n",
       "      <td>80250730</td>\n",
       "      <td>ARNT_HUMAN.H11MO.0.B.pwm.trim</td>\n",
       "      <td>10.397177</td>\n",
       "      <td>+</td>\n",
       "      <td>ARNT</td>\n",
       "      <td>pro_1+loop</td>\n",
       "      <td>ARNT::SP1</td>\n",
       "      <td>A431-CTRLi</td>\n",
       "      <td>pro_loop</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           chr     start      stop                      TSS  chr_m   start_m  \\\n",
       "1025338  chr17  79480000  79485000  chr17_79480000_79485000  chr17  79480116   \n",
       "1025346  chr17  79480000  79485000  chr17_79480000_79485000  chr17  79480126   \n",
       "1025365  chr17  79480000  79485000  chr17_79480000_79485000  chr17  79480189   \n",
       "1025374  chr17  79480000  79485000  chr17_79480000_79485000  chr17  79480193   \n",
       "1025393  chr17  79480000  79485000  chr17_79480000_79485000  chr17  79480579   \n",
       "1025404  chr17  79480000  79485000  chr17_79480000_79485000  chr17  79480589   \n",
       "1025523  chr17  79480000  79485000  chr17_79480000_79485000  chr17  79482054   \n",
       "1025556  chr17  79485000  79490000  chr17_79485000_79490000  chr17  79486071   \n",
       "1025587  chr17  79485000  79490000  chr17_79485000_79490000  chr17  79486628   \n",
       "1025616  chr17  79485000  79490000  chr17_79485000_79490000  chr17  79486648   \n",
       "1022252  chr17  78385000  78390000  chr17_78385000_78390000  chr17  78388964   \n",
       "1024289  chr17  79090000  79095000  chr17_79090000_79095000  chr17  79093208   \n",
       "1026790  chr17  79865000  79870000  chr17_79865000_79870000  chr17  79868575   \n",
       "1028492  chr17  80170000  80175000  chr17_80170000_80175000  chr17  80171372   \n",
       "1029136  chr17  80250000  80255000  chr17_80250000_80255000  chr17  80250724   \n",
       "\n",
       "           stop_m                        id_trim      score strand    tf  \\\n",
       "1025338  79480137   SP1_HUMAN.H11MO.0.A.pwm.trim  14.500829      -   SP1   \n",
       "1025346  79480136   SP1_HUMAN.H11MO.1.A.pwm.trim  13.382828      -   SP1   \n",
       "1025365  79480199   SP1_HUMAN.H11MO.1.A.pwm.trim  18.026685      +   SP1   \n",
       "1025374  79480214   SP1_HUMAN.H11MO.0.A.pwm.trim   7.210513      +   SP1   \n",
       "1025393  79480600   SP1_HUMAN.H11MO.0.A.pwm.trim  10.309607      -   SP1   \n",
       "1025404  79480599   SP1_HUMAN.H11MO.1.A.pwm.trim  14.347540      -   SP1   \n",
       "1025523  79482064   SP1_HUMAN.H11MO.1.A.pwm.trim  11.402480      +   SP1   \n",
       "1025556  79486081   SP1_HUMAN.H11MO.1.A.pwm.trim  13.620915      -   SP1   \n",
       "1025587  79486649   SP1_HUMAN.H11MO.0.A.pwm.trim   6.760533      -   SP1   \n",
       "1025616  79486658   SP1_HUMAN.H11MO.1.A.pwm.trim  15.451729      +   SP1   \n",
       "1022252  78388970  ARNT_HUMAN.H11MO.0.B.pwm.trim  10.397177      +  ARNT   \n",
       "1024289  79093214  ARNT_HUMAN.H11MO.0.B.pwm.trim  10.397177      +  ARNT   \n",
       "1026790  79868581  ARNT_HUMAN.H11MO.0.B.pwm.trim  10.397177      -  ARNT   \n",
       "1028492  80171378  ARNT_HUMAN.H11MO.0.B.pwm.trim  10.397177      -  ARNT   \n",
       "1029136  80250730  ARNT_HUMAN.H11MO.0.B.pwm.trim  10.397177      +  ARNT   \n",
       "\n",
       "               type      vocab      tissue    config  \n",
       "1025338       pro_1  ARNT::SP1  A431-CTRLi  pro_loop  \n",
       "1025346       pro_1  ARNT::SP1  A431-CTRLi  pro_loop  \n",
       "1025365       pro_1  ARNT::SP1  A431-CTRLi  pro_loop  \n",
       "1025374       pro_1  ARNT::SP1  A431-CTRLi  pro_loop  \n",
       "1025393       pro_1  ARNT::SP1  A431-CTRLi  pro_loop  \n",
       "1025404       pro_1  ARNT::SP1  A431-CTRLi  pro_loop  \n",
       "1025523       pro_1  ARNT::SP1  A431-CTRLi  pro_loop  \n",
       "1025556       pro_2  ARNT::SP1  A431-CTRLi  pro_loop  \n",
       "1025587       pro_2  ARNT::SP1  A431-CTRLi  pro_loop  \n",
       "1025616       pro_2  ARNT::SP1  A431-CTRLi  pro_loop  \n",
       "1022252  pro_1+loop  ARNT::SP1  A431-CTRLi  pro_loop  \n",
       "1024289  pro_1+loop  ARNT::SP1  A431-CTRLi  pro_loop  \n",
       "1026790  pro_1+loop  ARNT::SP1  A431-CTRLi  pro_loop  \n",
       "1028492  pro_1+loop  ARNT::SP1  A431-CTRLi  pro_loop  \n",
       "1029136  pro_1+loop  ARNT::SP1  A431-CTRLi  pro_loop  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=======================\n",
      "ARNT::SP1 A431-CTRLi\n",
      "ACTN4\n",
      "anc_prom (3, 9)\n",
      "num all anchors 28624\n",
      "num promoter anchors 3\n",
      "num looped anchors 28621\n",
      "chr19_39125000_39130000 39\n",
      "chr19_39130000_39135000 17\n",
      "NO INSTANCES FOUND - no looped footprints\n",
      "=======================\n",
      "ARNT::SP1 A431-CTRLi\n",
      "ADAP1\n",
      "anc_prom (3, 9)\n",
      "num all anchors 28628\n",
      "num promoter anchors 3\n",
      "num looped anchors 28625\n",
      "chr7_985000_990000 48\n",
      "chr7_990000_995000 3\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-121-a516d7f97800>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mgene\u001b[0m \u001b[0;32min\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m'AARS2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ABCA7'\u001b[0m\u001b[0;34m,\u001b[0m  \u001b[0;34m'ACBD4'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ACER2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ACLY'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ACTG1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ACTN4'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ADAP1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ADM'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ADM5'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ADRA1B'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'AHNAK'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'AHNAK2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'AL109659.1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'AL118506.1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'AL136376.1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'AL136531.1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'AL590714.1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ALDH3B1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ANKRD9'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ANKZF1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ANO9'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'AOC2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'AP000349.1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'AP3D1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'APH1A'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'APOC1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ARAP1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ARHGEF12'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ARHGEF4'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ARRB1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ARSG'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ASAP3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ASS1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ATAD3B'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ATG9A'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ATP11C'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ATP2A2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ATP2C2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ATP5B'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'B4GALT3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'B9D2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'BIRC3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'BMF'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'BTBD18'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'C12orf36'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'C16orf3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'C17orf100'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'C17orf96'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'C19orf24'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'C19orf55'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'C1QTNF6'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'C1orf200'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'C1orf216'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'C22orf15'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'C2CD4D'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'C4A'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'C5orf55'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'C6orf1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'C8orf82'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'C9orf114'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'C9orf16'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'C9orf173'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'C9orf50'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'C9orf69'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'C9orf78'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'CACNG6'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'CARD10'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'CARD9'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'CASP16'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'CCDC120'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'CCDC24'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'CCDC51'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'CCDC78'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'CD52'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'CDC20'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'CDC42SE1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'CDK4'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'CDT1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'CEMP1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'CHAF1A'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'CHCHD1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'CHMP2A'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'CIRBP'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'CISH'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'CLDN7'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'CLK3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'CNIH2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'CNTD2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'COA7'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'COL9A3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'COMMD6'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'CRIP1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'CRTC2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'CTB-96E2.2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'CTD-2521M24.10'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'CTNND1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'CTRC'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'CUEDC2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'CUL9'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'CYB561D1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'CYB561D2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'CYHR1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'DAZAP1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'DCAKD'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'DDB1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'DDX5'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'DEGS2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'DIEXF'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'DISP2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'DKC1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'DLG3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'DNAJB1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'DNAJC28'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'DNAJC5G'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'DNPEP'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'DPM2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'DPP9'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'DRAXIN'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'DSCR3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'DVL1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'DYNC1LI2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'E2F1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'EFCAB4B'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'EFNA3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'EFNA4'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'EHF'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'EIF3H'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'EMD'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'EML3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ENTPD8'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ERCC1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'EXOC3L1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'EXOC3L4'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'FAM129C'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'FAM175B'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'FAM189B'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'FAM213B'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'FAM83F'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'FAU'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'FBXL16'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'FES'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'FKBP10'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'FKBP2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'FKSG48'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'FLII'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'FOSL1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'FRMD1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'FTL'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'FTSJ1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'FTSJ2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'FUT6'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'G6PC2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'GABARAPL2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'GABPB2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'GEMIN2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'GFER'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'GJA4'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'GJC2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'GKAP1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'GNB5'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'GPATCH4'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'GPHA2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'GPR107'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'GSDMA'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'GSDMD'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'GSTM2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'HAGHL'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'HIST1H1D'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'HIST1H2AD'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'HIST1H2BF'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'HIST1H3D'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'HLCS'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'HNRNPH1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'HOGA1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'HOXB8'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'HOXC10'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'HOXD1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'HPN'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'HSD17B10'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'HSPB6'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ICAM1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ID1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'IDH1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'IDH3B'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'IER5L'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'IFI30'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'IFITM5'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'IGIP'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'IKZF4'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'INF2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'INMT-FAM188B'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'IZUMO1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'JAK1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'JPH2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'KBTBD4'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'KCNJ9'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'KIAA0895L'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'KIFAP3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'KIFC2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'KLHL21'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'KMT2E'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'KRI1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'LAMTOR1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'LEPREL4'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'LGI3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'LLGL2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'LMF1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'LMTK3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'LPAR2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'LRRC25'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'LRRC46'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'LRRC71'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'LSM10'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'LTBR'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'LUZP1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'LY6G6C'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'LY6G6D'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'LY6G6E'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'MAD2L2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'MAF1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'MAK16'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'MAPKAPK3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'MARCH9'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'MB'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'MED31'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'MEGF8'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'MESP2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'MGAT1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'MGAT4B'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'MIOX'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'MLLT11'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'MMP25'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'MOCS1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'MPDU1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'MPHOSPH9'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'MR1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'MROH1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'MRPL10'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'MRPL45'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'MRPL49'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'MSL1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'MT3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'MYADML2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'MYL4'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'N4BP1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'NANOS3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'NCR3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'NECAB3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'NEFH'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'NFIX'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'NHLRC4'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'NHSL1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'NISCH'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'NLGN2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'NOL3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'NOXO1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'NPB'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'NPRL2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'NR2C2AP'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'NRN1L'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'NTMT1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'NUBP2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'NUDT1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'NUPR1L'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'NXNL1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'NYAP1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ODF3L1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'OGFR'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'OPN1MW'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'OPN1MW2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'OSBPL2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'OSR1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'OTUB1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'OXTR'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PABPN1L'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PACSIN3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PAGR1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PAQR7'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PCDHGA1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PCSK4'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PHF13'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PI4K2A'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PIANP'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PIDD'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PIGQ'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PIKFYVE'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PIM3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PLCB3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PLCD3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PLEKHB2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PMPCA'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PNLDC1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'POU2F2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'POU5F1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PPIL2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PPP1R16A'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PPP4C'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PPP6R1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PRAP1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PRCC'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PRDX2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PRMT5'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PROC'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PRPF6'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PRSS33'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PRSS36'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PSMB3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PSMC6'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PSMD8'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PSRC1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PTCD2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PTRH1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'QPCTL'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'RAB17'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'RAB3A'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'RABEP2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'RBAK-RBAKDN'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'REC8'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'REEP4'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'REEP6'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'RGL2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'RHCG'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'RINT1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'RIPK1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'RNASE13'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'RNF175'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'RNF224'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ROM1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ROPN1B'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'RP1-130H16.18'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'RP1-4G17.5'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'RP11-293M10.1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'RP11-342M21.2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'RP11-345J4.3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'RP11-347C12.3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'RP11-51F16.8'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'RP11-834C11.12'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'RP13-512J5.1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'RPL10'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'RPL22'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'RPL24'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'RPL3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'RPLP2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'RUSC1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'S100A11'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SALL2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SAMD10'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SCAMP3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SCN4A'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SCNM1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SDCCAG3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SDR39U1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SEPT1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SERPINB10'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SERPINF2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SERTAD3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SIRT4'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SIRT7'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SIX1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SLC10A5'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SLC16A3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SLC22A18AS'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SLC23A3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SLC25A41'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SLC30A2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SLC30A3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SLC50A1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SLC5A2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SLC6A11'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SLC9A5'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SLC9A6'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SMCO1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SMTNL1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SNRPD1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SNRPD2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SOX8'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SP2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SPATA24'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SPSB3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SPTBN2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SQSTM1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SRA1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SRPK3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SRSF2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SSBP3-AS1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SSC5D'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SSRP1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SSSCA1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ST3GAL2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'STARD10'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'STMN1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'STMN3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SYNGAP1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SYNGR3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'TAB1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'TAOK2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'TBX10'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'TEX28'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'TGFB1I1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'TGFBR3L'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'TGM2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'THAP3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'TINF2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'TKTL1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'TLCD1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'TLX2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'TMA7'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'TMC4'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'TMEM145'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'TMEM206'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'TMEM52'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'TMEM63A'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'TMEM88'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'TMIE'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'TMPO'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'TNFAIP2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'TNFAIP8L2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'TNFRSF4'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'TPD52L2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'TRAF7'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'TRIM36'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'TRPM2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'TTC39A'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'TTC6'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'TTLL10'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'TUBB1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'TUBB3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'UBA7'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'UBXN11'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'UCHL3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'USP20'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'UTP23'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'VGF'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'VPS11'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'WDR24'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'WDR64'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'WDR74'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'WNT3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'WNT9B'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'XIRP1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'XXYLT1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'XXbac-BPG32J3.20'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'XXcos-LUCA11.5'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ZFHX3'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ZIC2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ZMYND19'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ZNF444'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ZNF513'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ZNF629'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ZNF821'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m     get_genomic_instances_pro_loop_single_gene(vocab, tissue, gene,anc_prom, loop_df, anc_foot,\n\u001b[0;32m----> 4\u001b[0;31m                                          verbose=True)\n\u001b[0m",
      "\u001b[0;32m<ipython-input-119-0650ef02da25>\u001b[0m in \u001b[0;36mget_genomic_instances_pro_loop_single_gene\u001b[0;34m(vocab, tissue, gene, anc_prom, loop_df, anc_foot, verbose, save_dir)\u001b[0m\n\u001b[1;32m     45\u001b[0m         \u001b[0manc_loop_arr_test\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mloop_loop_dict\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0manc_pro\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     46\u001b[0m         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0manc_pro\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0manc_loop_arr_test\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 47\u001b[0;31m         \u001b[0manc_foot1\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0manc_foot_sel\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0manc_foot_sel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0manc_loop_arr_test\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m&\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0manc_foot\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtf\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0manc_loop_tf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     48\u001b[0m         \u001b[0manc_foot1\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'type'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m=\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtype\u001b[0m \u001b[0;34m+\u001b[0m\u001b[0;34m'+loop'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     49\u001b[0m         \u001b[0manc_foot_loop\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconcat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0manc_foot_loop\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0manc_foot1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/pandas/core/ops/__init__.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(self, other)\u001b[0m\n\u001b[1;32m   1321\u001b[0m         \u001b[0mres_values\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mna_op\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0movalues\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1322\u001b[0m         \u001b[0munfilled\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_constructor\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mres_values\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mres_name\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1323\u001b[0;31m         \u001b[0mfilled\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfiller\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0munfilled\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1324\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mfinalizer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilled\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1325\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/pandas/core/ops/__init__.py\u001b[0m in \u001b[0;36m<lambda>\u001b[0;34m(x)\u001b[0m\n\u001b[1;32m   1285\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1286\u001b[0m     \u001b[0mfill_int\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfillna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1287\u001b[0;31m     \u001b[0mfill_bool\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfillna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbool\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1288\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1289\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mwrapper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mother\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/pandas/core/series.py\u001b[0m in \u001b[0;36mfillna\u001b[0;34m(self, value, method, axis, inplace, limit, downcast, **kwargs)\u001b[0m\n\u001b[1;32m   4343\u001b[0m             \u001b[0mlimit\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlimit\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4344\u001b[0m             \u001b[0mdowncast\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdowncast\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 4345\u001b[0;31m             \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   4346\u001b[0m         )\n\u001b[1;32m   4347\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36mfillna\u001b[0;34m(self, value, method, axis, inplace, limit, downcast)\u001b[0m\n\u001b[1;32m   6256\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6257\u001b[0m                 new_data = self._data.fillna(\n\u001b[0;32m-> 6258\u001b[0;31m                     \u001b[0mvalue\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mvalue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlimit\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlimit\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minplace\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0minplace\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdowncast\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdowncast\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   6259\u001b[0m                 )\n\u001b[1;32m   6260\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36mfillna\u001b[0;34m(self, **kwargs)\u001b[0m\n\u001b[1;32m    573\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    574\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mfillna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 575\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"fillna\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    576\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    577\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mdowncast\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36mapply\u001b[0;34m(self, f, axes, filter, do_integrity_check, consolidate, **kwargs)\u001b[0m\n\u001b[1;32m    436\u001b[0m                     \u001b[0mkwargs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreindex\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mb_items\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0malign_copy\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    437\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 438\u001b[0;31m             \u001b[0mapplied\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mb\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    439\u001b[0m             \u001b[0mresult_blocks\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_extend_blocks\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mapplied\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mresult_blocks\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    440\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/pandas/core/internals/blocks.py\u001b[0m in \u001b[0;36mfillna\u001b[0;34m(self, value, limit, inplace, downcast)\u001b[0m\n\u001b[1;32m    404\u001b[0m         \u001b[0minplace\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mvalidate_bool_kwarg\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minplace\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"inplace\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    405\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 406\u001b[0;31m         \u001b[0mmask\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0misna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    407\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mlimit\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    408\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mis_integer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlimit\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/pandas/core/dtypes/missing.py\u001b[0m in \u001b[0;36misna\u001b[0;34m(obj)\u001b[0m\n\u001b[1;32m    120\u001b[0m     \u001b[0mName\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mbool\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    121\u001b[0m     \"\"\"\n\u001b[0;32m--> 122\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0m_isna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    123\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    124\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/pandas/core/dtypes/missing.py\u001b[0m in \u001b[0;36m_isna_new\u001b[0;34m(obj)\u001b[0m\n\u001b[1;32m    145\u001b[0m         ),\n\u001b[1;32m    146\u001b[0m     ):\n\u001b[0;32m--> 147\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0m_isna_ndarraylike\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    148\u001b[0m     \u001b[0;32melif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mABCGeneric\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    149\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_constructor\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfunc\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0misna\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/pandas/core/dtypes/missing.py\u001b[0m in \u001b[0;36m_isna_ndarraylike\u001b[0;34m(obj)\u001b[0m\n\u001b[1;32m    247\u001b[0m             \u001b[0;31m# object array of non-strings\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    248\u001b[0m             \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mempty\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mbool\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 249\u001b[0;31m             \u001b[0mvec\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlibmissing\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misnaobj\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mravel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    250\u001b[0m             \u001b[0mresult\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m...\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mvec\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreshape\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    251\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# # gene = 'ABCA7'\n",
    "# for gene in ['AARS2', 'ABCA7',  'ACBD4', 'ACER2', 'ACLY', 'ACTG1', 'ACTN4', 'ADAP1', 'ADM', 'ADM5', 'ADRA1B', 'AHNAK', 'AHNAK2', 'AL109659.1', 'AL118506.1', 'AL136376.1', 'AL136531.1', 'AL590714.1', 'ALDH3B1', 'ANKRD9', 'ANKZF1', 'ANO9', 'AOC2', 'AP000349.1', 'AP3D1', 'APH1A', 'APOC1', 'ARAP1', 'ARHGEF12', 'ARHGEF4', 'ARRB1', 'ARSG', 'ASAP3', 'ASS1', 'ATAD3B', 'ATG9A', 'ATP11C', 'ATP2A2', 'ATP2C2', 'ATP5B', 'B4GALT3', 'B9D2', 'BIRC3', 'BMF', 'BTBD18', 'C12orf36', 'C16orf3', 'C17orf100', 'C17orf96', 'C19orf24', 'C19orf55', 'C1QTNF6', 'C1orf200', 'C1orf216', 'C22orf15', 'C2CD4D', 'C4A', 'C5orf55', 'C6orf1', 'C8orf82', 'C9orf114', 'C9orf16', 'C9orf173', 'C9orf50', 'C9orf69', 'C9orf78', 'CACNG6', 'CARD10', 'CARD9', 'CASP16', 'CCDC120', 'CCDC24', 'CCDC51', 'CCDC78', 'CD52', 'CDC20', 'CDC42SE1', 'CDK4', 'CDT1', 'CEMP1', 'CHAF1A', 'CHCHD1', 'CHMP2A', 'CIRBP', 'CISH', 'CLDN7', 'CLK3', 'CNIH2', 'CNTD2', 'COA7', 'COL9A3', 'COMMD6', 'CRIP1', 'CRTC2', 'CTB-96E2.2', 'CTD-2521M24.10', 'CTNND1', 'CTRC', 'CUEDC2', 'CUL9', 'CYB561D1', 'CYB561D2', 'CYHR1', 'DAZAP1', 'DCAKD', 'DDB1', 'DDX5', 'DEGS2', 'DIEXF', 'DISP2', 'DKC1', 'DLG3', 'DNAJB1', 'DNAJC28', 'DNAJC5G', 'DNPEP', 'DPM2', 'DPP9', 'DRAXIN', 'DSCR3', 'DVL1', 'DYNC1LI2', 'E2F1', 'EFCAB4B', 'EFNA3', 'EFNA4', 'EHF', 'EIF3H', 'EMD', 'EML3', 'ENTPD8', 'ERCC1', 'EXOC3L1', 'EXOC3L4', 'FAM129C', 'FAM175B', 'FAM189B', 'FAM213B', 'FAM83F', 'FAU', 'FBXL16', 'FES', 'FKBP10', 'FKBP2', 'FKSG48', 'FLII', 'FOSL1', 'FRMD1', 'FTL', 'FTSJ1', 'FTSJ2', 'FUT6', 'G6PC2', 'GABARAPL2', 'GABPB2', 'GEMIN2', 'GFER', 'GJA4', 'GJC2', 'GKAP1', 'GNB5', 'GPATCH4', 'GPHA2', 'GPR107', 'GSDMA', 'GSDMD', 'GSTM2', 'HAGHL', 'HIST1H1D', 'HIST1H2AD', 'HIST1H2BF', 'HIST1H3D', 'HLCS', 'HNRNPH1', 'HOGA1', 'HOXB8', 'HOXC10', 'HOXD1', 'HPN', 'HSD17B10', 'HSPB6', 'ICAM1', 'ID1', 'IDH1', 'IDH3B', 'IER5L', 'IFI30', 'IFITM5', 'IGIP', 'IKZF4', 'INF2', 'INMT-FAM188B', 'IZUMO1', 'JAK1', 'JPH2', 'KBTBD4', 'KCNJ9', 'KIAA0895L', 'KIFAP3', 'KIFC2', 'KLHL21', 'KMT2E', 'KRI1', 'LAMTOR1', 'LEPREL4', 'LGI3', 'LLGL2', 'LMF1', 'LMTK3', 'LPAR2', 'LRRC25', 'LRRC46', 'LRRC71', 'LSM10', 'LTBR', 'LUZP1', 'LY6G6C', 'LY6G6D', 'LY6G6E', 'MAD2L2', 'MAF1', 'MAK16', 'MAPKAPK3', 'MARCH9', 'MB', 'MED31', 'MEGF8', 'MESP2', 'MGAT1', 'MGAT4B', 'MIOX', 'MLLT11', 'MMP25', 'MOCS1', 'MPDU1', 'MPHOSPH9', 'MR1', 'MROH1', 'MRPL10', 'MRPL45', 'MRPL49', 'MSL1', 'MT3', 'MYADML2', 'MYL4', 'N4BP1', 'NANOS3', 'NCR3', 'NECAB3', 'NEFH', 'NFIX', 'NHLRC4', 'NHSL1', 'NISCH', 'NLGN2', 'NOL3', 'NOXO1', 'NPB', 'NPRL2', 'NR2C2AP', 'NRN1L', 'NTMT1', 'NUBP2', 'NUDT1', 'NUPR1L', 'NXNL1', 'NYAP1', 'ODF3L1', 'OGFR', 'OPN1MW', 'OPN1MW2', 'OSBPL2', 'OSR1', 'OTUB1', 'OXTR', 'PABPN1L', 'PACSIN3', 'PAGR1', 'PAQR7', 'PCDHGA1', 'PCSK4', 'PHF13', 'PI4K2A', 'PIANP', 'PIDD', 'PIGQ', 'PIKFYVE', 'PIM3', 'PLCB3', 'PLCD3', 'PLEKHB2', 'PMPCA', 'PNLDC1', 'POU2F2', 'POU5F1', 'PPIL2', 'PPP1R16A', 'PPP4C', 'PPP6R1', 'PRAP1', 'PRCC', 'PRDX2', 'PRMT5', 'PROC', 'PRPF6', 'PRSS33', 'PRSS36', 'PSMB3', 'PSMC6', 'PSMD8', 'PSRC1', 'PTCD2', 'PTRH1', 'QPCTL', 'RAB17', 'RAB3A', 'RABEP2', 'RBAK-RBAKDN', 'REC8', 'REEP4', 'REEP6', 'RGL2', 'RHCG', 'RINT1', 'RIPK1', 'RNASE13', 'RNF175', 'RNF224', 'ROM1', 'ROPN1B', 'RP1-130H16.18', 'RP1-4G17.5', 'RP11-293M10.1', 'RP11-342M21.2', 'RP11-345J4.3', 'RP11-347C12.3', 'RP11-51F16.8', 'RP11-834C11.12', 'RP13-512J5.1', 'RPL10', 'RPL22', 'RPL24', 'RPL3', 'RPLP2', 'RUSC1', 'S100A11', 'SALL2', 'SAMD10', 'SCAMP3', 'SCN4A', 'SCNM1', 'SDCCAG3', 'SDR39U1', 'SEPT1', 'SERPINB10', 'SERPINF2', 'SERTAD3', 'SIRT4', 'SIRT7', 'SIX1', 'SLC10A5', 'SLC16A3', 'SLC22A18AS', 'SLC23A3', 'SLC25A41', 'SLC30A2', 'SLC30A3', 'SLC50A1', 'SLC5A2', 'SLC6A11', 'SLC9A5', 'SLC9A6', 'SMCO1', 'SMTNL1', 'SNRPD1', 'SNRPD2', 'SOX8', 'SP2', 'SPATA24', 'SPSB3', 'SPTBN2', 'SQSTM1', 'SRA1', 'SRPK3', 'SRSF2', 'SSBP3-AS1', 'SSC5D', 'SSRP1', 'SSSCA1', 'ST3GAL2', 'STARD10', 'STMN1', 'STMN3', 'SYNGAP1', 'SYNGR3', 'TAB1', 'TAOK2', 'TBX10', 'TEX28', 'TGFB1I1', 'TGFBR3L', 'TGM2', 'THAP3', 'TINF2', 'TKTL1', 'TLCD1', 'TLX2', 'TMA7', 'TMC4', 'TMEM145', 'TMEM206', 'TMEM52', 'TMEM63A', 'TMEM88', 'TMIE', 'TMPO', 'TNFAIP2', 'TNFAIP8L2', 'TNFRSF4', 'TPD52L2', 'TRAF7', 'TRIM36', 'TRPM2', 'TTC39A', 'TTC6', 'TTLL10', 'TUBB1', 'TUBB3', 'UBA7', 'UBXN11', 'UCHL3', 'USP20', 'UTP23', 'VGF', 'VPS11', 'WDR24', 'WDR64', 'WDR74', 'WNT3', 'WNT9B', 'XIRP1', 'XXYLT1', 'XXbac-BPG32J3.20', 'XXcos-LUCA11.5', 'ZFHX3', 'ZIC2', 'ZMYND19', 'ZNF444', 'ZNF513', 'ZNF629', 'ZNF821']:\n",
    "#     get_genomic_instances_pro_loop_single_gene(vocab, tissue, gene,anc_prom, loop_df, anc_foot,\n",
    "#                                          verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
