{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# crms_irf.ipynb\n",
    "\n",
    "trying iterative random forests on random forest generated from both \n",
    "1. try irf on housekeeping gene subset of model for each cell type\n",
    "\n",
    "\n",
    "based on https://github.com/Yu-Group/iterative-Random-Forest/blob/master/demo/28_iRF_demo_sklearn.ipynb\n",
    "\n",
    "and the Iteratve Random Forest (Basu paper)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# basic packages\n",
    "import os, glob\n",
    "import pandas as pd\n",
    "import numpy as np; np.random.seed(0)\n",
    "import itertools\n",
    "from collections import Counter, defaultdict\n",
    "import time\n",
    "\n",
    "# machine learning packages from sklearn\n",
    "from sklearn.preprocessing import MinMaxScaler #StandardScaler \n",
    "from sklearn import preprocessing, metrics\n",
    "from sklearn.feature_selection import VarianceThreshold\n",
    "from sklearn.model_selection import train_test_split, KFold, cross_validate, cross_val_score, StratifiedKFold\n",
    "from sklearn.linear_model import LogisticRegression, Lasso, LassoCV\n",
    "from sklearn.naive_bayes import GaussianNB\n",
    "from sklearn.svm import LinearSVC\n",
    "from sklearn.ensemble import RandomForestClassifier, RandomForestRegressor, VotingClassifier, AdaBoostClassifier\n",
    "from sklearn.neighbors import KNeighborsClassifier\n",
    "from sklearn.neural_network import MLPClassifier\n",
    "from sklearn.metrics import roc_auc_score, auc, roc_curve, plot_roc_curve, confusion_matrix, accuracy_score\n",
    "from scipy import interp\n",
    "\n",
    "# for IRF\n",
    "from functools import reduce\n",
    "# Needed for the scikit-learn wrapper function\n",
    "import irf\n",
    "from irf import (irf_utils, utils,\n",
    "                 irf_jupyter_utils)\n",
    "from irf.ensemble.wrf import RandomForestClassifierWithWeights\n",
    "from math import ceil\n",
    "\n",
    "# Import our custom utilities\n",
    "from imp import reload\n",
    "\n",
    "\n",
    "# Import tools needed for visualization\n",
    "import seaborn as sns; sns.set()\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.tree import export_graphviz\n",
    "import pydot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TF annotations\n",
    "hoco_annon_file = 'annon/HOCOMOCOv11_core_annotation_HUMAN_mono.tsv'\n",
    "hoco_annon = pd.read_table(hoco_annon_file,header=0)\n",
    "hoco_annon[['motif_abbr','motif_info']]=hoco_annon['Model'].str.split(\"_\",expand=True)\n",
    "hoco_motif_to_tf = pd.Series(hoco_annon['Transcription factor'].values, index=hoco_annon.motif_abbr.values).to_dict()\n",
    "tfs = list(hoco_annon['Transcription factor'])\n",
    "\n",
    "\n",
    "# RNA\n",
    "THRES=1\n",
    "rna_tpm_file = 'rnaseq/tissue_tpm_sym.csv'\n",
    "rna_df = pd.read_csv(rna_tpm_file, index_col=0,header=0)\n",
    "rna_df.index = [x.upper() for x in rna_df.index.values]\n",
    "all_genes = rna_df.index\n",
    "tissue_diffloop_to_rna = {\n",
    "    'Air':'Airway', 'Astro':'Astrocyte', 'Blad':'Bladder', 'Colon':'Colon', 'Eso':'Esophageal', \n",
    "    'GDSD0':'GDSD0', 'GDSD3':'GDSD3', 'GDSD6':'GDSD6', \n",
    "    'GM12878':'GM12878', 'HMEC':'HMEC', 'MC':'Melanocytes', 'Ova':'Ovarian',\n",
    "    'Panc':'Pancreas', 'Pros':'Prostate', 'Renal':'Renal', 'Thy':'Thyroid', 'Uter':'Uterine'\n",
    "}\n",
    "\n",
    "def get_exp(rna_df, sym, tissue,verbose=True):\n",
    "    try:\n",
    "        return rna_df.loc[sym,tissue]\n",
    "    except KeyError:\n",
    "        if verbose:\n",
    "            print(sym, 'not found')\n",
    "        return 0\n",
    "   \n",
    "# loop and anchor and crms\n",
    "loop_files = sorted(glob.glob('diffloop_data/bedpe_files_csvs_rna/*merged.loops.csv'))\n",
    "anchor_files = sorted(glob.glob('diffloop_data/bedpe_files_csvs_rna/*merged.anchors.csv'))\n",
    "crm_files = sorted(glob.glob('networks/diffgene/all_no_thres/*_crm_tissue.csv'))\n",
    "\n",
    "\n",
    "tissues = []\n",
    "for crm in crm_files:\n",
    "    tissues.append(os.path.basename(crm).split('_')[1])\n",
    "tissue_dict = dict(zip(tissues,range(len(tissues))))\n",
    "\n",
    "# print(loop_files)\n",
    "# print(anchor_files)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 0. Data Wrangling\n",
    "- import\n",
    "- preprocess\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((291006, 824), (291006, 1), (291006, 1))"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# import\n",
    "data_all = pd.read_csv('networks/representations_no_tfs/data_all.csv',index_col=0,header=0)\n",
    "labels_all = np.array(pd.read_csv('networks/representations_no_tfs/labels_all.csv',index_col=0,header=None))\n",
    "genes_all = np.array(pd.read_csv('networks/representations_no_tfs/genes_all.csv',index_col=0,header=None))\n",
    "\n",
    "\n",
    "data_all.shape, labels_all.shape,genes_all.shape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # labels for problem 1: `labels_all_bin`\n",
    "# labels_all_bin = np.array([1 if x > THRES else 0 for x in labels_all])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Counter({'Air': 17118, 'Astro': 17118, 'Blad': 17118, 'Colon': 17118, 'Eso': 17118, 'GDSD0': 17118, 'GDSD3': 17118, 'GDSD6': 17118, 'GM12878': 17118, 'HMEC': 17118, 'MC': 17118, 'Ova': 17118, 'Panc': 17118, 'Pros': 17118, 'Renal': 17118, 'Thy': 17118, 'Uter': 17118})\n"
     ]
    }
   ],
   "source": [
    "# labels for problem 2: `tissue_labels`\n",
    "gene_to_num_dict = dict(zip(np.unique(genes_all),range(len(np.unique(genes_all)))))\n",
    "genes_num_all = np.vectorize(gene_to_num_dict.get)(genes_all)\n",
    "\n",
    "\n",
    "tissue_to_num_dict = dict(zip(tissue_dict.keys(), range(len(tissue_dict.keys()))))\n",
    "tissue_df = data_all[tissue_dict.keys()].reset_index().melt('index')\n",
    "tissue_df.columns = ['index','tissue','value']\n",
    "tissue_df = tissue_df[tissue_df.value==1].reset_index(drop=True)\n",
    "tissue_df['tissue_num'] = tissue_df.tissue.map(tissue_to_num_dict)\n",
    "\n",
    "tissue_labels = np.array(tissue_df.tissue)\n",
    "tissue_num_labels = np.array(tissue_df.tissue_num)\n",
    "\n",
    "print(Counter(tissue_labels))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pd.DataFrame([tissue_diffloop_to_rna[x] for x in tissue_labels]).to_csv(os.path.join(save_dir, 'tissues_all.csv'), header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# labels_all = np.array(pd.read_csv('networks/representations_irf_cca/tissues_all.csv',index_col=0,header=None))\n",
    "# tissue_bool_idx = np.where(labels_all=='Airway')[0]\n",
    "# len(tissue_bool_idx) #17118"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set up data collection, remove the tissues and uninformative columns\n",
    "\n",
    "data_all_no_tissue = data_all\n",
    "    \n",
    "if '_loop' in data_all_no_tissue.columns:\n",
    "    data_all_no_tissue.drop('_loop',axis=1,inplace=True)\n",
    "\n",
    "for tissue in tissue_dict.keys():\n",
    "    if tissue in data_all_no_tissue.columns:\n",
    "        data_all_no_tissue.drop(tissue,axis=1,inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(291006, 802) (291006, 648) 648\n",
      "(291006, 648) (291006, 648)\n"
     ]
    }
   ],
   "source": [
    "# data for problem 1: `data_all_norm_`\n",
    "data_all_1 = data_all_no_tissue#.copy()\n",
    "\n",
    "# # make sure expression is removed\n",
    "# if 'log_exp' in data_all_1.columns:\n",
    "#     data_all_1.drop('log_exp',axis=1,inplace=True)\n",
    "    \n",
    "if 'num_atac_regions' in data_all_1.columns:\n",
    "    data_all_1.drop('num_atac_regions',axis=1,inplace=True)\n",
    "\n",
    "if 'num_snp' in data_all_1.columns:\n",
    "    data_all_1.drop('num_snp',axis=1,inplace=True)\n",
    "\n",
    "    \n",
    "if 'num_loop_counts' in data_all_1.columns:\n",
    "    data_all_1.drop('num_loop_counts',axis=1,inplace=True)\n",
    "\n",
    "if 'num_loops' in data_all_1.columns:\n",
    "    data_all_1.drop('num_loops',axis=1,inplace=True)\n",
    "\n",
    "    \n",
    "# remove zero variance columns and scale rest\n",
    "selector = VarianceThreshold()\n",
    "data_all_varfilt_1 = selector.fit_transform(data_all_1)\n",
    "data_all_varfilt_cols_1 = data_all_1.columns[selector.get_support()]\n",
    "print(data_all_1.shape, data_all_varfilt_1.shape, len(data_all_varfilt_cols_1))\n",
    "scaler = MinMaxScaler()\n",
    "data_all_norm_1 = scaler.fit_transform(data_all_varfilt_1)\n",
    "print(data_all_varfilt_1.shape, data_all_norm_1.shape)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # data for problem 2: `data_all_norm_2`\n",
    "# data_all_2 = data_all_no_tissue.copy()\n",
    "\n",
    "# # keep gene expression value and the id of the gene itself\n",
    "# data_all_2['log_exp'] = labels_all\n",
    "# data_all_2['gene_num'] = genes_num_all\n",
    "\n",
    "# selector = VarianceThreshold()\n",
    "# data_all_varfilt_2 = selector.fit_transform(data_all_2)\n",
    "# data_all_varfilt_cols_2 = data_all_2.columns[selector.get_support()]\n",
    "# print(data_all_2.shape, data_all_varfilt_2.shape, len(data_all_varfilt_cols_2))\n",
    "# scaler = MinMaxScaler()\n",
    "# data_all_norm_2 = scaler.fit_transform(data_all_varfilt_2)\n",
    "# print(data_all_varfilt_2.shape, data_all_norm_2.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_dir = 'networks/representations_irf_cell_type'\n",
    "if not os.path.exists(save_dir):\n",
    "    os.makedirs(save_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pd.DataFrame(data_all_norm_1, columns=data_all_varfilt_cols_1).to_csv(os.path.join(save_dir, 'data.csv'))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# A. Problem 1: try irf on housekeeping gene subset of model for each cell type\n",
    "\n",
    "try GDSD3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## A.1 split data and run RF\n",
    "\n",
    "based on previous analysis, ideally running irf on 20% of the data will save time and energy\n",
    "\n",
    "so we will only be trying housekeeping genes (n=398)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7091\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "107780"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_expr_genes_tpm1 = list(pd.read_csv('rnaseq/all_expr_genes_tpm1.csv')['x'])\n",
    "print(len(all_expr_genes_tpm1))\n",
    "\n",
    "gene_in_highexpr_bool = np.isin(genes_all, all_expr_genes_tpm1)\n",
    "gene_in_highexpr_bool_idx = np.where(gene_in_highexpr_bool)[0]\n",
    "len(gene_in_highexpr_bool_idx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "398"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get housekeeping genes\n",
    "housekeeping_df = pd.read_table(os.path.join('annon', 'HSIAO_HOUSEKEEPING_GENES.txt'),skiprows=2, header=None)\n",
    "housekeeping_df.columns = ['genes']\n",
    "housekeeping_genes_list = list(housekeeping_df.genes)\n",
    "len(housekeeping_genes_list)\n",
    "\n",
    "# tss_df_tissue = pd.read_csv('atac_footprinting/tss_footprints/GDSD3_tss_hoco_tf_exp_1000.csv', index_col=0)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "for reference use values (rna spelling)\n",
    "tissue_diffloop_to_rna = {\n",
    "    'Air':'Airway', 'Astro':'Astrocyte', 'Blad':'Bladder', 'Colon':'Colon', 'Eso':'Esophageal', \n",
    "    'GDSD0':'GDSD0', 'GDSD3':'GDSD3', 'GDSD6':'GDSD6', \n",
    "    'GM12878':'GM12878', 'HMEC':'HMEC', 'MC':'Melanocytes', 'Ova':'Ovarian',\n",
    "    'Panc':'Pancreas', 'Pros':'Prostate', 'Renal':'Renal', 'Thy':'Thyroid', 'Uter':'Uterine'\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [],
   "source": [
    "rf_tissue = 'GM12878'\n",
    "gene_tissue = 'GM12878'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "506\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "4607"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_type_genes_tpm1 = pd.read_csv('rnaseq/unique_gene_lists/'+gene_tissue+'_genes.txt',header=None).loc[:,0]\n",
    "print(len(cell_type_genes_tpm1))\n",
    "\n",
    "cell_type_genes_tpm1_bool = np.isin(genes_all, cell_type_genes_tpm1)\n",
    "cell_type_genes_tpm1_bool_idx = np.where(cell_type_genes_tpm1_bool)[0]\n",
    "len(cell_type_genes_tpm1_bool_idx)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6205"
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# takes a long time\n",
    "gene_in_hk_bool = np.isin(genes_all, housekeeping_genes_list)\n",
    "gene_in_hk_bool_idx = np.where(gene_in_hk_bool)[0]\n",
    "len(gene_in_hk_bool_idx)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(17118,)"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# loop through all data\n",
    "tissue_bool_labels = (tissue_labels==rf_tissue)\n",
    "tissue_bool_idx = np.where(tissue_bool_labels)[0]\n",
    "tissue_bool_idx.shape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4607 291006 0.015831288701951163\n"
     ]
    }
   ],
   "source": [
    "select_idx = cell_type_genes_tpm1_bool_idx#list(set(tissue_bool_idx).intersection(gene_in_hk_bool_idx))\n",
    "print(len(select_idx), data_all_norm_1.shape[0], len(select_idx)/data_all_norm_1.shape[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.05 14550 291006\n"
     ]
    }
   ],
   "source": [
    "#### TESTING #####\n",
    "perc = .05\n",
    "num_sample = int(data_all_norm_1.shape[0]*perc)\n",
    "random_idx = np.random.choice(range(data_all_norm_1.shape[0]),num_sample,replace=False)\n",
    "print(perc, random_idx.shape[0], data_all_norm_1.shape[0])\n",
    "# num_sample = int(select_idx.shape[0]*perc)\n",
    "# random_idx = np.random.choice(select_idx,num_sample,replace=False)\n",
    "# print(perc, random_idx.shape[0], select_idx.shape[0])\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({False: 4336, True: 271})"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Counter(tissue_bool_labels[select_idx])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Training Features Shape: (3455, 648)\n",
      "Training Labels Shape: (3455,)\n",
      "Testing Features Shape: (1152, 648)\n",
      "Testing Labels Shape: (1152,)\n"
     ]
    }
   ],
   "source": [
    "# Split the data into training and testing sets\n",
    "# train_features, test_features, train_labels, test_labels = train_test_split(data_all_norm_1, labels_all_bin,\n",
    "#                                                                             test_size = 0.25, random_state = 42)\n",
    "\n",
    "# train_features, test_features, train_labels, test_labels = train_test_split(data_all_norm_1[random_idx,:], \n",
    "#                                                                             tissue_bool_labels[random_idx], \n",
    "#                                                                             test_size = 0.25, random_state = 42)\n",
    "\n",
    "train_features, test_features, train_labels, test_labels = train_test_split(data_all_norm_1[select_idx,:], \n",
    "                                                                            tissue_bool_labels[select_idx], \n",
    "                                                                            test_size = 0.25, random_state = 42)\n",
    "\n",
    "\n",
    "\n",
    "print('Training Features Shape:', train_features.shape)\n",
    "print('Training Labels Shape:', train_labels.shape)\n",
    "print('Testing Features Shape:', test_features.shape)\n",
    "print('Testing Labels Shape:', test_labels.shape)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 3.08 s, sys: 116 ms, total: 3.2 s\n",
      "Wall time: 580 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Create the model with 100 trees\n",
    "model = RandomForestClassifier(n_estimators=100,#50, \n",
    "                               max_features = 'sqrt',\n",
    "                               n_jobs=-1, verbose = 0)\n",
    "# Fit on training data\n",
    "model.fit(train_features, train_labels)\n",
    "# Actual class predictions\n",
    "rf_predictions = model.predict(test_features)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1152, 2)"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.predict_proba(test_features).shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.9765625\n",
      "[[1082    0]\n",
      " [  27   43]]\n",
      "0.9036836546078691\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<sklearn.metrics._plot.roc_curve.RocCurveDisplay at 0x7fc6b9393a10>"
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYQAAAEJCAYAAACUk1DVAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAgAElEQVR4nO3deXxU5dn/8c9MFrIRCGEgJIiIwOXCpiKiSOvCoz+11vqI1SIqqOBG1VqtC5UiuGsR3KiKW/WhFrWtC6DUYmtxB8GdW1xQ2SQbQkIgy+T3x5mESQjJBDMzJPm+Xy9f5ixzznVnyLnOuc+9+KqrqxEREfHHOwAREdkzKCGIiAighCAiIiFKCCIiAighiIhISGK8A9hNHYBDgfVAVZxjERFpLRKAHsB7wPb6G1trQjgU+G+8gxARaaVGAkvqr2ytCWE9QHFxKcFg8/tRZGdnUFhY0uJB7clU5vZBZW4fdrfMfr+PrKx0CF1D62utCaEKIBis3q2EUPPZ9kZlbh9U5vbhR5a5wap2vVQWERFACUFEREKUEEREBIjBOwQzywTeBH7mnFtdb9sQYA6QCbwOXOScq4x2TCIisrOoPiGY2WF4TZv672KXp4BJzrn+gA+YEM14RERk16JdZTQBuBRYV3+Dme0NpDrn3g6tehw4PcrxiIi0iOrqaoJx+i9aolpl5Jy7AMDMGtqcS922sOuBntGMR0SkpTwy/zPe/HhDzM/r88E1Zx9K/9yOLX7sePZD8APhqc4HBJtzgOzsjN0+eSDQ8r/MPZ3K3D6ozLFRsHkbPbqmc/Qhe8X0vH4/7Nc7i+xOqS1+7HgmhDV4Y2rUyKGBqqXGFBaW7FbnjECgI/n5W5r9udZMZW4f2kOZg8Fq8n8oY3u517eqS5d0iopKYx5H6dYKsjM7MOqg3JifO7tT6m59z36/r9Eb6bglBOfcN2a2zcxGOOfeAM4GFsYrHhHZs1RXV7OppJy1BSWs2VjK2vwS1uSXsq6wlIrKZlUmRE33LmnxDqFFxTwhmNkCYIpzbilwFvBwqGnq+8A9sY5HROKvbHslawtKWZNfwtqN3v/X5JdQum1HK/RO6cn0DKRz9EF55AXSSeuQ5K3vlMIPP2yLS9x9cjPjct5oiUlCcM71Dvv5xLCfPwCGxSIGEYm/yqogG4q2ehf+/FLWbPTu+gs377igd0hOoGfXdA6xbvQMpNMzkEFeIJ2OackNHrM9VJPFSmsd3E5E9mDV1dUUbt7GmvwdVT1r8kvYULiVqtB7vwS/j5wuaeybl8lPh+TWXvizO6Xg9/niXIL2SQlBRH6UkrKKOhf9mrv/beU7BtTMzuxAXiCDwft2JS90198jO43EBI2esydRQhCRiFRUVrGuYGudi/6a/BI2lZTX7pOekkheIIPDB+TQM5BBz0A6eV0zSEvRpaY10LckInUEg9Xkbyqrc9Ffk1/K98Vbqekkm5jgJzc7jf337kLPbumhi38GnTOS8am6p9VSQhBpp6qrq9lcWr5TPf+6glLKQ806fUAgK5W8rukcul83enbz7vq7ZaWS4Fd1T1ujhCDSDmwr95p17mjZ4yWAkrKK2n0y05LIC2Tw0yF5XuuebhnkZqfTITkhjpFLLCkhiLQhlVXB0IU/dNEPtekvCGunn5zkJ69rBgf167qjnj+QQWZ6w806pf1QQhBphaqrqynesr32Tr/m4r+haCuVVV51j9/no3uXVPbpkcnIQT1qm3V27ZyqZp3SICUEkT3c1m0VYU06d9T3l23f0Ys3q2MHegYyOGxADlnpSbXNOpMSVd0jkVNCENlDVFQGWV8YfuH3fi7esr12n9QOifQMpHPYAd3r9OJNT/GGcVCvXfkxlBBEYixYXU3BprKdWvd8X1RWO/lJgt9Hj+x0rFfn2nr+noEMsjp2ULNOiRolBJEo2lxaXq8XbynrCkrZXrGjF2/XTin0DGRwiAVCd/wZdM9KVS9eiTklBJEWsL28inWFOwZr8zp1lbB5645mnRmpSfQMpHsveLt5VT252emkdtCfoewZ9C9RpBmqgkE2FnvVPTXt+dfml5K/qax2+r/kRD+5XdMZtG/X2iadPQPpZKarF6/s2ZQQRBpQMznLmrD2/GvzS1hXuKNZp88H3bPS6NW9Zuwer54/0DkVv18Xfml9lBCk3du6rdKblau2qse7+IdPztI5I5m8QAbH9s6qHbenR3YayUlq1ilthxKCtBuVVUHWF9YdrXNtfgmFm3c060xJTiAvkM7Q/brV6cWbkZoUx8hFYkMJQdqcYHU1hT9s46vvS/jkywLWhi7+G4rqTc6SnUbfnp05KqyePzszRfX80m4pIUirtmVr3dE61+aXsKaglO11JmdJoWcgnSH9QpOzdM0gR5OziOxECUFahe0VVV4v3o07mnSuyS/lh9K6k7P0DGRw5IAe5AXSGdCvG+lJPjXrFImQ/lJkjxIMVrNxU1mdJp1r8kvYuKmsdnKWpEQ/udnpHLhPlzr1/PUnZ9EwDiLNo4QgcVFdXc0PpeV1mnSuyS9lXWEpFfUmZ+kZyAiN3eN15uqelaZmnSJRoIQgUVe23ZucZU1+CWtrqnwK6k3Okp5Mz0A6Rx+UVzsJe27XdDqoWadIzCghSIuprAqyoWhrWJPOnSdn6ZDkNes8uH/XUMse764/M02Ts4jEmxKCNFt1dTWFm7fVtuqpufCvL9zRrNPv85p19snNZOTgXHp2TSevWwZdO6VochaRPZQSQitRURlkxRcF/PeDdXy57ofaF6yR8vl8VDf3Q7tQFayurecH6JLpTc4ysE927R1/j+x0khLVrFOkNVFC2MOtyS/hvx+s561PNlBSVkFWxw4MPyCn2Rfb1NRkysrKm94xAj4fdOucWtuZKy1FvXhF2gIlhD1Q2fZK3vnse/77wXq+Xr+ZBL+Pg/p1ZeTgXA7s3WW3WtioCaaINEUJYQ8065kP+HzND+QF0jnz2H4cfmB3Ouqlq4hEmRLCHmhd4VZGDMjhvJP217g6IhIzSggxsGVrOV+u2xzRvlVVQUrKKgh0TlUyEJGYimpCMLMxwO+BJGCmc+7+etsPBh4EkoHvgLHOuU3RjCnWvtmwhZnPfsAPJc17odstKzVKEYmINCxqCcHM8oCbgUOA7cCbZvaac+7TsN1mAVOccwvN7I/AVXgJpE348MtCZj//MekpiVx22iA6ZUT2HiAxwU9eID3K0YmI1BXNJ4RRwGLnXBGAmT0LjAamhe2TAGSGfk4DiqIYT0z9Z8Vannzlc3oG0rn89MFkdewQ75BERBoVzYSQC6wPW14PDKu3z5XAIjObCZQChzXnBNnZGbsdXCDQcbc/25jq6mqeXPgZz/xrFQfv141rzh66x7TTj1aZ92Qqc/ugMreMaCYEPxDeNdYH1HZvNbNU4BFglHPuXTO7EvgzcFKkJygsLCEYbH7v22i1ya+oDPLYgs94+9Pv+cngXM4+vj+lW7ZRumVb0x+OsvbYD0Flbh9U5sj5/b5Gb6SjmRDWACPDlnOAdWHLA4Ay59y7oeUHgelRjCeq1hWU8vCLn/LN91s47ad9OHH43molJCKtSjQHm3kVONbMAmaWBpwGvBy2/QtgLzOz0PIpwHtRjCdqCn/Yxl1PL2ddYSkTf34AJx3eW8lARFqdqCUE59xaYDLwGrACmBuqGlpgZkOdc8XAOGCemX0InAeMj1Y80VJSVsGMeSvYXhHkhnOGMvyAnHiHJCKyW6LaD8E5NxeYW2/diWE/LwQWRjOGlrJ05UZeeGM11dR9Z1GytYLSbRX89owh9Oy2+y+5RUTiTT2VI7C5tJzHFq7E74P9emXV3dgFfjI4F6u/XkSklVFCiMCz//6S8ooqpp0/jB7Z6jAmIm1Tu0wI985bwaJ3vmnWZ044rJeSgYi0ae0yIXz3/Ra6dkrhiAGRvQBOT0nip0NyoxyViEh8tcuEsLm0nG5ZqfxiZJ94hyIisseIKCGYWU9gEPAKkOec+zaqUUVRRWWQtfklJCdmNr2ziEg70mQ/BDM7CXgTuB/oBnxqZqdEO7BoqazyRs/om9cpzpGIiOxZIumYNgVv0LlNzrn1wJHUHbG0VakKjX2k0UdFROqKJCEkhBIBAM65FUDzR5TbQ7hviwF26mAmItLeRZIQtppZL0JJwMxGAvEfvnM31QyOOrBPdnwDERHZw0TyUvlaYBHQw8zeAvrhDVTXqmnoORGRuppMCM65N81sOHA43gxnbzvnCqIemYiIxFSTCcHMFjrnTiBsEDoze9s5NzyqkYmISEztMiGE5kDuD+wbGp66RhKwPdqBiYhIbDX2hHAV0Bt4GPh12PpK4NMoxiQiInGwy4TgnFsNrDYzc84Fw7eZmUZ5ExFpYyJpZXSymU0DMvAa5yQAXYCO0QxMRERiK5J+CHcBtwDfApfgzYv8p2gGJSIisRdJQih1zv0VeBuvQ9rFwM+iGpWIiMRcJAlhm5l1AL4AhoTeJ2jcBxGRNiaSdwgvAPOBc4G3QkNXqGOaiEgb0+QTgnPuFuA859xa4BfA67SBoStERKSuRhOCmfU3sx41E+I4594HngFmxiI4ERGJnV0mBDO7GngfWGVmPwmtuwL4DOgRm/BERCRWGnuHcCGwP7AXcJWZXQwcBVzsnJsbg9hERCSGGksIpc6574DvQi+S3wL2d85tik1oIiISS40lhKqwnzcDZzjnyqIcj4iIxEkk/RAAflAyEBFp2xp7QuhmZlc28DMAzrkZ0QtLRERirbGE8E9gYAM/g3oqi4i0OY0Nfz3+xx7czMYAv8ebVGemc+7+etsNeBDIAjYAZzrnin/seUVEpPkifYfQbGaWB9wMHAkMASaa2QFh2314w2Lc5pwbDCwHro1WPCIi0rioJQRgFLDYOVfknCsFngVGh20/GK9p68uh5VuA+xERkbiIZHC73ZULrA9bXg8MC1vuC2wws0eAg/B6QIdP1SkiIjEUUUIws2F4F+3HgEOcc29F8DE/dV8++4DwqTgT8Xo+/8Q5t9TMpgMzgHGRxASQnZ0R6a61MtdtBiCrSzqBQPua9K29lRdU5vZCZW4ZTSYEMxsHXA2kAH8Hnjezyc65h5v46BpgZNhyDrAubHkDsMo5tzS0/Be8aqWIFRaWEAw2r8HT5s3bACguKiUtwdesz7ZmgUBH8vO3xDuMmFKZ2weVOXJ+v6/RG+lI3iFcBhwObHbObQQOAa6I4HOvAseaWcDM0vCGzH45bPubQMDMBoeWTwaWRXBcERGJgkgSQpVzbnPNQmh8o8qmPhSaP2Ey8BqwApjrnHvXzBaY2dBQz+dTgYfN7BPgGOC3u1MIERH58SJ5h1BkZkMIvQ8ws7OAokgOHhoVdW69dSeG/fwOdV80i4hInESSEK7AmxRnXzNbD5QBp0Q1KhERiblIEsJKYDDQH0gAnHOuIqpRiYhIzEWSEL4DHgEedc59E+V4REQkTiJ5qXws0AFYYmavmNloM4tmhzYREYmDJhOC81wL7A3MAq4C1kY7MBERia1Ieyp3A8YC5+L1OL4pmkGJiEjsRdJT+QVgBPA3YGKoqaiIiLQxkTwhvAiMcc6VRDsYERGJn10mBDMb65x7CsjEm8ugznZNoSki0rY09oTQL/T/AQ1s0xSaIiJtTGNTaP4h9OM/nHPPh28zs7OjGpWIiMRcY1VGJ+PNhXynmfnxWhcRWncj8GT0wxMRkVhprMpoCN4IpN3whsCuUQncHc2gREQk9hqrMpoOTDezS5xzD8QwJhERiYNIWhmlmtmV9berlZGISNuyu62MRESkjWmylZFzbnzNOjNLBnKcc9/GIDYREYmhJge3M7NTzexeM+sIfA58YGaXRz80ERGJpUiGv74OeAg4DXgLb9RT9UMQEWljIkkIPufcR8AoYKFzbnOEnxMRkVYkkgt70Mx+Cfw/YJGZnQgEoxuWiIjEWiQJ4bfAROA659wGYDJ1O6qJiEgb0OTw1865JcAoM9vbzPo650bEIC4REYmxSCbI6Qf8A8gF/GZWAJzknFsZ7eBERCR2Iqkyuhe4wzmX5ZzrhDd9poayEBFpYyJJCN2dc0/ULDjnHgMC0QtJRETiIZKEkGhmXWoWzKwrmiBHRKTNiWRO5XuBt83sr3iJ4Ew0/LWISJvT5BOCc+4h4EIgGUgDLnHOzY52YCIiEluNPiGEOqHtB/zHOXdNbEISEZF42OUTgpldi1dddBjwkpmNiVlUIiISc41VGY0BhjjnzgCOAiY19+BmNsbMPjWzVWZ2aSP7nWRmXzf3+CIi0nIaSwiVzrktAM45B2Q058BmlgfcDByJNz/zRDM7oIH9ugN3Ab7mHF9ERFpWc0YtrWzmsUcBi51zRc65UuBZYHQD+80BbmzmsUVEpIU19lI5wcyy2HHnXmfZOVfUxLFzgfVhy+uBYeE7mNllwPvA280JukZ2drMeWgDIXLcZgKwu6QQCHXfntK1WeysvqMzthcrcMhpLCAOBAupW5RSG/l8NJDRxbD91O7D5CBs228wG4E26cyzQM8J46ygsLCEYbF4fuc2btwFQXFRKWkL7qaUKBDqSn78l3mHElMrcPqjMkfP7fY3eSDc2p/KPnQRnDTAybDkHWBe2fDrQA1iK18ch18z+65wL/4yIiMRIJD2Vd9erwFQzCwCleE8DE2s2Ouf+APwBwMx6A/9WMhARiZ+oTYXpnFuLN5nOa8AKYK5z7l0zW2BmQ6N1XhER2T3RfELAOTcXmFtv3YkN7Lca6B3NWEREpHERJQQzSwX6Ah8Dqc65rVGNSkREYq7JKiMzGw58CcwH8oDvzOyIaAcmIiKxFck7hDvxOpkVOufWAGcDs6IalYiIxFwkCSHNOfdpzYJzbgFRfvcgIiKxF0lCqAj1UK4GMDOLbkgiIhIPkdzp3wT8B8gxs78AxxHWn0BERNqGJhOCc+4lM1sJ/A/ecBXTnHOfRT0yERGJqUhaGXUBioC/4vUp+D60TkRE2pBIqowKqDtIHXgjl+7WgHQiIrJniqTKqPYpwsyS8WZS04tlEZE2plljGTnnyp1zj+O9TxARkTakySeEeu8LfMBQICtqEYmISFw05x1CzWwyG4HLohaRiIjERSQJ4VDn3LKoRyIiInEVyTuEp6IehYiIxF0kTwgfmtkYYAlQUrPSOVcUtahERCTmIkkIp+DNfxyuGq/XsoiItBG7TAhm1sE5t905lxLLgEREJD4ae4fwVsyiEBGRuGssIfga2SYiIm1MY+8QUszsIHaRGJxz70cnJBERiYfGEkIf4DkaTgjVoe0iItJGNJYQPnXOHRSzSEREJK6aNbidiIi0XY0lhNdjFoWIiMTdLhOCc+7yWAYiIiLxpSojEREBlBBERCRECUFERAAlBBERCYlktNPdFho2+/dAEjDTOXd/ve2nADfidX77GhjvnCuOZkwiItKwqD0hmFkecDNwJDAEmGhmB4RtzwRmAyc55wYDHwJToxWPiIg0LppVRqOAxc65IudcKfAsMDpsexJwqXNubWj5Q6BXFOMREZFGRLPKKBdYH7a8HhhWs+CcKwT+DmBmqcC1wL1RjEdERBoRzYTgxxsEr4YPCNbfycw64SWGD5xzTzTnBNnZGc0OKnPdZgCyuqQTCHRs9udbs/ZWXlCZ2wuVuWVEMyGsAUaGLecA68J3MLMewCvAYuA3zT1BYWEJwWB10zuG2bx5GwDFRaWkJbSfKR8CgY7k52+JdxgxpTK3Dypz5Px+X6M30tFMCK8CU80sAJQCpwETazaaWQLwIjDPOXdTFOMQEZEIRC0hOOfWmtlk4DUgGZjjnHvXzBYAU4C9gIOBRDOredm81Dl3QbRiEhGRXYtqPwTn3Fxgbr11J4Z+XIo6xomI7DF0QRYREUAJQUREQpQQREQEUEIQEZEQJQQREQGUEEREJEQJQUREACUEEREJUUIQERFACUFEREKUEEREBFBCEBGRECUEEREBlBBERCRECUFERAAlBBERCVFCEBERQAlBRERClBBERARQQhARkZDEeAcgbVNVVSXFxflUVpbHLYaNG/0Eg8G4nT8eVOb2oaky+/0JpKZmkJHRCZ/PF/FxlRAkKoqL80lJSSM9PadZ/yBbUmKin8rK9nWhUJnbh8bKXF1dTVVVJVu2bKK4OJ8uXbpFfFxVGUlUVFaWk56eGbdkINJe+Xw+EhOT6Nw5m/Lybc36rBKCRI2SgUj8+Hx+oLpZn1FCEBERQAlB2oH169dx1FHDGTduDOPGjeHcc89k9OiTeeSRB1vk+AsWvMjNN09tkeOccMIxtXGOGzeGK6+c9OMD3IVPP/2YBx64p3a5oKCAadNuYOzYX3Luub/id7+7grVr19TG1hJlrDFu3BgASktLOP/8szn77F/yzDNPM2fOn37Uce+7byarVrna5SVLXufII4eycuVndfYbPfpk1q9fV2fdpEkTef/9pQBUVFTw0EMPcM45ZzBu3BgmThzHe++9s9txbdmyhauvvpyzzhrNpZdOoLCwYKd9ysrKmD79BsaOPZ3zzhvLkiWv125btOhlxo49nTPPPJXnnpsHwNatW7n++qupqqra7bjq00tlaRe6dg3w+ONza5cLCvI588xTOfbY4+jde584RlbXkUf+hMmTp8bkXKtXf01xcRHgXYwmTZrIr341lhtumIbP52PRooX85jeXMnfucy1+7prvYtWqz0lKSuKRR5780cdctepziooK6dfPatctWPAiRx89iuef/xv77Tc54mPdfPNUkpOTefjhP9OhQwe+/PILfvObS5k1azb77NOn2bE9/PADDBp0EHfeOYuXX57PrFl/ZNq0W+vs8+STj5GQkMif//xXtmzZzIUXnofZfrWff+SRJ0lKSuaii85j2LBh7LVXb4YOHcbzz/+N//3f05sdU0OUEKRdKigooLq6mrS0NCorK/njH2/jq6++pKioiL59+zJ16s0UFRVx/fVX0afPvnz+uaNLl2ymT7+NzMxOvPzyfJ544hHS0zPIyckhNTUNgI8//ohZs+6ivLyczp07c/XV19Oz515MmjQRs/348MMVlJeXc9FFv+aZZ55m9eqvOOOMMZxxxlmNxtvYcTMzO/H1118ybdqtbNpUxEMPzaayspIePfK45prJdOrUmfvum8l7772D3+9j5MijOP30XzFnzp8oKyvjiSceITs7m6ysLE455X9rz3nccSeQlJREeXndpsOLF7/K008/xfbt26moKOe666YwcOBgnn76KRYunI/f72P//Q/kd7+bzBdfrOKOO26mqqqK5ORkrr/+D+y1Vy+OPHIoL764iFtvnUZRUSHXXPMbfvrTY1i+fBmTJ0/ls88+4Z57ZrB9+zY6dfLKm5ubt1N5wy/+Tz/9FMcdd0Lt8qZNm3j//fd47LG5jB8/hkmTLic9PaPJfxtr1nzH66+/xvz5/6JDhw4A7Luv928iJSWlzr6ffPIxd955S511aWlpPPDAnDrr3nrrDe677yEARo06nhkz7qCyspLExB2X4FWrHKeeejp+v59OnTrTr19/3nnnLRISEjj44KFkZnYC4Oijj2Xx4lc599wLGDXqeCZOHMepp45ukXd2SggSE298tJ4lH66PyrGPHNSDEQN7NLpPQUE+48aNobx8Oz/8sIn99juQW265i27durNixfskJibx4IOPEQwGueyyi3jrrTcw258vvljFdddNoX///Zg8+WoWLVrIUUcdy+zZ9/DYY3PJzOzE7353BampaVRUVDB16vVMn34b++9/IIsXv8rUqZOZM+fPgNcc8OGH/8yjjz7EzJl38sQTT7NpUzHjxu1ICEuWvF5bnQJw2WVXMnDg4EaPu+++fbnlljspLi7mlltuZNasP5GZmck//vEcs2ffy7hxF/D222/y1FPz2LZtG7fcciPJyclccMFFLF++jHPPPZ8ZM26vvRsNd/TRo+osB4NBnn/+Oe64YyadO3fmpZee58knH+fWW+/iqace5x//eBm/389tt00nP38j8+bN5cwzx3LMMaNYuPAlPvnkI/baqxcAWVlduOaa3/Poow9x++13s2DBi4BXXXPbbTdx++13k5OTwzvvvMXtt9/MrFkP1ClvuOrqat58cwlXXXVd7bpFixZw6KHD6dEjF7MDWLToZU49dXST/55WrXLstVcvUlNT66w/+OChO+174IED6jx57kpBQT7Z2V0BSExMJD09nU2biunaNVC7T//++/Haa68ybNhwiouL+OijD+jXrz8+n6/2swDZ2V1ZufITADIzM0lLS+WLL1bRr1//JuNoihKCtAs1VUbBYJD77rub1au/5tBDDwNgyJCDyczsxHPPzePbb1ezZs13lJWVAd5Fq39/70LZp09fNm/ezEcffcCAAYPo0iUb8O6kly17j++++4aOHTuy//4HAnDMMaO4446bKSkpAWD48BEA5OT04MADB5KSkkJOTg9KSrbUxtlQldFXX33R6HEPOGAA4L0T+P77DVx22UUABINVZGZ2omvXAB06dODii8/jiCNGcvHFv669863h9/tJTk5u8vfo9/u55ZY7eeON//Ltt9+wfPky/H4/CQkJDBgwiAsuOIeRI3/KmWeeRSDQjcMPH8GMGXfwzjtvMmLETxgxYmST5/juu29Yt24N1157Ze260tLS2p9ryhtu06ZNAHUu4gsWvMT48RMAOPbY/+G55+bVJgSvBU7D5Yv0dwGRPyFUV1fvtFz/jv7ss8dzzz1/5LzzzqJXr94MGzacxMQkKirK6+xbXV2N378j/u7de7BmzbdKCNJ6jBjY9F18LPj9fi655HLGjx/DX/7yJGeddS5LlvyHOXMe5PTTz+TEE3/Opk2bav+A618Yav6Qw/++ExISAAgGG2riV00w6L30C68eqPlMJJo6bs3FPRisYtCgIdx22wwAtm/fTllZGYmJiTz00OOsWPE+b731BhddNJ57732oztHM9mfhwpd2Osttt03nl7/c8cSydetWJkw4l+OOO4HBgw9i33371r7kvPXWP/LJJx/x9ttv8tvfXsaUKdM5+uhRDBgwiDfe+C/z5s3lrbeWcM01v2+0vFVVQXJz82rvvKuqqmrfdYSXN5zX9n7H79e5lXz11RfMmnUX9947g2AwSEFBPv1O8m8AAA6ISURBVB9//BEDBgykY8eOdRIxQHFxER07ZpKTk8vq1avZvn0bHTrsqCKaN28uXbpkM2rU8bXrIn1CCAS6UVRUSLdu3amsrGTr1q106tS5zj4lJSVceOGlteuvvvpyRoz4CVu3lvLBB8tr9ysqKqzzZJGQkFAnQfwYUW1lZGZjzOxTM1tlZpc2sH2ImS01s8/NbI6ZKUFJ1CUmJnLppVfw+OOPUFhYwNKl73LMMaM46aSfk5GRwfLly2ovtg0ZNGgIn3zyIfn5GwkGgyxe/E8AevXamx9++IHPPvMe5//1r3/SvXuP2rrf3RXpcQ84YAAff/wh3377DQCPPz6H+++fyeefr2TSpIkMHnwQkyZdQe/effj2229ISEiobaFyzDGjWL9+PS+99I/a482f/wLLly+jZ8+9atd99923+Hw+zjnnPA4+eCj/+c9rBINBiouLGTv2dPr06csFF1zEoYcexpdfrmLKlOv47LNP+cUvTuOCCy7CuZVNlnfvvXuzefPm2ovg/PkvMHVq4y+EO3fuTDBYxdatWwFYsOAFfv7zU/nb3+bz7LMv8re/zef440/k+ee9F+RDhx7K/Pkv1Cb+5cuXUVZWxt579yYnJ4cjjhjB3Xffyfbt2wH4/POV/N//PUGfPvs2GX9Dhg8fwcsvzwdg8eJ/MnjwkDoJDODf/36VOXO8lm9ffLGKzz9fyaGHDmPo0GEsW/YexcXFbNu2jX//ezHDhx9R+7kNG9aTl7cXLSFqF2AzywNuBg4BtgNvmtlrzrlPw3Z7CrjAOfe2mT0CTABmRysmkRrDhx/BgAEDmTPnT4wefSY33jiZV199hcTEJAYOHMS6des45JCGP9ulSzZXXHE1V1xxCSkpqbWtlJKTk5k27VZmzLiDbdvKyMzstFNLkt0R6XGzs7syefIfmDLlOoLBKgKB7kyZMo1OnTozYMAgzjnnDFJSUhg4cDDDhx/BunVrefTRh5g9+14uvvjXzJx5P/feO4Onn56Lzwe5uXnMmHFfnaekvn370bdvf8aMGY3f72PYsMP58MMVZGVl8fOfn8qECefQoUMKvXrtzUknncLgwQdz++038fjjD5OYmMRVV10bUXmnT7+t9iV6Wlo6v//9jU1+bvjwI/jgg/cZOvQwXn31Fe65p26z4jPOOIsLLxzHr399JePGXcDMmXdx9tln4PNBZmYnbr31j7UX6euu+wOzZ9/D+PFjSEpKJiUlhRtumE6fPn2bjKMhEyZcxM03T2Xs2F/SsWMGU6bcBMCSJf9hyZLXufbaGzj55FOZPv0GzjnnDBITE5k27TbS0tJJS0tnwoRLuOyyC6moqOTkk0/hwAMHUFkZZMuWLZSUlNC3b7/diqs+X/26rZZiZucCP3HOnR9avgHwOeemhZb3BhY75/YNLY8EbnTOHRPB4XsDXxcWluzicXrX3lu5kdn/+Jjp5w8jL9B0i4O2IhDoSH7+lqZ3bCEbNnxDTs7eMTtfQzTGTftQU+ZVqz7niSce4aabbo93SFFXU+Z58/5CQkICp532ywb3q/936Pf7yM7OANgHWL3TcaMTLgC5QHizkvXAsCa292zOCUIFa5a9t5STmOCjV88sumSmNP2BNiQQ6Bizc23c6CcxMf79HveEGGKtvZZ5//33IyenO1984dhvv/3jHVLUlZdvY9myd7n99j/u8jv3+/3N+ruPZkKoP5CGDwg2Y3uTducJoVvHZP5v2gmUbtlGfn5Fsz7bmsX6CSEYDMb9TrU93y23J+FlnjTJa5nU1n8HiYl+kpNTuP32u4FdlzcYDNb5uw97QmhQNG8l1gDhzUpygHXN2B41aSlJsTiNiEirEs2E8CpwrJkFzCwNOA14uWajc+4bYJuZjQitOhtYGMV4JMai9X5KRJpWXR3Eq3iJXNQSgnNuLTAZeA1YAcx1zr1rZgvMrKbL31nA3Wa2EsgA7mn4aNLaJCYmU1q6WUlBJMaqq6uprKxg06YCkpOb9540aq2Moqw3u9nKCGJfn74niHWZ94QpNP3+9je1osrcPjRV5l1NoRnPVkbSjiUkJNK1a3x7Jivxtw8qc8tpf+3TRESkQUoIIiICtN4qowTw6sN214/5bGulMrcPKnP7sDtlDvtMg6MrttaXykcC/413ECIirdRIYEn9la01IXQADsUb7qLlJhQVEWnbEvA6BL+HN+hoHa01IYiISAvTS2UREQGUEEREJEQJQUREACUEEREJUUIQERFACUFEREKUEEREBGi9Q1dExMzGAL8HkoCZzrn7620fAswBMoHXgYucc5UxD7QFRVDmU4Ab8WbO+BoY75wrjnmgLaipMoftdxJwn3Nun1jGFw0RfM8GPAhkARuAM9v692xmB+OVORn4DhjrnNsU80BbkJllAm8CP3POra63rcWvX232CcHM8oCb8Ya5GAJMNLMD6u32FDDJOdcf7wI5IbZRtqymyhz6xzUbOMk5Nxj4EJgah1BbTITfM2bWHbiL5k4htQeK4Hv2AS8At4W+5+XAtfGItaVE+D3PAqaEyuyAq2IbZcsys8Pwhpfov4tdWvz61WYTAjAKWOycK3LOlQLPAqNrNprZ3kCqc+7t0KrHgdNjHmXLarTMeHdWl4ZmswMvIfSKcYwtraky15iD92TUFjRV5oOBUudczZS1twANPjW1IpF8zwl4d8sAaUBZDOOLhgnApTQw13y0rl9tucooF2+soxrrgWFNbO8Zg7iiqdEyO+cKgb8DmFkq3l3jvbEMMAqa+p4xs8uA94G3aRuaKnNfYIOZPQIcBHwG/Dp24UVFk98zcCWwyMxmAqXAYTGKLSqccxcAeLV/O4nK9astPyH4gfCBmnxAsBnbW6OIymRmnYD5wAfOuSdiFFu0NFpmMxsAnAZMj3Fc0dTU95wIHAXMds4dDHwFzIhZdNHR1PecCjwCjHLO9QAeAP4c0whjKyrXr7acENbgjepXI4e6j15NbW+NmiyTmfXAGzr8Q+CC2IUWNU2V+fTQ9qXAAiDXzFr70OlNlXkDsMo5tzS0/Bd2vptubZoq8wCgzDn3bmj5Qbyk2FZF5frVlhPCq8CxZhYwszS8u8SaOlWcc98A28xsRGjV2cDC2IfZohots5klAC8C85xzVzjn2sJQt019z39wzvV3zg0BTgTWOedGxinWltJomfFapQTMbHBo+WRgWYxjbGlNlfkLYC/bUb9yCt4Qz21StK5fbTYhhF6cTgZeA1YAc51z75rZAjMbGtrtLOBuM1sJZAD3xCfalhFBmX+O98JxtJmtCP03J44h/2gRfs9tSlNlds6VAacCD5vZJ8AxwG/jF/GPF0GZi4FxwDwz+xA4Dxgft4CjJNrXL82HICIiQBt+QhARkeZRQhAREUAJQUREQpQQREQEUEIQEZGQtjx0hbQyZlYNfAxUha1eWtOFfxefGQeMds79rAXOPxVv7Ji1eL1AE4CNwCXOuc9343i5wLPOuSPMbB/gLufcaeHrWyDm3sCXwEdhqzPwOi6d55z7qonPT8Hrsf78j41FWj8lBNnTHO2cK4jj+f/qnJtUs2BmvwbmAs3u0+CcWwfUXPT3BqyB9S2hLNTxDqgd7fQevNFBf9XEZ48BPm3BWKQVU0KQVsHMzgMuxBvrvgve0M6z6+3zv3jj5QfxnjKuds69Hhq7aRYwEG/E13+FtkUydvy/gFtDx++JN3x4b7yxY55wzt1pZol4gwSOACrwxg4aD3TFe+LphDfaap6ZvRIqR8361cAvnHPLQuf4K/Bv59xsM5uM1yPXH9rvklAyaUoK3uBnG0LH7I832mlHvOEOVgBnAOfjJbo7zawKb3yr24Gf4j0dLQcuc85tjuCc0gboHYLsaV4L60W9wsy6mVkG3lDAJzrnDsK7mN3RwGfvxLtoDgVuYMdYNncDy5xzh+CN/tkVb2TMRoUu9Ofj9Y4F+D/gNefcQLyL/1gzOxM4PHSuwaFzfAUMqjmOc64Kb9yoL51zx9db/yihHrVmloU3zPNcMzsHL4ENC939L8BLKg1JDf2uPjKz7/FGdl0JXBPaPgEveQ3HGwl1H7w5Me7HG+Ppaufc3/FGv60EDgnNKbAOuK2p35O0HXpCkD1Ng1VGZvYz4CQz64c3QUpGA599Gvi7mc0H/smOpPEzYJiZnR9aTm3k/GeY2ZGhn5PxxgCaYGbpeEngOADn3A9m9jhwAnA53hPJO6EngOdCwyr0jqC8jwLvmdmVeNU7L4SO/TO8AemWhobnScAb478htVVGZnY83sQpLzrnSkLbrwH+x8x+hzfZSi4N//5+BnQO7VtT/o0RlEHaCD0hyB4vVFWzAq8efgletdBOnHOT8WbUWoo3rs3roU0JwOnOuSGhC+dhwKSGjoH3DmFI6L8DnHNnO+c24P2t1J9tzQ8khaZpHIw3Q1cV8FczuySSsoUGKXsf72I8nh1PAQnA7WExD8VLSE0d7xW8oa6fCc2QB95opxOBb/Celt5voCw157w87JzDaHiyIWmjlBCkNRgK5AM3AYvwLp41o7cS+jnRzFYDac65PwGXAIPMrAPwCvAbM/OFll9g1wmhQc65LXgT7FwaOl8n4Bzgn6G7+X8BbzrnpuKNw39ovUNU4r2/aMjDeHfx6c65N0LrXgEuCLuoTwOejDDcu4At7Jgh7nhgmnPur6Hlw/Au/vXjegWYZGbJZuYPxXVrhOeUNkAJQVqDRXjNKB3e7F+98BJE35odQi+Ir8Crf38feAav2eV24DIgHa9p5oeh/zf0DqIpZ+ENwfwR8C7wN7ypCxcCnwAfm9lSvBZE9afr/BRvuOJ32fnu/AW8F9Xh7wjmAC8Bb4dGLB2E99TTJOdcBV7CmxSaIOh6vKq0j/DmCfgPO353LwC3mtm5eJMIrcZ7mfxpKM5WPUqqNI9GOxUREUBPCCIiEqKEICIigBKCiIiEKCGIiAighCAiIiFKCCIiAighiIhIiBKCiIgA8P8BA90HmdidGF0AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# evaluation\n",
    "acc = accuracy_score(test_labels, rf_predictions)\n",
    "print(acc)\n",
    "# Probabilities for each class\n",
    "rf_probs = model.predict_proba(test_features)[:, 1]\n",
    "# Calculate roc auc\n",
    "print(confusion_matrix(test_labels, rf_predictions))\n",
    "roc_value = roc_auc_score(test_labels, rf_probs)\n",
    "print(roc_value)\n",
    "plot_roc_curve(model, test_features, test_labels) \n",
    "# Extract feature importances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Colon\n",
      "GDSD0\n",
      "Airway\n",
      "GDSD6\n",
      "Pancreas\n",
      "Uterine\n",
      "Ovarian\n",
      "Bladder\n",
      "Melanocytes\n",
      "HMEC\n",
      "Prostate\n",
      "GM12878\n",
      "Renal\n",
      "GDSD3\n",
      "Esophageal\n",
      "Astrocyte\n",
      "Thyroid\n"
     ]
    }
   ],
   "source": [
    "for file in glob.glob('rnaseq/unique_gene_lists/*txt'):\n",
    "    print(os.path.basename(file).split('_')[-2])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Log\n",
    "\n",
    "when looking at 5% of highly expressed gene crm data, get about auc .8, , if use all data get auc .87, this is predicting GDSD3\n",
    "0.9465875370919882\n",
    "[[1267    0]\n",
    " [  72    9]]\n",
    "0.8061475050425327\n",
    "\n",
    "(5% of data)\n",
    "\n",
    "pretty good\n",
    "\n",
    "\n",
    "for predicting GDSD6 with GDSD6 specific genes, get auc of .85 (num genes is 256)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## A.2: Get all Random Forest and Decision Tree Data\n",
    "Extract in a single dictionary the random forest data and for all of it's decision trees\n",
    "This is as required for RIT purposes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 35.7 s, sys: 2.54 s, total: 38.3 s\n",
      "Wall time: 38.5 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "all_rf_tree_data = utils.get_rf_tree_data(\n",
    "    rf=model, X_train=train_features, X_test=test_features, y_test=test_labels)\n",
    "\n",
    "# all_rf_tree_data = get_rf_tree_data(\n",
    "#     rf=model, X_train=train_features, X_test=test_features, y_test=test_labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## A.3: Get the RIT data and produce RITs\n",
    "\n",
    "RIT = random interaction tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(12)\n",
    "all_rit_tree_data = irf_utils.get_rit_tree_data(\n",
    "    all_rf_tree_data=all_rf_tree_data,\n",
    "    bin_class_type=1,\n",
    "    M=100,\n",
    "    max_depth=2,\n",
    "    noisy_split=False,\n",
    "    num_splits=2)\n",
    "# Perform Manual CHECKS on the irf_utils\n",
    "# These should be converted to unit tests and checked with nosetests -v test_irf_utils.py"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## A.4: Plot and Evaluate Data\n",
    "### List Ranked Feature Importances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rank</th>\n",
       "      <th>feature</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>SPIB_loop</td>\n",
       "      <td>0.073172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>EGR2_loop</td>\n",
       "      <td>0.061718</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>IRF4_loop</td>\n",
       "      <td>0.044318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>TBX21_loop</td>\n",
       "      <td>0.032977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>SP4_loop</td>\n",
       "      <td>0.029375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>643</td>\n",
       "      <td>644</td>\n",
       "      <td>HIC1_anc</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>644</td>\n",
       "      <td>645</td>\n",
       "      <td>SOX5_anc</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>645</td>\n",
       "      <td>646</td>\n",
       "      <td>GRHL2_anc</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>646</td>\n",
       "      <td>647</td>\n",
       "      <td>GFI1_anc</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>647</td>\n",
       "      <td>648</td>\n",
       "      <td>ZSCAN31_loop</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>648 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     rank       feature     score\n",
       "0       1     SPIB_loop  0.073172\n",
       "1       2     EGR2_loop  0.061718\n",
       "2       3     IRF4_loop  0.044318\n",
       "3       4    TBX21_loop  0.032977\n",
       "4       5      SP4_loop  0.029375\n",
       "..    ...           ...       ...\n",
       "643   644      HIC1_anc  0.000000\n",
       "644   645      SOX5_anc  0.000000\n",
       "645   646     GRHL2_anc  0.000000\n",
       "646   647      GFI1_anc  0.000000\n",
       "647   648  ZSCAN31_loop  0.000000\n",
       "\n",
       "[648 rows x 3 columns]"
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Print the feature ranking (from irf)\n",
    "# print(\"Feature ranking:\")\n",
    "\n",
    "feature_importances_rank_idx = all_rf_tree_data['feature_importances_rank_idx']\n",
    "feature_importances = all_rf_tree_data['feature_importances']\n",
    "\n",
    "pd.DataFrame({'rank': range(1,train_features.shape[1]+1),\n",
    "             'feature': data_all_varfilt_cols_1[feature_importances_rank_idx],\n",
    "             'score':feature_importances[feature_importances_rank_idx]})\n",
    "# for f in range(train_features.shape[1]):\n",
    "#     print(\"%d. feature %s (%f)\" % (f + 1\n",
    "#                                    , data_all_varfilt_cols_1[feature_importances_rank_idx][f]\n",
    "#                                    , feature_importances[feature_importances_rank_idx[f]]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature</th>\n",
       "      <th>importance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>443</td>\n",
       "      <td>SPIB_loop</td>\n",
       "      <td>0.073172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>77</td>\n",
       "      <td>EGR2_loop</td>\n",
       "      <td>0.061718</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>197</td>\n",
       "      <td>IRF4_loop</td>\n",
       "      <td>0.044318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>467</td>\n",
       "      <td>TBX21_loop</td>\n",
       "      <td>0.032977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>441</td>\n",
       "      <td>SP4_loop</td>\n",
       "      <td>0.029375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>73</td>\n",
       "      <td>EBF1_loop</td>\n",
       "      <td>0.023383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>257</td>\n",
       "      <td>MEF2C_loop</td>\n",
       "      <td>0.023310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>513</td>\n",
       "      <td>TP73_loop</td>\n",
       "      <td>0.021849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>189</td>\n",
       "      <td>IKZF1_loop</td>\n",
       "      <td>0.016745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>229</td>\n",
       "      <td>LEF1_loop</td>\n",
       "      <td>0.014944</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        feature  importance\n",
       "443   SPIB_loop    0.073172\n",
       "77    EGR2_loop    0.061718\n",
       "197   IRF4_loop    0.044318\n",
       "467  TBX21_loop    0.032977\n",
       "441    SP4_loop    0.029375\n",
       "73    EBF1_loop    0.023383\n",
       "257  MEF2C_loop    0.023310\n",
       "513   TP73_loop    0.021849\n",
       "189  IKZF1_loop    0.016745\n",
       "229   LEF1_loop    0.014944"
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# compare to what sklearn directly outputs\n",
    "fi = pd.DataFrame({'feature': list(data_all_varfilt_cols_1),\n",
    "                   'importance': model.feature_importances_}).\\\n",
    "                    sort_values('importance', ascending = False)\n",
    "\n",
    "# Display\n",
    "fi.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot Ranked Feature Importances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Plot the feature importances of the forest\n",
    "# feature_importances_std = all_rf_tree_data['feature_importances_std']\n",
    "\n",
    "# plt.title(\"Feature importances\")\n",
    "# plt.bar(range(train_features.shape[1])\n",
    "#         , feature_importances[feature_importances_rank_idx]\n",
    "#         , color=\"r\"\n",
    "#         , yerr = feature_importances_std[feature_importances_rank_idx], align=\"center\")\n",
    "# plt.xticks(range(train_features.shape[1]), feature_importances_rank_idx)\n",
    "# plt.xlim([-1, train_features.shape[1]])\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## A.5: Run the iRF function!\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will run the iRF with the following parameters\n",
    "Data:\n",
    "- CRM data\n",
    "- random state (for reproducibility): 2018\n",
    "Weighted RFs\n",
    "- K: 5 iterations\n",
    "- number of trees: 20\n",
    "Bootstrap RFs\n",
    "- proportion of bootstrap samples: 20%\n",
    "- B: 30 bootstrap samples\n",
    "- number of trees (bootstrap RFs): 5 iterations\n",
    "RITs (on the bootstrap RFs)\n",
    "- M: 20 RITs per forest\n",
    "- filter label type: 1-class only\n",
    "- Max Depth: 5\n",
    "- Noisy Split: False\n",
    "- Number of splits at Node: 2 splits\n",
    "\n",
    "\n",
    "#### Running the iRF is easy - single function call\n",
    "All of the bootstrap, RIT complexity is covered through the key parameters passed through in the main algorithm (as listed above)\n",
    "This function call returns the following data:\n",
    "m score\n",
    "\n",
    "This is a lot of data returned! Will be useful when they build the interface later"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 5min, sys: 16.7 s, total: 5min 17s\n",
      "Wall time: 5min 17s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "all_rf_weights, all_K_iter_rf_data, \\\n",
    "all_rf_bootstrap_output, all_rit_bootstrap_output, \\\n",
    "stability_score = irf_utils.run_iRF(X_train=train_features,\n",
    "                                    X_test=test_features,\n",
    "                                    y_train=train_labels,\n",
    "                                    y_test=test_labels,\n",
    "                                    K=10,\n",
    "                                    rf=RandomForestClassifierWithWeights(n_estimators=100),\n",
    "                                    B=30,\n",
    "                                    random_state_classifier=2018,\n",
    "                                    propn_n_samples=.2,\n",
    "                                    bin_class_type=1,\n",
    "                                    M=25,\n",
    "                                    max_depth=5,\n",
    "                                    noisy_split=False,\n",
    "                                    num_splits=2,\n",
    "                                    n_estimators_bootstrap=5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "THIS IS WHAT WE CARE ABOUT `stability_score`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['AHR_anc', 'AHR_loop', 'ALX1_anc', 'ALX1_loop', 'ARID5B_anc',\n",
       "       'ARID5B_loop', 'ARNTL_anc', 'ARNTL_loop', 'ARNT_anc', 'ARNT_loop',\n",
       "       ...\n",
       "       'ZNF76_anc', 'ZNF76_loop', 'ZNF770_anc', 'ZNF770_loop', 'ZNF816_anc',\n",
       "       'ZNF816_loop', 'ZNF85_anc', 'ZNF85_loop', 'ZSCAN31_anc',\n",
       "       'ZSCAN31_loop'],\n",
       "      dtype='object', length=648)"
      ]
     },
     "execution_count": 138,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_all_varfilt_cols_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "networks/representations_irf_cell_type/test_GM12878_GM12878_stability_score.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:8: FutureWarning: The signature of `Series.to_csv` was aligned to that of `DataFrame.to_csv`, and argument 'header' will change its default value from False to True: please pass an explicit value to suppress this warning.\n",
      "  \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "SPIB_loop                                              0.866667\n",
       "EGR2_loop                                              0.800000\n",
       "EGR2_loop::SPIB_loop                                   0.666667\n",
       "EGR2_loop::PATZ1_loop                                  0.566667\n",
       "EGR2_loop::PATZ1_loop::SPIB_loop                       0.400000\n",
       "                                                         ...   \n",
       "PATZ1_loop::RUNX3_loop::SPIB_loop                      0.033333\n",
       "IRF4_loop::KLF5_loop::PATZ1_loop::SP4_anc::SP4_loop    0.033333\n",
       "FLI1_loop::IRF4_loop::LYL1_loop::SP4_loop              0.033333\n",
       "IRF4_loop::PATZ1_loop::SP4_anc::SP4_loop               0.033333\n",
       "EGR2_loop::IRF7_loop::OSR2_anc::SPIB_loop              0.033333\n",
       "Length: 259, dtype: float64"
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stability_score_names = {}\n",
    "for feat_idx, score in sorted(stability_score.items(),key=lambda x:x[1],reverse=True):\n",
    "    feat_names_arr = [data_all_varfilt_cols_1[int(x)] for x in feat_idx.split('_')]\n",
    "    feat_names = '::'.join(feat_names_arr)\n",
    "    stability_score_names[feat_names] = score\n",
    "stability_df = pd.Series(stability_score_names)\n",
    "save_file = os.path.join(save_dir, 'test_'+rf_tissue+'_' + gene_tissue + '_stability_score.csv')\n",
    "stability_df.to_csv(save_file)\n",
    "print(save_file)\n",
    "stability_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [],
   "source": [
    "# freq_feat = pd.DataFrame(train_features, columns=data_all_varfilt_cols_1).sum(axis=0).sort_values(ascending=False)\n",
    "# pd.DataFrame(freq_feat)[:20]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>SPIB_loop</td>\n",
       "      <td>0.866667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>EGR2_loop</td>\n",
       "      <td>0.800000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>EGR2_loop::SPIB_loop</td>\n",
       "      <td>0.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>EGR2_loop::PATZ1_loop</td>\n",
       "      <td>0.566667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>EGR2_loop::PATZ1_loop::SPIB_loop</td>\n",
       "      <td>0.400000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>IRF4_loop</td>\n",
       "      <td>0.300000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>EGR2_loop::IRF4_loop::SPIB_loop</td>\n",
       "      <td>0.300000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>EGR2_loop::ZNF467_loop</td>\n",
       "      <td>0.300000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>IRF4_loop::SPIB_loop</td>\n",
       "      <td>0.300000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>EGR2_loop::SPIB_loop::TP73_loop</td>\n",
       "      <td>0.200000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>EGR2_loop::SP4_loop</td>\n",
       "      <td>0.200000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>EGR2_loop::SP4_loop::SPIB_loop</td>\n",
       "      <td>0.166667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>EGR2_loop::SPIB_loop::TBX21_loop</td>\n",
       "      <td>0.166667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SPIB_loop::TBX21_loop</td>\n",
       "      <td>0.166667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>EGR2_loop::IRF4_loop</td>\n",
       "      <td>0.166667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>EGR2_loop::SPIB_loop::TBX21_anc</td>\n",
       "      <td>0.133333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>PATZ1_loop</td>\n",
       "      <td>0.133333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>EGR2_loop::IRF4_loop::PATZ1_loop</td>\n",
       "      <td>0.133333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>MEF2C_loop::SPIB_loop</td>\n",
       "      <td>0.133333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>EGR2_loop::LYL1_loop::SPIB_loop</td>\n",
       "      <td>0.100000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         0\n",
       "SPIB_loop                         0.866667\n",
       "EGR2_loop                         0.800000\n",
       "EGR2_loop::SPIB_loop              0.666667\n",
       "EGR2_loop::PATZ1_loop             0.566667\n",
       "EGR2_loop::PATZ1_loop::SPIB_loop  0.400000\n",
       "IRF4_loop                         0.300000\n",
       "EGR2_loop::IRF4_loop::SPIB_loop   0.300000\n",
       "EGR2_loop::ZNF467_loop            0.300000\n",
       "IRF4_loop::SPIB_loop              0.300000\n",
       "EGR2_loop::SPIB_loop::TP73_loop   0.200000\n",
       "EGR2_loop::SP4_loop               0.200000\n",
       "EGR2_loop::SP4_loop::SPIB_loop    0.166667\n",
       "EGR2_loop::SPIB_loop::TBX21_loop  0.166667\n",
       "SPIB_loop::TBX21_loop             0.166667\n",
       "EGR2_loop::IRF4_loop              0.166667\n",
       "EGR2_loop::SPIB_loop::TBX21_anc   0.133333\n",
       "PATZ1_loop                        0.133333\n",
       "EGR2_loop::IRF4_loop::PATZ1_loop  0.133333\n",
       "MEF2C_loop::SPIB_loop             0.133333\n",
       "EGR2_loop::LYL1_loop::SPIB_loop   0.100000"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(stability_df)[:20]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Examine the stability scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "irf_jupyter_utils._get_histogram(stability_score, sort = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Examine feature importances\n",
    "In particular, let us see how they change over the K iterations of random forest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for k in range(1, 6): \n",
    "    \n",
    "#     iteration = \"rf_iter{}\".format(k)\n",
    "    \n",
    "#     feature_importances_std = all_K_iter_rf_data[iteration]['feature_importances_std']\n",
    "#     feature_importances_rank_idx = all_K_iter_rf_data[iteration]['feature_importances_rank_idx']\n",
    "#     feature_importances = all_K_iter_rf_data[iteration]['feature_importances']\n",
    "\n",
    "#     plt.figure(figsize=(8, 6))\n",
    "#     title = \"Feature importances; iteration = {}\".format(k)\n",
    "#     plt.title(title)\n",
    "#     plt.bar(range(train_features.shape[1])\n",
    "#             , feature_importances[feature_importances_rank_idx]\n",
    "#             , color=\"r\"\n",
    "#             , yerr = feature_importances_std[feature_importances_rank_idx], align=\"center\")\n",
    "#     plt.xticks(range(train_features.shape[1]), feature_importances_rank_idx, rotation='vertical')\n",
    "#     plt.xlim([-1, train_features.shape[1]])\n",
    "#     plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## A.6 Explore iRF Data Further\n",
    "We can look at the decision paths of the Kth RF\n",
    "Let's look at the final iteration RF - the key validation metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{   'accuracy_score': 0.7256,\n",
      "    'confusion_matrix': array([[ 777,  370],\n",
      "       [ 316, 1037]]),\n",
      "    'f1_score': 0.7514492753623188,\n",
      "    'hamming_loss': 0.2744,\n",
      "    'log_loss': 9.477558583183162,\n",
      "    'precision_score': 0.7370291400142146,\n",
      "    'recall_score': 0.7664449371766445,\n",
      "    'zero_one_loss': 0.2744}\n"
     ]
    }
   ],
   "source": [
    "irf_jupyter_utils.pretty_print_dict(all_K_iter_rf_data['rf_iter5']['rf_validation_metrics'])\n",
    "# {   'accuracy_score': 0.9649122807017544,\n",
    "#     'confusion_matrix': array([[12,  2],\n",
    "#        [ 0, 43]]),\n",
    "#     'f1_score': 0.9772727272727273,\n",
    "#     'hamming_loss': 0.03508771929824561,\n",
    "#     'log_loss': 1.2119149470996806,\n",
    "#     'precision_score': 0.9555555555555556,\n",
    "#     'recall_score': 1.0,\n",
    "#     'zero_one_loss': 0.03508771929824561}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now plot the trees individually"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "#irf_jupyter_utils.draw_tree(decision_tree = all_K_iter_rf_data['rf_iter5']['rf_obj'].estimators_[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can get this data quite easily in a convenient format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# irf_jupyter_utils.pretty_print_dict(\n",
    "#     all_K_iter_rf_data['rf_iter5']['dtree0']['all_leaf_paths_features'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notes comparing above combination against decision paths"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In fact - we can go further and plot some interesting data from the Decision Trees. This can help us understand variable interactions better"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "irf_jupyter_utils.pretty_print_dict(\n",
    "    all_K_iter_rf_data['rf_iter5']['dtree0']['all_leaf_node_values'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can also look at the frequency that a feature appears along a decision path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "irf_jupyter_utils._hist_features(all_K_iter_rf_data['rf_iter5'], n_estimators = 20, \\\n",
    "                         title = 'Frequency of features along decision paths : iteration = 5')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run some Sanity Checks\n",
    "Run iRF for just 1 iteration - should be the uniform sampling version\n",
    "This is just a sanity check: the feature importances from iRF after 1 iteration should match the feature importance from running a standard RF\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_K_iter_rf_data.keys()\n",
    "print(all_K_iter_rf_data['rf_iter1']['feature_importances'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Compare to the original single fitted random forest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# rf = RandomForestClassifierWithWeights(n_estimators=20, random_state=2018)\n",
    "# rf.fit(X=X_train, y=y_train)\n",
    "print(model.feature_importances_)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notes: And they match ____(well?) as expected.\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
